---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r setup, include=FALSE,results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(scater)
library(SingleCellExperiment)
library(purrr)
library(stringr)
library(mvoutlier)
library(dropbead)
library(scater)
library(stringr)
library(scran)
library(cowplot)
library(rtracklayer)
library(org.Mmu.eg.db)
library(graphics)
library(RCy3)

source("../utils/02_sc_utils.R")

ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))
result_path <- file.path("/home/luisas/Desktop/cluster/data/RObjects")
data_dir <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/04_DigitalExpressionMatrix_OK"
```

## Single Cell Preprocessing Ex Vivo 

1. Merge all quantification files into one Seurat Object 
2. Cell QC - remove cells with too little or high UMIs counts
3. Gene QC - remove genes not showing up in at least 10 cells
4. Mitochondrial QC 
5. Doublet detection - scrublet 



```{r cars}
cell_mask_scrublet <- read.table(file.path(result_path, "EXVIVO_predicted_doublet_mask.mtx"))
doublet_scores_scrublet <- read.table(file.path(result_path, "EXVIVO_predicted_doublet_scores.txt"))

list <- as.list(as.character(cell_mask_scrublet$V1))
mask <- list == "True"
new <- pbmc[,!mask]

saveRDS(new, file.path(result_path,"EXVIVO_rhemac10_aftercellandgeneQC_afterScrublet.rds"))
```

##-------------------------------------------------------------------------------
##                         Stimulated data analysis 
##-------------------------------------------------------------------------------
```{r cars}
# -----------------------------------
#         LncRNAs
# -----------------------------------
# Extract known lncrnas 
immune.combined <- as.Seurat(pbmc)
mrna_ref_id <- unique(ref$gene_name[ref$gene_biotype == "protein_coding"])
lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype == "lncRNA"])
ribodepl_genes <- rownames(immune.combined)[grepl("ribodepl*", rownames(immune.combined))]
polya_genes <- rownames(immune.combined)[grepl("polya*", rownames(immune.combined))]
novel_lncrnas <- c(ribodepl_genes, polya_genes)
annotated_lncrna <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(lnc_ref_id, "_" , "-")]
all_lncrnas <- c(annotated_lncrna, novel_lncrnas)
```


# No Integration 


```{r cars}
#immune.combined <- readRDS(file.path(result_path,"EXVIVO_rhemac10_aftercellandgeneQC_afterScrublet.rds"))
# Normalize 
immune.combined <- NormalizeData(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindClusters(immune.combined, resolution = 0.5)
DefaultAssay(immune.combined) <- "RNA"
immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "-")[[1]][1]))))

saveRDS(immune.combined, file.path(result_path, "EXVIVO_preprocessed_ready.rds"))

# ----------------------------
# FIND MARKERS  for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(immune.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
saveRDS(pbmc.markers, file.path(result_path,"EXVIVO_markers_rhemac10_2k.rds"))

```


