---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
library(dplyr)
library(Seurat)
library(Matrix)
library(scater)
library(SingleCellExperiment)
library(purrr)
library(stringr)
library(mvoutlier)
library(dropbead)
library(scater)
library(stringr)
library(scran)
library(cowplot)
library(rtracklayer)
library(org.Mmu.eg.db)
library(graphics)
library(RCy3)

source("../utils/02_sc_utils.R")

theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/01_QC/immune.combined_qc.rds")

# Load annotations
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))


# Doublet detection results 
cell_mask_scrublet <- read.table("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/02_DoubletDetection/immune.combined_scrublet_mask.txt")


result_path<- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep"
```


# ---------------------------------------------------------
#    1. REMOVE DOUBLETS: Use Scrublet mask to remove doublets 
# ---------------------------------------------------------

```{r cars}
list <- as.list(as.character(cell_mask_scrublet$V1))
mask <- list == "True"
immune.combined <- immune.combined[,!mask]
saveRDS(immune.combined, file.path(result_path,"immune.combined_post_scrublet.rds"))
```

# ---------------------------------------------------------
#    Remove MDCKs
# ---------------------------------------------------------

```{r cars}

# Refine Labeling 
immune.combined$orig.ident <- gsub("D_", "D-", gsub("-", "_",immune.combined$orig.ident))
immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "_")[[1]][1]))))
immune.combined$dpi <- immune.combined$cond <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][2])))
immune.combined$batch <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][3])))
immune.combined$batch <- gsub("FRZ", "frozen", immune.combined$batch)
immune.combined$analysis <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][4])))
immune.combined$sample <- paste0(immune.combined$individual," ", immune.combined$dpi)


# Prepare (Normalize and Scale) dataset  
immune.combined <- NormalizeData(immune.combined)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.5)

# Visualize clustering
DimPlot(immune.combined, reduction = "umap", group.by = "batch")
DimPlot(immune.combined, reduction = "umap", label = TRUE)

# Identify MCDK clusters by expression of known markers
mcdk <- c("COL5A2", "EMP1")
plot_genes(immune.combined, mcdk, threshold = 2 , title = "MCDK-Cell markers expression", col = "#CA4866")

# Remove MCDK clusters
idents_to_keep <- setdiff(unique(Idents(immune.combined)), c(9,11))
immune.combined <- subset(immune.combined, idents = idents_to_keep)

# Visualize and check clusters have been removed
DimPlot(immune.combined, reduction = "umap", group.by = "batch")
```

# -------------------------------------
#    Integration
# -------------------------------------

```{r cars}
#------------------
# Integration 
# -----------------
immune.combined.list <- SplitObject(immune.combined, split.by = "batch")

immune.combined.list <- lapply(X = immune.combined.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
# Find anchors 
immune.anchors <- FindIntegrationAnchors(object.list = immune.combined.list, dims = 1:20)
saveRDS(immune.anchors, file.path(robjectsdir, "immune_anchors_invivo.rds"))

# Perform integration 
immune.combined <- IntegrateData(anchorset = immune.anchors, dims = 1:20)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
DefaultAssay(immune.combined) <- "integrated"
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.5)

# Save results 
saveRDS(immune.combined, file.path(robjectsdir, "invivo_immunecombined_cellqc_aftercellandgeneqc_afterScrublet_integrated.rds"))


# Visualize the clusters
immune.combined$group_dpi <- gsub("D", "DPI", immune.combined$dpi)
p1 <- DimPlot(immune.combined, reduction = "umap", group.by = "group_dpi", pt.size = 0.1, cols = brewer.pal(10, "Paired"))+theme_sc
p2 <- DimPlot(immune.combined, reduction = "umap", group.by = "individual")+theme_sc
p3 <- DimPlot(immune.combined, reduction = "umap", label = TRUE)
p4 <- DimPlot(immune.combined, reduction = "umap", group.by = "batch", pt.size = 0.2)+theme_sc
p5 <- DimPlot(immune.combined, reduction = "umap", group.by = "sample", pt.size = 0.00001)+theme_sc
p1; p2; p3; p4; p5
```

# -------------------------------------
#    identify Markers
# -------------------------------------

```{r cars}
# FIND MARKERS  for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(immune.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
saveRDS(pbmc.markers, file.path(robjectsdir,"markers_invivo_normal.rds"))
pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)

b <- c("CD79B", "MS4A1", "CD19")
monocytes <- c("CD14", "LYZ", "CTSB", "PSAP")
t <- c("IL7R", "CD28", "TRIB2", "SPOCK2")
nk <- c( "KLRB1")
dc <- c("IRF8")
neutrophil <- c("CD177", "SOD2")
platelet <- c("PPBP", "PF4")
mdk <- c("COL5A2", "EMP1")



marker.genes <- c(monocytes, b,t)
plot_genes(immune.combined,monocytes,threshold= 2, title = "Monocytes marker expression", col = "purple")
plot_genes(immune.combined, nk, threshold = 2 , title = "NK markers expression", col = "purple")
plot_genes(immune.combined, b, threshold = 2 , title = "B markers expression", col = "purple")
plot_genes(immune.combined, t, threshold = 2 , title = "T-Cell markers expression", col = "purple")
plot_genes(immune.combined, neutrophil, threshold = 4 , title = "Neutrophil-Cell markers expression", col = "purple")

immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "B", `2` = "T", 
   `3` = "T", `4` = "B", `5` = "Myeloid", `6` = "Myeloid", `7` = "Myeloid", `8` = "Myeloid", `9` = "Myeloid",
    `10` = "Myeloid", `11` = "Myeloid", `12` = "Myeloid", `13` = "Myeloid")


# Visualize Markers
DotPlot(immune.combined, features = rev(marker.genes), cols = c("blue", "red", "purple"), dot.scale = 8) + RotatedAxis()+theme(axis.text = element_text(size = 20), axis.title = element_blank())
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(3, "Set2"), label.size = 6)+theme_sc

```

