---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---







# ---------------------------------------
#          STATS: mapped reads
#              Our data 
# ---------------------------------------
```{r cars}
# How many samples
dir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/04_stringtie/"
gtf_our <- iterate_files(dir, "*.stringtie.gtf")


# -----------------------------------------------------------------------------------------
# Collect info about our samples 
# How many tissues do we have, separated by dataset 
tissues <- unlist(lapply(gtf_our, function(x) str_split(basename(x), "_")[[1]][2]))
tissues <- unlist(lapply(tissues, function(x) (str_split(x, "-")[[1]])[1]))
tissues <- gsub("LN", "Lymph node", tissues)

dataset <- unlist(lapply(gtf_our, function(x) str_split(basename(x), "_")[[1]][1]))
df_our_summary <- data.frame(tissue = tissues, dataset= dataset, stringsAsFactors = F)
# -----------------------------------------------------------------------------------------

# 1. Plot number of samples
tot_n_samples <- nrow(df_our_summary)
dataset_n_samples <- as.data.frame(df_our_summary %>% group_by(dataset) %>% dplyr::count())
rownames(dataset_n_samples) <- dataset_n_samples$dataset
ggplot(df_our_summary, aes(x = tissue, fill = dataset ))+geom_bar(width = 0.7)+theme_paper+scale_fill_brewer(palette = "Set3")+theme(axis.text.x = element_text(size = 12, angle =24, vjust =0.9, hjust = 1.0))+ylab("#Samples")+xlab("")+ggtitle(paste("Total number of samples: ", tot_n_samples, sep = " "))

# 2. Coverage
dir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/03b_rseqc/"
stat_files_hisat_our <- iterate_files(dir, "*.bam_stat.txt")
length(unique(stat_files_hisat_our))

# --- Find the samples which were filtered out and do not use bam stats from those 
bams_names <- gsub(".bam", "", unlist(lapply(stat_files_hisat_our, function(x) paste(unlist(str_split(basename(x),"_")[[1]][1:4]), collapse = "_"))))
gtf_names <- gsub(".UMI.f3.q60.stringtie.gtf", "", unlist(lapply(gtf_our, function(x) paste(unlist(str_split(basename(x),"_")[[1]][1:4]), collapse = "_"))))
gtf_names <- gsub(".UMI.f3.q60.umi", "", gtf_names)
filtered_out_samples <- bams_names[!(bams_names %in% gtf_names)]

stat_files_hisat_our_filtered <- stat_files_hisat_our [!(bams_names %in% filtered_out_samples)]


mapped_ours <- data.frame(mapped_reads = unlist(lapply(stat_files_hisat_our_filtered, function(lines) get_mapped((lines)))))
mapped_ours$source <- "Our"
mapped_ours$file <- gtf_names
mapped_ours$tissue <- unlist(lapply(mapped_ours$file, function(x) str_split(x, "_")[[1]][2]))
mapped_ours$tissue <- unlist(lapply(mapped_ours$tissue, function(x) (str_split(x, "-")[[1]])[1]))
mapped_ours$tissue <- gsub("LN", "Lymph node", mapped_ours$tissue)
mapped_ours$dataset <- unlist(lapply(mapped_ours$file, function(x) str_split(x, "_")[[1]][1]))
mapped_ours$dpi <- unlist(lapply(mapped_ours$file, function(x) str_split(x, "_")[[1]][3]))

nb.cols <- 14
mycolors <- colorRampPalette(brewer.pal(8, "Set3"))(nb.cols)
# Average mapped reads per tissue
ylab <- c(20,40,60)
ggplot(mapped_ours, aes( x = file, fill = tissue, y = mapped_reads))+geom_bar(stat = "identity", width= 0.8)+theme_paper+scale_fill_manual(values = mycolors)+theme(axis.text.x = element_text(size = 12, angle =24, vjust =0.9, hjust = 1.0))+ylab("# Mapped reads")+xlab("")+theme(axis.text.x = element_blank(), axis.ticks.x =element_blank() )+ facet_grid(~dataset, scales = "free_x", space= "free_x")+scale_y_continuous( labels = paste0(ylab, "M"),breaks = 10^6 * ylab)

```



# -------------------------------------------------------------------------------------------------------------------------------------
#                                                    OLD
# -------------------------------------------------------------------------------------------------------------------------------------
# Rseqc summary 
# Intron mapping reads
```{r cars}
library(ggpubr)
dir <- "/home/luisas/Desktop/cluster/data/99_BroadAnnotation_Feb2021/02_rseqc//"
read_distribution_files_broad <- iterate_files(dir, "*.read_distribution.txt")
read_distribution_files_our <- iterate_files("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/03b_rseqc/", "*q60.read_distribution.txt")

get_intron_perc <- function(file){
  lines <- readLines(file)
  total_reads <- as.double(strsplit(lines[grep(pattern = "Total Assigned Tags", x = lines)], " +")[[1]][4])
  introns <- as.double(strsplit(lines[grep(pattern = "Introns", x = lines)], " +")[[1]][3])
  perc <- introns/total_reads
  return(perc)
}

get_exon_perc <- function(file){
  lines <- readLines(file)
  total_reads <- as.double(strsplit(lines[grep(pattern = "Total Assigned Tags", x = lines)], " +")[[1]][4])
  introns <- as.double(strsplit(lines[grep(pattern = "CDS_Exons", x = lines)], " +")[[1]][3])
  perc <- introns/total_reads
  return(perc)
}

introns_broad <- data.frame(introns = unlist(lapply(read_distribution_files_broad, function(lines) get_intron_perc((lines)))))
introns_broad$source <- "Broad"
introns_broad$type <- "introns"
introns_our <- data.frame(introns = unlist(lapply(read_distribution_files_our, function(lines) get_intron_perc((lines)))))
introns_our$source <- "Our"
introns_our$type <- "introns"


exons_broad <- data.frame(introns = unlist(lapply(read_distribution_files_broad, function(lines) get_exon_perc((lines)))))
exons_broad$source <- "Broad"
exons_broad$type <- "exons"
exons_our <- data.frame(introns = unlist(lapply(read_distribution_files_our, function(lines) get_exon_perc((lines)))))
exons_our$source <- "Our"
exons_our$type <- "exons"

df <- rbind(introns_broad, introns_our, exons_our, exons_broad)
ggboxplot(df, x = "type", y = "introns",
                color = "source", palette =c("#00AFBB", "#E7B800", "#FC4E07"),
                add = "jitter", shape = "source")+theme_paper+ylab("")


ggboxplot(rbind(introns_broad, introns_our), x = "type", y = "introns",
                color = "source", fill = "source", alpha = 0.2,  palette =c("#00AFBB", "#E7B800", "#FC4E07"),
                add = "jitter", shape = "source")+theme_paper+xlab("")+ylab("Proportion intron mapping reads")

```





