---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---



# -------------------------------------
#       DEA
# -------------------------------------

```{r results='hide', message=FALSE, warning=FALSE}
library(edgeR)
library(stringr)
library(dplyr)
library(ggplot2)
library("wesanderson")
library(RColorBrewer)


# 0. Palettes & themes 
pal1 <- wes_palette("Darjeeling1", 5, type = "discrete")
pal2 <- wes_palette("Darjeeling2", 5, type = "discrete")
colfunc <- colorRampPalette(c("grey", "red"))

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))


# 1. Input Variables and Files
tissue <- "Lymph node"
outpath <- paste0("/home/luisas/Desktop/cluster/data/99_BroadAnnotation_Feb2021/05_DEA/Tissues/",tissue,"/")

# Sade objects
my_data <- readRDS(paste0(outpath,tissue,".voom_limma.data_objects.rds"))

dge <- my_data[["dge"]] 
v <- my_data[["v"]] 
fit <- my_data[["fit"]] 

metadata_tissue$batch.extraction <- as.factor(metadata_tissue$batch.extraction)
metadata_tissue$infection_status <- as.factor(metadata_tissue$infection_status)
```



# Explore dataset

```{r prcomp}
ggplot(metadata_tissue, aes(x = id.cohort,  y = pc.viral.reads, col =pc.viral.reads ))+geom_point(size = 2)+theme_paper+scale_color_continuous()
ggplot(metadata_tissue, aes(x = id.cohort,  y = batch.extraction ))+geom_point(size = 2)+theme_paper+scale_color_continuous()

metadata_tissue$id.cohort <- factor(metadata_tissue$id.cohort, levels = c("D000", "D001", "D002", "D003", "D004", "D005","D006","D007","D008" ))

# Check dataframe 


#-------------------------
#   Samples overview
#-------------------------

# Order samples by DPI 
df <- metadata_tissue[order(metadata_tissue$dpi_time_factor),]
df$X <- factor(df$X,levels = df$X)

variable <- "subtissue"
ggplot(df, aes(x = X, y = pc.viral.reads, col  =  eval(parse(text = variable)), fill = eval(parse(text = variable))))+geom_bar(stat = "identity")+theme_paper+xlab("Samples sorted by DPI")+ylab("% Viral read per sample")+theme(axis.ticks.x = element_blank(), axis.text.x = element_blank())+ggtitle("Lymph node")+theme(title = element_text(size = 20))+scale_color_brewer(palette = "Set3")+scale_fill_brewer(palette = "Set3") +facet_grid(. ~ id.cohort, scales = "free_x") 


```



```{r prcomp}

df_ebola <- merge(data.frame(viralcount = colSums(counts[grepl("EBOV", rownames(counts)),])), metadata_tissue, by = "row.names", all.x = F) 
ggplot(df_ebola, aes(x = viralReadCount,  y = viralcount ))+geom_point(size = 2)+theme_paper+scale_color_continuous()

df_ebola$viralReadCount

df_ebola[,c("id.cohort", "viralcount", "viralReadCount")]

# PROBLEM: my quantification not consistent
#ggplot(metadata_tissue, aes(x = id.cohort,  y = viralcount, col =viralcount ))+geom_point(size = 2)+theme_paper+scale_color_continuous()

# Find a D000 samples with a lot of viral reads ( according to my quantification )
df_ebola[df_ebola$id.cohort == "D000", c("X","viralcount"), drop = F ]
sample_toinvestigate <- "A0276_S127_L002"
df_ebola[df_ebola$X == sample_toinvestigate,]




counts_list <- lapply(quantification_tissue, get_df_count)
rn <- rownames(counts_list[[1]])
dat <- counts_list[[1]]
i <- 2
for(i in 2:length(counts_list)) {
  dat <- merge(dat, counts_list[[i]],  by= "row.names", all.x= T, all.y= F)
  #rownames(dat) <- rn
}

merged.data.frame = Reduce(function(...) merge(..., all=T), list.of.data.frames)


counts <- dat
counts

# Check 
mm <-counts[,grepl(sample_toinvestigate, colnames(counts) ), drop = F ]
mm[grepl("EBOV", rownames(mm)),]

```



```{r prcomp}

FDRTHRESHOLD <- 0.05
# Run test 
efit <- eBayes(fit)
summary(decideTests(efit), coef = "infection_statusInfected")

metadata_tissue$showing_infection 

metadata_tissue$viralreadsfound <- "notfound"
metadata_tissue[metadata_tissue$pc.viral.reads > 0.01,]$viralreadsfound <-  "found"
table(metadata_tissue$viralreadsfound)

(Reduce("cbind",lapply(quantification_tissue, get_df_count)))
```




```{r prcomp}

FDRTHRESHOLD <- 0.05
# Run test 
efit <- eBayes(fit)
summary(decideTests(efit), coef = "infection_statusInfected")


res <- topTable(efit, coef = "infection_statusInfected", adjust.method = "fdr", confint = TRUE, n = Inf)

sig_de <- res[res$adj.P.Val < FDRTHRESHOLD, ]


rownames(sig_de[sig_de$logFC < 0 ,])

library(org.Mmu.eg.db)
library(rtracklayer)
library(cl)

do_go <- function(list, universe, keytype = "ENSEMBL"){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =  org.Mmu.eg.db, keyType = keytype,ont = "BP",universe =universe)
  barplot(ego, showCategory=10)
  clusterProfiler::dotplot(ego)
}
gene_annotation  <- import("/home/luisas/Desktop/cluster/data/99_BroadAnnotation_Feb2021/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf")

universe <- rownames(fit)
list <- rownames(sig_de[sig_de$logFC > 0,])


list <- unlist(lapply(list, function(x) strsplit(x,".", fixed = T)[[1]][1]))
do_go(rownames(sig_de[sig_de$logFC > 0,]), rownames(fit))

efit <- eBayes(fit)
summary(decideTests(efit), coef = "infection_statusInfected")


table(metadata_tissue$id.cohort)
```


















