---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)


# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))

col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
```

## Read in all predictions from CPC2, CNIT and CPAT

```{r cpc2}
# -------------------------------------------
#                     CPC2 
# -------------------------------------------
candidates <- import("/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/03_novel_lncRNAs_list/02_novel_expressed/candidates.gtf")
# Load cpc2 prediction 
cpc2 <- read.table("/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/03_novel_lncRNAs_list/03_predictions/CPC2/cpc2_pred.txt")
names(cpc2) <- c("ID","transcript_length ","peptide_length","Fickett_score" , " pI", "ORF_integrity", "coding_probability","label")
# Extract transcript ids 
cpc2$transcript_id <- unlist(lapply(as.character(cpc2$ID), function(x) strsplit(x,"[(]")[[1]][1]))
# Extract non coding predictions
pred_nc_ids <- cpc2[cpc2$label == "noncoding",]$transcript_id
cpc2_pre_lnc <- candidates[candidates$transcript_id %in% pred_nc_ids, ]

# -----------------------------------------------
# Remove uncorcondant pred
cpc2$gene_id <- unlist(lapply(as.character(cpc2$transcript_id), function(x) paste(unlist(strsplit(x, "[.]"))[1:2], collapse = ".")))

unconcordant_prediction <- cpc2 %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(label)) %>%  dplyr::filter( Unique_Elements > 1)
cpc2_pre_lnc <- cpc2_pre_lnc[!(cpc2_pre_lnc$gene_id %in% unconcordant_prediction$gene_id)]
length(unique(cpc2_pre_lnc$gene_id))
# -----------------------------------------------


# -------------------------------------------
#                     CPAT 
# -------------------------------------------
# Compare which ones and how many overlap 
cpat <- read.table("/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/03_novel_lncRNAs_list/03_predictions//CPAT/cpat_pred.ORF_prob.tsv", header = TRUE)
# Cutoff defined by the paper of CPAT as best for discerning coding and lnc
cutoff <-  0.364
cpat$label <- ifelse(cpat$Coding_prob > cutoff, "coding", "noncoding")
cpat$transcript_id <- unlist(lapply(as.character(cpat$ID), function(x) strsplit(x,"[(]")[[1]][1]))
# extract lncRNAs
cpat_nc <- cpat[cpat$label == "noncoding",]
cpat_pre_lnc <- candidates[candidates$transcript_id %in% cpat_nc$transcript_id, ]

# -----------------------------------------------
# Remove uncorcondant pred
cpat$gene_id <- unlist(lapply(as.character(cpat$ID), function(x) paste(unlist(strsplit(x, "[.]"))[1:2], collapse = ".")))
unconcordant_prediction <- cpat %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(label)) %>%  dplyr::filter( Unique_Elements > 1)
cpat_pre_lnc <- cpat_pre_lnc[!(cpat_pre_lnc$gene_id %in% unconcordant_prediction$gene_id)]
length(unique(cpat_pre_lnc$gene_id))
# -----------------------------------------------



# -------------------------------------------
#                     CNIT 
# -------------------------------------------
# Compare which ones and how many overlap 
cnit <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/03_predictions/CNIT/cnit_pred/CNCI2.index", header = TRUE, ,sep="\t")
cnit$label <- cnit$index
cnit$transcript_id <- unlist(lapply(as.character(cnit$Transcript.ID), function(x) strsplit(x,"[(]")[[1]][1]))
cnit_nc <- cnit[cnit$label == "noncoding",]
cnit_pre_lnc <- candidates[candidates$transcript_id %in% cnit_nc$transcript_id, ]

# -----------------------------------------------
# Remove uncorcondant pred
cnit$gene_id <- unlist(lapply(as.character(cnit$Transcript.ID), function(x) paste(unlist(strsplit(x, "[.]"))[1:2], collapse = ".")))
cnit$transcript_id <- cnit$ID
unconcordant_prediction <- cnit %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(label)) %>%  dplyr::filter( Unique_Elements > 1)
cnit_pre_lnc <- cnit_pre_lnc[!(cnit_pre_lnc$gene_id %in% unconcordant_prediction$gene_id)]
length(unique(cnit_pre_lnc$gene_id))
# -----------------------------------------------
```


# Obtain the intersection of all the predictions

```{r cpc2vscpan}
length(unique(cpat_pre_lnc$transcript_id))
length(unique(cpc2_pre_lnc$transcript_id))

intersection <- unique(intersect(cpat_pre_lnc$transcript_id, cpc2_pre_lnc$transcript_id))
intersection_pre_lnc <- candidates[candidates$transcript_id %in% intersection, ]
length(unique(intersection_pre_lnc$transcript_id))
length(unique(intersection_pre_lnc$gene_id))
plot_stats_annotation(intersection_pre_lnc,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)
length(unique(intersection_pre_lnc$transcript_id))
length(unique(intersection_pre_lnc$gene_id))



# Remove monoexonic
summary_exons <- exon_nr_summary(intersection_pre_lnc)
monoexonic_transcript_ids <- summary_exons[summary_exons$max_exon == 1,]$transcript_id
intersection_pre_lnc_multi <- intersection_pre_lnc[!(intersection_pre_lnc$transcript_id %in% monoexonic_transcript_ids),]

plot_stats_annotation(intersection_pre_lnc_multi,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)

length(unique(intersection_pre_lnc_multi$transcript_id))
length(unique(intersection_pre_lnc_multi$gene_id))


gffcompare <- import("/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/01_stringtie_assembly_merged/01_gffCompare/merged.annotated.gtf")
gff_compare_pre_lnc <- gffcompare[gffcompare$transcript_id %in% intersection_pre_lnc_multi$transcript_id & gffcompare$type == "transcript",]
antisense_ids <- gff_compare_pre_lnc[gff_compare_pre_lnc$class_code == "x", ]$transcript_id
intergenic_ids <- gff_compare_pre_lnc[gff_compare_pre_lnc$class_code == "u", ]$transcript_id


intergenic_lnc <- intersection_pre_lnc_multi[intersection_pre_lnc_multi$transcript_id %in% intergenic_ids, ]
antisense_lnc <- intersection_pre_lnc_multi[intersection_pre_lnc_multi$transcript_id %in% antisense_ids, ]
library(RColorBrewer)
plot_stats_annotation_separated(intergenic_lnc,antisense_lnc ,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)


```



# Visualize overlap 
```{r cpc2vscpan}
library(UpSetR)
library(ComplexHeatmap)
# example of list input (list of named vectors)
!(intersection_pre_lnc$transcript_id %in% monoexonic_transcript_ids)
cpat_pre_lnc_multi <- cpat_pre_lnc[!(cpat_pre_lnc$transcript_id %in% monoexonic_transcript_ids)]
cpc2_pre_lnc_multi <- cpc2_pre_lnc[!(cpc2_pre_lnc$transcript_id %in% monoexonic_transcript_ids)]


listInput <- list(CPAT = unique(cpat_pre_lnc_multi$gene_id), CPC2 = unique(cpc2_pre_lnc_multi$gene_id))
upset(fromList(listInput), order.by = "freq", main.bar.color = c("darkorange", rep("gray23",2)),text.scale = 2.4)

```

```{r venn}
# ---------------------------
#   Add Names 
# ---------------------------
add_gene_name <- function(gtf){
  missing_gene_name <- gtf[gtf$type == "exon"][is.na(gtf[gtf$type == "exon"]$gene_name)]                                
  transcripts <- gtf[gtf$type == "transcript"]
  # To the ones that have no gene name, find the gene name associated with the transcript ID and use it
  test <-  lapply(seq_along(missing_gene_name), function(x) transcripts[transcripts$transcript_id == missing_gene_name[x]$transcript_id]$gene_name)
  gtf[gtf$type == "exon"][is.na(gtf[gtf$type == "exon"]$gene_name)]$gene_name <- unlist(test)
  gtf[is.na(gtf$gene_name)]$gene_name <- paste(gtf[is.na(gtf$gene_name)]$gene_id, "-unknown", sep = "")
  return(gtf)
}
add_prefix <- function(gtf, prefix){
  gtf$gene_id <- paste(prefix, gtf$gene_id, sep = "") 
  gtf$transcript_id <- paste(prefix, gtf$transcript_id, sep = "") 
  return(gtf)
}


#gtf_ribodepl_newids <- add_prefix(ribodepl_novel, "ribodepl-")

ref_gtf <- import("/home/luisas/Desktop/cluster/gene_annotations/UCSC/rheMac10/rheMac10.gtf")
ribodepl_novel <- intersection_pre_lnc
# ---------------------------------------
# Merge Novel and Ref  together ( just append ) 
# ---------------------------------------
glst_ref_and_novel <-list(ref_gtf,ribodepl_novel)
gtf_ref_and_novel <- do.call(c, as(glst_ref_and_novel, "GRangesList"))

outfile <- "/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/03_novel_lncRNAs_list/rheMac10_novel.gtf"
gtf_ref_and_novel_sort <- sort(gtf_ref_and_novel)
export(gtf_ref_and_novel_sort, outfile)  


outfile_name <- "/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/03_novel_lncRNAs_list/rheMac10_novel_genenames.gtf"


table(gtf_ref_and_novel_sort$source)

mask <- is.na(gtf_ref_and_novel_sort$gene_name)
gtf_ref_and_novel_sort[mask]$gene_name <- paste(gtf_ref_and_novel_sort[mask]$gene_id, "-unknown", sep = "") 
mask_t <- is.na(gtf_ref_and_novel_sort$transcript_name)
gtf_ref_and_novel_sort[mask_t]$transcript_name <-paste(gtf_ref_and_novel_sort[mask_t]$transcript_id, "-unknown", sep = "")  
export(gtf_ref_and_novel_sort, outfile_name)  


length(unique(intersection_pre_lnc_multi$transcript_id))
length(unique(intersection_pre_lnc_multi$gene_id))

u_ids <- gffcompare[gffcompare$type == "transcript" & gffcompare$class_code == "u",]$transcript_id
x_ids <- gffcompare[gffcompare$type == "transcript" & gffcompare$class_code == "x",]$transcript_id

length(unique(intersection_pre_lnc_multi[intersection_pre_lnc_multi$transcript_id %in% u_ids,]$transcript_id))
length(unique(intersection_pre_lnc_multi[intersection_pre_lnc_multi$transcript_id %in% x_ids,]$transcript_id))
```


# Supplementary figures

## Compare number of genes in macaques vs Human annotation 
```{r results='hide', message=FALSE, warning=FALSE}

palette_expression <-c("#E8C2D8", "#D4549C", "#8195D7")

iterate_files <- function(inpath, pattern_string){
  files <- list.files(path=inpath, pattern= pattern_string, full.names=TRUE, recursive=TRUE)
  return(files)
}


dir_counts_ref <- "/home/luisas/Desktop/cluster/data/99_BroadAnnotation_rerunningassemly_old/03_novel_lncRNAs_list/04_quantification_new_new/"


abundance_files <- iterate_files(dir_counts_ref, "*.tsv")

# Read all files
abundances <- list()
for(i in 1:length(abundance_files)) {
  file <- readr::read_tsv(abundance_files[i])
  # Sum up values for double entries (https://github.com/gpertea/stringtie/issues/192)
  file <- file %>% dplyr::group_by(`Gene ID`)%>% dplyr::summarise(TPM = sum(TPM))
  abundances[[i]] <- data.frame(file$TPM, row.names =file$`Gene ID` )
}

# Summarize all TPMs from all quantification files 
rn <- rownames(abundances[[1]])
dat <- abundances[[1]]
for(i in 2:length(abundances)) {
  dat <- merge(dat, abundances[[i]],  by= "row.names", all.x= F, all.y= F) [,-1]
  rownames(dat) <- rn
}

expression <- dat
expression <- as.matrix(expression)

ref <- import("/home/luisas/Desktop/cluster/gene_annotations/UCSC/rheMac10/rheMac10.gtf")
matching <- read.table("/home/luisas/Desktop/slncky-master/annotations/ensemblSource.txt")
lnc_ids <- matching[matching$V2 == "lncRNA",]$V1
pc_ids <- matching[matching$V2 == "protein_coding",]$V1

lnc_gene_ids <- ref[ref$type == "transcript" & ref$transcript_id %in% lnc_ids, ]$gene_id
pc_gene_ids <- ref[ref$type == "transcript" & ref$transcript_id %in% pc_ids, ]$gene_id

lncRNAs_ref <- ref[ref$gene_id %in% lnc_gene_ids, ]
mRNAs_ref <- ref[ref$gene_id %in% pc_gene_ids, ]
library(matrixStats)
# ----------------------------------------------------------------------
# Same for the median
# ----------------------------------------------------------------------
median_expression <- data.frame(id=rownames(expression), expr=log(rowMedians(as.matrix(expression))))
median_expression['type'] <- "0"
median_expression <- add_type(median_expression, intersection_pre_lnc_multi$gene_id, "Novel lncRNAs")
median_expression <- add_type(median_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
median_expression <- add_type(median_expression, mRNAs_ref$gene_id, "mRNAs")
median_expression <- median_expression[!median_expression$type == "0",]
plot_expression(median_expression, palette_expression, title = "Median Expression")+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))+labs(y = "logTPM")
```