---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Differential expression analysis IN VIVO 

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(ggplot2)
library(scales)

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation 
mtx <- "pre"
genes <- "GSE158055_covid19_features.csv.gz"
genes <- "genes_mod.tsv"
barcodes <- "GSE158055_covid19_barcodes.csv"
path <- "/home/luisas/Desktop/cluster/data/00_RawData/COVID_scRNAseq/"
a <- data.frame(a = rownames(Genes), rownames(Genes)) 
write.table(a, file.path(path, "genes_mod.tsv"), col.names = F, row.names = F ,sep = "\t")


library(edgeR)
covid.dataset <- read10X_alt(mtx = mtx, genes = genes, path = "/home/luisas/Desktop/cluster/data/00_RawData/COVID_scRNAseq/", DGEList = FALSE)


read10X_alt <- function(mtx = mtx, genes = genes, path = "/home/luisas/Desktop/cluster/data/00_RawData/COVID_scRNAseq/", DGEList = FALSE){
    mtx <- file.path(path, mtx)
    genes <- file.path(path, genes)
    if (!is.null(barcodes)) 
        barcodes <- file.path(path, barcodes)
    N <- scan(mtx, skip = 2, what = 0L, sep = " ", nmax = 3, 
        quiet = TRUE)
    ngenes <- N[1]
    ncells <- N[2]
    nmtx <- N[3]
    Genes <- read.table(genes, header = FALSE, comment.char = "", 
        sep = "\t", row.names = 1, colClasses = "character")
    
    if (nrow(Genes) != ngenes) 
        stop("Number of feature IDs doesn't agree with header information in mtx file")
    names(Genes)[1] <- "Symbol"

    m <- read.table(mtx, skip = 3, header = FALSE, comment.char = "", 
        sep = " ", colClasses = "real", nrows = nmtx)
    y <- matrix(0L, ngenes, ncells)
    i <- m[, 1] + (m[, 2] - 1L) * ngenes
    y[i] <- m[, 3]
    dimnames(y) <- list(Gene = row.names(Genes), Cell = 1:ncells)
    if (is.null(barcodes)) {
        Samples <- NULL
    }
    else {
        Barcodes <- scan(barcodes, what = "", quiet = TRUE)
        if (length(Barcodes) != ncells) 
            stop("Number of barcodes doesn't agree with header information in mtx file")
        Samples <- data.frame(Barcode = Barcodes)
    }
    if (DGEList) {
        DGEList(count = y, genes = Genes, samples = Samples)
    }
    else {
        list(counts = y, samples = Samples, genes = Genes)
    }
}

```






