---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# ----------------------------------------------------
# General Expression Patterns of lncRNAs IN VIVO
# ----------------------------------------------------


### Imports 
```{r include=FALSE}
library(ggplot2)
library(ggpubr)
library(reshape2)
library(cowplot)
library(MatchIt)

source(file.path("../utils/00_datapaths.R"))
source("../utils/02_sc_utils.R")

# 0. Palette used throughout the scripts
col_lnc = "#f72631"
col_mrna = "#3153a3"
palette_plot_percentage <- c(col_lnc, col_mrna)

# 1. Theme for plots
theme_sc_paper <- theme(panel.background = element_rect(fill = "white", colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())+theme_paper

theme_matching <- theme(panel.background = element_rect(fill = "white", color = "black"))+
                  theme(legend.position ="",panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.ticks.length=unit(.2, "cm"))+
                  theme(axis.text = element_text(size = 15, color ="black"), axis.title = element_text(size = 17))

# 2. Directories
# Where stats about lnc and pc computed on cluster are stored
robjectsdir<- file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/")

# 3. Load stats
df_complete_celltype <- readRDS(file.path(robjectsdir, "df_celltype.rds"))
df_complete_celltype <- df_complete_celltype[df_complete_celltype$n_cells > 30,]



df_complete <- readRDS(file.path(robjectsdir, "df_complete.rds"))
df_lnc <- readRDS(file.path(robjectsdir, "df_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))
df_lnc$gene_id <- as.character(df_lnc$gene_id)
df_mrna$gene_id <- as.character(df_mrna$gene_id)

# Apply same filters
df_lnc <- df_lnc[df_lnc$n_cells > 60, ]
df_mrna <- df_mrna[df_mrna$n_cells > 60, ]
df_lnc$type <- "lncRNA"
df_mrna$type <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)

df_complete_celltype <- df_complete_celltype[df_complete_celltype$gene_id %in% c(df_lnc$gene_id, df_mrna$gene_id),]
df_complete_celltype$gene_id <- as.character(df_complete_celltype$gene_id)
df_complete_celltype$type <- "NA"
df_complete_celltype[df_complete_celltype$gene_id %in% df_lnc$gene_id,]$type <- "lnc"
df_complete_celltype[df_complete_celltype$gene_id %in% df_mrna$gene_id,]$type <- "pc"

# Import Seurat object 
file <- file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
immune.combined <- readRDS(file)
```


# ---------------------------------------------------------
# 1. Visualzation: # Cells and median expression of all genes 
# ---------------------------------------------------------
```{r quartiles}


#  ------- 1 :  PERCENTAGE OF CELLS EXPRESSING A GENE -----------
# 1a. Density plot of % cells expressing 
density_cells <- ggdensity(df_complete, x = "perc_cells_expressing",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("% cells")+ylab("density")+scale_x_log10()+theme(text = element_text(size = 20))

# Save plot 
pdf(file.path(plots, "02/density_cells_scalexlog.pdf"), width = 6, height = 6)
density_cells
dev.off()


pdf(file.path(plots, "02/density_cells_NOscalexlog.pdf"), width = 6, height = 6)
ggdensity(df_complete, x = "perc_cells_expressing",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("% cells")+ylab("density")+theme(text = element_text(size = 20))
dev.off()


# Cumulative density expression 
pdf(file.path(plots, "02/cum_density_cells.pdf"), width = 6, height = 6)
cum_density_cells <- ggplot(df_complete, aes(perc_cells_expressing, colour = type)) + stat_ecdf(size = 1.5)+xlab("% cells")+ylab("cumulative density")+scale_color_manual(values=palette_plot_percentage)+theme_sc_paper+ theme(panel.border = element_rect(colour = "black", fill=NA))+theme(text = element_text(size = 30, color = "black"), legend.position = "")
cum_density_cells
dev.off()

# 1b. boxplot median expression 
density_expr <- ggboxplot(df_complete, x = "type", y = "perc_cells_expressing",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("")+ylab("% cells where gene is expressed")+theme(text = element_text(size = 20))+stat_compare_means()


# Save plot 
pdf(file.path(plots, "02/boxplot_cells.pdf"), width = 4, height = 6)
density_expr
dev.off()


# Wilcox test 

#  ------- 2: VISUALIZE MEDIAN EXPRESSION  -----------

# 1a. Density plot of % cells expressing 
density_cells <- ggdensity(df_complete, x = "medianexpr",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("median expression")+ylab("density")+theme(text = element_text(size = 20))+scale_x_log10()

# Save plot 
pdf(file.path(plots, "02/density_medianexpression_scalexlog.pdf"), width = 6, height = 6)
density_cells
dev.off()

# Save plot 
pdf(file.path(plots, "02/density_medianexpression_NOscalexlog.pdf"), width = 6, height = 6)
ggdensity(df_complete, x = "medianexpr",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("median expression")+ylab("density")+theme(text = element_text(size = 20))
dev.off()


# 1b. boxplot median expression 
density_expr <- ggboxplot(df_complete, x = "type", y = "medianexpr",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("")+ylab("median expression")+theme(text = element_text(size = 20))+stat_compare_means()

# Save plot 
pdf(file.path(plots, "02/boxplot_medianexpression.pdf"), width = 4, height = 6)
density_expr
dev.off()


# Cumulative density expression 
pdf(file.path(plots, "02/cum_density_median.pdf"), width = 6, height = 6)
cum_median <- ggplot(df_complete, aes(medianexpr, colour = type)) + stat_ecdf(size = 1.5)+xlab("median expression")+ylab("cumulative density")+scale_color_manual(values=palette_plot_percentage)+theme_sc_paper+ theme(panel.border = element_rect(colour = "black", fill=NA))+theme(text = element_text(size = 30, color = "black"), legend.position = "")
cum_median
dev.off()


#  ------- 3: VISUALIZE MAX EXPRESSION  -----------

# 1a. Density plot of % cells expressing 
density_cells <- ggdensity(df_complete, x = "maxexpr",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("maximum expression")+ylab("density")+theme(text = element_text(size = 20))

# Save plot 
pdf(file.path(plots, "02/density_maximumexpression.pdf"), width = 6, height = 6)
density_cells
dev.off()

# 1b. boxplot median expression 
density_expr <- ggboxplot(df_complete, x = "type", y = "maxexpr",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("")+ylab("maximum expression")+theme(text = element_text(size = 20))+stat_compare_means()

# Save plot 
pdf(file.path(plots, "02/boxplot_maximumexpression.pdf"), width = 4, height = 6)
density_expr
dev.off()



# Cumulative density expression 
pdf(file.path(plots, "02/cum_density_max.pdf"), width = 6, height = 6)
cum_median <- ggplot(df_complete, aes(maxexpr, colour = type)) + stat_ecdf(size = 1.5)+xlab("max expression")+ylab("cumulative density")+scale_color_manual(values=palette_plot_percentage)+theme_sc_paper+ theme(panel.border = element_rect(colour = "black", fill=NA))+theme(text = element_text(size = 30, color = "black"), legend.position = "")
cum_median
dev.off()

# WILCOX test on complete distributions
wilcox.test(df_lnc$perc_cells_expressing, df_mrna$perc_cells_expressing, paired = F)
wilcox.test(df_lnc$medianexpr, df_mrna$medianexpr, paired = F, alternative = "two.sided")
wilcox.test(df_lnc$maxexpr, df_mrna$maxexpr, paired = F)

```
# ---------------------------------------------------------
# 2. QUARTILES
# ---------------------------------------------------------

```{r quartiles}

# ---------------- 1:  general inspection 
# Scatter plot MEDIAN
pdf(file.path(plots, "02/SUPPL_scatterplot.pdf"), width = 6, height = 6)
scatter_plot_both(df_lnc,df_mrna, y = "% cells", type = "median",palette = palette_plot_percentage)
dev.off()

df_lnc_celltype <- df_complete_celltype[df_complete_celltype$type == "lnc",]
df_pc_celltype <- df_complete_celltype[df_complete_celltype$type == "pc",]


# Scatter plot MAX
pdf(file.path(plots, "02/SUPPL_scatterplot_max.pdf"), width = 6, height = 6)
scatter_plot_both(df_lnc,df_mrna, y = "% cells", type = "max",palette = palette_plot_percentage)
dev.off()

# ------------------2: quartiles 
# Quartiles 
pdf(file.path(plots, "02/SUPPL_quartiles.pdf"), width = 6, height = 6)
length(unique(df_mrna$gene_id))
plot_quartiles(df_lnc, df_mrna, "proportion", palette = palette_plot_percentage)+theme(text = element_text(size = 15))
dev.off()


# ----------------- 3: cell-types 

# Boxplot separated by cell-type
pdf(file.path(plots, "02/SUPPL_celltypes.pdf"), width = 6, height = 6)
ggplot(df_complete_celltype, aes( x= type, y = perc_cells_expressing, fill = type, col = type))+geom_boxplot(alpha= 0.4)+scale_fill_manual(values = palette_plot_percentage)+scale_color_manual(values = palette_plot_percentage)+theme_sc_paper+xlab ("")+ylab("% cells")+ facet_wrap(vars(ident))+theme(strip.text.x = element_text(size = 20, colour = "black", angle = 0))+theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "")
dev.off()




```


# --- Expression levels 

```{r quartiles}
# Boxplot separated by cell-type



df_complete_celltype$type <- gsub("lnc", "lncRNA", df_complete_celltype$type)
df_complete_celltype$type <- gsub("pc", "mRNA", df_complete_celltype$type)
df_complete_celltype$type <- factor(df_complete_celltype$type, levels = c("lncRNA","mRNA"))

get_expression_plot <- function(ident,df_complete_celltype, metric = "medianexpr"){
  df_ident <- df_complete_celltype[df_complete_celltype$ident == ident, ]
  if(metric == "medianexpr"){
      df_ident <- df_ident[!is.na(df_ident$medianexpr),]
  }else if(metric == "maxexpr"){
      df_ident <- df_ident[!is.na(df_ident$maxexpr),]
  }
  df_ident <- df_ident[!is.na(df_ident$medianexpr),]
  my_comparisons <- list( c("lncRNA", "mRNA"))
  p <- ggboxplot(df_ident, x = "type", y = metric, fill = "type", col = "type",alpha=1)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab(paste(gsub("expr","",metric), " expression (log(CP10K+1))"))+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,10), expand = c(0,0))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons)+ggtitle(ident)+theme(plot.title = element_text(hjust = 0.5))
  return(p)
}

test_median_wilcox <- function(ident, df_complete_celltype){
  # ---------------
  df_ident <- df_complete_celltype[df_complete_celltype$ident == ident, ]
  df_ident <- df_ident[!is.na(df_ident$medianexpr),]
  return(wilcox.test(df_ident[as.character(df_ident$type) == "lncRNA", ]$medianexpr, df_ident[as.character(df_ident$type) == "mRNA", ]$medianexpr))
  # ---------------
}
pdf(file.path(plots, "02/SUPPL_medianMono.pdf"), width = 2, height = 4)
get_expression_plot("Monocyte", df_complete_celltype)
dev.off()

pdf(file.path(plots, "02/SUPPL_medianB.pdf"), width = 2, height = 4)
get_expression_plot("B", df_complete_celltype)
dev.off()

pdf(file.path(plots, "02/SUPPL_medianT.pdf"), width = 2, height = 4)
get_expression_plot("T", df_complete_celltype)
dev.off()


pdf(file.path(plots, "02/SUPPL_medianNeutrophil.pdf"), width = 2, height = 4)
get_expression_plot("Neutrophil", df_complete_celltype)
dev.off()


# -----------------------

pdf(file.path(plots, "02/SUPPL_maxMono.pdf"), width = 2, height = 4)
get_expression_plot("Monocyte", df_complete_celltype,"maxexpr")
dev.off()

pdf(file.path(plots, "02/SUPPL_maxB.pdf"), width = 2, height = 4)
get_expression_plot("B", df_complete_celltype, "maxexpr")
dev.off()

pdf(file.path(plots, "02/SUPPL_maxT.pdf"), width = 2, height = 4)
get_expression_plot("T", df_complete_celltype, "maxexpr")
dev.off()


pdf(file.path(plots, "02/SUPPL_maxNeutrophil.pdf"), width = 2, height = 4)
get_expression_plot("Neutrophil", df_complete_celltype,"maxexpr")
dev.off()

get_cells_plot <- function(ident,df_complete_celltype){
  df_ident <- df_complete_celltype[df_complete_celltype$ident == ident, ]
  df_ident <- df_ident[!is.na(df_ident$perc_cells_expressing),]
  my_comparisons <- list( c("lncRNA", "mRNA"))
  p <- ggboxplot(df_ident, x = "type", y = "perc_cells_expressing", fill = "type", col = "type",alpha=1)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("% cells")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,1.1), expand = c(0,0))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons)+ggtitle(ident)+theme(plot.title = element_text(hjust = 0.5))
  return(p)
}


pdf(file.path(plots, "02/SUPPL_cellsMono.pdf"), width = 2, height = 4)
get_cells_plot("Monocyte", df_complete_celltype)
dev.off()

pdf(file.path(plots, "02/SUPPL_cellsB.pdf"), width = 2, height = 4)
get_cells_plot("B", df_complete_celltype)
dev.off()

pdf(file.path(plots, "02/SUPPL_cellsT.pdf"), width = 2, height = 4)
get_cells_plot("T", df_complete_celltype)
dev.off()

pdf(file.path(plots, "02/SUPPL_cellsNeutrophil.pdf"), width = 2, height = 4)
get_cells_plot("Neutrophil", df_complete_celltype)
dev.off()

# ------------------ GET all n for plots 
# Only retain genes considered for the plot 
df_complete_celltype
df_complete_celltype <- df_complete_celltype[!is.na(df_complete_celltype$medianexpr),]
df_complete_celltype <- df_complete_celltype[!is.na(df_complete_celltype$perc_cells_expressing),]

df_complete_celltype_red <- df_complete_celltype[,c("")]

df_complete_celltype %>% dplyr::group_by(ident) %>% dplyr::count(type)

```


# ------------------------------
#     Match median expression 
# ------------------------------

```{r prepForMatching}
set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
head(c)
```


```{r prepForMatching}
# ------------------------
# 1. Match expression 
# ------------------------
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, data = c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1, alternative = "greater")
wilcox.test(perc_cells_0, perc_cells_1, paired = T)

matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
matches_median$type <- factor(matches_median$type, levels = c("lnc", "mrna"))

table(matches_median$type)
levels(matches_median$type)


mi_max <- matchit(type ~ maxexpr,c)
matches_max <- get_matches(mi_max, data = c)
matches_max$type = ifelse(matches_max$type == 0, "mrna", "lnc")
matches_max$type <- as.factor(matches_max$type)
matches_max$type <- factor(matches_max$type, levels = c("lnc", "mrna"))

```



```{r prepForMatching}
# ------------------------------
#  2. Match perc cells
# ------------------------------
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, data = c)

matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
matches_perc$type <- factor(matches_perc$type, levels = c("lnc", "mrna"))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1,alternative = "less" )
wilcox.test(median_0, median_1, paired = T, alternative = "two.sided" )
wilcox.test(median_0, median_1, paired = T, alternative = "less" )

```

# Visualize
```{r matchperc}
plot_matching_scatter <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "median expression (logCP10K)", y = "% cells")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =medianexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  #+stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}

pdf(file.path(plots, "02/matches_perc_medianplot.pdf"), width = 6, height = 6)
plot_matching_scatter(matches_perc)
dev.off()

pdf(file.path(plots, "02/matches_median.pdf"), width = 6, height = 6)
plot_matching_scatter(matches_median)
dev.off()




pdf(file.path(plots, "02/MedianExpression_matchedByCells.pdf"), width = 4, height = 6)
ggboxplot(matches_perc, x = "type", y = "medianexpr",
   color = "type", fill = "type", alpha = 0.8,
   palette = palette_plot_percentage)+xlab("")+ylab("median expression")+theme(text = element_text(size = 20))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank(), axis.text.x =  element_blank() )+theme(legend.position = "")
dev.off()

pdf(file.path(plots, "02/PercCells_matchedByExpr.pdf"), width = 4, height = 6)
ggboxplot(matches_median, x = "type", y = "perc_cells_expressing",
   color = "type", fill = "type", alpha = 0.8, 
   palette = palette_plot_percentage)+xlab("")+ylab("% cells")+theme(text = element_text(size = 20))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank(), axis.text.x =  element_blank() )+scale_y_continuous(limits = c(0,1), expand = c(0,0))+theme(legend.position = "")
dev.off()




```



```{r matchperc}
plot_matching_scatter_max <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=maxexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "maximum expression (logCP10K)", y = "% cells")

  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =maxexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  #+stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}

pdf(file.path(plots, "02/matches_perc_max.pdf"), width = 6, height = 6)
plot_matching_scatter_max(matches_perc)
dev.off()

pdf(file.path(plots, "02/matches_max.pdf"), width = 6, height = 6)
plot_matching_scatter_max(matches_max)
dev.off()

```

# ------------------------------------------------------
# Pick an example 
# highly expressed in low proportion of cells 
# ------------------------------------------------------

```{r cellt type specificity}
candidates  <- df_lnc[ df_lnc$medianexpr>1.5 & df_lnc$n_cells > 50& df_lnc$n_cells < 1200,]
gene <- candidates$gene_id[11]
print(gene)
gene2 <- "NUDT9"
# ----------------------------------------------
# 1 . Find the matching (median expr) Gene
# ----------------------------------------------
df_lnc_filtered <- df_lnc[df_lnc$gene_id == gene,]
df_complete<- rbind(df_lnc_filtered, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
set.seed(125)
mi <- matchit(type ~ medianexpr,c)
matches <- get_matches(mi, data = c)
rownames(matches) <- gsub("-unknown", "", matches$gene_id)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)
matches[,c("medianexpr", "n_cells", "type")]

# Check the 2 genes 
genes <-  c("ENSMMUG00000045507-unknown", "NUDT9")
df_complete[df_complete$gene_id %in% genes,]
# ----------------------------------------------
#  2. Visualize that the expression of the 2 genes BOXPLOT
# ----------------------------------------------
expr_matrix <- immune.combined@assays$RNA@data
df_genes <- data.frame(t(expr_matrix[genes, ]))
colnames(df_genes) <-  gsub(".unknown", "", colnames(df_genes))
df_genes_melted <- melt(df_genes)
df_genes_melted <- df_genes_melted[df_genes_melted$value > 0, ]
bp <- ggboxplot(df_genes_melted, x = "variable", y = "value", fill = "variable", col = "black",alpha=1)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = rev(palette_plot_percentage))+stat_compare_means(comparisons = list(c(colnames(df_genes))))+xlab("")+ylab("logCP10K")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,5), expand = c(0,0))+theme(axis.line.x = element_blank(), legend.position = "", axis.ticks.x = element_blank()) + theme(axis.text.x = element_blank())

pdf(file.path(plots, "02/genes_boxplot.pdf"), width = 4, height = 6)
bp
dev.off()
# ----------------------------------------------
#  2. Visualize that the expression of the 2 genes FEATURE PLOT
# ----------------------------------------------
f1 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[2]), pt.size = 0.8, order = TRUE,max.cutoff = 3.5,  cols = c("lightgrey", "#0D3A75"))+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = genes[2])+theme_void()+theme(title = element_blank(), text = element_text(size = 20))+ scale_color_gradientn( colours = c('lightgrey', '#0D3A75'),  limits = c(-0.1, 3.5))
f1
f2 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[1]), pt.size = 0.8, order = TRUE, max.cutoff = 3.5,cols = c("lightgrey", "#A91818"))+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = gsub("-unknown", "", genes[1]))+theme_void()+theme(title = element_blank(), text = element_text(size = 20))+ scale_color_gradientn( colours = c('lightgrey', '#A91818'),  limits = c(-0.1, 3.5))
f2
CombinePlots(list(f1,f2))

pdf(file.path(plots, "02/gene_1.pdf"), width = 6, height = 6)
f1
dev.off()


pdf(file.path(plots, "02/gene_2.pdf"), width = 6, height = 6)
f2
dev.off()
```


# ------------------------------
#     4. Validate: Match Median per celltype 
# ------------------------------
```{r matchmedian}

set.seed(123)

# Filter
# Select the ones expressed at least in 50 cells 
df_complete_celltype<- df_complete_celltype[df_complete_celltype$n_cells >= 30, ]
df_complete_celltype$type <- gsub("mRNA", "protein coding", df_complete_celltype$type)

c <- df_complete_celltype[complete.cases(df_complete_celltype),]
c$type <-  ifelse(c$type == "lncRNA", 1, 0)


# ---------------new 
match_in_one_celltype_median_boxplotonly <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_median <- MatchIt::matchit(type ~ medianexpr, data = c)
  matches_median <-df_ident <-  get_matches(mi_median, data = c)

  df_ident <- df_ident[!is.na(df_ident$perc_cells_expressing),]
  df_ident$type <- ifelse(df_ident$type == 1, "lncRNA", "protein coding" )
  df_ident$type <- factor(df_ident$type, levels = c("lncRNA", "protein coding"))

  my_comparisons <- list( c("lncRNA", "protein coding"))
  p <- ggboxplot(df_ident, x = "type", y = "perc_cells_expressing", fill = "type", col = "type",alpha=0.7)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("% cells")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,1.1), expand = c(0,0))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons)+ggtitle(celltype_test)+theme(plot.title = element_text(hjust = 0.5))+theme(axis.text.x.bottom =element_blank() )
  return(p)
}


pdf(file.path(plots, "02/SUPPL_cellsMono_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_median_boxplotonly(c, "Monocyte")
dev.off()

pdf(file.path(plots, "02/SUPPL_cellsB_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_median_boxplotonly(c, "B")
dev.off()

pdf(file.path(plots, "02/SUPPL_cellsT_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_median_boxplotonly(c, "T")
dev.off()

pdf(file.path(plots, "02/SUPPL_cellsNeutrophil_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_median_boxplotonly(c, "Neutrophil")
dev.off()

# --------------------- now with % cells 
match_in_one_celltype_cells_boxplotonly <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_perc <- matchit(type ~ perc_cells_expressing, data = c)
  matches_perc <-df_ident <- get_matches(mi_perc, data = c)

  df_ident <- df_ident[!is.na(df_ident$perc_cells_expressing),]
  df_ident$type <- ifelse(df_ident$type == 1, "lncRNA", "protein coding" )
  df_ident$type <- factor(df_ident$type, levels = c("lncRNA", "protein coding"))

  my_comparisons <- list( c("lncRNA", "protein coding"))
  p <- ggboxplot(df_ident, x = "type", y = "perc_cells_expressing", fill = "type", col = "type",alpha=0.7)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("% cells")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,1.1), expand = c(0,0))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+ggtitle(celltype_test)+theme(plot.title = element_text(hjust = 0.5))+theme(axis.text.x.bottom =element_blank() )
  return(p)
}

pdf(file.path(plots, "02/SUPPL_medianMono_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_cells_boxplotonly(c, "Monocyte")
dev.off()

pdf(file.path(plots, "02/SUPPL_medianB_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_cells_boxplotonly(c, "B")
dev.off()

pdf(file.path(plots, "02/SUPPL_medianT_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_cells_boxplotonly(c, "T")
dev.off()

pdf(file.path(plots, "02/SUPPL_medianNeutrophil_matched_boxplot.pdf"), width = 2, height = 4)
match_in_one_celltype_cells_boxplotonly(c, "Neutrophil")
dev.off()



```

