---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---


# Replicate general expression patterns 
# General Expression Patterns of lncRNAs Ex Vivo


```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(MatchIt)
library(rtracklayer)
library(Seurat)

# Palette used throughout the script
# 0. Palette used throughout the scripts
col_lnc = "#f72631"
col_mrna = "#3153a3"
palette_plot_percentage <- c(col_lnc, col_mrna)

source("../utils/00_datapaths.R")
source("../utils/02_sc_utils.R")

# Read reference files 
ref<- import(file.path(gene_annotation_path,"/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))
table(ref$gene_biotype)
lnc <- ref[ref$gene_biotype == "lncRNA",]$gene_name
length(unique(lnc))
pc <- ref[ref$gene_biotype == "protein_coding",]$gene_name

# Import Seurat object 
immune.combined <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/01_immunecombined_idents.rds"))
annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc]
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% pc]

theme_matching <- theme(panel.background = element_rect(fill = "white", color = "black"))+
                  theme(legend.position ="",panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.ticks.length=unit(.2, "cm"))+
                  theme(axis.text = element_text(size = 15, color ="black"), axis.title = element_text(size = 17))

```

# Visualize how many lncRNAs and PC genes are present in the object 

```{r check}
#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)) ))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Greys")
```


# Calculate expression 

```{r quartiles}
# Calculate expression POOLED
# (Big, can be run on cluster)
df_lnc <- readRDS( file.path(data_path,"/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/df_lnc.rds"))
df_mrna <-readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/df_mrna.rds"))
```



#----------------------------------
# 1. Visualzation: Expression across Quartiles
# ----------------------------------
```{r quartiles}
# Only select genes expressed in at least 60 cells
df_lnc <- df_lnc[df_lnc$n_cells > 60, ]
df_mrna <- df_mrna[df_mrna$n_cells > 60, ]

df_lnc$type <- "lncRNA"
df_mrna$type <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)



#  ------- 1 :  PERCENTAGE OF CELLS EXPRESSING A GENE -----------
# 1a. Density plot of % cells expressing 
density_cells <- ggdensity(df_complete, x = "perc_cells_expressing",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("% cells")+ylab("density")+scale_x_log10()+theme(text = element_text(size = 20))

# Save plot 
pdf(file.path(plots, "02/replication33k/density_cells.pdf"), width = 6, height = 6)
density_cells
dev.off()

# WILCOX test on complete distributions
wilcox.test(df_lnc$perc_cells_expressing, df_mrna$perc_cells_expressing, paired = F)
wilcox.test(df_lnc$medianexpr, df_mrna$medianexpr, paired = F, alternative = "two.sided")
wilcox.test(df_lnc$maxexpr, df_mrna$maxexpr, paired = F)
```


```{r prepForMatching}
set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
head(c)
```

```{r prepForMatching}
# ------------------------
# 1. Match expression 
# ------------------------
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, data = c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
wilcox.test(perc_cells_0, perc_cells_1, paired = T)

matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
matches_median$type <- factor(matches_median$type, levels = c("lnc", "mrna"))

table(matches_median$type)
levels(matches_median$type)


mi_max <- matchit(type ~ maxexpr,c)
matches_max <- get_matches(mi_max, data = c)
matches_max$type = ifelse(matches_max$type == 0, "mrna", "lnc")
matches_max$type <- as.factor(matches_max$type)
matches_max$type <- factor(matches_max$type, levels = c("lnc", "mrna"))

```




```{r prepForMatching}
# ------------------------------
#  2. Match perc cells
# ------------------------------
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, data = c)

matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
matches_perc$type <- factor(matches_perc$type, levels = c("lnc", "mrna"))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1,alternative = "less" )
wilcox.test(median_0, median_1, paired = T)

matches_perc$type <- gsub("lnc", "lncRNA", matches_perc$type )
matches_perc$type <- gsub("mrna", "mRNA", matches_perc$type )
```

# MATCHING MEDIAN 
```{r matchperc}
plot_matching_scatter <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "median expression (logCP10K)", y = "% cells")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =medianexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  #+stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}

pdf(file.path(plots, "02/replication33k/matches_perc_medianplot.pdf"), width = 6, height = 6)
plot_matching_scatter(matches_perc)
dev.off()

pdf(file.path(plots, "02/replication33k/matches_median.pdf"), width = 6, height = 6)
plot_matching_scatter(matches_median)
dev.off()

matches_median$type <- gsub("lnc", "lncRNA", matches_median$type )
matches_median$type <- gsub("mrna", "mRNA", matches_median$type )

# Only visualize boxplots: matched by median
pdf(file.path(plots, "02/replication33k/matches_median_boxplot.pdf"), width = 3, height = 4)

my_comparisons <- list( c("lncRNA", "mRNA"))
ggboxplot(matches_median, x = "type", y = "perc_cells_expressing", fill = "type", col = "type",alpha=1)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("% cells")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,1.1), expand = c(0,0))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+theme(plot.title = element_text(hjust = 0.5))
dev.off()


pdf(file.path(plots, "02/replication33k/matches_cells_boxplot.pdf"), width = 3, height = 4)
my_comparisons <- list( c("lncRNA", "mRNA"))
ggboxplot(matches_perc, x = "type", y = "medianexpr", fill = "type", col = "type",alpha=1, )+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("median expression (logCP10K)")+theme(text = element_text(size=15))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+theme(plot.title = element_text(hjust = 0.5))
dev.off()



```




```{r matchperc}
plot_matching_scatter_max <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=maxexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "maximum expression (logCP10K)", y = "% cells")

  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =maxexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  #+stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}

pdf(file.path(plots, "02/replication33k/matches_perc_max.pdf"), width = 6, height = 6)
plot_matching_scatter_max(matches_perc)
dev.off()

pdf(file.path(plots, "02/replication33k/matches_max.pdf"), width = 6, height = 6)
plot_matching_scatter_max(matches_max)
dev.off()

matches_max$type <- gsub("lnc", "lncRNA", matches_max$type )
matches_max$type <- gsub("mrna", "mRNA", matches_max$type )

# Only visualize boxplots: matched by median
pdf(file.path(plots, "02/replication33k/matches_max_boxplot.pdf"), width = 3, height = 4)

my_comparisons <- list( c("lncRNA", "mRNA"))
ggboxplot(matches_max, x = "type", y = "perc_cells_expressing", fill = "type", col = "type",alpha=1)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("% cells")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,1.1), expand = c(0,0))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+theme(plot.title = element_text(hjust = 0.5))
dev.off()


pdf(file.path(plots, "02/replication33k/matches_maxcells_boxplot.pdf"), width = 3, height = 4)
# Only visualize boxplots: matched by %cells
my_comparisons <- list( c("lncRNA", "mRNA"))
ggboxplot(matches_max, x = "type", y = "maxexpr", fill = "type", col = "type",alpha=0.7)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("maximum expression (logCP10K)")+theme(text = element_text(size=15))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+theme(plot.title = element_text(hjust = 0.5))
dev.off()


```

# --------------old

```{r prepForMatching}
# ---------------- 1:  general inspection 
# Scatter plot MEDIAN
pdf(file.path(plots, "02/replication33k/SUPPL_scatterplot.pdf"), width = 6, height = 6)
scatter_plot_both(df_lnc,df_mrna, y = "% cells", type = "median",palette = palette_plot_percentage)
dev.off()


# Scatter plot MAX
pdf(file.path(plots, "02/replication33k/SUPPL_scatterplot_max.pdf"), width = 6, height = 6)
scatter_plot_both(df_lnc,df_mrna, y = "% cells", type = "max",palette = palette_plot_percentage)
dev.off()
```


```{r prepForMatching}
set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
table(c$type)

# ------------------------
# 1. Match expression 
# ------------------------
mi_median <-MatchIt::matchit(type ~ medianexpr,c)
matches_median <- MatchIt::get_matches(mi_median, data = c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1)


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
matches_median$type <- factor(matches_median$type, levels = c("lnc", "mrna"))

table(matches_median$type)
levels(matches_median$type)

# ------------------------------
#  2. Match perc cells
# ------------------------------

# matched perc cells
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, data = c)

matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
matches_perc$type <- factor(matches_perc$type, levels = c("lnc", "mrna"))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1)

```

# Visualize
```{r matchperc}

theme_matching <- theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(legend.position ="",legend.title = element_blank())+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 13), legend.text = element_text(size = 13))


plot_matching_scatter <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.7)+
                            labs( subtitle = paste0("# Matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "Median Expression", y = "% Cells expressing")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =medianexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)+ stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}


plot_matching_scatter(matches_perc)
plot_matching_scatter(matches_median)

```
