---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# ----------------------------------------------------
# General Expression Patterns of lncRNAs IN VIVO
# ----------------------------------------------------


### Imports 
```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(MatchIt)
library(rtracklayer)
library(Seurat)
source("../utils/02_sc_utils.R")

# Palette used throughout the script
col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

theme_sc_paper <- theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())+theme_paper


# Where stats about lnc and pc computed on cluster are stored
robjectsdir<- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"


# Read reference files 
ref<- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- rtracklayer::import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Import Seurat object 
file <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"
immune.combined <- readRDS(file)
```


# -------------- Check genes in object-------------------------
# Visualize how many lnc and mrnas are present in the object 

```{r check}
# ------------------------------------
#     PC genes present in object 
# -------------------------------------
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% unique(ref$gene_name[ref$gene_biotype == "protein_coding"])]

# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# Annotated lncrnas id 
lnc_ref_id <- gsub("_","-",unique(ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "lncRNA",]$gene_name))

# Annotated lncRNAs: remove the gene names so it can match 
annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc_ref_id]
# Novel lncRNAs
novel_lncrnas <- rownames(immune.combined)[unlist(lapply(rownames(immune.combined), function(x) startsWith(x, "MSTRG")))]
# Complete Set 
all_lncrnas <- c(annotated_lncrnas, novel_lncrnas)

# Save
#saveRDS(all_lncrnas, file.path(robjectsdir, "all_lncrnas.rds"))
#saveRDS(annotated_mrnas, file.path(robjectsdir, "annotated_mrnas.rds"))

#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated", "novel")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)), length(unique(novel_lncrnas))))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Dark2")
lnc_donut <- df.donut[df.donut$type == "lncRNAs",]
names(lnc_donut) <- c("-", "type", "count")
donut_plot(lnc_donut, "Greens")
```


# ---------------------------------------------------
# Play w/ pca
# ----------------------------------------------------

```{r check}

# Plot with correct
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(4, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

immune.combined_lnc <- subset(immune.combined, features = all_lncrnas)
DefaultAssay(immune.combined_lnc) <- "RNA"

immune.combined_lnc <- NormalizeData(immune.combined_lnc)
immune.combined_lnc <- FindVariableFeatures(immune.combined_lnc)
immune.combined_lnc <- ScaleData(immune.combined_lnc)
immune.combined_lnc <- RunPCA(immune.combined_lnc, npcs = 30, verbose = FALSE)
immune.combined_lnc <- RunUMAP(immune.combined_lnc, reduction = "pca", dims = 1:20)
immune.combined_lnc <- FindNeighbors(immune.combined_lnc, reduction = "pca",dims = 1:20 )

immune.combined_lnc$id <- Idents(immune.combined)
DimPlot(immune.combined_lnc, reduction = "umap", group.by = "id")


# UMAP dimension 2 separates the different cell-types 
```


# ---------------------------------------------------------
#  Load Metrics: ACROSS cell types
# (#cells in which the gene is expressed, median, etc) 
# ---------------------------------------------------------
```{r quartiles}
df_lnc <- readRDS(file.path(robjectsdir, "df_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))


df_lnc <- df_lnc[df_lnc$n_cells > 30, ]
df_mrna <- df_mrna[df_mrna$n_cells > 30, ]


# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNAs"
df_mrna$type <- "mRNAs"
df_complete <- rbind(df_lnc, df_mrna)

```

# ---------------------------------------------------------
#  Load Metrics: per cell type individually
# (#cells in which the gene is expressed, median, etc) 
# ---------------------------------------------------------
```{r celltype}
df_lnc_celltype <- readRDS(file.path(robjectsdir, "df_celltype_lnc.rds"))
df_mrna_celltype <- readRDS(file.path(robjectsdir, "df_mrna_celltype.rds"))

df_lnc_celltype <- df_lnc_celltype[df_lnc_celltype$n_cells > 10, ]
df_mrna_celltype <- df_mrna_celltype[df_mrna_celltype$n_cells > 10, ]

df_mrna_celltype$type <- "mRNAs"
df_lnc_celltype$type <-  "lncRNAs"

df_complete_celltype <- rbind(df_lnc_celltype, df_mrna_celltype)
```



# ---------------------------------------------------------
# 1. Visualzation: # Cells  across Quartiles and Cell-types 
# ---------------------------------------------------------
```{r quartiles}

#plot_quartiles(df_lnc, df_mrna, "number", palette = palette_plot_percentage )+theme_sc_paper
plot_quartiles(df_lnc, df_mrna, "proportion", palette = palette_plot_percentage)+theme_sc_paper
plot_quartiles(df_lnc, df_mrna, "median", palette = palette_plot_percentage)+theme_sc_paper

plot_quartiles(df_lnc, df_mrna, "max", palette = palette_plot_percentage)+theme_sc_paper


ggplot(df_complete_celltype, aes( x= type, y = perc_cells_expressing, fill = type, col = type))+geom_boxplot(alpha= 0.4)+scale_fill_manual(values = palette_plot_percentage)+scale_color_manual(values = palette_plot_percentage)+theme_sc_paper+xlab ("")+ylab("Proportion of cells")+ facet_wrap(vars(ident))


ggplot(df_complete, aes(x = perc_cells_expressing, col = type)) + stat_ecdf(geom = "step")+scale_color_manual(values = palette_plot_percentage)+theme_sc_paper+ylab("")+xlab("Proportion of cells")

ggboxplot(df_complete, y="medianexpr", x = "type", fill = "type", color = "type", alpha = 0.5)+
                scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)+
                stat_compare_means(comparisons = list(c("lncRNAs", "mRNAs")))+
                labs(x = "", y = "Median expression")+
                theme(legend.title = element_blank(), legend.position = "none")


```




# ------------------------------------------
# Calculate the EXPRESSION
# ------------------------------------------

```{r percentage ebov reads}
# Scatter plot both 
scatter_plot_both(df_lnc,df_mrna, y = "Cell Percentage", type = "Median",palette = palette_plot_percentage)

```


# ------------------------------
#     Match median expression 
# ------------------------------

```{r matchmedian}

set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# match expression 
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1)


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)


table(matches_median$type)
c
```

# Visualize matching

```{r matchmedianplot}

scatter_plot_matching <- ggplot(matches_median, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.5)+labs( subtitle = paste0("# Matches ", n_matches))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
          scale_color_manual(values = palette_plot_percentage)+theme(legend.position = "none")+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank(), legend.position = "")+theme(legend.title = element_blank(), legend.position = "")+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))


scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot

```




# --------------------------------------
#      4. Validate : run it w/ multiple seeds
# --------------------------------------
```{r matchvalidation}
run_matchit_median <- function(c, seed){
    set.seed(seed)
    mi_median <- matchit(type ~ medianexpr,c)
    matches_median <- get_matches(mi_median, c)
    matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
    n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
    matches_median$type <- as.factor(matches_median$type)
    output <- matches_median[matches_median$type  == "mrna",]$perc_cells_expressing
    df <- as.data.frame(output)
    df$seed <- as.character(seed)
    return(df)
}


#seeds <- 1:10
#list <- do.call(rbind,(lapply( seeds, function(seed) run_matchit_median(c,seed))))
#lncRNAs <- data.frame(output = c[c$type == 1,]$perc_cells_expressing, seed= as.character("0-lncRNAs"))
#lncRNAs$seed <- as.character(lncRNAs$seed)
#benchmark_df_matched_median <- rbind(lncRNAs, list)
#table(benchmark_df_matched_median$seed)

#ggplot(benchmark_df_matched_median, aes (y = output, x = seed, fill = seed, color = seed))+geom_boxplot(alpha = 0.6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+scale_fill_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+scale_color_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+labs(y = "% Cells", Title = "")+theme(legend.position = "")

```


# ------------------------------
#     Match perc cells
# ------------------------------

```{r matchperc}
# matched perc cells
set.seed(123)
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, c)
matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1)

```

# Visualize
```{r matchperc}

scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+
                         geom_point(size = 0.8, alpha = 0.8)+labs(title = "", subtitle = paste0("# Matches ", n_matches))+
                         theme(legend.title = element_blank(),
                               legend.position = "none",
                               panel.background = element_rect(fill = "white", colour = "grey50"),
                               axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))+
                         scale_fill_manual(values =palette_plot_percentage)+
                         scale_color_manual(values = palette_plot_percentage)+
                         labs(x = "Median Expression", y = "% Cells expressing")+theme()

scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.5, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot


```


# ------------------------------
#     4. Validate: Match Median per celltype 
# ------------------------------
```{r matchmedian}

set.seed(123)
df_lnc_celltype$type <- "lnc"
df_mrna_celltype$type <- "mrna"
df_complete<- rbind(df_lnc_celltype, df_mrna_celltype)

# Filter
# Select the ones expressed at least in 50 cells 
df_complete <- df_complete[df_complete$n_cells >= 50, ]

c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# -------------------------------
#   Select Only one celltype 
# -------------------------------
match_in_one_celltype_median <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_median <- matchit(type ~ medianexpr,c)
  matches_median <- get_matches(mi_median, c)
  
  # Calculate p-value 
  perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
  perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
  test <- wilcox.test(perc_cells_0, perc_cells_1, paired = TRUE)
  pval <- test$p.value
  
  matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
  matches_median$type <- as.factor(matches_median$type)
  
  title = paste0(celltype_test)
  scatter_plot_matching <- ggplot(matches_median, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( title=title, subtitle = paste0("n: ", n_matches, "\n", "p-value: ", pval))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))
  
  
  scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
  scatter_plot_matching_withboxplot

}

match_in_one_celltype_perccells <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  
  # match expression 
  mi_perc <- matchit(type ~ perc_cells_expressing,c)
  matches_perc <- get_matches(mi_perc, c)
  matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
  
  median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
  median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
  test <- wilcox.test(median_0, median_1, paired = TRUE)
  pval <- test$p.value

  
  title = paste0(celltype_test)
  scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, fill = factor(type, level = c("lnc", "mrna")), col = factor(type, level = c("lnc", "mrna"))))+geom_point(size = 0.8)+labs( title=title, subtitle = paste0("n: ", n_matches, "\n", "p-value: ", pval))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))
  
  
  scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot', margins = "both",size = 3, alpha = 0.3, groupColour = TRUE, groupFill = TRUE, margMapping = aes(colour = factor(type, level = c("lnc", "mrna"))))
  scatter_plot_matching_withboxplot

}


match_in_one_celltype_median(c, "Monocyte")
match_in_one_celltype_median(c, "T")
match_in_one_celltype_median(c, "B")

match_in_one_celltype_perccells(c, "B")
match_in_one_celltype_perccells(c, "Monocyte")
match_in_one_celltype_perccells(c, "T")


match_in_one_celltype_perccells_get_median <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  
  # match expression 
  mi_perc <- matchit(type ~ perc_cells_expressing,c)
  matches_perc <- get_matches(mi_perc, c)
  matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
  
  median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
  median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
  test <- wilcox.test(median_0, median_1, paired = TRUE)
  pval <- test$p.value
  
  df_0 <- data.frame(val = median_0, type = "lnc", celltype = celltype_test, stringsAsFactors = F)
  df_1 <- data.frame(val = median_1, type = "mrna", celltype = celltype_test, stringsAsFactors = F)
  return(rbind(df_0, df_1))
}

df_medians_matched <- do.call(rbind, lapply(unique(Idents(immune.combined)), function(celltype) match_in_one_celltype_perccells_get_median(c, celltype)))


ggplot(df_medians_matched, aes(x = celltype,col = type,  fill = type, y = val))+geom_boxplot(alpha = 0.5)+theme_paper+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+ylab("Median Expression")+ggtitle("Matched by the percentage of cells expressing")




match_in_one_celltype_median_get_perc <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_median <- matchit(type ~ medianexpr,c)
  matches_median <- get_matches(mi_median, c)
  matches_perc$type = ifelse(matches_perc$type == 1, "lnc", "mrna")

  # Calculate p-value 
  perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
  perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
  test <- wilcox.test(perc_cells_0, perc_cells_1, paired = TRUE)
  pval <- test$p.value
  
  df_0 <- data.frame(val = perc_cells_0, type = "mrna", celltype = celltype_test, stringsAsFactors = F)
  df_1 <- data.frame(val = perc_cells_1, type = "lnc", celltype = celltype_test, stringsAsFactors = F)
  return(rbind(df_0, df_1))
}



df_perc_matched <- do.call(rbind, lapply(unique(Idents(immune.combined)), function(celltype) match_in_one_celltype_median_get_perc(c, celltype)))


ggplot(df_perc_matched, aes(x = celltype,col = type,  fill = type, y = val))+geom_boxplot(alpha = 0.5)+theme_paper+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+ylab("Percentage of cells expressing")+ggtitle("Matched by the median")
```



# Validate


```{r matchperc}
run_matchit_perc<- function(c, seed){
    set.seed(seed)
    mi_perc <- matchit(type ~ perc_cells_expressing,c)
    matches_perc <- get_matches(mi_perc, c)
    matches_perc$type = ifelse(matches_perc$type == 0, "lnc", "mrna")
    matches_perc$type <- as.factor(matches_perc$type)
    output <- matches_perc[matches_perc$type  == "mrna",]$medianexpr
    df <- as.data.frame(output)
    df$seed <- as.character(seed)
    return(df)
}

#seeds <- 1:10
#list <- do.call(rbind,(lapply( seeds, function(seed) run_matchit_perc(c,seed))))
#lncRNAs <- data.frame(output = c[c$type == 1,]$medianexpr, seed= as.character("0-lncRNAs"))
#lncRNAs$seed <- as.character(lncRNAs$seed)
#benchmark_df_matched_median <- rbind(lncRNAs, list)
#table(benchmark_df_matched_median$seed)
#ggplot(benchmark_df_matched_median, aes (y = output, x = seed, fill = seed, color = seed))+geom_boxplot(alpha = 0.6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+scale_fill_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+scale_color_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+labs(y = "Median", Title = "")+theme(legend.position = "")
```

# ------------------------------------------------------
# Cherry pick an example 
# highly expressed in low proportion of cells 
# ------------------------------------------------------

```{r cellt type specificity}
a <- df_lnc[df_lnc$perc_cells_expressing < 0.1 & df_lnc$medianexpr>1.5 & df_lnc$n_cells > 50,]
gene <- a$gene_id[1]

# Find the matching (median expr) Gene 
df_lnc_filtered <- df_lnc[df_lnc$gene_id == gene,]
df_complete<- rbind(df_lnc_filtered, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
set.seed(125)
mi <- matchit(type ~ medianexpr,c)
matches <- get_matches(mi, c)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)
matches[,c("medianexpr", "n_cells", "type")]
rownames(matches) <- gsub("-unkown", "", matches$gene_id)

# Visualize that the expression 
expr_matrix <- immune.combined@assays$RNA@data
df_genes <- data.frame(t(expr_matrix[genes, ]))
df_genes_melted <- melt(df_genes)
df_genes_melted <- df_genes_melted[df_genes_melted$value > 0, ]
ggplot(df_genes_melted, aes(x = variable, y = value, fill = variable, col = variable))+geom_boxplot(alpha=0.5)+theme_sc_paper+theme(legend.position = "")+scale_fill_manual(values = rev(palette_plot_percentage))+scale_color_manual(values = rev(palette_plot_percentage))



# Visualize the 
f1 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[1]), pt.size = 0.6, order = TRUE, cols = c("lightgrey", "#bf2e2e"), max.cutoff = 3.5)+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = genes[1])

f2 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[2]), pt.size = 0.6, order = TRUE, cols = c("lightgrey", "#bf2e2e"), max.cutoff = 3.5)+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = gsub("-unkown", "", genes[2]))


f1
f2
CombinePlots(list(f1,f2))
```


# ------------------------------------------------------
# Cherry pick an example the other way around 
# highly expressed in low proportion of cells 
# ------------------------------------------------------

```{r cellt type specificity}
a <- df_lnc[df_lnc$perc_cells_expressing < 0.1 & df_lnc$medianexpr>1.5 & df_lnc$n_cells > 50,]
gene <- a$gene_id[1]

# Find the matching (median expr) Gene 
df_lnc_filtered <- df_lnc[df_lnc$gene_id == gene,]
df_complete<- rbind(df_lnc_filtered, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
table(c$type)
set.seed(125)
mi <- matchit(type ~ perc_cells_expressing,c)
matches <- get_matches(mi, c)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)
matches[,c("medianexpr", "n_cells", "type")]
rownames(matches) <- gsub("-unkown", "", matches$gene_id)

# Visualize that the expression 
expr_matrix <- immune.combined@assays$RNA@data
df_genes <- data.frame(t(expr_matrix[genes, ]))
df_genes_melted <- melt(df_genes)
df_genes_melted <- df_genes_melted[df_genes_melted$value > 0, ]
ggplot(df_genes_melted, aes(x = variable, y = value, fill = variable, col = variable))+geom_boxplot(alpha=0.5)+theme_sc_paper+theme(legend.position = "")+scale_fill_manual(values = rev(palette_plot_percentage))+scale_color_manual(values = rev(palette_plot_percentage))



# Visualize the 
f1 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[1]), pt.size = 0.6, order = TRUE, cols = c("lightgrey", "#bf2e2e"), max.cutoff = 3.5)+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = genes[1])

f2 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[2]), pt.size = 0.6, order = TRUE, cols = c("lightgrey", "#bf2e2e"), max.cutoff = 3.5)+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = gsub("-unkown", "", genes[2]))


f1
f2
CombinePlots(list(f1,f2))


df_complete[df_complete$gene_id == as.character(gene),]

test <- df_complete[df_complete$perc_cells_expressing > 0.01 & df_complete$perc_cells_expressing < 0.08,]
ggplot(test, aes( x= medianexpr))+geom_density()

table(df_complete$type)

ggplot(test, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.3)+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))


```


