---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---


# Replicate general expression patterns 
# General Expression Patterns of lncRNAs Ex Vivo


```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(MatchIt)
library(rtracklayer)
library(Seurat)

# Palette used throughout the script
col_lnc = "navy"
col_mrna = "#8DC3A7"
palette_plot_percentage <- c(col_lnc, col_mrna)

source("../utils/02_sc_utils.R")

# Read reference files 
ref<- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))
table(ref$gene_biotype)
lnc <- ref[ref$gene_biotype == "lncRNA",]$gene_name
length(unique(lnc))
pc <- ref[ref$gene_biotype == "protein_coding",]$gene_name

# Import Seurat object 
data <- Read10X("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/filtered_gene_bc_matrices/hg19/")
immune.combined = CreateSeuratObject(counts = data)
dim(immune.combined)
immune.combined <- NormalizeData(immune.combined)
annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc]
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% pc]
```

# Visualize how many lncRNAs and PC genes are present in the object 

```{r check}
#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)) ))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Greys")
```






# Calculate expression 

```{r quartiles}
# Calculate expression POOLED
# (Big, can be run on cluster)
df_lnc <- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(), tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = annotated_lncrnas), "",df_lnc, 1))
df_mrna<- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(),tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = annotated_mrnas), "",df_mrna, 1))
saveRDS(df_lnc, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/df_lnc.rds")
saveRDS(df_mrna, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/df_mrna.rds")
#saveRDS(df_pseudogenes, "/home/luisas/Desktop/cluster/data/RObjects_old/df_pseudogenes.rds")
#df_lnc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/df_lnc.rds")
#df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))
```



#----------------------------------
# 1. Visualzation: Expression across Quartiles
# ----------------------------------
```{r quartiles}
# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNAs"
df_mrna$type <- "mRNAs"
df_lnc <- df_lnc[df_lnc$n_cells > 30, ]
df_mrna <- df_mrna[df_mrna$n_cells > 30, ]

# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNA"
df_mrna$type <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)

df_complete <- rbind(df_lnc, df_mrna)
# Figure 3A (Density plot of % cells expressing )
ggdensity(df_complete, x = "perc_cells_expressing",
   add = "median",
   color = "type", fill = "type",
   palette = palette_plot_percentage)+theme_paper+xlab("Proportion of cells (log10))")+scale_x_log10()+ylab("Density")

# Quartiles plot
plot_quartiles(df_lnc, df_mrna, "proportion", palette = palette_plot_percentage)



scatter_plot_both(df_lnc,df_mrna, y = "Cell Percentage", type = "Median",palette = palette_plot_percentage)
```



```{r prepForMatching}
set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# ------------------------
# 1. Match expression 
# ------------------------
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1)


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
matches_median$type <- factor(matches_median$type, levels = c("lnc", "mrna"))

table(matches_median$type)
levels(matches_median$type)

# ------------------------------
#  2. Match perc cells
# ------------------------------

# matched perc cells
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, c)

matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
matches_perc$type <- factor(matches_perc$type, levels = c("lnc", "mrna"))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1)

```

# Visualize
```{r matchperc}

theme_matching <- theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(legend.position ="",legend.title = element_blank())+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 13), legend.text = element_text(size = 13))


plot_matching_scatter <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.7)+
                            labs( subtitle = paste0("# Matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "Median Expression", y = "% Cells expressing")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =medianexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)+ stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}


plot_matching_scatter(matches_perc)
plot_matching_scatter(matches_median)

```
