---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# General Expression Patterns of lncRNAs Ex Vivo


```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(MatchIt)
library(lme4)
library(edgeR)
library(DESeq2)
library(gtools)

robjectsdir<- "/home/luisas/Desktop/cluster/data/RObjects/"
source("../utils/02_sc_utils.R")

file <- "/home/luisas/Desktop/cluster/data/RObjects_old/results/seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet_dim_red.rds"
immune.combined <- readRDS(file)
```


# Try with normalization and FC super basic
```{r cars}
immune.combined <- NormalizeData(immune.combined)
```



```{r cars}

# Change Identifiers 
immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "Myeloid", `2` = "NK", 
    `3` = "B", `4` = "T", `5` = "T", `6` = "B", `7` = "Myeloid", `8` = "B", `9` = "B", 
    `10` = "Myeloid", `11` = "B", `12` = "T", `13` = "Myeloid",`14` = "Myeloid", `15` = "T", `16` = "NK")

#saveRDS(immune.combined, paste0(robjectsdir,"EXVIVO_immune.combined_idents.rds" ))

immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "-")[[1]][1]))))

# calculate count matrix
count_matrix <- immune.combined@assays$RNA@counts
count_matrix[1:10,1:10]
count_matrix_small <- count_matrix[1:3000,1:3000]

# Calculate number of genes per cell
get_n_genes <- function(count_matrix, threshold = 0 ){
  n_genes <- as.vector((Matrix::colSums(count_matrix>threshold)))
  df_new <- data.frame(n_genes = n_genes)
  df_new$cell_id <- colnames(count_matrix)
  return(df_new)
}

number_genes_per_cell <- get_n_genes(count_matrix)
colnames(count_matrix) <- Idents(immune.combined)

number_genes_per_cell[colnames(count_matrix) == "NK", ]
number_genes_per_cell[colnames(count_matrix) == "B", ]
number_genes_per_cell[colnames(count_matrix) == "Myeloid", ]


```

```{r cars}


# ----------------------------------------------------------------
#   In order to rertieve the number of cells in which they express 
# ----------------------------------------------------------------
df_lnc <- readRDS( "/home/luisas/Desktop/cluster/data/RObjects_old/df_celltype_lnc.rds")
df_mrna <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna_celltype.rds")


B <- subset(immune.combined, idents = c('B'))
Myeloid <- subset(immune.combined, idents = c('Myeloid'))
T <- subset(immune.combined, idents = c('T'))
NK <- subset(immune.combined, idents = c('NK'))


df_n_cells <- rbind(df_lnc, df_mrna)
#saveRDS(all_lncrnas, "/home/luisas/Desktop/cluster/data/RObjects/all_lncrnas.rds")
all_lncrnas <- readRDS("/home/luisas/Desktop/cluster/data/RObjects/all_lncrnas.rds")
#saveRDS(annotated_mrnas, "/home/luisas/Desktop/cluster/data/RObjects/annotated_mrnas.rds")
annotated_mrnas <- readRDS("/home/luisas/Desktop/cluster/data/RObjects/annotated_mrnas.rds")
#saveRDS(marker.genes,"/home/luisas/Desktop/cluster/data/RObjects/marker.genes.rds" )
```

```{r cars}
gene <- "CD14"
calc_spec <- function(gene){
  # Extract the normalized counts
  Bvector <- as.vector(mean(B[gene,]@assays$RNA@counts))
  Tvector <- as.vector(mean(T[gene,]@assays$RNA@counts))
  Mvector <- as.vector(mean(Myeloid[gene,]@assays$RNA@counts))
  NKvector <- as.vector(mean(NK[gene,]@assays$RNA@counts))
  
  # Weight the mean by the proportion of cells in which it express 
  weight_M <- df_n_cells[df_n_cells$ident == "Myeloid",][gene,]$perc_cells_expressing
  weight_T <- df_n_cells[df_n_cells$ident == "T",][gene,]$perc_cells_expressing
  weight_B <- df_n_cells[df_n_cells$ident == "B",][gene,]$perc_cells_expressing
  weight_NK <- df_n_cells[df_n_cells$ident == "NK",][gene,]$perc_cells_expressing

  # Multiply the expression vector by the fraction of cells expressed in each cell tyoe
  Bvector <- Bvector*weight_B
  Tvector <- Tvector*weight_T
  Mvector <- Mvector*weight_M
  NKvector <-NKvector*weight_NK
  
  # Calc specificity scores per celltype 
  Bscore <- min(abs(foldchange(Bvector,Tvector)),abs(foldchange(Bvector,Mvector)),abs(foldchange(Bvector,NKvector)))
  Mscore <- min(abs(foldchange(Mvector,NKvector)), abs(foldchange(Mvector,Tvector)), abs(foldchange(Mvector,Bvector)))
  NKscore <- min(abs(foldchange(NKvector,Tvector)), abs(foldchange(NKvector,Bvector)), abs(foldchange(NKvector,Mvector)))
  Tscore <- min(abs(foldchange(Tvector,NKvector)), abs(foldchange(Tvector,Bvector)), abs(foldchange(Tvector,Mvector)))
  
  # Keep the max
  specscore <- max(Bscore, Mscore, NKscore, Tscore)
  
  df <- data.frame(Bscore,Mscore,Tscore,NKscore, specscore)
  colnames(df) <- c("B", "Myeloid", "T", "NK", "Score")
  rownames(df) <- gene
  return(df)
}


calc_spec_2<- function(gene){
  # Extract the normalized counts
  Bvector <- as.vector(mean(B[gene,]@assays$RNA@counts))
  Tvector <- as.vector(mean(T[gene,]@assays$RNA@counts))
  Mvector <- as.vector(mean(Myeloid[gene,]@assays$RNA@counts))
  NKvector <- as.vector(mean(NK[gene,]@assays$RNA@counts))
  
  
  Bvector <- as.vector(mean(B[gene,]@assays$RNA@counts))
  Tvector <- as.vector(mean(T[gene,]@assays$RNA@counts))
  Mvector <- as.vector(mean(Myeloid[gene,]@assays$RNA@counts))
  NKvector <- as.vector(mean(NK[gene,]@assays$RNA@counts))
  
  # Weight the mean by the proportion of cells in which it express 
  weight_M <- df_n_cells[df_n_cells$ident == "Myeloid",][gene,]$perc_cells_expressing
  weight_T <- df_n_cells[df_n_cells$ident == "T",][gene,]$perc_cells_expressing
  weight_B <- df_n_cells[df_n_cells$ident == "B",][gene,]$perc_cells_expressing
  weight_NK <- df_n_cells[df_n_cells$ident == "NK",][gene,]$perc_cells_expressing

  # Multiply the expression vector by the fraction of cells expressed in each cell tyoe
  Bvector <- Bvector*weight_B
  Tvector <- Tvector*weight_T
  Mvector <- Mvector*weight_M
  NKvector <-NKvector*weight_NK
  
  # Calc specificity scores per celltype 
  Bscore <- min(abs(foldchange(Bvector,Tvector)),abs(foldchange(Bvector,Mvector)),abs(foldchange(Bvector,NKvector)))
  Mscore <- min(abs(foldchange(Mvector,NKvector)), abs(foldchange(Mvector,Tvector)), abs(foldchange(Mvector,Bvector)))
  NKscore <- min(abs(foldchange(NKvector,Tvector)), abs(foldchange(NKvector,Bvector)), abs(foldchange(NKvector,Mvector)))
  Tscore <- min(abs(foldchange(Tvector,NKvector)), abs(foldchange(Tvector,Bvector)), abs(foldchange(Tvector,Mvector)))
  
  # Keep the max
  specscore <- max(Bscore, Mscore, NKscore, Tscore)
  
  df <- data.frame(Bscore,Mscore,Tscore,NKscore, specscore)
  colnames(df) <- c("B", "Myeloid", "T", "NK", "Score")
  rownames(df) <- gene
  return(df)
}

colnames(count_matrix)
number_genes_per_cell <- get_n_genes(count_matrix)$n_genes
spec_res <- do.call(rbind, lapply(rownames(immune.combined), function(x) calc_spec(x)))
#saveRDS(spec_res, "/home/luisas/Desktop/cluster/data/RObjects/spec_res_weighted.rds")


spec_res <- readRDS("/home/luisas/Desktop/cluster/data/RObjects/spec_res_weighted.rds")

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)


spec_res$type <- "other"
spec_res[all_lncrnas,]$type<- "lnc"
spec_res[annotated_mrnas,]$type<- "pc"
spec_res$log <- log(spec_res$Score)



# ----------------------------------------------------------------
#   In order to rertieve the number of cells in which they express 
# ----------------------------------------------------------------
df_lnc <- readRDS( "/home/luisas/Desktop/cluster/data/RObjects_old/df_lnc.rds")
df_mrna <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna.rds")

df_n_cells <- rbind(df_lnc, df_mrna)
spec_res_int <- spec_res[spec_res$type %in% c("lnc", "pc"),]
df_test <- merge(spec_res_int, df_n_cells, by = "row.names")

# Only keep genes that show expression in at least 30 cells.
df_test <- df_test[df_test$n_cells > 30, ]
df_test <- df_test[df_test$log != "Inf",]
df_test <- df_test[!is.na(df_test$Row.names),]

rownames(df_test) <- df_test$`Row.names`
df_test$scaledScore <- rescale(df_test$Score)
```


```{r plots}
# How does the general distribution of the scores look like
ggdensity(df_test, x ="log",col="#666A86", fill = "#666A86", rug = TRUE)+
                theme_paper+scale_color_manual(values = (palette_plot_percentage))+
                scale_fill_manual(values = (palette_plot_percentage))+labs(x = "log(SpecScore)")


# How does the general distribution of the scores look like : PC vs LNC
ggdensity(df_test, x ="log",col="type", fill = "type", rug = TRUE)+
                theme_paper+scale_color_manual(values = (palette_plot_percentage))+
                scale_fill_manual(values = (palette_plot_percentage))+labs(x = "log(SpecScore)")

my_comparisons <- list( c("lnc", "pc") )
ggboxplot(df_test, x = "type",y= "log",col="type", fill = "type", alpha= 0.7) +
                theme_paper+scale_color_manual(values = rev(palette_plot_percentage))+
                scale_fill_manual(values = rev(palette_plot_percentage))+
                theme(legend.position = "none")+
                stat_compare_means(comparisons = my_comparisons)+
                labs(y = "log(specScore)", x = "")


# ---------------------------------------------------------
#                    Plot Markers 
# ---------------------------------------------------------
highlight_df <- df_test[rownames(df_test) %in% marker.genes, ]
p1 <- ggplot(df_test, aes(x = log , y = n_cells, col = type))+scale_color_manual(values = (palette_plot_percentage))
p1 <- p1+geom_point(size = 0.1)+theme_classic()+theme_paper+
         geom_point(data=highlight_df, aes(x=log,y=n_cells), color='red',size=3)

p1

# -----------------------------------------------------------
#                       Some examples 
# -----------------------------------------------------------

df_test_max <- df_test[df_test$log %in% tail(sort(df_test$log),15), ]
df_test_min <- df_test[df_test$log %in% head(sort(df_test$log),15), ]


p <- lapply(rownames(df_test_max), function(x) FeaturePlot(immune.combined, features = x))
ggsave(
   filename = "plots_max.pdf", 
   plot = marrangeGrob(p, nrow=5, ncol=3), 
   width = 9, height = 15
)


p <- lapply(rownames(df_test_min), function(x) FeaturePlot(immune.combined, features = x))
ggsave(
   filename = "plots_min.pdf", 
   plot = marrangeGrob(p, nrow=5, ncol=3), 
   width = 9, height = 15
)
```

# Matching ananlysis


```{r matchmedian}

set.seed(123)
c <- df_test[complete.cases(df_test),]
c$type = ifelse(c$type == "lnc", 1, 0)

# match expression 
mi_median <- matchit(type ~ n_cells,c)
matches_median <- get_matches(mi_median, c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$log
perc_cells_1 <- matches_median[matches_median$type == 1,]$log
t.test(perc_cells_0, perc_cells_1)


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)


scatter_plot_matching <- ggplot(matches_median, aes(x=log, y =n_cells, col = as.factor(type)))+geom_point(size = 0.8)+labs( subtitle = paste0("# Matches ", n_matches))+scale_fill_manual(values = palette_plot_percentage)+
          scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))

scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot
```







```{r matchmedian}

min <- spec_res[order(spec_res$log),][1:5,]
max <- spec_res[order(-spec_res$log),][1:15,]



lapply(rownames(min), function(x) FeaturePlot(immune.combined, x)) 


```
















```{r cellt type specificity}

# Calculate the number of genes expressed in each cell to use as covariate 
number_genes_per_cell <- get_n_genes(count_matrix)$n_genes
individual <- immune.combined$individual
celltype <- as.vector(Idents(immune.combined))
files <- colnames(count_matrix)



# Create object 
d <- DGEList(counts = count_matrix, group = factor(celltype))

# Normalize
d <- calcNormFactors(d)


# I need two different models depending on the reference celltype
# -----------------------------------
#    First One: Reference Myeloid
# -----------------------------------
celltype <- factor(celltype)
levels(celltype) <- c("Myeloid","NK","T", "B")
# Set up Design 
design <- model.matrix(~ celltype + number_genes_per_cell)
# Fit the NB GLMs
fit <- glmFit(d, design, dispersion=dispersion.true)
# Get contrasts for Myeloid 
contrasts <- makeContrasts(celltypeB-celltypeNK, celltypeB-celltypeT,  levels=fit$coefficients)

BvsNK <- glmLRT(fit, contrast = contrasts[,"celltypeB - celltypeNK"])
saveRDS(BvsNK,"/home/luisas/Desktop/cluster/data/RObjects/BvsNK.rds")
BvsNK <- readRDS("/home/luisas/Desktop/cluster/data/RObjects/BvsNK.rds")
BvsT <- glmLRT(fit, contrast = contrasts[,"celltypeB - celltypeT"])
saveRDS(BvsT,"/home/luisas/Desktop/cluster/data/RObjects/BvsT.rds")
BvsT <- readRDS("/home/luisas/Desktop/cluster/data/RObjects/BvsT.rds")
# -----------------------------------
#    Second One: Reference T
# -----------------------------------
celltype <- factor(celltype)
levels(celltype) <- c("T","NK","Myeloid", "B")
# Set up Design 
design <- model.matrix(~ celltype + number_genes_per_cell)
# Fit the NB GLMs
fit <- glmFit(d, design, dispersion=dispersion.true)
# Get contrasts for Myeloid 
contrasts <- makeContrasts(celltypeMyeloid-celltypeB,levels=fit$coefficients)
MvsB <- glmLRT(fit, contrast = contrasts[,"celltypeMyeloid - celltypeB"])
saveRDS(MvsB,"/home/luisas/Desktop/cluster/data/RObjects/MvsB.rds")



# -----------------------------------
#    Third One: Reference B
# -----------------------------------
celltype <- factor(celltype)
levels(celltype) <- c("B","NK","Myeloid", "T")
# Set up Design 
design <- model.matrix(~ celltype + number_genes_per_cell)
# Fit the NB GLMs
fit <- glmFit(d, design, dispersion=dispersion.true)
fit$coefficients
# Get contrasts for Myeloid 
contrasts <- makeContrasts(celltypeMyeloid-celltypeNK, celltypeMyeloid-celltypeT, celltypeT-celltypeNK, levels=fit$coefficients)

MvsNK <- glmLRT(fit, contrast = contrasts[,"celltypeMyeloid - celltypeNK"])
saveRDS(MvsNK,"/home/luisas/Desktop/cluster/data/RObjects/MvsNK.rds")

MvsT <- glmLRT(fit, contrast = contrasts[,"celltypeMyeloid - celltypeT"])
saveRDS(MvsT,"/home/luisas/Desktop/cluster/data/RObjects/MvsT.rds")

NKvsT <- glmLRT(fit, contrast = contrasts[,"celltypeT - celltypeNK"])
saveRDS(NKvsT,"/home/luisas/Desktop/cluster/data/RObjects/NKvsT.rds")



calc_spec_score <- function(gene){
  
  mb <- MvsB[gene,]$table$logFC
  mt <- MvsT[gene,]$table$logFC
  mnk <- MvsNK[gene,]$table$logFC
  nkt <- NKvsT[gene,]$table$logFC
  bnk <- BvsNK[gene,]$table$logFC
  bt <- BvsT[gene,]$table$logFC
  
  m <- min(abs(mb),abs(mt),abs(mnk))
  b <-  min(abs(mb),abs(bt),abs(bnk))
  t <- min(abs(mt), abs(tnk),abs(bt))
  nk <- min(abs(mnk), abs(tnk),abs(bnk))
  
  
  spec_score <- max(m,b,t,nk)
  
  df <- data.frame(b,t,nk,m, spec_score)
  colnames(df) <- c("B", "T", "NK", "Myeloid", "Score")

  return(df)
}




# Test Gene 
gene <- "A1BG"
FeaturePlot(immune.combined, features = c(gene))
calc_spec_score("CD14")
calc_spec_score("A1BG")


```





```{r cellt type specificity}

DGEList(counts = count_matrix_small, group = factor(celltype))
dds <- DESeqDataSetFromMatrix(countData = count_matrix_small, colData = sampleTable, design = design)
dds <- DESeq(dds)

resultsNames(dds)[2:length(resultsNames(dds))-1]
for (a in resultsNames(dds)){
  print(a)
}
design(dds)
r1 <- results(dds)
comp1 <- results(dds, name=c("celltype_Myeloid_vs_B"))
rownames(res)

aa <- results(dds, contrast=list( c("celltype_Myeloid_vs_B","celltype_NK_vs_B")))

head(aa)


markers <- FindAllMarkers(immune.combined)

saveRDS(markers, "/home/luisas/Desktop/cluster/data/RObjects_old/mk.rds")

plot(density(markers$avg_logFC))

gene_test <- markers[markers$avg_logFC== max(abs(markers$avg_logFC)),]

plot_genes(immune.combined,gene_test$gene,threshold= 3, title = "Monocytes marker expression", col = "purple")

# Other test 
edgeR::glmFit(count_matrix_small)

x <- 0:2
design <- model.matrix(~x)


nlibs <- 3
ngenes <- 100
dispersion.true <- 0.1

# Make first gene respond to covariate x
x <- 0:2
x
design <- model.matrix(~x)
design
beta.true <- cbind(Beta1=2,Beta2=c(2,rep(0,ngenes-1)))
mu.true <- 2^(beta.true %*% t(design))


d <- DGEList(y)

# Normalize
d <- calcNormFactors(d)

# Fit the NB GLMs
fit <- glmFit(d, design, dispersion=dispersion.true)
```



```{r a }
subset <- subset(immune.combined, feature = c("CD14"))
m_vs_t <- FindMarkers(immune.combined, ident.1 = "Myeloid", ident.2 = "T")
m_vs_nk <- FindMarkers(immune.combined, ident.1 = "Myeloid", ident.2 = "NK")
m_vs_B <- FindMarkers(immune.combined, ident.1 = "Myeloid", ident.2 = "B")
T_vs_B <- FindMarkers(immune.combined, ident.1 = "T", ident.2 = "B")
gene <- "CD79B"
spec_score <- min(abs(m_vs_t[gene,]$avg_logFC), abs(m_vs_nk[gene,]$avg_logFC), abs(m_vs_B[gene,]$avg_logFC))
spec_score
rbind(m_vs_t, m_vs_B, m_vs_nk)

# Two example genes
count_matrix_small <- count_matrix[c("CD14", "CD79B"), ]
FeaturePlot(immune.combined, features = c("CD14", "CD79B"))

# Try for one gene 
gene_counts <-  as.vector(count_matrix["CD14",])
gene_counts <- as.matrix(t(gene_counts))

dgelistobj <- DGEList(counts = gene_counts, group = sampleTable$celltype, genes = rownames(gene_counts))
library(MASS)
counts <- as.vector(gene_counts)[1:1000]
celltype <- factor(celltype)
model <- glmer.nb(counts ~ celltype + number_genes_per_cell + (1|individual), contrasts = cbind(c("B", "T"), c("NK", "B")))
summary(model)
require("gmodels")
fit.contrast(model)
emt <- emtrends(model)

glmFit()



```



# Expression of ebola genes 
```{r overallstats}
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unkown", sep ="-")]
ebola_genes
plot_genes(immune.combined, ebola_genes,3, col = "darkred", "Ebola RNA expression")
plot_genes(immune.combined, ebola_genes,1, col = "#DA0202", "Cells expressing Ebola RNA")

# How many cells present read in general 
mask <- unlist(lapply(rownames(immune.combined), function(x) any(grepl(x, ebola_genes) == TRUE ) ))
expr <- FetchData(immune.combined, rownames(immune.combined[mask,]))
g1 <- WhichCells(immune.combined, cells = colnames(immune.combined[, which(x = expr > 2)]))

# How many cells present read in myeloids
myeloids <- subset(immune.combined, idents  = "NK")
mask <- unlist(lapply(rownames(myeloids), function(x) any(grepl(x, ebola_genes) == TRUE ) ))
expr <- FetchData(myeloids, rownames(myeloids[mask,]))
g1_m <- WhichCells(myeloids, cells = colnames(myeloids[, which(x = expr > 2)]))

length(unique(g1))
(length(unique(g1_m))/length(unique(colnames(myeloids))))*100
```


# Calculate expression 

```{r quartiles}
# Calculate expression POOLED
df_lnc <- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(), tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = all_lncrnas), "",df_lnc, 1))
df_mrna<- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(),tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = annotated_mrnas), "",df_mrna, 1))

df_pseudogenes<- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(),tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = annotated_pseudogenes), "",df_pseudogenes, 1))


saveRDS(df_lnc, "/home/luisas/Desktop/cluster/data/RObjects_old/df_lnc.rds")
saveRDS(df_mrna, "/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna.rds")
saveRDS(df_pseudogenes, "/home/luisas/Desktop/cluster/data/RObjects_old/df_pseudogenes.rds")

df_lnc <- readRDS( "/home/luisas/Desktop/cluster/data/RObjects_old/df_lnc.rds")
df_mrna <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna.rds")
df_pseudogenes <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/df_pseudogenes.rds")

```

```{r celltype}
df_lnc_celltype <- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = all_lncrnas)
invisible(lapply(as.vector(unique(Idents(subset))), function(x) calc_mean_and_percentage_cell(subset, x, df_lnc_celltype)))

# mRNA
df_mrna_celltype <- data.frame( gene_id = c(),meanexpr = c(), not_na_cells = c(), maxexpr = c(), medianexpr = c(), var = c(), var = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
invisible(lapply(as.vector(unique(Idents(subset))), function(x) calc_mean_and_percentage_cell(subset, x, df_mrna_celltype)))

saveRDS(df_lnc_celltype, "/home/luisas/Desktop/cluster/data/RObjects_old/df_celltype_lnc.rds")
saveRDS(df_mrna_celltype, "/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna_celltype.rds")

df_lnc_celltype <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/df_celltype_lnc.rds")
df_mrna_celltype <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna_celltype.rds")

df_mrna_celltype$type <- NA
df_lnc_celltype$type <-  NA
```

# Define palette 

```{r colors}
col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)
```
# Expression across Quartiles
```{r quartiles}
# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNAs"
df_mrna$type <- "mRNAs"
df_complete <- rbind(df_lnc, df_mrna)

plot_quartiles(df_lnc, df_mrna, "number", palette = palette_plot_percentage)
plot_quartiles(df_lnc, df_mrna, "proportion", palette = palette_plot_percentage)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())


ggboxplot(df_complete, y="perc_cells_expressing", x = "type", fill = "type", color = "type", alpha = 0.5)+
                scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)+
                stat_compare_means(comparisons = list(c("lncRNAs", "mRNAs")))+
                labs(x = "", y = "% Cells")+
                theme(legend.title = element_blank(), legend.position = "none")

```




# ------------------------------------------
# Calculate the EXPRESSION
# ------------------------------------------



```{r percentage ebov reads}
# Scatter plot both 
scatter_plot_both(df_lnc,df_mrna, y = "Cell Percentage", type = "Max", palette = palette_plot_percentage)
scatter_plot_both(df_lnc,df_mrna, y = "Cell Percentage", type = "Median",palette = palette_plot_percentage)


ref_human <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/gencode/release_26/gencode.v26.GRCh38.annotation.gtf"))
table(ref_human$gene_type)

table(ref$gene_biotype)


df_lnc$ident <- "lncRNA"
df_mrna$ident <- "mRNA"
df_pseudogenes$type <- "pseudogene"
df_pseudogenes$ident <- "pseudogene"
df_compl <- rbind(df_lnc, df_mrna, df_pseudogenes)

ggplot(df_compl, aes(x = perc_cells_expressing, col = ident, fill= ident))+
  geom_density(alpha = 0.5)+
  theme_classic()+
  xlim(0,0.5)+
  labs(x = "% of cells expressing the gene")+
  scale_fill_manual(values= c(palette_plot_percentage, "darkgreen"))+
  scale_color_manual(values= c(palette_plot_percentage, "darkgreen"))+
  theme(legend.title = element_blank() ) 


ggboxplot(df_compl, y="perc_cells_expressing", x = "ident", fill = "ident", color = "ident", alpha = 0.5)+
                scale_fill_manual(values = c(palette_plot_percentage, "darkgreen"))+
                scale_color_manual(values = c(palette_plot_percentage, "darkgreen"))+
                stat_compare_means(comparisons = list(c("lncRNA", "mRNA"), c("pseudogene", "mRNA"),  c("pseudogene", "lncRNA")))+
                labs(x = "", y = "% Cells")+
                theme(legend.title = element_blank(), legend.position = "none")
```

# ------------------------------
# Linear model instead of matching
# ------------------------------

```{r linearModelInsetadoofMatching}
df_lnc$ident <- "lncRNA"
df_mrna$ident <- "protein_coding"
df_complete <- rbind(df_lnc, df_mrna)

genetype <- as.factor(ifelse(df_complete$ident == "lncRNA", 0, 1))
medianexpression <- df_complete$medianexpr
perc_cells_expressing <- df_complete$perc_cells_expressing

model2 <- lm(medianexpression ~  perc_cells_expressing + genetype )

summary(model2)

library(car)
library(gvlma)

car::vif(model2)
model <- glm(genetype ~ perc_cells_expressing+ medianexpression , family = binomial(link = "logit") )
summary(model)

```

# ------------------------------
#     Match median expression 
# ------------------------------

```{r matchmedian}

set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# match expression 
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1)


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
```

# Visualize matching

```{r matchmedianplot}

scatter_plot_matching <- ggplot(matches_median, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( subtitle = paste0("# Matches ", n_matches))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
          scale_color_manual(values = palette_plot_percentage)+theme(legend.position = "none")+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank(), legend.position = "")+theme(legend.title = element_blank(), legend.position = "")+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))


scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot
palette_plot_percentage

# Just to check 
ggboxplot(matches_median, x = "type", y = "perc_cells_expressing", fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("mrna", "lnc")),  label = "p.signif")+  stat_compare_means(label.y = 1)+ labs(x = "", y =" % Cells expressing")+theme_classic()+theme(legend.position = "none")

ggboxplot(matches_median, x = "type", y = "maxexpr", fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("mrna", "lnc")),  label = "p.signif")+  stat_compare_means(label.y = 7)+ labs(x = "")+theme_classic()+theme(legend.position = "none")

ggboxplot(matches_median, x = "type", y = "minexpr", fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("mrna", "lnc")),  label = "p.signif")+  stat_compare_means(label.y = 1.8)+ labs(x = "")+theme_classic()+theme(legend.position = "none")

ggplot(matches_median, aes(y=medianexpr, col = as.factor(type)))+geom_boxplot()+theme_classic()+labs(title = "Comparison of median expression values of matched genes by median (Validation)")
ggplot(matches_median, aes(y=perc_cells_expressing, col = as.factor(type)))+geom_boxplot()+theme_classic()+labs(title = "Comparison of %cells expressing of matched genes by median (Validation)")

```

# Validate 

```{r matchvalidation}
run_matchit_median <- function(c, seed){
    set.seed(seed)
    mi_median <- matchit(type ~ medianexpr,c)
    matches_median <- get_matches(mi_median, c)
    matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
    n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
    matches_median$type <- as.factor(matches_median$type)
    output <- matches_median[matches_median$type  == "mrna",]$perc_cells_expressing
    df <- as.data.frame(output)
    df$seed <- as.character(seed)
    return(df)
}


#seeds <- 1:10
#list <- do.call(rbind,(lapply( seeds, function(seed) run_matchit_median(c,seed))))
#lncRNAs <- data.frame(output = c[c$type == 1,]$perc_cells_expressing, seed= as.character("0-lncRNAs"))
#lncRNAs$seed <- as.character(lncRNAs$seed)
#benchmark_df_matched_median <- rbind(lncRNAs, list)
#table(benchmark_df_matched_median$seed)

#ggplot(benchmark_df_matched_median, aes (y = output, x = seed, fill = seed, color = seed))+geom_boxplot(alpha = 0.6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+scale_fill_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+scale_color_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+labs(y = "% Cells", Title = "")+theme(legend.position = "")

```

# ------------------------------
#     Match Median per celltype 
# ------------------------------
```{r matchmedian}

set.seed(123)
df_lnc_celltype$type <- "lnc"
df_mrna_celltype$type <- "mrna"
df_complete<- rbind(df_lnc_celltype, df_mrna_celltype)

# Filter
# Select the ones expressed at least in 50 cells 
df_complete <- df_complete[df_complete$n_cells >= 50, ]

c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

length(unique(unlist(de_lists_mrna)))
# -------------------------------
#   Select Only one celltype 
# -------------------------------
match_in_one_celltype_median <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_median <- matchit(type ~ medianexpr,c)
  matches_median <- get_matches(mi_median, c)
  
  # Calculate p-value 
  perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
  perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
  test <- wilcox.test(perc_cells_0, perc_cells_1, paired = TRUE)
  pval <- test$p.value
  
  matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
  matches_median$type <- as.factor(matches_median$type)
  
  title = paste0(celltype_test)
  scatter_plot_matching <- ggplot(matches_median, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( title=title, subtitle = paste0("n: ", n_matches, "\n", "p-value: ", pval))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))
  
  
  scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
  scatter_plot_matching_withboxplot

}

match_in_one_celltype_perccells <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  
  # match expression 
  mi_perc <- matchit(type ~ perc_cells_expressing,c)
  matches_perc <- get_matches(mi_perc, c)
  matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
  
  median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
  median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
  test <- wilcox.test(median_0, median_1, paired = TRUE)
  pval <- test$p.value

  
  title = paste0(celltype_test)
  scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, fill = factor(type, level = c("lnc", "mrna")), col = factor(type, level = c("lnc", "mrna"))))+geom_point(size = 0.8)+labs( title=title, subtitle = paste0("n: ", n_matches, "\n", "p-value: ", pval))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))
  
  
  scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot', margins = "both",size = 3, alpha = 0.3, groupColour = TRUE, groupFill = TRUE, margMapping = aes(colour = factor(type, level = c("lnc", "mrna"))))
  scatter_plot_matching_withboxplot

}

match_in_one_celltype_median(c, "NK")
match_in_one_celltype_median(c, "Myeloid")
match_in_one_celltype_median(c, "T")
match_in_one_celltype_median(c, "B")

match_in_one_celltype_perccells(c, "B")
match_in_one_celltype_perccells(c, "Myeloid")
match_in_one_celltype_perccells(c, "T")
match_in_one_celltype_perccells(c, "NK")
```


# ------------------------------
#     Match perc cells
# ------------------------------

```{r matchperc}
# matched perc cells
set.seed(123)
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, c)
matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1)

```

# Visualize
```{r matchperc}

scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+
                         geom_point(size = 0.8, alpha = 0.8)+labs(title = "", subtitle = paste0("# Matches ", n_matches))+
                         theme(legend.title = element_blank(),
                               legend.position = "none",
                               panel.background = element_rect(fill = "white", colour = "grey50"),
                               axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))+
                         scale_fill_manual(values =palette_plot_percentage)+
                         scale_color_manual(values = palette_plot_percentage)+
                         labs(x = "Median Expression", y = "% Cells expressing")+theme()

scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.5, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot

# Just checking
ggboxplot(matches_perc, x = "type", y = "medianexpr", fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("lnc", "mrna")),  label = "p.signif")+  stat_compare_means(label.y = 3.5)+ labs(x = "")

ggboxplot(matches_perc, x = "type", y = "maxexpr",fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("lnc", "mrna")),  label = "p.signif")+  stat_compare_means(label.y = 7)

ggboxplot(matches_perc, x = "type", y = "minexpr",fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("lnc", "mrna")),  label = "p.signif")+  stat_compare_means(label.y = 3)

ggboxplot(matches_perc, x = "type", y = "perc_cells_expressing",fill = "type", alpha = 0.5,
                color = "type", palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("lnc", "mrna")),  label = "p.signif")+  stat_compare_means(label.y = 1)


scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( subtitle = paste0("# Matches ", n_matches))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
          scale_color_manual(values = palette_plot_percentage)+theme(legend.position = "none")+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank(), legend.position = "")+theme(legend.title = element_blank(), legend.position = "")+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))


scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot

```

# Validate


```{r matchperc}
run_matchit_perc<- function(c, seed){
    set.seed(seed)
    mi_perc <- matchit(type ~ perc_cells_expressing,c)
    matches_perc <- get_matches(mi_perc, c)
    matches_perc$type = ifelse(matches_perc$type == 0, "lnc", "mrna")
    matches_perc$type <- as.factor(matches_perc$type)
    output <- matches_perc[matches_perc$type  == "mrna",]$medianexpr
    df <- as.data.frame(output)
    df$seed <- as.character(seed)
    return(df)
}

#seeds <- 1:10
#list <- do.call(rbind,(lapply( seeds, function(seed) run_matchit_perc(c,seed))))
#lncRNAs <- data.frame(output = c[c$type == 1,]$medianexpr, seed= as.character("0-lncRNAs"))
#lncRNAs$seed <- as.character(lncRNAs$seed)
#benchmark_df_matched_median <- rbind(lncRNAs, list)
#table(benchmark_df_matched_median$seed)
#ggplot(benchmark_df_matched_median, aes (y = output, x = seed, fill = seed, color = seed))+geom_boxplot(alpha = 0.6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+scale_fill_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+scale_color_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+labs(y = "Median", Title = "")+theme(legend.position = "")
```

# ------------------------------------------------------
# Cherry pick an example - highly expressed in low proportion of cells : TSNE
# ------------------------------------------------------

```{r cellt type specificity}
a <- df_lnc[df_lnc$perc_cells_expressing < 0.1 & df_lnc$medianexpr>1.5 & df_lnc$n_cells > 50,]
gene_example_1 <- "ENSMMUG00000038067-unkown"
gene <- a$gene_id[21]
a
df_lnc_filtered <- df_lnc[df_lnc$gene_id == gene,]
df_complete<- rbind(df_lnc_filtered, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

set.seed(125)
mi <- matchit(type ~ medianexpr,c)
matches <- get_matches(mi, c)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)
matches[,c("medianexpr", "n_cells", "type")]
rownames(matches) <- gsub("-unkown", "", matches$gene_id)
f1 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[1]), pt.size = 1.2, order = TRUE, cols = c("lightgrey", "#bf2e2e"), max.cutoff = 3.5)+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = genes[1])

f2 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[2]), pt.size = 1.2, order = TRUE, cols = c("lightgrey", "#bf2e2e"), max.cutoff = 3.5)+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = gsub("-unkown", "", genes[2]))


f1
f2
CombinePlots(list(f1,f2))
```


# cell type specificity : Glm 



```{r norm}
fit <- fitdistr(my_data, densfun = "normal")
hist(data.q, pch = 20 , breaks = 25, prob = TRUE, main = "")
curve(dnorm(x, fit$estimate[1], fit$estimate[2]), col = "red", lwd = 2, add = T)
```


# Cell type specificity : TAU 

```{r cellt type specificity}
library(knitr)
library(tispec)
knitr::opts_chunk$set(
    out.width = "600px", # sets image sizes
    fig.align = "center",
    collapse = TRUE,
    comment = ">"
)

wd <- paste(getwd(), '/images/', sep = '') # should work on any machine
immune.combined_new <- NormalizeData(immune.combined, normalization.method = "RC", scale.factor = 1e6)
expression_matrix <- immune.combined_new@assays$RNA@data

colnames(expression_matrix) <- Idents(immune.combined_new)
meanExpression <- as.data.frame(sapply(unique(colnames(expression_matrix)), function(celltype) rowMeans(expression_matrix[,colnames(expression_matrix) == celltype])))
# check that the matrix does not contain any string etc == > All fine
checkInput(meanExpression[,1:4])
meanExpression
# Log 2 Transformation
log2Exp <- log2Tran(meanExpression)
qnExp <- quantNorm(log2Exp)
tauExp <- calcTau(qnExp)

# Chek marker genes
tauExp[rownames(tauExp)%in% marker.genes,]


df_lnc$type <- "lncRNAs"
df_mrna$type <- "mRNAs"
df_complete<- rbind(df_lnc, df_mrna)


df_perc_tau <- merge(tauExp,df_complete, by = "row.names")
df_perc_tau <- df_perc_tau[df_perc_tau$n_cells > 200,]
rownames(df_perc_tau) <- df_perc_tau$Row.names
```


```{r cellt type specificity}
# Double check they make sense
library(ggpubr)
#tauMarkers <- tauExp[rownames(tauExp) %in% marker.genes,]
#tauMarkers
tauLnc <- df_perc_tau[rownames(df_perc_tau) %in% all_lncrnas,]
tauLnc$type <- "lncRNAs"
tauMrna <-  df_perc_tau[rownames(df_perc_tau) %in% annotated_mrnas,]
tauMrna$type <- "protein coding genes"

taus <- rbind(tauLnc, tauMrna)

t.test(tauLnc$tau, tauMrna$tau)
# General boxplot
ggviolin(taus, x = "type", y = "tau",
                color = "type", fill = "type", alpha = 0.4, palette =palette_plot_percentage)+stat_compare_means(comparisons =list(c("lnc", "mrna")),  label = "p.signif")+  stat_compare_means(label.y = 1.1)+labs(title = "Tau score comparison lncRNAs vs mRNAs", x = "")+theme(legend.position = "none")+geom_boxplot(width = 0.1, alpha = 0.9)+ylim(0,1)
palette_plot_percentage <- c( "#870052","#2e88bf")
ggplot(df_perc_tau, aes(x = tau, col = type, fill = type))+geom_density(alpha = 0.5, size =1)+theme_classic()+scale_fill_manual(values = palette_plot_percentage)+scale_color_manual(values = palette_plot_percentage)+theme(legend.title = element_blank())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=13), axis.title = element_text(size = 18), axis.text = element_text(size = 15),legend.title = element_blank(), plot.title = element_text(size = 22, hjust = 0.5))+labs(title = "Cell Type Specificity index", x = "Tau Score")


```

```{r }
df_mrna$type<- "mrna"
df_lnc$type <- "lncrna"
df_complete <- rbind(df_mrna, df_lnc)
df_complete$type <- as.factor(df_complete$type)
genes <-intersect(df_complete$gene_id, rownames(taus))

df_taus_perc <- merge(as.data.frame(tauExp[rownames(tauExp) %in% genes, 1:5 ]), as.data.frame(df_complete[df_complete$gene_id %in% genes,]), by='row.names', all=TRUE)

# -------------------------------------
# Boxplot quantiles specificity 
# -------------------------------------
quantiles = as.table(c(0,.25,.5,.75,1))
quantiles = quantile(df_complete_quantiles$perc_cells_expressing, probs  = 0:4/4)
df_complete_quantiles <- within(df_taus_perc, quartile <- as.integer(cut(df_complete$perc_cells_expressing, quantiles, include.lowest = TRUE)))

df_complete_quartiles <- within(df_taus_perc, quartile <- as.integer(cut(df_complete$perc_cells_expressing, quantile(df_complete$perc_cells_expressing, probs=0:4/4), include.lowest=TRUE)))
df_complete_quartiles$quartile <- as.numeric(df_complete_quartiles$quartile)

ggviolin(df_complete_quartiles, x = "quartile", y = "tau",
                color = "type", fill = "type", alpha = 0.4, palette =palette_plot_percentage)+labs(title = "Tau score comparison lncRNAs vs mRNAs", x = "")+
theme(legend.title = element_blank())+
    scale_x_discrete(name ="Percentage of Cells Expressing", 
                     labels=c("1-25%","25-50%","50-75%", "75-100%"))

# -----------------------------------------
#    Correlation plots
# -----------------------------------------
df_cor <- df_complete_quartiles[,c("tau", "perc_cells_expressing", "maxexpr")]
M <-cor(df_cor)
library(corrplot)
corrplot(M)
# DOT correlation plot
source("http://www.sthda.com/upload/rquery_cormat.r")
require("corrplot")
rquery.cormat(df_cor)
library("PerformanceAnalytics")
chart.Correlation(df_cor, pch=1)



```


#-----------------------------
# Matchit analysis with taus
#-----------------------------
```{r matchit taus}
library(MatchIt)
df_perc_tau$type
c <- df_perc_tau[complete.cases(df_perc_tau),]
c$type = ifelse(c$type == "lnc", 1, 0)

mi <- matchit(type ~ perc_cells_expressing,c)
matches <- get_matches(mi, c)
n_matches <- nrow(matches)
table(matches$type)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)
matches[,c("medianexpr", "n_cells", "type")]

scatter_plot_matching <- ggplot(matches, aes(x=perc_cells_expressing, y =tau, col = as.factor(type)))+geom_point(size = 0.8)+labs(title = " Matched % cells values", subtitle = paste0("# Matches ", n_matches))+ theme(legend.title = element_blank())+ geom_smooth(method = "loess")+scale_fill_brewer(palette = "Set1")+scale_color_brewer(palette = "Set1")+theme()+theme(panel.background = element_rect(fill = "white", colour = "grey50"))

scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 5,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot

# ----------
mi <- matchit(type ~ tau,c)
matches <- get_matches(mi, c)
n_matches <- nrow(matches)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)


scatter_plot_matching <- ggplot(matches, aes(x=perc_cells_expressing, y =tau, col = as.factor(type)))+geom_point(size = 0.8)+labs(title = " Matched tau  values", subtitle = paste0("# Matches ", n_matches))+ theme(legend.title = element_blank())+ geom_smooth(method = "loess")+scale_fill_brewer(palette = "Set1")+scale_color_brewer(palette = "Set1")+theme()+theme(panel.background = element_rect(fill = "white", colour = "grey50"))

scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 5,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot

```




