---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# ----------------------------------------------------
# General Expression Patterns of lncRNAs IN VIVO
# ----------------------------------------------------


### Imports 
```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(rtracklayer)

source(file.path("../utils/00_datapaths.R"))
source("../utils/02_sc_utils.R")

# 0. Palette used throughout the scripts
col_lnc = "#f72631"
col_mrna = "#3153a3"
palette_plot_percentage <- c(col_lnc, col_mrna)

# 1. Theme for plots
theme_sc_paper <- theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())+theme_paper


# 2. Directories
# Where stats about lnc and pc computed on cluster are stored
robjectsdir<- file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/")
# Read reference files 
ref<- import(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- rtracklayer::import(file.path(data_path, "/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Import Seurat object 
file <- file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
orthologs <- readRDS(file.paht(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))
immune.combined <- readRDS(file)
```


# -------------- Check genes in object-------------------------
# Visualize how many lnc and mrnas are present in the object 

```{r check}
# ------------------------------------
#     PC genes present in object 
# -------------------------------------
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% unique(ref$gene_name[ref$gene_biotype == "protein_coding"])]

# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# Annotated lncrnas id 
lnc_ref_id <- gsub("_","-",unique(ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "lncRNA",]$gene_name))

# Annotated lncRNAs: remove the gene names so it can match 
annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc_ref_id]
# Novel lncRNAs
novel_lncrnas <- rownames(immune.combined)[unlist(lapply(rownames(immune.combined), function(x) startsWith(x, "MSTRG")))]
# Complete Set 
all_lncrnas <- c(annotated_lncrnas, novel_lncrnas)

# Save
#saveRDS(all_lncrnas, file.path(robjectsdir, "all_lncrnas.rds"))
#saveRDS(annotated_mrnas, file.path(robjectsdir, "annotated_mrnas.rds"))

#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated", "novel")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)), length(unique(novel_lncrnas))))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Dark2")
lnc_donut <- df.donut[df.donut$type == "lncRNAs",]
names(lnc_donut) <- c("-", "type", "count")
donut_plot(lnc_donut, "Greens")
```


# ---------------------------------------------------
# Play w/ pca
# ----------------------------------------------------

```{r check}
pal_celltypes <-wes_palette("GrandBudapest1", 4)
pal_celltypes[3] <- "#AE4E4E"

# Plot with correct
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =pal_celltypes, label.size = 8)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
```


# ---------------------------------------------------------
#  Load Metrics: ACROSS cell types
# (#cells in which the gene is expressed, median, etc) 
# ---------------------------------------------------------
```{r quartiles}
df_lnc <- readRDS(file.path(robjectsdir, "df_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))


df_lnc <- df_lnc[df_lnc$n_cells > 30, ]
df_mrna <- df_mrna[df_mrna$n_cells > 30, ]


# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNA"
df_mrna$type <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)
df_complete$prop_cells <- df_complete$perc_cells_expressing
df_complete$perc_cells_expressing <- df_complete$prop_cells*100

table(df_complete$type)
```

# ---------------------------------------------------------
#  Load Metrics: per cell type individually
# (#cells in which the gene is expressed, median, etc) 
# ---------------------------------------------------------
```{r celltype}
df_lnc_celltype <- readRDS(file.path(robjectsdir, "df_celltype_lnc.rds"))
df_mrna_celltype <- readRDS(file.path(robjectsdir, "df_mrna_celltype.rds"))

df_lnc_celltype <- df_lnc_celltype[df_lnc_celltype$n_cells > 30, ]
df_mrna_celltype <- df_mrna_celltype[df_mrna_celltype$n_cells > 30, ]

df_mrna_celltype$type <- "Protein Coding"
df_lnc_celltype$type <-  "lncRNA"

df_complete_celltype <- rbind(df_lnc_celltype, df_mrna_celltype)

saveRDS(df_complete_celltype, file.path(robjectsdir, "df_celltype.rds"))
```



# ---------------------------------------------------------
# 1. Visualzation: # Cells  across Quartiles and Cell-types 
# ---------------------------------------------------------
```{r quartiles}

# Figure 3A (Density plot of % cells expressing )
density_cells <- ggdensity(df_complete, x = "perc_cells_expressing",
   color = "type", fill = "type", alpha = 0.8, add = "median",
   palette = palette_plot_percentage)+xlab("% cells where gene is expressed")+ylab("density")+scale_x_log10()+theme(text = element_text(size = 15))

density_cells

# Quartiles plot
plot_quartiles(df_lnc, df_mrna, "proportion", palette = palette_plot_percentage)+theme(text = element_text(size = 15))

# Boxplot separated by cell-type
ggplot(df_complete_celltype, aes( x= type, y = perc_cells_expressing, fill = type, col = type))+geom_boxplot(alpha= 0.4)+scale_fill_manual(values = palette_plot_percentage)+scale_color_manual(values = palette_plot_percentage)+theme_sc_paper+xlab ("")+ylab("Proportion of cells")+ facet_wrap(vars(ident))+theme(strip.text.x = element_text(size = 20, colour = "black", angle = 0))


scatter_plot_both(df_lnc,df_mrna, y = "Cell Percentage", type = "Median",palette = palette_plot_percentage)

table(df_complete$type)
```






# ------------------------------
#     Match median expression 
# ------------------------------

```{r prepForMatching}
set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# ------------------------
# 1. Match expression 
# ------------------------
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1, alternative = "greater")


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
matches_median$type <- factor(matches_median$type, levels = c("lnc", "mrna"))

table(matches_median$type)
levels(matches_median$type)

# ------------------------------
#  2. Match perc cells
# ------------------------------

# matched perc cells
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, c)

matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
matches_perc$type <- factor(matches_perc$type, levels = c("lnc", "mrna"))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1,alternative = "less" )

```

# Visualize
```{r matchperc}

theme_matching <- theme(panel.background = element_rect(fill = "white", color = "black"))+
                  theme(legend.position ="",panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
              axis.ticks.length=unit(.2, "cm"))+
                  theme(axis.text = element_text(size = 15, color ="black"), axis.title = element_text(size = 17))


plot_matching_scatter <- function(matches_perc){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = "median expression", y = "% cells")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =medianexpr, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)+ stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}


plot_matching_scatter(matches_perc)
plot_matching_scatter(matches_median)


df_lnc[order(df_lnc$medianexpr),]
```





# ------------------------------------------------------
# Cherry pick an example 
# highly expressed in low proportion of cells 
# ------------------------------------------------------

```{r cellt type specificity}
candidates  <- df_lnc[ df_lnc$medianexpr>1.5 & df_lnc$n_cells > 50& df_lnc$n_cells < 1200,]
gene <- candidates$gene_id[11]
print(gene)
# ----------------------------------------------
# 1 . Find the matching (median expr) Gene
# ----------------------------------------------
df_lnc_filtered <- df_lnc[df_lnc$gene_id == gene,]
df_complete<- rbind(df_lnc_filtered, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)
set.seed(125)
mi <- matchit(type ~ medianexpr,c)
matches <- get_matches(mi, c)
rownames(matches) <- gsub("-unknown", "", matches$gene_id)
matches$type = ifelse(matches$type == 0, "mrna", "lnc")
genes <- as.character(matches$gene_id)
matches[,c("medianexpr", "n_cells", "type")]


# ----------------------------------------------
#  2. Visualize that the expression of the 2 genes BOXPLOT
# ----------------------------------------------
expr_matrix <- immune.combined@assays$RNA@data
df_genes <- data.frame(t(expr_matrix[genes, ]))
colnames(df_genes) <-  gsub(".unknown", "", colnames(df_genes))
df_genes_melted <- melt(df_genes)
df_genes_melted <- df_genes_melted[df_genes_melted$value > 0, ]
bp <- ggboxplot(df_genes_melted, x = "variable", y = "value", fill = "variable", col = "black",alpha=1)+theme()+scale_fill_manual(values = rev(palette_plot_percentage))+scale_color_manual(values = rev(palette_plot_percentage))+stat_compare_means(comparisons = list(c(colnames(df_genes))))+xlab("")+ylab("logCP10K")+theme(text = element_text(size=15))+scale_y_continuous(limits = c(0,5), expand = c(0,0))+theme(axis.line.x = element_blank(), legend.position = "", axis.ticks.x = element_blank(), axis.text.x = element_blank())

bp
# ----------------------------------------------
#  2. Visualize that the expression of the 2 genes FEATURE PLOT
# ----------------------------------------------
f1 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[1]), pt.size = 0.6, order = TRUE,max.cutoff = 3.5,  cols = c("lightgrey", "navy"))+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = genes[1])+theme_void()+theme(title = element_blank(), text = element_text(size = 20))
f1
f2 <- FeaturePlot(immune.combined,label = FALSE, features = c(genes[2]), pt.size = 0.6, order = TRUE, max.cutoff = 3.5,cols = c("lightgrey", "#B0052D"))+theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))+labs(title = gsub("-unknown", "", genes[2]))+theme_void()+theme(title = element_blank(), text = element_text(size = 20))

CombinePlots(list(f1,f2))


get_orthologname_(genes[2])
```


# ------------------------------
#     4. Validate: Match Median per celltype 
# ------------------------------
```{r matchmedian}

set.seed(123)
df_lnc_celltype$type <- "lnc"
df_mrna_celltype$type <- "mrna"
df_complete<- rbind(df_lnc_celltype, df_mrna_celltype)

# Filter
# Select the ones expressed at least in 50 cells 
df_complete <- df_complete[df_complete$n_cells >= 30, ]

c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# -------------------------------
#   Select Only one celltype 
# -------------------------------
match_in_one_celltype_median <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_median <- matchit(type ~ medianexpr,c)
  matches_median <- get_matches(mi_median, c)
  
  # Calculate p-value 
  perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
  perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
  test <- wilcox.test(perc_cells_0, perc_cells_1, paired = TRUE)
  pval <- test$p.value
  
  matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
  matches_median$type <- as.factor(matches_median$type)
  
  title = paste0(celltype_test)
  scatter_plot_matching <- ggplot(matches_median, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( title=title, subtitle = paste0("n: ", n_matches, "\n", "p-value: ", pval))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))
  
  
  scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, groupColour = TRUE, groupFill = TRUE)
  scatter_plot_matching_withboxplot

}

match_in_one_celltype_perccells <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  
  # match expression 
  mi_perc <- matchit(type ~ perc_cells_expressing,c)
  matches_perc <- get_matches(mi_perc, c)
  matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
  
  median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
  median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
  test <- wilcox.test(median_0, median_1, paired = TRUE)
  pval <- test$p.value

  
  title = paste0(celltype_test)
  scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, fill = factor(type, level = c("lnc", "mrna")), col = factor(type, level = c("lnc", "mrna"))))+geom_point(size = 0.8)+labs( title=title, subtitle = paste0("n: ", n_matches, "\n", "p-value: ", pval))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"), legend.position = "none")+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 22), plot.title = element_text(size =27), plot.subtitle  = element_text(size =18))
  
  
  scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot', margins = "both",size = 3, alpha = 0.3, groupColour = TRUE, groupFill = TRUE, margMapping = aes(colour = factor(type, level = c("lnc", "mrna"))))
  scatter_plot_matching_withboxplot

}


match_in_one_celltype_median(c, "Monocyte")
match_in_one_celltype_median(c, "T")
match_in_one_celltype_median(c, "B")

match_in_one_celltype_perccells(c, "B")
match_in_one_celltype_perccells(c, "Monocyte")
match_in_one_celltype_perccells(c, "T")

match_in_one_celltype_perccells_get_median <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  
  # match expression 
  mi_perc <- matchit(type ~ perc_cells_expressing,c)
  matches_perc <- get_matches(mi_perc, c)
  matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
  n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
  
  median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
  median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
  test <- wilcox.test(median_0, median_1, paired = TRUE)
  pval <- test$p.value
  
  df_0 <- data.frame(val = median_0, type = "lnc", celltype = celltype_test, stringsAsFactors = F)
  df_1 <- data.frame(val = median_1, type = "mrna", celltype = celltype_test, stringsAsFactors = F)
  return(rbind(df_0, df_1))
}

df_medians_matched <- do.call(rbind, lapply(unique(Idents(immune.combined)), function(celltype) match_in_one_celltype_perccells_get_median(c, celltype)))


ggplot(df_medians_matched, aes(x = celltype,col = type,  fill = type, y = val))+geom_boxplot(alpha = 0.5)+theme_paper+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+ylab("Median Expression")+ggtitle("Matched by the percentage of cells expressing")




match_in_one_celltype_median_get_perc <- function(c, celltype_test){
  
  c <- c[c$ident == celltype_test,]
  # match expression 
  mi_median <- matchit(type ~ medianexpr,c)
  matches_median <- get_matches(mi_median, c)
  matches_perc$type = ifelse(matches_perc$type == 1, "lnc", "mrna")

  # Calculate p-value 
  perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
  perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
  test <- wilcox.test(perc_cells_0, perc_cells_1, paired = TRUE)
  pval <- test$p.value
  
  df_0 <- data.frame(val = perc_cells_0, type = "mrna", celltype = celltype_test, stringsAsFactors = F)
  df_1 <- data.frame(val = perc_cells_1, type = "lnc", celltype = celltype_test, stringsAsFactors = F)
  return(rbind(df_0, df_1))
}



df_perc_matched <- do.call(rbind, lapply(unique(Idents(immune.combined)), function(celltype) match_in_one_celltype_median_get_perc(c, celltype)))


# When matched by the median 
ggboxplot(df_perc_matched, col = "type",  fill = "type", y = "val",alpha = 0.5)+theme_paper+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+ylab("% Cells expressing")+ggtitle("Matched by the median")+facet_grid(cols = vars(celltype))+ stat_compare_means(aes(group=type))+theme(legend.position = "")


# When matched by the prop of cells
ggboxplot(df_medians_matched, col = "type",  fill = "type", y = "val",alpha = 0.5)+theme_paper+scale_fill_manual(values = palette_plot_percentage)+
            scale_color_manual(values = palette_plot_percentage)+ylab("Median Expression (logCP10K")+ggtitle("Matched by the median")+facet_grid(cols = vars(celltype))+ stat_compare_means(aes(group=type))+theme(legend.position = "")
```