---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# General Expression Patterns of lncRNAs Ex Vivo


```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(MatchIt)
library(rtracklayer)
library(Seurat)

col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

source("../utils/02_sc_utils.R")

# Read reference files 
ref<- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))
table(ref$gene_biotype)
lnc <- ref[ref$gene_biotype == "lncRNA",]$gene_name
pc <- ref[ref$gene_biotype == "protein_coding",]$gene_name
# Import Seurat object 
file <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/pbmc_tutorial.rds"
immune.combined <- readRDS(file)

annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc]
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% pc]
# Check that there is no overlap 
intersect(annotated_lncrnas, annotated_mrnas ) 

length(unique(annotated_lncrnas))
length(unique(annotated_mrnas))
```

# Visualize how many lnc and mrnas are present in the object 

```{r check}




#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)) ))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Dark2")
lnc_donut <- df.donut[df.donut$type == "lncRNAs",]
names(lnc_donut) <- c("-", "type", "count")
donut_plot(lnc_donut, "Greens")
```






# Calculate expression 

```{r quartiles}
# Calculate expression POOLED
# (Big, can be run on cluster)
df_lnc <- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(), tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = annotated_lncrnas), "",df_lnc, 1))
df_mrna<- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c(),tot_cells = c())
invisible(calc_mean_and_percentage_cell(subset(immune.combined, features = annotated_mrnas), "",df_mrna, 1))
#saveRDS(df_lnc, "/home/luisas/Desktop/cluster/data/RObjects_old/df_lnc.rds")
#saveRDS(df_mrna, "/home/luisas/Desktop/cluster/data/RObjects_old/df_mrna.rds")
#saveRDS(df_pseudogenes, "/home/luisas/Desktop/cluster/data/RObjects_old/df_pseudogenes.rds")
#df_lnc <- readRDS(file.path(robjectsdir, "df_lnc.rds"))
#df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))
```

```{r celltype}
df_lnc_celltype <- data.frame( gene_id = c(), meanexpr = c(), perc_cells_expressing = c(), maxexpr = c(), medianexpr = c(), var = c(), n_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_lncrnas)
invisible(lapply(as.vector(unique(Idents(subset))), function(x) calc_mean_and_percentage_cell(subset, x, df_lnc_celltype)))
# mRNA
df_mrna_celltype <- data.frame( gene_id = c(),meanexpr = c(), not_na_cells = c(), maxexpr = c(), medianexpr = c(), var = c(), var = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
invisible(lapply(as.vector(unique(Idents(subset))), function(x) calc_mean_and_percentage_cell(subset, x, df_mrna_celltype)))
#saveRDS(df_lnc_celltype, file.path(robjectsdir, "df_celltype_lnc.rds"))
#saveRDS(df_mrna_celltype, file.path(robjectsdir, "df_mrna_celltype.rds"))

#df_lnc_celltype <- readRDS(file.path(robjectsdir, "df_celltype_lnc.rds"))
#df_mrna_celltype <- readRDS(file.path(robjectsdir, "df_mrna_celltype.rds"))

df_mrna_celltype$type <- NA
df_lnc_celltype$type <-  NA
```


#----------------------------------
# 1. Visualzation: Expression across Quartiles
# ----------------------------------
```{r quartiles}
# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNAs"
df_mrna$type <- "mRNAs"
df_complete <- rbind(df_lnc, df_mrna)

plot_quartiles(df_lnc, df_mrna, "number", palette = palette_plot_percentage)
plot_quartiles(df_lnc, df_mrna, "proportion", palette = palette_plot_percentage)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())


ggboxplot(df_complete, y="perc_cells_expressing", x = "type", fill = "type", color = "type", alpha = 0.5)+
                scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)+
                stat_compare_means(comparisons = list(c("lncRNAs", "mRNAs")))+
                labs(x = "", y = "% Cells")+
                theme(legend.title = element_blank(), legend.position = "none")

```




# ------------------------------------------
# Calculate the EXPRESSION
# ------------------------------------------

```{r percentage ebov reads}
# Scatter plot both 
scatter_plot_both(df_lnc,df_mrna, y = "Cell Percentage", type = "Median",palette = palette_plot_percentage)

```


# ------------------------------
#     Match median expression 
# ------------------------------

```{r matchmedian}

set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)

# match expression 
mi_median <- matchit(type ~ medianexpr,c)
matches_median <- get_matches(mi_median, c)

# Calculate p-value 
perc_cells_0 <- matches_median[matches_median$type == 0,]$perc_cells_expressing
perc_cells_1 <- matches_median[matches_median$type == 1,]$perc_cells_expressing
t.test(perc_cells_0, perc_cells_1)


matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
matches_median$type <- as.factor(matches_median$type)
```

# Visualize matching

```{r matchmedianplot}

scatter_plot_matching <- ggplot(matches_median, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( subtitle = paste0("# Matches ", n_matches))+ theme(legend.title = element_blank())+scale_fill_manual(values = palette_plot_percentage)+
          scale_color_manual(values = palette_plot_percentage)+theme(legend.position = "none")+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+labs(x = "Median Expression", y = "% Cells expressing")+theme(legend.title = element_blank(), legend.position = "")+theme(legend.title = element_blank(), legend.position = "")+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))


scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot
palette_plot_percentage

```




# --------------------------------------
#      4. Validate : run it w/ multiple seeds
# --------------------------------------
```{r matchvalidation}
run_matchit_median <- function(c, seed){
    set.seed(seed)
    mi_median <- matchit(type ~ medianexpr,c)
    matches_median <- get_matches(mi_median, c)
    matches_median$type = ifelse(matches_median$type == 0, "mrna", "lnc")
    n_matches <- length(unique(matches_median[matches_median$type == "lnc",]$gene_id))
    matches_median$type <- as.factor(matches_median$type)
    output <- matches_median[matches_median$type  == "mrna",]$perc_cells_expressing
    df <- as.data.frame(output)
    df$seed <- as.character(seed)
    return(df)
}


#seeds <- 1:10
#list <- do.call(rbind,(lapply( seeds, function(seed) run_matchit_median(c,seed))))
#lncRNAs <- data.frame(output = c[c$type == 1,]$perc_cells_expressing, seed= as.character("0-lncRNAs"))
#lncRNAs$seed <- as.character(lncRNAs$seed)
#benchmark_df_matched_median <- rbind(lncRNAs, list)
#table(benchmark_df_matched_median$seed)

#ggplot(benchmark_df_matched_median, aes (y = output, x = seed, fill = seed, color = seed))+geom_boxplot(alpha = 0.6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+scale_fill_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+scale_color_manual(values = c("#E41A1C",rep("#377EB8", 10) ))+labs(y = "% Cells", Title = "")+theme(legend.position = "")

```


# ------------------------------
#     Match perc cells
# ------------------------------

```{r matchperc}
# matched perc cells
set.seed(123)
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, c)
matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))

median_0 <- matches_perc[matches_perc$type == "lnc",]$medianexpr
median_1 <- matches_perc[matches_perc$type == "mrna",]$medianexpr
t.test(median_0, median_1)

```

# Visualize
```{r matchperc}

scatter_plot_matching <- ggplot(matches_perc, aes(x=medianexpr, y =perc_cells_expressing, col = as.factor(type)))+
                         geom_point(size = 0.8, alpha = 0.8)+labs(title = "", subtitle = paste0("# Matches ", n_matches))+
                         theme(legend.title = element_blank(),
                               legend.position = "none",
                               panel.background = element_rect(fill = "white", colour = "grey50"),
                               axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))+
                         scale_fill_manual(values =palette_plot_percentage)+
                         scale_color_manual(values = palette_plot_percentage)+
                         labs(x = "Median Expression", y = "% Cells expressing")+theme()

scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.5, ,groupColour = TRUE, groupFill = TRUE)
scatter_plot_matching_withboxplot


```












