---
title: "Specificity"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# Specificity 

### Import 
```{r include=FALSE}
library(Seurat)
library(gtools)
library(ggplot2)
library(ggpubr)
library(rtracklayer)
library(MatchIt)
library(gridExtra)
library(reshape2)


theme_sc_paper <- theme(panel.background = element_rect(fill = "white", colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())+theme_paper


# Set themes and palette
theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
theme_sc <- theme(axis.title.x.top = element_text())+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(axis.title = element_text(size=15), plot.title = element_text(size =22, hjust = 0.5))

col_lnc = "#f72631"
col_mrna = "#3153a3"
palette_plot_percentage <- c(col_lnc, col_mrna)

# Load objects
robjectsdir <- file.path(data_path,"02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs//")
all_lncrnas <- readRDS(file.path(robjectsdir, "05_stats/all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir,"05_stats/annotated_mrnas.rds"))

# Select object on which to calculate Specificity score
# Neutrophils or not? 
#file <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
#immune.combined <- readRDS(file)

# Load stats ( about # cells, and median expression )
df_lnc <- readRDS(file.path(robjectsdir,"05_stats/df_lnc.rds"))
df_lnc_celtype <- readRDS(file.path(robjectsdir,"05_stats/df_celltype_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir,"05_stats/df_mrna.rds"))
df_mrna_celltype <- readRDS(file.path(robjectsdir,"05_stats/df_mrna_celltype.rds"))
df_n_cells <- rbind(df_lnc, df_mrna)
df_celltype <- rbind(df_mrna_celltype,df_lnc_celtype )
#saveRDS(df_celltype, file.path(robjectsdir,"df_celltype.rds"))


df_lnc <- df_lnc[df_lnc$n_cells > 60, ]
df_mrna <- df_mrna[df_mrna$n_cells > 60, ]



range01 <- function(x){(x-min(x))/(max(x)-min(x))}


get_df_summary <-function(spec_res){
  rownames(spec_res) <- spec_res$gene
  spec_res$type <- "other"
  spec_res[all_lncrnas,]$type<- "lnc"
  spec_res[annotated_mrnas,]$type<- "pc"
  spec_res$log <- log(spec_res$score)

  spec_res_int <- spec_res[spec_res$type %in% c("lnc", "pc"),]
  df_test <- merge(spec_res_int, df_n_cells, by = "row.names")
  
  # Only keep genes that show expression in at least 30 cells.
  df_test <- df_test[df_test$n_cells > 30, ]
  df_test <- df_test[df_test$log != "Inf",]
  df_test <- df_test[!is.na(df_test$Row.names),]
  
  rownames(df_test) <- df_test$`Row.names`
  return(df_test)
}

calc_mean_and_percentage_cell <- function(expression_matrix_permuted,ident = "", df, threshold = 1 ){
  
  # When not calculated on a specific cell-type, ident == ""
  # Otherwise it corresponds to the cell-type 
  # Only subsetting when doing the calculation on a specific cell-type
  if(ident != ""){
    expression_matrix <- expression_matrix_permuted[,colnames(expression_matrix_permuted)==ident]
  }
  
  # Selecting normalized data 
  # expression_matrix <- as.matrix(subset@assays$RNA@data)
  
  
  # -------------------- TOTAL NUMBER OF CELLS ----------------------
  # Cells are columns and genes are rows.
  # The total number of cells is the same for each gene. 
  # Tot_cells:= a vector as long as the number of genes present in the matrix. 
  # All values are the same: the total number of columns ( cells ) in the matrix. 
  tot_cells <- rep(ncol(expression_matrix), nrow(expression_matrix))
  
  
  # ---------------- NUMBER OF CELLS SHOWING EXPRESSION ----------------
  # Count the number of cells per gene showing some expression
  # Only select cells that are expressed (> threshold) 
  # Assign NAs to other cells, just for ease of calculation later
  # and compute the number of cells that show some expression 
  expression_matrix[expression_matrix < threshold] <- NA
  not_na_cells <- tot_cells- rowCounts(expression_matrix, value = NA)
  perc_cells_expressing <- not_na_cells/tot_cells
  
  
  # ---------------- CALCULATE MEAN, MEDIAN, MAX, MIN EXPRESSION and VARIANCE ------------
  # Per gene, calculate the above mentioned metrics, EXCLUSIVELY on the 
  # cells where the gene shows expression 
  meanexpr <- rowMeans(expression_matrix, na.rm = TRUE)
  medianexpr <- rowMedians(expression_matrix, na.rm = TRUE)
  maxexpr <- rowMaxs(expression_matrix, na.rm = TRUE)
  minexpr <- rowMins(expression_matrix, na.rm = TRUE)
  var <- rowVars(expression_matrix, na.rm = TRUE)
  
  # Vector of all gene ids 
  gene_id <- rownames(expression_matrix)
  
  # Save 
  dfo <- deparse(substitute(df))
  df_new <- data.frame(gene_id = gene_id, meanexpr = meanexpr,medianexpr = medianexpr, perc_cells_expressing = perc_cells_expressing, maxexpr = maxexpr, var = var, n_cells = not_na_cells, tot_cells
                       = tot_cells)
  df_new$minexpr <- minexpr
  df_new$ident <- as.factor(ident)
  df_complete <- rbind(df, df_new)
  assign(dfo, df_complete, envir = .GlobalEnv);
}

```


# -------------------------------------------------
# -------------------------------------------------
#             FuNCTIONS
# -------------------------------------------------
# -------------------------------------------------

# Match by the percentage of cells expressing
```{r matchplot}


match_plots <- function(df_test, title = "specificity score"){
  set.seed(123)
  c <- df_test[complete.cases(df_test),]
  c$type = ifelse(c$type == "lnc", 1, 0)
  # Match % cells 
  mi_perc <- matchit(type ~ perc_cells_expressing,c)
  matches_perc <- get_matches(mi_perc, data = c)
  matches_perc$type = ifelse(matches_perc$type == 0, "mRNA", "lncRNA")
  n_matches <- length(unique(matches_perc[matches_perc$type == "lncRNA",]$gene_id))
  matches_perc$type <- as.factor(matches_perc$type)
  
  # Plot 
  #scatter_plot_matching <- ggplot(matches_perc, aes(x=score, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( subtitle = paste0("# Matches ", n_matches))+scale_fill_manual(values = palette_plot_percentage)+theme(legend.position = "")+scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))+xlab("Specificity Score")+ylab("Prop cells")
  #scatter_plot_matching_withboxplot<- ggExtra::ggMarginal(scatter_plot_matching,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3,groupColour = TRUE, groupFill = TRUE)
  #scatter_plot_matching_withboxplot
  
  scatter_plot_matching <- ggplot(matches_perc, aes(x=score, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = title, y = "% cells")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =score, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  #+stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  p <- ggdraw(p2)
  
  test1 <- wilcox.test(matches_perc[matches_perc$type == "lncRNA", ]$score, matches_perc[matches_perc$type == "mRNA", ]$score, paired = T)
  
  
  # -------------------- matched no scatter -------------------
  my_comparisons <- list( c("lncRNA", "mRNA"))
  boxplot_cells_matched <- ggboxplot(matches_perc, x = "type", y = "score", fill = "type", col = "type",alpha=0.7)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab(title)+theme(text = element_text(size=15))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+theme(plot.title = element_text(hjust = 0.5))
  
  
  
  
  # .-------------------------------------------------------------------
  # Match by the score
  c <- df_test[complete.cases(df_test),]
  #c$type = ifelse(c$type == "lncRNAs", 1, 0)
  c$type = ifelse(c$type == "lnc", 1, 0)
  
  # match expression 
  mi_median <- matchit(type ~ score,c)
  matches_score <- get_matches(mi_median, data = c)
  
  matches_score$type = ifelse(matches_score$type == 0, "mRNA", "lncRNA")
  n_matches <- length(unique(matches_score[matches_score$type == "lncRNA",]$gene_id))
  matches_score$type <- as.factor(matches_score$type)
  
  
   scatter_plot_matching <- ggplot(matches_score, aes(x=score, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.6)+
                            labs( subtitle = paste0("# matches ", n_matches))+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching+labs(x = title, y = "% cells")
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_score, aes(x=factor(type), y =score, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  #+stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_score, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p3 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p4 <- insert_yaxis_grob(p3, ybox, grid::unit(1, "in"), position = "right")
  p5 <- ggdraw(p4)
  
  test2 <- wilcox.test(matches_score[matches_score$type == "lncRNA", ]$perc_cells_expressing, matches_score[matches_score$type == "mRNA", ]$perc_cells_expressing, paired = T)
  
    my_comparisons <- list( c("lncRNA", "mRNA"))
  boxplot_score_matched <- ggboxplot(matches_score, x = "type", y = "perc_cells_expressing", fill = "type", col = "type",alpha=0.7)+theme()+scale_fill_manual(values = (palette_plot_percentage))+scale_color_manual(values = (palette_plot_percentage))+xlab("")+ylab("% cells")+theme(text = element_text(size=15))+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank()) + theme(axis.text.x = element_text(size=10), legend.position = "")+stat_compare_means(comparisons = my_comparisons, label = "p.signif")+theme(plot.title = element_text(hjust = 0.5))

  # Plot 
  #scatter_plot_matching2 <- ggplot(matches_score, aes(x=score, y =perc_cells_expressing, col = as.factor(type)))+geom_point(size = 0.8)+labs( subtitle = paste0("# Matches ", n_matches))+scale_fill_manual(values = palette_plot_percentage)+
 #           scale_color_manual(values = palette_plot_percentage)+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 22), legend.text = element_text(size = 18))+theme(legend.position = "")+ylab("Prop cells")
  
  #scatter_plot_matching_withboxplot2<- ggExtra::ggMarginal(scatter_plot_matching2,type = 'boxplot',margins = "both",size = 3,colour = "darkgrey", alpha = 0.3,groupColour = TRUE, groupFill = TRUE)
  #scatter_plot_matching_withboxplot2
  
  return(list(p, p5, test1, test2, boxplot_cells_matched, boxplot_score_matched))
}

stats_specificity <- function(specificity_scores, plots = F, title = "specificity score"){

  # -----------------------------
  #     Load real Specificity Scores
  # -----------------------------
  specificity_scores_summary <- get_df_summary(specificity_scores)
  specificity_scores_summary$permutation <- "Real"
  specificity_scores_summary$type <- ifelse(specificity_scores_summary$type == "lnc", "lncRNA", "mRNA")
  df_tot <- specificity_scores_summary
  
  
  # Extract genes to highglight
  housekeeping_genes_3k <- specificity_scores_summary[housekeeping_genes, ][!is.na(specificity_scores_summary[housekeeping_genes, ]$n_cells) & specificity_scores_summary[housekeeping_genes, ]$n_cells>3000,]$gene
  highlight_df <- specificity_scores_summary[rownames(specificity_scores_summary) %in% marker.genes, ]
  highlight_df_hk <- specificity_scores_summary[rownames(specificity_scores_summary) %in% housekeeping_genes_3k, ]
  
  # ------------------ PLOT 1: permutaiton and markers ------------------------------
  # Compare Lnc and PC between permuted and real 
  my_comparisons <- list( c("lncRNA", "mRNA"))
  df_tot$type <- factor(df_tot$type, levels = c("lncRNA", "mRNA"))
  p1b <- ggboxplot(df_tot, x = "type", y = "score",  col="black", fill = "type", alpha= 1, short.panel.labs = T) +    scale_color_manual(values = (palette_plot_percentage))+
                  scale_fill_manual(values = (palette_plot_percentage))+
                  theme(legend.position = "none")+
                  stat_compare_means(comparisons = my_comparisons, label = "p.signif")+xlab("")+ylab(title)+
                  theme(axis.ticks.x = element_blank(), text = element_text(size = 15), axis.line.x = element_blank())
  

  
  # Add Markers
  p1 <- p1b+geom_point(data=highlight_df, aes(x=type,y=score), color='#02A237',size=3)+
    geom_point(data=highlight_df, aes(x=type,y=score), color='black',size=0.5)+
    geom_point(data=highlight_df_hk, aes(x=type,y=score), color='dark grey',size=3)+
    geom_point(data=highlight_df_hk, aes(x=type,y=score), color='black',size=0.5)
  p1
  
  
  
  # ------------------ PLOT 2: Visualize relation to % of cells  ------------------------------
  p2 <- ggscatter(specificity_scores_summary, x = "score", y = "perc_cells_expressing",  conf.int = T, 
            cor.coef = TRUE, cor.method = "spearman", size=0.3, col = "#999999",add.params = list(color = "red", size = 0.3))+theme_sc+labs(x = "Specificity Score", y = "Number of cells")+theme_paper
  
  p2

  
  # ------------------ PLOT 4: Examples  ------------------------------
  
  if(plots == T){
      # Create subsets for plotting
    df_test_max <- specificity_scores_summary[specificity_scores_summary$score %in% tail(sort(specificity_scores_summary$score),6), ]
    df_test_min <- specificity_scores_summary[specificity_scores_summary$score %in% head(sort(specificity_scores_summary$score),6), ]
    df_test_max_lnc <- specificity_scores_summary[specificity_scores_summary$type == "lncRNAs",]
    df_test_max_lnc <- df_test_max_lnc[df_test_max_lnc$score %in% tail(sort(df_test_max_lnc$score),6),]
    df_test_max_pc <- specificity_scores_summary[specificity_scores_summary$type == "PC genes",]
    df_test_max_pc <- df_test_max_pc[df_test_max_pc$score %in% tail(sort(df_test_max_pc$score),6),]
    
    # Plot top specificity genes in LNC
    p4 <- lapply(as.character(df_test_max_lnc$gene), function(gene) FeaturePlot(immune.combined, gene, pt.size=0.8, order = T, cols= c("gray", col_lnc))+theme_sc+labs(title = gsub("-unknown","", gene)))
    print("aalskh")
    # Plot top specificity genes in PC
    p5 <- lapply(as.character(df_test_max_pc$gene), function(gene) FeaturePlot(immune.combined, gene, pt.size=0.8, order = T, cols = c("gray", col_mrna ))+theme_sc+labs(title = gsub("-unknown","", gene)))
    print("aalskh")
    # The ones that have the min specificity score
    p6 <- lapply(as.character(df_test_min$gene), function(gene) FeaturePlot(immune.combined, gene, order = T )+theme_sc+theme_sc+labs(title = gsub("-unknown","", gene)))
    print("aalskh")
    return(list(p1,p2,p3,p4,p5,p6))
  }else{
    return(list(p1,p2))
  }

}

```

# -----------------------------------
#        TAU  - IN VIVO 
# -----------------------------------
```{r matchmedian}

tau <-  readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/03_specificity_tau.rds"))
colnames(tau) <- gsub("tau", "score", colnames(tau))
tau$gene <- rownames(tau)
tau <- tau[tau$gene %in% c(df_lnc$gene_id, df_mrna$gene_id),]


# Plot Stats
stats_tau <- stats_specificity(tau, plots = F, title = "repurposed tau score")

pdf(file.path(plots, "02/replication33k/SUPPL_spec_matching_cells_tau.pdf"), width = 6, height = 6)
match_plots(get_df_summary(tau), title  = "repurposed tau score")[[1]]
dev.off()

pdf(file.path(plots, "02/replication33k/SUPPL_sepc_matching_score_tau.pdf"), width = 6, height = 6)
match_plots(get_df_summary(tau), title = "repurposed tau score")[[2]]
dev.off()


match_plots(get_df_summary(tau))[[3]]
match_plots(get_df_summary(tau))[[4]]


pdf(file.path(plots, "02/replication33k/SUPPL_spec_tau_boxplotmarkers.pdf"), width = 3, height = 4)
stats_tau[[1]]
dev.off()



pdf(file.path(plots, "02/replication33k/SUPPL_specREP_tau_boxplotmatched1.pdf"), width = 3, height = 4)
match_plots(get_df_summary(tau), title = "repurposed tau score")[[5]]

dev.off()


pdf(file.path(plots, "02/replication33k/SUPPL_specREP_tau_boxplotmatched2s.pdf"), width = 3, height = 4)
match_plots(get_df_summary(tau), title = "repurposed tau score")[[6]]

dev.off()


match_plots(get_df_summary(tau), title = "repurposed tau score")

```




# --------------------------------------------
# --------------------------------------------
#               Approach 1 b
#      simple % cell ranged
# --------------------------------------------
# --------------------------------------------


```{r matchmedian}

specificity_scores <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/03_validation_pbmcs_external/01_10xGenomic_10kPBMCs/04_specificity_alternativescore.rds"))
specificity_scores<- specificity_scores[!is.na(specificity_scores$gene),c("gene", "score")]
rownames(specificity_scores) <- as.character(specificity_scores$gene)


specificity_scores <- specificity_scores[specificity_scores$gene %in% c(df_lnc$gene_id, df_mrna$gene_id),]

# Plot Stats
alt_stats <- stats_specificity(specificity_scores, plots = F, title = "alternative specificity score")

pdf(file.path(plots, "02/replication33k/SUPPL_specREP_alt_boxplotmarkers.pdf"), width = 3, height = 4)
alt_stats[[1]]
dev.off()


pdf(file.path(plots, "02/replication33k/SUPPL_specREP_matching_cells_alt.pdf"), width = 6, height = 6)
match_plots(get_df_summary(specificity_scores), title = "alternative specificity score")[[1]]
dev.off()

pdf(file.path(plots, "02/replication33k/SUPPL_sepREP_matching_score_alt.pdf"), width = 6, height = 6)
match_plots(get_df_summary(specificity_scores), title = "alternative specificity score")[[2]]
dev.off()


match_plots(get_df_summary(specificity_scores))[[3]]
match_plots(get_df_summary(specificity_scores))[[4]]


pdf(file.path(plots, "02/replication33k/SUPPL_sepREP_matching_score_altBoxplot1.pdf"), width = 3, height = 4)
match_plots(get_df_summary(specificity_scores), title = "alternative specificity score")[[5]]
dev.off()


pdf(file.path(plots, "02/replication33k/SUPPL_sepREP_matching_score_altBoxplot2.pdf"), width = 3, height = 4)
match_plots(get_df_summary(specificity_scores), title = "alternative specificity score")[[6]]
dev.off()



```



