---
title: "Specificity"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# Simulate single-cell dataset 

### simulation

```{r include=FALSE}
library(Seurat)
library(multiplyr)
# Set-up general information 
# Global variable 
celltypes <- identities <- c("B", "T", "M") 
props <- c(0.2, 0.5, 0.3)
tot_cells <- 40000

p <- c("#A30808", "#ff0000", "#ff4000", "#ff8000", "#ffbf00", "#CBCA1C",  "#6C8DBF", "#1D56AB", "#23265A")

# Row: genes 
# Column: cell-types
colnames <- unlist(lapply(1:length(celltypes), function(i) rep(celltypes[i], props[i]*tot_cells)))

# Simulate an expression Vector (For one celltype)
# n = Length of output vector
# p = Proportion of cells showing expression
# v = value of expression 
get_expr_vector_celltype <- function(p, n, v = 2){
  expressed <- rep(v, p*n)
  not_expressed <- rep(0, (1-p)*n)
  return(c(expressed, not_expressed))
}

# pc = proportion of cells expressed in that celltype 
# Real cell-type proportions
# total number of cells
# Genes expressed in a cell-type only pc <- c(1,0,0) or c(0,1,0) or c(0,0,1)
get_expression_vector <- function(pc, props.=props, tot_cells.=tot_cells){
  return(unlist(lapply(1:length(pc), function(i) get_expr_vector_celltype(pc[i],props.[i]*tot_cells.))))
}

# Genes expressed in a cell-type only
genes_proportions <- list(c(1,0,0),  #Highly specific:100%
                          c(0,1,0),  #Highly specific:100%
                          c(0,0,1),  #Highly specific:100%
                          
                          c(0.5,0,0),  #Highly specific:50%
                          c(0,0.5,0),  #Highly specific:50%
                          c(0,0,0.5),  #Highly specific:50%
                          
                          c(0.2,0,0),  #Highly specific:20%
                          c(0,0.2,0),  #Highly specific:20%
                          c(0,0,0.2),  #Highly specific:20%
                          
                          c(0.5,0.1,0.1), #Medium specific:50%, 10%, 10%
                          c(0.1,0.5,0.1), #Medium specific:10%, 50%, 10%
                          c(0.1,0.1,0.5), #Medium specific:50%, 10%, 50%
                          
                          c(0.5,0.5,0.1), #Medium-low specific:50%, 50%, 10%
                          c(0.5,0.1,0.5), #Medium-low specific:10%, 50%, 50%
                          c(0.1,0.5,0.5), #Medium-low specific:50%, 10%, 50%
                          
                          c(0.5,0.1,0), #Medium-high specific
                          c(0.5,0.1,0), #Medium-high specific
                          c(0.1,0.5,0), #Medium-high specific
                          c(0,0.5,0.1), #Medium-high specific
                          c(0.1,0,0.5), #Medium-high specific
                          c(0,0.1,0.5), #Medium-high specific
                          
                          c(0.1,0.1,0.1), #Low specific: 10%
                          c(0.5,0.5,0.5), #Low specific: 50%
                          c(1,1,1))       #Low specific: 100%

# Case1: super specific. All cells of one cell-type. (One per celltype)
# Case2: super specfici.  0.5 cells of one cell-type. (One per celltype)
# Case3: super specfici.  0.2 cells of one cell-type. (One per celltype)

labels <- c(rep("High",9), rep("Medium", 3), rep("Medium",3), rep("Medium", 6), rep("Low", 3))
labels <- factor(labels, levels= c("Low", "Medium", "High"))
labels_prop <- c(rep("100", 3), rep("50", 3), rep("20",3), rep("-",3),  rep("low",3),rep("high",6), "10", "50", "100")
labels_full <- paste(labels, labels_prop)
labels_full  <- factor(labels_full, levels=c("High 100","High 50","High 20", "Medium high","Medium -","Medium low","Low 100","Low 50", "Low 10"  ))


genes <- unlist(lapply(genes_proportions, get_expression_vector))
n_genes <- length(genes)/tot_cells

# Finally create the matrix for calculation 
m <- matrix(genes,
            nrow = n_genes,
            ncol = tot_cells,
            byrow = T, 
            dimnames = list(1:n_genes, colnames))

# Create and save seurat object 
m_seurat <- CreateSeuratObject(
  m,
  project = "Simulations",
  assay = "RNA"
)
Idents(m_seurat) <- colnames
#saveRDS(m_seurat, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/04_simulation/m_seurat.rds")


# Create summary statistics 
# Gene, Celltype, n_cells, tot_cells
names <- c("gene_id", "Celltype", "n_cells", "tot_cells")

get_df_celltype_pergene <- function(gene_index){
  df <- Reduce("rbind", lapply(1:length(celltypes),function(i) data.frame(rownames(m)[gene_index], celltypes[i], unlist(genes_proportions[gene_index])[i]*(props*tot_cells)[i], (props*tot_cells)[i], stringsAsFactors = F)))
  names(df) <-names
  return(df)
}

df_celltype <- Reduce("rbind", lapply(1:nrow(m),get_df_celltype_pergene))

# Add 1 to any cell where expressin is not shown 
#saveRDS(df_celltype, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/04_simulation/df_celltype.rds")


```


# --------------------------------------------
# --------------------------------------------
#             TAU 
# --------------------------------------------
# --------------------------------------------


```{r matchmedian}
library(tispec)

mean_exp <- function(expression_matrix, gene, ident){
  # Extract only the expression of the gene in the cell-type which we aer analyzing
  expression <- as.vector(expression_matrix[gene,colnames(expression_matrix) == ident])
  # Calculate the mean expression
  
  mean_expression <- mean(expression, na.rm=F)
  if(is.na(mean_expression)){ mean_expression <- 0}
  # Weight the mean:  Multiply the expression vector by the fraction of cells expressed in each cell tyoe
  return(mean_expression)
  #return(mean_expression)
}


get_gene_mean_expressions <- function(gene, m){
  df <- data.frame( value = unlist(lapply(identities, function(ident) mean_exp(m, gene, ident))), ident = identities, stringsAsFactors = F ) 
  return(unlist(lapply(identities, function(ident) mean_exp(m, gene, ident))))
}

# Now apply TAU 
m_exp <-matrix(Reduce("cbind",lapply(rownames(m), function(gene) get_gene_mean_expressions(gene, m))), ncol = length(identities), byrow = T)
colnames(m_exp) <- identities
rownames(m_exp) <- rownames(m)
m_exp

log2Exp <- log2Tran(m_exp)
qnExp <- quantNorm(log2Exp)
#m_exp <- qnExp

df <- as.data.frame(m_exp)
rownames(df) <- rownames(m)
tauExp <- calcTau(as.data.frame(m_exp)) 

test <- m_exp
rownames(test) <- labels_full

tauExp <- tauExp[order(as.double(rownames(tauExp))),]
specificity_scores  <- tauExp
specificity_scores$gene <- rownames(specificity_scores)
specificity_scores$score <- specificity_scores$tau
specificity_scores$label <- labels
specificity_scores$label_prop <- labels_prop
specificity_scores$label_full <- labels_full

specificity_scores$gene <- factor(specificity_scores$gene, levels = specificity_scores$gene[rev(order(specificity_scores$score))])

pdf(file.path(plots, "02/SUPPL_SPECSCORE_tau.pdf"), width = 8, height = 6)
ggplot(specificity_scores[order(specificity_scores$score),], aes(y = score, x = gene, col = label_full ))+geom_point(size = 6)+theme_paper+theme(axis.text.x = element_blank(), axis.ticks = element_blank())+scale_color_manual(values=p)+ylab("reproposed tau score")+ xlab("simulated gene expression")
dev.off()
```



# --------------------------------------------
# --------------------------------------------
#             PROPO modified 
# --------------------------------------------
# --------------------------------------------


```{r matchmedian}
df_celltype$observed <- df_celltype$n_cells

  
get_score_prop_mod <- function(df_celltype, gene){
  df_celltype_gene <- df_celltype[df_celltype$gene_id == gene, ]
  df_celltype_gene$prop <- df_celltype_gene$tot_cells/sum(df_celltype_gene$tot_cells)
  # Calculate the expected 
  df_celltype_gene$expected <- sum(df_celltype_gene$n_cells)*df_celltype_gene$prop
  df_celltype_gene$obs_prop <- df_celltype_gene$observed/sum(df_celltype_gene$observed)
  
  calc_score_1  <- function(df_celltype_gene, ident){
    df_celltype_gene_celltype <- df_celltype_gene[df_celltype_gene$Celltype == ident,];  
    s <- (df_celltype_gene_celltype$obs_prop -df_celltype_gene_celltype$prop)
    return(data.frame(score = s, identity = ident,prop = df_celltype_gene_celltype$prop, stringsAsFactors = F))
  }
  score <- Reduce("rbind", (lapply(identities, function(ident){calc_score_1(df_celltype_gene, ident)} ))) 
  score$gene <- gene
  return(score)
}

specificity_scores <- Reduce("rbind", lapply(unique(df_celltype$gene_id), function(gene) get_score_prop_mod(df_celltype, gene)))
#specificity_scores$label_full <- labels_full
# I now have per gene 3 scores 
# Select the row with the maximum
specificity_scores_filt <- specificity_scores %>% 
  dplyr::group_by(gene) %>% 
   dplyr::arrange(desc(score)) %>% 
   dplyr::slice(1) %>% 
   dplyr::ungroup()


df_labels <- data.frame(label_full = labels_full, gene = rownames(m), stringsAsFactors = F)
specificity_scores_filt <- merge(specificity_scores_filt, df_labels, by="gene")
specificity_scores_filt$gene <- factor(specificity_scores_filt$gene, levels = specificity_scores_filt$gene[rev(order(specificity_scores_filt$score))])
range01 <- function(x){(x-min(x))/(max(x)-min(x))}

range_celltype_score <- function(specificity_scores_filt, ident){
  score_celltype <- specificity_scores_filt[specificity_scores_filt$identity == ident,]
  score_celltype$score_ranged <- range01(score_celltype$score)
  return(score_celltype)
}

specificity_scores_filt <- Reduce("rbind", lapply(unique(specificity_scores_filt$identity), function(ident) range_celltype_score(specificity_scores_filt, ident)))



ggplot(specificity_scores_filt[order(specificity_scores_filt$score),], aes(y = score, x = gene, col = label_full ))+geom_point(size = 6)+theme_paper+theme(axis.text.x = element_blank(), axis.ticks = element_blank())+scale_color_manual(values=p)+ggtitle("Not ranged")


specificity_scores_filt$gene <- factor(specificity_scores_filt$gene, levels = specificity_scores_filt$gene[rev(order(specificity_scores_filt$score_ranged))])

pdf(file.path(plots, "02/SUPPL_SPECSCORE_alternative.pdf"), width = 8, height = 6)
ggplot(specificity_scores_filt[order(specificity_scores_filt$score_ranged),], aes(y = score_ranged, x = gene, col = label_full ))+geom_point(size = 6)+theme_paper+theme(axis.text.x = element_blank(), axis.ticks = element_blank())+scale_color_manual(values=p)+ylab("alternative specificity score")+xlab("simulated gene expression")
dev.off()



```


