
#  Benchmark
## Benchmarking CPC2, CNIT and CPAT lncRNAs prediction on rheMac10 reference 


#### Imports
```{r include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer);  library(stringr); library(ggplot2); library(grid); library(gridExtra); library(RColorBrewer); library(readr); library(matrixStats)
library(GenomicRanges); library(dplyr); library(zeallot); library(tximport); library(ggpubr); library(plyr); 
library(formattable)
library(DT)
library("htmltools")
library("webshot")    

export_formattable <- function(f, file, width = "100%", height = NULL, 
                               background = "white", delay = 0.2)
    {
      w <- as.htmlwidget(f, width = width, height = height)
      path <- html_print(w, background = background, viewer = NULL)
      url <- paste0("file:///", gsub("\\\\", "/", normalizePath(path)))
      webshot(url,
              file = file,
              selector = ".formattable_widget",
              delay = delay)
    }

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))
```

#### Read all the prediciton and reference files

```{r readPredictions}

# rheMac10 reference (download)
# ftp://ftp.ensembl.org/pub/release-100/gtf/macaca_mulatta/
ref <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf"))

# Table correspondence between transcripts and gene IDs 
df_ref <- distinct(data.frame(ref[!is.na(ref$transcript_id),]$transcript_id, ref[!is.na(ref$transcript_id),]$gene_id, stringsAsFactors = F))
names(df_ref) <- c("transcript_id", "gene_id")
rownames(df_ref) <- df_ref$transcript_id

ref_lnc <- ref[ref$transcript_biotype %in% c("lncRNA", "antisense"),]
ref_lnc <- ref_lnc[ref_lnc$type == "transcript", ]
ref_pc <- ref[ref$type == "transcript" & ref$transcript_biotype == "protein_coding",]

# ----------------------
#   CPC2 predictions
# ----------------------
cpc2 <- read.table(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/99_benchmark_annotation/01_predictions/CPC2/cpc2_pred.txt"))
names(cpc2) <- c("ID","transcript_length ","peptide_length","Fickett_score" , " pI", "ORF_integrity", "coding_probability","label")
# Extract transcript ids 
cpc2$transcript_id <- unlist(lapply(as.character(cpc2$ID), function(x) strsplit(x,"[(]")[[1]][1]))

# ----------------------
#   CPAT predictions
# ----------------------
cpat <- read.table(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/99_benchmark_annotation/01_predictions/CPAT/cpat_pred.ORF_prob.tsv"), header = T)
# Threshold reported by https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3616698/
# "score threshold of 0.364 gave the highest sensitivity and specificity (0.966 for both) for human data"
cutoff <-  0.364
cpat$label <- ifelse(cpat$Coding_prob > cutoff, "coding", "noncoding")

# ----------------------
#   CNIT predictions
# ----------------------
cnit <- read.table(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/99_benchmark_annotation/01_predictions/CNIT/cnit_pred/CNCI2.index"), header = T, sep="\t", stringsAsFactors = F)
# Compare which ones and how many overlap 
cnit$label <- cnit$index
cnit$transcript_id <- unlist(lapply(as.character(cnit$Transcript.ID), function(x) strsplit(x,"[(]")[[1]][1]))



# How many of the known lncRNAs (TRANSCRIPTS) are identified as such 
noncoding_cpc2 <- unlist(lapply(cpc2[cpc2$label == "noncoding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))
coding_cpc2 <- unlist(lapply(cpc2[cpc2$label == "coding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))

noncoding_cpat <- unlist(lapply(cpat[cpat$label == "noncoding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))
coding_cpat <- unlist(lapply(cpat[cpat$label == "coding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))

noncoding_cnit <- unlist(lapply(cnit[cnit$index == "noncoding",]$Transcript.ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))
coding_cnit <- unlist(lapply(cnit[cnit$index == "coding",]$Transcript.ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))

```



```{r tableStats}
# ---------------------------------------------------------------------------------------------------------------
# Given a set of predicted lncRNAs and predicted PC genes, compares it with the references (ref_lnc, ref_pc)
# Calculates sensitivity, specificity and accuracy of a se
# Name: name of the tool, appears in the resulting dataframe
# ---------------------------------------------------------------------------------------------------------------

calc_stats_prediction <- function(noncoding_pred,coding_pred, name, ref_lnc.=ref_lnc, ref_pc.=ref_pc){
  # Predicted as lnc and and are lnc
  TP <- length(unique(noncoding_pred[noncoding_pred %in% ref_lnc.$transcript_id ]))
  print(TP)
  # Predicted as pc but are lnc
  FN <- length(unique(coding_pred[coding_pred %in% ref_lnc.$transcript_id ]))
  print(FN)
  # Predicted as pc and are pc
  TN <- length(unique(coding_pred[coding_pred %in% ref_pc.$transcript_id ]))
  # Predicted as lnc but are pc
  FP <- length(unique(noncoding_pred[noncoding_pred %in% ref_pc.$transcript_id ]))
  sensitivity <- (TP/(FN+TP))*100
  specificity <- (TN/(TN+FP))*100
  accuracy <- ((TP + TN)/(TP + TN + FP + FN))*100 
  df <- data.frame(name = name, sensitivity = sensitivity, specificity = specificity, accuracy = accuracy, stringsAsFactors = F)
  return(df)
}

# Intersection of the predictions
noncoding_intersection <- unique(intersect(noncoding_cpc2,intersect(noncoding_cnit, noncoding_cpat)))
coding_intersection <- unique(ref[!(ref$gene_id %in% noncoding_intersection), ]$transcript_id)

# Union of the predictions
noncoding_union <- unique(union(noncoding_cpc2,union(noncoding_cnit, noncoding_cpat)))
coding_union <- unique(ref[!(ref$gene_id %in% noncoding_intersection), ]$transcript_id)

cpc2_stats <- calc_stats_prediction(noncoding_cpc2, coding_cpc2, "CPC2")
cpat_stats <- calc_stats_prediction(noncoding_cpat, coding_cpat, "CPAT")
cnit_stats <- calc_stats_prediction(noncoding_cnit, coding_cnit, "CNIT")
intersection_stats <- calc_stats_prediction(noncoding_intersection, coding_intersection, "Intersection")
union_stats <- calc_stats_prediction(noncoding_union, coding_union, "Union")

stats_summary <- rbind(cpc2_stats, cpat_stats, cnit_stats, intersection_stats, union_stats)

# Visualize stats in a table
table <- formattable(stats_summary, 
            align = c("l",rep("r", NCOL(stats_summary) - 1)),
            list(`name` = formatter("span", style = ~ style(color = "grey", font.weight = "bold")), 
                 area(col = c(accuracy)) ~ normalize_bar("#E5CCFF", 0.2)))
table
export_formattable(table,"Fig_S1A.png")
```



# Upset plot
```{r upset}
library(UpSetR)

# All lncRNAs annotated
annotated_lnc <- unique(ref_lnc$transcript_id)
length(unique(annotated_lnc))

# Upset Plot
listInput <- list(cpat = unique(noncoding_cpat[noncoding_cpat %in% annotated_lnc]), CPC2 = unique(noncoding_cpc2[noncoding_cpc2 %in% annotated_lnc]), CNIT = unique(noncoding_cnit[noncoding_cnit %in% annotated_lnc]),  annotated = annotated_lnc )

upset(fromList(listInput), order.by = "freq", main.bar.color = c("#6600cc", rep("gray23",7)),  text.scale = c(2.5, 2.5, 2.5, 1.0, 2, 2.15))


# Also export 
pdf(file = "Fig_S1B.pdf",   # The directory you want to save the file in
    width = 7, # The width of the plot in inches
    height = 7)

upset(fromList(listInput), order.by = "freq", main.bar.color = c("#6600cc", rep("gray23",7)),  text.scale = c(2.5, 2.5, 2.5, 1.0, 2, 2.15))
dev.off()
```


