---
title: "01_lncrnas_plots_paper"
author: "Luisa Santus"
date: "4/23/2020"
output: html_document
  keep_md: true
---

## 01 lncRNAs annotation summary

#test 111

### Summary of features of novel lncRNAs compared to human and macaque annotate lncRNAs and mRNAs

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer);  library(stringr); library(ggplot2); library(grid); library(gridExtra); library(RColorBrewer); library(readr); library(matrixStats)
library(GenomicRanges); library(dplyr); library(zeallot); library(tximport); library(ggpubr); library(plyr)

# Reoccurring paths
datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))

# Human reference for comparison
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_lncrna.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_proteincoding.gtf"))

# Macaque reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_proteincoding.gtf"))
mRNAs_ref <- mRNAs_ref[!is.na(mRNAs_ref$gene_biotype)]
mRNAs_ref <- mRNAs_ref[mRNAs_ref$gene_biotype == "protein_coding"]


all <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))

```
### Comparison of exon length, transcript length and number ofe exons distributions


```{r stats}
# How many novel lncRNAs do i identify (Genes)
all_novel_lnc <- all[substr(all$gene_id,1,4) %in% c( "MSTR"),]
length(unique(all_novel_lnc$gene_id))
```

# Check how many for each class 
```{r stats}
classes <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/05_feelNC_prediction/feelnc_gencode_linc/02_feelnc_classifier/rheMac10_EBOV-Kikwit_lncRNA_classes.txt", header = TRUE)
novel <- all_novel_lnc
novel_classes <- classes[classes$lncRNA_gene %in% novel$gene_id, ]
#novel_classes <- novel_classes[novel_classes$isBest ==1, ]

# Keep entry with lowest distance
novel_classes <- novel_classes %>% group_by(lncRNA_gene) %>% dplyr::mutate(mindistance = min(distance)) %>%  ungroup() %>% 
  filter(distance==mindistance)
novel_classes_type <- novel_classes %>% group_by(lncRNA_gene, subtype, direction, location) %>% dplyr::summarise(type = paste(type, collapse=", "))
novel_classes_type$types_summarized <- unlist(lapply(novel_classes_type$type, function(x) paste(unique(trimws(unlist(str_split(x, ",")))), collapse = ",")))
table(novel_classes_type$types_summarized)
# Check the subtypes 
subtypes <- (table(novel_classes_type[novel_classes_type$types_summarized  == "intergenic",]$subtype))
genic_antisense <- (table(novel_classes_type[novel_classes_type$types_summarized  == "genic" & novel_classes_type$direction == "antisense",]$location))

print(paste("Total novel lncRNA genes: ", length(unique(all_novel_lnc$gene_id))))
print(paste("INTERGENIC: ", table(novel_classes_type$types_summarized)[2]))
print("INTERGENIC subtypes : ")
print(subtypes[subtypes != 0 ])
print("-------------------")
print(paste("GENIC ANTISENSE: ", sum(genic_antisense)))
print("GENIC location : ")
print(genic_antisense[genic_antisense != 0 ])

```

# Plot summary statistics
```{r stats}
plot_stats_annotation(all_novel_lnc,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)
```



```{r stats}
# ----------------------------------------------------
#   Check how many we lose each step
# ----------------------------------------------------

novel_ribo<- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/05_feelNC_prediction/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf")

novel_transcripts <- novel_ribo[novel_ribo$type == "transcript", ]
n_novel_transcripts <- table(substr(unique(novel_transcripts$gene_id),1,3))[[2]]

novel_transcripts_ux <- novel_transcripts[novel_transcripts$class_code %in% c("u", "x"),]
n_novel_transcripts_ux <- table(substr(unique(novel_transcripts_ux$gene_id),1,3))[[1]]

conc_ribo <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/00_all_novels/novel_rhemac10_concordant_ribodepleted.gtf")
n_conc_ribo <- table(substr(unique(conc_ribo$gene_id),1,3))[[1]]

expre_ribo <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/02_novel_expressed/novel_expressed_ribodepleted.gtf")
n_expre_ribo <- table(substr(unique(expre_ribo$gene_id),1,3))[[1]]


# STATS 
print("-----------------------")
print(paste("Total novel transcripts",n_novel_transcripts, sep = ": "))
print(paste("Total novel transcripts intergenic or x",n_novel_transcripts_ux, sep = ": "))
print(paste("Total novel concordant",n_conc_ribo, sep = ": "))
print(paste("Total novel expressed",n_expre_ribo, sep = ": "))
print("-----------------------")
print(paste("Total lost for Intergenic/Antisense filter",n_novel_transcripts-n_novel_transcripts_ux , sep= ": "))
print(paste("Total lost for concordance filter",n_novel_transcripts_ux-n_conc_ribo , sep= ": "))
print(paste("Total lost for expression filter",n_conc_ribo-n_expre_ribo , sep= ": "))
print("-----------------------")







# Distribution of novel lncRNAs across chromosomes 
col_lnc = "#870052"
df <- NULL
df$chr <- seqnames(all[substr(all$gene_id,1,4) =="ribo"])
df$gene_id <- (all[substr(all$gene_id,1,4) =="ribo"]$gene_id)
df <- as.data.frame(df)
df <- df[str_length(as.character(df$chr)) < 3, ]
ggplot(unique(df), aes(x = chr))+geom_bar(color = col_lnc, fill =col_lnc, alpha = 0.75)+theme_paper+labs(x = "", y = "")+theme(legend.position = "")+theme(axis.text.x = element_text(angle = 45, hjust = 1) , axis.text.x.bottom = element_text(size = 12))



```

```{r stats}

# -------------------------------------
#       Number of exons distributions 
# -------------------------------------

df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(all_novel_lnc, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Annotated LncRNAs - Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Annotated LncRNAs - Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Annotated mRNAs - Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Annotated mRNAs - Macaque")))

# Numer of exons distribution
exon_nr_distr <- function(gr){
  gr <- gr[gr$type =="exon",]
  gr <- get_only_max_transcript(gr)
  df_l <- data.frame(get_nr_exons(gr))
  return(as.vector(as.numeric(df_l$max_exon)))
}

wilcox.test(exon_nr_distr(all_novel_lnc), exon_nr_distr(lncRNAs_ref))
wilcox.test(exon_nr_distr(all_novel_lnc), exon_nr_distr(mRNAs_ref))


# -------------------------------------
#      Transcript lengths  
# -------------------------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_transcript_length(all_novel_lnc, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref, "Annotated LncRNAs - Macaque ")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref_human, "Annotated LncRNAs - Human")))
df <- rbind(df,data.frame(calc_transcript_length(mrna_ref_human, "Annotated mRNAs - Human")))
df <- rbind(df,data.frame(calc_transcript_length(mRNAs_ref, "Annotated mRNAs - Macaque")))


wilcox.test(as.numeric(df[df$type == "Novel LncRNAs",]$range), as.numeric(df[df$type == "Annotated LncRNAs - Macaque ",]$range))
wilcox.test(as.numeric(df[df$type == "Novel LncRNAs",]$range), as.numeric(df[df$type == "Annotated LncRNAs - Human",]$range))

wilcox.test(as.numeric(df[df$type == "Novel LncRNAs",]$range), as.numeric(df[df$type == "Annotated mRNAs - Macaque",]$range))


```

```{r expression}

palette_expression <-c("#E8C2D8", "#D4549C", "#8195D7")

iterate_files <- function(inpath, pattern_string){
  files <- list.files(path=inpath, pattern= pattern_string, full.names=TRUE, recursive=TRUE)
  return(files)
}


dir_counts_ref <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/04_quantification/"


abundance_files <- iterate_files(dir_counts_ref, "*.tsv")

# Read all files
abundances <- list()
for(i in 1:length(abundance_files)) {
  file <- readr::read_tsv(abundance_files[i])
  # Sum up values for double entries (https://github.com/gpertea/stringtie/issues/192)
  file <- file %>% dplyr::group_by(`Gene ID`)%>% dplyr::summarise(TPM = sum(TPM))
  abundances[[i]] <- data.frame(file$TPM, row.names =file$`Gene ID` )
}

# Summarize all TPMs from all quantification files 
rn <- rownames(abundances[[1]])
dat <- abundances[[1]]
for(i in 2:length(abundances)) {
  dat <- merge(dat, abundances[[i]],  by= "row.names", all.x= F, all.y= F) [,-1]
  rownames(dat) <- rn
}

expression <- dat
expression <- as.matrix(expression)
# ----------------------------------------------------------------------
# Extract expression ALL
# ----------------------------------------------------------------------
max_expression <- data.frame(id=rownames(expression), expr=log(rowMaxs(as.matrix(expression))))
max_expression['type'] <- "0"
max_expression <- add_type(max_expression, all_novel_lnc$gene_id, "Novel lncRNAs")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mRNAs")
max_expression <- max_expression[!max_expression$type == "0",]
plot_expression(max_expression, palette_expression)+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))

# ----------------------------------------------------------------------
# Same for the median
# ----------------------------------------------------------------------
median_expression <- data.frame(id=rownames(expression), expr=log(rowMedians(as.matrix(expression))))
median_expression['type'] <- "0"
median_expression <- add_type(median_expression, all_novel_lnc$gene_id, "Novel lncRNAs")
median_expression <- add_type(median_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
median_expression <- add_type(median_expression, mRNAs_ref$gene_id, "mRNAs")
median_expression <- median_expression[!median_expression$type == "0",]
plot_expression(median_expression, palette_expression, title = "Median Expression")+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))+labs(y = "logTPM")

median_expression_novel <- median_expression[median_expression$type == "Novel lncRNAs",]$expr
median_expression_annot <- median_expression[median_expression$type == "Annotated lncRNAs",]$expr
median_expression_pc <- median_expression[median_expression$type == "mRNAs",]$expr
wilcox.test(median_expression_novel, median_expression_pc)
wilcox.test(median_expression_novel, median_expression_annot)
wilcox.test(median_expression_pc, median_expression_annot)

#level=  c("Novel lncRNAs","Annotated lncRNAs", "mRNAs")
#ggboxplot(median_expression, x = "type", y = "expr", fill = factor("type",level = level), alpha = 0.5,
#                color = "type", palette =palette_expression)+stat_compare_means(comparisons =list(c("Novel lncRNAs", #"mRNAs"), c("Novel lncRNAs", "Annotated lncRNAs"), c("Annotated lncRNAs", "mRNAs")),  label = "p.signif")+ labs(x = "")


```

# -----------------------------
# Tissue Sharing
# -----------------------------


```{r tissuespecificity}

tissues <- unlist(lapply(abundance_files, function(x) rev(str_split(x, "/")[[1]])[4]))
tissues <- lapply(tissues, function(x) (str_split(x, "-")[[1]])[1])
unique(tissues)
colnames(expression) <- tissues
tissues <- gsub("Testes", "Testis", tissues)
tissues <- gsub("Wholeblood", "Blood", tissues)
tissues <- gsub("Serum", "Blood", tissues)

# Check how many tissues express a gene
expressed_booleans <- ifelse(log(expression) > 1,1,0)
colnames(expressed_booleans) <- tissues
grouped_expression <- ifelse(t(rowsum(t(expressed_booleans), group = colnames(expressed_booleans), na.rm = TRUE))>0, 1,0)
n_tissues_expresseing_gene <- rowSums(grouped_expression)

# Create DF for plotting
df <- as.data.frame(as.data.frame(n_tissues_expresseing_gene))
df$id <- rownames(df)


df$type <- "0"
df <- add_type(df, all_novel_lnc$gene_id, "Novel lncRNAs")
df <- add_type(df, lncRNAs_ref$gene_id, "Annotated lncRNAs")
df <- add_type(df, mRNAs_ref$gene_id, "mRNAs")
df <- df[!df$type == "0",]



type <- c("Novel lncRNAs", "Annotated lncRNAs", "mRNAs")
palette <- c("#E8C2D8", "#D4549C", "#8195D7")
tl1 <-barplot_tissues(df,"Novel lncRNAs",palette[1])+
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl2 <-barplot_tissues(df,"Annotated lncRNAs",palette[2]) +
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl3 <-barplot_tissues(df,"mRNAs",palette[3])
b <- ggarrange( tl1,tl2,tl3,  ncol=1, nrow=3)

b <- annotate_figure(b, bottom = text_grob("Number of tissues", size  = 22), left = text_grob("Frequency", size = 22, rot = 90))
b



```



