---
title: "01_lncrnas_plots_paper"
author: "Luisa Santus"
date: "4/23/2020"
output: html_document
  keep_md: true 
---

## 01 lncRNAs annotation summary

### Summary of features of novel lncRNAs compared to human and macaque annotate lncRNAs and mRNAs

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer);  library(stringr); library(ggplot2); library(grid); library(gridExtra); library(RColorBrewer); library(readr); library(matrixStats)
library(GenomicRanges); library(dplyr); library(zeallot); library(tximport); library(ggpubr); library(plyr)

# Reoccurring paths
datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"
datadir_old <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/old/"

# Import Utils 
source(file.path("../utils/01_lncrna_annotation_utils.R"))

# Human reference for comparison
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownlncRNAs.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownMrnas.gtf"))

# Macaque reference 
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit.gtf"))


# Novel 
#all <- import(file.path(datadir, "03_novel_lncRNAs_list/rheMac10_EBOV_and_novel.gtf"))
#ribo_expr <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/02_novel_expressed/novel_expressed_ribodepleted.gtf"))
#polya_expr <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/02_novel_expressed/novel_expressed_polya.gtf"))

#Old
all <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/old/03_novel_lncrnas/02_final_catalogue/rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names.gtf"))
ribo_expr <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/old/03_novel_lncrnas/01_novel_expressed/novel_expressed_ribodepleted.gtf"))
polya_expr <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/old/03_novel_lncrnas/01_novel_expressed/novel_expressed_polyA.gtf"))

#levels(seqnames(all))
#seqlevelsStyle(all) <- "ucsc"
#export(all, file.path(datadir, "03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_ucsc.gtf"))
```
### Comparison of exon length, transcript length and number ofe exons distributions


```{r stats}
# Testing
ref <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit.gtf")
ref_mrna <- ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "protein_coding",]
ref_mrna_exons <- ref_mrna[ref_mrna$type == "exon",]

```


```{r stats}
polya_lnc  <- all[substr(all$gene_id,1,4) == "poly",]
length(unique(polya_lnc$gene_id))
ribo_lnc  <- all[substr(all$gene_id,1,4) == "ribo",]
length(unique(ribo_lnc$gene_id))

all_novel_lnc <- all[substr(all$gene_id,1,4) %in% c("poly", "ribo"),]
length(unique(all_novel_lnc$gene_id))

# Just chek there is no overlap with protein coding genes
mRNAs_ref <- mRNAs_ref[!is.na(mRNAs_ref$gene_biotype)]
mRNAs_ref <- mRNAs_ref[mRNAs_ref$gene_biotype == "protein_coding"]


plot_stats_annotation(all_novel_lnc,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)
plot_stats_annotation_separated(polya_lnc,ribo_lnc ,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)
```

```{r expression}

#palette_expression <-c("#E8C2D8", "#D4549C", "#A1D76A")
palette_expression <-c("#E8C2D8", "#D4549C", "#8195D7")

iterate_files <- function(inpath, pattern_string){
  files <- list.files(path=inpath, pattern= pattern_string, full.names=TRUE, recursive=TRUE)
  return(files)
} 

# ----------------------------------------------------------------------
# Extract expression ALL 
# ----------------------------------------------------------------------
dir_counts_ref = file.path(datadir, "04_quantification/")
#OLD
dir_counts_ref =file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/old/04_Stringtie_correct_with_names_gene_ab/")
abundance_files <- iterate_files(dir_counts_ref, "t_data.ctab")
# new
tissues <- unlist(lapply(abundance_files, function(x) rev(str_split(x, "/")[[1]])[4]))
#old
tissues <- unlist(lapply(abundance_files, function(x) rev(str_split(x, "/")[[1]])[5]))

# Create Map for Transcript names to Gene names 
tmp <- read_tsv(abundance_files[1])
tx2gene <- tmp[,c("t_name", "gene_id")]
txi <- tximport(abundance_files, type = "stringtie", tx2gene = tx2gene, abundanceCol = tissues)

# Remove non expressed genes 
expression <- txi$abundance
mask<- rowSums(as.data.frame(log(expression)) > 0) > 2
expression <- expression[mask,]


# ----------------------------------------------------------------------
# Extract expression ALL 
# ----------------------------------------------------------------------
max_expression <- data.frame(id=rownames(expression), expr=log(rowMaxs(expression)))
max_expression['type'] <- "0"
max_expression <- add_type(max_expression, all_novel_lnc$gene_id, "Novel lncRNAs")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mRNAs")
max_expression <- max_expression[!max_expression$type == "0",]
plot_expression(max_expression, palette_expression)+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))

# ----------------------------------------------------------------------
# Same for the median 
# ----------------------------------------------------------------------
median_expression <- data.frame(id=rownames(expression), expr=log(rowMedians(expression)))
median_expression['type'] <- "0"
median_expression <- add_type(median_expression, all_novel_lnc$gene_id, "Novel lncRNAs")
median_expression <- add_type(median_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
median_expression <- add_type(median_expression, mRNAs_ref$gene_id, "mRNAs")
median_expression <- median_expression[!median_expression$type == "0",]
plot_expression(median_expression, palette_expression, title = "Median Expression")+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))

median_expression_novel <- median_expression[median_expression$type == "Novel lncRNAs",]$expr
median_expression_annot <- median_expression[median_expression$type == "Annotated lncRNAs",]$expr
median_expression_pc <- median_expression[median_expression$type == "mRNAs",]$expr
wilcox.test(median_expression_novel, median_expression_pc)
wilcox.test(median_expression_novel, median_expression_annot)
wilcox.test(median_expression_pc, median_expression_annot)

#level=  c("Novel lncRNAs","Annotated lncRNAs", "mRNAs")
#ggboxplot(median_expression, x = "type", y = "expr", fill = factor("type",level = level), alpha = 0.5,
#                color = "type", palette =palette_expression)+stat_compare_means(comparisons =list(c("Novel lncRNAs", #"mRNAs"), c("Novel lncRNAs", "Annotated lncRNAs"), c("Annotated lncRNAs", "mRNAs")),  label = "p.signif")+ labs(x = "")

# ----------------------------------------------------------------------
# Extract compared
# ----------------------------------------------------------------------
palette_expression <-c("orange","#994C00", "#D4549C", "#8195D7")
max_expression <- data.frame(id=rownames(expression), expr=log(rowMaxs(expression)))
max_expression['type'] <- "0"
max_expression <- add_type(max_expression, polya_lnc$gene_id, "Poly(A) Novel lncRNAs")
max_expression <- add_type(max_expression, ribo_lnc$gene_id, "Ribodepleted Novel lncRNAs")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mRNAs")
max_expression <- max_expression[!max_expression$type == "0",]

levels <-  c("Poly(A) Novel lncRNAs", "Ribodepleted Novel lncRNAs","Annotated lncRNAs", "mRNAs")
plot_expression(max_expression, palette_expression, level = levels)+theme(plot.title = element_text(size = 22), axis.text = element_text( angle =0, vjust = 0.9),axis.title.y  = element_text(size = 15))





palette_expression <-c("orange","#994C00", "#D4549C", "#8195D7")
median_expression <- data.frame(id=rownames(expression), expr=log(rowMedians(expression)))
median_expression['type'] <- "0"
median_expression <- add_type(median_expression, polya_lnc$gene_id, "Poly(A) Novel lncRNAs")
median_expression <- add_type(median_expression, ribo_lnc$gene_id, "Ribodepleted Novel lncRNAs")
median_expression <- add_type(median_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
median_expression <- add_type(median_expression, mRNAs_ref$gene_id, "mRNAs")
median_expression <- median_expression[!median_expression$type == "0",]

levels <-  c("Poly(A) Novel lncRNAs", "Ribodepleted Novel lncRNAs","Annotated lncRNAs", "mRNAs")
plot_expression(median_expression, palette_expression, level = levels, title = "Median Expression")+theme(plot.title = element_text(size = 22), axis.text = element_text( angle =0, vjust = 0.9),axis.title.y  = element_text(size = 15))
```

# -----------------------------
# Tissue Sharing 
# -----------------------------


```{r tissuespecificity}
colnames(expression) <- tissues
tissues <- gsub("Testes", "Testis", tissues)
tissues <- gsub("Wholeblood", "Blood", tissues)


# Check how many tissues express a gene
expressed_booleans <- ifelse(log(expression) > 1,1,0)
colnames(expressed_booleans) <- tissues
grouped_expression <- ifelse(t(rowsum(t(expressed_booleans), group = colnames(expressed_booleans), na.rm = TRUE))>0, 1,0)
n_tissues_expresseing_gene <- rowSums(grouped_expression)

# Create DF for plotting
df <- as.data.frame(as.data.frame(n_tissues_expresseing_gene))
df$id <- rownames(df)


df$type <- "0"
df <- add_type(df, all_novel_lnc$gene_id, "Novel lncRNAs")
df <- add_type(df, lncRNAs_ref$gene_id, "Annotated lncRNAs")
df <- add_type(df, mRNAs_ref$gene_id, "mRNAs")
df <- df[!df$type == "0",]



type <- c("Novel lncRNAs", "Annotated lncRNAs", "mRNAs")
palette <- c("#E8C2D8", "#D4549C", "#8195D7")
tl1 <-barplot_tissues(df,"Novel lncRNAs",palette[1])+
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl2 <-barplot_tissues(df,"Annotated lncRNAs",palette[2]) +
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl3 <-barplot_tissues(df,"mRNAs",palette[3]) 
b <- ggarrange( tl1,tl2,tl3,  ncol=1, nrow=3)

b <- annotate_figure(b, bottom = text_grob("Number of tissues", size  = 22), left = text_grob("Frequency", size = 22, rot = 90))
b

```


# Check which ones are ribodepleted

```{r ribodepl-enrichment}

dir_counts_ref =file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/old/04_Stringtie_correct_with_names_gene_ab/")
abundance_files <- iterate_files(dir_counts_ref, "t_data.ctab")
#old
tissues <- unlist(lapply(abundance_files, function(x) rev(str_split(x, "/")[[1]])[5]))
tissues <- gsub("Testes", "Testis", tissues)
tissues <- gsub("Wholeblood", "Blood", tissues)

# Identify library prep
libraryprep <- unlist(lapply(abundance_files, function(x) rev(str_split(x, "/")[[1]])[6]))
libraryprep <-gsub("Batch01", "Ribodepleted", libraryprep) 
libraryprep <-gsub("Zyagen", "Ribodepleted", libraryprep) 
libraryprep <-gsub("RhesusmacaqueChinese", "Polya", libraryprep) 
libraryprep <-gsub("SRP016501-rhesus", "Polya", libraryprep)
table(libraryprep)

# Select only tissues present in both 
df <- data.frame(t = (tissues), lib = libraryprep)
summary_proprieties_samples <- df %>% dplyr::group_by(t) %>% dplyr::summarise(num = n_distinct(lib))
tissues_available_both <- summary_proprieties_samples[summary_proprieties_samples$num > 1,]$t

# Create Map for Transcript names to Gene names 
tmp <- read_tsv(abundance_files[1])
tx2gene <- tmp[,c("t_name", "gene_id")]
txi <- tximport(abundance_files, type = "stringtie", tx2gene = tx2gene, abundanceCol = libraryprep)

# Remove non expressed genes 
expression <- txi$abundance
colnames(expression) <- libraryprep

# Ecpressed in Brain 
tissue <- "Brain"

find_enrichment_libraryprep <- function(expression , tissue){
 expression <- expression[ ,tissues == tissue]
  # every gene to be consifered is required to be expressed in at least one sample abofe log(FPKM) of 1
  mask<- rowSums(as.data.frame(log(expression)) > 1) > 0
  expression <- expression[mask,]
  # Calculate mean expression per libraries
  expression_libraries <- as.data.frame(do.call(cbind, by(t(expression),INDICES=colnames(expression),FUN=colMeans)))
  
  # Calculate enrichment Ribodepleted/Polya
  expression_libraries$enrichmentPolyAnegative <- expression_libraries$Ribodepleted/expression_libraries$Polya
  
  # I use 5 fold expression 
  enriched <- expression_libraries[expression_libraries$enrichmentPolyAnegative > 5,] 
  enriched$tissue <- tissue
  return(enriched)
}



# Check that i get NO poly a selected ones enriched 
all_enriched <- do.call(rbind, (lapply(tissues_available_both, function(tissue) find_enrichment_libraryprep(expression, tissue))))


# Ribodepleted which may lack a poly A tail 
ribo_unique_ids_enriched <- unique(rownames(all_enriched))[grep("ribo", unique(rownames(all_enriched)))]

length(unique(ribo_unique_ids_enriched))
# Out of all the DE Ones 
list_de <- gsub("_unkown", "", gsub("-", "_", unique(unlist(de_lists_lnc))))
ribodepl_list_de <- list_de[grepl("ribo*", list_de)]

# Enichment of differentially expressed genes identified by ribodepletion 
enriched_polyneg_de <- all_enriched[rownames(all_enriched) %in% ribodepl_list_de, ]
enriched_polyneg_de <- all_enriched[rownames(all_enriched) %in% list_de, ]
length(unique(rownames(enriched_polyneg_de)))

df <- data.frame( foldEnrichment= enriched_polyneg_de$enrichmentPolyAnegative)
ggplot(df, aes(x = foldEnrichment))+geom_histogram(fill = "grey", color = "grey")+xlim(0,50)+ theme_classic()


# Check the same for protein coding genes 
list_de_mrna <- gsub("_unkown", "", gsub("-", "_", unique(unlist(de_lists_mrna))))

# Enichment of differentially expressed genes identified by ribodepletion 
enriched_polyneg_de <- all_enriched[rownames(all_enriched) %in% list_de_mrna, ]
length(unique(rownames(enriched_polyneg_de)))

df <- data.frame( foldEnrichment= enriched_polyneg_de$enrichmentPolyAnegative)
ggplot(df, aes(x = foldEnrichment))+geom_histogram(fill = "grey", color = "grey")+xlim(0,50)+ theme_classic()
```
