---
title: "01_lncrnas_plots_paper"
author: "Luisa Santus"
date: "4/23/2020"
output: html_document
  keep_md: true
---

## 01 lncRNAs annotation summary


### Summary of features of novel lncRNAs compared to human and macaque annotate lncRNAs and mRNAs

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer);  library(stringr); library(ggplot2); library(grid); library(gridExtra); library(RColorBrewer); library(readr); library(matrixStats)
library(GenomicRanges); library(dplyr); library(zeallot); library(tximport); library(ggpubr); library(plyr)

# Reoccurring paths
datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))

# Human reference for comparison
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_lncrna.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_proteincoding.gtf"))

# Macaque reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_proteincoding.gtf"))
mRNAs_ref <- mRNAs_ref[!is.na(mRNAs_ref$gene_biotype)]
mRNAs_ref <- mRNAs_ref[mRNAs_ref$gene_biotype == "protein_coding"]


all <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))



```
### Comparison of exon length, transcript length and number ofe exons distributions


```{r stats}
# How many novel lncRNAs do i identify (Genes)
all_novel_lnc <- all[substr(all$gene_id,1,4) %in% c( "MSTR"),]
length(unique(all_novel_lnc$gene_id))
length(unique(all_novel_lnc$transcript_id))
plot_stats_annotation(all_novel_lnc,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)

export(all_novel_lnc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/novelOnly.gtf")
```


# Separate by intergenic and antisense 
```{r cpc2vscpan}
gffcompare <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/04_stringtie_gffcompare/01_gffCompare/merged.annotated.gtf")

# Check everything is allright 
length(unique(all_novel_lnc$transcript_id))
length(unique(gffcompare[gffcompare$transcript_id %in% all_novel_lnc$transcript_id,]$transcript_id))

plot_stats_benchmark(gffcompare[gffcompare$transcript_id %in% all_novel_lnc$transcript_id & gffcompare$type == "transcript",])

gff_compare_pre_lnc <- gffcompare[gffcompare$transcript_id %in% all_novel_lnc$transcript_id & gffcompare$type == "transcript",]
antisense_ids <- gff_compare_pre_lnc[gff_compare_pre_lnc$class_code == "x", ]$transcript_id
intergenic_ids <- gff_compare_pre_lnc[gff_compare_pre_lnc$class_code == "u", ]$transcript_id


intergenic_lnc <- all_novel_lnc[all_novel_lnc$transcript_id %in% intergenic_ids, ]
antisense_lnc <- all_novel_lnc[all_novel_lnc$transcript_id %in% antisense_ids, ]

plot_stats_annotation_separated(intergenic_lnc,antisense_lnc ,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref)


```


```{r expression}

palette_expression <-c("#E8C2D8", "#D4549C", "#8195D7")

iterate_files <- function(inpath, pattern_string){
  files <- list.files(path=inpath, pattern= pattern_string, full.names=TRUE, recursive=TRUE)
  return(files)
}


dir_counts_ref <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/04_quantification/"


abundance_files <- iterate_files(dir_counts_ref, "*.tsv")

# Read all files
abundances <- list()
for(i in 1:length(abundance_files)) {
  file <- readr::read_tsv(abundance_files[i])
  # Sum up values for double entries (https://github.com/gpertea/stringtie/issues/192)
  file <- file %>% dplyr::group_by(`Gene ID`)%>% dplyr::summarise(TPM = sum(TPM))
  abundances[[i]] <- data.frame(file$TPM, row.names =file$`Gene ID` )
}

# Summarize all TPMs from all quantification files 
rn <- rownames(abundances[[1]])
dat <- abundances[[1]]
for(i in 2:length(abundances)) {
  dat <- merge(dat, abundances[[i]],  by= "row.names", all.x= F, all.y= F) [,-1]
  rownames(dat) <- rn
}

expression <- dat
expression <- as.matrix(expression)
# ----------------------------------------------------------------------
# Extract expression ALL
# ----------------------------------------------------------------------
max_expression <- data.frame(id=rownames(expression), expr=log(rowMaxs(as.matrix(expression))))
max_expression['type'] <- "0"
max_expression$type
max_expression <- add_type(max_expression, all_novel_lnc$gene_id, "Novel lncRNAs")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mRNAs")
max_expression <- max_expression[!max_expression$type == "0",]
plot_expression(max_expression, palette_expression)+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))


# ----------------------------------------------------------------------
# Same for the median
# ----------------------------------------------------------------------
median_expression <- data.frame(id=rownames(expression), expr=log(rowMedians(as.matrix(expression))))
median_expression['type'] <- "0"
median_expression <- add_type(median_expression, all_novel_lnc$gene_id, "Novel lncRNAs")
median_expression <- add_type(median_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
median_expression <- add_type(median_expression, mRNAs_ref$gene_id, "mRNAs")
median_expression <- median_expression[!median_expression$type == "0",]
plot_expression(median_expression, palette_expression, title = "Median Expression")+theme(plot.title = element_text(size = 26), axis.text = element_text(size = 18),axis.title.y  = element_text(size = 18))+labs(y = "logTPM")

median_expression_novel <- median_expression[median_expression$type == "Novel lncRNAs",]$expr
median_expression_annot <- median_expression[median_expression$type == "Annotated lncRNAs",]$expr
median_expression_pc <- median_expression[median_expression$type == "mRNAs",]$expr
wilcox.test(median_expression_novel, median_expression_pc)
wilcox.test(median_expression_novel, median_expression_annot)
wilcox.test(median_expression_pc, median_expression_annot)

#level=  c("Novel lncRNAs","Annotated lncRNAs", "mRNAs")
#ggboxplot(median_expression, x = "type", y = "expr", fill = factor("type",level = level), alpha = 0.5,
#                color = "type", palette =palette_expression)+stat_compare_means(comparisons =list(c("Novel lncRNAs", #"mRNAs"), c("Novel lncRNAs", "Annotated lncRNAs"), c("Annotated lncRNAs", "mRNAs")),  label = "p.signif")+ labs(x = "")

palette_expression <-c("orange","#994C00", "#D4549C", "#8195D7")
median_expression <- data.frame(id=rownames(expression), expr=log(rowMedians(expression)))
median_expression['type'] <- "0"
median_expression <- add_type(median_expression, intergenic_lnc$gene_id, "Intergenic Novel lncRNAs")
median_expression <- add_type(median_expression, antisense_lnc$gene_id, "Antisense Novel lncRNAs")
median_expression <- add_type(median_expression, lncRNAs_ref$gene_id, "Annotated lncRNAs")
median_expression <- add_type(median_expression, mRNAs_ref$gene_id, "mRNAs")
median_expression <- median_expression[!median_expression$type == "0",]

levels <-  c("Intergenic Novel lncRNAs", "Antisense Novel lncRNAs","Annotated lncRNAs", "mRNAs")
plot_expression(median_expression, palette_expression, level = levels, title = "Median Expression")+theme(plot.title = element_text(size = 22), axis.text = element_text( angle =0, vjust = 0.9),axis.title.y  = element_text(size = 15))+labs(y = "logTPM")

```

# -----------------------------
# Tissue Sharing
# -----------------------------


```{r tissuespecificity}

tissues <- unlist(lapply(abundance_files, function(x) rev(str_split(x, "/")[[1]])[4]))
tissues <- lapply(tissues, function(x) (str_split(x, "-")[[1]])[1])
unique(tissues)
colnames(expression) <- tissues
tissues <- gsub("Testes", "Testis", tissues)
tissues <- gsub("Wholeblood", "Blood", tissues)
tissues <- gsub("Serum", "Blood", tissues)
table(tissues)
# Check how many tissues express a gene
expressed_booleans <- ifelse(log(expression) > 1,1,0)
colnames(expressed_booleans) <- tissues
grouped_expression <- ifelse(t(rowsum(t(expressed_booleans), group = colnames(expressed_booleans), na.rm = TRUE))>0, 1,0)
n_tissues_expresseing_gene <- rowSums(grouped_expression)

# Create DF for plotting
df <- as.data.frame(as.data.frame(n_tissues_expresseing_gene))
df$id <- rownames(df)


df$type <- "0"
df <- add_type(df, all_novel_lnc$gene_id, "Novel lncRNAs")
df <- add_type(df, lncRNAs_ref$gene_id, "Annotated lncRNAs")
df <- add_type(df, mRNAs_ref$gene_id, "mRNAs")
df <- df[!df$type == "0",]



type <- c("Novel lncRNAs", "Annotated lncRNAs", "mRNAs")
palette <- c("#E8C2D8", "#D4549C", "#8195D7")
tl1 <-barplot_tissues(df,"Novel lncRNAs",palette[1])+
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl2 <-barplot_tissues(df,"Annotated lncRNAs",palette[2]) +
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl3 <-barplot_tissues(df,"mRNAs",palette[3])
b <- ggarrange( tl1,tl2,tl3,  ncol=1, nrow=3)

b <- annotate_figure(b, bottom = text_grob("Number of tissues", size  = 22), left = text_grob("Frequency", size = 22, rot = 90))
b


df$type <- "0"
df <- add_type(df, intergenic_lnc$gene_id, "Intergenic Novel lncRNAs")
df <- add_type(df, antisense_lnc$gene_id, "Antisense Novel lncRNAs")
df <- add_type(df, lncRNAs_ref$gene_id, "Annotated lncRNAs")
df <- add_type(df, mRNAs_ref$gene_id, "mRNAs")
df <- df[!df$type == "0",]

type <-  c("Intergenic Novel lncRNAs", "Antisense Novel lncRNAs","Annotated lncRNAs", "mRNAs")
palette <-c("orange","#994C00", "#D4549C", "#8195D7")

tl0 <-barplot_tissues(df,"Intergenic Novel lncRNAs",palette[1])+
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())

tl1 <-barplot_tissues(df,"Antisense Novel lncRNAs",palette[2])+
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl2 <-barplot_tissues(df,"Annotated lncRNAs",palette[3]) +
                          theme(axis.ticks.x = element_blank(),
                          axis.text.x = element_blank(),
                          axis.line.x = element_blank())
tl3 <-barplot_tissues(df,"mRNAs",palette[4])
b <- ggarrange( tl0,tl1,tl2,tl3,  ncol=1, nrow=4)

b <- annotate_figure(b, bottom = text_grob("Number of tissues", size  = 22), left = text_grob("Frequency", size = 22, rot = 90))
b

```

# Supplementary figures

## Compare number of genes in macaques vs Human annotation 
```{r results='hide', message=FALSE, warning=FALSE}
ref_macaque <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf"))
ref_human <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))


ref_macaque_gene <- ref_macaque[ref_macaque$type =="gene",]

table(ref_macaque[ref_macaque$source == "ensembl",]$gene_biotype)
table(ref_macaque[ref_macaque$source == "RefSeq",]$gene_biotype)

table(ref_human$gene_biotype)
ref_macaque_gene <- ref_macaque[ref_macaque$type == "gene",]
ref_human_gene <- ref_human[ref_human$type == "gene",]

lnc_macaque <- sum(ref_macaque_gene$gene_biotype == "lncRNA")
pc_macaque <- sum(ref_macaque_gene$gene_biotype == "protein_coding")
lnc_human <- sum(ref_human_gene$gene_biotype == "lncRNA")
pc_human <- sum(ref_human_gene$gene_biotype == "protein_coding")

count <- c(lnc_macaque, pc_macaque, lnc_human, pc_human ) 
species <- c(rep("Macaque", 2), rep("Human",2))
gene_biotype <- c(rep(c("LncRNA", "Protein Coding"), 2 ) )
df_n_genes <- data.frame(count, species, gene_biotype)
```

```{r results='hide', message=FALSE, warning=FALSE}
ggplot(df_n_genes, aes(fill = gene_biotype, y = count, x = species))+
                geom_bar(stat="identity", position = position_dodge2())+
                theme_paper+scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)+
                labs(x = "", y = "")+
                scale_y_continuous( breaks = c(lnc_macaque, lnc_human))+
                geom_abline(slope=0, intercept=lnc_macaque,  col = "darkgrey",lty=2)+
                geom_abline(slope=0, intercept=lnc_human,  col = "darkgrey",lty=2)
```


