---
title: "01_lncrnas_plots_paper"
author: "Luisa Santus"
date: "4/23/2020"
output: html_document
  keep_md: true
---

# LncRNAs annotation summary of filtering steps 
## Quantify the loss of transcipts and genes in each step of the de novo anntoation pipeline

# imports 
```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer);  library(stringr); library(ggplot2); library(grid); library(gridExtra); library(RColorBrewer); library(readr); library(matrixStats)
library(GenomicRanges); library(dplyr); library(zeallot); library(tximport); library(ggpubr); library(plyr)

# Reoccurring paths
# Define paths for data
source(file.path("../utils/00_datapaths.R"))
# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))

datadir <- file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/")
dir_counts_ref <- file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/04_quantification/")



# Human reference for comparison
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_lncrna.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_proteincoding.gtf"))

# Macaque reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_proteincoding.gtf"))
mRNAs_ref <- mRNAs_ref[!is.na(mRNAs_ref$gene_biotype)]
mRNAs_ref <- mRNAs_ref[mRNAs_ref$gene_biotype == "protein_coding"]

all <- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
```

# calculate loss
```{r stats}
all_novel<- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/02_RNA-Seq_ribodepl/04_stringtie_gffcompare/merged.annotated.gtf"))
novel_gene <- all_novel[all_novel$type == "transcript", ]
n_novel_gene <- table(substr(unique(novel_gene$gene_id),1,3))[[2]]

novel_gene_ux <- novel_gene[novel_gene$class_code %in% c("u", "x"),]
n_novel_gene_ux <- table(substr(unique(novel_gene_ux$gene_id),1,3))[[1]]

conc_ribo <- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/00_prefilter_candidates/prefilter_candidates.gtf"))
n_conc_ribo <- table(substr(unique(conc_ribo$gene_id),1,3))[[1]]

expre_ribo <- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/02_novel_expressed/candidates.gtf"))
n_expre_ribo <- table(substr(unique(expre_ribo$gene_id),1,3))[[1]]
```

# print
```{r stats}
# STATS 
print("-----------------------")
print(paste("Total novel gene",n_novel_gene, sep = ": "))
print(paste("Total novel gene intergenic or x",n_novel_gene_ux, sep = ": "))
print(paste("Total novel concordant",n_conc_ribo, sep = ": "))
print(paste("Total novel expressed",n_expre_ribo, sep = ": "))
print("-----------------------")
print(paste("Total lost for Intergenic/Antisense filter",n_novel_gene-n_novel_gene_ux , sep= ": "))
print(paste("Total lost for concordance filter",n_novel_gene_ux-n_conc_ribo , sep= ": "))
print(paste("Total lost for expression filter",n_conc_ribo-n_expre_ribo , sep= ": "))
print("-----------------------")

```
