---
title: "01_lncrnas_plots_paper"
author: "Luisa Santus"
date: "4/23/2020"
output: html_document
  keep_md: true
---

## 01 lncRNAs annotation summary


### Summary of features of novel lncRNAs compared to human and macaque annotate lncRNAs and mRNAs

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer);  library(stringr); library(ggplot2); library(grid); library(gridExtra); library(RColorBrewer); library(readr); library(matrixStats)
library(GenomicRanges); library(dplyr); library(zeallot); library(tximport); library(ggpubr); library(plyr)

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))


# Human reference for comparison
ref <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf"))
cpc2 <- read.table(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/99_benchmark_annotation/01_predictions/CPC2/cpc2_pred.txt"))
names(cpc2) <- c("ID","transcript_length ","peptide_length","Fickett_score" , " pI", "ORF_integrity", "coding_probability","label")
# Extract transcript ids 
cpc2$transcript_id <- unlist(lapply(as.character(cpc2$ID), function(x) strsplit(x,"[(]")[[1]][1]))
cpat <- read.table(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/99_benchmark_annotation/01_predictions/CPAT/cpat_pred.ORF_prob.tsv"), header = T)
cutoff <-  0.364
cpat$label <- ifelse(cpat$Coding_prob > cutoff, "coding", "noncoding")
cnit <- read.table(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/99_benchmark_annotation/01_predictions/CNIT/cnit_pred/CNCI2.index"), header = T, sep="\t")

ref_lnc <- ref[ref$transcript_biotype %in% c("lncRNA", "antisense"),]
ref_lnc <- ref_lnc[ref_lnc$type == "transcript", ]

# How many of the known lncRNAs are identified as such 
noncoding_cpc2 <- unlist(lapply(cpc2[cpc2$label == "noncoding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))
coding_cpc2 <- unlist(lapply(cpc2[cpc2$label == "coding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))

noncoding_cpat <- unlist(lapply(cpat[cpat$label == "noncoding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))
coding_cpat <- unlist(lapply(cpat[cpat$label == "coding",]$ID, function(x) strsplit(as.character(x),'(', fixed = T)[[1]][1]))

# How many lncRNAs in total in the annotation 
length(unique(ref_lnc$transcript_id))
length(unique(ref_lnc$gene_id))



identified_lnc <- ref_lnc[ref_lnc$transcript_id %in% noncoding_cpc2,]
length(unique(identified_lnc$transcript_id))
length(unique(identified_lnc$gene_id))

missed_lnc <- ref_lnc[ref_lnc$transcript_id %in% coding_cpc2,]
length(unique(missed_lnc$transcript_id))
length(unique(missed_lnc$gene_id))

TP <- length(unique(identified_lnc$transcript_id))
FN <- length(unique(missed_lnc$gene_id))
Sensitivity <- (TP/(FN+TP))*100
print(Sensitivity)


ref_pc <- ref[ref$type == "transcript" & ref$transcript_biotype == "protein_coding",]
length(unique(ref_pc$transcript_id))
# Specificity 
TN <- length(unique(ref_pc[ref_pc$transcript_id %in% coding_cpc2,]))
FP <- length(unique(ref_pc[ref_pc$transcript_id %in% noncoding_cpc2,]))
specificity <- TN/(TN+FP)
specificity


annotated_reduced <- unique(ref_lnc[ref_lnc$transcript_id %in% c(noncoding_cpat, noncoding_cpc2),]$transcript_id)
length(unique(annotated_reduced))
listInput <- list(cpat = unique(noncoding_cpat[noncoding_cpat %in% annotated_reduced]), CPC2 = unique(noncoding_cpc2[noncoding_cpc2 %in% annotated_reduced]), annotated = annotated_reduced )

library(eulerr)
library(ComplexHeatmap)
plot(euler(listInput, shape = "ellipse"),
     quantities = TRUE,
     edges = list(lty = 3), 
     legend = list(side = "right")
    )
upset(fromList(listInput), order.by = "freq", main.bar.color = c("#CC0066", rep("gray23",2)))


```


