---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---


# Supplementary figures

## Compare number of genes in macaques vs Human annotation 
```{r results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)

col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))

ref_macaque <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf"))
ref_human <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))


ref_macaque_gene <- ref_macaque[ref_macaque$type =="gene",]

table(ref_macaque[ref_macaque$source == "ensembl",]$gene_biotype)
table(ref_macaque[ref_macaque$source == "RefSeq",]$gene_biotype)

table(ref_human$gene_biotype)
ref_macaque_gene <- ref_macaque[ref_macaque$type == "gene",]
ref_human_gene <- ref_human[ref_human$type == "gene",]

lnc_macaque <- sum(ref_macaque_gene$gene_biotype == "lncRNA")
pc_macaque <- sum(ref_macaque_gene$gene_biotype == "protein_coding")
lnc_human <- sum(ref_human_gene$gene_biotype == "lncRNA")
pc_human <- sum(ref_human_gene$gene_biotype == "protein_coding")


ref_macaque_transcript <- ref_macaque[ref_macaque$type == "transcript",]
ref_human_transcript <- ref_human[ref_human$type == "transcript",]

lnc_macaque_transcript <- sum(ref_macaque_transcript$transcript_biotype == "lncRNA")
pc_macaque_transcript <- sum(ref_macaque_transcript$transcript_biotype == "protein_coding")
lnc_human_transcript <- sum(ref_human_transcript$transcript_biotype == "lncRNA")
pc_human_transcript<- sum(ref_human_transcript$transcript_biotype == "protein_coding")

count <- c(lnc_macaque, pc_macaque, lnc_human, pc_human, lnc_macaque_transcript, pc_macaque_transcript, lnc_human_transcript, pc_human_transcript ) 
species <- c(rep("Macaque", 2), rep("Human",2,),rep("Macaque", 2), rep("Human",2))
gene_biotype <- c(rep(c("LncRNA", "Protein Coding"), 2 ),rep(c("LncRNA", "Protein Coding"), 2 ) )
gene_type <- c(rep("gene", 4), rep("transcript",4,))
df_n_genes <- data.frame(count, species, gene_biotype, gene_type)

df_n_genes
```



```{r results='hide', message=FALSE, warning=FALSE}
ggplot(df_n_genes[df_n_genes$gene_type == "gene", ], aes(fill = gene_biotype,  y = count, x = species))+
                geom_bar(stat="identity", position = position_dodge2(), alpha = 0.8)+
                theme_paper+scale_fill_manual(values = palette_plot_percentage)+
                labs(x = "", y = "")+
                scale_y_continuous( breaks = c(lnc_macaque, lnc_human))+
                geom_abline(slope=0, intercept=lnc_macaque,  col = "darkgrey",lty=2)+
                geom_abline(slope=0, intercept=lnc_human,  col = "darkgrey",lty=2)
```




