---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---

# ORTHOLOGS ANALYSIS 


### Imports and prepare transcripts
```{r beg, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)
library(networkD3)
library(dplyr)


# Import Utils
source(file.path("../utils/00_datapaths.R"))
source(file.path("../utils/01_lncrna_annotation_utils.R"))


theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
```

## Import files

```{r cpc2}

# Check for all orthologs 
orthologs <- read.table(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.txt"))
orthologs <- read.table(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.top.txt"))
names(orthologs)<- c("lnc","lncGeneSymbol","ortholog","orthologGeneSymbol","alignScore","exonID","locusID","indelRate(exon)" ,"indelRate(intron)","lncExonsAligned","orthExonsAligned","spliceConserved","spliceTotal","category(mmul10)","category(hg38)")

all <- import(file.path(data_path,"01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ref <- all
df_ref <- distinct(data.frame(ref[!is.na(ref$transcript_id),]$transcript_id, ref[!is.na(ref$transcript_id),]$gene_id, stringsAsFactors = F))
names(df_ref) <- c("lnc", "gene_id")
rownames(df_ref) <- df_ref$transcript_id
orthologs <- merge(orthologs, df_ref, by = "lnc")

saveRDS(orthologs,file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))


check_df <- distinct(orthologs[,c("gene_id", "orthologGeneSymbol")])
check <- check_df %>% dplyr::group_by(orthologGeneSymbol) %>% dplyr::count()
check[check$n > 1, ]
```


# Look for nice examples 

```{r cpc2}
all_stringtie_novellnc <- all[all$source == "StringTie", ]
all_annot <- all[all$source != "StringTie", ]
all_annot_lnc <- all_annot[all_annot$type == "transcript" & all_annot$transcript_biotype == "lncRNA",]
length(unique(all_annot_lnc$gene_id))
length(unique(all_stringtie_novellnc$gene_id))

orthologs_novel <- orthologs[substr(orthologs$lnc,1,4) == "MSTR",]
orthologs_lncAnnot <- orthologs[orthologs$lnc %in% all_annot_lnc$transcript_id,]

novel_with_ortholog <- length(unique(orthologs_novel$gene_id))
annotated_with_ortholog <- length(unique(orthologs_lncAnnot$gene_id))
novel_with_ortholog/sum(novel_with_ortholog,annotated_with_ortholog)

```



# ----- Gene level
```{r cpc2}


orthologs_novel_found <- sum(orthologs$gene_id %in% unique(all_stringtie_novellnc$gene_id))
novel_no_ortholog <- length(unique(all_stringtie_novellnc$gene_id))-orthologs_novel_found

orthologs_annot_found <- sum(orthologs$gene_id %in% unique(all_annot_lnc$gene_id))
annot_no_ortholog <- length(unique(all_annot_lnc$gene_id))-orthologs_annot_found

freqs <- c( 
           novel_no_ortholog/length(unique(all_stringtie_novellnc$gene_id)), 
           orthologs_novel_found/length(unique(all_stringtie_novellnc$gene_id)),
           annot_no_ortholog/length(unique(all_annot_lnc$gene_id)),
            orthologs_annot_found/length(unique(all_annot_lnc$gene_id)))


df <- data.frame(count = c(novel_no_ortholog,orthologs_novel_found,annot_no_ortholog,orthologs_annot_found ), type = c("novel", "novel", "annotated", "annotated"), orth = c( "Ortholog NOT found","Ortholog found", "Ortholog NOT found", "Ortholog found"),freq = freqs) 
df$orth <- relevel(factor(df$orth), ref = "Ortholog NOT found") 

# Plot with proportions
p_o <- ggplot(df, aes(x = type, fill = orth, y = freq))+geom_bar(stat = "identity", alpha  = 0.6, position = "stack")+xlab("")+scale_fill_manual(values = c("grey","#525252"))+ylab("proportion of genes with human ortholog")+
          theme(panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"),
              axis.ticks.length=unit(.2, "cm"),
              axis.text= element_text(size = 15, color = "black"),
              axis.title = element_text(size = 15),
              plot.title = element_text(hjust = 0.5, size = 17),
              axis.line.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.background = element_rect(fill = "white"))+
        scale_y_continuous(limits =c(0,1), labels = seq(0,1, 0.25),breaks = seq(0,1, 0.25), expand = c(0,0))

pdf(file.path(plots, "01/orthologs.pdf"), width = 4.5, height = 6)
p_o
dev.off()
```



