


```{r tissuespecificity}
library(rtracklayer)
library(dplyr)
library(tximport)
library(readr)
library(matrixStats)
```
# How many of our novel lncRNAs are expressed in Broad data ? 

## Also check genes in reference annotation - how many are expressed ( lnc vs protein coding )

```{r tissuespecificity}
# ----------------------
#   Get Expression 
# ----------------------
iterate_files <- function(inpath, pattern_string){
  files <- list.files(path=inpath, pattern= pattern_string, full.names=TRUE, recursive=TRUE)
  return(files)
} 

dir_counts_ref <- "/home/luisas/Desktop/cluster/data/99_BroadAnnotation/04_quantification_our_on_broad/"
novel_concordant <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_UCSC.gtf")

abundance_files <- iterate_files(dir_counts_ref, "*.tsv")

# Read all files
abundances <- list()
for(i in 1:length(abundance_files)) {
  file <- readr::read_tsv(abundance_files[i])
  # Sum up values for double entries (https://github.com/gpertea/stringtie/issues/192)
  file <- file %>% dplyr::group_by(`Gene ID`)%>% dplyr::summarise(TPM = sum(TPM))
  abundances[[i]] <- data.frame(file$TPM, row.names =file$`Gene ID` )
}

# Summarize all TPMs from all quantification files 
rn <- rownames(abundances[[1]])
dat <- abundances[[1]]
for(i in 2:length(abundances)) {
  dat <- merge(dat, abundances[[i]],  by= "row.names", all.x= F, all.y= F) [,-1]
  rownames(dat) <- rn
}

expression <- dat

# ---------------------------
#   Get the expressed ones from the concordant set
# ---------------------------


novel <- novel_concordant[novel_concordant$source == "StringTie",]
length(unique(novel$transcript_id))

annot_lnc <- novel_concordant[novel_concordant$source != "StringTie" & novel_concordant$type == "transcript" & novel_concordant$transcript_biotype == "lncRNA",]

annot_pc<- novel_concordant[novel_concordant$source != "StringTie" & novel_concordant$type == "transcript" & novel_concordant$transcript_biotype == "protein_coding",]

# How many annotated lncRNAs are found to be expressed 
expression_pc <- expression[rownames(expression) %in% annot_pc$gene_id,]
table(rowSums(as.data.frame(expression_pc > 0.0)) > 0)

expression_lnc <- expression[rownames(expression) %in% annot_lnc$gene_id,]
table(rowSums(as.data.frame(expression_lnc > 0.0)) > 0)

expression_novel <- expression[rownames(expression) %in% novel$gene_id,]
table(rowSums(as.data.frame(expression_novel > 0.5)) > 3)
```


# Novel from Broad 
```{r tissuespecificity}
candidates_broad <- import("/home/luisas/Desktop/cluster/data/99_BroadAnnotation/03_novel_lncRNAs_list/00_prefilter_candidates/prefilter_candidates.gtf")
# Out of 10 k candidate transcripts 
length(unique(candidates_broad$transcript_id))
dir_counts_ref <- "/home/luisas/Desktop/cluster/data/99_BroadAnnotation/04_requantification_broad_notstrand/"
abundance_files <- iterate_files(dir_counts_ref, "*.tsv")

# Read all files
abundances <- list()
for(i in 1:length(abundance_files)) {
  file <- readr::read_tsv(abundance_files[i])
  # Sum up values for double entries (https://github.com/gpertea/stringtie/issues/192)
  file <- file %>% dplyr::group_by(`Gene ID`)%>% dplyr::summarise(TPM = sum(TPM))
  abundances[[i]] <- data.frame(file$TPM, row.names =file$`Gene ID` )
}

# Summarize all TPMs from all quantification files 
rn <- rownames(abundances[[1]])
dat <- abundances[[1]]
for(i in 2:length(abundances)) {
  dat <- merge(dat, abundances[[i]],  by= "row.names", all.x= F, all.y= F) [,-1]
  rownames(dat) <- rn
}

expression <- dat

expression_novel <- expression[rownames(expression) %in% candidates_broad$gene_id,]
table(rowSums(as.data.frame(expression_novel > 0.0)) > 3)




table(seqnames(candidates_broad[candidates_broad$gene_id %in% rownames(expression_novel[rowSums(as.data.frame(expression_novel > 0.0)) > 0,]), ]))

```
