---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---

# ORTHOLOGS ANALYSIS 


### Imports and prepare transcripts
```{r beg, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)
# Library
library(networkD3)
library(dplyr)

col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))


theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

plot_sankey <- function(df_pairs, group = c("type_a","type_a","type_b","type_b","type_b", "type_b", "type_b")){
  # A connection data frame is a list of flows with intensity for each flow
links <- data.frame(
  source=df_pairs$source, 
  target=df_pairs$target, 
  value=df_pairs$value
  )
 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes <- data.frame(
  name=c(as.character(links$source), 
  as.character(links$target)) %>% unique()
)
 
# With networkD3, connection must be provided using id, not using real name like in the links dataframe.. So we need to reformat it.
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1

links$group <- as.factor(group)
 
# Add a 'group' column to each node. Here I decide to put all of them in the same group to make them grey
nodes$group <- as.factor(c("my_unique_group"))
my_color <- 'd3.scaleOrdinal() .domain(["type_a", "type_b", "type_c", "my_unique_group"]) .range(["orange","#994C00", "navy", "darkgrey"])'


# Make the Network
 p <- sankeyNetwork(Links = links, Nodes = nodes,
              Source = "IDsource", Target = "IDtarget",
              Value = "value", NodeID = "name", 
              sinksRight=FALSE, colourScale=my_color, nodeWidth=2, fontSize=15, nodePadding=35, LinkGroup="group", NodeGroup="group", fontFamily = "Arial")
return(p)
}

```

## Import files

```{r cpc2}

# Check for all orthologs 
orthologs <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.txt")
orthologs <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.top.txt")
names(orthologs)<- c("lnc","lncGeneSymbol","ortholog","orthologGeneSymbol","alignScore","exonID","locusID","indelRate(exon)" ,"indelRate(intron)","lncExonsAligned","orthExonsAligned","spliceConserved","spliceTotal","category(mmul10)","category(hg38)")

all <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/OLD/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf")
ref <- all
```


# Look for nice examples 

```{r cpc2}
all_stringtie_novellnc <- all[all$source == "StringTie", ]
length(unique(all_stringtie_novellnc))
all_annot <- all[all$source != "StringTie", ]
all_annot_lnc <- all_annot[all_annot$type == "transcript" & all_annot$transcript_biotype == "lncRNA",]
length(unique(all_annot_lnc$gene_id))

orthologs_novel <- orthologs[substr(orthologs$lnc,1,4) == "MSTR",]
orthologs_lncAnnot <- orthologs[orthologs$lnc %in% all_annot_lnc$transcript_id,]
length(unique(orthologs_lncAnnot$lnc))

all[!is.na(all$transcript_id) & all$transcript_id == "MSTRG.193074.1",]

orthologs_novel[startsWith(as.character(orthologs_novel$orthologGeneSymbol), "IFN"),]
```

# Plot subtypes correspondence

```{r cpc2}
prep_pairs <- function(orthologs_novel){
  pairs <- table(orthologs_novel %>% group_by(lnc) %>% summarise(a = paste(`category(mmul10)`, `category(hg38)`)))
  df_pairs <- data.frame(pairs)
  names(df_pairs) <- c("pair", "value")
  df_pairs$pair <- unlist(lapply(as.character(df_pairs$pair), function(x) str_replace_all(x, "divergent", "antisense")))
  df_pairs$source <- unlist(lapply(df_pairs$pair, function(x) str_split(x, " ")[[1]][1]))
  df_pairs$target <- unlist(lapply(df_pairs$pair, function(x) str_split(x, " ")[[1]][2]))
  df_pairs$target <- paste(df_pairs$target, "_human")
  return(df_pairs)
}


#---------------------------------------------
# Check the orthologs for novel lncRNAs 
#---------------------------------------------
df_pairs_novel <- prep_pairs(orthologs_novel)
df_pairs_red_novel <- df_pairs_novel[df_pairs_novel$source != "sno_host",]
# How many do have an ortholog? 
(sum(df_pairs_red_novel$value)/length(unique(all_stringtie_novellnc$transcript_id)))*100

# Plot
plot_sankey(df_pairs_red_novel)

df_pairs_novel_complete <- df_pairs_novel
df_pairs_novel_complete$type <- "Novel"




#---------------------------------------------
# Check the orthologs for novel lncRNAs 
#---------------------------------------------
df_pairs_annotated <- prep_pairs(orthologs_lncAnnot)
df_pairs_red_annotated <- df_pairs_annotated[!(df_pairs_annotated$source %in%  c("sno_host", "mir_host_exon", "mir_host_intron")),]
(sum(df_pairs_red_annotated$value)/length(unique(all_annot_lnc$transcript_id)))*100
plot_sankey(df_pairs_red_annotated,as.character(as.integer(as.factor(df_pairs_red_annotated$source))))


saveRDS(orthologs, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs.rds")

```




# --- Compare numbers 
# ----- Transcript level 

```{r cpc2}


orthologs_novel_found <- sum(orthologs$lnc %in% unique(all_stringtie_novellnc$transcript_id))
novel_no_ortholog <- length(unique(all_stringtie_novellnc$transcript_id))-orthologs_novel_found

orthologs_annot_found <- sum(orthologs$lnc %in% unique(all_annot_lnc$transcript_id))
annot_no_ortholog <- length(unique(all_annot_lnc$transcript_id))-orthologs_annot_found

freqs <- c( 
           novel_no_ortholog/length(unique(all_stringtie_novellnc$transcript_id)), 
           orthologs_novel_found/length(unique(all_stringtie_novellnc$transcript_id)),
           annot_no_ortholog/length(unique(all_annot_lnc$transcript_id)),
            orthologs_annot_found/length(unique(all_annot_lnc$transcript_id)))


df <- data.frame(count = c(novel_no_ortholog,orthologs_novel_found,annot_no_ortholog,orthologs_annot_found ), type = c("Novel", "Novel", "Annotated", "Annotated"), orth = c( "Ortholog NOT found","Ortholog found", "Ortholog NOT found", "Ortholog found"),freq = freqs) 

df$orth <- relevel(df$orth, ref = "Ortholog NOT found") 




# Plot with proportions
ggplot(df, aes(x = type, fill = orth, y = freq))+geom_bar(stat = "identity", alpha  = 0.6, position = "stack")+theme_paper+xlab("")+scale_fill_manual(values = c("grey","red"))+ylab("Proportion")


```
# --- Compare numbers 
# ----- Gene level  (not in paper)

```{r cpc2}

# Define a gene to be an ortholog when any of its transcripts is an ortholog


# Table correspondence between transcripts and gene IDs 
df_ref <- distinct(data.frame(ref[!is.na(ref$transcript_id),]$transcript_id, ref[!is.na(ref$transcript_id),]$gene_id, stringsAsFactors = F))
names(df_ref) <- c("transcript_id", "gene_id")
rownames(df_ref) <- df_ref$transcript_id


length(unique(orthologs$lnc))
length(df_ref[orthologs$lnc,]$gene_id)
length(unique(df_ref[orthologs$lnc,]$gene_id))

# Genes for which we identified an ortholog
orthologs_novel <- orthologs[orthologs$lnc %in% all_stringtie_novellnc$transcript_id,]
orthologs_annot <- orthologs[orthologs$lnc %in% all_annot_lnc$transcript_id,]




# For how many novel lncRNAs did we find an ortholog (GENE) 
orthologs_novel_found <- sum(unique(df_ref[df_ref$transcript_id %in% orthologs_novel$lnc,]$gene_id) %in% unique(all_stringtie_novellnc$gene_id))
novel_no_ortholog <- length(unique(all_stringtie_novellnc$gene_id))-orthologs_novel_found

orthologs_annot_found <- sum(unique(df_ref[df_ref$transcript_id %in% orthologs_annot$lnc,]$gene_id)   %in% unique(all_annot_lnc$gene_id))
annot_no_ortholog <- length(unique(all_annot_lnc$gene_id))-orthologs_annot_found

freqs <- c( 
           novel_no_ortholog/length(unique(all_stringtie_novellnc$gene_id)), 
           orthologs_novel_found/length(unique(all_stringtie_novellnc$gene_id)),
           annot_no_ortholog/length(unique(all_annot_lnc$gene_id)),
            orthologs_annot_found/length(unique(all_annot_lnc$gene_id)))


df <- data.frame(count = c(novel_no_ortholog,orthologs_novel_found,annot_no_ortholog,orthologs_annot_found ), type = c("Novel", "Novel", "Annotated", "Annotated"), orth = c( "Ortholog NOT found","Ortholog found", "Ortholog NOT found", "Ortholog found"),freq = freqs) 
df$orth <- relevel(df$orth, ref = "Ortholog NOT found") 

# Plot with proportions
ggplot(df, aes(x = type, fill = orth, y = freq))+geom_bar(stat = "identity", alpha  = 0.6, position = "stack")+theme_paper+xlab("")+scale_fill_manual(values = c("grey","red"))+ylab("Proportion")


ggplot(df[df$orth == "Ortholog found",], aes(x = type, fill = type, y = freq*100))+geom_bar(stat = "identity", alpha  = 0.6)+theme_paper+xlab("")+scale_fill_manual(values = c("dark grey", "grey"))+ylab("% lncRNAs (GENEs) ")




```