---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---

# ORTHOLOGS ANALYSIS 
## How many orthologs do we find? 
## Plus prepare object for future analyses 

### Imports
```{r beg, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)
library(networkD3)
library(dplyr)


# Import Utils
source(file.path("../utils/00_datapaths.R"))
source(file.path("../utils/01_lncrna_annotation_utils.R"))


theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
```

## Import files

```{r cpc2}

# Check for all orthologs 
orthologs <- read.table(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.txt"))
orthologs <- read.table(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.top.txt"))
names(orthologs)<- c("lnc","lncGeneSymbol","ortholog","orthologGeneSymbol","alignScore","exonID","locusID","indelRate(exon)" ,"indelRate(intron)","lncExonsAligned","orthExonsAligned","spliceConserved","spliceTotal","category(mmul10)","category(hg38)")

# Import files 
all <- import(file.path(data_path,"01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ref <- all
all_novel_lnc <- all[substr(all$gene_id,1,4) %in% c( "MSTR"),]

# Dataframe correspondence transcripts - genes
df_ref <- distinct(data.frame(ref[!is.na(ref$transcript_id),]$transcript_id, ref[!is.na(ref$transcript_id),]$gene_id, stringsAsFactors = F))
names(df_ref) <- c("lnc", "gene_id")
rownames(df_ref) <- df_ref$transcript_id

# Get unique dataframe
orthologs <- merge(orthologs, df_ref, by = "lnc")
#saveRDS(orthologs,file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))

```

# Filter table to have 1 to one
```{r cpc2}
orthologs <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))

# Only keep relevant columns 
orthologs_red <- distinct(orthologs[,c("gene_id","lnc", "orthologGeneSymbol","alignScore")])

# 1. Retain maximum score of alignment per ortholog
orthologs_max <- orthologs_red %>% dplyr::group_by(orthologGeneSymbol) %>% dplyr::summarise(alignScore=max(alignScore))
orthologs_max$key <- paste(orthologs_max$orthologGeneSymbol, orthologs_max$alignScore, sep ="_")
orthologs_red$key <- paste(orthologs_red$orthologGeneSymbol, orthologs_red$alignScore, sep ="_")

# 2. Retain maximum score of alignment per gene
#orthologs_max_gene <- orthologs_red %>% dplyr::group_by(gene_id) %>% dplyr::summarise(alignScore=max(alignScore))
#orthologs_max_gene$key_gene <- paste(orthologs_max_gene$gene_id, orthologs_max_gene$alignScore, sep ="_")
#orthologs_red$key_gene <- paste(orthologs_red$gene_id, orthologs_red$alignScore, sep ="_")


# Join 
orthologs_max_ortholog <- orthologs_max %>% dplyr::left_join(orthologs_red, by = c("key"))

colnames(orthologs_max_ortholog) <- gsub("orthologGeneSymbol.x","orthologGeneSymbol",colnames(orthologs_max_ortholog))
colnames(orthologs_max_ortholog) <- gsub("gene_id.x","gene_id",colnames(orthologs_max_ortholog))
orthologs_max_ortholog <- distinct(orthologs_max_ortholog[,c("orthologGeneSymbol", "gene_id", "alignScore.x")])

#  Retain maximum score of alignment per gene
orthologs_max_gene <- orthologs_max_ortholog %>% dplyr::group_by(gene_id) %>% dplyr::summarise(alignScore=max(alignScore.x))
orthologs_max_gene$key_gene <- orthologs_max_gene$key_gene <- paste(orthologs_max_gene$gene_id, orthologs_max_gene$alignScore, sep ="_")
orthologs_max_ortholog$key_gene <- paste(orthologs_max_ortholog$gene_id, orthologs_max_ortholog$alignScore.x, sep ="_")
orthologs_max_gene <- orthologs_max_gene %>% dplyr::left_join(orthologs_max_ortholog, by = c("key_gene"))

# Check all is fine 
orthologs_max_gene %>% dplyr::group_by(gene_id.x) %>% dplyr::count(orthologGeneSymbol) %>% dplyr::filter(n>1)


colnames(orthologs_max_gene) <- gsub("orthologGeneSymbol.x","orthologGeneSymbol",colnames(orthologs_max_gene))
colnames(orthologs_max_gene) <- gsub("gene_id.x","gene_id",colnames(orthologs_max_gene))
orthologs_max_gene <- distinct(orthologs_max_gene[,c("orthologGeneSymbol", "gene_id", "alignScore.x")])

orthologs <- orthologs_max_gene
#saveRDS(orthologs_max_gene, file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
```


# How many ortholog did we identify? (GENE LEVEL)
```{r cpc2}


counts <- c( 
           length(novel_lncrnas)-length(novel_with_ortholog),
           length(novel_with_ortholog),
           length(annotated_lncrnas)-length(annotated_with_ortholog),
          length(annotated_with_ortholog))

freqs <- c( 
           (length(novel_lncrnas)-length(novel_with_ortholog))/length(novel_lncrnas), 
           length(novel_with_ortholog)/length(novel_lncrnas),
           (length(annotated_lncrnas)-length(annotated_with_ortholog))/length(annotated_lncrnas),
          length(annotated_with_ortholog)/length(annotated_lncrnas))


df <- data.frame(counts = counts, type = c("novel", "novel", "annotated", "annotated"), orth = c( "Ortholog NOT found","Ortholog found", "Ortholog NOT found", "Ortholog found"),freq = freqs) 
df$orth <- relevel(factor(df$orth), ref = "Ortholog NOT found") 

# Plot with proportions
p_o <- ggplot(df, aes(x = type, fill = orth, y = freq))+geom_bar(stat = "identity", alpha  = 0.6, position = "stack")+xlab("")+scale_fill_manual(values = c("grey","#525252"))+ylab("proportion of genes with human ortholog")+
          theme(panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"),
              axis.ticks.length=unit(.2, "cm"),
              axis.text= element_text(size = 15, color = "black"),
              axis.title = element_text(size = 15),
              plot.title = element_text(hjust = 0.5, size = 17),
              axis.line.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.background = element_rect(fill = "white"))+
        scale_y_continuous(limits =c(0,1), labels = seq(0,1, 0.25),breaks = seq(0,1, 0.25), expand = c(0,0))

#pdf(file.path(plots, "01/orthologs.pdf"), width = 4.5, height = 6)
p_o
#dev.off()
```

# Check lengths of novel lncRNAs with an ortholog and their ortholog ( To check we identify full transcripts )

```{r summary}

# 1. Raw data
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
orthologs$orthologGeneSymbol <- as.character(orthologs$orthologGeneSymbol )

orthologs_full <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))
orthologs_full <- orthologs_full[,c("ortholog", "orthologGeneSymbol", "gene_id")]
orthologs_full$ortholog <- as.character(orthologs_full$ortholog)
orthologs_full$orthologGeneSymbol <- as.character(orthologs_full$orthologGeneSymbol)
human_transcripts_orthologs_set <- unique(orthologs_full$ortholog)

human_reference <- import(file.path(datadir, "05_orthologs/gencode.v23.annotation.gtf"))
# Only keep transcripts found 
human_reference <- human_reference[human_reference$transcript_id %in% human_transcripts_orthologs_set, ]
length(unique(human_reference$transcript_id)) == length(unique(orthologs_full$ortholog))

corr_human <- distinct(data.frame(ortholog = human_reference[human_reference$type == "transcript", ]$transcript_id, ortholog_gene_id = human_reference[human_reference$type == "transcript", ]$gene_id, stringsAsFactors = F)) 

complete_orthologs_human<- distinct(merge(orthologs_full, corr_human))
complete_orthologs_human <- unique(complete_orthologs_human[, c("gene_id", "ortholog_gene_id")])

# 2. Extract only lncRNAs (novel and annotated) with human ortholog
#novel_with_ortholog
#annotated_with_ortholog
length(unique(novel_with_ortholog))
length(unique(annotated_with_ortholog))

novel <- complete_orthologs_human[complete_orthologs_human$gene_id %in% novel_with_ortholog,]
annotated <- complete_orthologs_human[complete_orthologs_human$gene_id %in% annotated_with_ortholog,]


novel_gr <- all[all$gene_id %in% novel_with_ortholog,]
annotated_gr <- all[all$gene_id %in% annotated_with_ortholog,]



# Calculate transcripts lengths 
novel_lengths <- data.frame(calc_transcript_length(novel_gr, "1.  novel lncRNA"), stringsAsFactors = F)
annotated_lengths <- data.frame(calc_transcript_length(annotated_gr, "2.  annotated lncRNA"))

lncrnas_lengths <- rbind(novel_lengths, annotated_lengths)



```


```{r summary}
novel_orthologs_df <- distinct(orthologs[orthologs$gene_id %in% unique(all_novel_lnc$gene_id),c("gene_id", "orthologGeneSymbol", "alignScore.x")])
novel_orthologs_df$orthologGeneSymbol <- as.character(novel_orthologs_df$orthologGeneSymbol)
novel_orthologs_df <- novel_orthologs_df[!is.na(novel_orthologs_df$alignScore.x),] 
length(unique(novel_orthologs_df$gene_id))
length(unique(novel_orthologs_df$orthologGeneSymbol))

corr_human <- distinct(data.frame(t = lncRNAs_ref_human[lncRNAs_ref_human$type == "transcript", ]$transcript_id, g = lncRNAs_ref_human[lncRNAs_ref_human$type == "transcript", ]$gene_id, stringsAsFactors = F)) 
rownames(corr_human) <- corr_human$t

# Extract from rthologs full the human ortholog transcript 
orthologs_full <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))
# Extract those lines representing a novel lnc 
orthologs_full <- orthologs_full[orthologs_full$gene_id %in% novel_orthologs_df$gene_id, ]
orthologs_full <- unique(orthologs_full[, c("gene_id", "ortholog")])
orthologs_full$t_orth <- unique(unlist(lapply(as.character(orthologs_full$ortholog) , function(x) strsplit(x, '.', fixed = T)[[1]][1])))
orthologs_full$human_gene_id <- corr_human[orthologs_full$t_orth,]$g

orthologs_full_reduced <- orthologs_full[,c("gene_id", "human_gene_id")]
orthologs_full_reduced <- orthologs_full_reduced[!is.na(orthologs_full_reduced$human_gene_id),]
orthologs_full_reduced <- unique(orthologs_full_reduced[!is.na(orthologs_full_reduced$gene_id),])

length(unique(orthologs_full_reduced$gene_id))
length(unique(orthologs_full_reduced$human_gene_id))
novel_with_ortholog_gr <- all_novel_lnc[all_novel_lnc$gene_id %in% orthologs_full_reduced$gene_id,]

human_orth_novel <- lncRNAs_ref_human[lncRNAs_ref_human$gene_id %in% orthologs_full_reduced$human_gene_id,]
length(unique(human_orth_novel$gene_id))
length(unique(novel_with_ortholog_gr$gene_id))

orthologs_full[orthologs_full$gene_id == "ENSG00000152931",]
orthologs_full[orthologs_full$gene_id == "MSTRG.202113",]
df[df$gene_id == "MSTRG.202113", ]
test <- all_novel_lnc[all_novel_lnc$gene_id == "MSTRG.202113",]
test <- human_orth_novel[human_orth_novel$gene_id == "ENSG00000152931",]
calc_transcript_length(test, "")

gr <- test
gr <- gr[gr$type =="exon",]
gr
gr <- get_only_max_transcript(gr)
gr
df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
df
collapsed <- df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
collapsed["type"] <- type
return(collapsed)

mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_proteincoding.gtf"))



orthologs[orthologs$gene_id == "MSTRG.202113",]
humans[humans$name == "PART1",]
```
