---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---

# ORTHOLOGS ANALYSIS 
## How many orthologs do we find? 
## Plus prepare object for future analyses 

### Imports
```{r beg, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)
library(networkD3)
library(dplyr)


# Import Utils
source(file.path("../utils/00_datapaths.R"))
source(file.path("../utils/01_lncrna_annotation_utils.R"))

# Reoccurring paths
datadir <- file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/")

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
```

## Import files

```{r cpc2}

# Check for all orthologs 
orthologs <- read.table(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.txt"))
orthologs <- read.table(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/slncky-master/out.orthologs.top.txt"))
names(orthologs)<- c("lnc","lncGeneSymbol","ortholog","orthologGeneSymbol","alignScore","exonID","locusID","indelRate(exon)" ,"indelRate(intron)","lncExonsAligned","orthExonsAligned","spliceConserved","spliceTotal","category(mmul10)","category(hg38)")

# Import files 
all <- import(file.path(data_path,"01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ref <- all
all_novel_lnc <- all[substr(all$gene_id,1,4) %in% c( "MSTR"),]

# Dataframe correspondence transcripts - genes
df_ref <- distinct(data.frame(ref[!is.na(ref$transcript_id),]$transcript_id, ref[!is.na(ref$transcript_id),]$gene_id, stringsAsFactors = F))
names(df_ref) <- c("lnc", "gene_id")
rownames(df_ref) <- df_ref$transcript_id

# Get unique dataframe
orthologs <- merge(orthologs, df_ref, by = "lnc")
#saveRDS(orthologs,file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))

```

# Filter table to have 1 to one
```{r cpc2}
orthologs <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))

# Only keep relevant columns 
orthologs_red <- distinct(orthologs[,c("gene_id","lnc", "orthologGeneSymbol","alignScore")])

# 1. Retain maximum score of alignment per ortholog
orthologs_max <- orthologs_red %>% dplyr::group_by(orthologGeneSymbol) %>% dplyr::summarise(alignScore=max(alignScore))
orthologs_max$key <- paste(orthologs_max$orthologGeneSymbol, orthologs_max$alignScore, sep ="_")
orthologs_red$key <- paste(orthologs_red$orthologGeneSymbol, orthologs_red$alignScore, sep ="_")

# 2. Retain maximum score of alignment per gene
#orthologs_max_gene <- orthologs_red %>% dplyr::group_by(gene_id) %>% dplyr::summarise(alignScore=max(alignScore))
#orthologs_max_gene$key_gene <- paste(orthologs_max_gene$gene_id, orthologs_max_gene$alignScore, sep ="_")
#orthologs_red$key_gene <- paste(orthologs_red$gene_id, orthologs_red$alignScore, sep ="_")


# Join 
orthologs_max_ortholog <- orthologs_max %>% dplyr::left_join(orthologs_red, by = c("key"))

colnames(orthologs_max_ortholog) <- gsub("orthologGeneSymbol.x","orthologGeneSymbol",colnames(orthologs_max_ortholog))
colnames(orthologs_max_ortholog) <- gsub("gene_id.x","gene_id",colnames(orthologs_max_ortholog))
orthologs_max_ortholog <- distinct(orthologs_max_ortholog[,c("orthologGeneSymbol", "gene_id", "alignScore.x")])

#  Retain maximum score of alignment per gene
orthologs_max_gene <- orthologs_max_ortholog %>% dplyr::group_by(gene_id) %>% dplyr::summarise(alignScore=max(alignScore.x))
orthologs_max_gene$key_gene <- orthologs_max_gene$key_gene <- paste(orthologs_max_gene$gene_id, orthologs_max_gene$alignScore, sep ="_")
orthologs_max_ortholog$key_gene <- paste(orthologs_max_ortholog$gene_id, orthologs_max_ortholog$alignScore.x, sep ="_")
orthologs_max_gene <- orthologs_max_gene %>% dplyr::left_join(orthologs_max_ortholog, by = c("key_gene"))

# Check all is fine 
orthologs_max_gene %>% dplyr::group_by(gene_id.x) %>% dplyr::count(orthologGeneSymbol) %>% dplyr::filter(n>1)


colnames(orthologs_max_gene) <- gsub("orthologGeneSymbol.x","orthologGeneSymbol",colnames(orthologs_max_gene))
colnames(orthologs_max_gene) <- gsub("gene_id.x","gene_id",colnames(orthologs_max_gene))
orthologs_max_gene <- distinct(orthologs_max_gene[,c("orthologGeneSymbol", "gene_id", "alignScore.x")])

orthologs <- orthologs_max_gene
#saveRDS(orthologs_max_gene, file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
annotated_lncrnas <- (unique(all[!is.na(all$gene_biotype) & all$gene_biotype == "lncRNA", ]$gene_id))
novel_lncrnas <- (unique(all_novel_lnc$gene_id))

novel_with_ortholog <- unique(orthologs[orthologs$gene_id %in% unique(all_novel_lnc$gene_id),]$gene_id)
length(unique(novel_with_ortholog))
annotated_with_ortholog <-unique(orthologs[orthologs$gene_id %in% annotated_lncrnas,]$gene_id)
length(unique(annotated_with_ortholog))

length(novel_with_ortholog)/sum(length(novel_with_ortholog),length(annotated_with_ortholog))

```


# How many ortholog did we identify? (GENE LEVEL)
```{r cpc2}


counts <- c( 
           length(novel_lncrnas)-length(novel_with_ortholog),
           length(novel_with_ortholog),
           length(annotated_lncrnas)-length(annotated_with_ortholog),
          length(annotated_with_ortholog))

freqs <- c( 
           (length(novel_lncrnas)-length(novel_with_ortholog))/length(novel_lncrnas), 
           length(novel_with_ortholog)/length(novel_lncrnas),
           (length(annotated_lncrnas)-length(annotated_with_ortholog))/length(annotated_lncrnas),
          length(annotated_with_ortholog)/length(annotated_lncrnas))


df <- data.frame(counts = counts, type = c("novel", "novel", "annotated", "annotated"), orth = c( "Ortholog NOT found","Ortholog found", "Ortholog NOT found", "Ortholog found"),freq = freqs) 
df$orth <- relevel(factor(df$orth), ref = "Ortholog NOT found") 

# Plot with proportions
p_o <- ggplot(df, aes(x = type, fill = orth, y = freq))+geom_bar(stat = "identity", alpha  = 0.6, position = "stack")+xlab("")+scale_fill_manual(values = c("grey","#525252"))+ylab("proportion of genes with human ortholog")+
          theme(panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              axis.line = element_line(colour = "black"),
              axis.ticks.length=unit(.2, "cm"),
              axis.text= element_text(size = 15, color = "black"),
              axis.title = element_text(size = 15),
              plot.title = element_text(hjust = 0.5, size = 17),
              axis.line.x=element_blank(),
              axis.ticks.x=element_blank(),
              panel.background = element_rect(fill = "white"))+
        scale_y_continuous(limits =c(0,1), labels = seq(0,1, 0.25),breaks = seq(0,1, 0.25), expand = c(0,0))

#pdf(file.path(plots, "01/orthologs.pdf"), width = 4.5, height = 6)
p_o
#dev.off()
```

# Check lengths of novel lncRNAs with an ortholog and their ortholog ( To check we identify full transcripts )

```{r summary}

# 1. Raw data
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
orthologs$orthologGeneSymbol <- as.character(orthologs$orthologGeneSymbol )

orthologs_full <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))
orthologs_full <- orthologs_full[,c("ortholog", "orthologGeneSymbol", "gene_id")]
orthologs_full$ortholog <- as.character(orthologs_full$ortholog)
orthologs_full$orthologGeneSymbol <- as.character(orthologs_full$orthologGeneSymbol)
human_transcripts_orthologs_set <- unique(orthologs_full$ortholog)
orthologs_full <- merge(orthologs, orthologs_full)

human_reference <- import(file.path(datadir, "05_orthologs/gencode.v23.annotation.gtf"))
# Only keep transcripts found 
human_reference <- human_reference[human_reference$transcript_id %in% human_transcripts_orthologs_set, ]
length(unique(human_reference$transcript_id)) == length(unique(orthologs_full$ortholog))

corr_human <- distinct(data.frame(ortholog = human_reference[human_reference$type == "transcript", ]$transcript_id, ortholog_gene_id = human_reference[human_reference$type == "transcript", ]$gene_id, stringsAsFactors = F)) 

complete_orthologs_human<- distinct(merge(orthologs_full, corr_human))
complete_orthologs_human <- unique(complete_orthologs_human[, c("gene_id", "ortholog_gene_id", "orthologGeneSymbol")])

# 2. Extract only lncRNAs (novel and annotated) with human ortholog
# Novel_with_ortholog
# Annotated_with_ortholog
length(unique(novel_with_ortholog))
novel <- complete_orthologs_human[complete_orthologs_human$gene_id %in% novel_with_ortholog,]
novel_gr <- all[all$gene_id %in% novel_with_ortholog,]
# Get lengths of human orthologs 
complete_orthologs_human_novel <- complete_orthologs_human[complete_orthologs_human$gene_id %in% novel_with_ortholog,]
humans_gr_novelorthologs <- human_reference[human_reference$gene_id %in% complete_orthologs_human_novel$ortholog_gene_id, ]
length(unique(humans_gr_novelorthologs$gene_id))

df_novel  <- data.frame(calc_transcript_length(novel_gr, "1.  novel lncRNA"))
df_human <- data.frame(calc_transcript_length(humans_gr_novelorthologs, "2. human ortholog "))
df <- rbind(df_novel, df_human)

# Remove outliers 
outliers <- df[df$range > 6000, ]$gene_id
novel_outliers <- complete_orthologs_human_novel[complete_orthologs_human_novel$gene_id %in% outliers, ]
outlier_ids_1 <- c(novel_outliers$gene_id, novel_outliers$ortholog_gene_id)
human_outliers <- complete_orthologs_human_novel[complete_orthologs_human_novel$ortholog_gene_id %in% outliers, ]
outlier_ids_2 <- c(human_outliers$gene_id, human_outliers$ortholog_gene_id)
outlier_ids <- c(outlier_ids_1, outlier_ids_2)
df_clean <- df[!(df$gene_id %in% outlier_ids),]

df <- df_clean
# ------------------- PLOT ---------------------------------
palette = c("#f72631", "dark grey", "#fc6d70", "grey")
df$type <- factor(df$type)
ylab <- seq(0,6,1)
my_comparisons <- list( c("1.  novel lncRNA", "2. human ortholog") )
p <- ggboxplot(df,x = "type",  y = "range", fill = "type", palette = palette) +
    labs( x = "", y = "transcript length (bp)" )+ stat_compare_means(aes(label = ..p.signif..),comparisons = my_comparisons)+
    theme(legend.position = "",axis.text.y = element_text(size = 20, color = "black"), axis.title = element_text(size = 20))
pdf(file.path(plots, "01/comparison_length_ortholog.pdf"), width = 4, height = 7)
p
dev.off()


rownames(df) <- df$gene_id
ordered_lengths_novel <- df[complete_orthologs_human_novel$gene_id,]$range
ordered_lengths_human <- df[complete_orthologs_human_novel$ortholog_gene_id,]$range
wilcox.test(ordered_lengths_novel, ordered_lengths_human, paired = T)

# How many datapoints for comparison 
nrow(df)/2

```
