---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---

```{r beg}
knitr::opts_chunk$set(echo = TRUE)
library(PCAtools)
library(Biobase)
library(GEOquery)
library(rtracklayer)
library(stringr)

col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/code/ebola/analysis/utils/")
source(file.path(scripts_dir,"03_de_utils.R"));

datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))

dir_counts_ref <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/04_quantification/"


# --------------------------------------------------------
# Obtain quantification for all genes across all samples 
# --------------------------------------------------------

abundance_files <- iterate_files(dir_counts_ref, "*.tsv")
# Read all files
abundances <- list()
for(i in 1:length(abundance_files)) {
  file <- readr::read_tsv(abundance_files[i])
  # Sum up values for double entries (https://github.com/gpertea/stringtie/issues/192)
  file <- file %>% dplyr::group_by(`Gene ID`)%>% dplyr::summarise(TPM = sum(TPM))
  abundances[[i]] <- data.frame(file$TPM, row.names =file$`Gene ID` )
}
# Summarize all TPMs from all quantification files 
rn <- rownames(abundances[[1]])
dat <- abundances[[1]]
for(i in 2:length(abundances)) {
  dat <- merge(dat, abundances[[i]],  by= "row.names", all.x= F, all.y= F) [,-1]
  rownames(dat) <- rn
}

expression <- dat
expression <- as.matrix(expression)
colnames(expression) <- unlist(lapply(abundance_files, function(file) strsplit(rev(strsplit(file, "/")[[1]])[1],".", fixed = T)[[1]][1]))

extract_info_sample_from_dir <- function(file){
  file_no_ext <- strsplit(file,".", fixed = T)[[1]][1]
  info <- rev(strsplit(file, "/")[[1]])
  dataset <- info[5]
  tissue <- info[4]
  dpo <- info[3]
  sample <- info[2]
  file_name <- info[1]
  sample_name=paste(dataset,tissue,dpo,sample,sep="_")
  df <- data.frame(id = sample_name, dataset = dataset , dpo = dpo , sample = sample, tissue = tissue , stringsAsFactors = F)
  return(df)
}
samples_info <- do.call(rbind,(lapply(abundance_files,extract_info_sample_from_dir)))
rownames(samples_info) <- samples_info$id

log_expression <- log(expression+0.01)
all(colnames(expression) == rownames(samples_info))


p <- pca(log_expression, metadata = samples_info, removeVar = 0.1)

screeplot(p, axisLabSize = 18, titleLabSize = 22)
biplot(p, lab = NULL)

p$metadata$dpo
library(stringr)
p$metadata$tissue <- unlist(lapply(p$metadata$tissue, function(x) str_split(x, "-")[[1]][1] ))
biplot(p,lab = NULL,
    colby = 'tissue',
    hline = 0, vline = 0,
    legendPosition = 'right')


biplot(p,lab = NULL,
    colby = 'tissue',
    hline = 0, vline = 0,
    legendPosition = 'right', 
    pointSize = 1.5)+theme_classic()


```


# Limma-voom for batch correction 

```{r beg}
library(edgeR)
# ----------------------------------
#           Import Utils 
# ----------------------------------
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/code/ebola/analysis/utils/")
source(file.path(scripts_dir,"03_de_utils.R"));

datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"

ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))

dir_counts_ref = file.path(datadir, "04b_quantification_htseq/")
htseq_files_ref <- iterate_files(dir_counts_ref, ".gene_counts.tab")

# Create matrix
list <- lapply(htseq_files_ref, function(x) readr::read_tsv(x, col_names=F))
df <- Reduce(
    function(x, y) merge(x, y, by = "X1", all = T),
    list)
df <- df[substr(df$X1,1,1) !=  "_", ]
rownames(df) <- df$X1
df <- df[,colnames(df)!="X1"]
expression_matrix <- data.matrix(df)
colnames(expression_matrix) <- unlist(lapply(htseq_files_ref, function(file) strsplit(rev(strsplit(file, "/")[[1]])[1],".", fixed = T)[[1]][1]))

d0 <- DGEList(expression_matrix)
d0 <- calcNormFactors(d0)
cutoff <- 1
drop <- which(apply(cpm(d0), 1, max) < cutoff)
d <- d0[-drop,] 
dim(d) # number of genes left
snames <- colnames(expression_matrix) # Sample names
cultivar <- substr(snames, 1, nchar(snames) - 2) 
time <- substr(snames, nchar(snames) - 1, nchar(snames) - 1)

dataset <-  unlist(lapply(snames, function(x) strsplit(x, "_")[[1]][1]))
tissue <-  unlist(lapply(snames, function(x) strsplit(x, "_")[[1]][2]))
dpo <-  unlist(lapply(snames, function(x) strsplit(x, "_")[[1]][3]))
sample <-  unlist(lapply(snames, function(x) strsplit(x, "_")[[1]][4]))
metadata <- data.frame(dataset=dataset, tissue = tissue, dpo = dpo , sample= sample, stringsAsFactors = F)
rownames(metadata) <- snames

all(colnames(expression_matrix) == rownames(metadata))


p <- pca(d$counts, metadata = metadata, removeVar = 0.1)

biplot(p,lab = NULL,
    colby = 'tissue',
    hline = 0, vline = 0,
    legendPosition = 'right', 
    pointSize = 1.5)+theme_classic()


# Specify the model to be fitted 
mm <- model.matrix(~0 + dataset)
y <- voom(d, mm, plot = T)
m <- y$weights
colnames(m)
p <- pca(y$weights, metadata = metadata, removeVar = 0.1)

biplot(p,lab = NULL,
    colby = 'dataset',
    hline = 0, vline = 0,
    legendPosition = 'right', 
    pointSize = 1.5)+theme_classic()
```



