---
title: "lncAnnotation performance Assessment"
author: "Luisa Santus"
date: "10/22/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data and libraries

```{r cars}
library(rtracklayer)
library(VennDiagram)
library(stringr)
library(ggplot2)
library(grid)
library(gridExtra)
library(UpSetR)
library(ComplexHeatmap)
library(GenomicRanges)
library(dplyr)
baseDir = "/home/luisas/Desktop/cluster/proj"
scripts_dir = file.path(baseDir,"code/ebola/src/scripts/")
source(paste0(scripts_dir,"de_utils.R"))

dir<- "/home/luisas/Desktop/cluster/proj/data"
dir_hisat_zyagen <- file.path(dir,"02_RNA-Seq/03_hisat/Zyagen")

lncRNAs_ref <- import(file.path(dir, "01_PreliminaryFiles/gene_annotations/rheMac8_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(dir, "01_PreliminaryFiles/gene_annotations/rheMac8_EBOV-Kikwit_mrnas.gtf"))
lncRNAs_out <- import(file.path(dir, "02_RNA-Seq/05_lncrnaAnnotation/feelnc_gencode_two_cut_offs_pred/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
candidates_lncRNA <- import(file.path(dir, "02_RNA-Seq/05_lncrnaAnnotation/feelnc_gencode_two_cut_offs/candidate_lncRNA.gtf"))
mRNAs_out <- import(file.path(dir, "02_RNA-Seq/05_lncrnaAnnotation/feelnc_gencode_two_cut_offs_pred/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
stringtie_gtf <- import(file.path(dir, "02_RNA-Seq/04_stringtie/Zyagen/Zyagen_stringtie_merged_reference_guided.gtf"))
```
We first Check if we have multiple isoforms per gene, we do we want to collapse them.
```{r duplicates}
df <- data.frame("gene" =  lncRNAs_out$gene_id, "transcript" = lncRNAs_out$transcript_id, stringsAsFactors = FALSE)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarise(number=dplyr::n()) %>% dplyr::filter(number > 1)
df[df$gene == "ENSMMUG00000037508",]
```
```{r duplicates}
library(dplyr)
df <- data.frame("gene" =  lncRNAs_ref$gene_id, "transcript" = lncRNAs_ref$transcript_id)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarise(number=dplyr::n()) %>% dplyr::filter(number > 1)
df[df$gene == "ENSMMUG00000000801",]
```

Keep Track of the number of novel genes identified: 
```{r}
get_novels <- function(lncRNAs_out, stringtie_gtf, lncRNAs_ref){
  intersect_feelnc_stringtie <- intersect(lncRNAs_out$gene_id, stringtie_gtf$gene_id)
  intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out$gene_id)
  novel_lncrnas = setdiff(intersect_feelnc_stringtie,intersect_reflnrna_out )
  return(novel_lncrnas)
}
novels <- list()
novel_lncrnas <- get_novels(lncRNAs_out, stringtie_gtf, lncRNAs_ref)
novels <- cbind(novels,"raw" = length(novel_lncrnas ))
```

Since there are unconcordant predictions (same gene - multiple transcripts - assigned to different classes : lncRNA AND mRNA)
```{r echo=T, results='hide', eval=FALSE}
## REMOVE UNCONCORDANT PREDICTIONS
# Get the genes with unconcordant prediction labels for transcripts 
df <- data.frame()
df <- rbind(df, data.frame(gene_id = lncRNAs_out$gene_id, transcript_id = lncRNAs_out$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)

# Remove unconcordant predictions from sets 
lncRNAs_out <-lncRNAs_out[!(lncRNAs_out$gene_id %in% unconcordant_prediction$gene_id), ]
mRNAs_out <- mRNAs_out[!(mRNAs_out$gene_id %in% unconcordant_prediction$gene_id), ]

# Double check there is for real no unconcordant prediction anymore
df <- data.frame()
df <- rbind(df, data.frame(gene_id = lncRNAs_out$gene_id, transcript_id = lncRNAs_out$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements = n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
nrow(unconcordant_prediction)

# NOVELS after filtering unconcordant predictions
novel_lncrnas <- get_novels(lncRNAs_out, stringtie_gtf, lncRNAs_ref)
novels <- cbind(novels,"after-unconcordant prediction" = length(novel_lncrnas ))
```
## Overlap of Gene_ids

we separate 

```{r}
grid::grid.newpage()
vd_draw(list(lncRNAs_ref$gene_id, lncRNAs_out$gene_id, stringtie_gtf$gene_id, mRNAs_ref$gene_id), c("LNCRNA-ref","feelnc","stringtie","MRNA-ref"))
```
```{r}
grid::grid.newpage()
vd_draw(list(lncRNAs_ref$gene_id, lncRNAs_out$gene_id, stringtie_gtf$gene_id), c("LNCRNA-ref","feelnc","stringtie"))
```
```{r}
grid::grid.newpage()
vd_draw(list(lncRNAs_ref$gene_id, lncRNAs_out$gene_id, stringtie_gtf$gene_id, mRNAs_ref$gene_id, mRNAs_out$gene_id), c("LNCRNA-ref","feelnc","stringtie","MRNA-ref", "Feelnc-mrna"))
```



```{r}
grid::grid.newpage()

lt <-list(lncRNAs_ref$gene_id, lncRNAs_out$gene_id, stringtie_gtf$gene_id, mRNAs_ref$gene_id, mRNAs_out$gene_id) 
names(lt) <- c("LNCRNA-ref","feelnc","stringtie","MRNA-ref", "Feelnc-mrna")
m1 = make_comb_mat(lt)
UpSet(m1)
```


```{r}
grid::grid.newpage()
intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out$gene_id)
intersect_reflnrna_stringtie <- intersect(lncRNAs_ref$gene_id, stringtie_gtf$gene_id)
intersect_feelnc_stringtie <- intersect(lncRNAs_out$gene_id, stringtie_gtf$gene_id)
intersect_reflnrna_mrnaout<- intersect(lncRNAs_ref$gene_id, mRNAs_out$gene_id)

correctly_identified = intersect(intersect_reflnrna_stringtie,lncRNAs_out$gene_id)
missed = setdiff(intersect_reflnrna_stringtie,correctly_identified)
total = intersect_reflnrna_stringtie
df <- data.frame()
df <- rbind(df, data.frame(dataset = "total", number = length(total), type = "lncrna"))
df <- rbind(df, data.frame(dataset = "correctly identified", number = length(correctly_identified), type = "lncrna"))
df <- rbind(df, data.frame(dataset = "missed", number = length(missed), type = "lncrna"))
df <- rbind(df, data.frame(dataset = "predicted as Mrna", number = length(intersect_reflnrna_mrnaout), type = "lncrna"))


p1 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge())+
  geom_text(aes(label=number), vjust=1.6, color="white",
            position = position_dodge(0.9), size=3.5)+
  scale_fill_brewer(palette="Paired")+
  theme_minimal() + 
  labs(title="Number genes identified as lncRNAs")

prop_missed = (length(missed)/length(total))*100
prop_correctly = (length(correctly_identified)/length(total))*100

df <- data.frame()
df <- rbind(df, data.frame(dataset = "correctly identified", number = prop_correctly, type = "lncrna"))
df <- rbind(df, data.frame(dataset = "missed", number = prop_missed, type = "lncrna"))
p2 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity")+
  geom_text(aes(label=number), vjust=1.8, color="white", size=3)+
  labs(title="Proportion of genes identified as lncRNAs")

grid.arrange(p1, p2, nrow = 1)
   
```

## About the novel ones

How many of the genes which were missed, were filtered out in the filter step? 
```{r}
grid::grid.newpage()
missed = setdiff(intersect_reflnrna_stringtie,correctly_identified)
candidates = candidates_lncRNA$gene_id
vd_draw(list(missed, candidates),c("missed","candidates"))
missed_because_of_filtering <- setdiff(missed,candidates)
print("Missed because of filtering")
print(length(missed_because_of_filtering))
```

# Collapse transcripts - only take the one with max length

```{r}

# Instead of collapsing we select the longest transcript among the transcripts in one gene
get_only_max_transcript <- function(gr){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  gene_with_multiple_isoforms <-df[!duplicated(df$transcript_id),] %>% dplyr::group_by(gene_id) %>% dplyr::summarise(number=dplyr::n()) %>% dplyr::filter(number > 1) 
  collapsed <-df[df$gene_id %in% gene_with_multiple_isoforms$gene_id,]%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width)) %>% dplyr::group_by(gene_id) %>% dplyr::slice(which.max(range))
  gr <- gr[gr$transcript_id %in% collapsed$transcript_id,]
  return(gr)
}
lncRNAs_out <- get_only_max_transcript(lncRNAs_out)
mRNAs_out <-  get_only_max_transcript(mRNAs_out)
```
## About the Novel ones

Distribution of number of exons in newly predicted ones
```{r}
grid::grid.newpage()

get_nr_exons <- function(gr){
  df <- data.frame("gene_id" = gr$gene_id,"exon_number" = as.numeric(gr$exon_number))
  number_exons <- df %>% dplyr::group_by(gene_id) %>%dplyr:: summarise(max_exon = max(exon_number))
  return(number_exons)
}


df_m <- data.frame(get_nr_exons(mRNAs_out))
df_m$type <- "mrnas"
df_l <- data.frame(get_nr_exons(lncRNAs_out))
df_l$type <- "lncrnas"
df <- rbind(df_m, df_l)
ggplot(df, aes(x=max_exon,col =type)) + 
  geom_histogram(aes(y=..density..), position="identity", alpha=0.5)+
  geom_density(alpha=0.6)+
  scale_color_manual(values=c("#999999", "#E69F00", "#56B4E9"))+
  scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9"))+
  theme_classic()

lncrnas_ref_exons <- get_only_max_transcript(lncRNAs_ref)
lncrnas_ref_exons <- lncRNAs_ref[lncRNAs_ref$type =="exon",]
df_ref_lncrnas <- get_nr_exons(lncrnas_ref_exons)

df_l$type <- "feelnc"
df_ref_lncrnas$type <- "ref"

df <- rbind(df_ref_lncrnas, df_l)
ggplot(df, aes(x=max_exon,col =type)) + 
  geom_histogram(position="identity", alpha=0.5)

ggplot(df, aes(x=max_exon,col =type)) + 
  geom_histogram(aes(y = stat(count / sum(count))),position="identity", alpha=0.5)

wilcox.test(df_l$max_exon, df_ref_lncrnas$max_exon)
```
Length of gene in novel predicted vs reference
```{r}
gr <- lncRNAs_out
df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
df$type <- "pred"
collapsed <-df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
collapsed$type <- "pred"
p1 <- ggplot(collapsed, aes(x=range)) + 
  geom_histogram(position="identity", alpha=0.5, fill = "blue")

gr <- lncrnas_ref_exons
df_ref <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
df_ref$type <- "ref"
collapsed_ref <-df_ref%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
collapsed_ref$type <- "ref"
p2 <- ggplot(collapsed_ref, aes(x=range)) + 
  geom_histogram(position="identity", alpha=0.5, fill= "pink")


all <- rbind(collapsed,collapsed_ref)
p3 <- ggplot(all, aes(x=range, col= type)) + 
  geom_histogram( position="identity", alpha=0.5) + 
  xlim(0,3000)

library("ggplotify")
library("grid")
g1 <- as.grob(p1)
g2 <- as.grob(p2)
g3 <- as.grob(p3)
grid.arrange(
  grobs = list(g1,g2,g3),
  layout_matrix = rbind(c(1, 2),
                        c(3,3))
)

```


## Expression Values 





## Load Needed
```{r echo=FALSE}
library(ggplot2)
library(GenomicFeatures)
library(zeallot)
library(tidyverse)
library(tximport)
library(tximportData)
library(edgeR)
library(SummarizedExperiment)
library(DESeq2)
library("RColorBrewer")
library("dplyr")
library(plyr)

# Load Utils
baseDir = "/home/luisas/Desktop/cluster/proj"
scripts_dir = file.path(baseDir,"code/ebola/src/scripts/")
source(paste0(scripts_dir,"de_utils.R"))
# Dirs
directory_zyagen = file.path(baseDir,"data/02_RNA-Seq/04_stringtie_counts/Zyagen")
```

## Create Summarized Experiment
```{r echo=FALSE, results='hide'}
ctabs<- iterate_files(directory_zyagen, 't_data.ctab')
ctabs_zyagen_umis <-ctabs[sapply(ctabs, function(str) str_detect(str,"umis"))]
se <-extract_se(ctabs_zyagen_umis)
```
We want to check WHY in the reference we have so many lncRNAS that are not present in stringtie.
1) I need to filter out the lowly expressed genes from the reference

```{r}
length(intersect(lncRNAs_ref$gene_id,stringtie_gtf$gene_id))
dge <- DGEList(counts = assays(se)$counts, group = se$tissue, genes = as.data.frame(mcols(se)))
dge <- calcNormFactors(dge)
saveRDS(dge, file.path("results", "dge_zyagen_umis.rds"))
```


Create the DGE List Object and calculate the normalization factors
```{r}
dge <- DGEList(counts = assays(se)$counts, group = se$tissue, genes = as.data.frame(mcols(se)))
dge <- calcNormFactors(dge)
saveRDS(dge, file.path("results", "dge_zyagen_umis.rds"))
```
We calc the logCPM values 
```{r}
assays(se)$logCPM <- cpm(dge, log = TRUE)
```
and filter out the lowly expressed genes
```{r}
mask <- rowMeans(assays(se)$logCPM) > 1
# How many we filtered out
length(mask[mask == FALSE])
dim(se)
se <- se[mask, ]
dge <- dge[mask, ]
dim(dge)
```

```{r}
htseq_files <- iterate_files(dir_hisat_zyagen, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")
dds <- create_dds(htseq_files,dir_hisat_zyagen)

sampleFiles = htseq_files
directory = "/home/luisas/Desktop/cluster/proj/data/02_RNA-Seq/03_hisat/"
samples_info <- sapply(sampleFiles,extract_info_sample_from_filename)
  # Create ddsHTSeq Object 
sampleTable <- data.frame(sampleName = sampleFiles,
                            fileName = sapply(sampleFiles,get_filename_and_dirs),
                            condition = samples_info[1,],
                            tissue = samples_info[2,], 
                            dpo = samples_info[3,], 
                            id = samples_info[4,],
                            complete_id = samples_info[5,]
  )

dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                    directory = directory,
                                    design= ~ 1)
diff_ids <- setdiff(rownames(dds), stringtie_gtf$gene_id )


dge <- DGEList(counts = assays(dds)$counts, group = dds$tissue, genes = as.data.frame(mcols(dds)))
dge <- calcNormFactors(dge)
assays(dds)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(dds)$logCPM) > 1
length(mask[mask == FALSE])
dds <- dds[mask, ]
dge <- dge[mask, ]
dim(dge)

diff_ids_2 <- setdiff(rownames(dds), stringtie_gtf$gene_id )
length(diff_ids)
length(diff_ids_2)

```



# --------------------------------------------------------------------------------

```{r}
grid::grid.newpage()
missed_because_of_filtering_ranges <- lncRNAs_ref[lncRNAs_ref$gene_id %in% missed_because_of_filtering]
length(unique(missed_because_of_filtering_ranges$gene_id))
missed_because_of_filtering_ranges <- missed_because_of_filtering_ranges[missed_because_of_filtering_ranges$type == "exon",]
# Extract 
width(ranges(missed_because_of_filtering_ranges))
df <- data.frame("gene_id" = missed_because_of_filtering_ranges$gene_id, "range_width" = width(ranges(missed_because_of_filtering_ranges)))
gene_lengths <- df %>% group_by(gene_id) %>% summarize(sum(range_width))
# Quick testing
gene_lengths[gene_lengths$gene_id == "ENSMMUG00000037481",]
df[df$gene_lengths == "ENSMMUG00000037481",]
length(intersect(missed_because_of_filtering_ranges$gene_id, gene_lengths$gene_id))

missed_because_of_filtering_ranges
ggplot(gene_lengths, aes(x=`sum(range_width)`)) +
  geom_histogram(fill="blue")+ 
  labs(title = "Distribution of gene lenghts misclassified because of filtering step ")


```




In terms of percentage how many genes we can retrieve
```{r}
grid::grid.newpage()
intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out$gene_id)
intersect_reflnrna_stringtie <- intersect(lncRNAs_ref$gene_id, stringtie_gtf$gene_id)
intersect_feelnc_stringtie <- intersect(lncRNAs_out$gene_id, stringtie_gtf$gene_id)
combo <- c(feelnc = length(unique(lncRNAs_out$gene_id)), ref_lnc = length(unique(lncRNAs_ref$gene_id)), stringtie = length(unique(stringtie_gtf$gene_id)), "feelnc&ref_lnc" = length(intersect_reflnrna_out), "feelnc&stringtie" = length(intersect_feelnc_stringtie), "ref_lnc&stringtie" = length(intersect_reflnrna_stringtie))

plot(euler(combo), quantities = TRUE)
```




Subselect only transcript ids
```{r }
grid::grid.newpage()
transcripts_ref <- lncRNAs_ref[!is.na(lncRNAs_ref$transcript_id),]
transcripts_out <- lncRNAs_out[!is.na(lncRNAs_out$transcript_id),]
transcripts_mRNAs <- mRNAs_ref[!is.na(mRNAs_ref$transcript_id),]
transcripts_stringtie <- stringtie_gtf[!is.na(stringtie_gtf$transcript_id),]
vd_draw(list(transcripts_ref$transcript_id, transcripts_out$transcript_id, transcripts_stringtie$transcript_id),c("ref","feelnc","stringtie"))
```

But how many in the newly detected one are MRG

```{r }
grid::grid.newpage()
transcripts_mRNAs <- mRNAs_ref[!is.na(mRNAs_ref$transcript_id),]
stringtie_mrg <- transcripts_stringtie[sapply(transcripts_stringtie$transcript_id, function(str) str_detect(str,"MSTRG")),]
stringtie_not_mrg <- transcripts_stringtie[sapply(transcripts_stringtie$transcript_id, function(str) !str_detect(str,"MSTRG")),]
vd_draw(list(transcripts_ref$transcript_id, transcripts_out$transcript_id, stringtie_mrg$transcript_id, stringtie_not_mrg$transcript_id, transcripts_mRNAs$transcript_id),c("ref","feelnc","stringtie_mrg", "stringtie_not_mrg","mrnasref"))
```
```{r }
grid::grid.newpage()

vd_draw(list(transcripts_ref$transcript_id, transcripts_out$transcript_id, stringtie_mrg$transcript_id, stringtie_not_mrg$transcript_id),c("ref","feelnc","stringtie_mrg", "stringtie_not_mrg"))
```



We are interested in the distribution of the numbers of exons
```{r }
grid::grid.newpage()
table(as.numeric(transcripts_out$exon_number))
```


