---
title: "OrthologAnalysis"
author: "Luisa Santus"
date: "2/26/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(tidyverse)
library(dplyr)
library(stringr)
library(scales)
```


```{r cars}

folder = "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs"
folder_lncrna = file.path(folder, "00_lncrna_05/")
filename = "rheMac10_EBOV_and_novel_genenames_exonid"
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))

ref[substr(ref$gene_id, 1,3 ) == "rib",]$exon_id <- paste(ref[substr(ref$gene_id, 1,3 ) == "rib",]$transcript_id, ref[substr(ref$gene_id, 1,3 ) == "rib",]$exon_number, sep = ".")
#export(ref,file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_exonid.gtf") )


# Try to add transcript and gene id 


get_entry_gene <- function(ref, gene){
  seqname <- as.character(seqnames(ref[ref$gene_id == gene, ]))[1]
  range <- ranges(range(ref[ref$gene_id == gene, ]))
  score <- as.character((ref[ref$gene_id == gene, ])$score)[1]
  source <- as.character((ref[ref$gene_id == gene, ])$source)[1]
  phase <- as.character((ref[ref$gene_id == gene, ])$phase)[1]
  gene_id <- as.character((ref[ref$gene_id == gene, ])$gene_id)[1]
  gene_name <- as.character((ref[ref$gene_id == gene, ])$gene_name)[1]
  
  
  newentry_gene <- GRanges(
      seqnames = seqname,
      ranges = range,
      source = source, 
      type = "gene", 
      score = score,
      phase = phase,
      gene_id = gene_id,
      gene_name = gene_name, 
      strand =strand(range(ref[ref$gene_id == gene, ]))
      )
  
  return(newentry_gene)
}

get_entry_transcript <- function(ref, transcript){
  seqname <- as.character(seqnames(ref[!is.na(ref$transcript_id) & ref$transcript_id == transcript, ]))[1]
  range <- ranges(range(ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ]))
  score <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$score)[1]
  source <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$source)[1]
  phase <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$phase)[1]
  gene_id <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$gene_id)[1]
  gene_name <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$gene_name)[1]
  transcript_id <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$transcript_id)[1]
  transcript_name <- as.character((ref[!is.na(ref$transcript_id) &ref$transcript_id == transcript, ])$transcript_name)[1]  
  
  newentry_transcript <- GRanges(
      seqnames = seqname,
      ranges = range,
      source = source, 
      type = "transcript", 
      score = score,
      phase = phase,
      gene_id = gene_id,
      gene_name = gene_name, 
      transcript_id = transcript_id, 
      transcipt_name = transcript_name,
      strand =strand(range(ref[ref$gene_id == gene, ]))
      )
  
  return(newentry_transcript)
}



genes_missing <- ref[substr(ref$gene_id, 1,3 ) == "rib",]$gene_id
transcripts_missing <- ref[substr(ref$gene_id, 1,3 ) == "rib",]$transcript_id


gene_missing_gr <-lapply(genes_missing, function(x) get_entry_gene(ref,x))
transcript_missing_gr <-lapply(transcripts_missing, function(x) get_entry_transcript(ref,x))
gene_missing_gr <- do.call("c",gene_missing_gr)
transcript_missing_gr <- do.call("c", transcript_missing_gr)

full <- append(append(ref, transcript_missing_gr), gene_missing_gr) 


full <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_full.gtf"))


full <- sortSeqlevels(full)
full_sorted <- sort(full)

export(full_sorted, file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_full_sorted.gtf"))

full





```





```{r cars}
# ---------------------------------
#      Analsysis Orthologs
# ---------------------------------

folder_input <- folder_lncrna
intersection_table <- read.table(paste(file.path(folder_input, "intersection.tab")))

# Rename columns so that we can assign them correctly 
bed_macaque <- read.table(paste(file.path(folder_input, filename), "sorted.bed6", sep = "."))
bed_human <- read.table(paste(file.path(folder_input, filename), "human.sorted.bed6", sep = "."))
bed_human_macaque <- read.table(paste(file.path(folder_input, filename), "human.macaque.sorted.bed6", sep = "."))
names(bed_macaque) <- c("chr", "start", "end", "complete_id", "score", "strand")
names(bed_human) <- c("chr", "start", "end", "complete_id", "score", "strand")
names(bed_human_macaque) <- c("chr", "start", "end", "complete_id", "score", "strand")
  
names(intersection_table) <- c("chrA", "startA", "endA", "complete_idA", "scoreA", "strandA", "chrB", "startB", "endB", "complete_idB", "scoreB", "strandB", "overlap")
intersection_table$complete_idA <- as.character(intersection_table$complete_idA)
intersection_table$complete_idB <- as.character(intersection_table$complete_idB)
  
# Check how many have the same gene overlap or exon overlap
exon_overlap <- intersection_table %>% dplyr::filter(complete_idA == complete_idB)
number_exons_with_exact_overlap <- length(unique(exon_overlap$complete_idB))



# Put a threshold 
exon_overlap$macaque_gene <- unlist(lapply(exon_overlap$complete_idA, function(x) strsplit(x,"_")[[1]][1]))
exon_overlap$macaque_exon <- unlist(lapply(exon_overlap$complete_idA, function(x) strsplit(x,"_")[[1]][2]))


# Each gene which ortholog exons it has 
orth_exons_per_gene <- distinct(exon_overlap) %>% group_by(macaque_gene) %>% dplyr::summarise(exon = paste(macaque_exon, collapse=", "))

exon_overlap[exon_overlap$macaque_gene =="ENSMMUG00000000258", ]
orth_exons_per_gene[substr(orth_exons_per_gene$macaque_gene, 1,3 ) == "rib",]

gene <- "ribodepl-MSTRG.102407"

calc_overlap_gene_exonorthologs <- function(gene){
  gene_length_spliced <- sum(width(ranges(ref[ref$gene_id == gene & ref$type == "exon", ])))
  sum_exons_orthologs <- sum(width(IRanges::reduce(ranges(ref[ref$exon_id %in% unlist(str_split(orth_exons_per_gene[orth_exons_per_gene$macaque_gene == gene,]$exon, ", ")), ]))))
  orth_exons_overlap_gene <- sum_exons_orthologs*100/gene_length_spliced
  return(orth_exons_overlap_gene)
}


# Total number of exons tested 
total_exons_beginning <- length(unique(bed_macaque$complete_id))
total_exons_after_first_liftover <- length(unique(bed_human$complete_id))
total_exons_after_second_liftover <- length(unique(bed_human_macaque$complete_id))
  
# ORTHOLOGS BIOTYPES 
intersection_table_biotype <- read.table(paste(file.path(folder_input, "intersection_human_macaquehuman.tab")))
names(intersection_table_biotype) <- c("chrA", "startA", "endA", "complete_idA", "scoreA", "strandA", "biotype", "chrB", "startB", "endB", "complete_idB", "scoreB", "strandB", "overlap")
  
# Get the biotypes for the overlapped exons( the ones matching after both liftovers)
full_correspondence <- intersection_table_biotype %>% right_join(exon_overlap, by=c("complete_idB"))



# Remove lines where there is no match 
full_correspondence <- full_correspondence[!is.na(full_correspondence$complete_idA.x), ]
full_correspondence_reduced <- data.frame(human_id_full = as.character(full_correspondence$complete_idA.x), 
                                             macaque_id_full = full_correspondence$complete_idB, 
                                             biotype =full_correspondence$biotype, stringsAsFactors = FALSE)
full_correspondence_reduced$human_gene_id <-  unlist(lapply(full_correspondence_reduced$human_id_full, function(x) strsplit(x,"_")[[1]][1]))
full_correspondence_reduced$macaque_gene_id <-  unlist(lapply(full_correspondence_reduced$macaque_id_full, function(x) strsplit(x,"_")[[1]][1]))

full_correspondence_reduced <- distinct(full_correspondence_reduced) %>% group_by(macaque_id_full, human_id_full) %>% dplyr::summarise(biotype = paste(biotype, collapse=", "))


full_correspondence_reduced$macaque_gene <- unlist(lapply(as.character(full_correspondence_reduced$macaque_id_full), function(x) strsplit(x,"_")[[1]][1]))
full_correspondence_reduced <- distinct(full_correspondence_reduced) %>% group_by(macaque_gene) %>% dplyr::summarise(biotype = paste(biotype, collapse=", "))
full_correspondence_reduced$biotype_summarized <- unlist(lapply(full_correspondence_reduced$biotype, function(x) paste(unique(trimws(unlist(str_split(x, ",")))), collapse = ",")))


full_correspondence_reduced$perc_gene_orth_exons <-unlist(lapply(full_correspondence_reduced$macaque_gene,calc_overlap_gene_exonorthologs))

full_correspondence_reduced_filter <- full_correspondence_reduced[full_correspondence_reduced$perc_gene_orth_exons > 50, ]
novel_orth_macaque <- full_correspondence_reduced[substr(full_correspondence_reduced$macaque_gene, 1,3 ) == "rib",]

table(full_correspondence_reduced$perc_gene_orth_exons > 50 ) 

# Orthologs for how many of my lnc do i find? 
novel_orth_macaque <- full_correspondence_reduced_filter[substr(full_correspondence_reduced_filter$macaque_gene, 1,3 ) == "rib",]
annot_orth_macaque <- full_correspondence_reduced_filter[substr(full_correspondence_reduced_filter$macaque_gene, 1,3 ) != "rib",]


ggplot(annot_orth_macaque, aes(y = biotype_summarized, fill = biotype_summarized))+geom_bar()+theme_classic()+theme(legend.position = "none")+labs(y = "Biotype")

ggplot(novel_orth_macaque, aes(y = biotype_summarized, fill = biotype_summarized))+geom_bar()+theme_classic()+theme(legend.position = "none")+labs(y = "Biotype")


```


````{r bio}
# I find only 900
correspondence_filtered$type <- unlist(lapply(correspondence_filtered$macaque_id, function(x) substr(x,1,4)))
length(unique(correspondence_filtered$macaque_id))
ggplot(correspondence_filtered, aes(y = biotype_summarized, fill = type))+geom_bar()+theme_classic()+scale_fill_brewer(palette = "Paired")

count_biotype <- as.data.frame(table(correspondence_filtered$biotype_summarized))
count_biotype <- count_biotype[count_biotype$Freq > 10, ] 
biotypes_keep <- as.vector(count_biotype$Var1)

ggplot(correspondence_filtered[correspondence_filtered$biotype_summarized %in% biotypes_keep, ], aes(y = biotype_summarized, fill = type))+geom_bar()+theme_classic()+scale_fill_brewer(palette = "Paired")+labs(y = "Biotype")

```

```{r de}

full_correspondence[full_correspondence$complete_idA.x %in% de, ]

# Check overlap with pan cancer lncRNA analysis
sign_pathway <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs/Lnc_Pathways_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
sign_cell <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs/Lnc_Immunecell_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
sign_cancer <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs/LncRNA_Cancer_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)



de_ortholgs_immlnc <- sign_pathway[sign_pathway$lncRNA_id %in% unlist(lapply(de_full_corr$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])),]
de_ortholgs_cancer <- sign_cancer[sign_cancer$lncRNA_id %in% unlist(lapply(de_full_corr$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])),]
de_ortholgs_cell <- sign_cell[sign_cell$lncRNA_id %in% unlist(lapply(de_full_corr$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])),]


de_ortholgs_immlnc_reduced <- data.frame(id = de_ortholgs_immlnc$lncRNA_id, 
                                         pathway = de_ortholgs_immlnc$immune_pathway, 
                                         pval = de_ortholgs_immlnc$p_adjust, 
                                         cancer  = de_ortholgs_immlnc$cancer_type
                                         )


imm_lnc_matches_distr <- de_lnc_orthologs[unlist(lapply(de_lnc_orthologs$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])) %in% de_ortholgs_immlnc_reduced$id,]

theme_paper <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())


ggplot(distinct(de_ortholgs_immlnc_reduced), aes(y = pathway, fill = id, color = cancer ))+geom_bar()+theme_classic()+labs(fill = "Gene ID")+scale_fill_brewer(palette = "Accent")+scale_color_manual(values = rep("grey", length(de_ortholgs_immlnc_reduced$cancer)))+guides( color = FALSE)+theme_paper


full_correspondence_orthologs <- full_correspondence_reduced[unlist(lapply(full_correspondence_reduced$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])) %in% de_ortholgs_immlnc_reduced$id, ]

my_de_ortholog <- unique(full_correspondence_orthologs$macaque_gene_id[as.character(full_correspondence_orthologs$macaque_gene_id) %in% gsub("-unkown", "", unlist(de_lists_lnc_my))])

length(my_de_ortholog)

T_de_ortholog <- unique(full_correspondence_orthologs$macaque_gene_id[as.character(full_correspondence_orthologs$macaque_gene_id) %in% gsub("-unkown", "", unlist(de_lists_lnc_T))])

length(T_de_ortholog)

B_de_ortholog <- unique(full_correspondence_orthologs$macaque_gene_id[as.character(full_correspondence_orthologs$macaque_gene_id) %in% gsub("-unkown", "", unlist(de_lists_lnc_B))])
length(B_de_ortholog)

full_correspondence_orthologs$human_id_simple <- unlist(lapply(full_correspondence_orthologs$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1]))


de_ortholgs_immlnc_reduced$human_id_simple <- as.character(de_ortholgs_immlnc_reduced$id)
full_correspondence_orthologs$human_id_simple

length(unique(de_ortholgs_immlnc_reduced$id))

full_for_join <- data.frame( human_id_simple = full_correspondence_orthologs$human_id_simple, 
                             macaque_id = full_correspondence_orthologs$macaque_gene_id, stringsAsFactors = FALSE ) 

de_full_corr <- de_ortholgs_immlnc_reduced %>% join(full_for_join, by=c("human_id_simple"))
length(unique(de_full_corr$macaque_gene_id))

unique(de_ortholgs_immlnc_reduced[de_ortholgs_immlnc_reduced$pathway == "Cytokines",]$id)
unique(de_full_corr[de_full_corr$pathway == "Cytokines",]$id)

nrow(de_full_corr)
de_full_corr$orth_id <- paste(de_full_corr$macaque_id, "-> ", de_full_corr$human_id_simple)

ggplot(distinct(de_full_corr), aes(y = pathway, fill = macaque_id, color = cancer ))+geom_bar()+theme_classic()+labs(fill = "Macaque Gene ID", y = "Pathway")+scale_fill_brewer(palette = "Set3")+scale_color_manual(values = rep("darkgrey", length(de_ortholgs_immlnc_reduced$cancer)))+guides( color = FALSE)+theme_paper


intersect(B_de_ortholog, T_de_ortholog)
intersect(my_de_ortholog, T_de_ortholog)
intersect(B_de_ortholog, my_de_ortholog)


myeloids_de[myeloids_de$primerid %in% my_de_ortholog, ]

# test ---- end

ggplot(correspondence_filtered[correspondence_filtered$biotype_summarized %in% biotypes_keep, ], aes(y = biotype_summarized, fill = type))+geom_bar()+theme_classic()+scale_fill_brewer(palette = "Paired")+theme(legend.position = "none")+labs(y = "Biotype")+theme(axis.text = element_text(size = 17))

ggplot(correspondence_filtered, aes(y = biotype_summarized, fill = type))+geom_bar()+theme_classic()+scale_fill_brewer(palette = "Paired")+theme(legend.position = "none")+labs(y = "Biotype")
ggplot(de_lnc_orthologs, aes(y = biotype_summarized, fill = type ))+geom_bar()+theme_classic()+scale_fill_brewer(palette = "Paired")+labs(y = "Biotype")+theme(axis.text = element_text(size = 17))


# How many orthlogs i find 
length(unique(correspondence_filtered$macaque_id))

# How many are novel and how many annotated
table_number_orthologs <-table(substr(unique(correspondence_filtered$macaque_id),1,3))
table_number_orthologs*100/sum(table_number_orthologs)


table(correspondence_filtered$biotype_summarized)*100/sum(table(correspondence_filtered$biotype_summarized))



```

```{r rest }

# Orthologs Stats 
# Out of all the lncRNAs for which we tried to get the ortholgs this is what we find. 

# In general stats about lnc we have in reference
ref[substr(ref$gene_id, 1,3) %in% c("pol", "rib"), ]$gene_biotype <- "lncRNA"
table(ref[substr(ref$gene_id, 1,3) %in% c("pol", "rib"), ]$gene_biotype)
table(ref$gene_biotype)
lnc<- ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "lncRNA", ]
length(unique(gtf[gtf$gene_biotype == "lncRNA", ]$gene_id)) # 7k
lnc_distr <- as.data.frame(table(unlist(lapply(unique(lnc$gene_id), function(x) substr(x,1,4)))))
names(lnc_distr) <- c("type", "count")
ggplot(lnc_distr, aes(x = "", y = count, fill = type))+geom_bar(stat =  "identity")+coord_polar("y", start = 0)+
  theme_minimal()+
  scale_fill_brewer(palette = "Paired")+
  geom_text(label = lnc_distr$count, size = 5, aes(y = c(6000,900,2100)))+
  theme(axis.text.x = element_blank())



lnc_orthologs <- full_correspondence_reduced[!is.na(full_correspondence_reduced$biotype), ]
lnc_orthologs_unique <- lnc_orthologs[!duplicated(lnc_orthologs$macaque_gene_id),]
# General plot
ggplot(lnc_orthologs_unique, aes(y = biotype, fill = biotype))+geom_bar()+theme_classic()+theme(legend.position = "none")

# I find only 900
lnc_orthologs_unique$type <- unlist(lapply(lnc_orthologs_unique$macaque_gene_id, function(x) substr(x,1,4)))
length(unique(lnc_orthologs_unique$macaque_gene_id))
ggplot(lnc_orthologs_unique, aes(y = biotype, fill = type))+geom_bar()+theme_classic()+scale_fill_brewer(palette = "Paired")

# How many of the De genes 
de_lnc_orthologs <- lnc_orthologs_unique[lnc_orthologs_unique$macaque_gene_id %in% de, ]
ggplot(de_lnc_orthologs, aes(y = biotype, fill = type ))+geom_bar()+theme_minimal()+scale_fill_brewer(palette = "Paired")


# Check overlap with pan cancer lncRNA analysis
sign_pathway <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs/Lnc_Pathways_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
sign_cell <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs/Lnc_Immunecell_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
sign_cancer <- read.table("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/04_orthologs/LncRNA_Cancer_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)

de_ortholgs_immlnc <- sign_pathway[sign_pathway$lncRNA_id %in% unlist(lapply(de_lnc_orthologs$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])),]
de_ortholgs_cancer <- sign_cancer[sign_cancer$lncRNA_id %in% unlist(lapply(de_lnc_orthologs$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])),]
de_ortholgs_cell <- sign_cell[sign_cell$lncRNA_id %in% unlist(lapply(de_lnc_orthologs$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])),]


de_ortholgs_immlnc_reduced <- data.frame(id = de_ortholgs_immlnc$lncRNA_id, 
                                         pathway = de_ortholgs_immlnc$immune_pathway
                                         )


imm_lnc_matches_distr <- de_lnc_orthologs[unlist(lapply(de_lnc_orthologs$human_gene_id, function(x) strsplit(x,"[.]")[[1]][1])) %in% de_ortholgs_immlnc_reduced$id,]



de_ortholgs_immlnc_reduced

ggplot(distinct(de_ortholgs_immlnc_reduced), aes(y = pathway, fill = id))+geom_bar()+theme_minimal()


distinct(de_ortholgs_immlnc_reduced) %>% group_by(id) %>% dplyr::summarise(pathway = paste(pathway, collapse=", "))

```
