---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# 0. Load and Explore reference 

```{r Imports}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)

# Gene annotation
ref <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_addedgenesNovel.gtf")

table(ref$type)
```

# 1. Extract promoter regions
```{r Imports}

# -----------------------------------------------------------------------------
#  Extract promoter Regions (80 bp upstream to 34 bp downstream of the TSS)
# -----------------------------------------------------------------------------

# 0. Only use gene entries 
genes <- ref[ref$type == "gene",]

# 1. Select strand specific promoter regions ( gene level )
# Most upsream TSS selected 
get_bedEntry <- function(gene_id, ref){
  gene <- ref[ref$gene_id == gene_id, ]
  chr <- as.character(seqnames(gene))
  strand <- as.character(strand(gene))
  
  if(strand == "+"){
    tss <- start(gene)
    start = tss - 80
    end = tss + 34
  }else{
    tss <- end(gene)
    end = tss + 80
    start = tss - 34
  }
  
  # Create Bed entry 
  entry <- data.frame(chr=chr, start = start, end = end, gene_id, 1, strand)
  return(entry)
}

bed <- Reduce("rbind",lapply( unique(genes$gene_id), function(x) get_bedEntry(x, genes)))
bed[bed$start < 0,]$start <- 0
write.table(bed,"/home/luisas/Desktop/cluster/data/03_tfbs/00_promoters.bed",sep = "\t", quote = F, col.names = F, row.names = F )

```





