---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# 0. Load and Explore reference 

```{r Imports}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(GenomicRanges)
FDRTHRESHOLD <- 0.05

ref<- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_addedgenesNovel.gtf"))

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "black"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 14))

# Promoter regions per gene
promoters <- read.table("/home/luisas/Desktop/cluster/data/03_tfbs/00_promoters.bed")
#promoters$start <- as.integer(promoters$start)
#write.table(promoters, "/home/luisas/Desktop/cluster/data/03_tfbs/00_promoters.bed", quote= F, col.names = F, row.names = F, sep = "\t")

#promoters <- promoters[,c(1,2,3,4,6)]
colnames(promoters) <- c("chr", "start", "end", "gene_id", "strand")

# Motifs identified 
fimo <- read.table("/home/luisas/Desktop/cluster/data/03_tfbs/fimo_final.tsv", header = T )
fimo <- fimo[fimo$q.value < FDRTHRESHOLD, ]
fimo$motif_bp <- nchar(as.character(fimo$matched_sequence))
colnames(fimo) <- gsub("sequence_name","gene_id", colnames(fimo))
fimo$gene_id <- as.character(fimo$gene_id)
fimo$motif_id <- paste(fimo$motif_id, fimo$strand, sep = "")
fimo_red <- unique(fimo[,c("gene_id", "motif_id", "motif_bp")])
promoter_vs_motif <- acast(fimo_red, gene_id ~ motif_id, fun.aggregate = length )


# Load gene information INVIVO
robjectsdir<- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
df_lnc <- readRDS(file.path(robjectsdir, "df_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))
df_lnc <- df_lnc[df_lnc$n_cells > 10, ]
df_mrna <- df_mrna[df_mrna$n_cells > 10, ]
# Only select genes expressed in at least 50 cells
df_lnc$type <- "lncRNA"
df_mrna$type <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)
colnames(df_complete) <- gsub("gene_id","gene_name", colnames(df_complete))

correspondence <- unique(data.frame(ref)[,c("gene_id", "gene_name")])

#gene_versions <- correspondence %>% dplyr::group_by(gene_name) %>% dplyr::summarise(max_gv = max(gene_version))
#correspondence <- merge(correspondence, gene_versions, by = "gene_name")
#correspondence <- correspondence[correspondence$gene_version == correspondence$max_gv,]

# Remove (for now) genes w/ multiple gene_ids
multiple_geneid <- correspondence %>% dplyr::group_by(correspondence$gene_name) %>% tally() 
multiple_geneid <- multiple_geneid[multiple_geneid$n>1, ]$`correspondence$gene_name`
correspondence <- correspondence[!(correspondence$gene_name %in% multiple_geneid),]
rownames(correspondence) <- correspondence$gene_name

# ADD gene id
df_complete <-merge(df_complete, correspondence, by = "gene_name")

n_lnc <- nrow(df_complete[df_complete$type == "lncRNA",])
n_pc <- nrow(df_complete[df_complete$type == "Protein Coding",])

df_complete$type_legend <- df_complete$type
df_complete[df_complete$type == "lncRNA", ]$type_legend <- paste("lncRNAs (n= ", n_lnc, ")",sep = "")
df_complete[df_complete$type == "Protein Coding", ]$type_legend <- paste("mRNAs (n= ", n_pc, ")",sep = "")

# Add subtypes annotation 
df_complete$annotation <- ifelse(substr(df_complete$gene_id,1,3) == "MST", "Novel", "Annotated")
df_complete$subtype <- df_complete$type
df_complete[df_complete$type == "lncRNA",]$subtype <- paste(df_complete[df_complete$type == "lncRNA",]$type, df_complete[df_complete$type == "lncRNA",]$annotation, sep = " ")

n_lncannot <- nrow(df_complete[df_complete$subtype == "lncRNA Annotated",])
n_lncnovel <- nrow(df_complete[df_complete$subtype == "lncRNA Novel",])
n_pc <- nrow(df_complete[df_complete$subtype == "Protein Coding",])

df_complete$subtype_legend <- df_complete$subtype
df_complete[df_complete$subtype == "lncRNA Annotated", ]$subtype_legend <- paste("lncRNAs Annotated (n= ", n_lncannot, ")",sep = "")
df_complete[df_complete$subtype == "lncRNA Novel", ]$subtype_legend <- paste("lncRNAs Novel (n= ", n_lncnovel, ")",sep = "")
df_complete[df_complete$subtype == "Protein Coding", ]$subtype_legend <- paste("mRNAs (n= ", n_pc, ")",sep = "")

```

# add real start 
```{r merge}
start_pos <- unique(data.frame(gene_id= ref[ref$type == "gene",]$gene_id, tss=start(ref[ref$type == "gene",])))
start_pos$gene_id <- as.character(start_pos$gene_id)
rownames(start_pos) <- start_pos$gene_id

fimo_updated <- merge(fimo, start_pos, by = "gene_id")

# -------------------
rownames(promoters) <- promoters$gene_id
colnames(promoters) <- c("chr", "start_promoter", "stop_promoter", "gene_id", "strand")
prom_red <- promoters[,c(1,2,3,4)]

fimo_updated_prom <- merge(fimo_updated, prom_red, by = "gene_id")
fimo_updated_prom$start_updated <- fimo_updated_prom$start + fimo_updated_prom$start_promoter
fimo_updated_prom$stop_updated <- fimo_updated_prom$stop + fimo_updated_prom$start_promoter

fimo_red <- unique(fimo_updated_prom[,c("gene_id", "motif_id", "motif_bp")])


saveRDS(fimo_updated_prom, "/home/luisas/Desktop/cluster/data/03_tfbs/fimo.rds")
#promoter_vs_motif <- acast(fimo_updated_prom, gene_id ~ motif_id, fun.aggregate = length )



# Prepare bed file
chr_cor <- unique(data.frame(chr = seqnames(ref), gene_id = as.character(ref$gene_id)))
chr_cor$gene_id <- as.character(chr_cor$gene_id)
fimo_chr <- merge(fimo, chr_cor, by = "gene_id")
motifs_bed <- fimo_chr[,c("chr", "start", "stop", "motif_id","score", "strand")]
write.table(motifs_bed, "/home/luisas/Desktop/cluster/data/03_tfbs/motifs.bed", quote= F, col.names = F, row.names = F, sep = "\t")
```




```{r merge}
n_motif_per_gene <- fimo_red %>% dplyr::group_by(gene_id) %>% dplyr::count() 
colnames(n_motif_per_gene) <- c("gene_id", "n_motifs")
summary <- merge(df_complete, n_motif_per_gene, by="gene_id")
summary
```

# ------------------------------------------
# 1-  NUMBER OF IDENTIFIED MOTIFS 
# -------------------------------------------
```{r merge}
# 0. There is one outlier! 
summary <- summary[summary$n_motifs != max(summary$n_motifs),]

# 1. Plot # Motifs in promoters 
ggplot(summary, aes(x = n_motifs))+geom_histogram(binwidth = 1, fill ="grey")+theme_paper+xlab("# Motifs")+ylab("# Genes")


# 2. Number of motifs per promoter (separaetd by gene class)
palette = c("#cc0000", "#6699ff")
ggplot(summary, aes(x = n_motifs, col = type_legend)) + stat_ecdf(geom = "line",size = 1)+theme_paper+theme_paper+xlab("# motifs")+ylab("cumulative density")+scale_color_manual(values =palette)


# 3. Number of motifs per promoter (separaetd by gene class + annotation )
palette_subtypes= c("grey", "black", "#6699ff")
ggplot(summary, aes(x = n_motifs, col = subtype_legend)) + stat_ecdf(geom = "line",size = 1)+theme_paper+theme_paper+xlab("# motifs")+ylab("cumulative density")+scale_color_manual(values =palette_subtypes)
```

# ------------------------------------------
# 2- # of bp covered by motifs
# -------------------------------------------

```{r merge}
summary

fimo_updated_prom

```



```{r merge}
# 1. There is one outlier! 
create_grange <- function(gene_motifs){
  return(Reduce("c", (lapply(split(gene_motifs, gene_motifs$motif_id), function(i){ GRanges(seqnames = as.character(i$chr),
                                                           strand = i$strand,
                                                           ranges = IRanges(start = i$start_updated,
                                                           end = i$stop_updated,
                                                           names = i$motif_id))}))))
}

get_nb_motifs <- function(gene, fimo){ data.frame(gene = gene, nbp = sum(width(reduce(create_grange(fimo[fimo$gene_id == gene, ])))))}

library(parallel)
cl <- makeCluster(detectCores())
clusterExport(cl,c('gene','fimo_updated_prom','get_nb_motifs', "create_grange"))
clusterEvalQ(cl, library("GenomicRanges"))
sumary_nbp <- parLapply(cl,unique(fimo$gene_id), function(gene) get_nb_motifs(gene, fimo_updated_prom))
stopCluster(cl)
summary_nbp <- Reduce("rbind",sumary_nbp )
saveRDS(summary_nbp,"/home/luisas/Desktop/cluster/data/03_tfbs/summary_nbp.rds")

```

```{r merge}
# Check example 
ref[grepl("45507",ref$gene_id), ]
ref[grepl("NUDT9",ref$gene_name), ]
a <- "ENSMMUG00000045507"
b <- "ENSMMUG00000019043"
summary_nbp[summary_nbp$gene_id == a, ]
summary_nbp[summary_nbp$gene_id == "ENSMMUG00000019043", ]

```



```{r merge}
colnames(summary_nbp) <- c("gene_id", "nbp")
summary_complete_nbp <- merge(summary, summary_nbp, by = "gene_id")
summary_complete_nbp$log_n_bp <- log(summary_complete_nbp$nbp)

# 1. NBP per promoter (General)
ggplot(summary_complete_nbp, aes(x = nbp))+geom_density(binwidth = 1, fill ="grey", col = "white")+theme_paper+xlab("# bp covered")

# 2. NBP per promoter (separaetd by gene class)
ggplot(summary_complete_nbp, aes(x = log_n_bp, col = type_legend)) + stat_ecdf(geom = "line",size = 1)+theme_paper+theme_paper+xlab("log (# bp covered)")+ylab("cumulative density")+scale_color_manual(values =palette)+xlab("log(# bp covered)")+xlim(c(1,5))

ggplot(summary_complete_nbp, aes(x = nbp, col = subtype_legend)) + stat_ecdf(geom = "line",size = 1)+theme_paper+theme_paper+xlab("log (# bp covered)")+ylab("cumulative density")+scale_color_manual(values =palette_subtypes)+xlab("log(# bp covered)")

```


```{r merge}
ggplot(summary_complete, aes(x = medianexpr, y = perc_cells_expressing, col = subtype))+geom_point()+theme_paper+scale_color_manual(values =palette_subtypes)


ggplot(summary_complete, aes(x = medianexpr, y = n_motifs, col = subtype))+geom_point()+theme_paper+scale_color_manual(values =palette_subtypes)


ggplot(summary_complete, aes(x = perc_cells_expressing, y = n_motifs, col = subtype))+geom_point()+theme_paper+scale_color_manual(values =palette_subtypes)

ggplot(summary_complete_nbp, aes(x = nbp, y = max_expression))+geom_point()


summary_complete_nbp

```



# How many overlappin motifs 
```{r merge}

gene <- "ENSMMUG00000005271"
gene_motifs <- fimo[fimo$gene_id == gene, ]


get_max_cluster <- function(gene, fimo){
  r <- create_grange(fimo[fimo$gene_id == gene, ])
  df<- as.data.frame(findOverlaps(r))
  # Identify all clusters ( reduntant but we don't care )
  # Which is the size of the cluster to which that motif belongs to
  clusters <- df %>% dplyr::group_by(queryHits) %>%  dplyr::summarise(n = n())
  colnames(clusters) <- c("queryHits", "sizeCluster")
  return(data.frame(gene_id= gene, max_cluster =max(clusters$sizeCluster)))
}

library(parallel)
cl <- makeCluster(detectCores())
clusterExport(cl,c('gene','fimo','get_max_cluster', "create_grange"))
clusterEvalQ(cl, library("GenomicRanges"))
clusterEvalQ(cl, library("dplyr"))
sumary_mc <- parLapply(cl,unique(fimo$gene_id), function(gene) get_max_cluster(gene, fimo))
stopCluster(cl)
summary_mc <- Reduce("rbind",sumary_mc )
saveRDS(summary_mc,"/home/luisas/Desktop/cluster/data/03_tfbs/summary_maxcluster.rds")

summary_mc <- readRDS("/home/luisas/Desktop/cluster/data/03_tfbs/summary_maxcluster.rds")
```



```{r merge}
summary_complete <- merge(summary_complete_nbp,summary_mc, by = "gene_id")
summary_complete$log_mc <- log(summary_complete$max_cluster)
summary_complete <- summary_complete[summary_complete$max_cluster != max(summary_complete$max_cluster),]

# 1. NBP per promoter (General)
ggplot(summary_complete, aes(x = max_cluster, col = type_legend))+geom_histogram(binwidth = 1, fill ="grey", col = "white")+theme_paper+xlab("Max cluster")

ggplot(summary_complete, aes(x = log_mc, col = type))+geom_density(binwidth = 1)+theme_paper+xlab("Max cluster")

# 2. NBP per promoter (separaetd by gene class)
ggplot(summary_complete, aes(x = log_mc, col = type_legend)) + stat_ecdf(geom = "line",size = 1)+theme_paper+theme_paper+xlab("log (# bp covered)")+ylab("cumulative density")+scale_color_manual(values =palette)+xlab("log(max coverage)")
ggplot(summary_complete, aes(x = log_mc, col = subtype_legend)) + stat_ecdf(geom = "line",size = 1)+theme_paper+theme_paper+xlab("log (# bp covered)")+ylab("cumulative density")+scale_color_manual(values =palette_subtypes)+xlab("log(max coverage)")

```

```{r merge}
table(summary_complete$type)

```




# --------------------------------------------------------------------------------------------------------------------

```{r merge}
library(MatchIt)
set.seed(123)

df <- summary[,c("gene_id", "perc_cells_expressing", "type", "n_motifs", "nbp")]
c <- df[complete.cases(df),]
c$t = ifelse(c$type == "lncRNA", 1, 0)
table(c$type)

mi_perc <- matchit(t ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, c)

ggplot(matches_perc, aes(x = n_motifs, y = perc_cells_expressing, col = type)) + geom_point(size = 0.2)+theme_paper+theme_paper+scale_color_manual(values =palette)

library(cowplot)
theme_matching <- theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(legend.position ="",legend.title = element_blank())+theme(legend.title = element_blank())+theme(axis.text = element_text(size = 15), axis.title = element_text(size = 13), legend.text = element_text(size = 13))

plot_matching_scatter <- function(matches_perc, palette_plot_percentage = palette){
    scatter_plot_matching <- ggplot(matches_perc, aes(x=n_motifs, y =perc_cells_expressing, col = factor(type)))+
                            geom_point(size = 1, alpha = 0.7)+
                            scale_color_manual(values = palette_plot_percentage)+
                            theme_matching
  
  
  xbox <- axis_canvas(scatter_plot_matching, axis = "x", coord_flip = TRUE) + scale_x_discrete()+
    geom_boxplot(data = matches_perc, aes(x=factor(type), y =n_motifs, col = factor(type), fill = factor(type)), alpha = 0.5) + coord_flip()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)+ stat_compare_means( label = "p.signif")
  
  ybox <- axis_canvas(scatter_plot_matching, axis = "y") + 
    geom_boxplot(data = matches_perc, aes(y = perc_cells_expressing, x = factor(type), color = factor(type), fill = factor(type), alpha = 0.5))+
    scale_x_discrete()+
    scale_fill_manual(values = palette_plot_percentage)+
    scale_color_manual(values = palette_plot_percentage)
  
  p1 <- insert_xaxis_grob(scatter_plot_matching, xbox, grid::unit(1, "in"), position = "top")
  p2 <- insert_yaxis_grob(p1, ybox, grid::unit(1, "in"), position = "right")
  ggdraw(p2)
}


plot_matching_scatter(matches_perc)

??axis_canvas

```







# ------------------------------------------------------------------------




```{r Imports}
get_prop <- function(propriety, input_id){
  propriety <- as.character(propriety)
  ids <- strsplit(propriety, "=|;")[[1]][seq_len(length(strsplit(propriety, "=|;")[[1]])) %% 2 == 1]
  id_df <- data.frame(id = ids, index = seq(1, length(ids)))
  if(!grepl(input_id, propriety)){
    return(1)
  }
  return(strsplit(strsplit(propriety, ";")[[1]][id_df[id_df$id == input_id,]$index], "=")[[1]][2])
}


fimo$name <- unlist(lapply(fimo$proprieties, function(x) get_prop(x, "Name")))
fimo$pval <- as.double(unlist(lapply(fimo$proprieties, function(x) get_prop(x, "pvalue"))))
fimo$qval <- as.double(unlist(lapply(fimo$proprieties, function(x) get_prop(x, "qvalue"))))

fimo$motif <- unlist(lapply(fimo$name, function(x) str_split(x, "_")[[1]][1]))
fimo$motif <- paste(fimo$motif, fimo$strand, sep = "")
# 1 . Only consider significant motifs 
#fimo <- fimo[fimo$pval < FDRTHRESHOLD, ]
fimo
```


```{r Imports}
acast(fimo, gene ~ motif )

unique(fimo$motif)
```



```{r Imports}
# Transform fimo in genomic ranges
motifs_range <- lapply(split(fimo, fimo$name), function(i){ GRanges(seqnames = i$chr,
                                                           strand = i$strand,
                                                           ranges = IRanges(start = i$start,
                                                           end = i$end,
                                                           names = i$name))})
motifs_range <- Reduce("c", motifs_range)


# Transform promoters in genomic ranges
promoters_range <- lapply(split(promoters, promoters$gene_id), function(i){ GRanges(seqnames = i$chr,
                                                           strand = i$strand,
                                                           ranges = IRanges(start = i$start,
                                                           end = i$end,
                                                           names = i$gene_id))})
promoters_range <- Reduce("c", promoters_range)

# Save 
saveRDS(promoters_range,"/home/luisas/Desktop/cluster/data/03_tfbs/02_promoters_range.rds")
saveRDS(motifs_range,"/home/luisas/Desktop/cluster/data/03_tfbs/03_motifs_range.rds")
fimo
```



```{r Imports}
# Mock 
motifs_range <- renameSeqlevels(motifs_range,mapSeqlevels(seqlevels(motifs_range), "ENSEMBL"))

pr<- Reduce("c",(promoters_range[c(1:1000)]))
test <-  GRanges(seqnames = "1", strand = "-", ranges = IRanges(start = 204758392,
                                                           end = 204758408,
                                                           names = "ENSMMUG00000036181"))

test2 <-  GRanges(seqnames = "1", strand = "-", ranges = IRanges(start = 204758392,
                                                           end = 204758408,
                                                           names = "ENSMMUG00000031200"))

p <- Reduce("c", list(test,test2))
```


```{r Imports}
# 1. Create Table columns motifs, rows genes (1: found motif in promoter )
motifs <- unique(names(motifs_range))
genes <- unique(names(p))

# 2. Find overlaps btw promoters and motifs 
overlaps <- findOverlaps(p, motifs_range, minoverlap=0L,ignore.strand=F)
df_overlap <- data.frame(overlaps)
df_overlap$gene <- genes[df_overlap$queryHits]
df_overlap$motif <- motifs[df_overlap$subjectHits]
df_overlap


# 3. Add summary #Motifs
df_overlap <-df_overlap %>% dplyr::group_by(gene) %>% dplyr::mutate(motifs = paste0(motif, collapse = ";"))
df_overlap <- merge(df_overlap, df_overlap %>% dplyr::group_by(gene) %>% dplyr::summarise(n_motif = n()))

df_overlap[,c(1,)]

melt(promoter_vs_motif)
# ---------------   DF  ------------------------
# Gene_ID, chr, strand, promoter_region, listOfMotifsinregion (MotifsIDs )
```




