---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
library(Seurat)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(RColorBrewer)


source("../../analysis/utils/02_sc_utils.R")
theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/01_QC/immune.combined_qc.rds")

# Load annotations
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))


# Doublet detection results 
cell_mask_scrublet <- read.table("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/02_DoubletDetection/immune.combined_scrublet_mask.txt")


result_path<- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep"
```

# ---------------------------------------------------------
#    1. REMOVE DOUBLETS: Use Scrublet mask to remove doublets 
# ---------------------------------------------------------

```{r cars}
list <- as.list(as.character(cell_mask_scrublet$V1))
mask <- list == "True"
immune.combined <- immune.combined[,!mask]
```



# -------------------------------------
#     Metadata ordering 
# -------------------------------------

```{r cars}
immune.combined$orig.ident <-  gsub("_moi1e_1", "", gsub("-", "_",immune.combined$orig.ident))
immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "_")[[1]][1]))))
immune.combined$cond <- immune.combined$cond <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][2])))
immune.combined$dpi <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][3])))
immune.combined$sample <- paste0(immune.combined$individual," ", immune.combined$dpi)
table(immune.combined$cond)
```

```{r cars}
immune.combined <- NormalizeData(immune.combined)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.5)
```

# -------------------------------------
#    Visualize Data through UMAP
# -------------------------------------

```{r UMAPvisualization}

# Day post infection 
p1 <- DimPlot(immune.combined, reduction = "umap", group.by = "dpi", pt.size = 0.1, cols = brewer.pal(10, "Paired"))+theme_sc

# Individual 
p2 <- DimPlot(immune.combined, reduction = "umap", group.by = "individual")+theme_sc

# Clusters 
p3 <- DimPlot(immune.combined, reduction = "umap", label = TRUE)+theme_sc

# Sample
p5 <- DimPlot(immune.combined, reduction = "umap", group.by = "sample", pt.size = 0.00001)+theme_sc

# Plot all
p1; p2; p3; p5

table(Idents(immune.combined))

```

# -------------------------------------
#     Preprocess and identify clusters
# -------------------------------------

```{r cars}



# Broad markers
b <- c("CD79B", "MS4A1", "CD19")
plasmablast <- c("MZB1")
tcd4 <- c("CD3D","IL7R")
tcd8 <- c("GZMB", "CD3D")
t <- c(tcd4, tcd8)
nk <- c("KLRB1", "GZMB","FCGR3")
mono <- c("CFD", "CD163", "KLRB1")
cDC <- c("IRF8", "FLT3")
pDC <- c("IRF8", "FLT3")
neut <- c( "LCN2") 


marker.genes <- c(plasmablast, b,tcd4, tcd8, nk, mono, cDC, pDC, neut)
marker.genes_red <- c(t,b,nk,mono,neut)

plot_genes(immune.combined,mono,threshold= 2, title = "Monocytes marker expression", col = "purple")
plot_genes(immune.combined, nk, threshold = 2 , title = "NK markers expression", col = "purple")
plot_genes(immune.combined, b, threshold = 2 , title = "B markers expression", col = "purple")
plot_genes(immune.combined, c(tcd4, tcd8), threshold = 2 , title = "T-Cell markers expression", col = "purple")
plot_genes(immune.combined, neut, threshold = 2 , title = "Neutrophil-Cell markers expression", col = "purple")

immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "B", `2` = "NK", 
   `3` = "B", `4` = "T", `5` = "B", `6` = "T", `7` = "B", `8` = "Myeloid", `9` = "T",
    `10` = "B", `11` = "Myeloid", `12` = "Myeloid",`13` = "B", `14` = "Myeloid", `15` = "Myeloid" )

saveRDS(immune.combined, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
```


# Visualize Marker genes

```{r Find markers}
# Visualize Markers
DotPlot(immune.combined, features = unique(marker.genes_red), cols = c("blue", "red", "purple"), dot.scale = 8) + RotatedAxis()+theme(axis.text = element_text(size = 20), axis.title = element_blank())

# Plot with correct
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(4, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
```

# -----------------------------
#     Identify EbolaGenes
#-----------------------------

```{r cars}
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unknown", sep ="-")]
ebola_genes
FeaturePlot(immune.combined, features  = ebola_genes)
plot_genes(immune.combined, ebola_genes,0, col = "#DA0202", "Cells expressing Ebola RNA")


```


