---
title: "OrthologAnalysis"
author: "Luisa Santus"
date: "2/26/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(tidyverse)
library(dplyr)
library(stringr)
library(scales)

get_stats_liftover <- function(folder, filename, lim = 100 ){
  intersection_table <- read.table(paste(file.path(folder, "intersection.tab")))
  
  bed_macaque <- read.table(paste(file.path(folder, filename), "sorted.bed6", sep = "."))
  bed_human <- read.table(paste(file.path(folder, filename), "human.sorted.bed6", sep = "."))
  bed_human_macaque <- read.table(paste(file.path(folder, filename), "human.macaque.sorted.bed6", sep = "."))
  names(bed_macaque) <- c("chr", "start", "end", "complete_id", "score", "strand")
  names(bed_human) <- c("chr", "start", "end", "complete_id", "score", "strand")
  names(bed_human_macaque) <- c("chr", "start", "end", "complete_id", "score", "strand")
  
  
  names(intersection_table) <- c("chrA", "startA", "endA", "complete_idA", "scoreA", "strandA", "chrB", "startB", "endB", "complete_idB", "scoreB", "strandB", "overlap")
  intersection_table$complete_idA <- as.character(intersection_table$complete_idA)
  intersection_table$complete_idB <- as.character(intersection_table$complete_idB)
  #intersection_table$geneA <- unlist(lapply(intersection_table$complete_idA, function(x) unlist(str_split(x, "_"))[[1]]))
  #intersection_table$geneB <- unlist(lapply(intersection_table$complete_idB, function(x) unlist(str_split(x, "_"))[[1]]))
  
  # Check how many have the same gene overlap or exon overlap
  exon_overlap <- intersection_table %>% dplyr::filter(complete_idA == complete_idB)
  number_exons_with_exact_overlap <- length(unique(exon_overlap$complete_idB))
   
  # total number of exons tested 
  total_exons_beginning <- length(unique(bed_macaque$complete_id))
  total_exons_after_first_liftover <- length(unique(bed_human$complete_id))
  total_exons_after_second_liftover <- length(unique(bed_human_macaque$complete_id))
  
 
   # ORTHOLOGS BIOTYPES 
  intersection_table_biotype <- read.table(paste(file.path(folder, "intersection_human_macaquehuman.tab")))
  names(intersection_table_biotype) <- c("chrA", "startA", "endA", "complete_idA", "scoreA", "strandA", "biotype", "chrB", "startB", "endB", "complete_idB", "scoreB", "strandB", "overlap")
  
  # Get the biotypes for the overlapped exons( the ones matching after both liftovers)
  new_dataset <- intersection_table_biotype %>% right_join(exon_overlap, by=c("complete_idB"))
  correspondence <- as.data.frame(c(new_dataset$complete_idB))
  correspondence$biotype <- as.character(new_dataset$biotype)
  names(correspondence) <- c("id", "biotype")
  
  print(length(unique(correspondence_filtered$id)))
  # remove duplicate rows (distinct) and get biotype's combination occurrence
  correspondence_filtered <- distinct(correspondence) %>% group_by(id) %>% summarise(biotype = paste(biotype, collapse=", "))
  
  # Reformat data for plotting
  df <- as.data.frame(as.matrix(table(correspondence_filtered$biotype)))
  df$type <- rownames(df)
  colnames(df) <- c("count", "biotype")
 
  biotype_na <- df[df$biotype == "NA",]$count
  biotype_not_na <- sum(df[df$biotype != "NA",]$count)
  
  # plot with no restrictions
  max = NA
  lim = lim
  p2 <- ggplot(data=df[df$count > lim,][df[df$count > lim,]$biotype != "NA",], aes(x=biotype, y=count, fill=biotype)) +geom_bar(stat="identity", width=0.5) + coord_flip()+  scale_fill_brewer(palette="Paired")+
    theme_minimal() + ylim(0,max) + theme(legend.position = "none") 
  
  # plot with ylim restrictions
  max = 1000
  p3 <- ggplot(data=df[df$count > lim,][df[df$count > lim,]$biotype != "NA",], aes(x=biotype, y=count, fill=biotype)) +geom_bar(stat="identity", width=0.5) + coord_flip()+  scale_fill_brewer(palette="Paired")+
    theme_minimal() + ylim(0,max) + theme(legend.position = "none") 

  labels <- c("1_Total", "2_After1stLiftover", "3_After2ndLiftover", "4_biotype NA", "5_biotypeFound")
  df1 <- as.data.frame(labels)
  df1$count <- c(total_exons_beginning,total_exons_after_first_liftover,total_exons_after_second_liftover,biotype_na, biotype_not_na)
  p1 <- ggplot(data=df1, aes(x = labels,  y=count, fill=labels)) +
    geom_bar(stat="identity")+
    scale_fill_brewer(palette="Paired")+
    theme_minimal()+scale_y_continuous(limits=c(number_exons_with_exact_overlap -1000, total_exons_beginning),oob = rescale_none, labels = function(x) format(x, comma = TRUE))
  
  p1b<- ggplot(data=df1, aes(x = labels,  y=count, fill=labels)) +
    geom_bar(stat="identity")+
    scale_fill_brewer(palette="Paired")+
    theme_minimal()
  
  p1
  p2
  p3
  return(list(p1,p1b,p2,p3))
}
```

## Compute number of orthologs

    TODO 

# if number is ok --> check intersect with human annotation to get the biotype, also check for strandness!!!
```{r cars}

folder = "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_final_catalogue"
folder_mrna = file.path(folder, "01_orthologs_known_mrnas/")
folder_lncrna = file.path(folder, "01_orthologs_known_lncrnas/")
folder_novel_polya = file.path(folder, "03_orthologs_novel/polya")
folder_novel_ribodepl = file.path(folder, "03_orthologs_novel/ribodepl")
filename = "rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names"


mrns_plot <- get_stats_liftover(folder_mrna, filename)
lncrnas_plot <- get_stats_liftover(folder_lncrna, filename, lim = 1)
novel_polya_plot <- get_stats_liftover(folder_novel_polya, filename, lim = 1 )
novel_ribodepl_plot <- get_stats_liftover(folder_novel_ribodepl, filename, lim = 1 )

#folder <- folder_lncrna

mrns_plot
lncrnas_plot
novel_polya_plot
novel_ribodepl_plot

```
