---
title: "Untitled"
author: "Luisa Santus"
date: "2/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# After GFF compare find the novel ones 

# ------------------------------------
# Merge them together 

```{r results='hide', message=FALSE, warning=FALSE}


datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all"
# Novel from ribodepletion compared to polyA 
riboonly <- import(file.path(datadir, "03_novel_lncrnas/01_novel_expressed/01_gffcompare/gffcompare_btw_novels.gtf.annotated.gtf"))
mask_u <- riboonly$class_code == "u";  mask_u[is.na(mask_u)] <- FALSE
table(riboonly[mask_u]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- riboonly[mask_u]$transcript_id
novel <- riboonly[riboonly$transcript_id %in% novel_transcript_ids]
print("Novel from ribo depletion - gff compare")
print(length(unique(novel$transcript_id)))
print(length(unique(novel$gene_id)))


mask_u <- riboonly$class_code == "m";  mask_u[is.na(mask_u)] <- FALSE
table(riboonly$class_code)


# ---------------
#       Save
# ---------------
export(novel, file.path(datadir, "03_novel_lncrnas/01_novel_expressed/novel_expressed_ribodepl_newtopolyA.gtf"))

# -----------------
# Stats 
plot_stats_annotation(novel,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref )
plot_stats_annotation(remove_one_exon(novel),remove_one_exon(lncRNAs_ref),remove_one_exon(lncRNAs_ref_human), remove_one_exon(mrna_ref_human), remove_one_exon(mRNAs_ref))


# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))

#dir_counts_ref = file.path(datadir, "02_RNA-Seq_BatchZyagen/06_counts_all")
dir_counts_ref = file.path(datadir, "04_Quantification_correct/")
#dir_counts_ref = file.path(datadir, "02_RNA-Seq_BatchZyagen/06_counts_all")
htseq_files_ref <- iterate_files(dir_counts_ref, ".gene_counts.tab")
batch <- htseq_files_ref[str_detect(htseq_files_ref, "Batch01")]
zyagen <- htseq_files_ref[str_detect(htseq_files_ref, "Zyagen")]
htseq_files_ref <- c(batch,zyagen)
# Expression Levels ( filter and plot )
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
# Calc logcpm 
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
# Create dataframe w/ logcpms
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
## ADD only if we want to select a certain tissue
#expression_logcpm <- expression_logcpm[,str_detect(colnames(expression_logcpm), "Liver")]
# select only when > than 1 at least 3 tissues
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
expression_logcpm <- expression_logcpm[mask,]

# Plotting
max_expression <- data.frame(id=rownames(expression_logcpm), expr=rowMax(expression_logcpm))

lncnras_expressed_ribodepl <- unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id)
mrnas_expressed_ribodepl <- unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id)

max_expression['type'] <- "0"
max_expression <- add_type(max_expression, novel$gene_id, "novel")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "lncrna")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mrna")
max_expression <- max_expression[!max_expression$type == "0",]

length(unique(novel_concordant[novel$gene_id %in% max_expression$id]$gene_id))
length(unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id))
length(unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id))

plot_expression(max_expression)

```