---
title: "lncAnnotation performance Assessment"
author: "Luisa Santus"
date: "10/22/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data and libraries

```{r echo=FALSE}
library(rtracklayer)
library(VennDiagram)
library(stringr)
library(ggplot2)
library(grid)
library(gridExtra)
library(UpSetR)
library(ComplexHeatmap)
library(GenomicRanges)
library(dplyr)
library(zeallot)
library(DESeq2)
library(edgeR)
library("ggplotify")
library("grid")
library(tidyverse)
library(tximport)
library(tximportData)

baseDir = "/home/luisas/Desktop/cluster/proj"
scripts_dir = file.path(baseDir,"code/ebola/src/scripts/")
source(paste0(scripts_dir,"de_utils.R"))

dir<- "/home/luisas/Desktop/cluster/proj/data"
dir_hisat_zyagen <- file.path(dir,"02_RNA-Seq/03_hisat/Zyagen")

lncRNAs_ref <- import(file.path(dir, "01_PreliminaryFiles/gene_annotations/rheMac8_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(dir, "01_PreliminaryFiles/gene_annotations/rheMac8_EBOV-Kikwit_mrnas.gtf"))
lncRNAs_out <- import(file.path(dir, "02_RNA-Seq/05_lncrnaAnnotation/feelnc_gencode_lncrna/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
candidates_lncRNA <- import(file.path(dir, "02_RNA-Seq/05_lncrnaAnnotation/feelnc_gencode_lncrna/candidate_lncRNA.gtf"))
mRNAs_out <- import(file.path(dir, "02_RNA-Seq/05_lncrnaAnnotation/feelnc_gencode_lncrna/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
stringtie_gtf <- import(file.path(dir, "02_RNA-Seq/04_stringtie/Zyagen/Zyagen_stringtie_merged_reference_guided.gtf"))
```
We first Check if we have multiple isoforms per gene, we do we want to collapse them.
```{r duplicates}
df <- data.frame("gene" =  lncRNAs_out$gene_id, "transcript" = lncRNAs_out$transcript_id, stringsAsFactors = FALSE)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
df[df$gene == "ENSMMUG00000037508",]
# same for reference 
df <- data.frame("gene" =  lncRNAs_ref$gene_id, "transcript" = lncRNAs_ref$transcript_id)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
df[df$gene == "ENSMMUG00000000801",]
```

do we have any lncrnas which have gene id as merged, but known transcript? YES! In that case we add the refernce id as gene id to the ones that have it.
```{r duplicates}
# For each 
stringtie_gtf <- import(file.path(dir, "02_RNA-Seq/04_stringtie/Zyagen/Zyagen_stringtie_merged_reference_guided.gtf"))
length(stringtie_gtf$gene_id)
stringtie_gtf[str_detect(stringtie_gtf$gene_id,"MSTRG") & !is.na(stringtie_gtf$ref_gene_id),]$gene_id <- stringtie_gtf[str_detect(stringtie_gtf$gene_id,"MSTRG") & !is.na(stringtie_gtf$ref_gene_id),]$ref_gene_id
length(stringtie_gtf$gene_id)

lncRNAs_out[str_detect(lncRNAs_out$gene_id,"MSTRG") & !is.na(lncRNAs_out$ref_gene_id),]$gene_id <- lncRNAs_out[str_detect(lncRNAs_out$gene_id,"MSTRG") & !is.na(lncRNAs_out$ref_gene_id),]$ref_gene_id

unique(lncRNAs_out$gene_id)
```
Keep Track of the number of novel genes identified: 
```{r}
get_novels <- function(lncRNAs_out, stringtie_gtf, lncRNAs_ref){
  intersect_feelnc_stringtie <- intersect(lncRNAs_out$gene_id, stringtie_gtf$gene_id)
  intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out$gene_id)
  novel_lncrnas = setdiff(intersect_feelnc_stringtie,intersect_reflnrna_out )
  return(novel_lncrnas)
}
novels <- list()
novel_lncrnas <- get_novels(lncRNAs_out, stringtie_gtf, lncRNAs_ref)
novels <- cbind(novels,"raw" = length(novel_lncrnas ))
```

## Remove unconcordant predictions
Since there are unconcordant predictions (same gene - multiple transcripts - assigned to different classes : lncRNA AND mRNA)
```{r echo=T, results='hide', eval=FALSE}
## REMOVE UNCONCORDANT PREDICTIONS
# Get the genes with unconcordant prediction labels for transcripts 
df <- data.frame()
df <- rbind(df, data.frame(gene_id = lncRNAs_out$gene_id, transcript_id = lncRNAs_out$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))


unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)

unconcordant_prediction

# Remove unconcordant predictions from sets 
lncRNAs_out <-lncRNAs_out[!(lncRNAs_out$gene_id %in% unconcordant_prediction$gene_id), ]
mRNAs_out <- mRNAs_out[!(mRNAs_out$gene_id %in% unconcordant_prediction$gene_id), ]

# Double check there is for real no unconcordant prediction anymore
df <- data.frame()
df <- rbind(df, data.frame(gene_id = lncRNAs_out$gene_id, transcript_id = lncRNAs_out$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarize(Unique_Elements = n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
nrow(unconcordant_prediction)

# NOVELS after filtering unconcordant predictions
novel_lncrnas <- get_novels(lncRNAs_out, stringtie_gtf, lncRNAs_ref)
novels <- cbind(novels,"after-unconcordant prediction" = length(novel_lncrnas ))
```
## Overlap of Gene_ids

we separate 

```{r}
grid::grid.newpage()
vd_draw(list(unique(lncRNAs_ref$gene_id), unique(lncRNAs_out$gene_id), unique(stringtie_gtf$gene_id),unique(mRNAs_ref$gene_id)), c("LNCRNA-ref","feelnc","stringtie","MRNA-ref"))
```

```{r}
grid::grid.newpage()
vd_draw(list(unique(lncRNAs_ref$gene_id), unique(lncRNAs_out$gene_id), unique(stringtie_gtf$gene_id)), c("LNCRNA-ref","feelnc","stringtie"))
```


```{r}
grid::grid.newpage()
vd_draw(list(lncRNAs_ref$gene_id, lncRNAs_out$gene_id, stringtie_gtf$gene_id, mRNAs_ref$gene_id, mRNAs_out$gene_id), c("LNCRNA-ref","feelnc","stringtie","MRNA-ref", "Feelnc-mrna"))
```



```{r}
grid::grid.newpage()

lt <-list(lncRNAs_ref$gene_id, lncRNAs_out$gene_id, stringtie_gtf$gene_id, mRNAs_ref$gene_id, mRNAs_out$gene_id) 
names(lt) <- c("LNCRNA-ref","feelnc","stringtie","MRNA-ref", "Feelnc-mrna")
m1 = make_comb_mat(lt)
UpSet(m1)
```


```{r}
grid::grid.newpage()
intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out$gene_id)
intersect_reflnrna_stringtie <- intersect(lncRNAs_ref$gene_id, stringtie_gtf$gene_id)
intersect_feelnc_stringtie <- intersect(lncRNAs_out$gene_id, stringtie_gtf$gene_id)
intersect_reflnrna_mrnaout<- intersect(lncRNAs_ref$gene_id, mRNAs_out$gene_id)

correctly_identified = intersect(intersect_reflnrna_stringtie,lncRNAs_out$gene_id)
missed = setdiff(intersect_reflnrna_stringtie,correctly_identified)
total = intersect_reflnrna_stringtie
df <- data.frame()
df <- rbind(df, data.frame(dataset = "total", number = length(total), type = "lncrna"))
df <- rbind(df, data.frame(dataset = "correctly identified", number = length(correctly_identified), type = "lncrna"))
df <- rbind(df, data.frame(dataset = "missed", number = length(missed), type = "lncrna"))
df <- rbind(df, data.frame(dataset = "predicted as Mrna", number = length(intersect_reflnrna_mrnaout), type = "lncrna"))


p1 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge())+
  geom_text(aes(label=number), vjust=1.6, color="white",
            position = position_dodge(0.9), size=3.5)+
  scale_fill_brewer(palette="Paired")+
  theme_minimal() + 
  labs(title="Number genes identified as lncRNAs")

prop_missed = (length(missed)/length(total))*100
prop_correctly = (length(correctly_identified)/length(total))*100

df <- data.frame()
df <- rbind(df, data.frame(dataset = "correctly identified", number = prop_correctly, type = "lncrna"))
df <- rbind(df, data.frame(dataset = "missed", number = prop_missed, type = "lncrna"))
p2 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity")+
  geom_text(aes(label=number), vjust=1.8, color="white", size=3)+
  labs(title="Proportion of genes identified as lncRNAs")

grid.arrange(p1, p2, nrow = 1)
   
```

## About the novel ones

How many of the genes which were missed, were filtered out in the filter step? 
```{r}
grid::grid.newpage()
missed = setdiff(intersect_reflnrna_stringtie,correctly_identified)
candidates = candidates_lncRNA$gene_id
vd_draw(list(missed, candidates),c("missed","candidates"))
missed_because_of_filtering <- setdiff(missed,candidates)
print("Missed because of filtering")
print(length(missed_because_of_filtering))
```

# Collapse transcripts - only take the one with max length

```{r}

# Instead of collapsing we select the longest transcript among the transcripts in one gene
get_only_max_transcript <- function(gr){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  gene_with_multiple_isoforms <-df[!duplicated(df$transcript_id),] %>% dplyr::group_by(gene_id) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1) 
  collapsed <-df[df$gene_id %in% gene_with_multiple_isoforms$gene_id,]%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width)) %>% dplyr::group_by(gene_id) %>% dplyr::slice(which.max(range))
  gr <- gr[gr$transcript_id %in% collapsed$transcript_id,]
  return(gr)
}
lncRNAs_out <- get_only_max_transcript(lncRNAs_out)
mRNAs_out <-  get_only_max_transcript(mRNAs_out)
```
# About the Novel ones

## Number Of exons distributon
Distribution of number of exons in newly predicted ones
```{r}
grid::grid.newpage()
get_nr_exons <- function(gr){
  df <- data.frame("gene_id" = gr$gene_id,"exon_number" = as.numeric(gr$exon_number))
  number_exons <- df %>% dplyr::group_by(gene_id) %>%dplyr::summarize(max_exon = max(exon_number))
  return(number_exons)
}

lncrnas_ref_exons <- lncRNAs_ref[lncRNAs_ref$type =="exon",]
lncrnas_ref_collapsed <- get_only_max_transcript(lncrnas_ref_exons)

df_l <- data.frame(get_nr_exons(lncRNAs_out))
df_l$type <- "pred"
df_ref_lncrnas <- get_nr_exons(lncrnas_ref_collapsed)
df_ref_lncrnas$type <- "ref"
df <- rbind(df_ref_lncrnas, df_l)

p1 <- ggplot(df, aes(x=max_exon,col =type, fill=type)) + 
  geom_histogram(position="identity", alpha=0.5, binwidth =1)
p1 <- p1 + scale_fill_manual(values=colors)

p2 <- ggplot(df, aes(x=max_exon,col =type, fill=type)) + 
  geom_histogram(aes(y=c(..count..[..group..==1]/sum(..count..[..group..==1]),
                         ..count..[..group..==2]/sum(..count..[..group..==2]))*100),position="identity", alpha=0.5,binwidth = 1)+
  labs(title="Number of exons Novel vs REF ")+ ylab ("Percentage of #exons") 
p2 <- p2 + scale_fill_manual(values=colors)

grid.arrange(p1,p2,nrow=1)

df_m <- data.frame(get_nr_exons(mRNAs_out))
df_m$type <- "ref-mrna"
df_m <- rbind(df, df_m)

p3 <- ggplot(df_m, aes(x=max_exon,col =type, fill=type)) + 
  geom_histogram(aes(y=c(..count..[..group..==1]/sum(..count..[..group..==1]),
                         ..count..[..group..==2]/sum(..count..[..group..==2]),
                         ..count..[..group..==3]/sum(..count..[..group..==3]))*100),position="identity", alpha=0.5,binwidth = 1)+
  labs(title="Number of exons Novel vs REF ")+ ylab ("Percentage of #exons") 


# Double check
df_m$type <- "mrnas"
df <- data.frame("gene" =  mRNAs_out$gene_id, "transcript" = mRNAs_out$transcript_id, stringsAsFactors = FALSE)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
```


# LENGTH OF GENES 
```{r}

# Double checking we have no duplicates - so we ave one transcript per gene - we sum up the length of the exons --> Length of the gene
df <- data.frame("gene" =  lncRNAs_out$gene_id, "transcript" = lncRNAs_out$transcript_id, stringsAsFactors = FALSE)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
# same for reference
df <- data.frame("gene" =  lncrnas_ref_collapsed$gene_id, "transcript" = lncrnas_ref_collapsed$transcript_id, stringsAsFactors = FALSE)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
lncrnas_ref_collapsed[lncrnas_ref_collapsed$gene_id == "ENSMMUG00000000801",]


# Define function for plotting
colors <- c("#D16103","#4E84C4")
histo_length <- function(gr,title,color){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  collapsed <-df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  p1 <- ggplot(collapsed, aes(x=range)) + 
    geom_histogram(position="identity", alpha=0.5, fill = color, binwidth = 40)+ 
    xlim(0,8000)+
    ylim(0,50)+
    labs(title=title)
  return(p1)
}

library(plyr)
plot_all <- function(ref, gr, colors){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  collapsed <-df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  collapsed$type <- "novel"
  df_ref <- data.frame("gene_id" = ref$gene_id,"transcript_id" = ref$transcript_id, "range_width" = width(ranges(ref)))
  collapsed_ref <-df_ref%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  collapsed_ref$type <- "ref"
  all <- rbind(data.frame(collapsed_ref),data.frame(collapsed))
  mu <- ddply(all, "type", summarise, grp.mean=mean(range))
  print(mu[mu$type == "novel",]$grp.mean)
  p3 <- ggplot(all, aes(x=range, col= type)) + 
    geom_histogram(position="identity",alpha=0.5, binwidth = 20) +
    ylim(0,50)+
    geom_vline(data=mu, aes(xintercept=grp.mean, color=type),
             linetype="dashed")+
    geom_text(aes(x = (mu[mu$type == "novel",]$grp.mean ), y = 20, label = as.integer(mu[mu$type == "novel",]$grp.mean))) +
    geom_text(aes(x = (mu[mu$type == "ref",]$grp.mean), y = 40, label = as.integer(mu[mu$type == "ref",]$grp.mean)), colour = colors[2]) +
    labs(title="GeneLength distribution NOVEL vs REFERENCE")
  p3 <- p3 + scale_color_manual(values=colors)
  return(p3)
}

plot_all(lncrnas_ref_collapsed,lncRNAs_out,colors)
p1 <-histo_length(lncRNAs_out,"GeneLength distribution  of NOVEL lncrnas",colors[1])
p2 <-histo_length(lncrnas_ref_collapsed,"GeneLength distribution  of REF lncrnas",colors[2])
p3 <- plot_all(lncrnas_ref_collapsed,lncRNAs_out,colors)
p3  
# plot ALL 
g1 <- as.grob(p1)
g2 <- as.grob(p2)

g3 <- as.grob(p3)

grid.arrange(
  grobs = list(g1,g2,g3),
  layout_matrix = rbind(c(1,2),
                        c(3,3))
)



```
The novel seem much longer or at least we have some VERY BIG outliers

```{r}
# Exploring the max length of transcripts! In novel much longer!
ref <- lncrnas_ref_collapsed
df_ref <- data.frame("gene_id" = ref$gene_id,"transcript_id" = ref$transcript_id, "range_width" = width(ranges(ref)))
collapsed_ref <-df_ref%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
maximum_length_reference <- max(collapsed_ref$range)


gr <- lncRNAs_out
df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
collapsed <-df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
collapsed[collapsed$range == max(collapsed$range),]

max_transcript <- gr[gr$transcript_id == "MSTRG.32406.1",]
ranges(max_transcript)


#filter the ones too long
collapsed <- collapsed[collapsed$range <maximum_length_reference,]
lncRNAs_out_outliers_filtered <- lncRNAs_out[lncRNAs_out$gene_id %in% collapsed$gene_id, ]
# how many we filtered out
length(lncRNAs_out$gene_id)- length(lncRNAs_out_outliers_filtered$gene_id)


lncRNAs_out_outliers_filtered
p1 <-histo_length(lncRNAs_out_outliers_filtered,"GeneLength distribution  of NOVEL lncrnas",colors[1])
p2 <-histo_length(lncrnas_ref_collapsed,"GeneLength distribution  of REF lncrnas",colors[2])
p3 <- plot_all(lncrnas_ref_collapsed,lncRNAs_out_outliers_filtered,colors)
  
# plot ALL 
g1 <- as.grob(p1)
g2 <- as.grob(p2)
g3 <- as.grob(p3)

grid.arrange(
  grobs = list(g1,g2,g3),
  layout_matrix = rbind(c(1,2),
                        c(3,3))
)

```

## Expression Levels of the novel ones
- novel_gene_ids: filtered set of gene ids which are considered novel by stringtie
Extra question: are all the novel MRG? if they are not, are they in mrna? if yes REMOVE. No they are not in mrna, there is one annotated and is a pseudogene
```{r}
length(novel_lncrnas)
# Check just to be suer if there is any novel one which is not a merged transcript
not_merged_but_novel <-novel_lncrnas[!str_detect(novel_lncrnas,"MSTRG")]
candidates_lncRNA[candidates_lncRNA$gene_id == "ENSMMUG00000045895",]
# it is a pseudogene
```
I need the expression Levels 
```{r echo=FALSE,  results='hide'}
directory_zyagen = file.path(baseDir,"data/02_RNA-Seq/04_stringtie_counts/Zyagen")
ctabs<- iterate_files(directory_zyagen, 't_data.ctab')
ctabs_zyagen_umis <-ctabs[sapply(ctabs, function(str) str_detect(str,"umis"))]
se <-extract_se(ctabs_zyagen_umis)
```

I only extract the ones in the novel - 
question : stringtie a transcript KNOWN assigned to merged gene - TO BE HANDLED!
this seem to happen when a merged transcript is assigned to a known gene, which has as well a known transcript.

```{r echo=FALSE,  results='hide'}
se_novel_only <- se[rowData(se)$gene_id %in% novel_lncrnas,]

rowData(se_novel_only[!str_detect(rowData(se_novel_only)$gene_id,"MSTRG")])
rowData(se_novel_only)
# How many we have
not_merged_but_novel <-novel_lncrnas[!str_detect(novel_lncrnas,"MSTRG")]

candidates_lncRNA[candidates_lncRNA$gene_id == "ENSMMUG00000045895",]

rowData(se_novel_only)$gene_id
table(unique(rowData(se_novel_only)$gene_id) %in% unique(novel_lncrnas))
table(unique(novel_lncrnas) %in% unique(rowData(se_novel_only)$gene_id))

dge <- DGEList(counts = assays(se_novel_only)$counts, group = se_novel_only$tissue, genes = as.data.frame(rowData(se_novel_only)$gene_id))

dge$counts
dge <- calcNormFactors(dge, method ="TMM")
?calcNormFactors
dge
assays(se_novel_only)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(se_novel_only)$logCPM) > 1
length(mask[mask == FALSE])
se_novel_only <- se_novel_only[mask, ]
dge <- dge[mask, ]
dim(dge)
mds <- plotMDS(dge, label = dge$samples$group)
```
Might be interesting to MDS the expression levels of NOVEL lncRNAs vs known lncRNAs

```{r echo=FALSE,  results='hide'}

lncrnas_ref_collapsed$gene_id
lncrnas_ref_collapsed$transcript_id

# might be interesting 

table(rowData(se)$t_name %in%lncrnas_ref_collapsed$transcript_id)
table(rowData(se)$gene_id %in% lncrnas_ref_collapsed$gene_id)
se_lnc_known <- se[rowData(se)$gene_id %in% lncrnas_ref_collapsed$gene_id,]


se_lnc_known <- se[rowData(se)$t_name %in%lncrnas_ref_collapsed$transcript_id,]
se_lnc_known_2 <- se[rowData(se)$gene_id %in% lncrnas_ref_collapsed$gene_id,]
setdiff(rowData(se_lnc_known_2)$gene_id, rowData(se_lnc_known)$gene_id)

# CAn it be we have so little ones in the ref??
dim(se_lnc_known)
dim(se)
rowData(se_lnc_known)

dge <- DGEList(counts = assays(se_novel_only)$counts, group = se_novel_only$tissue, genes = as.data.frame(mcols(se_novel_only)))
dim(dge)
dge <- calcNormFactors(dge)
assays(se_novel_only)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(se_novel_only)$logCPM) > 1
length(mask[mask == FALSE])
se_novel_only <- se_novel_only[mask, ]
dge <- dge[mask, ]
dim(dge)
mds <- plotMDS(dge, label = dge$samples$group)

```







## Other questions
## Why so many gene_ids from reference not in Stringtie? 
Solved: stringtie applies filtering on transcript lengths and coverage etc and so some are not present in the final one.

We want to check WHY in the reference we have so many lncRNAS that are not present in stringtie.
Maybe it was beacause the 
1) I need to filter out the lowly expressed genes from the reference

```{r}
# Number of gene ids present in ref and not in stringtie
ref_min_stringtie <- setdiff(lncRNAs_ref$gene_id,stringtie_gtf$gene_id)
stringtie_min_ref <- setdiff(stringtie_gtf$gene_id,lncRNAs_ref$gene_id) 
print("Not present in stringtie but present in reference")
print(length(ref_min_stringtie))
```
I want to check how long are the ones not present in stringtie
```{r}
#ref_min_stringtie
ref_min_stringtie

length(lncRNAs_ref[lncRNAs_ref$gene_id %in% ref_min_stringtie])
gr <- lncrnas_ref_exons
#gr <- get_only_max_transcript(gr)
df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
transcript_length <- df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))

t_in_dff <- transcript_length[unique(max_transcript_length$gene_id) %in% ref_min_stringtie,]
removed_because_length <- t_in_dff[t_in_dff$range < 200,]
ref_min_stringtie <- setdiff(ref_min_stringtie,removed_because_length$gene_id)
length(ref_min_stringtie)
```
some are eliminated because of that but not all.

So the idea is that we go and get the rawmapping counts (for instance, htseq counts) and see if we still had the missed genes there. If we had, we expect them to be lowly expressed or not expressed at all.
```{r}
htseq_files <- iterate_files(dir_hisat_zyagen, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")
directory = "/home/luisas/Desktop/cluster/proj/data/02_RNA-Seq/03_hisat/"
dds <- create_dds(htseq_files,directory)
```
Now  we look if in htse counts we still have all the ids
```{r}
# Number of gene ids present in ref and not in httseq
ref_min_htseq <- setdiff(lncRNAs_ref$gene_id,rownames(dds))
print("Not present in hhtseq but present in reference")
print(length(ref_min_htseq))
```
Now we are interested in the expression levels of the ones NOT present in stringtie but present in ref
```{r}
#ref_min_stringtie
dim(dds)
dds_f<- dds[rownames(dds) %in% ref_min_stringtie]
dim(dds_f)
```

```{r}
# Create DGE list object
dge <- DGEList(counts = assays(dds)$counts, group = dds$tissue, genes = as.data.frame(mcols(dds)))
dge <- calcNormFactors(dge)
assays(dds)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(dds)$logCPM) > 1


dds_expression_ok <-  dds[mask, ]
dds_low <- dds[!mask,]

# How many of the ones NOT present in stringtie are lowly expressd
length(intersect(ref_min_stringtie, rownames(dds_low)))

```



# Expression Levels of novel transcripts


## Load Needed
```{r echo=FALSE}
library(ggplot2)
library(GenomicFeatures)
library(zeallot)
library(tidyverse)
library(tximport)
library(tximportData)
library(edgeR)
library(SummarizedExperiment)
library(DESeq2)
library("RColorBrewer")
library("dplyr")
library(plyr)

# Load Utils
baseDir = "/home/luisas/Desktop/cluster/proj"
scripts_dir = file.path(baseDir,"code/ebola/src/scripts/")
source(paste0(scripts_dir,"de_utils.R"))
# Dirs

```

## Create Summarized Experiment
```{r echo=FALSE, results='hide'}
directory_zyagen = file.path(baseDir,"data/02_RNA-Seq/04_stringtie_counts/Zyagen")
ctabs<- iterate_files(directory_zyagen, 't_data.ctab')
ctabs_zyagen_umis <-ctabs[sapply(ctabs, function(str) str_detect(str,"umis"))]
se <-extract_se(ctabs_zyagen_umis)
```


```{r}
dge <- DGEList(counts = assays(se)$counts, group = se$tissue, genes = as.data.frame(mcols(se)))
dge <- calcNormFactors(dge)
saveRDS(dge, file.path("results", "dge_zyagen_umis.rds"))
```


Create the DGE List Object and calculate the normalization factors
```{r}
dge <- DGEList(counts = assays(se)$counts, group = se$tissue, genes = as.data.frame(mcols(se)))
dge <- calcNormFactors(dge)
saveRDS(dge, file.path("results", "dge_zyagen_umis.rds"))
```
We calc the logCPM values 
```{r}
assays(se)$logCPM <- cpm(dge, log = TRUE)
```
and filter out the lowly expressed genes
```{r}
mask <- rowMeans(assays(se)$logCPM) > 1
# How many we filtered out
length(mask[mask == FALSE])
dim(se)
se <- se[mask, ]
dge <- dge[mask, ]
dim(dge)
```

```{r}
htseq_files <- iterate_files(dir_hisat_zyagen, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")
dds <- create_dds(htseq_files,dir_hisat_zyagen)

sampleFiles = htseq_files
directory = "/home/luisas/Desktop/cluster/proj/data/02_RNA-Seq/03_hisat/"
samples_info <- sapply(sampleFiles,extract_info_sample_from_filename)
  # Create ddsHTSeq Object 
sampleTable <- data.frame(sampleName = sampleFiles,
                            fileName = sapply(sampleFiles,get_filename_and_dirs),
                            condition = samples_info[1,],
                            tissue = samples_info[2,], 
                            dpo = samples_info[3,], 
                            id = samples_info[4,],
                            complete_id = samples_info[5,]
  )

dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                    directory = directory,
                                    design= ~ 1)

rownames(dds)
colnames(dds)

assays(dds)$counts


diff_ids <- setdiff(rownames(dds), stringtie_gtf$gene_id )


dge <- DGEList(counts = assays(dds)$counts, group = dds$tissue, genes = as.data.frame(mcols(dds)))
dge <- calcNormFactors(dge)
assays(dds)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(dds)$logCPM) > 1
length(mask[mask == FALSE])
dds <- dds[mask, ]
dge <- dge[mask, ]
dim(dge)

diff_ids_2 <- setdiff(rownames(dds), stringtie_gtf$gene_id )
length(diff_ids)
length(diff_ids_2)

```



# --------------------------------------------------------------------------------

```{r}
grid::grid.newpage()
missed_because_of_filtering_ranges <- lncRNAs_ref[lncRNAs_ref$gene_id %in% missed_because_of_filtering]
length(unique(missed_because_of_filtering_ranges$gene_id))
missed_because_of_filtering_ranges <- missed_because_of_filtering_ranges[missed_because_of_filtering_ranges$type == "exon",]
# Extract 
width(ranges(missed_because_of_filtering_ranges))
df <- data.frame("gene_id" = missed_because_of_filtering_ranges$gene_id, "range_width" = width(ranges(missed_because_of_filtering_ranges)))
gene_lengths <- df %>% group_by(gene_id) %>% summarize(sum(range_width))
# Quick testing
gene_lengths[gene_lengths$gene_id == "ENSMMUG00000037481",]
df[df$gene_lengths == "ENSMMUG00000037481",]
length(intersect(missed_because_of_filtering_ranges$gene_id, gene_lengths$gene_id))

missed_because_of_filtering_ranges
ggplot(gene_lengths, aes(x=`sum(range_width)`)) +
  geom_histogram(fill="blue")+ 
  labs(title = "Distribution of gene lenghts misclassified because of filtering step ")


```




In terms of percentage how many genes we can retrieve
```{r}
grid::grid.newpage()
intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out$gene_id)
intersect_reflnrna_stringtie <- intersect(lncRNAs_ref$gene_id, stringtie_gtf$gene_id)
intersect_feelnc_stringtie <- intersect(lncRNAs_out$gene_id, stringtie_gtf$gene_id)
combo <- c(feelnc = length(unique(lncRNAs_out$gene_id)), ref_lnc = length(unique(lncRNAs_ref$gene_id)), stringtie = length(unique(stringtie_gtf$gene_id)), "feelnc&ref_lnc" = length(intersect_reflnrna_out), "feelnc&stringtie" = length(intersect_feelnc_stringtie), "ref_lnc&stringtie" = length(intersect_reflnrna_stringtie))

plot(euler(combo), quantities = TRUE)
```




Subselect only transcript ids
```{r }
grid::grid.newpage()
transcripts_ref <- lncRNAs_ref[!is.na(lncRNAs_ref$transcript_id),]
transcripts_out <- lncRNAs_out[!is.na(lncRNAs_out$transcript_id),]
transcripts_mRNAs <- mRNAs_ref[!is.na(mRNAs_ref$transcript_id),]
transcripts_stringtie <- stringtie_gtf[!is.na(stringtie_gtf$transcript_id),]
vd_draw(list(transcripts_ref$transcript_id, transcripts_out$transcript_id, transcripts_stringtie$transcript_id),c("ref","feelnc","stringtie"))
```

But how many in the newly detected one are MRG

```{r }
grid::grid.newpage()
transcripts_mRNAs <- mRNAs_ref[!is.na(mRNAs_ref$transcript_id),]
stringtie_mrg <- transcripts_stringtie[sapply(transcripts_stringtie$transcript_id, function(str) str_detect(str,"MSTRG")),]
stringtie_not_mrg <- transcripts_stringtie[sapply(transcripts_stringtie$transcript_id, function(str) !str_detect(str,"MSTRG")),]
vd_draw(list(transcripts_ref$transcript_id, transcripts_out$transcript_id, stringtie_mrg$transcript_id, stringtie_not_mrg$transcript_id, transcripts_mRNAs$transcript_id),c("ref","feelnc","stringtie_mrg", "stringtie_not_mrg","mrnasref"))
```
```{r }
grid::grid.newpage()

vd_draw(list(transcripts_ref$transcript_id, transcripts_out$transcript_id, stringtie_mrg$transcript_id, stringtie_not_mrg$transcript_id),c("ref","feelnc","stringtie_mrg", "stringtie_not_mrg"))
```



We are interested in the distribution of the numbers of exons
```{r }
grid::grid.newpage()
table(as.numeric(transcripts_out$exon_number))
```


