---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Analysis on genes correlated with the viral load 

# 0. Imports 
```{r include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr); library(Seurat); library(Matrix); library(SingleCellExperiment)
library(stringr); library(rtracklayer); library(RColorBrewer); library(scales)
library(ggthemes); library(org.Mmu.eg.db); library(ggplot2); library(ggvenn)
library(ggExtra); library(Gviz); library(ggpubr); library(ggsci); library(gridExtra)


# Define paths for data
source("../../utils/00_datapaths.R")
source("../../utils/01_lncrna_annotation_utils.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")
source("../../utils/05_utils_corr.R")

theme_paper <- theme(panel.background = element_rect(fill = "white"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=10), legend.title = element_blank())

# --------------------------------
# Set thresholds 
RHOTHRESHOLD = 0.0
FDRTHRESHOLD<- 0.05

# ---------------------------------------------------
#           Load  references 
# ---------------------------------------------------
ref <- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_addedgenesNovel.gtf"))
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))

# files ex vivo 
robjectsdir_stats <- file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/")
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))
ebola_genome_percentage_df_exvivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))


all_lncrnas_invivo <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/", "all_lncrnas.rds"))
annotated_mrnas_invivo <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/", "annotated_mrnas.rds"))
ebola_genome_percentage_df_invivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))


```



# 1. Load viral load correaltion files 
```{r imports}
# 1. Ex vivo no ebola 
# 1. Ex vivo no ebola 
correlations_exvivo_noebola <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved_correlation.rds"))
filtered <- filter_significant_correlations(correlations_exvivo_noebola, ref, all_lncrnas, annotated_mrnas, FDRTHRESHOLD, RHOTHRESHOLD, usep = F)
correlations_exvivo_noebola <- filtered[[1]]
table(correlations_exvivo_noebola$type)
sig_cor_lnc_exvivo_noebola <- filtered[[2]]
#write.xlsx(sig_cor_lnc_exvivo_noebola, "/home/mariasr/cluster/02.ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/sig_corr_lnc.xls", sheetName = "significant_lnc", col.names = TRUE, row.names = TRUE, append = FALSE)
mono_live_h24_inf_exvivo <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved.rds"))


# 2. In vivo no ebola
mono_live_h24_inf_invivo_noebola <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_log10TPT_late_ebolaremoved.rds"))
correlations_invivo_noebola <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved_correlation_withzeros_log10TPT.rds"))
sig_cor_lnc_invivo_noebola <- filter_significant_correlations(correlations_invivo_noebola, ref, all_lncrnas_invivo, annotated_mrnas_invivo, FDRTHRESHOLD, RHOTHRESHOLD, usep = T)[[2]]
```

#2. Plots figure paper 
```{r}

# load object 
immune.combined <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready_infectionstatus_noebola.rds"))

# subset data for expression plot 
monocytes <- subset(immune.combined, idents= "Monocyte")
mono <- monocytes[,monocytes$cond != "irrad"]
class <- cbind(as.data.frame(mono$cond),as.data.frame(mono$dpi), as.data.frame(mono$infection))
class$type <- ""
class$type <- ifelse(class$`mono$cond`=="media", "baseline", ifelse(class$`mono$cond`=="live" & class$`mono$dpi`== "H004", "remove", ifelse(class$`mono$infection` == "Infected", "infected", "bystander")))
class <- class[class$type != "remove",]
type <- class$type
mono <- mono[,colnames(mono) %in% rownames(class)]
mono <- AddMetaData(mono, type, col.name = "type" )
table(mono$type, mono$infection)


plot_correlation <- function(gene,mono,ebola_genome_percentage_df_exvivo, keepzeros = T){
  #get expression
  genes_expression <-  get_expression_summary_gene(gene, mono, ebola_genome_percentage_df_exvivo, T)
  genes_expression$gene <- as.character(genes_expression$gene)
  genes_expression$orth <- unlist(lapply(as.character(genes_expression$gene), get_orthologname_))
  
  #plot correlation
  correlation <- ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = orth))+stat_smooth(method = "loess", formula = y ~ x, n = 15, span =2, se = T)+theme_bw()+theme(legend.position = "none", plot.title = element_text(size = 12), text = element_text(size = 17))+ xlab("% viral load")+ylab("log(CP10K+1)")+theme_paper+ ggtitle(get_orthologname_(gene))

  #subset data
  expression <- mono[gene,]
  bystander <-expression[,expression$type == "bystander"]@assays$RNA@data
  baseline <-expression[,expression$type == "baseline"]@assays$RNA@data
  infected <-expression[,expression$type == "infected"]@assays$RNA@data
  
  get_average <- function(name){
    v <- as.vector(get(name))
    b <- data.frame(t(Rmisc::CI(v)), type = name)
    b$pct.exp <- PercentAbove(v,0)*100
    return(b)
    }
  
    exp <- Reduce(rbind,lapply(as.character(unique(expression$type)), function(x) get_average(x)))
    exp$type <- factor(exp$type, levels = c("baseline", "bystander", "infected"))

  expression_plot <- ggplot(exp, aes(x= type, y = mean,))+geom_point(shape = 1,colour = "black", fill = "white", alpha = 6, stroke = 1, aes(size = pct.exp),   show.legend = T)+scale_size(range = c(5,15), limits = c(1, 7), breaks =c())+theme_classic()+geom_errorbar(aes(ymax = upper, ymin = lower), width=.2)+ylab("log(CP10K+1)")+theme(text = element_text( size = 18))+xlab("")+theme(legend.title = "% of cells", legend.position = "right")+theme(axis.line.x = element_blank(), axis.ticks.x = element_blank())+theme_paper+theme_paper
  return(list(correlation, expression_plot))
  }


# genes DE 
genes_DE <- c("MSTRG.181870-unknown","ENSMMUG00000061044-unknown", "ENSMMUG00000058644-unknown","ENSMMUG00000064224-unknown", "MSTRG.238227-unknown" )
lapply(genes_DE, function(gene) plot_correlation(gene, mono, ebola_genome_percentage_df_exvivo, T))
```



# 3. Other plots
```{r imports}

# 1. Check how many are significantly up/down
plot_corr_genes_stat(correlations_exvivo_noebola)

# 2. Plot ex vivo genes
genes <- unique(sig_cor_lnc_exvivo_noebola$gene)
p_corr_exvivo_noebola <- visualize_correlation(genes, infected,ebola_genome_percentage_df_exvivo, keepzeros = T)+ggtitle("EXVIVO")
p_corr_exvivo_noebola

#Plot lncRNA 
genes_positive <- c("ENSMMUG00000061044-unknown",  "MSTRG.238227-unknown")
p_corr_exvivo_positive <- visualize_correlation(genes_positive, infected,ebola_genome_percentage_df_exvivo, keepzeros = T)+ggtitle("EXVIVO")
p_corr_exvivo_positive

# Plot 2 lncRNA negative
genes_negative <- c("ENSMMUG00000064224-unknown", "IFIT2", "ISG15", "MX1")
p_corr_exvivo_negative <- visualize_correlation(genes_negative, mono_live_h24_inf_exvivo_noebola,ebola_genome_percentage_df_exvivo, keepzeros = T)+ggtitle("EXVIVO")
p_corr_exvivo_negative

# 3. Visualize in vivo
visualize_correlation(genes, mono_live_h24_inf_invivo_noebola,ebola_genome_percentage_df_invivo, keepzeros = T)+ggtitle("INVIVO")
visualize_correlation_window(genes_positive, mono_live_h24_inf_exvivo_noebola, keepzeros = T)+ggtitle("INVIVO")

# 4.lncRNA statistics 
significant_lnc <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/significant_lnc.rds"))

significant_lnc$type <- ifelse(str_detect(significant_lnc$gene, regex("MST")), "novel","annotated")
significant_lnc$corr <- ifelse(significant_lnc$rho < 0, "negative","positive")
ggplot(significant_lnc, aes( x=significant_lnc$corr, fill=as.factor(type))) + geom_bar()+ scale_y_continuous(breaks = seq(0, 14, by = 2))+ theme_paper

```




# 4. Plot bystander vs infected 

```{r imports}
#------------------------ EX VIVO -----------------------------
# 1. Load Ex vivo 
immune.combined <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready_infectionstatus_noebola.rds"))
print(table(colnames(immune.combined) == rownames(ebola_genome_percentage_df_exvivo)))
# 2. Separate  infected and bystanders
infected_cells <- subset(immune.combined, infection == "Infected")
infected_cells <- colnames(infected_cells)
monocytes <- subset(immune.combined, ident = "Monocyte")
baseline <- monocytes[,monocytes$cond == "media"]
baseline <- baseline[, baseline$infection == "Not Infected"]
late <- monocytes[,monocytes$cond == "live" & monocytes$dpi == "H024"]
infected <- late[,infected_cells]
bystanders <- late[,setdiff(colnames(late), infected_cells)]

# Check 
length(unique(colnames(bystanders))) + length(unique(colnames(infected))) == length(unique(colnames(late)))
#saveRDS(infected, file.path(data_path,"/05_RObjects/03_prep/infected.rds"))
#saveRDS(bystanders, file.path(data_path,"/05_RObjects/03_prep/bystander.rds"))
#saveRDS(baseline,file.path(data_path,"/05_RObjects/03_prep/baseline.rds") )


#------------------------ IN VIVO -----------------------------
immune.combined_invivo <- readRDS(file.path("/home/mariasr/cluster/02.ebola_sc/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"))
print(table(colnames(immune.combined_invivo) == rownames(ebola_genome_percentage_df_invivo)))
# 2. Add infection information
immune.combined_invivo$infection <- ebola_genome_percentage_df_invivo$classification 
immune.combined_invivo$viral_load <- ebola_genome_percentage_df_invivo$percentage_viral_reads 
infected_cells <- rownames(ebola_genome_percentage_df_invivo[ebola_genome_percentage_df_invivo$classification == "Infected",])
monocytes_invivo <- subset(immune.combined_invivo, ident = "Monocyte")
baseline_invivo <- monocytes_invivo[,monocytes_invivo$group_dpi %in%c("DPI-04","DPI-30", "DPI000")]
late_invivo <- monocytes_invivo[,monocytes_invivo$group_dpi %in% c("DPI007","DPI008")]
infected_invivo <- late_invivo[,infected_cells]
bystanders_invivo <- late_invivo[,setdiff(colnames(late_invivo), infected_cells)]

# Check 
length(unique(colnames(bystanders))) + length(unique(colnames(infected))) == length(unique(colnames(late)))
#saveRDS(infected_invivo, file.path(data_path,"/05_RObjects/03_prep/infected_invivo.rds"))
#saveRDS(bystanders_invivo, file.path(data_path,"/05_RObjects/03_prep/bystander_invivo.rds"))
#saveRDS(baseline_invivo, file.path(data_path,"/05_RObjects/03_prep/baseline_invivo.rds"))

#plot infected vs bystander 
test_boxplots <- function(gene){
  p1 <- plot_boxplot_cells(gene,baseline, bystanders, infected, "ex vivo")
  #p2 <- plot_boxplot_cells(gene,baseline_invivo, bystanders_invivo, infected_invivo, "in vivo")
  p3 <- plot_boxplot_cells_two(gene, baseline, late, "ex vivo")
  return(list(p1, p3))
}

# Test example 
genes <- c("MSTRG.238227-unknown", "ENSMMUG00000064224-unknown", "ENSMMUG00000061044-unknown", "ENSMMUG00000058644-unknown", "MSTRG.181870-unknown")
#positive corr
test_boxplots("ENSMMUG00000061044-unknown")
test_boxplots("ENSMMUG00000058644-unknown")
test_boxplots("MSTRG.181870-unknown")

#negative corr
test_boxplots("MSTRG.238227-unknown")

test_boxplots("ENSMMUG00000064224-unknown")
test_boxplots(genes)

```

