---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r imports}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(RColorBrewer)
library(scales)
library(ggthemes)
library(org.Mmu.eg.db)
library(ggplot2)
library(ggvenn)
library(ggExtra)
library(Gviz)
library(ggpubr)

# Define paths for data
source("../../utils/00_datapaths.R")
source("../../utils/01_lncrna_annotation_utils.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

# Load needed files EX VIVO 
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds"))
mono_live_h24_inf_exvivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds"))
robjectsdir_stats <- file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/")
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))
ebola_genome_percentage_df <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))
ref <- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_addedgenesNovel.gtf"))

# --------------------------------
# 1. LOAD and prep CORRELATIONS
RHOTHRESHOLD = 0.05
FDRTHRESHOLD<- 0.05
correlations_exvivo_all <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds"))
correlations_exvivo <- correlations_exvivo_all[!is.na(correlations_exvivo_all$pval),]
correlations_exvivo <- correlations_exvivo[!is.na(correlations_exvivo$pval) & correlations_exvivo$pval < FDRTHRESHOLD & abs(correlations_exvivo$rho) > RHOTHRESHOLD, ]

# Add Gene Type
correlations_exvivo$type <- "-"
correlations_exvivo[correlations_exvivo$gene %in% all_lncrnas ,]$type <- "lnc"
correlations_exvivo[correlations_exvivo$gene %in% annotated_mrnas ,]$type <- "pc"
# Extract lnc only
sig_cor_lnc_exvivo <- correlations_exvivo[correlations_exvivo$type == "lnc",]
sig_cor_lnc_exvivo$gene_name <- unlist(lapply(sig_cor_lnc_exvivo$gene, get_orthologname_))
# Remove ebola gene 
sig_cor_lnc_exvivo <- sig_cor_lnc_exvivo[sig_cor_lnc_exvivo$gene != "MSTRG.252489-unknown",]


# ------- load in vivo 
immune.combined <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"))
mono_live_h24_inf_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds"))

correlations_invivo_all <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds"))
correlations_invivo <- correlations_invivo_all[!is.na(correlations_invivo_all$pval) & !is.na(correlations_invivo_all$pval) & correlations_invivo_all$pval < FDRTHRESHOLD & abs(correlations_invivo_all$rho) > RHOTHRESHOLD, ]

correlations_invivo$type <- "-"
correlations_invivo[correlations_invivo$gene %in% all_lncrnas ,]$type <- "lnc"
correlations_invivo[correlations_invivo$gene %in% annotated_mrnas ,]$type <- "pc"


sig_cor_lnc_invivo <- correlations_invivo[correlations_invivo$type == "lnc",]
sig_cor_lnc_invivo$gene_name <- unlist(lapply(sig_cor_lnc_invivo$gene, get_orthologname_))
# Remove ebola gene 
sig_cor_lnc_invivo <- sig_cor_lnc_invivo[sig_cor_lnc_invivo$gene != "MSTRG.252489-unknown",]
```


# Visualize umap plot 
```{r lala}
infected_cells <- rownames(ebola_genome_percentage_df[ebola_genome_percentage_df$classification == "Infected",])
immune.combined$viral_load <- ebola_genome_percentage_df[colnames(immune.combined),]$percentage_viral_reads
immune.combined$log_viralload <- log10(immune.combined$viral_load+1)

pdf(file.path(plots, "05/viralload.pdf"), width = 7, height = 6)
FeaturePlot(immune.combined, features = "log_viralload", cols = c("grey", "#9A0000"), pt.size = 0.1, order = T)+theme_void()+theme(text = element_text(size = 15), title = element_blank())
dev.off()

FeaturePlot(immune.combined, features = "log_viralload", cols = c("grey", "#9A0000"), pt.size = 0.1, order = T)+theme_void()+theme(text = element_text(size = 15), title = element_blank())

```



# Are the lncRNAs found DE very lowly expressed? 
```{r imports}
PercentAbove <- function(x, threshold){
  return(length(x = x[x > threshold]) / length(x = x))
}

# Compute mean expression and % cells 
get_average <- function(gene, mono_live_h24_inf){
  exp <- mono_live_h24_inf[gene,]@assays$RNA@data
  v <- as.vector(exp)
  b <- data.frame(t(Rmisc::CI(v[v>1])), gene = gene)
  b$pct.exp <- PercentAbove(v,1)*100
  b$n_cells <- count(v > 1)
  return(b)
}


# ---------------------   What i need to know ---------------------


# 1 ----- How many lnc and pc are found correlated in vivo and ex vivo
print("lnc significantly correlated ex vivo")
length(unique(sig_cor_lnc_exvivo$gene_name))
print("pc significantly correlated ex vivo")
length(unique( correlations_exvivo[correlations_exvivo$type == "pc",]$gene))

# proportion
length(unique(sig_cor_lnc_exvivo$gene_name))*100/length(unique( correlations_exvivo[correlations_exvivo$type == "pc",]$gene))

print("lnc significantly correlated IN vivo")
length(unique(sig_cor_lnc_invivo$gene_name))
print("pc significantly correlated IN vivo")
length(unique( correlations_invivo[correlations_invivo$type == "pc",]$gene))
length(unique(sig_cor_lnc_invivo$gene_name))*100/length(unique( correlations_invivo[correlations_invivo$type == "pc",]$gene))


# 1b: how many novel 
table(substr(sig_cor_lnc_exvivo$gene,1,3))


# 2 -----In how many cells is each of the genes expressed
summary_exvivo <- Reduce(rbind, lapply(sig_cor_lnc_exvivo$gene, function(x) get_average(x, mono_live_h24_inf_exvivo)))
summary_invivo <- Reduce(rbind, lapply(sig_cor_lnc_invivo$gene, function(x) get_average(x, mono_live_h24_inf_invivo)))
min(summary_exvivo$n_cells)
ggplot(summary_exvivo, aes(x = n_cells))+geom_histogram()+theme_classic()+xlab("number of cells")+theme(text = element_text(size = 15))+scale_fill_brewer(palette = "Set1")


# 3 --------- How big is the overlap 
x <- list(invivo = summary_invivo$gene, exvivo = summary_exvivo$gene)
ggvenn(
  x, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 6
  )

# ----- Are the ones not identified in vivo missing? 
missing_invivo <- setdiff(summary_exvivo$gene,summary_invivo$gene)
summary_in_vivo_missing <- Reduce(rbind, lapply(missing_invivo, function(x) get_average(x, mono_live_h24_inf_invivo)))

summary_invivo$found <- "in vivo - found"
summary_in_vivo_missing$found <- "in vivo - missing"
summary_exvivo$found <- "ex vivo"
s <- rbind(summary_invivo, summary_in_vivo_missing, summary_exvivo)

pdf(file.path(plots, "05/ditribution_cells.pdf"), width = 7, height = 6)
ggplot(s, aes(x = n_cells, fill = found, color = found))+geom_density(alpha = 0.7)+theme_classic()+xlab("number of cells (log10)")+theme(text = element_text(size = 15))+scale_fill_brewer(palette = "Set1")+scale_color_brewer(palette = "Set1")+scale_x_log10()
dev.off()
wilcox.test(summary_invivo$n_cells, summary_exvivo$n_cells)
wilcox.test(summary_in_vivo_missing$n_cells, summary_exvivo$n_cells)

# ------ The one not identified ex vivo is there, low p value but not enough to be significant 
missing_exvivo <- setdiff(summary_invivo$gene,summary_exvivo$gene)
summary_ex_vivo_missing <- Reduce(rbind, lapply(missing_exvivo, function(x) get_average(x, mono_live_h24_inf_exvivo)))
correlations_exvivo_all[correlations_exvivo_all$gene == summary_ex_vivo_missing$gene, ]
```

# General functional imputation 
```{r imports}


get_cis <- function(gr_gene, gr_hits =mrnas_ranges, maxgap =1000000L ){
  p <- findOverlapPairs(gr_gene, get_gene_only(gr_hits), maxgap=maxgap, ignore.strand = TRUE)
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  if(length(p@second) == 0 ){
    df <- data.frame(gene=gr_gene$gene_id, distances=NA, second = NA, second_name = NA, chr = seqnames(gr_gene), stringsAsFactors = F )
  }else{
    distances <- unlist(lapply(1:length(p@second), function(index) GenomicRanges::distance(gr_gene, p@second[index], ignore.strand = TRUE) ))
    df <- data.frame(gene=gr_gene$gene_id, distances=distances, second = p@second$gene_id,second_name = p@second$gene_name, stringsAsFactors = F, chr = seqnames(gr_gene) )
  }
  return(df)
}

do_go <- function(list, universe, keytype = "SYMBOL", p = 0.05 ){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =  org.Mmu.eg.db, keyType = keytype,ont = "BP",universe =universe,  pvalueCutoff = p, qvalueCutoff = p )
  a1 <- barplot(ego, showCategory=5)
  a2 <- clusterProfiler::dotplot(ego)
  a3 <- clusterProfiler::heatplot(ego)
  ego@result
  return(list(a1,a2,a3))
}

# --------------- 1 ---------------------
# Investigate what is going on
genes <- sig_cor_lnc_exvivo$gene
genes_expressed <- rownames(mono_live_h24_inf_exvivo)
genes_correlated <- correlations_exvivo[correlations_exvivo$type == "pc",]$gene
ref_genes <- ref[ref$type == "gene", ]
# ref correspondence
corresp <- distinct(data.frame(id = ref$gene_id, name = ref$gene_name))
rownames(corresp) <- corresp$id
genes_pool <- ref_genes[ref_genes$gene_name %in% genes_expressed, ]
genes_pool_corr <- ref_genes[ref_genes$gene_name %in% genes_correlated, ]

cis_exvivo <- Reduce(rbind,lapply(genes, function(gene) get_cis(ref_genes[ref_genes$gene_name == gene, ], genes_pool_corr)))
cis_exvivo <- cis_exvivo[cis_exvivo$gene != cis_exvivo$second,]
cis_exvivo <- cis_exvivo[!is.na(cis_exvivo$gene), ]

#do_go(unique(cis_exvivo$second_name), genes_expressed)
rownames(correlations_exvivo) <- correlations_exvivo$gene
correlations_exvivo$direction <- ifelse(correlations_exvivo$rho > 0, "up","down")

cis_exvivo$direction_second <- correlations_exvivo[cis_exvivo$second_name,]$direction
cis_exvivo$direction <- correlations_exvivo[corresp[cis_exvivo$gene,]$name,]$direction
cis_exvivo$conc <- paste(cis_exvivo$direction, cis_exvivo$direction_second, sep = "-")

ggplot(cis_exvivo, aes(conc))+geom_bar(stat = "count")

#print(cis_exvivo[order(cis_exvivo$second_name),])
#do_go(cis_exvivo$second_name, universe = genes_expressed,  p = 0.1)
```

```{r plots}
# --------------------------------
# 2. PLOT MAIN 
#    Examples 
dim_big <- 0.8
#genes_expression <- Reduce("rbind", lapply(sig_cor_lnc_exvivo$gene, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo)))

cols <- c('ENSMMUG00000064224-unknown' = 'red',  "ENSMMUG00000056793-unknown" = "#6C0202", "NDUFA10" = "grey", "HDLBP" = "black")
highlights <- c('ENSMMUG00000064224-unknown' ,  "ENSMMUG00000056793-unknown", "HDLBP", "NDUFA10")
sizes <- c('ENSMMUG00000064224-unknown' = dim_big,  "ENSMMUG00000056793-unknown" = dim_big, "NDUFA10" = dim_big, "HDLBP" = dim_big, "other" = 0.2)
genes_expression <-  Reduce("rbind", lapply(highlights, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo)))


# 2. Plot 
pdf(file.path(plots, "05/main_corr.pdf"), width = 9, height = 5)
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$id <- ifelse(genes_expression$gene %in% highlights, genes_expression$gene, "other")
nlnc = length(unique(genes_expression$gene))
correaltion <-ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene, size = id))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 15, span =0.7)+
      theme_classic()+
      theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
      scale_color_manual(values = cols)+ggtitle(paste("n = ", nlnc))+
      scale_size_manual(values = sizes)

correaltion
dev.off()

correaltion
```

```{r correlations}

# -------------------------------------
# 3.  IDENTIFY GENE CLUSTER 
# -------------------------------------
# -------- PLOT REGION 
pdf(file.path(plots, "05/region.pdf"), width = 10, height = 6)
plot_region("ENSMMUG00000064224-unknown", range = 1200000, palette_plot_percentage = rep("grey",4), getmax = T )
dev.off()

gene_id <- "ENSMMUG00000064224"
range = 1000000
chr <- paste("chr",unique(as.character(seqnames(ref[ref$gene_id == gene_id,]))), sep = "")
# Define region nearby 
from  <- max(0,(min(start(ref[ref$gene_id == gene_id,])) -range))
to <- (max(end(ref[ref$gene_id == gene_id,]))+range)
ref_small<- ref[start(ref) > from  & end(ref) < to   ,]
# GENE CLUSTER 
genes_nearby <- ref_small[seqnames(ref_small) == 12 & ref_small$type == "gene",]$gene_name
#de_mono <- de_all[de_all$celltype == "Monocyte",]
#overlap <- genes_nearby[genes_nearby %in% de_mono$primerid]
cluster_names_ex <- genes_nearby[genes_nearby %in% correlations_exvivo$gene]
cluster_names_in <- genes_nearby[genes_nearby %in% correlations_invivo$gene]
overlap <- intersect(cluster_names_in, cluster_names_ex)
#EX VIVO
correlations_exvivo[correlations_exvivo$gene %in% overlap, ]
# IN VIVO 
correlations_invivo[correlations_invivo$gene %in% overlap, ]

lapply(cluster_names_ex, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf_exvivo))
#lapply(cluster_names_in, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf_invivo))


```

```{r plots}
# --------------------------------
# 2. PLOT MAIN 
cols <- c('highlight' = 'red', 'other' = 'grey')
genes_expression <- Reduce("rbind", lapply(sig_cor_lnc_exvivo$gene, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo)))
pdf(file.path(plots, "05/main_corr.pdf"), width = 7, height = 5)
# 2. Plot 
ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene))+stat_smooth(method = "loess", formula = y ~ x, size = 0.4, se = F, n = 7, span = 0.5)+
  theme_classic()+
  theme( legend.position = "", text = element_text(size = 17))+ 
  xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
  scale_color_manual(values = c(rep("#848484", length(unique(genes_expression$gene)))))
dev.off()




genes_expression <-  readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/genesexpr.rds"))

cols <- c('ENSMMUG00000064224-unknown' = 'red',  "ENSMMUG00000056793-unknown" = "orange", "ENSMMUG00000028689-unknown" = "dark green")
highlights <- c('ENSMMUG00000064224-unknown' ,  "ENSMMUG00000056793-unknown", "ENSMMUG00000028689-unknown", "NDUFA10")
sizes <- c('ENSMMUG00000064224-unknown' = 1,  "ENSMMUG00000056793-unknown" = 1, "ENSMMUG00000028689-unknown" = 1, "other" = 0.2)
# 1. Get the expression of all genes 
# Genes: sig_cor_lnc$gene


# 2. Plot 
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$id <- ifelse(genes_expression$gene %in% highlights, genes_expression$gene, "other")
nlnc = length(unique(genes_expression$gene))
correaltion <-ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene, size = id))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 6, span = 0.4)+
      theme_classic()+
      theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
      scale_color_manual(values = cols)+ggtitle(paste("n = ", nlnc))+
      scale_size_manual(values = sizes)
correaltion

```



