---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r imports}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr); library(Seurat); library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(reshape2)
library(circlize)
library(ggthemes)
library(wesanderson)
library(Gviz)
library(ggpubr)
library(ggExtra)

source("../../utils/00_datapaths.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation
ref <- import(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path(data_path, "/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE/")

immune.combined <- readRDS(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")

ebola_genes <- paste(ebola_ref$gene_id, "-unknown", sep ="")
orthologs <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
robjectsdir_stats <- file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/")
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))


# Load de 
de_all<- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds"))
de_lnc <- unique(de_all[de_all$type == "lnc" & de_all$celltype == "Monocyte",]$primerid)
ebola_genome_percentage_df <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)

# Prepare metadata
immune.combined$cond <- factor(immune.combined$cond, levels = c("media", "irrad", "live"))
immune.combined$condition <- paste(immune.combined$cond, immune.combined$infection)
immune.combined$celltype <- Idents(immune.combined)
immune.combined$celltype_group <- paste(immune.combined$celltype, immune.combined$cond, sep = "_")

mono <-  subset(immune.combined, ident="Monocyte")
mono <- mono[,mono$cond %in% c("media","live")]
mono$condition <- factor(mono$condition, levels = c("media Not Infected", "live Not Infected", "live Infected"))

mono$labels  <- as.character(mono$condition)
mono$labels  <- gsub("media Not Infected", "Control", mono$labels )
mono$labels  <- gsub("live Not Infected", "Bystanders", mono$labels )
mono$labels  <- gsub("live Infected", "Infected", mono$labels )
mono$labels <- factor(mono$labels, levels = c("Control", "Bystanders", "Infected"))
mono$log_viraload <- log(mono$viraload)


# Prepare subsets
mono_live <- mono[,mono$cond == "live"]
mono_live_h24 <- mono_live[,mono_live$dpi == "H024"]
mono_live_h24_inf <- mono_live_h24[,mono_live_h24$infection == "Infected" ]
mono_live_h24_notinf <- mono_live_h24[,mono_live_h24$infection == "Not Infected" ]
saveRDS(mono_live_h24_inf, file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds"))
print("")



#DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(4, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))


summary_gene <- function(gene_id, mono, ebola_genome_percentage_df ){
  # Visual representation

  p2 <- FeaturePlot(mono, feature = gene_id, cols =  c("lightgrey", "dark blue"), order = T)+theme_sc+ggtitle(get_orthologname_(gene_id))
  # General inspections
  #p3 <- plot_viral_load_gene(gene_id, mono, ebola_genome_percentage_df)
  #p4 <- get_prop(gene_id)
  #p5 <- boxplot_bystanders(get_expression_summary_gene(gene_id,mono))
  p6 <- check_expressed_vs_infection(gene_id, mono, ebola_genome_percentage_df)
  return(list(p2,p6))
}
```



# Correlation with viral load

```{r correlations}
RHOTHRESHOLD = 0.0
FDRTHRESHOLD<- 0.1
correlations_infected_24 <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds"))
correlations_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval),]
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]

# Add Gene Type
sig_correlations_pc_infected_24$type <- "-"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"
sig_cor_lnc <- sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$type == "lnc",]
sig_cor_lnc$gene_name <- unlist(lapply(sig_cor_lnc$gene, get_orthologname_))
# remove ebola gene 
sig_cor_lnc <- sig_cor_lnc[sig_cor_lnc$gene != "MSTRG.252489-unknown",]
```


# 1. Collect some information about correlations
```{r InfoCorrelation}
#sig_cor_lnc <- sig_cor_lnc[abs(sig_cor_lnc$rho) > 0.1,]
# remove ebola gene 
sig_cor_lnc <- sig_cor_lnc[sig_cor_lnc$gene != "MSTRG.252489-unknown",]

# 1. How many lncrnas correlated with the viral load do we find? 
length(unique(sig_cor_lnc$gene))

# 2. Which is the directionality of correlation? 
table(sig_cor_lnc$rho>0) # All downregulated 

# 3. Which is the biotype of the genes downregulated? 
table(substr(sig_cor_lnc$gene,1,3))

# 4. How many have an ortholog? 
table(substr(sig_cor_lnc$gene_name,1,3))

unique(sig_cor_lnc$gene)
```

# 2. Collect some information about correlations
```{r InfoCorrelation}
#lapply(sig_cor_lnc$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24))
```

# 2. Collect some information about correlations
```{r InfoCorrelation}
#lapply(sig_cor_lnc$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf))
```




# MAIN FIGURE 1

```{r InfoCorrelation}

#genes_expression <- Reduce("rbind", lapply(sig_cor_lnc$gene, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf)))
#saveRDS(genes_expression,file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/genesexpr.rds"))

genes_expression <-  readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/genesexpr.rds"))

cols <- c('ENSMMUG00000064224-unknown' = 'red',  "ENSMMUG00000056793-unknown" = "orange", "ENSMMUG00000028689-unknown" = "dark green")
highlights <- c('ENSMMUG00000064224-unknown' ,  "ENSMMUG00000056793-unknown", "ENSMMUG00000028689-unknown")
sizes <- c('ENSMMUG00000064224-unknown' = 1,  "ENSMMUG00000056793-unknown" = 1, "ENSMMUG00000028689-unknown" = 1, "other" = 0.2)
# 1. Get the expression of all genes 
# Genes: sig_cor_lnc$gene


# 2. Plot 
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$id <- ifelse(genes_expression$gene %in% highlights, genes_expression$gene, "other")
nlnc = length(unique(genes_expression$gene))
correaltion <-ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene, size = id))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 6, span = 0.4)+
      theme_classic()+
      theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
      scale_color_manual(values = cols)+ggtitle(paste("n = ", nlnc))+
      scale_size_manual(values = sizes)


pdf(file.path(plots, "05/main_corr.pdf"), width = 8.5, height = 5)
correaltion
dev.off()



genes_expression_red <- genes_expression[genes_expression$gene %in% highlights,]
correaltion <-ggplot(genes_expression_red, aes(x = percentage_viral_reads, y = value, col = gene, size = id))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 100, span = 0.4)+
      theme_classic()+
      theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
      scale_color_manual(values = cols)+ggtitle(paste("n = ", nlnc))+
      scale_size_manual(values = sizes)


pdf(file.path(plots, "05/main_corr_red.pdf"), width = 8.5, height = 5)
correaltion
dev.off()

```


# Check HDLBP ex vivo and in vivo 

```{r inVivoimports}
# Compute 
exvivo_hdlbp <- Reduce("rbind", lapply("HDLBP", function(gene) get_expression_summary_gene(gene, mono_live_h24_inf)))
saveRDS(exvivo_hdlbp ,file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/exvivohdlbp.rds"))

exvivo_hdlbp  <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/exvivohdlbp.rds"))


pdf(file.path(plots, "05/main_hdbpl.pdf"), width = 8.5, height = 5)
ggplot(exvivo_hdlbp, aes(x = percentage_viral_reads, y = value, col = gene))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 20, span = 0.4)+
      theme_classic()+
      theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
      scale_color_manual(values = c("grey"))
dev.off()

```



# ------------------------------------------
#       In vivo
# ------------------------------------------
```{r inVivoimports}
immune.combined_invivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
ebola_genome_percentage_df_invivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds")


table(colnames(immune.combined_invivo) == rownames(ebola_genome_percentage_df_invivo))
immune.combined_invivo$viraload <- ebola_genome_percentage_df_invivo$percentage_viral_reads
immune.combined_invivo$infection <- ebola_genome_percentage_df_invivo$classification
immune.combined_invivo$celltype <- Idents(immune.combined_invivo)
immune.combined_invivo$group <- gsub("late", "ulate", immune.combined_invivo$group)
immune.combined_invivo$celltype_group <- paste(immune.combined_invivo$celltype, immune.combined_invivo$group, sep = "_")
mono_invivo <- subset(immune.combined_invivo, ident = "Monocyte")


table(rownames(ebola_genome_percentage_df_invivo) == colnames(immune.combined_invivo))
ebola_genome_percentage_df_invivo$cond <-immune.combined_invivo$group

#mono_invivo <- mono_invivo[,mono_invivo$group %in% c("baseline","late")]
mono_invivo$condition <- paste(mono_invivo$group, mono_invivo$infection)
mono_invivo$condition <- factor(mono_invivo$condition, levels = c("baseline Not Infected", "late Not Infected", "late Infected"))

mono_invivo$labels  <- as.character(mono_invivo$condition)
mono_invivo$labels  <- gsub("baseline Not Infected", "aControl", mono_invivo$labels )
mono_invivo$labels  <- gsub("late Not Infected", "Bystanders", mono_invivo$labels )
mono_invivo$labels  <- gsub("late Infected", "Infected", mono_invivo$labels )
mono_invivo$labels <- factor(mono_invivo$labels, levels = c("aControl", "Bystanders", "Infected"))
mono_invivo$viral_load <- ebola_genome_percentage_df_invivo[colnames(mono_invivo), ]$percentage_viral_reads
mono_invivo$log_viralload <- log(mono_invivo$viral_load)


# Monocyte subtypes
dp_dn_invivo <- as.data.frame(t(mono_invivo[c("CD14", "FCGR3"),]@assays$RNA@data))
dp_dn_invivo$status <- ""
dp_dn_invivo[dp_dn_invivo$CD14 > 1 & dp_dn_invivo$FCGR3 > 1, ]$status <- "DP"
dp_dn_invivo[dp_dn_invivo$CD14 < 1 & dp_dn_invivo$FCGR3 < 1, ]$status <- "DN"
dp_dn_invivo[dp_dn_invivo$CD14 > 1 & dp_dn_invivo$FCGR3 < 1, ]$status <- "CD14+"
dp_dn_invivo[dp_dn_invivo$CD14 < 1 & dp_dn_invivo$FCGR3 > 1, ]$status <- "CD16+"


# Prepare subsets
mono_late_invivo <- mono_invivo[,mono_invivo$group == "ulate"]
mono_late_invivo_infected <- mono_invivo[,mono_invivo$infection == "Infected"]
table(mono_invivo$infection)
mono_late_invivo_infected$viraload
#saveRDS(mono_late_invivo_infected,"/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds")

table(mono_late_invivo_infected$infection)
print("")

```


# Check if we can see also in vivo 

```{r correlations}

# 0. READ in correlations and pick the significant ones
RHOTHRESHOLD = 0
FDRTHRESHOLD = 0.1
correlations_infected_24 <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds"))
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & !is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]
# Add Gene Type
sig_correlations_pc_infected_24$type <- "-"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"
sig_cor_lnc_invivo <- sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$type == "lnc",]
sig_cor_lnc_invivo$gene_name <- unlist(lapply(sig_cor_lnc_invivo$gene, get_orthologname_))

unique(sig_cor_lnc_invivo$gene)

```


# Check overlap 
```{r correlations}
intersect(sig_cor_lnc_invivo$gene, sig_cor_lnc$gene)
```

```{r correlations}

# -------------------------------------
# 3.  IDENTIFY GENE CLUSTER 
# -------------------------------------
plot_region("ENSMMUG00000064224-unknown", range = 500000, palette_plot_percentage = rep("grey",4))
gene_id <- "ENSMMUG00000064224"
range = 1000000
chr <- paste("chr",unique(as.character(seqnames(ref[ref$gene_id == gene_id,]))), sep = "")
# Define region nearby 
from  <- max(0,(min(start(ref[ref$gene_id == gene_id,])) -range))
to <- (max(end(ref[ref$gene_id == gene_id,]))+range)
ref_small<- ref[start(ref) > from  & end(ref) < to   ,]
# GENE CLUSTER 
genes_nearby <- ref_small[seqnames(ref_small) == 12 & ref_small$type == "gene",]$gene_name
#de_mono <- de_all[de_all$celltype == "Monocyte",]
#overlap <- genes_nearby[genes_nearby %in% de_mono$primerid]
cluster_names <- genes_nearby[genes_nearby %in% correlations_infected_24$gene]

# Are PC in this cluster correlated? 
sig_cluster_names <- cluster_names[cluster_names %in% correlations_infected_24$gene]
lapply(overlap_exvivo$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf)+scale_color_manual(values = "grey"))
lapply(overlap_exvivo$gene, function(gene) plot_viral_load_gene(gene, mono_late_invivo_infected, ebola_genome_percentage_df_invivo))





# -------- PLOT REGION 
plot_region("ENSMMUG00000064224-unknown", range = 1200000, palette_plot_percentage = rep("grey",4), getmax = T )

plot_viral_load_gene <- function(gene_id, mono_live_h24_inf){
  df <- get_expression_summary_gene(gene_id, mono =mono_live_h24_inf)
  df_2 <- get_expression_summary_gene(gene_id, mono =mono_live_h24_notinf)
  p1 <- ggplot(df, aes( x = classification, y = value, col = classification, fill = classification))+geom_boxplot(alpha = 1)+theme_paper+scale_y_log10()+ylab("logCP10K")+xlab("")+theme(axis.text.x = element_blank())+theme(legend.position = "")+theme(axis.ticks.x.bottom = element_blank(), plot.margin = unit(c(2.33,-0.9,0.95,0), "cm"))+scale_color_manual(values = c("#848484"))+scale_fill_manual(values = c("grey"))
  # Visualize viral load and gene expression 
  p2 <- ggMarginal(ggplot(df, aes(x = percentage_viral_reads, y = value, col = classification))+geom_point(alpha = 0.5)+theme_minimal()+scale_color_manual(values = c("#848484"))+theme_paper+theme(axis.text.y = element_blank(), legend.position = "")+xlab("viral load")+ylab("")+ggtitle(get_orthologname_(unique(df$gene)))+geom_line(stat = "summary_bin", binwidth = 0.1)+geom_smooth()+scale_x_log10()+scale_y_log10()+theme(plot.margin = unit(c(0,0,0,0), "cm")),type = "histogram", bins = 40)
  ggarrange(p1, p2, widths = c(0.15,2))
}

plots <- lapply(overlap_exvivo$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf))

```





```{r correlations}
m <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/m.rds")
m[overlap_exvivo$gene,]
hm(m= m,row_title = "", features  = overlap_exvivo$gene, first = TRUE, row_names = T)

```


```{r correlations}

gene1 <- overlap_exvivo$gene[2]
gene2 <- overlap_exvivo$gene[3]
gene3 <- overlap_exvivo$gene[1]
# All significantly correlated 
p1 <- calc_correlation_genes(as.vector(mono[gene1, ]@assays$RNA@data), as.vector(mono[gene2, ]@assays$RNA@data))
p2 <- calc_correlation_genes(as.vector(mono[gene2, ]@assays$RNA@data), as.vector(mono[gene3, ]@assays$RNA@data))
p3 <- calc_correlation_genes(as.vector(mono[gene1, ]@assays$RNA@data), as.vector(mono[gene3, ]@assays$RNA@data))

rbind(p1,p2,p3)
```





