---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Imports 

#-----------------------
# Method number 1 
# ----------------------
# Remove viral reads previous to normalization 
# To identify genes whose expression correlates with changes in the amount of viral transcripts in the cell, we only considered monocytes at lage stage of infection ( 24h post infection ex vivo, days post-infection 5-8 in vivo). We computed the spearman correlation coefficient between the viral load (log10) and each geneâ€™s averaged normalized expression (log(CP10K+1)) over a 10-cell sliding window. We only considered cells presenting at least 1 viral read. Normalized expression values of the genes were calculated after removing viral transcripts in order to avoid library size normalization biases. P-values were corrected with Benjamin and Hochberg multiple testing correction

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr); library(Seurat); library(Matrix); library(SingleCellExperiment)
library(stringr); library(rtracklayer); library(RColorBrewer); library(scales)
library(ggthemes); library(org.Mmu.eg.db); library(ggplot2); library(ggvenn)
library(ggExtra); library(Gviz); library(ggpubr)

# Define paths for data
source("../../utils/00_datapaths.R")
source("../../utils/01_lncrna_annotation_utils.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")
source("../../utils/05_utils_corr.R")
# --------------------------------
# Set thresholds 
RHOTHRESHOLD = 0.25
FDRTHRESHOLD<- 0.05

# ---------------------------------------------------
#           Load  references 
# ---------------------------------------------------
ref <- import(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames_addedgenesNovel.gtf"))
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))


```

# ---------------------------------------------------
#          EX VIVO 
# ---------------------------------------------------
```{r LoadExVivo}
#immune.combined_exvivo <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"))
# 0- Read in information files
robjectsdir_stats <- file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/")
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
all_lncrnas <- setdiff(all_lncrnas, paste(unique(ref[seqnames(ref) == "EBOV_Kikwit", ]$gene_id), "-unknown", sep = ""))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))
final_summary <- readRDS(file.path(data_path, "00_Metadata/final_DE_summary.rds"))
mono_live_h24_inf_exvivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved.rds"))
ebola_genome_percentage_df_exvivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))


#-- in vivo 
mono_live_h24_inf_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved.rds"))
# 0 - Load files 
all_lncrnas_invivo <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/", "all_lncrnas.rds"))
all_lncrnas_invivo <- setdiff(all_lncrnas_invivo, paste(unique(ref[seqnames(ref) == "EBOV_Kikwit", ]$gene_id), "-unknown", sep = ""))

annotated_mrnas_invivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/", "annotated_mrnas.rds"))
```

# Load all 4 metrics and check expected behaviour

```{r LoadExVivo}

# Load different results EX VIVO 
Approach_A <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_A.rds"))
Approach_B <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_B.rds"))
Approach_C <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_C.rds"))
Approach_D <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_D.rds"))


# Load different results EX VIVO 
Approach_A_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_A.rds"))
Approach_B_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_B.rds"))
Approach_C_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_C.rds"))
Approach_D_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/Approach_D.rds"))


# Filter 1: How do the test genes look like? 
test_genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
Approach_A[Approach_A$gene %in% test_genes, ]
Approach_A_invivo[Approach_A_invivo$gene %in% test_genes, ]

# B out 
Approach_B[Approach_B$gene %in% test_genes, ]
Approach_B_invivo[Approach_B_invivo$gene %in% test_genes, ]

# C ok 
Approach_C[Approach_C$gene %in% test_genes, ]
Approach_C_invivo[Approach_C_invivo$gene %in% test_genes, ]

# D somehoe fails ( i think to few datapoints )
Approach_D[Approach_D$gene %in% test_genes, ]
Approach_D_invivo[Approach_D_invivo$gene %in% test_genes, ]

# Subsequently exclude approaches: B and D out 

lapply("MX1", function(gene) calc_correlation_viralload_average(expression_matrix, viral_load, as.character(gene), dropzeros = T))

```

# Pick visualization 

```{r LoadExVivo}

genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")

# Ex vivo 
pdf(file.path(plots, "06/EXVIVO_broad_test_all.pdf"), width = 7, height = 6)
visualize_correlation(genes, mono_live_h24_inf_exvivo,ebola_genome_percentage_df_exvivo, keepzeros = T)+ggtitle("EXVIVO")
dev.off()

pdf(file.path(plots, "06/EXVIVO_broad_test_avgwindow.pdf"), width = 7, height = 6)
visualize_correlation_window(genes, mono_live_h24_inf_exvivo, keepzeros = T)
dev.off()

# In vivo 
pdf(file.path(plots, "06/INVIVO_broad_test_all.pdf"), width = 7, height = 6)
visualize_correlation(genes, mono_live_h24_inf_invivo, ebola_genome_percentage_df_invivo, keepzeros = T)+ggtitle("INVIVO")
dev.off()

pdf(file.path(plots, "06/INVIVO_broad_test_avgwindow.pdf"), width = 7, height = 6)
visualize_correlation_window(genes, mono_live_h24_inf_invivo, keepzeros = T)+ggtitle("INVIVO")
dev.off()
```
# My current choice would be method C with first visualization method
```{r LoadExVivo}
# Calculate FDR 
Approach_A$qval <- p.adjust(Approach_A$pval, method = "BH", n = length(Approach_A$pval))
Approach_A_invivo$qval <- p.adjust(Approach_A_invivo$pval, method = "BH", n = length(Approach_A_invivo$pval))
Approach_B$qval <- p.adjust(Approach_B$pval, method = "BH", n = length(Approach_B$pval))
Approach_B_invivo$qval <- p.adjust(Approach_B_invivo$pval, method = "BH", n = length(Approach_B_invivo$pval))
Approach_C$qval <- p.adjust(Approach_C$pval, method = "BH", n = length(Approach_C$pval))
Approach_C_invivo$qval <- p.adjust(Approach_C_invivo$pval, method = "BH", n = length(Approach_C_invivo$pval))
Approach_D$qval <- p.adjust(Approach_D$pval, method = "BH", n = length(Approach_D$pval))
Approach_D_invivo$qval <- p.adjust(Approach_D_invivo$pval, method = "BH", n = length(Approach_D_invivo$pval))


RHOTHRESHOLD <- 0.2
Approach_C_significant_exvivo <- filter_significant_correlations(Approach_C, ref, all_lncrnas, annotated_mrnas, 0.05, RHOTHRESHOLD)[[1]]
Approach_C_significant_invivo <- filter_significant_correlations(Approach_C_invivo, ref, all_lncrnas_invivo, annotated_mrnas_invivo, 0.05, RHOTHRESHOLD)[[1]]

RHOTHRESHOLD_EASY <- 0.1
Approach_A_significant_exvivo <- filter_significant_correlations(Approach_A, ref, all_lncrnas, annotated_mrnas, 0.05, RHOTHRESHOLD_EASY)[[1]]
Approach_A_significant_invivo <- filter_significant_correlations(Approach_A_invivo, ref, all_lncrnas_invivo, annotated_mrnas_invivo, 0.05, RHOTHRESHOLD_EASY)[[1]]
RHOTHRESHOLD_EASY <- 0.03
Approach_B_significant_exvivo <- filter_significant_correlations(Approach_B, ref, all_lncrnas, annotated_mrnas, 0.05, RHOTHRESHOLD_EASY)[[1]]
Approach_B_significant_invivo <- filter_significant_correlations(Approach_B_invivo, ref, all_lncrnas_invivo, annotated_mrnas_invivo, 0.05, RHOTHRESHOLD_EASY)[[1]]



# Check how many lnc and pc 
plot_corr_genes_stat(Approach_A_significant_exvivo, "06/EXVIVO_A_stats.pdf")
plot_corr_genes_stat(Approach_A_significant_invivo, "06/INVIVO_A_stats.pdf")

plot_corr_genes_stat(Approach_B_significant_exvivo, "06/EXVIVO_B_stats.pdf")
plot_corr_genes_stat(Approach_B_significant_invivo, "06/INVIVO_B_stats.pdf")

plot_corr_genes_stat(Approach_C_significant_exvivo, "06/EXVIVO_C_stats.pdf")
plot_corr_genes_stat(Approach_C_significant_invivo, "06/INVIVO_C_stats.pdf")

#plot_corr_genes_stat(Approach_A_significant_exvivo, "06/EXVIVO_A_stats.pdf")
#plot_corr_genes_stat(Approach_A_significant_invivo, "06/INVIVO_A_stats.pdf")

# I am officially going for the C approach cause A has really low coefficients, i do not trust it
plot_corr_genes_stat(Approach_C_significant_exvivo, "06/EXVIVO_C_stats.pdf")
plot_corr_genes_stat(Approach_C_significant_invivo, "06/INVIVO_C_stats.pdf")


# -------------- Visualize the plot with the genes going up or down 


genes <- Approach_C_significant_exvivo[Approach_C_significant_exvivo$type == "lnc" & Approach_C_significant_exvivo$rho > 0,]$gene
pdf(file.path(plots, "06/EXVIVO_lncup_zeros.pdf"), width = 7, height = 6)
visualize_correlation(genes, mono_live_h24_inf_exvivo,ebola_genome_percentage_df_exvivo, keepzeros = T)+ggtitle("EXVIVO")
dev.off()


pdf(file.path(plots, "06/EXVIVO_lncup_NOzeros.pdf"), width = 7, height = 6)
visualize_correlation(genes[1], mono_live_h24_inf_exvivo,ebola_genome_percentage_df_exvivo, keepzeros = F)+ggtitle("EXVIVO")
dev.off()



genes <- Approach_C_significant_exvivo[Approach_C_significant_exvivo$type == "lnc" & Approach_C_significant_exvivo$rho < 0,]$gene
pdf(file.path(plots, "06/EXVIVO_lncdown.pdf"), width = 7, height = 6)
visualize_correlation(genes, mono_live_h24_inf_exvivo,ebola_genome_percentage_df_exvivo, keepzeros = T)+ggtitle("EXVIVO")
dev.off()

genes <- "MSTRG.213929-unknown"
genes_expression <-  Reduce("rbind", lapply(genes, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo, ebola_genome_percentage_df_exvivo,  T )))
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$orth <- unlist(lapply(as.character(genes_expression$gene), get_orthologname_))
ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = orth))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 10, span =0.6)+
    theme_bw()+theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
    xlab("viral load (log10)")+ylab("log(CP10K+1)")+scale_x_log10()





```






# ---------------OLD 

```{r LoadExVivo}
# 1 - Correlations 

# ----------------  Method 1 -----------------
old_correlations_exvivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds"))
old_correlations_exvivo <- old_correlations_exvivo[!is.na(old_correlations_exvivo$pval),]
old_correlations_exvivo$qval <- p.adjust(old_correlations_exvivo$pval, method = "BH", n = length(old_correlations_exvivo$pval))
old_sig_cor_lnc_exvivo <- filter_significant_correlations(old_correlations_exvivo, ref, all_lncrnas, annotated_mrnas, 0.05, 0.05)[[2]]

# ----------------- Method2 --------------------
correlations_exvivo_all <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved_correlation.rds"))
correlations_exvivo <- correlations_exvivo_all[!is.na(correlations_exvivo_all$pval),]
# Adjust p values
correlations_exvivo$qval <- p.adjust(correlations_exvivo$pval, method = "BH", n = length(correlations_exvivo$pval))
correlations_exvivo$correlation <- ifelse( correlations_exvivo$rho > 0, "positive", "negative")
correlations_exvivo <- filter_significant_correlations(correlations_exvivo, ref, all_lncrnas, annotated_mrnas, FDRTHRESHOLD, RHOTHRESHOLD)[[1]]
correlations_exvivo_lnc <- filter_significant_correlations(correlations_exvivo, ref, all_lncrnas, annotated_mrnas, FDRTHRESHOLD, RHOTHRESHOLD)[[2]]
correlations_exvivo_lnc$gene_id <- gsub("-unknown", "", correlations_exvivo_lnc$gene)


# 2 - BROAD CHECK ---- Check that it behaves as it should 
genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
correlations_exvivo[correlations_exvivo$gene %in% genes,]
old_correlations_exvivo[old_correlations_exvivo$gene %in% genes,]
pdf(file.path(plots, "05/EXVIVO_broad_test.pdf"), width = 7, height = 6)
genes_expression <-  Reduce("rbind", lapply(genes, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo, ebola_genome_percentage_df_exvivo, T)))
genes_expression$gene <- as.character(genes_expression$gene)
ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 10, span =0.6)+
      theme_bw()+theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load (log10)")+ylab("log(CP10K+1)")+scale_x_log10()+ggtitle(" Test genes Kotliar et al.")
dev.off()
```

```{r LoadExVivo}
# 3- See how many lncRNA we find and get stats
table(correlations_exvivo$type)
table(correlations_exvivo$direction)
final_summary <- readRDS(file.path(data_path, "00_Metadata/final_correlation_summary.rds"))


# 4 --------- Plot up and down regulation 
mono_live_h24_inf_exvivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved.rds"))
ebola_genome_percentage_df_exvivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))

genes <- correlations_exvivo_lnc[correlations_exvivo_lnc$direction == "negative", ]$gene
lnc_old <- old_sig_cor_lnc_exvivo$gene

# ------------- Plot ------------------------

genes <- old_sig_cor_lnc_exvivo$gene
pdf(file.path(plots, "05/EXVIVO_old.pdf"), width = 7, height = 6)
genes_expression <-  Reduce("rbind", lapply(genes, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo, ebola_genome_percentage_df_exvivo, T)))
genes_expression$gene <- as.character(genes_expression$gene)
ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 10, span =0.6)+
      theme_bw()+theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load (log10)")+ylab("log(CP10K+1)")+scale_x_log10()+ggtitle(" OLD")
dev.off()    



genes <- correlations_exvivo_lnc[correlations_exvivo_lnc$direction == "negative", ]$gene
pdf(file.path(plots, "05/EXVIVO_new_negative.pdf"), width = 7, height = 6)
genes_expression <-  Reduce("rbind", lapply(genes, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo, ebola_genome_percentage_df_exvivo, T)))
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$ortholog <- unlist(lapply(genes_expression$gene, function(gene) get_orthologname_(gene)))
ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = ortholog))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 10, span =0.6)+
      theme_bw()+theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load (log10)")+ylab("log(CP10K+1)")+scale_x_log10()+ggtitle(" New negative")
dev.off()    


genes <- correlations_exvivo_lnc[correlations_exvivo_lnc$direction == "positive", ]$gene
pdf(file.path(plots, "05/EXVIVO_new_positive.pdf"), width = 7, height = 6)
genes_expression <-  Reduce("rbind", lapply(genes, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo, ebola_genome_percentage_df_exvivo, F)))
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$ortholog <- unlist(lapply(genes_expression$gene, function(gene) get_orthologname_(gene)))
ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = ortholog))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 10, span =0.6)+
      theme_bw()+theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load (log10)")+ylab("log(CP10K+1)")+scale_x_log10()+ggtitle(" New positive")
dev.off()   

unique(genes_expression$gene)
g <- "MSTRG.36496-unknown"
get_expression_summary_gene(g, mono_live_h24_inf_exvivo, ebola_genome_percentage_df_exvivo, T)

cellsubset <- mono_live_h24_inf_exvivo
test <- calc_correlation_viralload(expression_matrix, viral_load, g, 10)[[2]]
ggplot(test, aes(x = viral_load, y = avg, col = bin))+geom_line()+
      theme_bw()+theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load (log10)")+ylab("log(CP10K+1)")+ggtitle(" New positive")

```


# Load In Vivo
```{r LoadInVivo}
# ---------------------------------------------------
#   Load IN VIVO 
# ---------------------------------------------------
#mono_live_h24_inf_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds"))
mono_live_h24_inf_invivo <- readRDS( file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved.rds"))

# 0 - Load files 
all_lncrnas_invivo <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/", "all_lncrnas.rds"))
annotated_mrnas_invivo <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/", "annotated_mrnas.rds"))

# 1 - Correlations 
correlations_invivo_all <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/mono_infected_late_ebolaremoved_correlation.rds"))
correlations_invivo <- correlations_invivo_all[!is.na(correlations_invivo_all$pval),]
# Adjust p values
correlations_invivo$qval <- p.adjust(correlations_invivo$pval, method = "BH", n = length(correlations_invivo$pval))
correlations_invivo$direction <- ifelse( correlations_invivo$rho > 0, "positive", "negative")
correlations_invivo <- filter_significant_correlations(correlations_invivo, ref, all_lncrnas, annotated_mrnas, FDRTHRESHOLD, 0.3)[[1]]
correlations_invivo_lnc <- sig_cor_lnc_invivo <-filter_significant_correlations(correlations_invivo, ref, all_lncrnas, annotated_mrnas, FDRTHRESHOLD, RHOTHRESHOLD)[[2]]


# 2 - Check that it behaves as it should 
test_genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1", "ISG15")
correlations_invivo[correlations_invivo$gene %in% test_genes,]


# 3- See how many lnc we find
table(correlations_invivo$type)
table(correlations_invivo$direction)
```





# STATS 
```{r imports}

# 1 ----- How many lnc and pc are found correlated in vivo and ex vivo
print("lnc significantly correlated ex vivo")
length(unique(sig_cor_lnc_exvivo$gene_name))
print("pc significantly correlated ex vivo")
length(unique( correlations_exvivo[correlations_exvivo$type == "pc",]$gene))

# proportion
length(unique(sig_cor_lnc_exvivo$gene_name))*100/length(unique( correlations_exvivo[correlations_exvivo$type == "pc",]$gene))

print("lnc significantly correlated IN vivo")
length(unique(sig_cor_lnc_invivo$gene_name))
print("pc significantly correlated IN vivo")
length(unique( correlations_invivo[correlations_invivo$type == "pc",]$gene))
length(unique(sig_cor_lnc_invivo$gene_name))*100/length(unique( correlations_invivo[correlations_invivo$type == "pc",]$gene))


# 1b: how many novel EX VIVO
table(substr(sig_cor_lnc_exvivo$gene,1,3))
```
```{r imports}
# 2 -----In how many cells is each of the genes expressed
summary_exvivo <- Reduce(rbind, lapply(sig_cor_lnc_exvivo$gene, function(x) get_average(x, mono_live_h24_inf_exvivo)))
summary_invivo <- Reduce(rbind, lapply(sig_cor_lnc_invivo$gene, function(x) get_average(x, mono_live_h24_inf_invivo)))


# 3 --------- How big is the overlap 
x <- list(invivo = summary_invivo$gene, exvivo = summary_exvivo$gene)
ggvenn(
  x, 
  fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#CD534CFF"),
  stroke_size = 0.5, set_name_size = 6
  )

# ----- Are the ones not identified in vivo missing? 
missing_invivo <- setdiff(summary_exvivo$gene,summary_invivo$gene)
summary_in_vivo_missing <- Reduce(rbind, lapply(missing_invivo, function(x) get_average(x, mono_live_h24_inf_invivo)))

summary_invivo$found <- "in vivo - found"
summary_in_vivo_missing$found <- "in vivo - missing"
summary_exvivo$found <- "ex vivo"
s <- rbind(summary_invivo, summary_in_vivo_missing, summary_exvivo)

pdf(file.path(plots, "05/ditribution_cells.pdf"), width = 7, height = 6)
ggplot(s, aes(x = n_cells, fill = found, color = found))+geom_density(alpha = 0.7)+theme_classic()+xlab("number of cells (log10)")+theme(text = element_text(size = 15))+scale_fill_brewer(palette = "Set1")+scale_color_brewer(palette = "Set1")+scale_x_log10()
dev.off()

# Significance tests
wilcox.test(summary_invivo$n_cells, summary_exvivo$n_cells)
wilcox.test(summary_in_vivo_missing$n_cells, summary_exvivo$n_cells)
wilcox.test(summary_in_vivo_missing$n_cells, summary_invivo$n_cells)

# ------ The one not identified ex vivo is there, low p value but not enough to be significant 
missing_exvivo <- setdiff(summary_invivo$gene,summary_exvivo$gene)
summary_ex_vivo_missing <- Reduce(rbind, lapply(missing_exvivo, function(x) get_average(x, mono_live_h24_inf_exvivo)))
correlations_exvivo_all[correlations_exvivo_all$gene == summary_ex_vivo_missing$gene, ]
```


```{r plots}
# --------------------------------
# 2. PLOT MAIN 
#    Examples 
dim_big <- 0.8
#genes_expression <- Reduce("rbind", lapply(sig_cor_lnc_exvivo$gene, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo)))

cols <- c('ENSMMUG00000064224-unknown' = 'red',  "ENSMMUG00000056793-unknown" = "#6C0202", "NDUFA10" = "grey", "HDLBP" = "black")
highlights <- c('ENSMMUG00000064224-unknown' ,  "ENSMMUG00000056793-unknown", "HDLBP", "NDUFA10")
sizes <- c('ENSMMUG00000064224-unknown' = dim_big,  "ENSMMUG00000056793-unknown" = dim_big, "NDUFA10" = dim_big, "HDLBP" = dim_big, "other" = 0.2)
genes_expression <-  Reduce("rbind", lapply(highlights, function(gene) get_expression_summary_gene(gene, mono_live_h24_inf_exvivo)))


# 2. Plot 
pdf(file.path(plots, "05/main_corr.pdf"), width = 9, height = 5)
genes_expression$gene <- as.character(genes_expression$gene)
genes_expression$id <- ifelse(genes_expression$gene %in% highlights, genes_expression$gene, "other")
nlnc = length(unique(genes_expression$gene))
correaltion <-ggplot(genes_expression, aes(x = percentage_viral_reads, y = value, col = gene, size = id))+stat_smooth(method = "loess", formula = y ~ x,  se = F, n = 15, span =0.7)+
      theme_classic()+
      theme( legend.title = element_blank(),plot.title = element_text(size = 12), text = element_text(size = 17))+ 
      xlab("viral load")+ylab("gene expression (logCP10K)")+scale_x_log10()+
      scale_color_manual(values = cols)+ggtitle(paste("n = ", nlnc))+
      scale_size_manual(values = sizes)

correaltion
dev.off()

correaltion
```

# Specific example
```{r plots}
monocytes <- subset(immune.combined, ident = "Monocyte")
baseline <- monocytes[,monocytes$cond == "media"]
late <- monocytes[,monocytes$cond == "live" & monocytes$dpi == "H024"]
infected <- late[,infected_cells]
bystanders <- late[,setdiff(colnames(late), infected_cells)]
# Check 
length(unique(colnames(bystanders))) + length(unique(colnames(infected))) == length(unique(colnames(late)))

gene <- "MX1"

get_expression_subset <- function(s, label, gene){
  s <- subset(s, features = gene)
  df <- data.frame(t(s@assays$RNA@data))
  df$stage <- label
  df$viral_load <- s$viral_load
  df$q <- cut(df$viral_load, breaks = 4)
  df$q <- as.numeric(df$q)
  colnames(df) <- c("exp", "stage", "viral_load","bin")
  return(df)
}
library(ggsci)
df_plotready <- rbind(get_expression_subset(baseline,"baseline",gene), get_expression_subset(bystanders,"bystander",gene), get_expression_subset(infected,"infected",gene)) 
df_plotready <- df_plotready[df_plotready$exp != 0,]

df_plotready$p <- ifelse(df_plotready$stage == "infected", df_plotready$bin
                         ,df_plotready$stage)

ggboxplot(df_plotready, x = "stage", y = "exp", col = "black", fill = "stage",palette = c(pal_jco()(2),rep("red",4)),
          add = "jitter")+theme_classic()+ggtitle(gene)


ggboxplot(df_plotready, x = "p", y = "exp", col = "black", fill = "p",palette = c(pal_jco()(2),rep("red",4)),
          add = "jitter")+theme_classic()+ggtitle(gene)


df_plotready$bin


max(df_plotready$viral_load)
df_plotready[df_plotready$viral_load != 0, ]





```

