---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r imports}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(reshape2)
library(circlize)
library(ggthemes)
library(wesanderson)
library(Gviz)
library(ggpubr)
library(ggExtra)

source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE/")

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.infectionstatus_neut.rds")

ebola_genes <- paste(ebola_ref$gene_id, "-unknown", sep ="")
saveRDS(ebola_genes, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds")
orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds")
# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)

# Prepare metadata
immune.combined$cond <- factor(immune.combined$cond, levels = c("media", "irrad", "live"))
immune.combined$condition <- paste(immune.combined$cond, immune.combined$infection)
immune.combined$celltype <- Idents(immune.combined)
immune.combined$celltype_group <- paste(immune.combined$celltype, immune.combined$cond, sep = "_")

mono <-  subset(immune.combined, ident="Myeloid")
mono <- mono[,mono$cond %in% c("media","live")]
mono$condition <- factor(mono$condition, levels = c("media Not Infected", "live Not Infected", "live Infected"))

mono$labels  <- as.character(mono$condition)
mono$labels  <- gsub("media Not Infected", "Control", mono$labels )
mono$labels  <- gsub("live Not Infected", "Bystanders", mono$labels )
mono$labels  <- gsub("live Infected", "Infected", mono$labels )
mono$labels <- factor(mono$labels, levels = c("Control", "Bystanders", "Infected"))



# Prepare subsets
mono_live <- mono[,mono$cond == "live"]
mono_live_h24 <- mono_live[,mono_live$dpi == "H024"]
mono_live_h24_inf <- mono_live_h24[,mono_live_h24$infection == "Infected" ]
mono_live_h24_notinf <- mono_live_h24[,mono_live_h24$infection == "Not Infected" ]
```


# Visualize EX vivo Dataset
```{r Dimplotoverview}
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(4, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
```

# Determine threshold to call a cell infected
```{r DetermineThreshold}
# Extract normalied expression values matrix
expression_matrix <- immune.combined@assays$RNA@counts
colnames(expression_matrix) <- Idents(immune.combined)


# Per cell, check the % of ebola reads ( not normalized - just calculate the % of raw counts )
ebola_reads <- expression_matrix[rownames(expression_matrix) %in% ebola_genes, ]
ebola_genome_reads <-colSums(ebola_reads)

total_reads <- colSums(expression_matrix)
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- (ebola_genome_reads/total_reads)*100

ebola_genome_percentage_df <- data.frame(percentage_viral_reads = ebola_genome_percentage,
                                         celltype = names(ebola_genome_percentage),
                                         cond =immune.combined$cond,
                                         hour = immune.combined$dpi)


# Define the threshold for the percentage of viral reads identified per cell for it to be called Infected
#threshold <- quantile(ebola_genome_percentage_df[ebola_genome_percentage_df$celltype != "Myeloid", ]$percentage_viral_reads, probs = c(0.99))
threshold <- max(unlist(lapply(setdiff(unique(ebola_genome_percentage_df$celltype), "Myeloid"), function(x) quantile(ebola_genome_percentage_df[ebola_genome_percentage_df$celltype == x, ]$percentage_viral_reads, probs = c(0.99)))))

print(paste("Threshold", threshold))
mono$viral_load <- ebola_genome_percentage_df[colnames(mono), ]$percentage_viral_reads
mono$log_viralload <- log(mono$viral_load)

```

# Discriminate infected vs Bystander cells
```{r CallInfectedCells}

# Separate the cells into infected and not infected based on the threshold
ebola_genome_percentage_df$classification <- ifelse(ebola_genome_percentage_df$percentage_viral_reads >= threshold, "Infected", "Not Infected")
immune.combined$infection <- ebola_genome_percentage_df$classification

saveRDS(ebola_genome_percentage_df, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds")

table(immune.combined[,Idents(immune.combined) == "Myeloid" ]$infection)
table(immune.combined[,Idents(immune.combined) == "B" ]$infection)
table(immune.combined[,Idents(immune.combined) == "T" ]$infection)
table(immune.combined[,Idents(immune.combined) == "NK" ]$infection)

# Save information about infection
#saveRDS(immune.combined, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")

# Add information about how many cells are found per celltype and, per celltype, which is the proportion of cells infected
# How many cells are found per celltype
summary_celltype <- ebola_genome_percentage_df%>% dplyr::group_by(celltype) %>% tally()
ebola_genome_percentage_df$celltype_count <- unlist(apply(ebola_genome_percentage_df,1,function(x) summary_celltype[summary_celltype$celltype ==x[2],]$n))


# Visualize the percentage of infected cells per celltype
percentage_cells_infected_per_celltype <- ebola_genome_percentage_df %>% dplyr::group_by(celltype, classification, celltype_count, cond, hour) %>% tally()

percentage_cells_infected_per_celltype <- percentage_cells_infected_per_celltype[percentage_cells_infected_per_celltype$classification == "Infected", ]

percentage_cells_infected_per_celltype$perc_infected <- (percentage_cells_infected_per_celltype$n / percentage_cells_infected_per_celltype$celltype_count)*100
# reset levels in right order
percentage_cells_infected_per_celltype$cond <- factor(percentage_cells_infected_per_celltype$cond , levels = c("media", "irrad", "live"))

# Just visualize the threshold for calling a cell infectedimmune.combine
ggplot(ebola_genome_percentage_df[ebola_genome_percentage_df$celltype == "Myeloid",], aes( x = percentage_viral_reads , col = classification, fill = classification))+geom_density( alpha = 0.6)+theme_paper+xlab("")+theme(axis.text.y = element_text(size = 8))+theme(axis.text.x = element_text(angle=45, hjust=1), axis.title.y = element_text(size = 15), axis.text.y = element_text(size =15))+scale_x_log10()+scale_fill_manual(values = c( "red", "grey"))+scale_color_manual(values = c( "red", "grey"))+xlab("Viral Load")+geom_vline(xintercept = threshold, lty =  2)

```

# Visualize Infected cells
```{r VisualizeInfectedCells}
percentage_cells_infected_per_celltype$condhr <- paste(percentage_cells_infected_per_celltype$cond, percentage_cells_infected_per_celltype$hour)

ggplot(percentage_cells_infected_per_celltype, aes( x= celltype, y = perc_infected , col = condhr, fill = condhr))+geom_bar(stat="identity",position = position_dodge(), alpha = 0.8)+theme_paper+xlab("")+ylab("% of cells infected")+theme(axis.text.y = element_text(size = 8))+scale_fill_manual(values = wes_palette("Zissou1", n = 3))+scale_color_manual(values = wes_palette("Zissou1", n = 3))+theme(axis.text.x = element_text(angle=45, hjust=1), axis.title.y = element_text(size = 15), axis.text.y = element_text(size =15))+geom_hline(yintercept =1, lty =2)


percentage_cells_infected <- ebola_genome_percentage_df %>% dplyr::group_by(celltype, classification, celltype_count) %>% tally()
percentage_cells_infected$perc_infected <- (percentage_cells_infected$n / percentage_cells_infected$celltype_count)*100
percentage_cells_infected <- percentage_cells_infected[percentage_cells_infected$classification == "Infected", ]

infected_cells <- rownames(ebola_genome_percentage_df[ebola_genome_percentage_df$classification == "Infected",])
# Or with labels (Same plot just with cell-type labels)
DimPlot(immune.combined, label=T, cells.highlight= list(infected_cells), sizes.highlight = 0.2, cols.highlight = c("#DA0202"), cols= "grey", pt.size = 0.1)+
    labs(title = "Infected cells" )+ NoLegend()+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+
    theme(legend.position = "none", plot.title = element_text(size = 22), axis.title = element_text(size = 18))

```

#  DE Analysis results: Bystander vs Infected  
```{r ReadDEInfectionResults}
# General expression of gene
mast_mono <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE_InfectionStatus/Myeloid_MAST_infectedVSnotinfected_LIVE24.rds")
# Corrected q-value .05
FDRTHRESHOLD<- 0.1
# Fold-Change of 30%
FCTHRESHOLD <- 0

de_mono <- mast_mono[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
#de_mono <- de_mono[!(de_mono$primerid == "MSTRG.182795-unknown"),]
de_mono$direction <- ifelse(de_mono$coef < 0 , "down", "up")
de_mono$direction <- factor(de_mono$direction, c("down", "up"))
de_mono_lnc <- de_mono[de_mono$primerid %in% all_lncrnas, ]
de_mono_pc <- de_mono[de_mono$primerid %in% annotated_mrnas, ]

de_mono_lnc_up <- de_mono_lnc[de_mono_lnc$coef> 0,]
de_mono_lnc_down <- de_mono_lnc[de_mono_lnc$coef< 0,]
```

```{r EnrichmentPCDE}
# 1. Check enrichment for DE pc genes
# UP
ego_pc_up <- clusterProfiler::enrichGO(gene = de_mono_pc[de_mono_pc$coef > 0 , ]$primerid,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
clusterProfiler::dotplot(ego_pc_up)
# DOWN
ego_de_down <- clusterProfiler::enrichGO(gene = de_mono_pc[de_mono_pc$coef < 0 , ]$primerid,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
clusterProfiler::dotplot(ego_de_down)

de_stats_direction_plot <- de_mono %>% group_by(primerid) %>% dplyr::summarise(direction = toString((direction)))
ggplot(de_stats_direction_plot, aes(x=direction, fill = direction))+geom_bar(stat="count")+theme_paper+scale_fill_manual(values = c( "#C0E4C6", "#FF6666"))+ylab("Number of DE genes")+xlab("")+theme(legend.position = "")

# 2. Check wether we can also identify some that Broad people can, and compare directionality
down_broad <- c("S100A9", "SELL", "MX1", "VCAN", "MS4A6A")
up_broad <- c("FTL", "C3", "CTSC")
de_mono[de_mono$primerid %in% down_broad,]
de_mono[de_mono$primerid %in% up_broad,]

de_mono_lnc_up <- de_mono_lnc[de_mono_lnc$coef> 0,]
de_mono_lnc_down <- de_mono_lnc[de_mono_lnc$coef< 0,]
```

# Visualize Boxplots
```{r BoxPlotsExpression}
# Visualize boxplot
monocyte <- mono
lapply(de_mono_lnc$primerid, function(gene) boxplot_bystanders(get_expression_summary_gene(gene)))
```

# Visualize Dotplots
```{r DotplotsUPDDOWN}
# Upregulated
plot.data_up <- Dotplot_data(mono, features = c(de_mono_lnc_up$primerid), group.by = "labels",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
heatmap_celltypes(plot.data_up)

# Downregulated
plot.data_down <- Dotplot_data(mono, features = c(de_mono_lnc_down$primerid), group.by = "labels",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
ggplot(plot.data_down, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
# Heatmap
heatmap_celltypes(plot.data_down)


plot.data_down[plot.data_down$name == "NEAT1",]

plot.data_up <- Dotplot_data(mono, features = c(get_gene_id("MIR22HG"), "STAT3"), group.by = "labels",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
heatmap_celltypes(plot.data_up)
```

```{r Heatmapscomplete}
plot.data_down_all <- Dotplot_data(immune.combined, features = c(de_mono_lnc_down$primerid), group.by = "celltype",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down_all$name <- unlist(lapply(plot.data_down_all$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down_all)
```

```{r MonocyteSubtypes}
dp_dn <- as.data.frame(t(mono[c("CD14", "FCGR3"),]@assays$RNA@data))
dp_dn$status <- ""
dp_dn[dp_dn$CD14 > 1 & dp_dn$FCGR3 > 1, ]$status <- "DP"
dp_dn[dp_dn$CD14 < 1 & dp_dn$FCGR3 < 1, ]$status <- "DN"
dp_dn[dp_dn$CD14 > 1 & dp_dn$FCGR3 < 1, ]$status <- "CD14+"
dp_dn[dp_dn$CD14 < 1 & dp_dn$FCGR3 > 1, ]$status <- "CD16+"

mono$monocyte_subset <- dp_dn[colnames(mono),]$status
DimPlot(mono, group.by = "monocyte_subset", cols =wes_palette("IsleofDogs1", 4, type = "discrete"))
saveRDS(dp_dn, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/dp_dn.rds")
```
# ------------------------------------------
#           Gene by gene inspection
# ------------------------------------------
```{r NEAT1}
# Visualize gene Ex Vivo
gene_id <- get_gene_id("NEAT1")

FeaturePlot(mono, feature = "log_viralload", cols =  c("lightgrey", "dark red"), order= T )+theme_sc

summary_gene <- function(gene_id, mono, ebola_genome_percentage_df = ebola_genome_percentage_df){
  # Visual representation

  p2 <- FeaturePlot(mono, feature = gene_id, cols =  c("lightgrey", "dark blue"), order = T)+theme_sc+ggtitle(get_orthologname_(gene_id))
  # General inspections
  p3 <- plot_viral_load_gene(gene_id, mono_live_h24_inf)
  #p4 <- get_prop(gene_id)
  #p5 <- boxplot_bystanders(get_expression_summary_gene(gene_id,mono))
  p6 <- check_expressed_vs_infection(gene_id, mono, ebola_genome_percentage_df)
  return(list(p2,p3,p6))
}

lapply(de_mono_lnc_down$primerid,summary_gene)

check_expressed_vs_infection(gene_id, mono = mono, ebola_genome_percentage_df, dp_dn_invivo)


```

```{r granular}
mono_granular <- FindNeighbors(mono, reduction = "pca",dims = 1:20,force.recalc = T )
mono_granular_c <- FindClusters(mono_granular, resolution = 0.05)

# Visualize novel clusters
DimPlot(mono_granular_c, reduction = "umap", label = TRUE, label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

# Check for batch effects
DimPlot(mono_granular_c, group.by = "individual")
DimPlot(mono_granular_c, group.by = "dpi", cols = c("blue", "red"))

# Identify subtypes
monocytes <- c("CD163", "CFD")
cDC <- c("FLT3", "DQA1")
marker.genes <- c(monocytes, cDC)

FeaturePlot(mono, monocytes)
FeaturePlot(mono, cDC)

mono_granular_c <- RenameIdents(mono_granular_c, `0` = "Monocyte", `1` = "Monocyte", `2` = "cDC")
DotPlot(mono_granular_c,features = marker.genes )
```

# Correlations
```{r correlations}
RHOTHRESHOLD = 0
correlations_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10//05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds")
correlations_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval),]
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]

# Add Gene Type
sig_correlations_pc_infected_24$type <- "-"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"

sig_cor_lnc <- sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$type == "lnc",]
sig_cor_lnc$gene_name <- unlist(lapply(sig_cor_lnc$gene, get_orthologname_))

lapply(sig_cor_lnc$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf))
```

# ------------------------------------------
# ------------------------------------------
# ------------------------------------------
#       In vivo
# ------------------------------------------
```{r inVivoimports}
immune.combined_invivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.infectionstatus_neut.rds")
ebola_genome_percentage_df_invivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds")
mono_invivo <- subset(immune.combined_invivo, ident = "Monocyte")
immune.combined_invivo$celltype <- Idents(immune.combined_invivo)
immune.combined_invivo$group <- gsub("late", "ulate", immune.combined_invivo$group)
immune.combined_invivo$celltype_group <- paste(immune.combined_invivo$celltype, immune.combined_invivo$group, sep = "_")

mono_invivo <- mono_invivo[,mono_invivo$group %in% c("baseline","late")]
mono_invivo$condition <- paste(mono_invivo$group, mono_invivo$infection)
mono_invivo$condition <- factor(mono_invivo$condition, levels = c("baseline Not Infected", "late Not Infected", "late Infected"))

mono_invivo$labels  <- as.character(mono_invivo$condition)
mono_invivo$labels  <- gsub("baseline Not Infected", "aControl", mono_invivo$labels )
mono_invivo$labels  <- gsub("late Not Infected", "Bystanders", mono_invivo$labels )
mono_invivo$labels  <- gsub("late Infected", "Infected", mono_invivo$labels )
mono_invivo$labels <- factor(mono_invivo$labels, levels = c("aControl", "Bystanders", "Infected"))
mono_invivo$viral_load <- ebola_genome_percentage_df_invivo[colnames(mono_invivo), ]$percentage_viral_reads
mono_invivo$log_viralload <- log(mono_invivo$viral_load)


# Monocyte subtypes
dp_dn_invivo <- as.data.frame(t(mono_invivo[c("CD14", "FCGR3"),]@assays$RNA@data))
dp_dn_invivo$status <- ""
dp_dn_invivo[dp_dn_invivo$CD14 > 1 & dp_dn_invivo$FCGR3 > 1, ]$status <- "DP"
dp_dn_invivo[dp_dn_invivo$CD14 < 1 & dp_dn_invivo$FCGR3 < 1, ]$status <- "DN"
dp_dn_invivo[dp_dn_invivo$CD14 > 1 & dp_dn_invivo$FCGR3 < 1, ]$status <- "CD14+"
dp_dn_invivo[dp_dn_invivo$CD14 < 1 & dp_dn_invivo$FCGR3 > 1, ]$status <- "CD16+"


# Prepare subsets
mono_late_invivo <- mono_invivo[,mono_invivo$group == "late"]
table(mono_invivo$labels)
```

# DE infected vs bystanders
```{r inVivoDotplotsInfectedBystanders}
mast_mono <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE_InfectionStatus/Monocyte_MAST_infectedVSnotinfected_LIVE24_DPI58.rds")
# Corrected q-value .05
FDRTHRESHOLD<- 0.5
# Fold-Change of 30%
FCTHRESHOLD <- 0

de_mono <- mast_mono[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
de_mono <- de_mono[!(de_mono$primerid == "MSTRG.182795-unknown"),]
de_mono$direction <- ifelse(de_mono$coef < 0 , "down", "up")
de_mono$direction <- factor(de_mono$direction, c("down", "up"))
de_mono_lnc <- de_mono[de_mono$primerid %in% all_lncrnas, ]
de_mono_pc <- de_mono[de_mono$primerid %in% annotated_mrnas, ]

de_mono_lnc_up <- de_mono_lnc[de_mono_lnc$coef> 0,]
de_mono_lnc_down <- de_mono_lnc[de_mono_lnc$coef< 0,]

unlist(lapply(de_mono_lnc$primerid, get_orthologname_))
```


# Visualize Dotplots
```{r DotplotsUPDDOWN}
# Upregulated
plot.data_up <- Dotplot_data(mono_invivo, features = c(de_mono_lnc_up$primerid), group.by = "labels",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
heatmap_celltypes(plot.data_up)

# Downregulated
plot.data_down <- Dotplot_data(mono_live, features = c(de_mono_lnc_down$primerid), group.by = "labels",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
ggplot(plot.data_down, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
# Heatmap
heatmap_celltypes(plot.data_down)

plot.data_up <- Dotplot_data(mono_live, features = c(get_gene_id("MIR22HG"), "STAT3"), group.by = "labels",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
heatmap_celltypes(plot.data_up)
```
```{r inVivoDotplotsCelltype}
# IN VIVO DOWN across celltypes
plot.data_down <- Dotplot_data(immune.combined_invivo, features = c(de_mono_lnc$primerid), group.by = "celltype_group", scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down)+ggtitle("In Vivo")

# In vivo up across celltypes
plot.data_down <- Dotplot_data(immune.combined_invivo, features = c(de_mono_lnc_up$primerid), group.by = "celltype_group", scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down)+ggtitle("In Vivo")
```


```{r exVivoDotplotsCelltype}
# Need to modify the object and add the clltype gruop

# EX VIVO DOWN across celltypes
plot.data_down <- Dotplot_data(immune.combined, features = c(de_mono_lnc_down$primerid), group.by = "celltype_group", scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down)+ggtitle("In Vivo")

# In vivo up across celltypes
plot.data_down <- Dotplot_data(immune.combined, features = c(de_mono_lnc_up$primerid), group.by = "celltype_group", scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down)+ggtitle("In Vivo")


# Check gene specific summary also in the in vivo settings
summary_gene(gene_id, mono = mono_invivo,ebola_genome_percentage_df_invivo)


check_expressed_vs_infection <- function(gene_id, mono = mono, ebola_genome_percentage_df = ebola_genome_percentage_df, dp_dn = dp_dn){
  # When neat is expressed check infected not infected, status
  gene_expressed <- as.data.frame(t(mono[c(gene_id),]@assays$RNA@data))
  colnames(gene_expressed) <- c("gene")
  gene_expressed$exp <- "Not Expressed"
  gene_expressed[gene_expressed$gene > 1, ]$exp <- "Expressed"
  gene_expressed$infection <- ebola_genome_percentage_df[rownames(gene_expressed),]$classification

  gene_expressed$infection <- factor(gene_expressed$infection, levels =c("Not Infected", "Infected"))

  gene_expressed$status <- dp_dn[rownames(gene_expressed),]$status


  summary_exp <- gene_expressed %>% dplyr::group_by(infection,exp ) %>%dplyr::summarise(n = n())%>% dplyr::mutate(freq = n *100 / sum(n))
  p1 <- ggplot(summary_exp, aes( x = exp, fill = infection, y = n ))+geom_bar(stat = "identity", position = "dodge")+theme_paper+scale_fill_manual(values = c("#F2D994", "#A93E00"))+ylab("%Cells")+xlab("")+ggtitle(get_orthologname_(gene_id))

  summary_exp <- gene_expressed %>% dplyr::group_by(status,exp ) %>%dplyr::summarise(n = n())%>% dplyr::mutate(freq = n *100 / sum(n))
  p2 <- ggplot(summary_exp, aes( x = exp, fill = status, y = n ))+geom_bar(stat = "identity", position = "dodge", alpha = 0.9)+theme_paper+scale_fill_manual(values = wes_palette("Darjeeling1",4))+ylab("%Cells")+xlab("")+ggtitle(get_orthologname_(gene_id))
  return(list(p1,p2))
}

check_expressed_vs_infection(gene_id, mono = mono_late_invivo, ebola_genome_percentage_df_invivo, dp_dn_invivo)


```

```{r correlations}
RHOTHRESHOLD = 0
correlations_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds")
correlations_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval),]
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]

# Add Gene Type
sig_correlations_pc_infected_24$type <- "-"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"

sig_cor_lnc <- sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$type == "lnc",]
sig_cor_lnc$gene_name <- unlist(lapply(sig_cor_lnc$gene, get_orthologname_))

lapply(sig_cor_lnc$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf))
```
