---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r imports}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr); library(Seurat); library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(reshape2)
library(circlize)
library(ggthemes)
library(wesanderson)
library(Gviz)
library(ggpubr)
library(ggExtra)

source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE/")

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")

ebola_genes <- paste(ebola_ref$gene_id, "-unknown", sep ="")
orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds")
# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)

# Prepare metadata
immune.combined$cond <- factor(immune.combined$cond, levels = c("media", "irrad", "live"))
immune.combined$condition <- paste(immune.combined$cond, immune.combined$infection)
immune.combined$celltype <- Idents(immune.combined)
immune.combined$celltype_group <- paste(immune.combined$celltype, immune.combined$cond, sep = "_")

mono <-  subset(immune.combined, ident="Monocyte")
mono <- mono[,mono$cond %in% c("media","live")]
mono$condition <- factor(mono$condition, levels = c("media Not Infected", "live Not Infected", "live Infected"))

mono$labels  <- as.character(mono$condition)
mono$labels  <- gsub("media Not Infected", "Control", mono$labels )
mono$labels  <- gsub("live Not Infected", "Bystanders", mono$labels )
mono$labels  <- gsub("live Infected", "Infected", mono$labels )
mono$labels <- factor(mono$labels, levels = c("Control", "Bystanders", "Infected"))
mono$log_viraload <- log(mono$viraload)



# Prepare subsets
mono_live <- mono[,mono$cond == "live"]
mono_live_h24 <- mono_live[,mono_live$dpi == "H024"]
mono_live_h24_inf <- mono_live_h24[,mono_live_h24$infection == "Infected" ]
mono_live_h24_notinf <- mono_live_h24[,mono_live_h24$infection == "Not Infected" ]
saveRDS(mono_live_h24_inf, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds")
print("")

de_all<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds")
de_lnc <- unique(de_all[de_all$type == "lnc" & de_all$celltype == "Monocyte",]$primerid)
ebola_genome_percentage_df <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds")
```


# Visualize EX vivo Dataset
```{r Dimplotoverview}
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(4, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
```


```{r MonocyteSubtypes}
dp_dn <- as.data.frame(t(mono[c("CD14", "FCGR3"),]@assays$RNA@data))
dp_dn$status <- ""
dp_dn[dp_dn$CD14 > 1 & dp_dn$FCGR3 > 1, ]$status <- "DP"
dp_dn[dp_dn$CD14 < 1 & dp_dn$FCGR3 < 1, ]$status <- "DN"
dp_dn[dp_dn$CD14 > 1 & dp_dn$FCGR3 < 1, ]$status <- "CD14+"
dp_dn[dp_dn$CD14 < 1 & dp_dn$FCGR3 > 1, ]$status <- "CD16+"

mono$monocyte_subset <- dp_dn[colnames(mono),]$status
DimPlot(mono, group.by = "monocyte_subset", cols =wes_palette("IsleofDogs1", 4, type = "discrete"))
saveRDS(dp_dn, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/dp_dn.rds")
```

# I am not finding any DE :( )
#  DE Analysis results: Bystander vs Infected  
```{r ReadDEInfectionResults}
# General expression of gene
mast_mono <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE_InfectionStatus/Monocyte_MAST_infectedVSnotinfected_LIVE24.rds")
# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- 0

de_mono <- mast_mono[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
#de_mono <- de_mono[!(de_mono$primerid == "MSTRG.182795-unknown"),]
de_mono$direction <- ifelse(de_mono$coef < 0 , "down", "up")
de_mono$direction <- factor(de_mono$direction, c("down", "up"))
de_mono_lnc <- de_mono[de_mono$primerid %in% all_lncrnas, ]
de_mono_pc <- de_mono[de_mono$primerid %in% annotated_mrnas, ]

de_mono_lnc_up <- de_mono_lnc[de_mono_lnc$coef> 0,]
de_mono_lnc_down <- de_mono_lnc[de_mono_lnc$coef< 0,]
de_mono_lnc_up
de_mono_lnc_down

monocyte <- mono
```


# Correlation with viral load

```{r correlations}
RHOTHRESHOLD = 0.0
correlations_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds")
correlations_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval),]
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]

# Add Gene Type
sig_correlations_pc_infected_24$type <- "-"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"

sig_cor_lnc <- sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$type == "lnc",]
sig_cor_lnc$gene_name <- unlist(lapply(sig_cor_lnc$gene, get_orthologname_))
```


```{r correlations}
sig_cor_lnc <- sig_cor_lnc[abs(sig_cor_lnc$rho) > 0.1,]
sig_cor_lnc[order(sig_cor_lnc$rho),]

lapply(sig_cor_lnc$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24))


a <- "ENSMMUG00000061044-unknown"
lapply(a, function(gene) plot_viral_load_gene(gene, mono_live_h24))
```


# -----------------------------------------------------------------------------------------------------------------
# -----------------------------------------------------------------------------------------------------------------
```{r EnrichmentPCDE}

# 1. Which lncRNA enriched in infected cells ####
# cell-type specific
#                      Infected              Not infected
# Gene expressed            x11               x12      
# Gene not expressed        x21               x22

get_infection_enrichment <- function(gene, mono, ebola_genome_percentage_df, plot = F, alternative = "greater"){
  print(gene)
  if(!(gene %in% rownames(mono))){
    return()
  }
  gene_expression <- as.data.frame(t(mono[c(gene),]@assays$RNA@data))
  colnames(gene_expression) <- c("geneExp")
  gene_expression$geneExp <- ifelse(gene_expression$geneExp > 0.5, "Expressed", "NotExpressed")
  gene_expression_ready <- merge(ebola_genome_percentage_df, gene_expression, by="row.names")
  df_exp <- gene_expression_ready %>% dplyr::group_by(classification) %>% dplyr::ise(geneExp)
  
  
  summary <- as.data.frame(table(df_exp))
  if(sum(summary[summary$geneExp == "Expressed",]$Freq) < 50){
   return(NA) 
  }else{
      m <- as.matrix(t(table(df_exp)))
      if(alternative == "smaller"){
        m <- m[,c(2,1)]
        alternative <- "greater"
      }
      ft <- fisher.test(m,alternative = alternative)
      df <- data.frame(gene = gene, pval = ft$p.value, or = ft$estimate)
      p <- ggplot(as.data.frame(m), aes(x = geneExp, y = Freq, fill = classification))+geom_bar(stat="Identity", position="dodge")+theme_paper
  
     p_df <- as.data.frame(m)
     p_df <- p_df %>% dplyr::group_by(classification) %>%  dplyr::mutate(prop = Freq / sum(Freq))
     p_df$geneExp <- relevel(p_df$geneExp, "NotExpressed")
     p1 <- ggplot(p_df, aes(x = classification, y = prop, fill = geneExp))+geom_bar(stat="Identity")+theme_paper+coord_flip()+xlab("")+ylab("")+scale_fill_manual(values = c("grey", "navy"))

    if(plot == T){
      return(list(p, p1))
    }else{
      return(df)
    }
  }
}

get_gene_id <- function(gene){
  gene_id <- unique(ref[!is.na(ref$transcript_id) & ref$transcript_id %in% unique(orthologs[orthologs$orthologGeneSymbol == gene,]$lnc),]$gene_name)
  if(length(gene_id) != 0 ){
    gene_id  <- rownames(subset(immune.combined, features = c(gene_id)))
  }else{
    gene_id <- gene
  }
  return(gene_id)
}

# EX VIVO 
ors <- Reduce("rbind", (lapply(de_lnc, function(x) get_infection_enrichment(get_gene_id(x), mono_live_h24, ebola_genome_percentage_df, alternative = "greater"))))
ors$gene_name <- unlist(lapply( ors$gene,get_orthologname_))
ors_sig_exvivo <- ors[!is.na(ors$gene) & ors$pval < FDRTHRESHOLD, ]
ors_sig_exvivo <- ors_sig_exvivo[order(ors_sig_exvivo$or, decreasing = T),]
ors_sig_exvivo$gene_name <- unlist(lapply( ors_sig_exvivo$gene,get_orthologname_))
ors_sig_exvivo

# IN VIVO 
ors_invivo <- Reduce("rbind", (lapply(de_lnc, function(x) get_infection_enrichment(get_gene_id(x), mono_invivo, ebola_genome_percentage_df_invivo))))
ors_invivo_sig <- ors_invivo[!is.na(ors_invivo$gene) & ors_invivo$pval < FDRTHRESHOLD, ]
ors_invivo_sig <- ors_invivo_sig[order(ors_invivo_sig$or, decreasing = T),]
ors_invivo_sig$gene_name <- unlist(lapply( ors_invivo_sig$gene,get_orthologname_))

intersection_inex <- intersect(ors_sig_exvivo[order(ors_sig_exvivo$or, decreasing = F), ][1:20,]$gene,ors_invivo_sig[order(ors_invivo_sig$or, decreasing = F), ][1:20,]$gene)

ors_sig_exvivo[ors_sig_exvivo$gene %in% intersection_inex,]
ors_invivo[ors_invivo$gene %in% intersection_inex,]



gene <-"ENSMMUG00000058644-unknown"
get_infection_enrichment(gene, mono, ebola_genome_percentage_df, plot = T)
get_infection_enrichment(gene, mono_invivo, ebola_genome_percentage_df_invivo, plot = T)

gene <- "ENSMMUG00000058566-unknown"

pall <- function(gene){
  get_infection_enrichment(get_gene_id(gene), mono_invivo, ebola_genome_percentage_df_invivo, plot = T)
  get_infection_enrichment(get_gene_id(gene), mono, ebola_genome_percentage_df, plot = T)
}

lapply(as.character(unique(ors_invivo_sig$gene)), function(x) get_infection_enrichment(get_gene_id(x), mono_invivo, ebola_genome_percentage_df_invivo, plot = T))



lapply(as.character(unique("ENSMMUG00000061044-unknown")), function(x) get_infection_enrichment(get_gene_id(x), mono, ebola_genome_percentage_df, plot = T))

```

# ------------------------------------------
#           Gene by gene inspection
# ------------------------------------------
```{r NEAT1}
# Visualize gene Ex Vivo
gene_id <- get_gene_id("NEAT1")
gene_id <- "MSTRG.181870-unknown"
FeaturePlot(mono_live_h24, feature = "log_viraload", cols =  c("lightgrey", "dark red"), order= T )+theme_sc
summary_gene(gene_id, mono_live_h24,ebola_genome_percentage_df)

summary_gene <- function(gene_id, mono, ebola_genome_percentage_df ){
  # Visual representation

  p2 <- FeaturePlot(mono, feature = gene_id, cols =  c("lightgrey", "dark blue"), order = T)+theme_sc+ggtitle(get_orthologname_(gene_id))
  # General inspections
  p3 <- plot_viral_load_gene(gene_id, mono, ebola_genome_percentage_df)
  #p4 <- get_prop(gene_id)
  #p5 <- boxplot_bystanders(get_expression_summary_gene(gene_id,mono))
  p6 <- check_expressed_vs_infection(gene_id, mono, ebola_genome_percentage_df)
  return(list(p2,p3,p6))
}

plot_region(gene_id, palette_plot_percentage = c(rep("gray", 4)))


```


# ------------------------------------------------------------------------------------------------------------------------------------------------------------------
# ------------------------------------------------------------------------------------------------------------------------------------------------------------------



# ------------------------------------------
#       In vivo
# ------------------------------------------
```{r inVivoimports}
immune.combined_invivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")
ebola_genome_percentage_df_invivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds")


table(colnames(immune.combined_invivo) == rownames(ebola_genome_percentage_df_invivo))
immune.combined_invivo$viraload <- ebola_genome_percentage_df_invivo$percentage_viral_reads
immune.combined_invivo$celltype <- Idents(immune.combined_invivo)
immune.combined_invivo$group <- gsub("late", "ulate", immune.combined_invivo$group)
immune.combined_invivo$celltype_group <- paste(immune.combined_invivo$celltype, immune.combined_invivo$group, sep = "_")
mono_invivo <- subset(immune.combined_invivo, ident = "Monocyte")


table(rownames(ebola_genome_percentage_df_invivo) == colnames(immune.combined_invivo))
ebola_genome_percentage_df_invivo$cond <-immune.combined_invivo$group

#mono_invivo <- mono_invivo[,mono_invivo$group %in% c("baseline","late")]
mono_invivo$condition <- paste(mono_invivo$group, mono_invivo$infection)
mono_invivo$condition <- factor(mono_invivo$condition, levels = c("baseline Not Infected", "late Not Infected", "late Infected"))

mono_invivo$labels  <- as.character(mono_invivo$condition)
mono_invivo$labels  <- gsub("baseline Not Infected", "aControl", mono_invivo$labels )
mono_invivo$labels  <- gsub("late Not Infected", "Bystanders", mono_invivo$labels )
mono_invivo$labels  <- gsub("late Infected", "Infected", mono_invivo$labels )
mono_invivo$labels <- factor(mono_invivo$labels, levels = c("aControl", "Bystanders", "Infected"))
mono_invivo$viral_load <- ebola_genome_percentage_df_invivo[colnames(mono_invivo), ]$percentage_viral_reads
mono_invivo$log_viralload <- log(mono_invivo$viral_load)


# Monocyte subtypes
dp_dn_invivo <- as.data.frame(t(mono_invivo[c("CD14", "FCGR3"),]@assays$RNA@data))
dp_dn_invivo$status <- ""
dp_dn_invivo[dp_dn_invivo$CD14 > 1 & dp_dn_invivo$FCGR3 > 1, ]$status <- "DP"
dp_dn_invivo[dp_dn_invivo$CD14 < 1 & dp_dn_invivo$FCGR3 < 1, ]$status <- "DN"
dp_dn_invivo[dp_dn_invivo$CD14 > 1 & dp_dn_invivo$FCGR3 < 1, ]$status <- "CD14+"
dp_dn_invivo[dp_dn_invivo$CD14 < 1 & dp_dn_invivo$FCGR3 > 1, ]$status <- "CD16+"


# Prepare subsets
mono_late_invivo <- mono_invivo[,mono_invivo$group == "late"]
mono_late_invivo_infected <- mono_invivo[,mono_invivo$infection == "Infected"]
table(mono_invivo$infection)
mono_late_invivo_infected$viraload
saveRDS(mono_late_invivo_infected,"/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_monocytes_infected_late.rds")

table(mono_late_invivo_infected$infection)
print("")
```




```{r exVivoDotplotsCelltype}
# EX VIVO DOWN across celltypes
plot.data_down <- Dotplot_data(mono_late_invivo_infected, features = c("MX1"), group.by = "celltype_group", scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down)+ggtitle("In Vivo")

# In vivo up across celltypes
plot.data_down <- Dotplot_data(immune.combined, features = c(de_mono_lnc_up$primerid), group.by = "celltype_group", scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_down$name <- unlist(lapply(plot.data_down$features.plot, get_orthologname))
heatmap_celltypes(plot.data_down)+ggtitle("In Vivo")


# Check gene specific summary also in the in vivo settings
summary_gene(gene_id, mono = mono_invivo,ebola_genome_percentage_df_invivo)


check_expressed_vs_infection <- function(gene_id, mono = mono, ebola_genome_percentage_df = ebola_genome_percentage_df, dp_dn = dp_dn){
  # When neat is expressed check infected not infected, status
  gene_expressed <- as.data.frame(t(mono[c(gene_id),]@assays$RNA@data))
  colnames(gene_expressed) <- c("gene")
  gene_expressed$exp <- "Not Expressed"
  gene_expressed[gene_expressed$gene > 1, ]$exp <- "Expressed"
  gene_expressed$infection <- ebola_genome_percentage_df[rownames(gene_expressed),]$classification

  gene_expressed$infection <- factor(gene_expressed$infection, levels =c("Not Infected", "Infected"))

  gene_expressed$status <- dp_dn[rownames(gene_expressed),]$status


  summary_exp <- gene_expressed %>% dplyr::group_by(infection,exp ) %>%dplyr::summarise(n = n())%>% dplyr::mutate(freq = n *100 / sum(n))
  p1 <- ggplot(summary_exp, aes( x = exp, fill = infection, y = n ))+geom_bar(stat = "identity", position = "dodge")+theme_paper+scale_fill_manual(values = c("#F2D994", "#A93E00"))+ylab("%Cells")+xlab("")+ggtitle(get_orthologname_(gene_id))

  summary_exp <- gene_expressed %>% dplyr::group_by(status,exp ) %>%dplyr::summarise(n = n())%>% dplyr::mutate(freq = n *100 / sum(n))
  p2 <- ggplot(summary_exp, aes( x = exp, fill = status, y = n ))+geom_bar(stat = "identity", position = "dodge", alpha = 0.9)+theme_paper+scale_fill_manual(values = wes_palette("Darjeeling1",4))+ylab("%Cells")+xlab("")+ggtitle(get_orthologname_(gene_id))
  return(list(p1,p2))
}

check_expressed_vs_infection(gene_id, mono = mono_late_invivo, ebola_genome_percentage_df_invivo, dp_dn_invivo)


```

```{r correlations}

# 0. READ in correlations and pick the significant ones
RHOTHRESHOLD = 0
correlations_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds")
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & !is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]
# Add Gene Type
sig_correlations_pc_infected_24$type <- "-"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"
sig_cor_lnc <- sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$type == "lnc",]
sig_cor_lnc$gene_name <- unlist(lapply(sig_cor_lnc$gene, get_orthologname_))

# 1. Example - check ISG related lncRNAs 
# ISG genes-related 
lapply(c("ENSMMUG00000064224-unknown","ENSMMUG00000062255-unknown"), function(gene) summary_gene(gene,mono_late_invivo_infected,ebola_genome_percentage_df_invivo))
lapply(c("ENSMMUG00000064224-unknown","ENSMMUG00000062255-unknown"), function(gene) summary_gene(gene,mono_live_h24_inf,ebola_genome_percentage_df))

mono
# 2. Visualize all significant ones 
lapply(sig_cor_lnc$gene, function(gene) plot_viral_load_gene(gene, mono_late_invivo, ebola_genome_percentage_df_invivo))


# -------------------------------------
# 3.  IDENTIFY GENE CLUSTER 
# -------------------------------------
plot_region("ENSMMUG00000064224-unknown", range = 500000, palette_plot_percentage = rep("grey",4))
gene_id <- "ENSMMUG00000064224"
range = 1000000
chr <- paste("chr",unique(as.character(seqnames(ref[ref$gene_id == gene_id,]))), sep = "")
# Define region nearby 
from  <- max(0,(min(start(ref[ref$gene_id == gene_id,])) -range))
to <- (max(end(ref[ref$gene_id == gene_id,]))+range)
ref_small<- ref[start(ref) > from  & end(ref) < to   ,]
# GENE CLUSTER 
genes_nearby <- ref_small[seqnames(ref_small) == 12 & ref_small$type == "gene",]$gene_name
#de_mono <- de_all[de_all$celltype == "Monocyte",]
#overlap <- genes_nearby[genes_nearby %in% de_mono$primerid]
cluster_names <- genes_nearby[genes_nearby %in% correlations_infected_24$gene]

# Are PC in this cluster correlated? 
sig_cluster_names <- cluster_names[cluster_names %in% correlations_infected_24$gene]
lapply(overlap_exvivo$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf))
lapply(overlap_exvivo$gene, function(gene) plot_viral_load_gene(gene, mono_late_invivo_infected, ebola_genome_percentage_df_invivo))


# 2 . Checks: do we also see it ex vivo ? 
correlations_infected_24_exvivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds")
correlations_infected_24_exvivo <- correlations_infected_24_exvivo[!is.na(correlations_infected_24_exvivo$pval),]
sig_correlations_pc_infected_24_exvivo <- correlations_infected_24_exvivo[!is.na(correlations_infected_24_exvivo$pval) & correlations_infected_24_exvivo$pval < FDRTHRESHOLD & abs(correlations_infected_24_exvivo$rho) > RHOTHRESHOLD, ]
overlap_exvivo <- sig_correlations_pc_infected_24_exvivo[sig_correlations_pc_infected_24_exvivo$gene %in% cluster_names,]
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% cluster_names,]
# Plot ex vivo 
lapply(overlap_exvivo$gene, function(gene) plot_viral_load_gene(gene, mono_live_h24_inf))

# -------- PLOTS 
# -------- PLOT REGION 
plot_region("ENSMMUG00000064224-unknown", range = 1200000, palette_plot_percentage = rep("grey",4), getmax = T )

lapply(get_gene_id("MIR22HG"), function(gene) summary_gene(gene,mono_live_h24_inf,ebola_genome_percentage_df))

check_expressed_vs_infection(get_gene_id("MIR22HG"), mono_live_h24, ebola_genome_percentage_df_invivo)
check_expressed_vs_infection
```





```{r correlations}
m <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/m.rds")
m[overlap_exvivo$gene,]
hm(m= m,row_title = "", features  = overlap_exvivo$gene, first = TRUE, row_names = T)

```


```{r correlations}

# Plot invivo 
invivoplots <- lapply(overlap_exvivo$gene, function(x) boxplot_bystanders(get_expression_summary_gene(x,mono_invivo, ebola_genome_percentage_df_ = ebola_genome_percentage_df_invivo),  baseline = "baseline", live = "ulate") )
# Plot 
exvivoplots <- lapply(overlap_exvivo$gene, function(x) boxplot_bystanders(get_expression_summary_gene(x,mono, ebola_genome_percentage_df_ = ebola_genome_percentage_df)))
exvivoplots
invivoplots


calc_correlation_genes <- function(gene1_vector,gene2_vector, type = "spearman"){
  mask <- gene1_vector !=0 & gene2_vector !=0
  # calculate the correlation coefficient
  if(sum(mask)>50){
    #pc <- bayes.cor.test(c1, c2)
    pc <- cor.test(gene1_vector[mask], gene2_vector[mask], method = c(type))
    df <- data.frame( pval=pc$p.value, rho=pc$estimate)
    return(df)
  }else{
    return(NA)
  }
}
gene1 <- overlap_exvivo$gene[2]
gene2 <- overlap_exvivo$gene[3]
gene3 <- overlap_exvivo$gene[1]
# All significantly correlated 
calc_correlation_genes(as.vector(mono[gene1, ]@assays$RNA@data), as.vector(mono[gene2, ]@assays$RNA@data))
calc_correlation_genes(as.vector(mono[gene2, ]@assays$RNA@data), as.vector(mono[gene3, ]@assays$RNA@data))
calc_correlation_genes(as.vector(mono[gene1, ]@assays$RNA@data), as.vector(mono[gene3, ]@assays$RNA@data))

```





