---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(enrichplot)
library(ggthemes)


source("../../utils/02_sc_utils.R")
theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE/")

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")

ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unknown", sep ="-")]

orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs.rds")

# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# annotated lncrnas id 
lnc_ref_id <- gsub("_","-",unique(ref$gene_name[ref$gene_biotype %in% c("lncRNA")]))
# Annotated lncRNAs: remove the gene names so it can match 
complete_set_annotated_lncrnas_in_object <- rownames(immune.combined)[rownames(immune.combined) %in% lnc_ref_id]
# Novel lncRNAs
novel_lncrnas <- rownames(immune.combined)[grepl("MST*", rownames(immune.combined))]
# Complete Set 
all_lncrnas <- c(complete_set_annotated_lncrnas_in_object, novel_lncrnas)

annotated_mrnas <- unique(ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "protein_coding", ]$gene_name)
```

```{r colocation}
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(4, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
```

# ----------------------------------------------------------------
#        DE Analysis results: Bystander vs Infected  
# ----------------------------------------------------------------

```{r cars}
# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)

mast_myeloids <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE_noebola/MAST_exvivo_model_infected_vs_notinfected_myeloids.rds")

de_myeloids <- mast_myeloids[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
de_myeloids$direction <- ifelse(de_myeloids$coef < 0 , "down", "up")
de_myeloids$direction <- factor(de_myeloids$direction, c("down", "up"))
de_myeloids_lnc <- de_myeloids[de_myeloids$primerid %in% all_lncrnas, ]

# Check wich lncRNAs are DE
print(de_myeloids_lnc)

# Check enrichment for DE pc genes
de_myeloids_pc <- de_myeloids[de_myeloids$primerid %in% annotated_mrnas, ]
ego_pc_up <- clusterProfiler::enrichGO(gene = de_myeloids_pc[de_myeloids_pc$coef > 0 , ]$primerid,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
dotplot(ego_pc_up)


ego_de_down <- clusterProfiler::enrichGO(gene = de_myeloids_pc[de_myeloids_pc$coef < 0 , ]$primerid,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
dotplot(ego_de_down)




de_stats_direction_plot <- de_myeloids %>% group_by(primerid) %>% dplyr::summarise(direction = toString((direction)))
ggplot(de_stats_direction_plot, aes(x=direction, fill = direction))+geom_bar(stat="count")+theme_paper+scale_fill_manual(values = c( "#C0E4C6", "#FF6666"))+ylab("Number of DE genes")+xlab("")+theme(legend.position = "")
```






# ------------------------------------------
#      MAST CONTINOUS
#--------------------------------------------

# Checking if there is anything - nothing really 
```{r cars}

mast_myeloids_continous <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE_noebola_viralload/MAST_exvivo_model_viralload_myeloids.rds")
mast_myeloids_continous <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE_noebola_viralload/MAST_exvivo_model_viralload_myeloids_LIVE.rds")
mast_myeloids_continous <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/04_DE_noebola_viralload/MAST_exvivo_model_viralload_myeloids_LIVE_24only.rds")

de_myeloids_continous <- mast_myeloids_continous[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
de_myeloids_continous$direction <- ifelse(de_myeloids_continous$coef < 0 , "down", "up")
de_myeloids_continous$direction <- factor(de_myeloids_continous$direction, c("down", "up"))
de_myeloids_lnc_continous <- de_myeloids_continous[de_myeloids_continous$primerid %in% all_lncrnas, ]
print(de_myeloids_lnc_continous)
de_myeloids_pc_continous <- de_myeloids_continous[de_myeloids_continous$primerid %in% annotated_mrnas, ]

```



# --------------------------------------------------------------------------------------------------------------------------------# -------------------------------------------------------------------------------------------------------------------------------# --------------------------------------------------------------------------------------------------------------------------------# 
#                                                          CORRELATION 
# --------------------------------------------------------------------------------------------------------------------------------# --------------------------------------------------------------------------------------------------------------------------------# --------------------------------------------------------------------------------------------------------------------------------# 


# ----------------------------------------------------------------
# Calculate viral load 
# ----------------------------------------------------------------
```{r cars}

# Extract myeloid
myeloids <- subset(immune.combined, idents = "Myeloid")
dim(myeloids)

# Extract normalied expression values matrix
expression_matrix <- myeloids@assays$RNA@counts


# Per cell, check the % of ebola reads ( not normalized - just calculate the % of raw counts )
table(rownames(expression_matrix) %in% ebola_genes)
ebola_reads <- expression_matrix[rownames(expression_matrix) %in% ebola_genes, ]
ebola_genome_reads <-colSums(ebola_reads)

total_reads <- colSums(expression_matrix)
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- (ebola_genome_reads/total_reads)*100

ebola_genome_percentage_df <- data.frame(percentage_viral_reads = ebola_genome_percentage,
                                         cond =myeloids$cond, 
                                         hour = myeloids$dpi, 
                                         condhr = paste(myeloids$cond, myeloids$dpi), 
                                         infection = myeloids$infection)



df <- ebola_genome_percentage_df
#ggplot(df,  aes(percentage_viral_reads+0.01, col = infection , fill = infection )) + 
# geom_histogram(aes(y=..density..), alpha=0.5, 
#                position="identity")+
# geom_density(alpha=.2) +theme_paper+xlab("Viral load")+scale_fill_manual(values =c("red", "grey"))+scale_color_manual(values =c("red", "grey"))+geom_vline(xintercept #=(threshold), size = 0.5, linetype = "dashed")+scale_x_continuous(trans = "log10")



ggplot(df[df$infection == "Infected",],  aes(percentage_viral_reads, col = infection , fill = infection )) + 
 geom_histogram(aes(y=..density..), alpha=0.5, 
                position="identity")+
 geom_density(alpha=.2) +theme_paper+xlab("Viral load")+scale_fill_manual(values =c("red", "grey"))+scale_color_manual(values =c("red", "grey"))+scale_x_continuous(trans = "log10")



```

# ----------------------------------------------------------------
# Correlation with viral load and lncrnas 
# ----------------------------------------------------------------
```{r cars}
immune.combined_noebola <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus_noebola.rds")

myeloids <- subset(immune.combined_noebola, ident = "Myeloid")
myeloids$viral_load <- viral_load <- df$percentage_viral_reads
#saveRDS(myeloids, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/myeloids.rds")

dim(myeloids)
infected_myeloids <- myeloids[,myeloids$infection == "Infected" ]
#saveRDS(infected_myeloids, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/myeloids_infected.rds")
dim(infected_myeloids)

infected_myeloids_24 <- infected_myeloids[,infected_myeloids$dpi == "H024" ]
saveRDS(infected_myeloids_24, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/myeloids_infected_24.rds")
dim(infected_myeloids_24)

lnc <- rownames(myeloids)[rownames(myeloids) %in% all_lncrnas ]
pc <- rownames(myeloids)[rownames(myeloids) %in% annotated_mrnas ]
#saveRDS(lnc, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/lnc.rds")
#saveRDS(pc, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/pc.rds")
```

# Collect all genes 
```{r cars}
FDRTHRESHOLD<- 0.05
RHOTHRESHOLD <- 0.1
correlations_lnc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload/spearman-correlations_lnc.rds")
correlations_lnc_infected <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected/spearman-correlations_lnc.rds")
correlations_lnc_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_24/spearman-correlations_lnc.rds")
sig_correlations_lnc <- correlations_lnc[!is.na(correlations_lnc$pval) & correlations_lnc$pval < FDRTHRESHOLD, ]
sig_correlations_lnc_infected <- correlations_lnc_infected[!is.na(correlations_lnc_infected$pval) & correlations_lnc_infected$pval < FDRTHRESHOLD, ]
sig_correlations_lnc_infected_24 <- correlations_lnc_infected_24[!is.na(correlations_lnc_infected_24$pval) & correlations_lnc_infected_24$pval < FDRTHRESHOLD, ]


correlations_pc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload/spearman-correlations_pc.rds")
correlations_pc_infected <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected/spearman-correlations_pc.rds")
correlations_pc_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_24/spearman-correlations_pc.rds")
sig_correlations_pc <- correlations_pc[!is.na(correlations_pc$pval) & correlations_pc$pval < FDRTHRESHOLD & abs(correlations_pc$rho) > RHOTHRESHOLD, ]
sig_correlations_pc_infected <- correlations_pc_infected[!is.na(correlations_pc_infected$pval) & correlations_pc_infected$pval < FDRTHRESHOLD & abs(correlations_pc$rho) > RHOTHRESHOLD, ]
sig_correlations_pc_infected_24 <- correlations_pc_infected_24[!is.na(correlations_pc_infected_24$pval) & correlations_pc_infected_24$pval < FDRTHRESHOLD & abs(correlations_pc$rho) > RHOTHRESHOLD, ]

sig_correlations_lnc$type <- "lnc"
sig_correlations_lnc_infected$type <- "lnc"
sig_correlations_lnc_infected_24$type <- "lnc"
sig_correlations_lnc$dataset <- "myeloids"
sig_correlations_lnc_infected$dataset <- "myeloids infected"
sig_correlations_lnc_infected_24$dataset <- "myeloids infected 24h"

sig_correlations_pc$type <- "mrna"
sig_correlations_pc_infected$type <- "mrna"
sig_correlations_pc_infected_24$type <- "mrna"
sig_correlations_pc$dataset <- "myeloids"
sig_correlations_pc_infected$dataset <- "myeloids infected"
sig_correlations_pc_infected_24$dataset <- "myeloids infected 24h"
df_all <- rbind(sig_correlations_lnc, sig_correlations_lnc_infected, sig_correlations_pc_infected_24, sig_correlations_pc,sig_correlations_pc_infected, sig_correlations_lnc_infected_24)

ggplot(df_all, aes(x = abs(rho), col = dataset, fill = dataset))+geom_density( alpha =0.7)+theme_paper+ggtitle("Correlation coefficients with viral load")+facet_grid()+scale_fill_calc()+scale_color_calc()

```




```{r cars}
library(org.Hs.eg.db)

expression_matrix <- infected_myeloids_24@assays$RNA@data
viral_load <- infected_myeloids_24$viral_load

# Check how many are correlated and in which direction 
length(unique(sig_correlations_lnc_infected_24$gene))
table(ifelse(sig_correlations_lnc_infected_24$rho > 0, "up", "down"))


# Check how many are correlated and in which direction (Protein coding) 
length(unique(sig_correlations_pc_infected_24$gene))
table(ifelse(sig_correlations_pc_infected_24$rho > 0, "up", "down"))

sig_correlations_pc_infected_24[unlist(lapply(sig_correlations_pc_infected_24$gene, function(x) grepl( "IFI.*", x ))),]
sig_correlations_pc_infected_24["MX1",]

# For the postively regulated ones
egoup <- clusterProfiler::enrichGO(gene = sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$rho > 0, ]$gene, OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(expression_matrix))))
#heatplot(egoup)
dotplot(egoup)

# For the negatively regulated ones
egodown <- clusterProfiler::enrichGO(gene =sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$rho < 0, ]$gene,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(expression_matrix))))
#heatplot(egodown)
dotplot(egodown)



# Check how mnay of the idenified correlated ones have human orthologs
transcripts_de_orthologs <- orthologs[orthologs$lnc %in% unique(ref[ref$gene_id %in% gsub("-unknown", "", sig_correlations_lnc_infected_24$gene), ]$transcript_id),]
transcripts_de_orthologs$gene_id <- unlist(lapply(transcripts_de_orthologs$lnc, function(t) unique(ref[!is.na(ref$transcript_id) &  ref$transcript_id == t]$gene_id)))
length(unique(transcripts_de_orthologs$gene_id))
unique(transcripts_de_orthologs$orthologGeneSymbol)


gene_gas5 <- paste(unique(ref[ref$transcript_id %in% unique(transcripts_de_orthologs[transcripts_de_orthologs$orthologGeneSymbol == "GAS5", ]$lnc),]$gene_id), "-unknown", sep = "")
#plot_corr(expression_matrix, gene_gas5, sig_correlations_lnc_infected_24[sig_correlations_lnc_infected_24$gene == gene_gas5,]$rho)



plot_corr_gene_box <- function(gene, n){
   gene_gas5 <- paste(unique(ref[ref$transcript_id %in% unique(transcripts_de_orthologs[transcripts_de_orthologs$orthologGeneSymbol == gene, ]$lnc),]$gene_id), "-unknown", sep = "")
   print(gene_gas5)
  df <- data.frame(exp = expression_matrix[gene_gas5, ], viral_load)
  df <- df[df$exp !=0, ]
  # Order cells by viral load 
  df_ordered <- df[order(df$viral_load),]
  n<-n
  df_ordered$bin <- cut(1:length(df_ordered$viral_load), breaks =c( seq(0, length(df_ordered$viral_load), n), ceiling(length(df_ordered$viral_load)/n)*n) )
  print(c( seq(0, length(df_ordered$viral_load), n), ceiling(length(df_ordered$viral_load)/n)*n))
  ggplot(df_ordered, aes(x= bin, y = exp))+geom_boxplot(fill = "#5bcfcb", alpha =0.9)+theme_paper+xlab("Viral Load")+ylab("logCP10K")+ggtitle(paste(gene, gene_gas5, sig_correlations_lnc_infected_24[sig_correlations_lnc_infected_24$gene == gene_gas5,]$rho, sep = "  "))
}


plot_corr_gene_box("GAS5", 20)

plot_corr_gene_box("SNHG6", 40)
gene_gas5 <- paste(unique(ref[ref$transcript_id %in% unique(transcripts_de_orthologs[transcripts_de_orthologs$orthologGeneSymbol == "SNHG6", ]$lnc),]$gene_id), "-unknown", sep = "")
plot_corr(expression_matrix, gene_gas5, sig_correlations_lnc_infected_24[sig_correlations_lnc_infected_24$gene == gene_gas5,]$rho)

```


# Visualize each of them 
```{r cars}
# Plot correlation 
plot_corr <- function(expression_matrix, gene, rho){
  df <- data.frame(exp = expression_matrix[gene, ], viral_load)
  df <- df[df$exp !=0, ]
  # Order cells by viral load 
  df_ordered <- df[order(df$viral_load),]
  ggplot(df_ordered, aes(x = viral_load, y = exp))+geom_point( size = 0.6, color = "orange")+geom_line(stat = "summary_bin", binwidth = 0.05,  color = "orange")+scale_x_continuous()+theme_paper+ylim(0,4)+ggtitle(paste(gene,rho ))+geom_smooth()
}

plot_corr_boxplot <- function(expression_matrix, gene, rho){
  df <- data.frame(exp = expression_matrix[gene, ], viral_load)
  df <- df[df$exp !=0, ]
  # Order cells by viral load 
  df_ordered <- df[order(df$viral_load),]
  ggplot(df_ordered, aes(x = viral_load, y = exp))+geom_boxplot( size = 0.6, color = "orange")+geom_line(stat = "summary_bin", binwidth = 0.05,  color = "orange")+scale_x_continuous()+theme_paper+ylim(0,4)+ggtitle(paste(gene,rho ))+geom_smooth()
}



# Plot each of them 
apply(sig_correlations_lnc_infected_24, 1,function(x) plot_corr(expression_matrix, x[3], x[2]))
apply(sig_correlations_lnc_infected_24, 1,function(x) plot_corr_boxplot(expression_matrix, x[3], x[2]))

```
# ----------------------------------------------------------------
# Calculate viral load correlation across windows
# ----------------------------------------------------------------


```{r cars}

corr_window_lnc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_24_window100/spearman-correlations_lnc.rds")
corr_window_pc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/06_correlation/03_viralload_infected_24_window100/spearman-correlations_pc.rds")


corr_window_lnc <- corr_window_lnc[substr(corr_window_lnc$rho,1,5) != "Error",]
sig_corr_window_lnc <- corr_window_lnc[corr_window_lnc$pval < FDRTHRESHOLD, ]

corr_window_pc <- corr_window_pc[substr(corr_window_pc$rho,1,5) != "Error" & !is.na(corr_window_pc$gene),]
sig_corr_window_pc <- corr_window_pc[corr_window_pc$pval < FDRTHRESHOLD, ]
sig_corr_window_pc <- sig_corr_window_pc[!is.na(sig_corr_window_pc$rho),]

apply(sig_corr_window_lnc, 1,function(x) plot_corr(expression_matrix, x[1], x[2]))


transcripts_de_orthologs <- orthologs[orthologs$lnc %in% unique(ref[ref$gene_id %in% gsub("-unknown", "", sig_corr_window_lnc$gene), ]$transcript_id),]
transcripts_de_orthologs$gene_id <- unlist(lapply(transcripts_de_orthologs$lnc, function(t) unique(ref[!is.na(ref$transcript_id) &  ref$transcript_id == t]$gene_id)))
length(unique(transcripts_de_orthologs$gene_id))
unique(transcripts_de_orthologs$orthologGeneSymbol)


plot_corr_gene <- function(gene){
  gene_gas5 <- paste(unique(ref[ref$transcript_id %in% unique(transcripts_de_orthologs[transcripts_de_orthologs$orthologGeneSymbol == gene, ]$lnc),]$gene_id), "-unknown", sep = "")
plot_corr(expression_matrix, gene_gas5, sig_corr_window_lnc[sig_corr_window_lnc$gene == gene_gas5,]$rho)+ggtitle(paste(gene, gene_gas5, sig_corr_window_lnc[sig_corr_window_lnc$gene == gene_gas5,]$rho, sep = "  "))
}

lapply(unique(transcripts_de_orthologs$orthologGeneSymbol), plot_corr_gene)


plot_corr_gene_box <- function(gene){
   gene_gas5 <- paste(unique(ref[ref$transcript_id %in% unique(transcripts_de_orthologs[transcripts_de_orthologs$orthologGeneSymbol == gene, ]$lnc),]$gene_id), "-unknown", sep = "")
  df <- data.frame(exp = expression_matrix[gene_gas5, ], viral_load)
  df <- df[df$exp !=0, ]
  # Order cells by viral load 
  df_ordered <- df[order(df$viral_load),]
  n<-20
  df_ordered$bin <- cut(1:length(df_ordered$viral_load), breaks =c( seq(0, length(df_ordered$viral_load), n), ceiling(length(df_ordered$viral_load)/n)*n) )
  print(c( seq(0, length(df_ordered$viral_load), n), ceiling(length(df_ordered$viral_load)/n)*n))
  ggplot(df_ordered, aes(x= bin, y = exp))+geom_boxplot(fill = "#5bcfcb", alpha =0.9)+theme_paper+xlab("Viral Load")+ylab("logCP10K")+theme(axis.text.x = element_blank())+ggtitle(paste(gene, gene_gas5, sig_corr_window_lnc[sig_corr_window_lnc$gene == gene_gas5,]$rho, sep = "  "))
}


lapply(unique(transcripts_de_orthologs$orthologGeneSymbol), plot_corr_gene_box)

min(viral_load)
max(viral_load)


ego_de_down <- clusterProfiler::enrichGO(gene = gsub("-unknown", "",sig_corr_window_pc$gene),OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
dotplot(ego_de_down)

```





