---
title: "Untitled"
author: "Luisa Santus"
date: "9/21/2020"
output: html_document
---

```{r , include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(seqinr)
library(rtracklayer)
library(Biostrings)

extract_transcript_sequences <- function(transcript_id, fasta, gtf_exons, type, output_file){
    
    transcript_exons <- gtf_exons[gtf_exons$transcript_id == transcript_id, ]
    gene_id <- transcript_exons$gene_id[1]
    strand <- as.vector(strand(transcript_exons))[1]
    chr <- as.vector(seqnames(transcript_exons)[1])
    mygrs <- GRanges(seqnames = chr,
                    ranges=IRanges::reduce(ranges(transcript_exons)),
                    strand=strand) 
    # Create ID 
    pos <- paste(chr, paste(start(mygrs)[1],rev(end(mygrs))[1], sep ="-"), sep = ":")
  
    
    id <- paste(transcript_id, gene_id,strand,pos, type, sep = "_")
    seq <- paste(as.character(getSeq(fasta, mygrs)), collapse = "")
    write.fasta(seq, id, nbchar = 60,output_file, open = "a")
    
}
```


```{r , include=FALSE}
gtf_macaque <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf"))
fasta_macaque <- readDNAStringSet(file.path("/home/luisas/Desktop/cluster/assemblies/ensembl/release-100/rheMac10/Macaca_mulatta.Mmul_10.dna.toplevel.fa"))


gtf_human <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))
fasta_human <- readDNAStringSet(file.path("/home/luisas/Desktop/cluster/assemblies/ensembl/release-100/homo_sapiens/Homo_sapiens.GRCh38.dna.toplevel.fa"))

# Change the names of the chromosomes so that they can match with the one in gtf. Instead of 1 Primary..Etc.. retain only 1 
names(fasta_macaque) <- unlist(lapply(names(fasta_macaque), function(x) strsplit(x, " ")[[1]][1]))
names(fasta_human) <- unlist(lapply(names(fasta_human), function(x) strsplit(x, " ")[[1]][1]))





  
```

```{r cars}

output_file_lncrna_macaque <-  '/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_lncrna.fa'
output_file_mrna_macaque <-  '/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/Macaca_mulatta.Mmul_10.100_known_proteincoding.fa'

output_file_lncrna_human <-  '/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_lncrna.fa'
output_file_mrna_human <-  '/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/Homo_sapiens.GRCh38.100_known_proteincoding.fa'

file.create(output_file_lncrna_macaque)
file.create(output_file_mrna_macaque)

file.create(output_file_lncrna_human)
file.create(output_file_mrna_human)
```

```{r cars}
# --------------------------
#   Prepare files Macaque 
# --------------------------
known_lnc_macaque <- gtf_macaque[gtf_macaque$gene_biotype =="lncRNA", ]
known_lnc_macaque <- known_lnc_macaque[known_lnc_macaque$type == "transcript", ]
  
known_pc_macaque <- gtf_macaque[gtf_macaque$gene_biotype =="protein_coding", ]
known_pc_macaque <- known_pc_macaque[known_pc_macaque$type == "transcript", ]

gtf_exons_macaque <- gtf_macaque[gtf_macaque$type %in% c("exon", "CDS", "start_codon", "stop_codon"),]

lapply(unique(known_pc_macaque$transcript_id), function(x) extract_transcript_sequences(x, fasta_macaque, gtf_exons_macaque, "protein_coding", output_file_mrna_macaque))
lapply(unique(known_lnc_macaque$transcript_id), function(x) extract_transcript_sequences(x,fasta_macaque, "lncRNA", output_file_lncrna_macaque ))

# --------------------------
#   Prepare files Human 
# --------------------------
known_lnc_human <- gtf_human[gtf_human$gene_biotype =="lncRNA", ]
known_lnc_human <- known_lnc_human[known_lnc_human$type == "transcript", ]
  
known_pc_human <- gtf_human[gtf_human$gene_biotype =="protein_coding", ]
known_pc_human <- known_pc_human[known_pc_human$type == "transcript", ]

gtf_exons_human <- gtf_human[gtf_human$type %in% c("exon", "CDS", "start_codon", "stop_codon"),]


lapply(unique(known_pc_human$transcript_id), function(x) extract_transcript_sequences(x, fasta_human, gtf_exons_human, "protein_coding", output_file_mrna_human))
lapply(unique(known_lnc_human$transcript_id), function(x) extract_transcript_sequences(x,fasta_human, "lncRNA", output_file_lncrna_human))


```

