---
title: "01_DE_results_newFilt"
author: "Luisa Santus"
output: html_document
date: "1/22/2020"
---

```{r Imports, message=FALSE,echo=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(dplyr)
library(Seurat)
library(wesanderson)
library(Matrix)
#library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(ggplot2)
library(scales)
library(circlize)
library(reshape2)
library(circlize)

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))
theme_umap <- theme(panel.background = element_rect(fill = "white"),
                    panel.grid.major = element_blank(),
                    legend.position = "", 
                    panel.grid.minor = element_blank(),
                    text = element_text(size=18))


# Imports
#source("../../utils/00_datapaths.R")
#source("../../utils/02_sc_utils.R")
#source("../../utils/03_de_utils.R")
#source("../../utils/04_utils_graph.R")


# Gene annotation

laptop <- "work"
if (laptop == "work"){
  datadir <- "/home/mariasr/cluster/data/"
}else{
datadir <- "/Users/mariasopenar/cluster/02.ebola_sc/data/"
}
robjectsdir <- paste0(datadir, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/")
plots <- paste0(datadir, "plots_new/")
#ref <- import(file.path(datadir, "/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
#ebola_ref <- import(file.path(datadir,"/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

all_de <- readRDS(paste0(robjectsdir, "04_DE/allCells_DE_table.rds"))
all_de$gene_biotype <- gsub("novel_lnc", "lnc", all_de$gene_biotype)
all_lncrnas <- readRDS(file.path(robjectsdir, "/05_stats/all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir,"/05_stats/annotated_mrnas.rds"))

# Check LncPedia
lncpedia_bed <- import(file.path(datadir,"/00_RawData/lncpedia/lncipedia_5_2_hc_hg38.bed"))
lncpedia <- unlist(lapply(lncpedia_bed$name, function(x) str_split(x, ":")[[1]][1]))
lncpedia_immune <- import(file.path(datadir, "/00_RawData/lncpedia/lncpedia_immune.gtf"))
lncpedia_infection <- import(file.path(datadir,"/00_RawData/lncpedia/lncpedia_infection.gtf"))

lncpedia <- unique(c(unique(unlist(lncpedia_immune@elementMetadata[, grepl("gene", colnames(lncpedia_immune@elementMetadata))])), unique(unlist(lncpedia_infection@elementMetadata[, grepl("gene", colnames(lncpedia_infection@elementMetadata))]))))

immlnc <- read.table(file.path(datadir, "00_RawData/ImmLnc/Lnc_Pathways_Sig.txt"), header = TRUE, sep = "\t", stringsAsFactors = FALSE)
reported_immlnc <- immlnc$lncRNA_symbol

orthologs <- readRDS(file.path(datadir, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))

```


## Prepare data  

```{r prepareHeatMap}
de_sig <- all_de[all_de$fdr < 0.05 & abs(all_de$logFC) > 0.1, ]

lnc_sig <- de_sig[de_sig$gene_biotype == "lnc", ]
table(lnc_sig$celltype, lnc_sig$stage)
pc_sig <- de_sig[de_sig$gene_biotype == "pc", ]
table(pc_sig$celltype, pc_sig$stage)

##### lnc ------

#mono 
de_lists_lnc_mono <- list(all_de[all_de$celltype == "Monocyte" & all_de$gene_biotype == "lnc" & all_de$stage == "early", ], 
                        all_de[all_de$celltype == "Monocyte" & all_de$gene_biotype == "lnc" & all_de$stage == "middle", ], 
                        all_de[all_de$celltype == "Monocyte" & all_de$gene_biotype == "lnc" & all_de$stage == "late", ])
names(de_lists_lnc_mono) <- c("lnc_Monocyte_late", "lnc_Monocyte_middle", "lnc_Monocyte_early")


de_lists_lnc_T <- list(all_de[all_de$celltype == "T" & all_de$gene_biotype == "lnc" & all_de$stage == "early", ], 
                        all_de[all_de$celltype == "T" & all_de$gene_biotype == "lnc" & all_de$stage == "middle", ], 
                        all_de[all_de$celltype == "T" & all_de$gene_biotype == "lnc" & all_de$stage == "late", ])
names(de_lists_lnc_T) <- c("lnc_T_late", "lnc_T_middle", "lnc_T_early")
de_lists_lnc_T <- de_lists_lnc_T[unlist(lapply(1:length(de_lists_lnc_T), function(index) length(de_lists_lnc_T[[index]]) != 0))]


de_lists_lnc_B <- list(all_de[all_de$celltype == "B" & all_de$gene_biotype == "lnc" & all_de$stage == "early", ], 
                        all_de[all_de$celltype == "B" & all_de$gene_biotype == "lnc" & all_de$stage == "middle", ], 
                        all_de[all_de$celltype == "B" & all_de$gene_biotype == "lnc" & all_de$stage == "late", ])
names(de_lists_lnc_B) <- c("lnc_B_late", "lnc_B_middle", "lnc_B_early")
de_lists_lnc_B <- de_lists_lnc_B[unlist(lapply(1:length(de_lists_lnc_B), function(index) length(de_lists_lnc_B[[index]]) != 0))]

de_lnc_all <- c(de_lists_lnc_mono, de_lists_lnc_T, de_lists_lnc_B)

```



#-------------------------
# Figure 4 - DE analysis 
#-------------------------

# Figure 4A-C: Heatmaps ###

```{r Heatmap }

#m <- readRDS(paste0(datadir,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/m.rds"))
#palettes
pal_celltypes <-wes_palette("GrandBudapest1", 4)
pal_celltypes[3] <- "#AE4E4E"
pal_celltypes <- pal_celltypes[c(2,1,3,4)]


#T cell ----------
#FC 0.1
pdf(file.path(plots, "Fig4/T_FC_0.1.pdf"), width = 6, height = 12)
FC_heatmap_subset_celltype(all_de, "T", "T cell",  "#A0A0A0",orthologs, lncpedia, reported_immlnc, legend = TRUE, height = 17, fcthreshold = 0.1)
dev.off()

# FC 0.2
pdf(file.path(plots, "03/T_FC_0.2.pdf"), width = 6, height = 12)
FC_heatmap_subset_celltype(all_de, "T", "T cell",  "#A0A0A0",orthologs, lncpedia, reported_immlnc, legend = TRUE, height = 17, fcthreshold = 0.2)
dev.off()

#B cell  ----------
#FC 0.1
pdf(file.path(plots, "03/B_FC_0.1.pdf"), width = 6, height = 12)
FC_heatmap_subset_celltype(all_de, "B", "B cell",  "#A0A0A0",orthologs, lncpedia, reported_immlnc, legend = TRUE, height = 20, fcthreshold = 0.1 )
dev.off()

#FC 0.2 
pdf(file.path(plots, "03/B_FC_0.2.pdf"), width = 6, height = 12)
FC_heatmap_subset_celltype(all_de, "B", "B cell",  "#A0A0A0",orthologs, lncpedia, reported_immlnc, legend = TRUE, height = 20, fcthreshold = 0.2 )
dev.off()

# Monocytes 
# FC 0.1
pdf(file.path(plots, "03/Mono_FC_0.1.pdf"), width = 6, height = 20)
FC_heatmap_subset_celltype(all_de, "Monocyte", "Monocyte",  "#A0A0A0",orthologs, lncpedia, reported_immlnc, legend = TRUE, height = 40, fcthreshold = 0.1 )
dev.off()

#FC 0.2
pdf(file.path(plots, "03/Mono_FC_0.2.pdf"), width = 6, height = 20)
FC_heatmap_subset_celltype(all_de, "Monocyte", "Monocyte",  "#A0A0A0",orthologs, lncpedia, reported_immlnc, legend = TRUE, height = 40, fcthreshold = 0.2 )
dev.off()

```


# Figure 4D: Number of genes DE per cell type (Uspet) ##

```{r Upset}
fc <- 0.1
de_sig <- all_de[all_de$fdr < 0.05 & abs(all_de$logFC) > fc, ]

lnc_sig <- de_sig[de_sig$gene_biotype == "lnc", ]
table(lnc_sig$celltype, lnc_sig$stage)
pc_sig <- de_sig[de_sig$gene_biotype == "pc", ]
table(pc_sig$celltype, pc_sig$stage)

library(UpSetR)
library("gridExtra")
# lnc ---
x = list( unique(lnc_sig[lnc_sig$celltype == "T",]$gene_name), unique(lnc_sig[lnc_sig$celltype == "B",]$gene_name), unique(lnc_sig[lnc_sig$celltype == "Monocyte",]$gene_name))
length(unique(unlist(x)))
names(x)  <-  c( "T " , "B  ","Monocyte" )


#pdf(paste0(plots, "03/4D_upset_de_coloredCelltype_, ",fc, ".pdf"), width = 7, height = 5)
upset_lnc<- upset(fromList(x), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color=pal_celltypes[c(3,1,2)],  main.bar.color=c(rep("black",3), rep("dark grey",4)), sets.x.label = "DE genes", mainbar.y.label = "Number of lncRNA")<- upset(fromList(x), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color=pal_celltypes[c(3,1,2)],  main.bar.color=c(rep("black",3), rep("dark grey",4)), sets.x.label = "DE genes", mainbar.y.label = "Number of lncRNA")
#dev.off()

#pc ---
y = list( unique(pc_sig[pc_sig$celltype == "T",]$gene_name), unique(pc_sig[pc_sig$celltype == "B",]$gene_name), unique(pc_sig[pc_sig$celltype == "Monocyte",]$gene_name))
length(unique(unlist(y)))
names(y)  <-  c( "T " , "B  ","Monocyte" )
#pdf(paste0(plots, "03/4D_PC_upset_de_coloredCelltype_, ",fc, ".pdf"), width = 7, height = 5)
upset_pc <- upset(fromList(y), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color=pal_celltypes[c(3,1,2)],  main.bar.color=c(rep("black",3), rep("dark grey",4)), sets.x.label = "DE genes", mainbar.y.label = "Number of pc")
#dev.off()

par(mfrow = c(1, 1)) 
upset(fromList(x), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color=pal_celltypes[c(3,1,2)],  main.bar.color=c(rep("black",3), rep("dark grey",4)), sets.x.label = "DE genes", mainbar.y.label = "Number of lncDE")
upset(fromList(y), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color=pal_celltypes[c(3,1,2)],  main.bar.color=c(rep("black",3), rep("dark grey",4)), sets.x.label = "DE genes", mainbar.y.label = "Number of genes")

cowplot::plot_grid(upset_pc, upset_lnc)

if (fc == 0.1){
  all_ln <- length(unique(unlist(x)))
  spec_lnc <- 144
  no_spec_lnc <- all_ln - spec_lnc
  all_pc <- length(unique(unlist(y)))
  spec_pc <- 3281
  no_spec_pc <- all_pc - spec_pc
}else{
  all_ln <- length(unique(unlist(x)))
  spec_lnc <- 50 
  no_spec_lnc <- all_ln - spec_lnc
  all_pc <- length(unique(unlist(y)))
  spec_pc <- 1891
  no_spec_pc <- all_pc - spec_pc
  
}
df <- data.frame("specific" = c(spec_lnc, spec_pc), "no_specific" = c(no_spec_lnc, no_spec_pc))
fisher.test(df)
rownames(df) <- c("lnc", "pc")
df


```


# Figure 4E: Specificity lnc DE 


```{r}

#load specificity data 
specificity_scores <- readRDS(file.path(datadir,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/00_specificity/04_specificity_alternativescore.rds"))
specificity_scores$gene_id <- as.character(specificity_scores$gene)
specificity_scores$identity <- as.character(specificity_scores$identity)
specificity_scores$type <- "other"
specificity_scores[specificity_scores$gene_id %in% all_lncrnas,]$type<- "lnc"
specificity_scores[specificity_scores$gene_id %in% annotated_mrnas,]$type<- "pc"
specificity_scores$log <- log(specificity_scores$score)
specificity_scores_lnc <- specificity_scores[specificity_scores$type == "lnc", ]
specificity_scores_pc <- specificity_scores[specificity_scores$type == "pc", ]

# load stats info
df_lnc <- readRDS(file.path(robjectsdir,"/05_stats/df_lnc.rds"))
df_lnc$gene_id <- as.character(df_lnc$gene_id)
df_lnc_celtype <- readRDS(file.path(robjectsdir,"/05_stats/df_celltype_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir,"/05_stats/df_mrna.rds"))
df_mrna$gene_id <- as.character(df_mrna$gene_id)
df_mrna_celltype <- readRDS(file.path(robjectsdir,"/05_stats/df_mrna_celltype.rds"))
df_n_cells <- rbind(df_lnc, df_mrna)
df_celltype <- rbind(df_mrna_celltype,df_lnc_celtype )

df_test <- merge(specificity_scores, df_n_cells, by = "gene_id")
  
# Only keep genes that show expression in at least 30 cells.
df_test <- df_test[df_test$n_cells > 30, ]
df_test <- df_test[df_test$log != "Inf",]
df_test <- df_test[!is.na(df_test$gene_id),]
rownames(df_test) <- df_test$`Row.names`

spec_info_all <- df_test
lnc_specificity_info <- df_test[df_test$type == "lnc",]
lnc_specificity_info$gene_name <- lnc_specificity_info$gene 

pc_specificity_info <- df_test[df_test$type == "pc",]
pc_specificity_info$gene_name <- pc_specificity_info$gene 

df_test$gene_name <- df_test$gene 
summary_de_specificity<- merge(de_sig, df_test, by = "gene_name")
summary_de_specificity$specificity <- ifelse(summary_de_specificity$score > 0.7, "specific", "intermediate")
summary_de_specificity[summary_de_specificity_lnc$score < 0.3,"specificity"] <- "ubiquitous"
spec_vs_de <- summary_de_specificity %>% group_by(celltype, specificity, type) %>% count()
colnames(spec_vs_de) <- c("de", "spec", "type", "n")
spec_vs_de$spec <- factor(spec_vs_de$spec, levels = c("ubiquitous", "intermediate", "specific"))
spec_vs_de$type <- gsub("pc", "protein coding", spec_vs_de$type)
spec_vs_de$type <- gsub("lnc", "lncRNA", spec_vs_de$type)


ggplot(spec_vs_de, aes(x = spec, y = n, fill = de))+geom_bar(stat = "identity")+theme_paper+ scale_fill_manual(values= c("#FD6467","#AE4E4E", "#F1BB7B"))+ labs(y = "# of genes DE", x = "")+theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5, size = 10),  legend.position ="none", )+facet_wrap(~type, scales = "free_y")+  theme(strip.text.x = element_text(size = 18), strip.background = element_blank(), axis.text.y = element_text(size=12), axis.title.y = element_text(size=14))



# lnc ---
summary_de_specificity_lnc <- merge(de_sig, lnc_specificity_info, by = "gene_name")
summary_de_specificity_lnc$specificity <- ifelse(summary_de_specificity_lnc$score > 0.7, "specific", "intermediate")
summary_de_specificity_lnc[summary_de_specificity_lnc$score < 0.3,"specificity"] <- "ubiquitous"
summary_de_specificity_lnc <- summary_de_specificity_lnc[summary_de_specificity_lnc$gene_biotype == "lnc",]
spec_vs_de_lnc <- summary_de_specificity_lnc %>% group_by(celltype, specificity) %>% count()
colnames(spec_vs_de_lnc) <- c("de", "spec", "n")
spec_vs_de_lnc$spec <- factor(spec_vs_de_lnc$spec, levels = c("ubiquitous", "intermediate", "specific"))

# -------------- DE lnc
#pdf(file.path(plots,  "03/DElnc_spec_0.1.pdf"), width = 5, height = 4)
lnc_spec <- ggplot(spec_vs_de_lnc, aes(x = spec, y = n, fill = de))+
       geom_bar(stat = "identity")+
       theme_paper+
       scale_fill_manual(values= c("#FD6467","#AE4E4E", "#F1BB7B"))+ 
       labs(y = "# of DE lncRNAs", x = "")+theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5, size = 10),  legend.position = "none")
#dev.off()



#  protein coding 
summary_de_specificity_pc <- merge(de_sig, pc_specificity_info, by = "gene_name")
summary_de_specificity_pc$specificity <- ifelse(summary_de_specificity_pc$score > 0.7, "specific", "intermediate")
summary_de_specificity_pc[summary_de_specificity_pc$score < 0.3,"specificity"] <- "ubiquitous"
spec_vs_de_pc <- summary_de_specificity_pc %>% group_by(celltype, specificity) %>% count()
colnames(spec_vs_de_pc) <- c("de", "spec", "n")
spec_vs_de_pc$spec <- factor(spec_vs_de_pc$spec, levels = c("ubiquitous", "intermediate", "specific"))



#pdf(file.path(plots,  "03/DEpc_spec_0.1.pdf"), width = 5, height = 4)
pc_spec <- ggplot(spec_vs_de_pc, aes(x = spec, y = n, fill = de))+
       geom_bar(stat = "identity")+
       theme_paper+
       scale_fill_manual(values= c("#FD6467","#AE4E4E", "#F1BB7B"))+ 
       labs(y = "# of DE mRNAs", x = "")+theme(axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5, size = 10),  legend.position = "none")
#dev.off()

plot_grid(lnc_spec, pc_spec)



```


# Figure 4F: Number of genes DE per stage (Uspet) ##

```{r}
fc <- 0.1
de_sig <- all_de[all_de$fdr < 0.05 & abs(all_de$logFC) > fc, ]
lnc_sig <- de_sig[de_sig$gene_biotype == "lnc", ]
table(lnc_sig$celltype, lnc_sig$stage)
pc_sig <- de_sig[de_sig$gene_biotype == "pc", ]
table(pc_sig$celltype, pc_sig$stage)

library(UpSetR)
# stage
x = list( unique(lnc_sig[lnc_sig$stage == "early",]$gene_name), unique(lnc_sig[lnc_sig$stage == "middle",]$gene_name), unique(lnc_sig[lnc_sig$stage == "late",]$gene_name))
length(unique(unlist(x)))
names(x)  <-  c( "early" , "middle","late" )


pdf(paste0(plots, "03/4F_upset_de_stage_.",fc, ".pdf"), width = 7, height = 5)
upset(fromList(x), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color="grey",  main.bar.color=c(rep("black",3), rep("dark grey",4)) ,sets.x.label = "DE genes", mainbar.y.label = "Number of lncRNA")
dev.off()


y =list( unique(pc_sig[lnc_sig$stage == "early",]$gene_name), unique(pc_sig[lnc_sig$stage == "middle",]$gene_name), unique(pc_sig[lnc_sig$stage == "late",]$gene_name))
names(y)  <-  c( "early" , "middle","late" )
pdf(paste0(plots, "03/4D_PC_upset_de_stage_",fc,".pdf.pdf"), width = 7, height = 5)
upset(fromList(y), order.by = "freq",text.scale = c(2.5, 2.5, 1.5, 1.5, 2.5, 2.15),sets.bar.color="grey",  main.bar.color=c(rep("black",3), rep("dark grey",4)) ,sets.x.label = "DE genes", mainbar.y.label = "Number of genes")
dev.off()


#fisher test

if (fc == 0.1){
  all_ln <- length(unique(unlist(x)))
  spec_lnc <- 109
  no_spec_lnc <- all_ln - spec_lnc
  all_pc <- length(unique(unlist(y)))
  spec_pc <- 2554
  no_spec_pc <- all_pc - spec_pc
}else{
  all_ln <- length(unique(unlist(x)))
  spec_lnc <- 41 
  no_spec_lnc <- all_ln - spec_lnc
  all_pc <- length(unique(unlist(y)))
  spec_pc <- 1587
  no_spec_pc <- all_pc - spec_pc
  
}

df <- data.frame("specific" = c(spec_lnc, spec_pc), "all" = c(no_spec_lnc, no_spec_pc))
fisher.test(df)
rownames(df) <- c("lnc", "pc")


```




# ------------------
#   Supplementary
# -----------------

# Stats 

```{r VisualizeDEnumbers}

# FIG S7A: number of lncRNA DE ----

lnc_de_all <- de_sig[de_sig$gene_biotype == "lnc", ]
lnc_de_all$gene_biotype<- ifelse(startsWith(lnc_de_all$gene_name, "MSTR"), "novel", "annotated")

df_de <- lnc_de_all %>% dplyr::group_by(gene_biotype,subtype, celltype) %>% dplyr::summarise(n=n_distinct(gene_id))
df_de$y_pos <- c(15+12,40+23,11+15,10,10,10)
df_de$color <- paste0(df_de$celltype, "_", df_de$subtype)


immune.combined$group <- factor(immune.combined$group, levels = c("baseline", "early", "middle", "late"))
pal_celltypes_short <- pal_celltypes[c(1,3,2)]
pal_celltypes_short_transparency <- c(pal_celltypes_short, alpha(pal_celltypes_short, 0.5))
names(pal_celltypes_short_transparency) <- c( "B_novel", "Monocyte_novel", "T_novel", "B_annotated", "Monocyte_annotated", "T_annotated")

pdf(file.path(plots, "Fig4/SUPPL_DElnc_0.1.pdf"), width = 5, height = 6)
ggplot(df_de , aes(x = celltype,y=n, fill = color))+geom_bar(stat = "identity")+scale_fill_manual(values = pal_celltypes_short_transparency)+scale_color_manual(values=c( "grey", "black"))+theme_paper+theme_classic()+ylab("DE lncRNAs")+xlab("")+theme(legend.position = "")+geom_text(aes(label=n), position = position_stack(vjust = 0.75), color="white", size=6)+theme(axis.ticks.x = element_blank(), text = element_text(size = 20)) +theme(axis.line.y.left  = element_line(color= "black"))
dev.off()




# FIG S7B: number of pc DE ----

mrna_de_all <- de_sig[de_sig$gene_biotype == "pc", ]
pdf(file.path(plots, "Fig4/SUPPL_DEpc_0.1.pdf"), width = 5, height = 6)
ggplot(mrna_de_all %>% dplyr::group_by(subtype, celltype) %>% dplyr::summarise(n=n_distinct(gene_id)), aes(x = celltype,y=n, col = celltype, fill = celltype))+geom_bar(stat = "identity")+scale_fill_manual(values = pal_celltypes_short)+scale_color_manual(values = pal_celltypes_short)+theme_paper+theme_classic()+ylab("DE mRNAs")+xlab("")+theme(legend.position = "")+geom_text(aes(y=n, label=rev(n)),position = position_stack(vjust = 0.5), color="white", size=6, )+theme(axis.ticks.x = element_blank(), text = element_text(size = 20), axis.line.x = element_blank()) +theme(axis.line.y.left  = element_line(color= "black"))
dev.off()

# How many DE lncRNAs
length(unique(lnc_de_all$gene_id))
length(unique(mrna_de_all$gene_id))
```
```{r}

# FIG S7C-D: directionality ----

de_stats_direction_plot <- de_sig %>%  dplyr::group_by(gene_name, celltype,stage,gene_biotype) %>% dplyr::summarise(direction = toString((direction)))
de_stats_direction_plot$stage_short <- factor(toupper(substr(de_stats_direction_plot$stage,1,1)), levels=c("E", "M", "L"))

de_stats_direction_plot$direction <- factor(de_stats_direction_plot$direction, c("down", "up"))
# In general, agnostic of celltype and stage, how many are up and how many are down


#mRNA
pdf(file.path(plots, "Fig4/SUPPL_directionalityPC_0.1.pdf"), width = 5, height = 3.45)
p1<- ggplot(de_stats_direction_plot[de_stats_direction_plot$gene_biotype == "pc",], aes(x = stage_short, fill = direction))+geom_bar(stat="count", position = "dodge")+theme_paper +facet_grid(. ~ celltype)+scale_fill_manual(values = rev(c(alpha("#50C878", 0.8), alpha("#F28C28", 0.8))))+ylab("number of protein coding DE")+xlab("")+theme(strip.text.x = element_text(size = 15), strip.background = element_blank(), axis.title = element_text(size=13), axis.text = element_text(size=15), legend.text = element_text(size=15), panel.grid = element_blank())
p1
dev.off()

# lncRNA 
pdf(file.path(plots, "Fig4/SUPPL_directionalityLNC_0.1.pdf"), width = 5, height = 3.45)
p2 <- ggplot(de_stats_direction_plot[de_stats_direction_plot$gene_biotype == "lnc",],  aes(x = stage_short, fill = direction))+geom_bar(stat="count", position = "dodge")+theme_paper +facet_grid(. ~ celltype)+scale_fill_manual(values = rev(c(alpha("#50C878", 0.8), alpha("#F28C28", 0.8))))+ylab("number of lncRNA DE")+xlab("")+theme(strip.text.x = element_text(size = 15), strip.background = element_blank(), axis.title = element_text(size=13), axis.text = element_text(size=15), legend.text = element_text(size=15), panel.grid = element_blank())
p2
dev.off()


```



```{r}

# FIG S7F: directionality ----
de_lnc <- de_sig[de_sig$gene_biotype == "lnc", ]
de_lnc$ortholog <- unlist(lapply(de_lnc$gene_name, get_orthologname_))
de_lnc$ortholog_found <- ifelse(startsWith(de_lnc$ortholog, "ENS"), FALSE,  TRUE)
de_lnc$ortholog_found <- ifelse(startsWith(de_lnc$ortholog, "MST"), FALSE,  de_lnc$ortholog_found)
de_lnc$ImmLnc <- ifelse(de_lnc$ortholog %in% immlnc$lncRNA_symbol, "present", "absent")
de_lnc$ImmLnc <- factor(de_lnc$ImmLnc, levels = c( "absent","present"))
de_lnc_red <- de_lnc[,c("celltype","ortholog","ortholog_found", "ImmLnc")]                    
de_lnc_red <- unique(de_lnc_red)
# How many DE genes have an ortholog 
de_lnc%>% dplyr::group_by(ortholog_found) %>% dplyr::summarise(n =  dplyr::n()) %>% dplyr::mutate(freq = n /sum(n))
# How many had been prev reported by ImmLnc
de_lnc[de_lnc$ortholog_found == TRUE, ] %>% dplyr::group_by(ImmLnc) %>% dplyr::summarise(n =  dplyr::n()) %>% dplyr::mutate(freq = n /sum(n))


ortholog_summary <-de_lnc_red[de_lnc_red$ortholog_found == TRUE, ] %>% dplyr::group_by( celltype,ortholog_found, ImmLnc) %>% dplyr::summarise(n =  dplyr::n()) %>% dplyr::mutate(freq = n / sum(n))

ortholog_summary$celltype <- gsub("B", "B cell", ortholog_summary$celltype )
ortholog_summary$celltype <- gsub("T", "T cell", ortholog_summary$celltype )
ortholog_summary$celltype <- factor(ortholog_summary$celltype, levels  = c("B cell", "Monocyte","T cell"))
pdf(file.path(plots, "Fig4/SUPPL_orthologs_0.1.pdf"), width = 7, height = 6)
p <- ggplot(ortholog_summary, aes(x = celltype, fill = ImmLnc, y = n))+geom_bar(stat = "identity")+theme_paper+scale_fill_manual(values = rev(c("#000033","grey")))+ylab("# genes")+xlab("")+labs(fill = "in ImmLnc")+theme(legend.title = element_text(size = 20))+theme(text = element_text(colour = "black"), panel.grid = element_blank())
p
dev.off()

```



# Fisher exact test when matched by the proportion of cells in which they express 

```{r Fisher}
library(MatchIt)
robjectsdir<- file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/")
df_lnc <- readRDS(file.path(robjectsdir, "df_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir, "df_mrna.rds"))

df_lnc <- df_lnc[df_lnc$gene_id %in% all_lnc,]
df_mrna <- df_mrna[df_mrna$gene_id %in% annotated_mrnas,]

set.seed(123)
df_lnc$type <- "lnc"
df_mrna$type <- "mrna"
df_complete<- rbind(df_lnc, df_mrna)
c <- df_complete[complete.cases(df_complete),]
c$type = ifelse(c$type == "lnc", 1, 0)


# --------------- MATCH BY PROP CELLS ------
mi_perc <- matchit(type ~ perc_cells_expressing,c)
matches_perc <- get_matches(mi_perc, data = c)

matches_perc$type = ifelse(matches_perc$type == 0, "mrna", "lnc")
n_matches <- length(unique(matches_perc[matches_perc$type == "lnc",]$gene_id))
matches_perc$type <- factor(matches_perc$type, levels = c("lnc", "mrna"))
matches_perc$gene_id <- as.character(matches_perc$gene_id)

# Only retain the matched one for the test 


fc <- 0.1
de_sig <- all_de[all_de$fdr < 0.05 & abs(all_de$logFC) > fc, ]

lnc_sig <- de_sig[de_sig$gene_biotype == "lnc", ]
table(lnc_sig$celltype, lnc_sig$stage)
pc_sig <- de_sig[de_sig$gene_biotype == "pc", ]
table(pc_sig$celltype, pc_sig$stage)

T_lnc <- lnc_sig[lnc_sig$celltype == "T",] ; T_pc <- pc_sig[pc_sig$celltype == "T",]
B_lnc <-  lnc_sig[lnc_sig$celltype == "B",] ; B_pc <- pc_sig[pc_sig$celltype == "B",]
M_lnc <-  lnc_sig[lnc_sig$celltype == "Monocyte",] ; M_pc <- pc_sig[pc_sig$celltype == "Monocyte",]

T_lnc <-  T_lnc[T_lnc$gene_name %in% matches_perc$gene_id,]$gene_name; T_pc <- T_pc[T_pc$gene_name %in% matches_perc$gene_id,]$gene_name;
B_lnc <- B_lnc[B_lnc$gene_name %in% matches_perc$gene_id,]$gene_name; B_pc <- B_pc[B_pc$gene_name %in% matches_perc$gene_id,]$gene_name;
M_lnc <- M_lnc[M_lnc$gene_name %in% matches_perc$gene_id,]$gene_name; M_pc <- M_pc[M_pc$gene_name %in% matches_perc$gene_id,]$gene_name;

all_lnc <- unique(c(T_lnc, B_lnc,M_lnc )); all_pc <- unique(c(T_pc, B_pc,M_pc ));
non_cell_type_specific_lnc <- unique(c(intersect(T_lnc, B_lnc), intersect(B_lnc,M_lnc), intersect(M_lnc, T_lnc)))
cell_type_specific_lnc <- setdiff(all_lnc, non_cell_type_specific_lnc)
non_cell_type_specific_pc <- unique(c(intersect(T_pc, B_pc), intersect(B_pc,M_pc), intersect(M_pc, T_pc)))
cell_type_specific_pc <- setdiff(all_pc, non_cell_type_specific_pc)

# 1. Are DE lncRNA more cell-type or cell-stage specific than DE protein-coding genes ####
# cell-type specific
#           cell-type_specific non_cell-type_specific
# lncRNA            x11               x12      
# PC                x21               x22
x11 <- length(cell_type_specific_lnc)
x21 <- length(cell_type_specific_pc)
x12 <- length(non_cell_type_specific_lnc)
x22 <- length(non_cell_type_specific_pc)
m_type <- matrix(c(x11, x12,x21,x22),2,2,byrow = TRUE)
rownames(m_type) <- c("lncRNA","PC")
colnames(m_type) <- c("type_specific","non_cell-type_specific")
m_type
fisher.test(m_type,alternative = "greater")


```


# same with stage 
```{r}

fc <- 0.1
de_sig <- all_de[all_de$fdr < 0.05 & abs(all_de$logFC) > fc, ]

lnc_sig <- de_sig[de_sig$gene_biotype == "lnc", ]
table(lnc_sig$celltype, lnc_sig$stage)
pc_sig <- de_sig[de_sig$gene_biotype == "pc", ]
table(pc_sig$celltype, pc_sig$stage)

E_lnc <- lnc_sig[lnc_sig$stage == "early",] ; E_pc <- pc_sig[pc_sig$stage == "early",]
mid_lnc <-  lnc_sig[lnc_sig$stage == "middle",] ; mid_pc <- pc_sig[pc_sig$stage == "middle",]
L_lnc <-  lnc_sig[lnc_sig$stage == "late",] ; L_pc <- pc_sig[pc_sig$stage == "late",]

E_lnc <-  E_lnc[E_lnc$gene_name %in% matches_perc$gene_id,]$gene_name; E_pc <- E_pc[E_pc$gene_name %in% matches_perc$gene_id,]$gene_name;
mid_lnc <- mid_lnc[mid_lnc$gene_name %in% matches_perc$gene_id,]$gene_name; mid_pc <- mid_pc[mid_pc$gene_name %in% matches_perc$gene_id,]$gene_name;
L_lnc <- L_lnc[L_lnc$gene_name %in% matches_perc$gene_id,]$gene_name; L_pc <- L_pc[L_pc$gene_name %in% matches_perc$gene_id,]$gene_name;

all_lnc <- unique(c(E_lnc, mid_lnc,L_lnc )); all_pc <- unique(c(E_pc, mid_pc,L_pc ));
non_cell_type_specific_lnc <- unique(c(intersect(E_lnc, mid_lnc), intersect(mid_lnc,L_lnc), intersect(L_lnc, E_lnc)))
cell_type_specific_lnc <- setdiff(all_lnc, non_cell_type_specific_lnc)
non_cell_type_specific_pc <- unique(c(intersect(E_pc, mid_pc), intersect(mid_pc,L_pc), intersect(L_pc, E_pc)))
cell_type_specific_pc <- setdiff(all_pc, non_cell_type_specific_pc)

# 1. Are DE lncRNA more cell-type or cell-stage specific than DE protein-coding genes ####
# cell-type specific
#           stage_specific non_stage_specific
# lncRNA            x11               x12      
# PC                x21               x22
x11 <- length(cell_type_specific_lnc)
x21 <- length(cell_type_specific_pc)
x12 <- length(non_cell_type_specific_lnc)
x22 <- length(non_cell_type_specific_pc)
m_type <- matrix(c(x11, x12,x21,x22),2,2,byrow = TRUE)
rownames(m_type) <- c("lncRNA","PC")
colnames(m_type) <- c("type_specific","non_cell-type_specific")
m_type
fisher.test(m_type,alternative = "greater")

```

