---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)

source("../../utils/02_sc_utils.R")

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjectsOLD/04_DE/")

# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjectsOLD/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))
print("--")

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjectsOLD/03_prep/03_immune.combined.ready_no_neut.rds")
```

```{r colocation}
library(zeallot)

mrnas_ranges <- ref[ref$gene_name %in% annotated_mrnas  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unknown", "",all_lncrnas))]
mrnas_ranges_de<- ref[ref$gene_name %in% unique(unlist(de_mrna_all))  ]


# Add missing gene entries (as novel lncrnas only prsent transcript and exon entries)
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(base::c,gr_list)
lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)

# Only select the DE interestin
lnc_ranges_complete_de <- lnc_ranges_complete[lnc_ranges_complete$gene_id %in% gsub("-", "_",gsub("-unknown", "",unique(unlist(de_lnc_all)))), ]
length(unique(lnc_ranges_complete_de$gene_id))

# --------------------------------------------------------------
#           Get 5 closest genes to each gene of interest
# --------------------------------------------------------------
# Find pairs under maximum overlap
get_all_colocations <- function(gr_gene, gr_hits =mrnas_ranges){
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  distances <- unlist(lapply(1:length(mrnas_ranges), function(index) GenomicRanges::distance(gr_gene, mrnas_ranges[index], ignore.strand = TRUE) ))
  df <- data.frame(gene=gr_gene$gene_id, distances=distances, second = p@second$gene_id, stringsAsFactors = F, chr = seqnames(gr_gene) )
  return(df)
}



get_n_closest_gr <- function(gr_gene, gr_hits =mrnas_ranges, maxgap =2000000L ){
  p <- findOverlapPairs(gr_gene, get_gene_only(gr_hits), maxgap=maxgap, ignore.strand = TRUE)
  
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  if(length(p@second) == 0 ){
    df <- data.frame(gene=gr_gene$gene_id, distances=NA, second = NA, chr = seqnames(gr_gene), stringsAsFactors = F )
  }else{
    distances <- unlist(lapply(1:length(p@second), function(index) GenomicRanges::distance(gr_gene, p@second[index], ignore.strand = TRUE) ))
    df <- data.frame(gene=gr_gene$gene_id, distances=distances, second = p@second$gene_id, stringsAsFactors = F, chr = seqnames(gr_gene) )
  }
  return(df)
}



get_nearest <- function(gr_gene, gr_hits =mrnas_ranges){
  p <- nearest(gr_gene, get_gene_only(gr_hits),ignore.strand = TRUE)
  if(is.na(p)){
    return(NA)
  }
  df <- data.frame(gene = gr_gene$gene_id, second =get_gene_only(gr_hits)[p]$gene_id, distance = GenomicRanges::distance(gr_gene, get_gene_only(gr_hits)[p], ignore.strand = TRUE) )
  return(df)
}
```
# Check where lncRNAs are located - compared to each other 

```{r colocation}
df_distances_de <- do.call(rbind, lapply(1:(length(lnc_ranges_complete_de)), function(index) (get_nearest(lnc_ranges_complete_de[index],gr_hits=lnc_ranges_complete_de[setdiff(1:length(lnc_ranges_complete_de), index)]))))
df_distances_de$type <- ifelse(substr(df_distances_de$gene,1,4)=="MSTR", "Novel", "Annotated")
df_distances_de <- df_distances_de[!is.na(df_distances_de$distance),]
options(scipen=999)
ggplot(df_distances_de, aes(x = distance, fill = type))+geom_histogram(alpha=0.6)+theme_paper+ggtitle("Distance to closest DE lncRNA")+scale_fill_brewer(palette = "Paired")
df_distances_de$bin <- (cut(df_distances_de$distance, breaks = c(-1,20000, 100000,500000,1000000, 2000000,10000000, Inf), labels = c("20kb", "100kb","500kb","1Mb","2Mb", "10Mb",">10Mb" )))


df_lnc_distance_count <- df_distances_de %>% group_by(bin) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100



ggplot(df_lnc_distance_count , aes(bin, y = perc))+geom_bar(stat = "identity", position = position_dodge())+theme_paper



ggplot(df_distances_de, aes(y = distance, x = type,col = type, fill = type))+geom_boxplot( alpha = 0.7)+theme_paper+scale_fill_brewer(palette = "Paired")+scale_color_brewer(palette = "Paired")




```

```{r colocation}
df_distances_de <- do.call(rbind, lapply(1:(length(lnc_ranges_complete_de)), function(index) (get_nearest(lnc_ranges_complete_de[index],gr_hits=mrnas_ranges_de))))
df_distances_de$type <- ifelse(substr(df_distances_de$gene,1,4)=="MSTR", "Novel", "Annotated")
df_distances_de <- df_distances_de[!is.na(df_distances_de$distance),]


options(scipen=999)
ggplot(df_distances_de, aes(x = distance, fill = type))+geom_histogram(alpha=0.6)+theme_paper+ggtitle("Distance to closest DE lncRNA")+scale_fill_brewer(palette = "Paired")
df_distances_de$bin <- (cut(df_distances_de$distance, breaks = c(-1,20000, 100000,500000,1000000, 2000000,10000000, Inf), labels = c("20kb", "100kb","500kb","1Mb","2Mb", "10Mb",">10Mb" )))

df_lnc_distance_count <- df_distances_de %>% group_by(bin) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100

ggplot(df_lnc_distance_count , aes(bin, y = perc))+geom_bar(stat = "identity", position = position_dodge(),fill = "Orange")+theme_paper+xlab("Distance to closest DE mRNA") + ylab("% DE lncRNAs")


df_lnc_distance_count <- df_distances_de %>% group_by(bin, type) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100

ggplot(df_lnc_distance_count , aes(bin, y = perc))+geom_bar(stat = "identity", position = position_dodge(),fill = "Orange")+theme_paper+xlab("Distance to closest DE mRNA") + ylab("% DE lncRNAs")

df_distances_de_novel <- df_distances_de[df_distances_de$type == "Novel",]
df_lnc_distance_count <- df_distances_de_novel %>% group_by(bin, type) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100

df_distances_de_annot <- df_distances_de[df_distances_de$type == "Annotated",]
df_lnc_distance_count_annot <- df_distances_de_annot %>% group_by(bin, type) %>% tally()
df_lnc_distance_count_annot$perc <- (df_lnc_distance_count_annot$n/sum(df_lnc_distance_count_annot$n))*100


ggplot(rbind(df_lnc_distance_count,df_lnc_distance_count_annot ) , aes(bin, y = perc, fill = type))+geom_bar(stat = "identity", position = position_dodge())+theme_paper+xlab("Distance to closest DE mRNA") + ylab("% DE lncRNAs")+scale_fill_brewer(palette = "Paired")

```

```{r colocation}

```



```{r colocation}
# maximum 1Mb
all_lnc_2mb <- do.call(rbind,lapply(1:(length(lnc_ranges_complete_de)), function(index) (get_n_closest_gr(lnc_ranges_complete_de[index], mrnas_ranges_de, maxgap = 2000000L))))
all_lnc_2mb$type <- ifelse(substr(all_lnc_2mb$gene,1,4)=="MSTR", "Novel", "Annotated")

# How many DE pc genes they have in 2Mbp range nearby 
ggplot(all_lnc_2mb %>% group_by(gene, type) %>% tally(), aes(x=n, fill = type ))+geom_histogram()+theme_paper+scale_fill_brewer(palette = "Paired")+ggtitle("How many DE mRNAs found in 2Mb")

# Visualize enrichment of PC genes found in those regions
all_lnc_2mb_names <- ref[ref$gene_id %in% all_lnc_2mb$second, ]$gene_name
logfc <- stats_complete[stats_complete$primerid %in% unlist(unique(all_lnc_2mb_names)),]$coef
library(org.Hs.eg.db)
ego <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(ref$gene_name))
library(clusterProfiler) 
cnetplot(ego, foldChange = 2^logfc, colorEdge = TRUE, node_label = "gene")
dotplot(ego)+theme(axis.text.y = element_text(size = 10))
length(unique(all_lnc_2mb_names))


# Visualize enrichment of PC genes found in those regions close to Novel 
all_lnc_2mb_novel <- all_lnc_2mb[all_lnc_2mb$type == "Novel",]
all_lnc_2mb_names <- ref[ref$gene_id %in% all_lnc_2mb_novel$second, ]$gene_name
ego <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(ref$gene_name))
cnetplot(ego, foldChange = 2^logfc, colorEdge = TRUE, node_label = "gene")
dotplot(ego)+theme(axis.text.y = element_text(size = 10))
length(unique(all_lnc_2mb_names))

# Visualize enrichment of PC genes found in those regions close to Annotated 
all_lnc_2mb_novel <- all_lnc_2mb[all_lnc_2mb$type == "Annotated",]
all_lnc_2mb_names <- ref[ref$gene_id %in% all_lnc_2mb_novel$second, ]$gene_name
logfc <- stats_complete[stats_complete$primerid %in% unlist(unique(all_lnc_2mb_names)),]$coef
ego <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(ref$gene_name))
cnetplot(ego, foldChange = 2^logfc, colorEdge = TRUE, node_label = "gene")
dotplot(ego)+theme(axis.text.y = element_text(size = 10))
length(unique(all_lnc_2mb_names))




#
all_lnc_2mb <- all_lnc_2mb[!is.na(all_lnc_2mb$distances),]
all_lnc_2mb$bin <- (cut(all_lnc_2mb$distances, breaks = c(-1,20000, 100000,500000,1000000, 2000000,10000000, Inf), labels = c("20kb", "100kb","500kb","1Mb","2Mb", "10Mb",">10Mb" )))
all_lnc_2mb <- all_lnc_2mb[!is.na(all_lnc_2mb$bin), ]
ggplot(all_lnc_2mb %>% group_by(gene, type, bin) %>% tally(), aes(y= n, col = type, fill = type, x = bin))+geom_boxplot(alpha = 0.5)+theme_paper+ scale_fill_brewer(palette = "Paired")+scale_color_brewer(palette = "Paired")+xlab("Distance from DE lnc allowed")+ylab("#DE mRNAs found")

```


# Correlations

```{r colocation}
correlations <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjectsOLD/06_correlation/pearson_correlations.rds")
correlations[correlations$pval < 0.05,]

pairs <- (expand.grid.unique(1:10, 1:10))
length(unique(pairs))

```


```{r colocation}




FDRTHRESHOLD = 0.1; FCTHRESHOLD = log(1.2)
stats_complete <- stats_complete[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
stats_complete$celltype <- unlist(lapply(stats_complete$comparison, function(x) str_split(x, "_")[[1]][1]))
stats_complete$stage <- unlist(lapply(stats_complete$comparison, function(x) str_split(x, "_")[[1]][2]))


lnc_stats <- stats_complete[stats_complete$primerid %in% unique(lnc_ranges_complete_de$gene_name), ]
pc_stats <- stats_complete[stats_complete$primerid %in% unique(mrnas_ranges_de$gene_name), ]

table(Idents(immune.combined))

myeloids <- subset(immune.combined, Idents = "Myeloid" ) 
count_matrix <- myeloids@assays$RNA@counts



calc_correlation <- function(pair, count_matrix, type = "pearson"){
  gene1 <- pair$first
  gene2 <- pair$second
  print(gene1)
  print(gene2)
   mask <- count_matrix[gene1,]!=0 & count_matrix[gene2,]!=0
    c1 <- count_matrix[gene1,mask]
    c2 <- count_matrix[gene2,mask]
  
  # calculate the correlation coefficient
  if(sum(mask)>50){ 
    #pc <- bayes.cor.test(c1, c2) 
    pc <- cor.test(c1, c2, method = c(type)) 
    return(pc)
  }else{ 
    return(NA)
  }
}
# Calc correlation values
length(unique(unlist(de_lists_mrna_my)))
pairs <- expand.grid(unlist(de_lists_lnc_my), unique(unlist(de_lists_mrna_my)), stringsAsFactors = F)
colnames(pairs) <- c("first", "second")


correlations <- apply(pairs, 1, function(x) calc_correlation(x, count_matrix, type = "spearman"))


mask <- !is.na(correlations)
correlations <- correlations[mask]
pvals <- (lapply(correlations, function(correlation) print(correlation$p.value)))
rhos <- as.vector(lapply(correlations, function(correlation) print(correlation$estimate)))


df <- cbind(pairs[mask,], pval = unlist(pvals), rho = unlist(rhos))
ggplot(df, aes( x =abs(rho)))+geom_density()+theme_paper+xlab("Correlation Coefficient")
```



# -------------------------------------------
# -------------------------------------------
# Analysis colocation in genomic regions

```{r colocation}
# 1) For each gene lncrna find neighbour protein coding genes. ( either overlapping or in close region) --> putative pairs
# 2) Filter: Check the pearson correlation of the identified pairs, reatain the ones w/ |pearson correlation coefficient| > 0.6
# ==> Only if something nice comes out 
# 3) Validation: Simulate random pairs 
# 4) Validation: Compare the correlation of the ones identified with the random ones. 
library(zeallot)
# --------------------------------------------------------------
#           Get 5 closest genes to each gene of interest
# --------------------------------------------------------------
# Find pairs under maximum overlap
get_n_closest_gr <- function(gr_gene, gr_hits =mrnas_ranges , n = 5, maxgap =1000000L ){
  p <- findOverlapPairs(gr_gene, get_gene_only(gr_hits), maxgap=maxgap, ignore.strand = TRUE)
  pairs_list <- zipup(p)
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  distances <- unlist(lapply(pairs_list, function(pair) GenomicRanges::distance(pair[1], pair[2], ignore.strand = TRUE) ))
  if(!is.null(distances)){
    n <- min(length(pairs_list), n)
    closest_5_pairs <-  pairs_list[order(distances)][1:n]
  }else{
    return(NULL)
  }
  return(closest_5_pairs)
}


get_closest_genes <- function(genes, original_range, hit_range, n = 10, maxgap = 100000L ){
  original_range <- original_range[original_range$gene_id %in% gsub("-", "_",gsub("-unkown", "",genes)), ]
  ranges_5_cl_genes <- lapply(1:(length(original_range)), function(index) unlist(get_n_closest_gr(original_range[index], hit_range, n = n, maxgap = maxgap))[c(rep(FALSE,1), TRUE)]$gene_name)
return(ranges_5_cl_genes)
}


# prepare ranges
mrnas_ranges <- ref[ref$gene_name %in% mrna_ref_id  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unkown", "",all_lncrnas))]
# add gene entries for novel lncrna, which only present transcripts and exons
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) (get_gene_range(x, lnc_ranges)))
missing_gene_ranges <- do.call(base::c,gr_list)
lnc_ranges_genes <-c(get_gene_only(lnc_ranges), missing_gene_ranges)


# Group DE lnc 
lnc_myeloid_late <- get_de_names(mast_myeloid_late, subset = all_lncrnas, FCTHRESHOLD = log(1.0))
lnc_myeloid_middle <- get_de_names(mast_myeloid_middle, subset = all_lncrnas, FCTHRESHOLD = log(1.0))
lnc_myeloid_early <- get_de_names(mast_myeloid_early, subset = all_lncrnas, FCTHRESHOLD = log(1.0))
de_myeloid <- c(lnc_myeloid_late, lnc_myeloid_middle, lnc_myeloid_early)

lnc_T_late <- get_de_names(mast_T_late, subset = all_lncrnas, FCTHRESHOLD = log(1.2))
lnc_T_middle <- get_de_names(mast_T_middle, subset = all_lncrnas, FCTHRESHOLD = log(1.2))
lnc_T_early <- get_de_names(mast_T_early, subset = all_lncrnas, FCTHRESHOLD = log(1.2))
de_t <- c(lnc_T_late, lnc_T_middle, lnc_T_early)

lnc_B_late <- get_de_names(mast_B_late, subset = all_lncrnas)
lnc_B_middle <- get_de_names(mast_B_middle, subset = all_lncrnas)
lnc_B_early <- get_de_names(mast_B_early, subset = all_lncrnas)
de_b<- c(lnc_B_late, lnc_B_middle, lnc_B_early)

de_all <- c(de_myeloid, de_t,de_b)
closest_m <- get_closest_genes(unique(de_myeloid), lnc_ranges_genes, get_gene_only(mrnas_ranges), Inf, 1000000L)

closest_t <- get_closest_genes(unique(de_t), lnc_ranges_genes, get_gene_only(mrnas_ranges), Inf, 1000000L)

closest_b <- get_closest_genes(unique(de_b), lnc_ranges_genes, get_gene_only(mrnas_ranges), Inf, 1000000L)

closest_myeloid <- get_closest_genes(unique(de_myeloid), lnc_ranges_genes, get_gene_only(mrnas_ranges), Inf, 1000000L)

closest_all <- get_closest_genes(unique(de_all), lnc_ranges_genes, get_gene_only(mrnas_ranges), Inf, 1000000L)


do_go(unique(unlist(closest_all)))
do_go(unique(unlist(closest_t)))
do_go(unique(unlist(closest_b)))
do_go(unique(unlist(closest_myeloid)))
do_go(unique(unlist(closest_m)))


# Extract mrnas from significant correlations 
do_go <- function(list, keytype = "SYMBOL"){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =  org.Hs.eg.db, keyType = keytype,ont = "BP")
  barplot(ego, showCategory=10)
  clusterProfiler::dotplot(ego)
}

```


# -------------------------------------------
# -------------------------------------------
# Analysis cis-action/coregulation 

```{r cis-regulation}
# 1) For each gene lncrna find neighbour protein coding genes. ( either overlapping or in close region) --> putative pairs
# 2) Filter: Check the pearson correlation of the identified pairs, reatain the ones w/ |pearson correlation coefficient| > 0.6
# ==> Only if something nice comes out 
# 3) Validation: Simulate random pairs 
# 4) Validation: Compare the correlation of the ones identified with the random ones. 
library(zeallot)

# 1. extract the count matrix
count_matrix <- log2(myeloids@assays$RNA@counts+.5)
count_matrix <- myeloids@assays$RNA@counts
rownames(count_matrix) <- gsub("-unkown","",rownames(count_matrix))
exp_matrix <-myeloids@assays$integrated@data

# 2. 
mrnas_ranges <- ref[ref$gene_name %in% mrna_ref_id  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unkown", "",all_lncrnas))]


lnc_ranges
# Add missing gene entries 
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) (get_gene_range(x, lnc_ranges)))
missing_gene_ranges <- do.call(base::c,gr_list)
c(gr_list[1][[1]],gr_list[2][[1]])

lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)

# Only select the DE interestin

lnc_ranges_complete_de <- lnc_ranges_complete[lnc_ranges_complete$gene_id %in% gsub("-", "_",gsub("-unkown", "",de_lnc_myelid_late_ordered)), ]

# ----------------------
# Find pairs allowing 10K gap on both sides

pairs <- findOverlapPairs(lnc_ranges_complete_de, get_gene_only(mrnas_ranges), maxgap=10000L, ignore.strand = TRUE)
# Get a list of ranges objects in pairs
pairs_list <- zipup(pairs)
length(pairs_list)

# How many lncrnas in total vs how many are in pairs
length(unique(lnc_ranges$gene_id)); length(unique(mrnas_ranges$gene_id)); length(pairs)

# get the counts of the interesting genes
calc_correlation <- function(pair, count_matrix, type = "pearson"){
  gene1 <- str_replace_all(pair[1]$gene_id, "_" , "-")
  gene2 <- str_replace_all(pair[2]$gene_id, "_" , "-")
  print(gene1)
  print(gene2)
  c1 <- count_matrix[rownames(count_matrix)== gene1,]
  c2 <- count_matrix[rownames(count_matrix)== gene2,]
  print("----")
  # calculate the correlation coefficient 
  if(length(c1)> 0 && length(c2)>0){ 
    #pc <- bayes.cor.test(c1, c2) 
    pc <- cor.test(c1, c2, method = c(type)) 
  }else{ pc = cor.test(c(1:100),c(1:100), method = c(type))  }
  return(pc)
}
# Calc correlation values

correlations <- lapply(pairs_list, function(x) calc_correlation(x, count_matrix, type = "spearman"))
pvals <- (lapply(correlations, function(correlation) print(correlation$p.value)))
rhos <- as.vector(lapply(correlations, function(correlation) print(correlation$estimate)))

# Obtain significant correlations
significant_correlations <- correlations[pvals < 0.1 & rhos !=1]
significant_pairs <- pairs_list[pvals < 0.1 & rhos !=1]
significant_correlations

# Extract mrnas from significant correlations 
do_go <- function(list, keytype = "SYMBOL"){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =  org.Mmu.eg.db, keyType = keytype,ont = "BP",universe =unique(ref$gene_name))
  barplot(ego, showCategory=10)
  clusterProfiler::dotplot(ego)
}
correlated_mrnas <-unlist(significant_pairs)[c(rep(FALSE,1), TRUE)]$gene_id
do_go(correlated_mrnas, keytype = "ENSEMBL")

keytypes(org.Mmu.eg.db)

```
