---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(GENIE3)
library(igraph)
library(RCy3)
library(Rgraphviz)
library(reshape2)
library(org.Mmu.eg.db)
library(enrichplot)
library(networkD3)

source("../../utils/02_sc_utils.R")

# IN VIVO 
immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")

de_all<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_genes.rds")
# get DE lnc names

ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

```

# Prepare the matrixes to calculate the networks on. One per celltype, only DE genes. 

```{r a}
myeloids <- immune.combined[,Idents(immune.combined)=="Myeloid"]
b <- immune.combined[,Idents(immune.combined)=="B"]
t <- immune.combined[,Idents(immune.combined)=="T"]

de_myeloids <- myeloids[rownames(myeloids) %in% de_all,]
de_b <- b[rownames(b) %in% de_all,]
de_t <- t[rownames(t) %in% de_all,]

exprMatr_myeloids_de <- de_myeloids@assays$integrated@data
exprMatr_b_de <- de_b@assays$integrated@data
exprMatr_t_de <- de_t@assays$integrated@data


universe_myeloids <- unique(rownames(myeloids))
universe_t <- unique(rownames(t))
universe_b <- unique(rownames(b))
```



# Prepare the matrixes with Ebola genes 

```{r a}
ebola_genes <- paste(ebola_ref$gene_id, "-unknown", sep = "")

de_myeloids_ebola <- myeloids[rownames(myeloids) %in% c(de_all,ebola_genes)]
de_b_ebola <- b[rownames(b) %in% c(de_all,ebola_genes)]
de_t_ebola <- t[rownames(t) %in% c(de_all,ebola_genes),]

exprMatr_myeloids_de_ebola <- de_myeloids_ebola@assays$integrated@data
exprMatr_b_de_ebola <- de_b_ebola@assays$integrated@data
exprMatr_t_de_ebola <- de_t_ebola@assays$integrated@data


write.csv(t(exprMatr_myeloids_de_ebola), "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_myeloids_de_ebola.csv", row.names = FALSE)

write.csv(t(exprMatr_b_de_ebola), "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_b_de_ebola.csv", row.names = FALSE)

write.csv(t(exprMatr_t_de_ebola), "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_t_de_ebola.csv", row.names = FALSE)


```

# Save all the expression matrices

```{r a}
dir.create(dirname(file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_myeloids_de.csv")), showWarnings = F)

write.csv(t(exprMatr_myeloids_de), "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_myeloids_de.csv", row.names = FALSE)


write.csv(t(exprMatr_b_de), "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_b_de.csv", row.names = FALSE)


write.csv(t(exprMatr_t_de), "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/matrix_final_t_de.csv", row.names = FALSE)

```