---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Build network based on values from grnBoost 


## 1. Imports 
```{r include=FALSE}
library(dplyr)
library(ggsci)
library(cowplot)
library(Seurat)
library(circlize)
library(Matrix)
library(stringr)
library(rtracklayer)
library(ggrepel)
library(scales)
library(GENIE3)
library(gplots)
library (plyr)
library(igraph)
library(reshape2)
library(org.Mmu.eg.db)
library(enrichplot)
library(networkD3)

source("../../utils/00_datapaths.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

# IN VIVO 
immune.combined <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"))
dim(immune.combined)

# Load objects 
ref <- import(file.path(file.path(data_path,"/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf")))
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))


de_all<- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds"))
de_lnc <- de_all[de_all$type == "lnc", ]

# Get DE lncRNAs gene names
robjectsdir_stats <- file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/")
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
de_all$type <- ifelse(de_all$primerid %in% all_lncrnas, "lnc", "pc")
de_lnc_names <- gsub("-unknown", "",de_all[de_all$type == "lnc", ]$primerid)

# DE per cell-type
de_lnc_mono <- gsub("-unknown", "",de_all[de_all$celltype == "Monocyte" & de_all$type == "lnc",]$primerid)

# Load graph info 
gD2 <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/gD2.rds"))
lou2 <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/lou.rds"))
linkList <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/linkList.rds"))
```


## 1. Prepare cell-type subset 
```{r a}
monocyte <- immune.combined[,Idents(immune.combined)=="Monocyte"]
de_monocyte<- monocyte[rownames(monocyte) %in% de_all[de_all$celltype == "Monocyte",]$primerid,]
# Universe needed for enrichment analysis (expressed genes in monocytes)
universe_myeloids <- unique(rownames(monocyte))
```


#  2. Inspect communities
```{r a}
set.seed(1)

do_go_community<- function(list, universe,n,de_lnc_names, genes = T, npathways=5 ){
    # ---------------------------------------------
    #       Check the enrichment per community 
    # --------------------------------------------
   ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
   #ego <- clusterProfiler::enrichKEGG(gene = list, organism = "mcc", keyType = "SYMBOL", pvalueCutoff = 0.05, pAdjustMethod = "BH", universe =universe, minGSSize = 10, maxGSSize = 500, qvalueCutoff = 0.2, use_internal_data = FALSE)


   p1 <- clusterProfiler::dotplot(ego,showCategory =5)
   p1b <-barplot(ego,showCategory =5)
   p2 <- heatplot(ego, showCategory =5)
   
   lnc_comm <- unlist(lou2[lou2$membership == n])[unlist(lou2[lou2$membership == n]) %in% de_lnc_names]
   orth <- Reduce("rbind", lapply(lnc_comm, get_orthologname_))
   if(genes == T ){
     go_genes <- unique(unlist(lapply(ego@result[1:npathways, ]$geneID, function(x) str_split(x,"/")[[1]])))
     return(go_genes)
   }else{
     return(list(p1,p1b,p2, orth))
   }
   
}

# Check community enrichments for each of them 
community_check <- function(n, genes = F){ do_go_community(lou2$names[lou2$membership == n],universe_myeloids,n, de_lnc_mono, genes = genes) }

# Leave commented the ones presenting no enrichment
c1 <- community_check(1)
c1
c2 <- community_check(2)
#c2
c3 <- community_check(3)
#c3
c4 <- community_check(4)
c4
c5 <- community_check(5)
#c5
c6 <- community_check(6)
#c6
c7 <- community_check(7)
#c7

# Save genes leading enrichments of communities
c1_genes <- community_check(1, genes = T)
c4_genes <- community_check(4, genes = T)
go_genes <- c(c1_genes, c4_genes)
```

# 3. Plot the graph (MAIN)

```{r colocation}
# Graph plotting Proprieties
V(gD2)$color <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_names, "ENSMMUG00000052424") ,  "dark red", "grey")
# highlight novel ones 
V(gD2)[startsWith(names(as.list(V(gD2))), "MSTRG")]$color <- "dark red"
# Highlight orthologs
V(gD2)$frame.color <- "black"
V(gD2)$community <- membership(lou2)



# ------ Set layout --------------------
#l <- layout.fruchterman.reingold(gD2, niter=5000)
#saveRDS(l, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/layout.rds")
l <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/layout.rds"))

# ------ Which genes to highlight --------------------
go_genes <- c(go_genes, "FCGR3", "MS4A7", "ZFP36", "ENSMMUG00000052424", "SRRM2")


# ------ General proprieties -------------------------
V(gD2)$size <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_mono, go_genes), 4, 1)
V(gD2)$label <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_mono, go_genes), unlist(lapply(names(as.list(V(gD2))), get_orthologname_)), NA)
V(gD2)$font <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_mono), 2, 1 )
V(gD2)$degree <- -pi/2 


# ------------- Change position of specific genes 
# Below
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("IL7R", "S100A9", "S100A8","GADD45B", "ISG15", "MSTRG.195139", "MS4A7", "LGALS3", "ENSMMUG00000058532"), pi/2, V(gD2)$degree )
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("FCGR3" ), pi/1.4, V(gD2)$degree )

# Right
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("CDKN1A", "GADD45B", "HBEGF", "MSTRG.168139", "LINC00861", "ENSMMUG00000052424", "SRRM2"), 0, V(gD2)$degree )
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c("CDKN1A", "GADD45B","HBEGF", "LINC00861", "ENSMMUG00000052424", "SRRM2"), 1.5,0.55 )

# Left
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("ENSMMUG00000058644", "MSTRG.228510", "MSTRG.17705"), pi, V(gD2)$degree )
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c("ENSMMUG00000058644" ), 3.5,V(gD2)$dist)
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c( "MSTRG.228510", "MSTRG.17705", "MSTRG.168139"), 2.5,V(gD2)$dist)

V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("KLF4"), pi, V(gD2)$degree )
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c("KLF4" ), 1,V(gD2)$dist)



# -------------------Plot the Graph 
png(filename = file.path(data_path,"plots/04/graph_main.png"), width = 3300, height = 2100)
plot(lou2, gD2,
    edge.arrow.size = .1,
    col=V(gD2)$color,
    layout=l,
    edge.color= "black", vertex.frame.width = 10,
    mark.col=adjustcolor("grey", alpha = 0.3),
    mark.border=adjustcolor("grey", alpha = 0.3),
    vertex.label.color="black", vertex.label.font=1,
    vertex.label.cex=3,
    vertex.label.dist=V(gD2)$dist , 
    vertex.label.degree=V(gD2)$degree, 
    vertex.label.family="Helvetica",
    vertex.label = V(gD2)$label )+ geom_text_repel()

legend("topright",
       legend=unique(lou2$membership),
       fill=adjustcolor(pal_jco()(8), alpha = 0.4), border=NA)
dev.off()


```

# 4. Plot graph for supplementary: Dyrectionality of dyregulation 


```{r readfilesDE}

# ---- 1. Get the direction ( UP, DOWN or MIXED)
genes <- names(as.list(V(gD2)))
de_all$id <- gsub("-unknown","",de_all$primerid) 
mono_de <- de_all[de_all$celltype == "Monocyte",]
mono_de <- mono_de %>% dplyr::group_by(id, type) %>% dplyr::summarise(direction = toString((direction)))
a_red <- lapply(mono_de$direction, function(x) unique(str_split(x, ", ")[[1]]))
summarized_direction <- unlist(ifelse(lapply(a_red, function(x) length(x))>1, "mixed", a_red))
mono_de$summarized_direction_lnc <- summarized_direction

mono_de$id <- gsub("MAMU-", "MAMU.", mono_de$id, fixed = T)
mono_de <- dplyr::distinct(mono_de[mono_de$id %in% genes, ])
mono_de$color <- "grey"
mono_de[mono_de$summarized_direction_lnc == "up", ]$color <- "red"
mono_de[mono_de$summarized_direction_lnc == "down", ]$color <- "blue"
rownames(mono_de) <- mono_de$id

V(gD2)$color  <- mono_de[names(as.list(V(gD2))),]$color

  
# 2. ------------- PLOT 
png(filename = file.path(data_path,"plots/04/SUPPL_graph_directionalityDysregulation.png"), width = 3300, height = 2100)
plot(lou2, gD2,
    edge.arrow.size = .1,
    col=V(gD2)$color,
    layout=l,
    edge.color= "black", vertex.frame.width = 10,
    mark.col=adjustcolor("grey", alpha = 0.3),
    mark.border=adjustcolor("grey", alpha = 0.3),
    vertex.label.color="black", vertex.label.font=1,
    vertex.label.cex=3,
    vertex.label.dist=V(gD2)$dist , 
    vertex.label.degree=V(gD2)$degree, 
    vertex.label.family="Helvetica",
    vertex.label = V(gD2)$label )+ geom_text_repel()
dev.off()
```

#  5. Plot graph for supplementary: Stage of dyregulation 

```{r readfilesDE}

# ---- 1. Get the direction 
# All genes in graph 
genes <- names(as.list(V(gD2)))
# Get table with DE information 
de_all$id <- gsub("-unknown","",de_all$primerid) 
# Only extract cell-type concerned and genes used
mono_de <- de_all[de_all$celltype == "Monocyte",]
mono_de$id <- gsub("MAMU-", "MAMU.", mono_de$id, fixed = T)
mono_de <- dplyr::distinct(mono_de[mono_de$id %in% genes, ])
length(unique(mono_de$primerid))
mono_de <- mono_de[,c("id", "type", "coef", "stage")]
# Retain DE info for maximum FC
mono_de <- mono_de %>% dplyr::group_by(id) %>%  dplyr::filter(abs(coef) == max(abs(coef)))
length(unique(mono_de$id))

# ---------------    Only get the stage with maximum FC ------------------------
mono_de <- mono_de %>% dplyr::group_by(id, type) %>% dplyr::summarise(stage = toString((stage)))
a_red <- lapply(mono_de$stage, function(x) unique(str_split(x, ", ")[[1]]))
summarized_direction <- unlist(ifelse(lapply(a_red, function(x) length(x))>1, "mixed", a_red))
mono_de$summarized_direction_lnc <- summarized_direction


mono_de$color <- "grey"
mono_de[mono_de$summarized_direction_lnc == "early", ]$color <- "yellow"
mono_de[mono_de$summarized_direction_lnc == "middle", ]$color <- "orange"
mono_de[mono_de$summarized_direction_lnc == "late", ]$color <- "red"
rownames(mono_de) <- mono_de$id

V(gD2)$color  <- mono_de[names(as.list(V(gD2))),]$color

# --------- PLOT 
png(filename = file.path(data_path,"plots/04/SUPPL_graph_stage.png"), width = 3300, height = 2100)
plot(lou2, gD2,
    edge.arrow.size = .1,
    col=V(gD2)$color,
    layout=l,
    edge.color= "black", vertex.frame.width = 10,
    mark.col=adjustcolor("grey", alpha = 0.3),
    mark.border=adjustcolor("grey", alpha = 0.3),
    vertex.label.color="black", vertex.label.font=1,
    vertex.label.cex=3,
    vertex.label.dist=V(gD2)$dist , 
    vertex.label.degree=V(gD2)$degree, 
    vertex.label.family="Helvetica",
    vertex.label = V(gD2)$label )+ geom_text_repel()
dev.off()


lou2
```

# 6. Dotplot ISGs 

```{r dotplotISG}

monocyte$group_red <- factor(toupper(substring(monocyte$group, 1,1)), levels = c("B", "E", "M", "L"))

# select genes 
genes <- c("ENSMMUG00000064224-unknown","PLAC8","ISG15","IFIT2","MX1")

# DOTPLOT standard 
#scale.func <- switch(EXPR = "radius", size = scale_size, 
#                     radius = scale_radius, stop("'scale.by' must be either 'size' or 'radius'"))
#plot.data <- Dotplot_data(monocyte, features = genes, dot.scale =25,group.by = "group",scale.by = "radius", cols = c( "blue", "red"))
#plot.data$name <- unlist(lapply(plot.data$features.plot, get_orthologname_))
#dotplot_isgs <- ggplot(plot.data, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_classic()+xlab("")+ylab("")+
#        scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+
#        scale.func(range = c(0, 13))+ggtitle("")+
#        theme(axis.text.x= element_text(size = 20, color = "black"), axis.text.y= element_text(size = 15, color = "black"))
#dotplot_isgs


# DOTPLOT when i calculate it !!
# ---------- ISGS
plot.data_up <- Reduce(rbind, lapply(genes, function(gene) plot_expression_gene(gene, monocyte)[3][[1]]))
plot.data_up$name <- unlist(lapply(plot.data_up$gene, get_orthologname_))
dotplot_isgs <- ggplot(plot.data_up, aes(x = type, y = name, size= pct.exp, col = mean))+geom_point()+theme_paper+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+scale_size(range = c(5,10))+
  theme(axis.ticks.x = element_blank(), axis.title.y = element_blank() )


pdf(file.path(plots, "04/dotplot_isgs.pdf"), width = 8, height = 6)
dotplot_isgs
dev.off()


# ----------- Leukocyte proliferation 
genes <- c(paste(groups(lou2)[[1]][groups(lou2)[[1]] %in% de_lnc_names], "unknown", sep = "-"), c1_genes)
plot.data_up <- Reduce(rbind, lapply(genes, function(gene) plot_expression_gene(gene, monocyte)[3][[1]]))
plot.data_up$name <- unlist(lapply(plot.data_up$gene, get_orthologname_))
dotplot_leukprol <- ggplot(plot.data_up, aes(x = type, y = name, size= pct.exp, col = mean))+geom_point()+theme_paper+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+scale_size(range = c(5,10))+
  theme(axis.ticks.x = element_blank(), axis.title.y = element_blank() )

pdf(file.path(plots, "04/dotplot_leuk.pdf"), width = 8, height = 6)
dotplot_leukprol
dev.off()

```


# 7. Store information about community 

```{r readfilesDE}
# GeneID, Community number
df <- ldply ( groups(lou2), data.frame)
colnames(df) <- c("Community", "gene_name")

saveRDS(df, file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/Graph_communitybelonging.rds"))

# Sort by size

```




