---
title: "KEGG enrichment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Build network based on values from grnBoost 


## 1. Imports 
```{r include=FALSE}
library(dplyr)
#library(tfse)
library(ggsci)
library(cowplot)
library(spatstat)
library(Seurat)
library(circlize)
library(Matrix)
library(stringr)
library(rtracklayer)
library(ggrepel)
library(scales)
library(GENIE3)
library(gplots)
library (plyr)
library(igraph)
library(reshape2)
library(org.Mmu.eg.db)
library(enrichplot)
library(networkD3)

data_path <- "~/cluster/02.ebola_sc/data"

#source("../../utils/00_datapaths.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

# IN VIVO 
immune.combined <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"))
dim(immune.combined)

# Load objects 
ref <- import(file.path(file.path(data_path,"01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf")))
orthologs <- readRDS(file.path(data_path, "/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))


de_all<- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds"))
de_lnc <- de_all[de_all$type == "lnc", ]

# Get DE lncRNAs gene names
robjectsdir_stats <- file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/")
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
de_all$type <- ifelse(de_all$primerid %in% all_lncrnas, "lnc", "pc")
de_lnc_names <- gsub("-unknown", "",de_all[de_all$type == "lnc", ]$primerid)

# DE per cell-type
de_lnc_mono <- gsub("-unknown", "",de_all[de_all$celltype == "Monocyte" & de_all$type == "lnc",]$primerid)

# Load graph info 
gD2 <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/gD2.rds"))
lou2 <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/lou.rds"))
linkList <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/linkList.rds"))
```


## 1. Prepare cell-type subset 
```{r a}
monocyte <- immune.combined[,Idents(immune.combined)=="Monocyte"]
de_monocyte<- monocyte[rownames(monocyte) %in% de_all[de_all$celltype == "Monocyte",]$primerid,]
# Universe needed for enrichment analysis (expressed genes in monocytes)
universe_myeloids <- unique(rownames(monocyte))
```


```{r}
do_go_community<- function(list, universe,n,de_lnc_names, genes = T, npathways=5 ){
    # ---------------------------------------------
    #       Check the enrichment per community 
    # --------------------------------------------
      set.seed(5)
  mcc <- org.Mmu.eg.db

  entrez <- AnnotationDbi::select(mcc, 
       keys = list,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

  entrez <- entrez[!duplicated(entrez$SYMBOL),]
  id <- entrez$ENTREZID

  universe_entrez <- AnnotationDbi::select(mcc, 
       keys = universe,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

  universe_entrez <- universe_entrez[unique(universe_entrez$SYMBOL),]
  universe_entrez <- universe_entrez$ENTREZID
  ego <- clusterProfiler::enrichKEGG(gene = id, organism = "mcc", keyType = "ncbi-geneid", pvalueCutoff = 0.05, pAdjustMethod = "BH", minGSSize = 3, universe = universe_entrez)

  
   p1 <- clusterProfiler::dotplot(ego,showCategory =5)
   p1b <-barplot(ego,showCategory =5)
   p2 <- heatplot(ego, showCategory =5)
   
   lnc_comm <- unlist(lou2[lou2$membership == n])[unlist(lou2[lou2$membership == n]) %in% de_lnc_names]
   orth <- Reduce("rbind", lapply(lnc_comm, get_orthologname_))
   if(genes == T ){
     go_genes <- unique(unlist(lapply(ego@result[1:npathways, ]$geneID, function(x) str_split(x,"/")[[1]])))
     return(go_genes)
   }else{
     return(list(p1,p1b,p2, orth))
   }
   
}

# Check community enrichments for each of them 
community_check <- function(n, genes = F){ do_go_community(lou2$names[lou2$membership == n],universe_myeloids,n, de_lnc_mono, genes = genes) }

# Leave commented the ones presenting no enrichment
#c1 <- community_check(1)
#c1
c2 <- community_check(2)
c2
c3 <- community_check(3)
c3
c4 <- community_check(4)
c4
c5 <- community_check(5)
c5
c6 <- community_check(6)
c6
c7 <- community_check(7)
c7

mcc <- org.Mmu.eg.db
 universe_entrez <- AnnotationDbi::select(mcc, 
       keys = c("106992282", "718909"),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "ENTREZID")


```

```{r}
 entrez <- AnnotationDbi::select(mcc, 
       keys = c("100424944", "699365", "717897"),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "ENTREZID")
```




```{r}

list =lou2$names[lou2$membership == 2]
 mcc <- org.Mmu.eg.db

  entrez <- AnnotationDbi::select(mcc, 
       keys = list,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
 entrez <-entrez[!duplicated(entrez$SYMBOL),]
  id <- entrez$ENTREZID
  universe_entrez <- AnnotationDbi::select(mcc, 
       keys = universe_myeloids,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
  universe_entrez <- universe_entrez$ENTREZID
ego <- clusterProfiler::enrichKEGG(gene = id, organism = "mcc", keyType = "ncbi-geneid", pvalueCutoff = 0.05, pAdjustMethod = "BH", minGSSize = 10, universe = universe_entrez)
  p1 <- clusterProfiler::dotplot(ego,showCategory =5)
   p1b <-barplot(ego,showCategory =5)
   p2 <- heatplot(ego, showCategory =5)
   
   lnc_comm <- unlist(lou2[lou2$membership == n])[unlist(lou2[lou2$membership == n]) %in% de_lnc_names]
   orth <- Reduce("rbind", lapply(lnc_comm, get_orthologname_))

```


############################
```{r}

list <- lou2$names[lou2$membership == 1]
EG_IDs = mget(list, revmap(org.Mmu.egSYMBOL),ifnotfound=NA)

   EG_IDs <- na_omit(EG_IDs) 
   
   KEGG_IDs = mget(as.character(EG_IDs), org.Mmu.egPATH,ifnotfound=NA)
   KEGG_IDs <- na_omit(KEGG_IDs)



gse <- gseKEGG(geneList     = KEGG_IDs,
               organism     = "mcc",
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "KEGG")













x <- org.Mmu.egPATH
# Get the entrez gene identifiers that are mapped to a KEGG pathway ID
mapped_genes <- mappedkeys(x)
# Convert to a list
xx <- as.list(x[mapped_genes])
if(length(xx) > 0) {
# Get the PATH for the first five genes
xx[1:5]
# Get the first one
xx[[1]]
}
   
   sym <- lou2$names[lou2$membership == 1]

   list <- select(mcc, 
       keys = my.symbols,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
   list <- list %>% drop_na()
   entrez <- list$KEGG
   
   
   EG_IDs = mget(list, revmap(org.Mmu.egSYMBOL),ifnotfound=NA)

   EG_IDs <- na_omit(EG_IDs) 
   
   KEGG_IDs = mget(as.character(EG_IDs), org.Mmu.egPATH,ifnotfound=NA)
   KEGG_IDs <- na_omit(KEGG_IDs) 
   
   
   
   
library(org.Hs.eg.db)

##Toy example symbols:
sym = c("AKT3","CDH1")

##Get the Entrez gene IDs associated with those symbols
EG_IDs = mget(sym, revmap(org.Hs.egSYMBOL),ifnotfound=NA)

##Then get the KEGG IDs associated with those entrez genes.
KEGG_IDs = mget(as.character(EG_IDs), org.Hs.egPATH,ifnotfound=NA)

   
```


