---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Build network based on values from grnBoost 


## 1. Imports 
```{r include=FALSE}
library(igraph)
library(stringr)

source("../../utils/00_datapaths.R")
source("../../utils/02_sc_utils.R")
source("../../utils/04_utils_graph.R")

# Load graph info 
gD2 <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/gD2.rds"))
lou2 <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/lou.rds"))
linkList <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/linkList.rds"))

de_all<- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds"))
de_lnc <- de_all[de_all$type == "lnc", ]
de_lnc_mono <- gsub("-unknown", "",de_all[de_all$celltype == "Monocyte" & de_all$type == "lnc",]$primerid)

graph_info <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/Graph_communitybelonging.rds"))
summary <- readRDS(file.path(data_path, "00_Metadata/final_correlation_summary.rds"))



```

```{r include=FALSE}
# Check which community is which number 

graph_info
comm_size <- graph_info %>% dplyr::group_by(Community) %>% dplyr::count()
comm_size[order(comm_size$Community),]$n
graph_info[graph_info$Community == 3, ]
```

```{r include=FALSE}

# Compute an color according to in how many cells they are DE
de_all
de_lnc

de_reduced <- unique(de_all[,c("primerid", "celltype")])
de_n_celltypes <- de_reduced %>% dplyr::group_by(primerid) %>% dplyr::count()
rownames(de_n_celltypes) <- de_n_celltypes$primerid

de_all$id <- gsub("-unknown","",de_all$primerid) 
de_all$id <- gsub("MAMU-", "MAMU.", de_all$id, fixed = T)
de_mono <- de_all[de_all$celltype == "Monocyte",]

de_graph <- de_mono %>% dplyr::group_by(id) %>%  dplyr::filter(abs(coef) == max(abs(coef)))
de_graph <- de_graph[,c("primerid", "direction", "type", "stage", "id")]
mono_de <-de_graph

#--- Compute color Up/DOWN
graph_info$id <- graph_info$gene_name
graph_info_de <- merge(graph_info, mono_de, by="id")
# Add community size 
comm_size <- graph_info_de %>% dplyr::group_by(Community) %>% dplyr::count()
names(comm_size) <- c("Community", "community_size")
graph_info_de <- merge(graph_info_de, comm_size, by="Community")
# % of UP 
graph_info_de_full <- graph_info_de %>% dplyr::group_by(Community,community_size, direction) %>% dplyr::count()
graph_info_de_up_perc <- graph_info_de_full[graph_info_de_full$direction == "up",]
graph_info_de_up_perc$perc_up <- graph_info_de_up_perc$n/ graph_info_de_up_perc$community_size
#Get color palette
```


# 3. Plot the graph (Shades for Different cell-types)

```{r grapg}

V(gD2)$frame.color <- "black"
V(gD2)$community <- membership(lou2)
# ------ Set layout --------------------
#l <- layout.fruchterman.reingold(gD2, niter=5000)
#saveRDS(l, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/layout.rds")
l <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/layout.rds"))

# ------ Which genes to highlight --------------------
#go_genes <- c(go_genes, "FCGR3", "MS4A7", "ZFP36", "ENSMMUG00000052424", "SRRM2")

# ------ General proprieties -------------------------
# 1. Dot's size 
V(gD2)$size <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_mono), 4, 2)

# 2. Label size
V(gD2)$label <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_mono), unlist(lapply(names(as.list(V(gD2))), get_orthologname_)), NA)
V(gD2)$font <- ifelse(names(as.list(V(gD2))) %in% c(de_lnc_mono), 2, 1 )
V(gD2)$degree <- -pi/2 


# ------------- Change position of specific genes 
# Below
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("IL7R", "S100A9", "S100A8","GADD45B", "ISG15", "MSTRG.195139", "MS4A7", "LGALS3", "ENSMMUG00000058532"), pi/2, V(gD2)$degree )
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("FCGR3" ), pi/1.4, V(gD2)$degree )

# Right
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("CDKN1A", "GADD45B", "HBEGF", "MSTRG.168139", "LINC00861", "ENSMMUG00000052424", "SRRM2"), 0, V(gD2)$degree )
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c("CDKN1A", "GADD45B","HBEGF", "LINC00861", "ENSMMUG00000052424", "SRRM2"), 1.5,0.55 )

# Left
V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("ENSMMUG00000058644", "MSTRG.228510", "MSTRG.17705"), pi, V(gD2)$degree )
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c("ENSMMUG00000058644" ), 3.5,V(gD2)$dist)
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c( "MSTRG.228510", "MSTRG.17705", "MSTRG.168139"), 2.5,V(gD2)$dist)

V(gD2)$degree <- ifelse(names(as.list(V(gD2))) %in% c("KLF4"), pi, V(gD2)$degree )
V(gD2)$dist <- ifelse(names(as.list(V(gD2))) %in% c("KLF4" ), 1,V(gD2)$dist)


# ------
graph_info <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/Graph_communitybelonging.rds"))
# Only retain genes in graph 
de_all$id <- gsub("-unknown","",de_all$primerid) 
de_all$id <- gsub("MAMU-", "MAMU.", de_all$id, fixed = T)
de_mono <- de_all[de_all$celltype == "Monocyte",]
de_graph <- de_mono %>% dplyr::group_by(id) %>%  dplyr::filter(abs(coef) == max(abs(coef)))
de_graph <- de_graph[,c("primerid", "direction", "type", "stage")]



de_all$id <- gsub("-unknown","",de_all$primerid) 
de_all$id <- gsub("MAMU-", "MAMU.", de_all$id, fixed = T)
de_mono <- de_all[de_all$celltype == "Monocyte",]

de_graph <- de_mono %>% dplyr::group_by(id) %>%  dplyr::filter(abs(coef) == max(abs(coef)))
de_graph <- de_graph[,c("primerid", "direction", "type", "stage", "id")]
mono_de <-de_graph

table(de_graph$primerid %in% graph_info$gene_name)
table(graph_info$gene_name %in% mono_de$primerid)
graph_info[!(graph_info$gene_name %in% mono_de$id),]

mono_de[mono_de$primerid == "ENSMMUG00000054038-unknown",]
de_graph

plot_expression_gene
```

# 4. Plot graph for supplementary: Dyrectionality of dyregulation 


```{r readfilesDE}

# ---- 1. Get the direction ( UP, DOWN or MIXED)
genes <- names(as.list(V(gD2)))


# 1 - Get wether up- or down- regulated 
de_all$id <- gsub("-unknown","",de_all$primerid) 
mono_de <- de_all[de_all$celltype == "Monocyte",]
mono_de <- mono_de %>% dplyr::group_by(id, type) %>% dplyr::summarise(direction = toString((direction)))
a_red <- lapply(mono_de$direction, function(x) unique(str_split(x, ", ")[[1]]))
summarized_direction <- unlist(ifelse(lapply(a_red, function(x) length(x))>1, "mixed", a_red))
mono_de$summarized_direction_lnc <- summarized_direction
mono_de$id <- gsub("MAMU-", "MAMU.", mono_de$id, fixed = T)
mono_de <- dplyr::distinct(mono_de[mono_de$id %in% genes, ])
mono_de_direction <- mono_de

# 2- Get stage
de_all$id <- gsub("-unknown","",de_all$primerid) 
# Only extract cell-type concerned and genes used
mono_de <- de_all[de_all$celltype == "Monocyte",]
mono_de$id <- gsub("MAMU-", "MAMU.", mono_de$id, fixed = T)
mono_de <- dplyr::distinct(mono_de[mono_de$id %in% genes, ])
length(unique(mono_de$primerid))
mono_de <- mono_de[,c("id", "type", "coef", "stage")]
# Retain DE info for maximum FC
mono_de <- mono_de %>% dplyr::group_by(id) %>%  dplyr::filter(abs(coef) == max(abs(coef)))
length(unique(mono_de$id))

# ---------------    Only get the stage with maximum FC ------------------------
mono_de <- mono_de %>% dplyr::group_by(id, type) %>% dplyr::summarise(stage = toString((stage)))
a_red <- lapply(mono_de$stage, function(x) unique(str_split(x, ", ")[[1]]))
summarized_direction <- unlist(ifelse(lapply(a_red, function(x) length(x))>1, "mixed", a_red))
mono_de$summarized_stage <- summarized_direction
mono_de_stage <- mono_de


mono_de <- merge(mono_de_stage, mono_de_direction, by = "id")

# -------------------
# Change Color 

# 0. Default Color 
mono_de$color <- "grey"

# 1. Up, shades 
up_late <- "#660000"
up_middle <- "#FF0000"
up_early <- "#FF9999"
mono_de[mono_de$summarized_direction_lnc == "up" & mono_de$summarized_stage == "late", ]$color <- up_late
mono_de[mono_de$summarized_direction_lnc == "up" & mono_de$summarized_stage == "middle", ]$color <- up_middle
mono_de[mono_de$summarized_direction_lnc == "up" & mono_de$summarized_stage == "early", ]$color <- up_early

# 2. Down, shades
down_late <- "#000033"
down_middle <- "#0000FF"
down_early <- "#CCE5FF"
mono_de[mono_de$summarized_direction_lnc == "down" & mono_de$summarized_stage == "late", ]$color <- down_late
mono_de[mono_de$summarized_direction_lnc == "down" & mono_de$summarized_stage == "middle", ]$color <- down_middle
mono_de[mono_de$summarized_direction_lnc == "down" & mono_de$summarized_stage == "early", ]$color <- down_early


# Update colors
rownames(mono_de) <- mono_de$id
V(gD2)$color  <- mono_de[names(as.list(V(gD2))),]$color


# -------------------Plot the Graph 
png(filename = file.path(data_path,"plots/04/graph_main_shades.png"), width = 3300, height = 2100)
plot(lou2, gD2,
    edge.arrow.size = .1,
    col=V(gD2)$color,
    layout=l,
    edge.color= "black", vertex.frame.width = 10,
    mark.col=adjustcolor("grey", alpha = 0.3),
    mark.border=adjustcolor("grey", alpha = 0.3),
    vertex.label.color="black", vertex.label.font=1,
    vertex.label.cex=3,
    vertex.label.dist=V(gD2)$dist , 
    vertex.label.degree=V(gD2)$degree, 
    vertex.label.family="Helvetica",
    vertex.label = V(gD2)$label )+ geom_text_repel()

legend("topright",
       legend=unique(lou2$membership),
       fill=adjustcolor(pal_jco()(8), alpha = 0.4), border=NA)
dev.off()

#Get communities belonging
summary <- readRDS(file.path(data_path, "00_Metadata/final_DE_summary.rds"))
summary_community <- unique(summary[,c("gene_name", "community")])
summary_community <- summary_community[!is.na(summary_community$community), ]
summary_community


# -----------------
de_all$id <- gsub("-unknown","",de_all$primerid) 
de_all$id <- gsub("MAMU-", "MAMU.", de_all$id, fixed = T)
de_mono <- de_all[de_all$celltype == "Monocyte",]
de_graph <- de_mono %>% dplyr::group_by(id) %>%  dplyr::filter(abs(coef) == max(abs(coef)))
de_graph <- de_graph[,c("primerid", "direction", "type", "stage", "id")]
mono_de <-de_graph

# Get Color code per community
graph_info$id <- graph_info$gene_name
graph_info_de <- merge(graph_info, mono_de, by="id")
# Add community size 
comm_size <- graph_info_de %>% dplyr::group_by(Community) %>% dplyr::count()
names(comm_size) <- c("Community", "community_size")
graph_info_de <- merge(graph_info_de, comm_size, by="Community")
# % of UP 
graph_info_de_full <- graph_info_de %>% dplyr::group_by(Community,community_size, direction) %>% dplyr::count()
graph_info_de_up_perc <- graph_info_de_full[graph_info_de_full$direction == "up",]
graph_info_de_up_perc$perc_up <- graph_info_de_up_perc$n/ graph_info_de_up_perc$community_size
#Get color palette
svals <-graph_info_de_up_perc$perc_up 
f <- colorRamp(c("blue", "red"))
colors <- rgb(f(svals)/255)


```



