---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Colocation/correlation analysis in vivo 

```{r Imports}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(ggplot2)
library(scales)
library(circlize)
library(reshape2)

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Imports 
source("../../utils/02_sc_utils.R")
source("../../utils/03_de_utils.R")
source("../../utils/04_utils_graph.R")

# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE_neut/")

# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))

# Seurat object
immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready_neut.rds")

# Ortholgs
orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid.rds")



# Check LncPedia
lncpedia_immune <- import("../../lncpedia_immune.gtf")
lncpedia_infection <- import("../../lncpedia_infection.gtf")

lncpedia <- unique(c(unique(unlist(lncpedia_immune@elementMetadata[, grepl("gene", colnames(lncpedia_immune@elementMetadata))])), unique(unlist(lncpedia_infection@elementMetadata[, grepl("gene", colnames(lncpedia_infection@elementMetadata))]))))

immlnc <- read.table("/home/luisas/Desktop/cluster/data/00_RawData/ImmLnc/Lnc_Pathways_Sig.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
reported_immlnc <- immlnc$lncRNA_symbol
# -------------------------------------------
#    Define threshold for the whole analysis
# -------------------------------------------

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)
print("-")


calc_correlation_genes <- function(gene1_vector,gene2_vector,gene1, gene2, type = "spearman"){
  mask <- gene1_vector !=0 & gene2_vector !=0
  # calculate the correlation coefficient
  if(sum(mask)>10){ 
    #pc <- bayes.cor.test(c1, c2) 
    pc <- cor.test(gene1_vector[mask], gene2_vector[mask], method = c(type))
    df <- data.frame( pval=pc$p.value, rho=pc$estimate)
    df$g1 <- gene1
    df$g2 <- gene2
    return(df)
  }else{ 
    return(NA)
  }
}

expand.grid.unique <- function(x, y, include.equals=FALSE){
  x <- unique(x)
  y <- unique(y)
  g <- function(i)
  {
    z <- setdiff(y, x[seq_len(i-include.equals)])
    if(length(z)) cbind(x[i], z, deparse.level=0)
  }
  do.call(rbind, lapply(seq_along(x), g))
}



mono <- subset(immune.combined, ident = "Monocyte")
expressionmatrix <- as.matrix(mono@assays$RNA@counts)
```






```{r colocation}
colocation <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/07_colocation/colocation_lncvsDE_df.rds")
de_all_genes<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE_neut/de_all_stages.rds")
de_pc <- de_all_genes[de_all_genes$type == "pc",]
de_pc_monocyte <- de_pc[de_pc$celltype == "Monocyte",]


distances <- data.frame(colocation, stringsAsFactors = F)
colnames(distances)  <-  c("lnc", "gene", "d")
distances$d <- as.integer(distances$d)

# Add gene names 
genes <- ref[ref$type =="gene",]
correspondence <- data.frame(genes$gene_id, genename=genes$gene_name, stringsAsFactors = F)
rownames(correspondence) <- correspondence$genes.gene_id
distances$gene_name <- unlist(correspondence[distances$gene, ]$genename)


# Add lnc ortholog
orth_lnc <- data.frame(lnc = unique(distances$lnc), stringsAsFactors = F) 
orth_lnc$orth <- unlist(lapply(orth_lnc$lnc, get_orthologname_))
rownames(orth_lnc) <- orth_lnc$lnc
distances$orth_lnc <- orth_lnc[distances$lnc, ]$orth
distances <- distances[!is.na(distances$d) ,]
```


```{r cis}
# Add information about our DE
cis_pairs <- distances[distances$d <cis_threshold,]

# Calculate the correlation for the close ones
df_reduced <- cis_pairs[,c("lnc","gene_name")]
df_reduced$lnc <- paste(df_reduced$lnc, "-unknown", sep = "")
pairs <- as.list(data.frame(t(df_reduced), stringsAsFactors = F))

correlations <- lapply(pairs, function(genes) {calc_correlation_genes(expressionmatrix[genes[1], ], expressionmatrix[genes[2],], genes[1], genes[2])})

correlations_df <- Reduce("rbind",correlations[!is.na(correlations)])
correlations_df$lnc <- gsub("-unknown","", correlations_df$g1)
correlations_df$pair <- paste(correlations_df$lnc, correlations_df$g2, sep = "-")
rownames(correlations_df) <- correlations_df$pair

cis_pairs$pair <- paste(cis_pairs$lnc, cis_pairs$gene_name, sep ="-")
cis_pairs_corr <- merge(cis_pairs, correlations_df, by = "pair", all = T)
```

```{r cis}

cis_pairs_corr_sig <- cis_pairs_corr[!is.na(cis_pairs_corr$pval) & cis_pairs_corr$pval < 0.05,]
cis_pairs_corr_sig_red <- cis_pairs_corr_sig[,c("orth_lnc", "gene_name", "rho")]


gD2 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut/gD2.rds")
lou2 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut/lou.rds")
linkList <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut/linkList.rds")


lnc_candidates <- unique(cis_pairs_corr_sig$lnc.x)

lnc_graphs <- linkList[as.character(linkList$regulatoryGene) %in% lnc_candidates |as.character(linkList$targetGene) %in% lnc_candidates, ]


lnc_reg <- lnc_graphs[lnc_graphs$regulatoryGene %in% lnc_candidates,]
lnc_reg$lnc <- lnc_reg$regulatoryGene
lnc_reg$pc <- lnc_reg$targetGene
lnc_target <- lnc_graphs[lnc_graphs$targetGene %in% lnc_candidates,]
lnc_target$lnc <- lnc_target$targetGene
lnc_target$pc <- lnc_target$regulatoryGene

lnc_graphs_pairs <- rbind(lnc_reg, lnc_target)

lnc_graphs_pairs$pair <- paste(lnc_graphs_pairs$lnc, lnc_graphs_pairs$pc, sep = "-")

cis_pairs_corr_sig$pair <- gsub("-unknown", "", cis_pairs_corr_sig$pair)
merge(cis_pairs_corr_sig,lnc_graphs_pairs,by = "pair",all = F )
```




```{r colocation}
# ---------------------------------------------------------------------------------------
# Import external datasets
isg <- read.table("../isg.csv", stringsAsFactors = F)
de_broad <- read.csv("../de_broad.tsv", sep = "\t", stringsAsFactors = F)
de_broad_sig <- de_broad[de_broad$Module_assignment != "Not significant",]
rownames(de_broad_sig) <- de_broad_sig$Gene_symbol
broad_de <- unique(de_broad[de_broad$Module_assignment != "Not significant",]$Gene_symbol)
# Check if ISG 
distances$ISG <- ifelse(distances$gene_name %in% isg$V1, TRUE, FALSE)
distances$BROAD <- ifelse(distances$gene_name %in% broad_de, TRUE, FALSE)
distances$BROAD_module <- de_broad_sig[distances$gene_name, ]$Module_assignment
# Explore
isg_dist <- distances[ distances$ISG == T & !is.na(distances$d) ,]
broad_dist <- distances[ distances$BROAD == T & !is.na(distances$d) ,]
close_isg <- isg_dist[isg_dist$d < cis_threshold ,]
close_isg_summary <- close_isg %>% dplyr::group_by(lnc) %>% dplyr::mutate(close = paste0(unique(gene_name), collapse = ","))
close_isg_summary <- distinct(close_isg_summary[,c("lnc","close")])
close_isg_summary <- as.data.frame(close_isg_summary)
rownames(close_isg_summary) <- close_isg_summary$lnc
# Add isg names 
close_broad <- broad_dist[broad_dist$d < cis_threshold ,]$lnc
broad_dist <- distances[ distances$BROAD == T & distances$d < cis_threshold ,]
# ---------------------------------------------------------------------------------------
```