---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)

source("../../utils/02_sc_utils.R")

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/")

# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))
print("--")

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")

de_all_genes<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_genes.rds")
length(unique(de_all_genes))

de_mrna_all <- de_all_genes[de_all_genes %in% annotated_mrnas]
length(unique(de_mrna_all))

de_lnc_all <- de_all_genes[de_all_genes %in% all_lncrnas]
length(unique(de_lnc_all))

stats_complete <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/stats_complete.rds")

# Check that i also have the novel lncRNAs
# OK - i also have the novel ones
table(substr(all_lncrnas, 1,4) == "MSTR")

orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs.rds")


df_lnc <- readRDS(file.path(robjectsdir_stats, "df_lnc.rds"))
df_mrna <- readRDS(file.path(robjectsdir_stats, "df_mrna.rds"))
df_lnc$type <- "lncRNAs"
df_mrna$type <- "mRNAs"
df_complete <- rbind(df_lnc, df_mrna)

col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)

all_genes <- unlist(lapply(rownames(df_complete), function(x) gsub( "-unknown", "",x)))
```


# Explore protein coding genes 

```{r a}
table(de_all$celltype)
library(ggupset)

ego_up <- clusterProfiler::enrichGO(gene = unique(de_all[de_all$type == "pc" & de_all$coef >0  ,]$primerid),OrgDb=org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))

ego_down <- clusterProfiler::enrichGO(gene = unique(de_all[de_all$type == "pc" & de_all$coef < 0,]$primerid),OrgDb=org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))

heatplot(ego_up)
heatplot(ego_down)

upsetplot(ego_up) 
upsetplot(ego_down) 
```

# Explore lncRNAs, how can I have some function 
```{r a}
de_lnc_up <- de_all[de_all$type == "lnc" & de_all$coef >0 ,]
de_lnc_down <- de_all[de_all$type == "lnc" & de_all$coef <0 ,]
de_lnc_up <- de_lnc_up[order(de_lnc_up$fdr),]
de_lnc_down <- de_lnc_up[order(de_lnc_down$fdr),]
# Add gene id to orthologs data frame
#orthologs$gene_id  <- apply( orthologs, 1, function(o) unique(ref[!is.na(ref$transcript_id) & ref$transcript_id == o[1], ]$gene_id))

find_ortholog <- function(gene){
  return(unique(orthologs[orthologs$gene_id == gene,]$orthologGeneSymbol))
}

orthologs_lnc_up<- unlist(lapply(gsub("-unknown", "",de_lnc_up$primerid), function(x) find_ortholog(x)))
orthologs_lnc_down<- unlist(lapply(gsub("-unknown", "",de_lnc_down$primerid), function(x) find_ortholog(x)))

orthologs_lnc_up
orthologs_lnc_down

intersect(orthologs_lnc_down, orthologs_lnc_up)

ortholog_name <- "ZFAS1"

unique(orthologs[orthologs$orthologGeneSymbol == ortholog_name, ]$gene_id)



rownames(immune.combined)[rownames(immune.combined) %in% unique(ref[seqnames(ref) == "7" & start(ranges(ref))> 169649612 & start(ranges(ref)) < 169657348,]$gene_name)]

plot_seperate_features(immune.combined, gene  = c("ENSMMUG00000052391-unknown"), ident = "group")
plot_seperate_features(immune.combined, gene  = c("MSTRG.132309-unknown"), ident = "group", colors =c('#DCD9DE','navy'))


library(biomaRt)
ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl") #uses human ensembl annotations
#gets gene symbol, transcript_id and go_id for all genes annotated with GO:0007507
gene.data <- getBM(attributes=c('hgnc_symbol', 'ensembl_transcript_id', 'go_id'),
                   filters = 'go_id', values = 'GO:0007507', mart = ensembl)
```






