---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Differential expression analysis IN VIVO

```{r Imports}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(dplyr))
suppressMessages(library(Seurat))
suppressMessages(library(Matrix))
suppressMessages(library(SingleCellExperiment))
suppressMessages(library(stringr))
suppressMessages(library(rtracklayer))
suppressMessages(library(RColorBrewer))
suppressMessages(library(ggplot2))
suppressMessages(library(scales))
suppressMessages(library(circlize))
suppressMessages(library(reshape2))
suppressMessages(library(ComplexHeatmap))

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Imports
source("../../utils/02_sc_utils.R")
source("../../utils/03_de_utils.R")
source("../../utils/04_utils_graph.R")

# Gene annotation
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE_neut//")

# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))

# Seurat object
immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready_neut.rds")

# Ortholgs
orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs.rds")

# -------------------------------------------
#    Define threshold for the whole analysis
# -------------------------------------------

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)


ref_nonovel <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release92/RheMac8/Macaca_mulatta.Mmul_8.0.1.92.gtf"))


DimPlot(immune.combined)

calc_correlation_genes <- function(gene1_vector,gene2_vector,gene1, gene2, type = "spearman"){
  mask <- gene1_vector !=0 & gene2_vector !=0
  # calculate the correlation coefficient
  if(sum(mask)>50){
    #pc <- bayes.cor.test(c1, c2)
    pc <- cor.test(gene1_vector[mask], gene2_vector[mask], method = c(type))
    df <- data.frame( pval=pc$p.value, rho=pc$estimate)
    df$g1 <- gene1
    df$g2 <- gene2
    return(df)
  }else{
    return(NA)
  }
}



expand.grid.unique <- function(x, y, include.equals=FALSE){
  x <- unique(x)
  y <- unique(y)
  g <- function(i)
  {
    z <- setdiff(y, x[seq_len(i-include.equals)])
    if(length(z)) cbind(x[i], z, deparse.level=0)
  }
  do.call(rbind, lapply(seq_along(x), g))
}

de_all<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds")
de_lnc <- de_all[de_all$type == "lnc", ]
calc_z <- function(expression, na.remove = F){
      mean <- rowMeans(expression, na.rm = na.remove); sd <- apply(expression, 1 , function(x) sd(x, na.rm = na.remove));
      z <- (expression-mean)/sd
      return(z)
}

mono <- subset(immune.combined, ident = "Monocyte")
pal <- c(brewer.pal(8, "Pastel2")[5:9])
pal_g <- c(rep(pal[1],3), pal[2], rep(pal[3],2), rep(pal[4],3))
```


# Load DE files
```{r readfilesDE}
de_all_genes <- readRDS(file.path(robjectsdir, "de_all_genes.rds"))
de_all <- readRDS( "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE//de_all_stages.rds")
stats_complete <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE//stats_complete.rds")
m_original <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE//m.rds")


de_lnc_names <- unique(de_all[de_all$type == "lnc",]$primerid)
de_pc_names <- unique(de_all[de_all$type == "pc",]$primerid)
de_names <- (c(de_pc_names, de_lnc_names))

types <- data.frame(distinct(de_all[, c("type", "primerid")]))
rownames(types) <- types$primerid
```


# ----------------------------------------------------------------------------------
#    LncRNAs are involved in co-regulatory network involved in MHCII presentation  
# ----------------------------------------------------------------------------------

# 1. Collect all genes involved in MHCII in rheMac10
```{r collectGenes}
library(plyranges)
mamu_dr <- c("MAMU-DRA", "MAMU-DRB1")
# Identify MHC II complex genes ( Prefix == "MAMU" )
# Not all the genes which are annotated in Human or in rheMac8 are present in rheMac10
# We just use the one present in the rheMac10 and should be a proxy for the mhc2 complex expression
mhc <- unique(ref$gene_name[grepl("MAMU", ref$gene_name)])
mhc2 <- mhc[grepl("D", mhc)]
mhc1 <- setdiff(mhc,mhc2)
# Code for using also the gene annotated in RheMac8
#genes_mhc2_v8 <- unique(ref_nonovel[ref_nonovel$gene_name %in% unique(ref_nonovel$gene_name[grepl("MAMU", ref_nonovel$gene_name)]),]$gene_id)
#unique(ref[ref$gene_id %in% genes_mhc2_v8 & seqnames(ref) == "4",]$gene_name)

# MHCII genes are DE
de_mhc2 <- de_all[de_all$primerid %in% mhc2, c("primerid","comparison") ]

# LNC OR CLOSE TO MHCII REGION  
# Check wether there are lncRNAs in MHCII region ( 2mbp up- and downstream )
mhc2_range <- c(min(start(ranges(ref[ref$gene_name %in% mhc2,])))-2000000, max(start(ranges(ref[ref$gene_name %in% mhc2,])))+2000000)
ref_mhc2 <- ref[seqnames(ref) == "4" & start(ref) > mhc2_range[1] & end(ref) < mhc2_range[2],]

novel_lnc_mhc2 <- ref_mhc2[is.na(ref_mhc2$gene_biotype),]
annot_lnc_mhc2 <- ref_mhc2[!is.na(ref_mhc2$gene_biotype) & ref_mhc2$gene_biotype == "lncRNA",]
lnc_mhc2_region <- c(novel_lnc_mhc2, annot_lnc_mhc2)
```


# Visualize that DR complex goes down
```{r heatmap}
# Calculate z-scores for MHCII genes
get_df_z <- function(gene, mono, z = F ){
  expression_gene <- mono[rownames(mono) == gene, ]@assays$RNA@data
  colnames(expression_gene) <- mono$dpi
  dpis <- mono$dpi
  zmatrix <- as.matrix(calc_z(expression_gene, T))
  df <- reshape2::melt(as.matrix(zmatrix))
  names(df) <- c("gene", "group", "value")
  df_real <- reshape2::melt(as.matrix(expression_gene))
  names(df_real) <- c("gene", "group", "value")
  df_real$gr <- mono$group
  df_real <- df_real[df_real$value > 0,]
  if(z == T ){
    return(df)
  }else{
    return(df_real)
  }

}


# Plot MHC II DR complex
df_mhc <-Reduce("rbind", lapply(unique(mhc2), function(gene2) get_df_z(gene2, mono, z = F ) ))
# Merge the baselines
df_mhc$group <- ifelse(as.character(df_mhc$group) %in% c("D-30","D-04"), "D000", as.character(df_mhc$group))
df_mhc$group <- factor(df_mhc$group, levels = c("D-30","D-04","D000","D003","D004","D005","D006","D007","D008" ))

df_mamu <- df_mhc[df_mhc$gene %in% mamu_dr, ]
pal <- c(brewer.pal(8, "Pastel2")[5:9])
pal_g <- c(rep(pal[1],3), pal[2], rep(pal[3],2), rep(pal[4],3))
ggplot(df_mamu, aes( x = group, y = value, color = group, fill = group ) )+theme_paper+theme(axis.text.x = element_text(size =10 ))+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+scale_fill_manual(values=pal_g )+
  stat_summary(
    fun.y = median,
    geom = 'line',
    aes(group = gene, colour = gene)
  )+xlab("")+ylab("Z-score")+ggtitle("MAMU-DR (MAMU-DRA + MAMU-DRB1)")+scale_color_manual(values= c(brewer.pal(8, "Set2")))

```




```{r readfilesDE}
# CORRELATIONS
# Check for gene showing correlation w/ mhcII complex genes in myeloids
# Create all the pairs
cor_lnc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/01_spearman/Spearman_correlations_MAMU_alllnc_ Monocyte .rds")
cor_pc <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/01_spearman/Spearman_correlations_MAMU_allpc_ Monocyte .rds")

# Check fo significant correlations
c_lnc <- Reduce("rbind", cor_lnc)
c_pc <- Reduce("rbind", cor_pc)
c_lnc$type <- "lnc"
c_pc$type <- "pc"
c <- rbind(c_lnc,c_pc)
df_cor <- c[!is.na(c$pval), ]
df_cor_sig <- df_cor[df_cor$pval < 0.05,]
```

```{r readfilesDE}
# -------------------------------------------
# Top 2 Positively correlated PC
# -------------------------------------------
df_cor_sig_pos <- df_cor_sig[df_cor_sig$rho > 0, ]
pos_10_pc <- unique(df_cor_sig_pos[order(df_cor_sig_pos$pval, decreasing = F),]$g1)[1:2]
df <-Reduce("rbind", lapply((pos_10_pc), function(gene2) get_df_z(gene2, mono, z = F ) ))
df$group <- factor(df$group, levels = c("D-30","D-04","D000","D003","D004","D005","D006","D007","D008"))
df$gr <- factor(df$gr, levels = c("baseline", "early", "middle","late"))

ggplot(df, aes( x = gr, y = value, color = gr, fill = gr ) )+theme_paper+theme(axis.text.x = element_text(size =10 ))+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+scale_fill_manual(values=pal_g )+
  stat_summary(
    fun.y = median,
    geom = 'line',
    aes(group = gene, colour = gene)
  )+xlab("")+ylab("logCP10K")
```

```{r readfilesDE}
# -------------------------------------------
# Top 10 Positively correlated lnc
# -------------------------------------------
df_cor_sig_pos <- df_cor_sig[df_cor_sig$type == "lnc" & df_cor_sig$rho > 0, ]
pos_10_lnc <- unique(df_cor_sig_pos[order(df_cor_sig_pos$pval, decreasing = F),]$g1)[1:5]
df <-Reduce("rbind", lapply((pos_10_lnc), function(gene2) get_df_z(gene2, mono, z = F ) ))
df$group <- factor(df$group, levels = c("D-30","D-04","D000","D003","D004","D005","D006","D007","D008"))
df$gr <- factor(df$gr, levels = c("baseline", "early", "middle","late"))

ggplot(df, aes( x = gene, y = value, color = gr, fill = gr ) )+theme_paper+theme(axis.text.x = element_text(size =10 ))+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+scale_fill_manual(values=pal )+
  stat_summary(
    fun.y = median,
    geom = 'line',
    aes(group = gene, colour = gene)
  )+xlab("")+ylab("logCP10K")
```

# Analysis: which are the genes close to MHCII complez

```{r readfilesDE}
# Colocation
# Correlation
cis_mono[grepl("MAMU", cis_mono$gene_name),]$orth
unlist(lapply(pos_10_lnc, get_orthologname_))

# Which ones do i find correlated with MAMU DR complex in Monocytes?
df_cor_sig_pos$orth <- unlist(lapply(df_cor_sig_pos$g1, get_orthologname_))
genes_corr  <- unique(df_cor_sig_pos[order(df_cor_sig_pos$pval, decreasing = F), "g1"])[1:5]


# Ones in graph comm antigen presentation

gD2 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut//gD2.rds")
lou <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut//lou.rds")

genes_antigen <- rownames(gD2[V(gD2)$community  == 3])
genes_mm <- rownames(gD2[V(gD2)$community  == 6])
saveRDS(genes_antigen , "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut//genes_antigen.rds")
saveRDS(genes_mm  , "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph_neut//genes_mm.rds")

```


```{r readfilesDE}
mono$group <- factor(mono$group, levels = c("baseline", "early", "middle", "late"))
DotPlot(mono, features = c(mamu_dr, pos_10_pc[1:2],unique(pos_10_lnc), get_gene_id("HCG24")), group.by = "group", dot.scale =10,scale.by = "radius", cols = c( "lightblue", "darkred"))+coord_flip()+theme_paper+xlab("")+ylab("")+theme_bw()+ggtitle("Expression in Monocytes")


plot.data_up <- Dotplot_data(mono, features = c(mamu_dr, pos_10_pc[1:2],unique(pos_10_lnc) ), group.by = "group",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname_))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))



get_gene_id <- function(gene){
  gene_id <- unique(ref[!is.na(ref$transcript_id) & ref$transcript_id %in% unique(orthologs[orthologs$orthologGeneSymbol == gene,]$lnc),]$gene_name)
  if(length(gene_id) != 0 ){
    gene_id  <- rownames(subset(immune.combined, features = c(gene_id)))
    return(gene_id)
  }else{
    return(gene)
  }
}

# Plot the strongest correlated ones
genes_mm <- ifelse(substr(genes_mm,1,3) %in% c("ENS", "MST"), paste( genes_mm,"-unknown", sep = ""), genes_mm)
genes_mm <- gsub(".", "-", genes_mm, fixed = T)
plot.data_up <- Dotplot_data(mono, features = c( genes_mm ), group.by = "group",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname_))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
plot.data_up$id <- gsub("late", "ulate", as.character(plot.data_up$id))
heatmap_celltypes(plot.data_up)

# Do they also show correlation values ?
df_cr_dra <- 	df_cor[df_cor$g2 == "MAMU-DRA",]
df_cr_dra[df_cr_dra$g1 %in% genes_antigen, ]


```


# NEAT 1
```{r readfilesDE}
plot.data_up <- Dotplot_data(mono, features = c(get_gene_id("NEAT1"), "CXCL8" ), group.by = "group",scale.by = "radius", cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
plot.data_up$id <- gsub("late", "ulate", as.character(plot.data_up$id))
heatmap_celltypes(plot.data_up)
```


```{r readfilesDE}
df <-Reduce("rbind", lapply(c( mamu_dr, get_gene_id("NEAT1")), function(gene2) get_df_z(gene2, mono, z = F ) ))
df$group <- factor(df$group, levels = c("D-30","D-04","D000","D003","D004","D005","D006","D007","D008"))
df$gr <- factor(df$gr, levels = c("baseline", "early", "middle","late"))

ggplot(df, aes( x = gene, y = value, color = group, fill = group ) )+theme_paper+theme(axis.text.x = element_text(size =10 ))+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+scale_fill_manual(values=pal_g )+
  stat_summary(
    fun.y = median,
    geom = 'line',
    aes(group = gene, colour = gene)
  )+xlab("")+ylab("logCP10K")




get_expression <- function(gene, mono = monocyte, zeros = F ){
  if(!(gene %in% rownames(mono) )){
    return("Gene id not found!")
  }
  expression_gene <- mono[rownames(mono)== gene, ]@assays$RNA@data
  df <- reshape2::melt(as.matrix(expression_gene))
  names(df) <- c("gene", "cell", "value")
  df_individual <- data.frame(mono$individual)
  df_individual$cell <- rownames(df_individual)

  df_dpi <- data.frame(mono$dpi)
  df_dpi$cell <- rownames(df_dpi)

  df_dpi <- data.frame(mono$dpi)
  df_dpi$cell <- rownames(df_dpi)

  df <- merge(df_individual, df, by="cell")
  df <- merge(df_dpi, df, by = "cell")

  if(zeros == F ){
    df <- df[df$value !=0,]
  }
  return(df)
}



df <- get_expression(get_gene_id("NEAT1"), mono =mono)
ggplot(df, aes( x = mono.dpi, y = value, color = mono.individual, fill = mono.individual ) )+theme_paper+theme(axis.text.x = element_text(size =10 ))+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+xlab("")+ylab("logCP10K")


p <- df %>% dplyr::group_by(mono.individual, mono.dpi) %>% summarise(  mean = mean(value))
p$type <- ifelse(as.integer(gsub("D", "", p$mono.dpi)) < 1, "Baseline", as.character(p$mono.dpi))
p$comp <- -1

for(i in 1:nrow(p)){
 v <- p[i,]$mean/ p[p$mono.individual == p[i,]$mono.individual & p$type == "Baseline",]$mean  
 if(length(p[i,]$mean/ p[p$mono.individual == p[i,]$mono.individual & p$type == "Baseline",]$mean) == 0 ){
   p[i,]$comp <- NA
 }else{
   p[i,]$comp <- p[i,]$mean/ p[p$mono.individual == p[i,]$mono.individual & p$type == "Baseline",]$mean  
 }

}

ggplot(p, aes( x = type, y = comp, col = mono.individual, group = mono.individual))+geom_point()+ geom_line()+geom_hline(yintercept= 1, col = "grey")+theme_paper

a <- data.frame(mono$individual, mono$dpi)
distinct(a)
```
