---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---


# Differential expression analysis IN VIVO

```{r Imports}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(wesanderson)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(ggplot2)
library(scales)
library(circlize)
library(reshape2)
library(Gviz)

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Imports
source("../../utils/00_datapaths.R")
source("../../utils/02_sc_utils.R")
source("../../utils/03_de_utils.R")
source("../../utils/04_utils_graph.R")

# Seurat object
immune.combined <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds"))

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)

orthologs <- readRDS(file.path(data_path, "01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds"))
mono <- subset(immune.combined, ident = "Monocyte")

# ----------------------------------------------------
# 1. Check ortholog 
# ----------------------------------------------------

neat1 <- get_gene_id("NEAT1")

infection_info <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))

table(colnames(immune.combined) == rownames(infection_info))
immune.combined$viralperc <- infection_info$percentage_viral_reads
immune.combined$infectionstatus <- infection_info$classification
```

# Is NEAT1 Downregulated? (IN VIVO )

```{r Fisher}

# ----------------------------------------------------
# 2. Visualize downregulation general 
# ----------------------------------------------------
plot.data_up <- Dotplot_data(mono, features = c(neat1, "IFI27" ), group.by = "group",scale.by = "radius", dot.scale = 20,cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp.scaled))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
plot.data_up$id <- gsub("late", "ulate", as.character(plot.data_up$id))

pdf(file.path(plots, "05/neat1.pdf"), width = 3, height = 2)
heatmap_celltypes(plot.data_up)
dev.off()



# ----------------------------------------------------
# 3. Visualize downregulation per human-primate
# ----------------------------------------------------
df <- get_expression(get_gene_id("NEAT1"), mono =mono)
df$mono.sample <- paste(df$mono.individual, df$mono.dpi, sep = " ")
p <- df %>% dplyr::group_by(mono.individual, mono.dpi, mono.sample) %>% dplyr::summarise(  mean = mean(value))
p$mono.dpi <- unlist(lapply( as.character(p$mono.dpi), function(x) paste("D",as.integer(gsub("D", "", x)),sep = "")))

p$type <- ifelse(as.integer(gsub("D", "", p$mono.dpi)) < 1, "Baseline", as.character(p$mono.dpi))

p$comp <- -1

for(i in 1:nrow(p)){
 v <- p[i,]$mean/ p[p$mono.individual == p[i,]$mono.individual & p$type == "Baseline",]$mean  
 if(length(p[i,]$mean/ p[p$mono.individual == p[i,]$mono.individual & p$type == "Baseline",]$mean) == 0 ){
   p[i,]$comp <- NA
 }else{
   p[i,]$comp <- p[i,]$mean/ p[p$mono.individual == p[i,]$mono.individual & p$type == "Baseline",]$mean  
 }
}

p <- p[!is.na(p$comp),]
plot_neat <- ggplot(p, aes( x = type, y = comp, col = mono.individual, group = mono.individual))+geom_point(col = "#666666")+ geom_line()+geom_hline(yintercept= 1, col = "grey", lty=2)+theme(legend.position = "")+ylab("mean expression / baseline expression")+xlab("")+
          theme(axis.text.x=element_text(angle=20, hjust=1))+
          theme(axis.text = element_text(size = 15), axis.title = element_text(size = 15))+
          scale_color_manual(values = rep("#666666",17))+
          theme( axis.line = element_line(colour = "black"),
              axis.ticks.length=unit(.2, "cm"),
              axis.text= element_text(size = 17, color = "black"),panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(), panel.background = element_rect(fill = "white"))
plot_neat


# ----------------------------------------------------
# 3. Check viral load per sample 
# ----------------------------------------------------
viral_reads_df <- data.frame(sample = immune.combined$sample, viralperc = immune.combined$viralperc, stringsAsFactors = F) 
viral_reads_df_summary <- viral_reads_df %>% dplyr::group_by(sample) %>% dplyr::summarise(sum_viral = sum(viralperc))
viral_reads_df_summary_plot <- viral_reads_df_summary[viral_reads_df_summary$sum_viral > 0, ]
viral_reads_df_summary_plot$sample <- factor( viral_reads_df_summary_plot$sample, levels = c(as.character(viral_reads_df_summary_plot[order(viral_reads_df_summary_plot$sum_viral), ]$sample)))
ggplot(viral_reads_df_summary_plot, aes( x = sample, y = sum_viral))+geom_bar(stat = "identity")+coord_flip()+theme_pander()

p$sample <- p$mono.sample
merged_table <- merge(p , viral_reads_df_summary_plot, by = "sample")
merged_table$sample <- factor( merged_table$sample, levels = c(as.character(merged_table[order(merged_table$sum_viral), ]$sample)))
ggplot(merged_table[order(merged_table$sum_viral),], aes(col = sample, x = sum_viral, y = comp))+geom_point()+theme_base()
```


# Is NEAT1 Downregulated? (EX VIVO )

# 1. Load dataset 
```{r Fisher}
immune.combined_exvivo <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
mono_exvivo <- subset(immune.combined_exvivo, ident = "Monocyte")
```

# 2. 
```{r Fisher}
mono_exvivo$group <- paste(mono_exvivo$cond, mono_exvivo$dpi)
#mono_exvivo$group <- paste(mono_exvivo$cond)
# ----------------------------------------------------
# 2. Visualize downregulation general 
# ----------------------------------------------------
plot.data_up <- Dotplot_data(mono_exvivo, features = c(neat1, "IFI27" ), group.by = "group",scale.by = "radius", dot.scale = 20,cols = c( "lightblue", "darkred"))
plot.data_up$name <- unlist(lapply(plot.data_up$features.plot, get_orthologname))
ggplot(plot.data_up, aes(x = id,y = name, size= pct.exp, col = avg.exp
))+geom_point()+theme_paper+xlab("")+xlab("")+scale_colour_gradient(low = "lightblue", high = "dark red", na.value = NA)+theme(axis.text.x = element_text(angle = 20, vjust = 0.9, hjust=1))
#Heatmap
heatmap_celltypes(plot.data_up)


```


# 3. Correlation NEAT1 and IFI

```{r Fisher}

expressionmatrix <- mono_exvivo@assays$RNA@counts
colnames(expressionmatrix) <- Idents(mono_exvivo)

calc_correlation_genes <- function(gene1_vector,gene2_vector,gene1, gene2, type = "spearman"){
  mask <- gene1_vector !=0 & gene2_vector !=0
  # calculate the correlation coefficient
  if(sum(mask)>30){
    #pc <- bayes.cor.test(c1, c2)
    pc <- cor.test(gene1_vector[mask], gene2_vector[mask], method = c(type))
    df <- data.frame( pval=pc$p.value, rho=pc$estimate)
    df$g1 <- gene1
    df$g2 <- gene2
    return(df)
  }else{
    return(NA)
  }
}

genes <- c(get_gene_id("NEAT1"), "IFI27")
calc_correlation_genes(expressionmatrix[genes[1], ], expressionmatrix[genes[2],], genes[1], genes[2])


```
