---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(scater)
library(SingleCellExperiment)
library(purrr)
library(stringr)
library(mvoutlier)
#library(dropbead)
library(scater)
library(stringr)
library(scran)
library(cowplot)
library(rtracklayer)
#library(org.Mmu.eg.db)
library(graphics)
#library(RCy3)


# Define paths for data
source(file.path("../utils/00_datapaths.R"))
# Import Utils
source("../utils/02_sc_utils.R")

theme_sc <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "black"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))


# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))


# Seurat Object - preprocessed
immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/01_immune.combined_post_scrublet_mdkremoved__02_integrated_dimred_all.rds")
dim(immune.combined)
DimPlot(immune.combined, reduction = "umap", label = TRUE)+theme_sc
sum(table(immune.combined$dpi))

```




# -----------------------------------------------------
#    Find markers thoufh DE against all other clusters 
# -----------------------------------------------------


```{r Find markers}

# FIND MARKERS  for every cluster compared to all remaining cells, report only the positive ones
#pbmc.markers <- FindAllMarkers(immune.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
#saveRDS(pbmc.markers, file.path(robjectsdir,"markers_invivo_normal.rds"))
#pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)



b <- c("CD79B", "MS4A1", "CD19", "IGHM")
CD8T <- c("CD3D", "GZMB", "GNLY")
CD4T <-  c("CD3D", "IL7R")
t <- c(CD4T, CD8T)
nk <- c("KLRB1", "GZMB","FCGR3")
mono <- c("LYZ", "PSAP", "CFD") 
neut <- c( "CD177","LCN2") 
pDC <- c("IRF8", "FLT3")

marker.genes_red <- c(t,b,mono,neut)

saveRDS(marker.genes_red, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/marker.genes.rds")


FeaturePlot(immune.combined, pDC, order = T )
FeaturePlot(immune.combined, nk, order = T )
plot_genes(immune.combined,mono,threshold= 2, title = "Monocytes marker expression", col = "purple")
plot_genes(immune.combined, b, threshold = 2 , title = "B markers expression", col = "purple")
plot_genes(immune.combined,t, threshold = 2 , title = "T-Cell markers expression", col = "purple")
plot_genes(immune.combined, neut, threshold = 3 , title = "Neutrophil-Cell markers expression", col = "purple")

immune.combined <- RenameIdents(immune.combined, `0` = "B", `1` = "T", `2` = "T", 
   `3` = "T", `4` = "B", `5` = "Monocyte", `6` = "Neutrophil", `7` = "Monocyte", `8` = "T", `9` = "Monocyte",
    `10` = "T", `11` = "T", `12` = "Monocyte", `13` = "T", `14` = "Neutrophil", `15` = "Monocyte",`15` ="T", `16` ="B", `17` ="T", `18` ="T")

saveRDS(immune.combined, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")

#immune.combined_noneut <- RenameIdents(immune.combined, `Neutrophil` = "Myeloid", `Monocyte` = "Myeloid")
#saveRDS(immune.combined_noneut, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")
```



# Visualize Marker genes

```{r Find markers}
# Visualize Markers
DotPlot(immune.combined, features = unique(marker.genes_red), cols = c("navy", "#FFD533", "grey"), dot.scale = 11) + RotatedAxis()+theme(axis.text = element_text(size = 20), axis.title = element_blank())+ theme(panel.background = element_rect(fill = "white", colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=21))


DotPlot(immune.combined, features = unique(marker.genes_red), cols = c("grey", "dark red", "white"), dot.scale = 11) + RotatedAxis()+theme(axis.text = element_text(size = 20), axis.title = element_blank())+ theme(panel.background = element_rect(fill = "white"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=21))

library(wesanderson)
pal_celltypes <- brewer.pal(4, "Set2")
pal_celltypes <-wes_palette("GrandBudapest1", 4)
pal_celltypes[3] <- "#AE4E4E"
pal_celltypes <- pal_celltypes[c(2,1,3,4)]
# Plot with correct
dimplot <- DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =pal_celltypes, label.size = 9)+theme_minimal()+ theme(panel.background = element_rect(fill = "white"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+theme(text = element_text(size=18))+theme_sc+theme(axis.line = element_blank())
```
# -------------------------------------
#    Visualize Data through UMAP
# -------------------------------------

```{r UMAPvisualization}

immune.combined$group_dpi <- gsub("DPI-04", "DPI000", immune.combined$group_dpi)
immune.combined$group_dpi <- gsub("DPI-30", "DPI000", immune.combined$group_dpi)
immune.combined$group_dpi_plot <- gsub("DPI00", "", immune.combined$group_dpi)
# Day post infection 
colfunc <- colorRampPalette(c("black", "white"))
pal_grad <-plot(rep(1,10),col=colfunc(10),pch=19,cex=3)

dpi <- DimPlot(immune.combined, reduction = "umap", group.by = "group_dpi_plot", pt.size = 0.1, cols =rev(colfunc(9)[1:7]))+theme_sc

# Individual 
DimPlot(immune.combined, reduction = "umap", group.by = "individual")+theme_sc


immune.combined$batch <- gsub("FRZ", "Frozen", immune.combined$batch)
immune.combined$batch <- gsub("fresh", "Fresh", immune.combined$batch)
# Batch 
DimPlot(immune.combined, reduction = "umap", group.by = "batch", pt.size = 0.01,cols = (brewer.pal(2, "Paired")[1:2]))+theme_sc

DefaultAssay(immune.combined)
# Sample
DimPlot(immune.combined, reduction = "umap", group.by = "sample", pt.size = 0.00001)+theme_sc
```
# -----------------------------
#     Identify EbolaGenes
#-----------------------------

```{r cars}
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unknown", sep ="-")]


lapply(ebola_genes, function(gene) FeaturePlot(immune.combined, features  = gene, sort.cell = T, pt.size = 0.7)+theme_sc)
plot_genes(immune.combined, ebola_genes,1, col = "#DA0202", "Cells expressing Ebola RNA")

novel <- ref[substr(ref$gene_id,1,2) =="MS", ]$gene_name
lnc <- ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "lncRNA",]$gene_name
found <- rownames(immune.combined)[rownames(immune.combined) %in% c(lnc, novel)]
table(substr(ref[ref$gene_name %in% found, ]$transcript_id,1,3))
```

# ---------------------------------
#     Identify lncRNAs and PC genes 
#----------------------------------

```{r cars}
# ------------------------------------
#     PC genes present in object 
# -------------------------------------
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% unique(ref$gene_name[ref$gene_biotype == "protein_coding"])]

# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# Annotated lncrnas id 
lnc_ref_id <- gsub("_","-",unique(ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "lncRNA",]$gene_name))
unique(ref[ref$gene_name == lnc_ref_id[substr(lnc_ref_id,1,3) != "ENS"],]$gene_biotype)

# Annotated lncRNAs: remove the gene names so it can match 
annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc_ref_id]
# Novel lncRNAs
novel_lncrnas <- rownames(immune.combined)[unlist(lapply(rownames(immune.combined), function(x) startsWith(x, "MSTRG")))]
# Complete Set 
all_lncrnas <- c(annotated_lncrnas, novel_lncrnas)

#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated", "novel")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)), length(unique(novel_lncrnas))))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Dark2")
lnc_donut <- df.donut[df.donut$type == "lncRNAs",]
names(lnc_donut) <- c("-", "type", "count")
donut_plot(lnc_donut, "Greens")



# Save
robjectsdir<- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats"
dir.create(file.path(robjectsdir), showWarnings = FALSE)
saveRDS(all_lncrnas, file.path(robjectsdir, "all_lncrnas.rds"))
saveRDS(annotated_mrnas, file.path(robjectsdir, "annotated_mrnas.rds"))
```
# ---------------------------------
#    Remove Ebola Genes  
#----------------------------------
```{r cars}
#length(setdiff(rownames(immune.combined_noneut), ebola_genes))
#immune.combined_noneut_ebolaremoved <- immune.combined_noneut[setdiff(rownames(immune.combined_noneut), ebola_genes),]
#immune.combined_noneut_ebolaremoved <- NormalizeData(immune.combined_noneut_ebolaremoved)
```

