---
title: "Untitled"
author: "Luisa Santus"
date: "2/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# After having merged them , plt stats!

# ------------------------------------
# stats 
```{r results='hide', message=FALSE, warning=FALSE}

novel_all <- import(file.path(datadir, "03_novel_lncrnas/novel_expressed_all.gtf"))
novel_transcripts_codes <- novel_all[novel_all$type == "transcript",]$class_code
total <- length(novel_transcripts_codes)
prop_intergenic <- (length(novel_transcripts_codes[novel_transcripts_codes == "u"])/total)*100
prop_antisense <- (length(novel_transcripts_codes[novel_transcripts_codes == "x"])/total)*100
print(prop_intergenic)
print(prop_antisense)
plots <- plot_stats_annotation(novel_all,lncRNAs_ref,lncRNAs_ref_human, mrna_ref_human, mRNAs_ref )

plot_stats_annotation(remove_one_exon(novel_all),remove_one_exon(lncRNAs_ref),remove_one_exon(lncRNAs_ref_human), remove_one_exon(mrna_ref_human), remove_one_exon(mRNAs_ref))
# --------------
#  Expression
dir_counts_ref = file.path(datadir, "04_Quantification_correct/")
htseq_files_ref <- iterate_files(dir_counts_ref, ".gene_counts.tab")


# Calc Expression 
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
expression_logcpm <- expression_logcpm[mask,]

max_expression <- data.frame(id=rownames(expression_logcpm ), expr=rowMax(expression_logcpm))

# Add labels
max_expression['type'] <- "0"; 
max_expression <- add_type(max_expression, novel_all$gene_id, "novel")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "lncrna") ; max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mrna") ; max_expression <- max_expression[!max_expression$type == "0",]


plot_expression(max_expression)

length(unique(novel_all$gene_id))

```

# -----------------------------
# Tissue Sharing 
# -----------------------------


```{r results='hide', message=FALSE, warning=FALSE}

palette_expression <-c( "#E9A3C9","#F4E3ED", "#E6F5D0")

dds_ref[,dds_ref$tissue == "Testis"]$tissue <- "Testes"

expression <- sapply(dds_ref, function(dds_ref) length(unique(dds_ref[,assays(dds_ref)$logCPM > 1]$tissue)))
saveRDS(expression, "results/expression_for_tissuesharing.rds")

expression <- readRDS("results/expression_for_tissuesharing.rds")

lncnras_expressed <- c(lncnras_expressed_polyA, lncnras_expressed_ribodepl) 
length(mrnas_expressed)
mrnas_expressed <- c(mrnas_expressed_polyA, mrnas_expressed_ribodepl)
# Create DF for plotting
df <- as.data.frame(as.data.frame(expression))
df$id <- rownames(df)

df$type <- "0"
df <- add_type(df, novel_all$gene_id, "novel")
df <- add_type(df, lncnras_expressed, "lncrna")
df <- add_type(df, mrnas_expressed, "mrna")
df <- df[!df$type == "0",]
table(df$type)

# Boxplot Tissue Sharing
p1 <- ggplot(df, aes(x = factor(type, levels = c("lncrna", "novel", "mrna")), y = expression, fill = type)) +
            geom_boxplot(color = "grey", fill = palette_expression)+theme_pubclean()+labs(x = "", y = "# Tissues", title = "Expression across tissues")+
            theme(legend.title=element_blank())+
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1


# Density plot tissue sharing

# Boxplot Tissue Sharing
p <- ggplot(df, aes(x = expression, fill = factor(type, levels = c("lncrna", "novel", "mrna")), color = factor(type, levels = c("lncrna", "novel", "mrna")))) +labs(y = "", x = "# Tissues", title = "Expression across tissues (Tissue specificity)")+
            theme_pubclean()+geom_density(alpha=0.7)+
            theme(legend.title=element_blank())+
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+theme(plot.title = element_text(hjust = 0.5))
p + scale_fill_manual(values=palette_expression) + scale_color_manual(values=palette_expression)


df

barplot_tissues <- function(df, type, col){
  ## extract the number of exons
  df_l <- df[df$type == type,]
  h <- ggplot(df_l, aes(x=expression, fill=type)) + 
    geom_histogram(position="identity", binwidth =1)+
    xlim(1,16)
  
  h_plotdata <- ggplot_build(h)$data[[1]]
  h_plotdata$group <- as.factor(h_plotdata$group)
  levels(h_plotdata$group) <- c(type)
  
  ## plot with geom_bar
  p1 <-ggplot(h_plotdata, aes(x=x, y=y, fill = group, palette = col)) +
    geom_bar(stat = "identity", width = 0.8) +
    theme(legend.title=element_blank())+
    labs(y = "", x = "")+
    theme(legend.title=element_blank())+ theme(legend.position = "none")+
    scale_x_continuous( labels = as.character(h_plotdata$x), breaks = (h_plotdata$x)) + 
    theme(plot.title = element_text(hjust = 0.5))+
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
    scale_fill_manual(values=c(col))
  return(p1)
}

palette <- palette_expression
tl1 <-barplot_tissues(df,"novel",palette[1]) 
tl2 <-barplot_tissues(df,"lncrna",palette[2]) + labs( y  = "Count")
tl3 <-barplot_tissues(df,"mrna",palette[3]) + labs( x = "#Tissues")
b <- ggarrange( tl1,tl2,tl3,  ncol=1, nrow=3, top = "Expression across tissues (Tissue specificity)",  legend="right")
b
```

```{r results='hide', message=FALSE, warning=FALSE}

gr <- novel_all
gr <- gr[gr$type =="exon",]
df <- data.frame("gene_id" = gr$gene_id,"exon_number" = as.numeric(gr$exon_number))
number_exons <- df %>% dplyr::group_by(gene_id) %>% dplyr::summarize(max_exon = max(exon_number))


df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
collapsed <- df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))

all <- merge(number_exons, collapsed, by=0, all=TRUE)


# Boxplot Tissue Sharing
p <- ggplot(all, aes(x = range, fill =factor(max_exon), color = factor(max_exon))) +labs(y = "", x = "Transcript Length", title = "Trancript length per number of exons")+
            geom_density(alpha=0.5)+theme_pubclean()+
            theme(legend.title=element_blank())+
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+theme(plot.title = element_text(hjust = 0.5))+xlim(0, 20000)
p 
```

