---
title: "lncrna annotation "
author: "Luisa Santus"
date: "12/19/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data and libraries

```{r results='hide', message=FALSE, warning=FALSE}
library(rtracklayer); library(VennDiagram); library(stringr); library(ggplot2); library(grid); library(gridExtra); library(UpSetR); library(ComplexHeatmap)
library(GenomicRanges); library(dplyr); library(zeallot); library(DESeq2); library(edgeR);  library("ggplotify"); library("grid"); library(tidyverse)
library(tximport); library(tximportData); library(plyr); library(Gviz); library(ggpubr)

# Reused paths
projdir = "/home/luisas/Desktop/cluster/proj/data/"
datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all"
assemblies <- "/home/luisas/Desktop/cluster/assemblies"

# Import Utils 
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/","code/ebola/src/scripts/Analysis/utils")
source(file.path(scripts_dir,"de_utils.R")); source(file.path(scripts_dir,"lncrna_annotation_utils.R"))

# Human training set
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownlncRNAs.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownMrnas.gtf"))
lncRNAs_ref_human_gene <- lncRNAs_ref_human[lncRNAs_ref_human$type == "gene",]

# The lncRNAs and mRNAs in the reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_mrnas.gtf"))

human_annotation <- import(file.path(assemblies, "ensembl/release-96/Homo_sapiens_hg38/Homo_sapiens.GRCh38.96.gtf"))
```


First, apply the filtering to each dataset separately 
 - identify novel transcripts ( gff compare "u", "x")
 - deplete non-concordant predictions
 - retain only ones expressed in more than 3 samples
 
```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------
#     RiboDepleted
# ---------------------


# -----------------
# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
dir_counts_ref = file.path(datadir, "02_RNA-Seq_BatchZyagen/06_counts_all")
htseq_files_ref <- iterate_files(dir_counts_ref, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

palette_expression <-c( "#E9A3C9","#F4E3ED", "#E6F5D0")

# -----------------
# Select only the NOVEL ones (gff compare )
mask_uix <- merged_annotated$class_code == "u" | merged_annotated$class_code == "x" ;  mask_uix[is.na(mask_uix)] <- FALSE
table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

# -----------------
# Remove Non-Concordant
df <- data.frame()
df <- rbind(df, data.frame(gene_id = novel$gene_id, transcript_id = novel$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
novel_concordant <- novel[!(novel$gene_id %in% unconcordant_prediction$gene_id)]
print("Novel - concordant")
print(length(unique(novel_concordant$transcript_id)))

# Export the novel Concordant 
export(novel_concordant, (file.path(datadir,"03_novel_lncrnas/00_gtf_batchzyagen/novel_rhemac10_concordant_ribodepleted.gtf")))

# -----------------
# Expression Levels ( filter and plot )
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
# Calc logcpm 
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
# Create dataframe w/ logcpms
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
colnames(expression_logcpm)
## ADD only if we want to select a certain tissue
#expression_logcpm <- expression_logcpm[,str_detect(colnames(expression_logcpm), "Liver")]
#expression_logcpm

# select only when > than 1 at least 3 tissues
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
expression_logcpm <- expression_logcpm[mask,]

# Plotting
max_expression <- data.frame(id=rownames(expression_logcpm), expr=rowMax(expression_logcpm))

lncnras_expressed_ribodepl <- unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id)
mrnas_expressed_ribodepl <- unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id)

print(head(max_expression))
max_expression['type'] <- "0"
max_expression <- add_type(max_expression, novel_concordant$gene_id, "novel")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "lncrna")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mrna")
max_expression <- max_expression[!max_expression$type == "0",]

length(unique(novel_concordant[novel_concordant$gene_id %in% max_expression$id]$gene_id))
length(unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id))
length(unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id))

ggplot(max_expression, aes(x=factor(type,level = c("lncrna", "novel", "mrna")) , y=expr, fill = type )) + 
      geom_boxplot(color="darkgrey",  alpha=1.0, fill = palette_expression )+
      labs(y = "logCPM", x = "", title = "Expression" )+
      theme(legend.title=element_blank())+
      theme(plot.title = element_text(hjust = 0.5))+
      theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))

novel_expressed <- novel_concordant[novel_concordant$gene_id %in% max_expression$id]

print(paste("Novel Genes riboDepleted: ",length(unique(novel_expressed$gene_id))))
print(paste("Novel Transcripts ribodepleted: ",length(unique(novel_expressed$transcript_id))))

length(unique(novel_expressed$transcript_id))
length(unique(novel_expressed$gene_id))

#------------------
#       Save 
# -----------------
export(novel_expressed, file.path(datadir, "03_novel_lncrnas/novel_expressed_batchzyagen.gtf"))
```


#Stats 

```{r results='hide', message=FALSE, warning=FALSE}
# Palette for all plots below
palette <-brewer.pal(5,"Paired")
palette <- c("#F4E3ED", "#E9A3C9", "#C51B7D", "#E6F5D0", "#A1D76A")
levels <- c('Novel LncRNAs', 'Known lncRNAs: Macaque ', 'Known lncRNAs: Human',"Known mRNAs: Macaque", "Known mRNAs: Human" )
# --------------------
## EXON COUNT
# --------------------

ec1 <- barplot_exon_count(novel_expressed, "Novel", palette[1])
ec2 <- barplot_exon_count(lncRNAs_ref, "lncRNAs - Reference Macaque", palette[2])
ec3 <- barplot_exon_count(lncRNAs_ref_human, "lncRNAs - Reference Human", palette[3])+labs(y = "Frequency")
ec4 <- barplot_exon_count(mrna_ref_human, "mRNAs - Reference Human", palette[4])
ec5 <- barplot_exon_count(mRNAs_ref, "mRNAs - Reference Macaque", palette[5])+labs(x = "Number of Exons") 
ggarrange( ec1,ec2,ec3,ec4,ec5,  ncol=1, nrow=5, top = "Transcripts Lengths") 
# Exon count generally lower in lncrnas than mrnas (Ok. same as sources)

# --------------------
## Transcript lengths: still weird, very long ones ...
# --------------------
tl1 <-transcript_length_density(novel_expressed," NOVEL",palette[1]) 
tl2 <-transcript_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
tl3 <-transcript_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
tl4 <-transcript_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
tl5 <-transcript_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Transcript length")
ggarrange( tl1,tl2,tl3,tl4,tl5,  ncol=1, nrow=5, top = "Transcripts Lengths",  legend="right")


# --------------------
## Transcript lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_transcript_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = levels),  y = range )) +
        labs( x = "", y = "" , title = "Transcript length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,10000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1

# --------------------
## Exon lengths - Densities 
# --------------------
p1 <-exon_length_density(novel_expressed," NOVEL",palette[1]) 
p2 <-exon_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
p3 <-exon_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
p4 <-exon_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
p5 <-exon_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Exon length")
ggarrange( p1,p2,p3,p4,p5,  ncol=1, nrow=5, top = "Exon Lengths",  legend="right")

# --------------------
## Exon lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = levels),  y = range )) +
        labs( x = "", y = "" , title = "Exon length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,1000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1
```


Repeat the same analysis with polyA dataset 

```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------
#    PolyA
# ---------------------


# -----------------
# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
dir_counts_ref = file.path(datadir, "02_RNA-Seq_external/06_counts_all")
htseq_files_ref <- iterate_files(dir_counts_ref, "UMI.f3.q60.HTseq.gene_counts.tab")

# -----------------
# Select only the NOVEL ones (gff compare )
mask_uix <- merged_annotated$class_code == "u" | merged_annotated$class_code == "x" ;  mask_uix[is.na(mask_uix)] <- FALSE

table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

# save file
#export(novel, file.path(datadir, "03_novel_lncrnas/novel_after_gffcompare_polyA.gtf"))
export(novel_concordant, (file.path(datadir,"03_novel_lncrnas/00_gtf_external/novel_rhemac10_concordant_polyA.gtf")))



##  here quantification step happens outside!! 

# -----------------
# Remove Non-Concordant
df <- data.frame()
df <- rbind(df, data.frame(gene_id = novel$gene_id, transcript_id = novel$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
novel_concordant <- novel[!(novel$gene_id %in% unconcordant_prediction$gene_id)]
print("Novel - concordant")
print(length(unique(novel_concordant$transcript_id)))


# -----------------
# Expression Levels ( filter and plot )
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
# Calc logcpm 
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
# Create dataframe w/ logcpms
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
colnames(expression_logcpm)
## ADD only if we want to select a certain tissue
#expression_logcpm <- expression_logcpm[,str_detect(colnames(expression_logcpm), "Liver")]
#expression_logcpm

# select only when > than 1 at least 3 tissues
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
table(mask)
expression_logcpm <- expression_logcpm[mask,]

# Plotting
max_expression <- data.frame(id=rownames(expression_logcpm ), Means=rowMax(expression_logcpm))

print(head(max_expression))
max_expression['type'] <- "0"
max_expression <- add_type(max_expression, novel_concordant$gene_id, "novel")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "lncrna")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mrna")
max_expression <- max_expression[!max_expression$type == "0",]

length(unique(novel_concordant[novel_concordant$gene_id %in% max_expression$id]$gene_id))
length(unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id))
length(unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id))

lncnras_expressed_polyA <- unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id)
mrnas_expressed_polyA <- unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id)

ggplot(max_expression, aes(x=factor(type,level = c("lncrna", "novel", "mrna")) , y=Means, fill = type )) + 
      geom_boxplot(color="darkgrey",  alpha=1.0, fill = palette_expression)+
      labs(y = "logCPM", x = "", title = "Expression" )+
      theme(legend.title=element_blank())+
      theme(plot.title = element_text(hjust = 0.5))+
      theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))

novel_expressed <- novel_concordant[novel_concordant$gene_id %in% max_expression$id]
length(unique(novel_expressed$gene_id))


length(unique(novel_expressed$transcript_id))
length(unique(novel_expressed$gene_id))


# ---------------
#       Save
# ---------------
export(novel_expressed, file.path(datadir, "03_novel_lncrnas/novel_expressed_polyA.gtf"))

print(paste("Novel Genes polyA: ",length(unique(novel_expressed$gene_id))))
print(paste("Novel Transcripts polyA: ",length(unique(novel_expressed$transcript_id))))



```

Things to try: 
- remove one exon for the stats
- remove b samples 
#Stats 

```{r results='hide', message=FALSE, warning=FALSE}
# Palette for all plots below

# --------------------
## EXON COUNT
# --------------------

ec1 <- barplot_exon_count(novel_expressed, "Novel", palette[1])
ec2 <- barplot_exon_count(lncRNAs_ref, "lncRNAs - Reference Macaque", palette[2])
ec3 <- barplot_exon_count(lncRNAs_ref_human, "lncRNAs - Reference Human", palette[3])+labs(y = "Frequency")
ec4 <- barplot_exon_count(mrna_ref_human, "mRNAs - Reference Human", palette[4])
ec5 <- barplot_exon_count(mRNAs_ref, "mRNAs - Reference Macaque", palette[5])+labs(x = "Number of Exons") 
ggarrange( ec1,ec2,ec3,ec4,ec5,  ncol=1, nrow=5, top = "Transcripts Lengths") 
# Exon count generally lower in lncrnas than mrnas (Ok. same as sources)

# --------------------
## Transcript lengths: still weird, very long ones ...
# --------------------
tl1 <-transcript_length_density(novel_expressed," NOVEL",palette[1]) 
tl2 <-transcript_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
tl3 <-transcript_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
tl4 <-transcript_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
tl5 <-transcript_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Transcript length")
ggarrange( tl1,tl2,tl3,tl4,tl5,  ncol=1, nrow=5, top = "Transcripts Lengths",  legend="right")


# --------------------
## Transcript lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_transcript_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = levels),  y = range )) +
        labs( x = "", y = "" , title = "Transcript length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,10000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1

# --------------------
## Exon lengths - Densities 
# --------------------
p1 <-exon_length_density(novel_expressed," NOVEL",palette[1]) 
p2 <-exon_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
p3 <-exon_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
p4 <-exon_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
p5 <-exon_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Exon length")
ggarrange( p1,p2,p3,p4,p5,  ncol=1, nrow=5, top = "Exon Lengths",  legend="right")

# --------------------
## Exon lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = levels),  y = range )) +
        labs( x = "", y = "" , title = "Exon length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,1000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1

```



# ------------------------------------
# Merge them together 
```{r results='hide', message=FALSE, warning=FALSE}

# Novel from ribodepletion compared to polyA 
riboonly <- import(file.path(datadir, "03_novel_lncrnas/01_gffcompare/gffcompare_btw_novels.gtf.annotated.gtf"))
mask_u <- riboonly$class_code == "u";  mask_u[is.na(mask_u)] <- FALSE
table(riboonly[mask_u]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- riboonly[mask_u]$transcript_id
novel <- riboonly[riboonly$transcript_id %in% novel_transcript_ids]
print("Novel from ribo depletion - gff compare")
print(length(unique(novel$transcript_id)))
print(length(unique(novel$gene_id)))



## I could check some stats about the new ones.

# I just append one to the otherm

# ---------------
#       Save
# ---------------
#export(novel, file.path(datadir, "03_novel_lncrnas/novel_expressed_ribodepl_newtopolyA.gtf"))
```

```{r results='hide', message=FALSE, warning=FALSE}
# Palette for all plots below

novel_all <- import(file.path(datadir, "03_novel_lncrnas/novel_expressed_all.gtf"))

novel_transcripts_codes <- novel_all[novel_all$type == "transcript",]$class_code
total <- length(novel_transcripts_codes)
prop_intergenic <- (length(novel_transcripts_codes[novel_transcripts_codes == "u"])/total)*100
prop_antisense <- (length(novel_transcripts_codes[novel_transcripts_codes == "x"])/total)*100
print(prop_intergenic)
print(prop_antisense)
# --------------------
## EXON COUNT
# --------------------
novel_expressed <- novel_all
ec1 <- barplot_exon_count(novel_expressed, "Novel", palette[1])
ec2 <- barplot_exon_count(lncRNAs_ref, "lncRNAs - Reference Macaque", palette[2])
ec3 <- barplot_exon_count(lncRNAs_ref_human, "lncRNAs - Reference Human", palette[3])+labs(y = "Frequency")
ec4 <- barplot_exon_count(mrna_ref_human, "mRNAs - Reference Human", palette[4])
ec5 <- barplot_exon_count(mRNAs_ref, "mRNAs - Reference Macaque", palette[5])+labs(x = "Number of Exons") 
ggarrange( ec1,ec2,ec3,ec4,ec5,  ncol=1, nrow=5, top = "Transcripts Lengths") 
# Exon count generally lower in lncrnas than mrnas (Ok. same as sources)

# --------------------
## Transcript lengths: still weird, very long ones ...
# --------------------
tl1 <-transcript_length_density(novel_expressed," NOVEL",palette[1]) 
tl2 <-transcript_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
tl3 <-transcript_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
tl4 <-transcript_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
tl5 <-transcript_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Transcript length")
ggarrange( tl1,tl2,tl3,tl4,tl5,  ncol=1, nrow=5, top = "Transcripts Lengths",  legend="right")


# --------------------
## Transcript lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_transcript_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = levels),  y = range )) +
        labs( x = "", y = "" , title = "Transcript length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,10000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1

# --------------------
## Exon lengths - Densities 
# --------------------
p1 <-exon_length_density(novel_expressed," NOVEL",palette[1]) 
p2 <-exon_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
p3 <-exon_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
p4 <-exon_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
p5 <-exon_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Exon length")
ggarrange( p1,p2,p3,p4,p5,  ncol=1, nrow=5, top = "Exon Lengths",  legend="right")

# --------------------
## Exon lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = levels),  y = range )) +
        labs( x = "", y = "" , title = "Exon length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,1000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1


# -----------------
# Expression Levels ( filter and plot )
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
# Calc logcpm 
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
# Create dataframe w/ logcpms
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
colnames(expression_logcpm)
## ADD only if we want to select a certain tissue
#expression_logcpm <- expression_logcpm[,str_detect(colnames(expression_logcpm), "Liver")]
#expression_logcpm

# select only when > than 1 at least 3 tissues
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
table(mask)
expression_logcpm <- expression_logcpm[mask,]

# Plotting
max_expression <- data.frame(id=rownames(expression_logcpm ), Means=rowMax(expression_logcpm))

print(head(max_expression))
max_expression['type'] <- "0"
max_expression <- add_type(max_expression, novel_expressed$gene_id, "novel")
max_expression <- add_type(max_expression, lncRNAs_ref$gene_id, "lncrna")
max_expression <- add_type(max_expression, mRNAs_ref$gene_id, "mrna")
max_expression <- max_expression[!max_expression$type == "0",]

length(unique(novel_expressed[novel_expressed$gene_id %in% max_expression$id]$gene_id))
length(unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% max_expression$id]$gene_id))
length(unique(mRNAs_ref[mRNAs_ref$gene_id %in% max_expression$id]$gene_id))

ggplot(max_expression, aes(x=factor(type,level = c("lncrna", "novel", "mrna")) , y=Means, fill = type )) + 
      geom_boxplot(color="darkgrey",  alpha=1.0, fill = palette_expression)+
      labs(y = "logCPM", x = "", title = "Expression" )+
      theme(legend.title=element_blank())+
      theme(plot.title = element_text(hjust = 0.5))+
      theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))

novel_expressed <- novel_expressed[novel_expressed$gene_id %in% max_expression$id]
length(unique(novel_expressed$gene_id))
```
# Amount of tissues in which each type of RNA is expressed 

```{r results='hide', message=FALSE, warning=FALSE}
#names <-lapply(colnames(assays(dds_ref)$logCPM), function(x) str_split(tail(str_split(x,"/")[[1]], n = 1), "_")[[1]][2])
#colnames(assays(dds_ref)$logCPM) <- names

palette_expression <-c( "#E9A3C9","#F4E3ED", "#E6F5D0")

expression <- sapply(dds_ref, function(dds_ref) length(unique(dds_ref[,assays(dds_ref)$logCPM > 1]$tissue)))

expression

# i have expression values from 

df <- as.data.frame(as.data.frame(expression))
df$id <- rownames(df)
df$type <- "0"

df <- add_type(df, novel_expressed$gene_id, "novel")
df <- add_type(df, lncnras_expressed, "lncrna")
df <- add_type(df, mrnas_expressed, "mrna")
df <- df[!df$type == "0",]

p <- ggplot(df, aes(x = factor(type, levels = c("lncrna", "novel", "mrna")), y = expression, fill = type)) +
            geom_boxplot(color = "grey", fill = palette_expression)+theme_pubclean()+labs(x = "", y = "# Tissues", title = "Expression across tissues")+
            theme(legend.title=element_blank())+
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p 

table(df$type)

```




```{r results='hide', message=FALSE, warning=FALSE}
blast_dir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_blast"
matrix_o <- as.matrix(read.table(file.path(blast_dir, "blast_result_lncpedia_0.0001.tab"), sep="\t",header=TRUE))

head(matrix_o)


lncpedia_001 = 6.8
lncpedia_00001 = 15
noncode_001 = 33
noncode_00001 = 44
df2 <- data.frame(Eval=rep(c("0.001", "0.00001"), each=2),
                db=rep(c("LncPedia-Human", "Noncode-Human"),2),
                count=c(lncpedia_001, lncpedia_00001, noncode_001, noncode_00001))
head(df2)
p <- ggplot(data=df2, aes(x=db, y=count, fill=Eval)) +labs( title="BlastN to obtain Orthologs", x = "Database", y = "# Orthologs")+
geom_bar(stat="identity", color="grey", position=position_dodge())+
  theme_minimal()
# Use brewer color palettes
p + scale_fill_brewer(palette="Reds")
```









