---
author: "Luisa Santus"
date: "2/10/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---
title: "lncrna annotation "
author: "Luisa Santus"
date: "12/19/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# The script gets the pipeline till the novel concordant for both polyA and ribodepleted 

## Load data and libraries

```{r results='hide', message=FALSE, warning=FALSE}
library(rtracklayer); library(stringr); library(ggplot2); library(grid); library(gridExtra); library(UpSetR); library(ComplexHeatmap)
library(GenomicRanges); library(dplyr); library(zeallot); library(DESeq2); library(edgeR);  library("ggplotify"); library("grid"); library(tidyverse)
library(tximport); library(tximportData); library(plyr); library(Gviz); library(ggpubr)

# Reused paths
projdir = "/home/luisas/Desktop/cluster/proj/data/"
datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"
assemblies <- "/home/luisas/Desktop/cluster/assemblies"

# Import Utils 
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/","code/ebola/src/scripts/Analysis/utils")
source(file.path(scripts_dir,"de_utils.R")); source(file.path(scripts_dir,"lncrna_annotation_utils.R"))

# Human training set
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownlncRNAs.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownMrnas.gtf"))
lncRNAs_ref_human_gene <- lncRNAs_ref_human[lncRNAs_ref_human$type == "gene",]

# The lncRNAs and mRNAs in the reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_mrnas.gtf"))

```


First, apply the filtering to each dataset separately 
 - identify novel transcripts ( gff compare "u", "x")
 - deplete non-concordant predictions
 - retain only ones expressed in more than 3 samples

# Preprocess Ribo depleted 
```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------
#     RiboDepleted
# ---------------------

# -----------------
# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_feelNC_prediction/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_feelNC_prediction/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))


# -----------------
# Select only the NOVEL ones (gff compare )
mask_uix <- merged_annotated$class_code == "u" | merged_annotated$class_code == "x" ;  mask_uix[is.na(mask_uix)] <- FALSE
table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

# -----------------
# Remove Non-Concordant
df <- data.frame()
df <- rbind(df, data.frame(gene_id = novel$gene_id, transcript_id = novel$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
novel_concordant <- novel[!(novel$gene_id %in% unconcordant_prediction$gene_id)]
print("Novel - concordant")
print(length(unique(novel_concordant$transcript_id)))

# Export the novel Concordant 
export(novel_concordant, (file.path(datadir,"03_novel_lncrnas/00_all_novels/novel_rhemac10_concordant_ribodepleted.gtf")))
```

# Pre process polyA

```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------
#    PolyA
# ---------------------


# -----------------
# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
dir_counts_ref = file.path(datadir, "02_RNA-Seq_external/06_counts_all")
htseq_files_ref <- iterate_files(dir_counts_ref, "UMI.f3.q60.HTseq.gene_counts.tab")

# -----------------
# Select only the NOVEL ones (gff compare )
mask_uix <- merged_annotated$class_code == "u" | merged_annotated$class_code == "x" ;  mask_uix[is.na(mask_uix)] <- FALSE

table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

# save file
#export(novel, file.path(datadir, "03_novel_lncrnas/novel_after_gffcompare_polyA.gtf"))



##  here quantification step happens outside!! 

# -----------------
# Remove Non-Concordant
df <- data.frame()
df <- rbind(df, data.frame(gene_id = novel$gene_id, transcript_id = novel$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
novel_concordant <- novel[!(novel$gene_id %in% unconcordant_prediction$gene_id)]
print("Novel - concordant")
print(length(unique(novel_concordant$transcript_id)))
export(novel_concordant, (file.path(datadir,"03_novel_lncrnas/00_all_novels/novel_rhemac10_concordant_polyA.gtf")))
```

# --------------------------------
Both the novel concordant gtf files were merged and quantification was run with htseq-count
03_novel_lncrnas/00_gtf_filtering_step_01
#-------------------------------------


