---
title: "Untitled"
author: "Luisa Santus"
date: "2/7/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
```




```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------------------------
# Add the prefix o both datasets to make identifiers unique
# ---------------------------------------
gtf_polya <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/01_novel_expressed/novel_expressed_polyA.gtf")
gtf_ribodepl <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/01_novel_expressed/novel_expressed_ribodepl_newtopolyA.gtf")
ref_gtf <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit.gtf")


add_prefix <- function(gtf, prefix){
  gtf$gene_id <- paste(prefix, gtf$gene_id, sep = "") 
  gtf$transcript_id <- paste(prefix, gtf$transcript_id, sep = "") 
  return(gtf)
}

# ---------------------------------------                        
# Add gene names and transcript name
# ---------------------------------------
add_gene_name <- function(gtf){
  missing_gene_name <- gtf[gtf$type == "exon"][is.na(gtf[gtf$type == "exon"]$gene_name)]                                
  transcripts <- gtf[gtf$type == "transcript"]
  # To the ones that have no gene name, find the gene name associated with the transcript ID and use it
  test <-  lapply(seq_along(missing_gene_name), function(x) transcripts[transcripts$transcript_id == missing_gene_name[x]$transcript_id]$gene_name)
  gtf[gtf$type == "exon"][is.na(gtf[gtf$type == "exon"]$gene_name)]$gene_name <- unlist(test)
  gtf[is.na(gtf$gene_name)]$gene_name <- paste(gtf[is.na(gtf$gene_name)]$gene_id, "_unkown", sep = "")
  return(gtf)
}

gtf_polya_newids <- add_prefix(gtf_polya, "polya_")
gtf_ribodepl_newids <- add_prefix(gtf_ribodepl, "ribodepl_")

# Identify ribodepleted transcripts, whose gene name was already present in the poly(A) ones.

gtf_polya_newids_names <- add_gene_name(gtf_polya_newids)
gtf_ribodepl_newids_names <- add_gene_name(gtf_ribodepl_newids)

unique(gtf_ribodepl_newids_names$gene_name[gtf_ribodepl_newids_names$gene_name %in% gtf_polya_newids_names$gene_name])

# ---------------------------------------
# Merge them together ( just append ) 
# ---------------------------------------
glst <-list(gtf_polya_newids, gtf_ribodepl_newids)
gtf <- do.call(c, as(glst, "GRangesList"))

# Check that the lengths correspond 
length(unique(gtf$gene_id))
length(unique(gtf_polya_newids$gene_id))+length(unique(gtf_ribodepl_newids$gene_id))

# ---------------------------------------
# Merge Novel and Ref  together ( just append ) 
# ---------------------------------------
glst_ref_and_novel <-list(ref_gtf,gtf)
gtf_ref_and_novel <- do.call(c, as(glst_ref_and_novel, "GRangesList"))

gtf_ref_and_novel[is.na(gtf_ref_and_novel$gene_name)]$gene_name <- paste(gtf_ref_and_novel[is.na(gtf_ref_and_novel$gene_name)]$gene_id, "_unkown", sep = "") 
gtf_ref_and_novel[is.na(gtf_ref_and_novel$transcript_name)]$transcript_name <-paste(gtf_ref_and_novel[is.na(gtf_ref_and_novel$transcript_name)]$transcript_id, "_unkown", sep = "")  

export(gtf_ref_and_novel, file.path("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_final_catalogue/rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names.gtf"))  

table(is.na(gtf_ref_and_novel$gene_name))
gtf_ref_and_novel[grepl("c\\(", gtf_ref_and_novel$gene_name)]
```
