---
title: "lncAnnotation performance Assessment"
author: "Luisa Santus"
date: "10/22/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data and libraries

```{r results='hide', message=FALSE, warning=FALSE}
library(rtracklayer)
library(VennDiagram)
library(stringr)
library(ggplot2)
library(grid)
library(gridExtra)
library(UpSetR)
library(ComplexHeatmap)
library(GenomicRanges)
library(dplyr)
library(zeallot)
library(DESeq2)
library(edgeR)
library("ggplotify")
library("grid")
library(tidyverse)
library(tximport)
library(tximportData)
library(plyr)
library(Gviz)

projdir = "/home/luisas/Desktop/cluster/proj/data/"
datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
assemblies <- "/home/luisas/Desktop/cluster/assemblies"

scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/","code/ebola/src/scripts/Analysis/utils")
source(file.path(scripts_dir,"de_utils.R"))

hisat <- file.path(datadir, "02_RNA-Seq_old/03_hisat")
dir_hisat_zyagen <- file.path(hisat, "Zyagen")
#dir_abundances_zyagen <- file.path(dir,"02_RNA-Seq_old/06_abundances/Zyagen")

# Human training set
lncRNAs_ref_human <- import(file.path(projdir, "01_PreliminaryFiles/gene_annotations/gencode.v26.GRCh38.annotation_knownlncRNAs.gtf"))

# The lncRNAs and mRNAs in the reference
lncRNAs_ref <- import(file.path(projdir, "01_PreliminaryFiles/gene_annotations/rheMac8_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(projdir, "01_PreliminaryFiles/gene_annotations/rheMac8_EBOV-Kikwit_mrnas.gtf"))

# Candidates lncRNAs 
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_old/05_lncrnaAnnotation/feelnc_gencode_lncrna/candidate_lncRNA.gtf"))

# Prediction 
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_old/05_lncrnaAnnotation/feelnc_gencode_lncrna/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_old/05_lncrnaAnnotation/feelnc_gencode_lncrna/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
stringtie_gtf <- import(file.path(datadir, "02_RNA-Seq_old/04_stringtie/Zyagen/Zyagen_stringtie_merged_reference_guided.gtf"))
```


## Exploring the data
We first Check if we have multiple isoforms per gene, we do we want to collapse them. We do.
```{r multiple_isoforms_check, results='hide', message=FALSE, warning=FALSE}
df <- data.frame("gene" =  lncRNAs_out$gene_id, "transcript" = lncRNAs_out$transcript_id, stringsAsFactors = FALSE)
#df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
#df[df$gene == "ENSMMUG00000037508",]
# same for reference 
df <- data.frame("gene" =  lncRNAs_ref$gene_id, "transcript" = lncRNAs_ref$transcript_id)
#df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
#df[df$gene == "ENSMMUG00000000801",]
```

### Remove unconcordant predictions
Since there are unconcordant predictions (same gene - multiple transcripts - assigned to different classes : lncRNA AND mRNA), we do remove them.
```{r echo=T, results='hide', eval=FALSE}
## REMOVE UNCONCORDANT PREDICTIONS
# Get the genes with unconcordant prediction labels for transcripts 
df <- data.frame()
df <- rbind(df, data.frame(gene_id = lncRNAs_out$gene_id, transcript_id = lncRNAs_out$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))

unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)

# Remove unconcordant predictions from sets 
lncRNAs_out <-lncRNAs_out[!(lncRNAs_out$gene_id %in% unconcordant_prediction$gene_id), ]
mRNAs_out <- mRNAs_out[!(mRNAs_out$gene_id %in% unconcordant_prediction$gene_id), ]

# Double check there is for real no unconcordant prediction anymore
df <- data.frame()
df <- rbind(df, data.frame(gene_id = lncRNAs_out$gene_id, transcript_id = lncRNAs_out$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
```

### Gene id merged, but identified transcript?
Do we have any lncrnas which have gene id as merged, but known transcript? YES! 
In that case we add the refernce id as gene id to the ones that have it.

```{r duplicates}
# Collect all the gene ids - merged and reference gene ids
# I add all the reference gene ids and remove NAs 
get_geneids_with_refids <- function(gr){
  gr_ref_ids <- unique(gr[str_detect(gr$gene_id,"MSTRG") &
                                          !is.na(gr$ref_gene_id),]$ref_gene_id)
  return(unique(union(gr_ref_ids, gr$gene_id)))
}
stringtie_gene_ids_withref<- get_geneids_with_refids(stringtie_gtf)
lncRNAs_out_gene_ids_withref <-  get_geneids_with_refids(lncRNAs_out)
candidates_lncRNA_gene_ids_withref <- get_geneids_with_refids(candidates_lncRNA)
mRNAs_out_gene_ids_withref <- get_geneids_with_refids(mRNAs_out)
```

### How many Novels? 
Keep Track of the number of novel genes identified. For the moment we just save the raw number.
They are calculated trough overlap of gene_ids.
```{r}

get_novels <- function(lncRNAs_out, stringtie_gtf, lncRNAs_ref){
  intersect_feelnc_stringtie <- intersect(lncRNAs_out, stringtie_gtf)
  intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out)
  novel_lncrnas = setdiff(intersect_feelnc_stringtie,intersect_reflnrna_out )
  return(novel_lncrnas)
}
novels <- list()
novel_lncrnas <- get_novels(lncRNAs_out_gene_ids_withref, stringtie_gene_ids_withref, lncRNAs_ref)
length(unique(novel_lncrnas))
novels <- cbind(novels,"raw" = length(novel_lncrnas ))

# Here we retrieve the ranges 
novel_lncrnas_ranges <- lncRNAs_out[lncRNAs_out$gene_id %in% novel_lncrnas]
  novel_lncrnas_ranges[str_detect(novel_lncrnas_ranges$gene_id,"MSTRG"),]
```
## Overlap of Gene_ids
we separate 
```{r}
grid::grid.newpage()
vd_draw(list(unique(lncRNAs_ref$gene_id), unique(lncRNAs_out_gene_ids_withref), unique(stringtie_gene_ids_withref), unique(mRNAs_ref$gene_id)), c("LNCRNA-ref","feelnc","stringtie", "mrnas ref"))

grid::grid.newpage()
vd_draw(list(unique(lncRNAs_ref$gene_id), unique(lncRNAs_out_gene_ids_withref), unique(stringtie_gene_ids_withref)), c("LNCRNA-ref","feelnc","stringtie"))

grid::grid.newpage()
vd_draw(list(unique(lncRNAs_ref$gene_id), unique(lncRNAs_out_gene_ids_withref), unique(stringtie_gene_ids_withref), unique(mRNAs_ref$gene_id), unique(mRNAs_out_gene_ids_withref)), c("LNCRNA-ref","feelnc","stringtie", "mrnas ref", "mrnapred"))

grid::grid.newpage()
lt <-list(unique(lncRNAs_ref$gene_id), unique(lncRNAs_out_gene_ids_withref), unique(stringtie_gene_ids_withref),  unique(mRNAs_ref$gene_id), unique(mRNAs_out_gene_ids_withref))
names(lt) <- c("LNCRNA-ref","feelnc","stringtie","MRNA-ref", "Feelnc-mrna")
m1 = make_comb_mat(lt)
UpSet(m1)


```



```{r}
# checking how many of the gene ids or ref gene ids in the predicted ones are in the reference
intersect_reflnrna_out <- intersect(lncRNAs_ref$gene_id, lncRNAs_out_gene_ids_withref)
intersect_reflnrna_stringtie <- intersect(lncRNAs_ref$gene_id, stringtie_gene_ids_withref)
intersect_feelnc_stringtie <- intersect(lncRNAs_out_gene_ids_withref, stringtie_gene_ids_withref)
intersect_reflnrna_mrnaout<- intersect(lncRNAs_ref$gene_id, mRNAs_out_gene_ids_withref)

# get the Interesting Numbers 
correctly_identified = intersect(intersect_reflnrna_stringtie,lncRNAs_out_gene_ids_withref)
missed = setdiff(intersect_reflnrna_stringtie,correctly_identified)
intersect_missed_mrnaout<- intersect(missed, mRNAs_out_gene_ids_withref)
candidates = unique(candidates_lncRNA_gene_ids_withref)
missed_for_unconcordant_prediction <- setdiff(intersect(missed,candidates), mRNAs_out_gene_ids_withref)
missed_because_of_filtering <- setdiff(missed,candidates)


# PLOT 
# -----------------
df <- data.frame()
df <- rbind(df, data.frame(dataset = "all NOVEL", number = length(novel_lncrnas), type = "Novel"))
df <- rbind(df, data.frame(dataset = "Correctly identified", number = length(correctly_identified), type = "Known correct"))
df <- rbind(df, data.frame(dataset = "Predicted as Mrna", number = length(intersect_missed_mrnaout), type = "Known missed"))
df <- rbind(df, data.frame(dataset = "Missed for Filter ", number = length(missed_because_of_filtering), type = "Known missed"))
df <- rbind(df, data.frame(dataset = "Missed for unconcordant Prediction ", number = length(missed_for_unconcordant_prediction), type = "Known missed"))

p1 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge(), width=0.85)+
  geom_text(aes(label=number),color="black",
            position = position_dodge(0.9), size=3.5)+
  scale_fill_brewer(palette="Paired")+
  labs(title="lncRNAs Prediction vs Reference") +
  ylab("Count") +
  xlab("") +
  theme(legend.title=element_blank())+
  theme(plot.title = element_text(hjust = 0.5))
p1 

# calculate the proportion of the correctly idententified ones
prop_correctly = (length(correctly_identified)/(length(correctly_identified) + length(missed)))*100
print(paste0("Proportion correclty identified: ", prop_correctly))
# -----------------
```

## INVESTIGATE THE NOVEL ONES

How many of the genes which were missed, were filtered out in the filter step?
The 118 that are in the candidates but are not predicted as mrnas? might be for unconcordant prediction


# Collapse transcripts - only take the one with max length
```{r}
# Instead of collapsing we select the longest transcript among the transcripts in one gene
get_only_max_transcript <- function(gr){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  gene_with_multiple_isoforms <-df[!duplicated(df$transcript_id),] %>% dplyr::group_by(gene_id) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
  collapsed <-df %>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width)) %>% dplyr::group_by(gene_id) %>% dplyr::slice(which.max(range))
  gene_with_one_isoform <-df[!duplicated(df$transcript_id),] %>% dplyr::group_by(gene_id) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number == 1) 
  gr <- gr[gr$transcript_id %in% collapsed$transcript_id ,]
  return(gr)
}


get_nr_exons <- function(gr){
  gr <- get_only_max_transcript(gr)
  df <- data.frame("gene_id" = gr$gene_id,"exon_number" = as.numeric(gr$exon_number))
  number_exons <- df %>% dplyr::group_by(gene_id) %>%dplyr::summarize(max_exon = max(exon_number))
  return(number_exons)
}
```

## Number Of exons distributon
Distribution of number of exons in newly predicted ones
```{r}
grid::grid.newpage()

colors <- c("#D16103","red","#4E84C4" )

df_l <- data.frame(get_nr_exons(novel_lncrnas_ranges))
df_l$type <- "Novel"

lncrnas_ref_exons <- lncRNAs_ref[lncRNAs_ref$type =="exon",]
df_ref_lncrnas <- get_nr_exons(lncrnas_ref_exons)
df_ref_lncrnas$type <- "Reference-macaque"

df <- rbind(df_ref_lncrnas, df_l)

df_ref_human_lncrnas_exons <- lncRNAs_ref_human[lncRNAs_ref_human$type =="exon",]
df_ref_human_lncrnas <- get_nr_exons(df_ref_human_lncrnas_exons)
df_ref_human_lncrnas$type <- "Reference-human "
df <- rbind(df_ref_human_lncrnas, df)

p1 <- ggplot(df, aes(x=max_exon, fill=type)) + 
  geom_histogram(position="identity", alpha=0.4, binwidth =1)+
  xlim(1,15)+
  theme(legend.title=element_blank())+
  theme(plot.title = element_text(hjust = 0.5))+
  xlab("Number of exons")+
  ylab("Count")+
  labs(title="Number of Exons")
p1 <- p1 + scale_fill_manual(values=colors)
p1



# Same but with proportions
p2 <- ggplot(df, aes(x=max_exon, fill=type)) + 
  geom_histogram(aes(y=c(..count..[..group..==1]/sum(..count..[..group..==1]),
                         ..count..[..group..==2]/sum(..count..[..group..==2]),
                         ..count..[..group..==3]/sum(..count..[..group..==3]))*100),position="identity", alpha=0.5,binwidth = 1)+
  labs(title="% Number of exons ")+ ylab ("Percentage of #exons")+
  theme(legend.title=element_blank())+
  theme(plot.title = element_text(hjust = 0.5))+
  xlab("Number of exons")+
  xlim(1,15)
p2 <- p2 + scale_fill_manual(values=colors)
p2


```


# LENGTH OF GENES 
```{r}

histo_length <- function(gr,title,color){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  collapsed <- df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  p1 <- ggplot(collapsed, aes(x=range)) + geom_histogram(position="identity", alpha=0.5, fill = color, binwidth = 100)+labs(title=title) + scale_y_log10()+ 
    ylab("Log10 Count") + xlab("#Bases")+  theme(legend.title=element_blank())+
  theme(plot.title = element_text(hjust = 0.5))+xlim(0,20000)
  return(p1)
}

histo_length_exon <- function(gr,title,color){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "exon_number" = gr$exon_number, "range_width" = width(ranges(gr)))
  collapsed <- df%>% dplyr::group_by(gene_id,transcript_id, exon_number) %>% dplyr::summarize("range" = sum(range_width))
  p1 <- ggplot(collapsed, aes(x=range)) + geom_histogram(position="identity", alpha=0.5, fill = color, binwidth = 100)+labs(title=title) + scale_y_log10()+ 
    ylab("Log10 Count") + xlab("#Bases")+  theme(legend.title=element_blank())+
  theme(plot.title = element_text(hjust = 0.5))+xlim(0,20000)
  return(p1)
}


plot_all_all <- function(ref, gr, gr1, colors){
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
  collapsed <-df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  collapsed$type <- "novel"
  
  df_human <- data.frame("gene_id" = gr1$gene_id,"transcript_id" = gr1$transcript_id, "range_width" = width(ranges(gr1)))
  collapsed_human <-df_human%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  collapsed_human$type <- "humanref"
  
  df_ref <- data.frame("gene_id" = ref$gene_id,"transcript_id" = ref$transcript_id, "range_width" = width(ranges(ref)))
  collapsed_ref <-df_ref%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
  collapsed_ref$type <- "refmacaque"
  
  all <- rbind(data.frame(collapsed),data.frame(collapsed_ref),data.frame(collapsed_human))
  mu <- ddply(all, "type", summarise, grp.mean=mean(range))
  
  p3 <- ggplot(all, aes(x=range, col= type))  + 
    geom_histogram(position="identity",alpha=0.5, binwidth = 50) +
    geom_vline(data=mu, aes(xintercept=grp.mean, color=type),
             linetype="dashed")+
    geom_text(aes(x = (mu[mu$type == "novel",]$grp.mean)+ 1000, y = 340, label = as.integer(mu[mu$type == "novel",]$grp.mean)), colour = colors[2]) +
    geom_text(aes(x = (mu[mu$type == "refmacaque",]$grp.mean)+ 1000, y = 740, label = as.integer(mu[mu$type == "refmacaque",]$grp.mean)), colour = colors[3]) +
    geom_text(aes(x = (mu[mu$type == "humanref",]$grp.mean)+ 1000, y = 540, label = as.integer(mu[mu$type == "humanref",]$grp.mean)), colour = colors[1]) +

    labs(title="Length distribution NOVEL vs REFERENCE")+
    theme(legend.title=element_blank())+
    theme(plot.title = element_text(hjust = 0.5))+xlim(0,20000)+ ylab("Log10 Count") + xlab("#Bases")
  p3 <- p3 + scale_color_manual(values=colors)
  return(p3)
}


plot_length_exons <- function(ref, gr, gr1, colors){
  gr <- novel_lncrnas_ranges_collapsed
  df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id,"exon_number" = gr$exon_number,  "range_width" = width(ranges(gr)))
  collapsed <- df%>% dplyr::group_by(gene_id,transcript_id, exon_number) %>% dplyr::summarize("range" = sum(range_width))
  collapsed$type <- "novel"
  
  df_human <- data.frame("gene_id" = gr1$gene_id,"transcript_id" = gr1$transcript_id,"exon_number" = gr1$exon_number,  "range_width" = width(ranges(gr1)))
  collapsed_human <-df_human%>% dplyr::group_by(gene_id,transcript_id, exon_number) %>% dplyr::summarize("range" = sum(range_width))
  collapsed_human$type <- "humanref"
  
  df_ref <- data.frame("gene_id" = ref$gene_id,"transcript_id" = ref$transcript_id,"exon_number" = ref$exon_number,  "range_width" = width(ranges(ref)))
  collapsed_ref <-df_ref%>% dplyr::group_by(gene_id,transcript_id, exon_number) %>% dplyr::summarize("range" = sum(range_width))
  collapsed_ref$type <- "refmacaque"
  
  all <- rbind(data.frame(collapsed),data.frame(collapsed_ref),data.frame(collapsed_human))
  mu <- ddply(all, "type", summarise, grp.mean=mean(range))
  
  p3 <- ggplot(all, aes(x=range, col= type))  + 
    geom_histogram(position="identity",alpha=0.5, binwidth = 50) +
    geom_vline(data=mu, aes(xintercept=grp.mean, color=type),
             linetype="dashed")+
    geom_text(aes(x = (mu[mu$type == "novel",]$grp.mean)+ 1000, y = 340, label = as.integer(mu[mu$type == "novel",]$grp.mean)), colour = colors[2]) +
    geom_text(aes(x = (mu[mu$type == "refmacaque",]$grp.mean)+ 1000, y = 2140, label = as.integer(mu[mu$type == "refmacaque",]$grp.mean)), colour = colors[3]) +
    geom_text(aes(x = (mu[mu$type == "humanref",]$grp.mean)+ 1000, y = 1240, label = as.integer(mu[mu$type == "humanref",]$grp.mean)), colour = colors[1]) +

    labs(title="Length distribution NOVEL vs REFERENCE")+
    theme(legend.title=element_blank())+
    theme(plot.title = element_text(hjust = 0.5))+xlim(0,20000)+ ylab("Log10 Count") + xlab("#Bases")
  p3 <- p3 + scale_color_manual(values=colors)
  return(p3)
}

# Select subsets to plot
# Get only exons of the reference, retain only the max transcript in term of length
lncrnas_ref_exons <- lncRNAs_ref[lncRNAs_ref$type =="exon",]
lncrnas_ref_collapsed <- get_only_max_transcript(lncrnas_ref_exons)

ref_human_lncrnas_exons <- lncRNAs_ref_human[lncRNAs_ref_human$type =="exon",]
lncrnas_ref_collapsed_human <- get_only_max_transcript(ref_human_lncrnas_exons)

novel_lncrnas_ranges <- lncRNAs_out[lncRNAs_out$gene_id %in% novel_lncrnas]
novel_lncrnas_ranges_collapsed <- get_only_max_transcript(novel_lncrnas_ranges)
# Double checking
# length(unique(novel_lncrnas_ranges$gene_id))
gr <- lncrnas_ref_collapsed
df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id, "range_width" = width(ranges(gr)))
collapsed <-df%>% dplyr::group_by(gene_id,transcript_id) %>% dplyr::summarize("range" = sum(range_width))
collapsed$type <- "novel"
  

colors <- c("red","#D16103","#4E84C4")
p3 <- plot_length_exons(lncrnas_ref_collapsed,novel_lncrnas_ranges_collapsed,lncrnas_ref_collapsed_human,colors)
p1 <-histo_length_exon(novel_lncrnas_ranges," NOVEL",colors[2])
p2 <-histo_length_exon(lncrnas_ref_collapsed," REF MACAQUE",colors[3])
p4 <-histo_length_exon(lncrnas_ref_collapsed_human,"REF HUMAN","red")
g1 <- as.grob(p1); g2 <- as.grob(p2) ; g3 <- as.grob(p3) ; g4 <- as.grob(p4)
grid.arrange( grobs = list(g1,g2,g3,g4),layout_matrix = rbind(c(1,2,4),c(3,3,3)), top = "Exon Lengths")

# plot ALL 
colors <- c("red","#D16103","#4E84C4")
p3 <- plot_all_all(lncrnas_ref_collapsed,novel_lncrnas_ranges_collapsed,lncrnas_ref_collapsed_human,colors)
p1 <-histo_length(novel_lncrnas_ranges," NOVEL",colors[2])
p2 <-histo_length(lncrnas_ref_collapsed," REF MACAQUE",colors[3])
p4 <-histo_length(lncrnas_ref_collapsed_human,"REF HUMAN","red")
g1 <- as.grob(p1); g2 <- as.grob(p2) ; g3 <- as.grob(p3) ; g4 <- as.grob(p4)
grid.arrange( grobs = list(g1,g2,g3,g4),layout_matrix = rbind(c(1,2,4),c(3,3,3)), top = "Transcripts Lengths")


# --------------------------------------------------
## CONTROL HOW I AM GETTING THE LENGHTS 
df <- data.frame("gene" =  lncrnas_ref_collapsed$gene_id, "transcript" = lncrnas_ref_collapsed$transcript_id, stringsAsFactors = FALSE)
df[!duplicated(df$transcript),] %>% dplyr::group_by(gene) %>% dplyr::summarize(number=dplyr::n()) %>% dplyr::filter(number > 1)
df[df$gene == "ENSMMUG00000037508",]

# Checking wether we  really have so long exons
gr <- novel_lncrnas_ranges_collapsed
df <- data.frame("gene_id" = gr$gene_id,"transcript_id" = gr$transcript_id,"exon_id" = gr$exon_number,  "range_width" = width(ranges(gr)))
collapsed <- df%>% dplyr::group_by(gene_id,transcript_id, exon_id) %>% dplyr::summarize("range" = sum(range_width))

abs(12282983-12285590)

```


## Expression Levels of the novel ones

```{r echo=FALSE, results='hide'}
boxplot_logcpms <- function(dge,title, outline= TRUE, priorcount = 0.25){
  logcounts <- cpm(dge,log=TRUE, prior.count = priorcount)
  # Check distributions of samples using boxplots
  colors <- brewer.pal(n = length(levels(dge$samples$group)), name = 'Paired')
  col.cell <- colors[dge$samples$group]
  
  colnames(logcounts) <- dge$samples$group
  boxplot(logcounts, xlab="", ylab="Log2 counts per million"
          ,las=2,
          col = col.cell, outline = outline
          )
  # Let's add a blue horizontal line that corresponds to the median logCPM
  title(title)
}
filter_and_plotMDS <- function(dds,dge,title){
  # Filter lowly expressed genes
  assays(dds)$logCPM <- cpm(dge, log = TRUE,prior.count = 0.25)
  mask <- rowMeans(assays(dds)$logCPM) > 1
  dds <- dds[mask, ]
  dge <- dge[mask, ]
  print(length(levels(dge$samples$group)))
  colors <- brewer.pal(n = length(levels(dge$samples$group)), name = 'Paired')
  mds <- plotMDS(cpm(dge, log = TRUE), label= dge$samples$group , col=colors[dge$samples$group],main=title)
  
  par(xpd=T, mar=par()$mar+c(0,0,0,6))
  mds <- plotMDS(cpm(dge, log = TRUE), pch = 20, col=colors[dge$samples$group],main=title)

  legend(4.5,0.9, inset=c(-0.2,0), legend=unique(dge$samples$group),  fill= unique(colors[dge$samples$group]))

  par(mar=c(5, 4, 4, 2) + 0.1)
}

```


First we need to get the counts into a readable format.
```{r echo=FALSE,  results='hide'}
directory_zyagen = file.path(datadir,"02_RNA-Seq_old/06_quantification_stringtie_prepde")
count_matrix<- iterate_files(directory_zyagen, 'gene_count_matrix.csv')

# Load gene(/transcript) count matrix and labels
countData <- as.matrix(read.csv(count_matrix, row.names="gene_id"))
colData <- get_colData(countData)
all(rownames(colData) %in% colnames(countData))
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
```



### Zyagen 
```{r echo=FALSE,  results='hide'}
dds_zyagen <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
dge_zyagen <- DGEList(counts = assays(dds_zyagen)$counts, group = dds_zyagen$tissue, genes = rownames(dds_zyagen))
assays(dds_zyagen)$logCPM <- cpm(dge_zyagen, log = TRUE,prior.count = 0.25)
mask <- rowMeans(assays(dds_zyagen)$logCPM) > 1


filter_and_plotMDS(dds_zyagen,dge_zyagen,"Zyagen - All")
boxplot_logcpms( dge_zyagen,"Zyagen - all - no filter ")
boxplot_logcpms( dge_zyagen[mask, ],"Zyagen - all - logCPM >1  filter ")


dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
se_novel_only_zyagen <- dds[rownames(dds) %in% novel_lncrnas,]
dge_novel_only_zyagen <- DGEList(counts = assays(se_novel_only_zyagen)$counts, group = se_novel_only_zyagen$tissue, genes = rownames(se_novel_only_zyagen))
assays(se_novel_only_zyagen)$logCPM <- cpm(dge_novel_only_zyagen, log = TRUE, prior.count = 0.25)
mask <- rowMeans(assays(se_novel_only_zyagen)$logCPM) > 1


filter_and_plotMDS(se_novel_only_zyagen,dge_novel_only_zyagen,"Zyagen - Novel")
boxplot_logcpms(dge_novel_only_zyagen, "Zyagen - Novel - no filter")
boxplot_logcpms( dge_novel_only_zyagen[mask, ],"Zyagen - Novel - logCPM >1  filter ")

# How many novels are we comparing 
print(paste0("Number of novel used = ", length(unique(rownames(se_novel_only_zyagen)))))


# ADD the ones that are in stringtie gtf that have ref gene id in reference (corresponds to the number of correctly identified previously)
merged_but_with_refid_in_ref <- unique(stringtie_gtf[stringtie_gtf$ref_gene_id %in%correctly_identified, ]$gene_id)

dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
dds_ref <- dds[rownames(dds) %in% merged_but_with_refid_in_ref,]
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$tissue, genes = rownames(dds_ref))

assays(dds_ref)$logCPM <- cpm(dge_ref, log = TRUE, prior.count = 0.01)
mask <- rowMeans(assays(dds_ref)$logCPM) > 1


filter_and_plotMDS(dds_ref,dge_ref,"Zyagen - ref")
boxplot_logcpms(dge_ref, "Zyagen - ref - no filter", priorcount = 0.01)
boxplot_logcpms( dge_ref[mask, ],"Zyagen - ref - logCPM >1  filter ",priorcount = 0.01)
dim(dge_ref[mask, ])

assays(dds_ref)$logCPM
assays(dds_ref)$count

# How many novels are we comparing 
print(paste0("Number of ref used = ", length(unique(rownames(dds_ref)))))
print(paste0("Total ref =  ", length(unique(lncRNAs_ref$gene_id))))

print(rownames(dds_ref))

```

## Batch
```{r echo=FALSE,  results='hide'}

directory_batch = file.path(datadir,"02_RNA-Seq_old/06_quantification_stringtie_prepde_batch")
count_matrix<- iterate_files(directory_batch, 'gene_count_matrix.csv')

# Load gene(/transcript) count matrix and labels
countData <- as.matrix(read.csv(count_matrix, row.names="gene_id"))
colData <- get_colData(countData)
all(rownames(colData) %in% colnames(countData))
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)


dds_batch <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
dge_batch <- DGEList(counts = assays(dds_batch)$counts, group = dds_batch$tissue, genes = rownames(dds_batch))
assays(dds_batch)$logCPM <- cpm(dge_batch, log = TRUE,prior.count = 0.25)
mask <- rowMeans(assays(dds_batch)$logCPM) > 1


filter_and_plotMDS(dds_batch,dge_batch,"batch - All")
boxplot_logcpms( dge_batch,"batch - all - no filter ")
boxplot_logcpms( dge_batch[mask, ],"batch - all - logCPM >1  filter ")


dds_batch <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
se_novel_only_batch <- dds_batch[rownames(dds_batch) %in% novel_lncrnas,]
dge_novel_only_batch <- DGEList(counts = assays(se_novel_only_batch)$counts, group = se_novel_only_batch$tissue, genes = rownames(se_novel_only_batch))
assays(se_novel_only_batch)$logCPM <- cpm(dge_novel_only_batch, log = TRUE, prior.count = 0.25)
mask <- rowMeans(assays(se_novel_only_batch)$logCPM) > 1

assays(se_novel_only_batch)$logCPM
assays(se_novel_only_batch)$count

filter_and_plotMDS(se_novel_only_batch,dge_novel_only_batch,"batch - Novel")
boxplot_logcpms(dge_novel_only_batch, "batch - Novel - no filter")
boxplot_logcpms( dge_novel_only_batch[mask, ],"batch - Novel - logCPM >1  filter ")
dim(dge_novel_only_batch[mask, ])
# How many novels are we comparing 
print(paste0("Number of novel used = ", length(unique(rownames(se_novel_only_batch)))))


# ADD the ones that are in stringtie gtf that have ref gene id in reference (corresponds to the number of correctly identified previously)
merged_but_with_refid_in_ref <- unique(stringtie_gtf[stringtie_gtf$ref_gene_id %in%correctly_identified, ]$gene_id)

dds_batch <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)
dds_ref_batch <- dds_batch[rownames(dds_batch) %in% merged_but_with_refid_in_ref,]
dge_ref_batch <- DGEList(counts = assays(dds_ref_batch)$counts, group = dds_ref_batch$tissue, genes = rownames(dds_ref_batch))


# I change the prior count to have the logCPM and count 
assays(dds_ref_batch)$logCPM <- cpm(dge_ref_batch, log = TRUE, prior.count = 0.001)
mask <- rowMeans(assays(dds_ref_batch)$logCPM) > 1


filter_and_plotMDS(dds_ref_batch,dge_ref_batch,"batch - ref")
boxplot_logcpms(dge_ref_batch, "batch - ref - no filter", priorcount = 0.001)
boxplot_logcpms( dge_ref_batch[mask, ],"batch - ref - logCPM >1  filter ", priorcount = 0.001)
dim(dge_ref_batch[mask, ])


dge_ref_batch
assays(dds_ref_batch)$logCPM
assays(dds_ref_batch)$count

# How many novels are we comparing 
print(paste0("Number of ref used = ", length(unique(rownames(dds_ref)))))
print(paste0("Total ref =  ", length(unique(lncRNAs_ref$gene_id))))

print(rownames(dds_ref))
```



##  Orthologs
```{r}
#chainfile <- file.path(projdir,"/data/01_PreliminaryFiles/gene_annotations/chains/hg19ToHg18.over.chain")
#chain <- import.chain(chainfile)
#library(TxDb.Hsapiens.UCSC.hg19.knownGene)
#tx_hg19 <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
#tx_hg18 <- as.data.frame(liftOver(tx_hg19, chain))


#I need to find the orthologs: macaque --> Human 
genome_human <-  file.path(projdir,"/01_PreliminaryFiles/gene_annotations/chains/rheMac8ToHg38.over.chain")
chainfile <- file.path(projdir,"/01_PreliminaryFiles/gene_annotations/chains/rheMac8ToHg38.over.chain")
chain <- import.chain(chainfile) # Macaque!
macaque <- novel_lncrnas_ranges
Human_Orthologs <- liftOver(macaque, chain)

# For how many we did not find the orthologs
MISSED <- length(unique(novel_lncrnas_ranges$gene_id)) - length(unique(unlist(Human_Orthologs)$gene_id))
print(paste0("#Genes where no ortholog was identified: ",MISSED))

#Visualizing
novel_lncrnas_ranges_chr3 <- novel_lncrnas_ranges[seqnames(novel_lncrnas_ranges) == "chr3"]
novel_lncrnas_ranges_chr3 <- novel_lncrnas_ranges_chr3[start(novel_lncrnas_ranges_chr3) >= 310061 & end(novel_lncrnas_ranges_chr3) <= 330061]

novel_lncrnas_ranges_chr3_gene1 <- novel_lncrnas_ranges[novel_lncrnas_ranges$gene_id == "MSTRG.48461"]


GenomicRanges::reduce(Human_Orthologs[[1]])
atrack <- AnnotationTrack(novel_lncrnas_ranges_chr3, name = "Novel lncRNAs range")
gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack),
           from = 310061, to = 330061)
gen<-genome(Human_Orthologs[[2]])
chr <- as.character(unique(seqnames(Human_Orthologs[[2]])))

itrack <- IdeogramTrack()
plotTracks(list(itrack, gtrack, atrack))

as.data.frame(hg38_annotation)

hg38_annotation

grtrack <- GeneRegionTrack(as.data.frame(hg38_annotation), genome = "hg38", chromosome = "chr3", name = "Gene Model")
plotTracks(list(itrack, gtrack, atrack, grtrack),
           from = 47455280, to = 47453386)

```



```{r}
# Q1: To which region of Human are they overlapping
hg38_annotation <- import(file.path(assemblies,"/ensembl/release-96/Homo_sapiens_hg38/Homo_sapiens.GRCh38.96.gtf"))
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
hg38 <- transcripts(TxDb.Hsapiens.UCSC.hg38.knownGene)
hg38chr3 <- hg38[seqnames(hg38) == "chr3"]

df <- as.data.frame(findOverlaps(GenomicRanges::reduce(Human_Orthologs[[1]]), hg38, select="all" ))
unique(df$subjectHits)
hg38[df$subjectHits[1]]


library(AnnotationHub)
ah <- AnnotationHub()

```



## Classification 




