---
title: "lncrna annotation paper"
author: "Luisa Santus"
date: "12/19/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data and libraries

```{r results='hide', message=FALSE, warning=FALSE}
library(rtracklayer); library(VennDiagram); library(stringr); library(ggplot2); library(grid); library(gridExtra); library(UpSetR); library(ComplexHeatmap)
library(GenomicRanges); library(dplyr); library(zeallot); library(DESeq2); library(edgeR);  library("ggplotify"); library("grid"); library(tidyverse)
library(tximport); library(tximportData); library(plyr); library(Gviz); library(ggpubr)

# Reused paths
projdir = "/home/luisas/Desktop/cluster/proj/data/"
datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
assemblies <- "/home/luisas/Desktop/cluster/assemblies"

# Import Utils 
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/","code/ebola/src/scripts/Analysis/utils")
source(file.path(scripts_dir,"de_utils.R")); source(file.path(scripts_dir,"lncrna_annotation_utils.R"))

# Human training set
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownlncRNAs.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownMrnas.gtf"))
lncRNAs_ref_human_gene <- lncRNAs_ref_human[lncRNAs_ref_human$type == "gene",]

# The lncRNAs and mRNAs in the reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_mrnas.gtf"))

# Candidates lncRNAs  & Prediction
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_rheMac10/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_rheMac10/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_rheMac10/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
stringtie_gtf <- import(file.path(datadir, "02_RNA-Seq_rheMac10/04_stringtie/stringtie_merged_reference_guided.gtf"))
```


# Novel Ones
```{r results='hide', message=FALSE, warning=FALSE}
# Obtain novels from gff compare 
# remove genes, whose transcript are predicted as both lncrnas and mrnas
#remove_unconcordant_prediction(lncRNAs_out, mRNAs_out)
#merged_annotated <- import(file.path(datadir, "02_RNA-Seq_rheMac10/04b_gffCompare/merged.annotated.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_rheMac10/05_lncrnaAnnotation_no_l_option/01_gffcompare/merged.annotated.gtf"))
mask_uix <- merged_annotated$class_code == "u" |  merged_annotated$class_code == "x"|  merged_annotated$class_code == "i" ;  mask_uix[is.na(mask_uix)] <- FALSE
mask_u <- merged_annotated$class_code == "u";  mask_u[is.na(mask_u)] <- FALSE
table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel <- merged_annotated[mask_uix]
table(novel$class_code)
length(unique(novel$gene_id))
novel
```

#Known ones 
```{r results='hide', message=FALSE, warning=FALSE}
# Obtain novels from gff compare 
# remove genes, whose transcript are predicted as both lncrnas and mrnas
#remove_unconcordant_prediction(lncRNAs_out, mRNAs_out)
#merged_annotated <- import(file.path(datadir, "02_RNA-Seq_rheMac10/04b_gffCompare/merged.annotated.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_rheMac10/05_lncrnaAnnotation_no_l_option/benchmark_known/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
mask_eq <- merged_annotated$class_code == "="  ;  mask_eq[is.na(mask_eq)] <- FALSE
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
known_identified <- (unique(merged_annotated[mask_eq]$gene_id))
length(known_identified)
missed <- setdiff(unique(lncRNAs_ref$gene_id), known_identified)

# -----------------
# PLOT 
# -----------------
df <- data.frame()
df <- rbind(df, data.frame(dataset = "all NOVEL", number = length(unique(novel$gene_id)), type = "Novel"))
df <- rbind(df, data.frame(dataset = "Correctly identified", number = length(known_identified), type = "Known correct"))
df <- rbind(df, data.frame(dataset = "Uncorrectly classified", number = length(missed), type = "Known missed"))


p1 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge(), width=0.85)+
  geom_text(aes(label=number),color="black",
            position = position_dodge(0.9), size=3.5)+
  scale_fill_brewer(palette="Paired")+
  labs(title="LncRNAs prediction: #genes ") +
  ylab("Count") +
  xlab("") +
  theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1 

# calculate the proportion of the correctly idententified ones
prop_correctly = (length(correctly_identified)/(length(correctly_identified) + length(missed)))*100
print(paste0("Proportion correclty identified: ", prop_correctly))

```

#Stats 


```{r results='hide', message=FALSE, warning=FALSE}
# Palette for all plots below
palette <-brewer.pal(5,"Paired")

# --------------------
## EXON COUNT
# --------------------

ec1 <- barplot_exon_count(novel, "Novel", palette[1])
ec2 <- barplot_exon_count(lncRNAs_ref, "Reference Macaque", palette[2])
ec3 <- barplot_exon_count(lncRNAs_ref_human, "Reference Human", palette[3])+labs(y = "Frequency")
ec4 <- barplot_exon_count(mrna_ref_human, "Reference Human- mrna", palette[4])
ec5 <- barplot_exon_count(mRNAs_ref, "Reference Macaque- mrna", palette[5])+labs(x = "Exon count") 
ggarrange( ec1,ec2,ec3,ec4,ec5,  ncol=1, nrow=5, top = "Transcripts Lengths", legend  = NULL) 
# Exon count generally lower in lncrnas than mrnas (Ok. same as sources)

# --------------------
## Transcript lengths: still weird, very long ones ...
# --------------------
tl1 <-transcript_length_density(novel," NOVEL",palette[1]) 
tl2 <-transcript_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
tl3 <-transcript_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
tl4 <-transcript_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
tl5 <-transcript_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Transcript length")
ggarrange( tl1,tl2,tl3,tl4,tl5,  ncol=1, nrow=5, top = "Transcripts Lengths",  legend="right")

# --------------------
## Exon lengths - Densities 
# --------------------
p1 <-exon_length_density(novel," NOVEL",palette[1]) 
p2 <-exon_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
p3 <-exon_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
p4 <-exon_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
p5 <-exon_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Exon length")
ggarrange( p1,p2,p3,p4,p5,  ncol=1, nrow=5, top = "Exon Lengths",  legend="right")


# --------------------
## Exon lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(novel, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = c('Novel LncRNAs', 'Known lncRNAs: Macaque ', 'Known lncRNAs: Human', "Known mRNAs: Human", "Known mRNAs: Macaque")),  y = range )) +
        labs( x = "", y = "" , title = "Exon length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,1000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1
```

#Expression HTSEQ 

```{r}
# --------------------
## Create dds and dge - MERGED 
# --------------------
dir_hisat=  file.path(datadir, "02_RNA-Seq_rheMac10/03_hisat")
dir_counts = file.path(datadir, "02_RNA-Seq_rheMac10/06_counts")
htseq_files <- iterate_files(dir_counts, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

dds <- create_dds(htseq_files,dir_counts)
dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))




# calc logcpm 
assays(dds)$logCPM <- cpm(dge, log = TRUE,prior.count = 0.5)

# create dataframe w/ logcpms
expression_logcpm <-  as.data.frame(assays(dds)$logCPM)
mask_expressed_in_3 <- rowSums(expression_logcpm > 1) > 2
#select the ones expressed in at least 3 samples
expression_logcpm_filtered <- expression_logcpm[mask_expressed_in_3,]

table(novel$gene_id %in% rownames(expression_logcpm_filtered))
novel_expressed <- novel[novel$gene_id %in% rownames(expression_logcpm_filtered)]
length(unique(novel_2$gene_id))

# --------------------
## Reference
# --------------------
dir_counts_ref = file.path(datadir, "02_RNA-Seq_rheMac10/06_counts_ref")
htseq_files_ref <- iterate_files(dir_counts_ref, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))

# calc logcpm 
assays(dds_ref)$logCPM <- cpm(dge_ref, log = TRUE,prior.count = 0.5)

counts(dds_ref)
assays(dds_ref)$logCPM
# create dataframe w/ logcpms
expression_logcpm_ref <-  as.data.frame(assays(dds_ref)$logCPM)
```



```{r}

unique(stringtie_gtf[(str_detect(stringtie_gtf$gene_id,"MSTRG", negate = TRUE))]$gene_id)


# calc mean expression across samples
mean_expression <- data.frame(id=rownames(expression_logcpm ), Means=rowMeans(expression_logcpm [,-1]))
mean_expression_ref <- data.frame(id=rownames(expression_logcpm_ref ), Means=rowMeans(expression_logcpm_ref[,-1]))
# add column to mean expression datatframe
mean_expression['type'] <- "0"
mean_expression <- add_type(mean_expression, novel$gene_id, "novel")
mean_expression <- mean_expression[!mean_expression$type == "0",]
table(mean_expression$type)
mean_expression_ref['type'] <- "0"
mean_expression_ref <- add_type(mean_expression_ref, lncRNAs_ref$gene_id, "lncrna")
mean_expression_ref <- add_type(mean_expression_ref, mRNAs_ref$gene_id, "mrna")
mean_expression_ref  <- mean_expression_ref[!mean_expression_ref$type == "0",]
table(mean_expression_ref$type)

me <- rbind(mean_expression, mean_expression_ref)

mean_expression <- mean_expression_ref
length(unique(mean_expression$id))
length(unique(mean_expression[mean_expression$id %in% mRNAs_ref$gene_id,]$id))

ggplot(me, aes(x=factor(type,level = c("lncrna", "novel", "mrna")) , y=Means, fill = type )) + 
    geom_boxplot(color="darkgrey",  alpha=0.7)+
    scale_fill_brewer(palette="Dark2")+
    labs(y = "", x = "", title = "Expression" )+
    theme(legend.title=element_blank())+
    theme(plot.title = element_text(hjust = 0.5))+
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))


```







