---
title: "lncrna annotation "
author: "Luisa Santus"
date: "12/19/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data and libraries

```{r results='hide', message=FALSE, warning=FALSE}
library(rtracklayer); library(VennDiagram); library(stringr); library(ggplot2); library(grid); library(gridExtra); library(UpSetR); library(ComplexHeatmap)
library(GenomicRanges); library(dplyr); library(zeallot); library(DESeq2); library(edgeR);  library("ggplotify"); library("grid"); library(tidyverse)
library(tximport); library(tximportData); library(plyr); library(Gviz); library(ggpubr)

# Reused paths
projdir = "/home/luisas/Desktop/cluster/proj/data/"
datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all"
assemblies <- "/home/luisas/Desktop/cluster/assemblies"

# Import Utils 
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/","code/ebola/src/scripts/Analysis/utils")
source(file.path(scripts_dir,"de_utils.R")); source(file.path(scripts_dir,"lncrna_annotation_utils.R"))

# Human training set
lncRNAs_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownlncRNAs.gtf"))
mrna_ref_human <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/gencode.v26.GRCh38.annotation_knownMrnas.gtf"))
lncRNAs_ref_human_gene <- lncRNAs_ref_human[lncRNAs_ref_human$type == "gene",]

# The lncRNAs and mRNAs in the reference
lncRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_lncrna.gtf"))
mRNAs_ref <- import(file.path(datadir, "01_PreliminaryFiles_rheMac10/gene_annotations/rheMac10_EBOV-Kikwit_mrnas.gtf"))

human_annotation <- import(file.path(dir_assemblies, "ensembl/release-96/Homo_sapiens_hg38/Homo_sapiens.GRCh38.96.gtf"))
```


First, apply the filtering to each dataset separately 
 - identify novel transcripts ( gff compare "u", "x")
 - deplete non-concordant predictions
 - retain only ones expressed in more than 3 samples
 
```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------
#     RiboDepleted
# ---------------------


# -----------------
# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_BatchZyagen/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
dir_counts_ref = file.path(datadir, "02_RNA-Seq_BatchZyagen/06_counts_all")
htseq_files_ref <- iterate_files(dir_counts_ref, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

# -----------------
# Select only the NOVEL ones (gff compare )
mask_uix <- merged_annotated$class_code == "u" | merged_annotated$class_code == "x" ;  mask_uix[is.na(mask_uix)] <- FALSE
table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

# -----------------
# Remove Non-Concordant
df <- data.frame()
df <- rbind(df, data.frame(gene_id = novel$gene_id, transcript_id = novel$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
novel_concordant <- novel[!(novel$gene_id %in% unconcordant_prediction$gene_id)]
print("Novel - concordant")
print(length(unique(novel_concordant$transcript_id)))

# -----------------
# Expression Levels ( filter and plot )
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
# Calc logcpm 
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
# Create dataframe w/ logcpms
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
colnames(expression_logcpm)
## ADD only if we want to select a certain tissue
#expression_logcpm <- expression_logcpm[,str_detect(colnames(expression_logcpm), "Liver")]
#expression_logcpm

# select only when > than 1 at least 3 tissues
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
expression_logcpm <- expression_logcpm[mask,]

# Plotting
mean_expression <- data.frame(id=rownames(expression_logcpm ), Means=rowMedians(expression_logcpm))

print(head(mean_expression))
mean_expression['type'] <- "0"
mean_expression <- add_type(mean_expression, novel_concordant$gene_id, "novel")
mean_expression <- add_type(mean_expression, lncRNAs_ref$gene_id, "lncrna")
mean_expression <- add_type(mean_expression, mRNAs_ref$gene_id, "mrna")
mean_expression <- mean_expression[!mean_expression$type == "0",]

length(unique(novel_concordant[novel_concordant$gene_id %in% mean_expression$id]$gene_id))
length(unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% mean_expression$id]$gene_id))
length(unique(mRNAs_ref[mRNAs_ref$gene_id %in% mean_expression$id]$gene_id))

ggplot(mean_expression, aes(x=factor(type,level = c("lncrna", "novel", "mrna")) , y=Means, fill = type )) + 
      geom_boxplot(color="darkgrey",  alpha=0.7, fill = c("#1F78B4","#A6CEE3","#FB9A99"))+
      labs(y = "logCPM", x = "", title = "Expression" )+
      theme(legend.title=element_blank())+
      theme(plot.title = element_text(hjust = 0.5))+
      theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))

novel_expressed <- novel_concordant[novel_concordant$gene_id %in% mean_expression$id]
length(unique(novel_expressed$gene_id))


export(novel_expressed, file.path(data_dir, "03_novel_lncrnas/novel_expressed_batchzyagen.gtf"))
```


#Stats 

```{r results='hide', message=FALSE, warning=FALSE}
# Palette for all plots below
palette <-brewer.pal(5,"Paired")
  # --------------------
## EXON COUNT
# --------------------

ec1 <- barplot_exon_count(novel_expressed, "Novel", palette[1])
ec2 <- barplot_exon_count(lncRNAs_ref, "lncRNAs - Reference Macaque", palette[2])
ec3 <- barplot_exon_count(lncRNAs_ref_human, "lncRNAs - Reference Human", palette[3])+labs(y = "Frequency")
ec4 <- barplot_exon_count(mrna_ref_human, "mRNAs - Reference Human", palette[4])
ec5 <- barplot_exon_count(mRNAs_ref, "mRNAs - Reference Macaque", palette[5])+labs(x = "Number of Exons") 
ggarrange( ec1,ec2,ec3,ec4,ec5,  ncol=1, nrow=5, top = "Transcripts Lengths") 
# Exon count generally lower in lncrnas than mrnas (Ok. same as sources)

# --------------------
## Transcript lengths: still weird, very long ones ...
# --------------------
tl1 <-transcript_length_density(novel_expressed," NOVEL",palette[1]) 
tl2 <-transcript_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
tl3 <-transcript_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
tl4 <-transcript_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
tl5 <-transcript_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Transcript length")
ggarrange( tl1,tl2,tl3,tl4,tl5,  ncol=1, nrow=5, top = "Transcripts Lengths",  legend="right")


# --------------------
## Transcript lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_transcript_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = c('Novel LncRNAs', 'Known lncRNAs: Macaque ', 'Known lncRNAs: Human', "Known mRNAs: Human", "Known mRNAs: Macaque")),  y = range )) +
        labs( x = "", y = "" , title = "Transcript length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,10000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1

# --------------------
## Exon lengths - Densities 
# --------------------
p1 <-exon_length_density(novel_expressed," NOVEL",palette[1]) 
p2 <-exon_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
p3 <-exon_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
p4 <-exon_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
p5 <-exon_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Exon length")
ggarrange( p1,p2,p3,p4,p5,  ncol=1, nrow=5, top = "Exon Lengths",  legend="right")

# --------------------
## Exon lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = c('Novel LncRNAs', 'Known lncRNAs: Macaque ', 'Known lncRNAs: Human', "Known mRNAs: Human", "Known mRNAs: Macaque")),  y = range )) +
        labs( x = "", y = "" , title = "Exon length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,1000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1
```


Repeat the same analysis with polyA dataset 

```{r results='hide', message=FALSE, warning=FALSE}
# ---------------------
#    PolyA
# ---------------------


# -----------------
# Feelnc output files
candidates_lncRNA <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/candidate_lncRNA.gtf"))
lncRNAs_out <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.lncRNA.gtf"))
mRNAs_out <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/feelnc_codpot_out/candidate_lncRNA.gtf.mRNA.gtf"))
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_external/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
dir_counts_ref = file.path(datadir, "02_RNA-Seq_external/06_counts_all")
htseq_files_ref <- iterate_files(dir_counts_ref, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

# -----------------
# Select only the NOVEL ones (gff compare )
mask_uix <- merged_annotated$class_code == "u" | merged_annotated$class_code == "x" ;  mask_uix[is.na(mask_uix)] <- FALSE
table(merged_annotated[mask_uix]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

# -----------------
# Remove Non-Concordant
df <- data.frame()
df <- rbind(df, data.frame(gene_id = novel$gene_id, transcript_id = novel$transcript_id, pred = "lncrna" ))
df <- rbind(df, data.frame(gene_id = mRNAs_out$gene_id, transcript_id = mRNAs_out$transcript_id, pred = "mrna" ))
unconcordant_prediction <- df %>%  dplyr::group_by(gene_id) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(pred)) %>%  dplyr::filter( Unique_Elements > 1)
novel_concordant <- novel[!(novel$gene_id %in% unconcordant_prediction$gene_id)]
print("Novel - concordant")
print(length(unique(novel_concordant$transcript_id)))

# -----------------
# Expression Levels ( filter and plot )
dds_ref <- create_dds(htseq_files_ref,dir_counts_ref)
dge_ref <- DGEList(counts = assays(dds_ref)$counts, group = dds_ref$dpo, genes = rownames(dds_ref), samples = colData(dds_ref))
# Calc logcpm 
assays(dds_ref)$logCPM <- edgeR::cpm(dge_ref, log = TRUE,prior.count = 0.5)
# Create dataframe w/ logcpms
expression_logcpm <-  as.matrix(assays(dds_ref)$logCPM)
colnames(expression_logcpm)
## ADD only if we want to select a certain tissue
#expression_logcpm <- expression_logcpm[,str_detect(colnames(expression_logcpm), "Liver")]
#expression_logcpm

# select only when > than 1 at least 3 tissues
mask<- rowSums(as.data.frame(expression_logcpm) > 1) > 2
expression_logcpm <- expression_logcpm[mask,]

# Plotting
mean_expression <- data.frame(id=rownames(expression_logcpm ), Means=rowMedians(expression_logcpm))

print(head(mean_expression))
mean_expression['type'] <- "0"
mean_expression <- add_type(mean_expression, novel_concordant$gene_id, "novel")
mean_expression <- add_type(mean_expression, lncRNAs_ref$gene_id, "lncrna")
mean_expression <- add_type(mean_expression, mRNAs_ref$gene_id, "mrna")
mean_expression <- mean_expression[!mean_expression$type == "0",]

length(unique(novel_concordant[novel_concordant$gene_id %in% mean_expression$id]$gene_id))
length(unique(lncRNAs_ref[lncRNAs_ref$gene_id %in% mean_expression$id]$gene_id))
length(unique(mRNAs_ref[mRNAs_ref$gene_id %in% mean_expression$id]$gene_id))

ggplot(mean_expression, aes(x=factor(type,level = c("lncrna", "novel", "mrna")) , y=Means, fill = type )) + 
      geom_boxplot(color="darkgrey",  alpha=0.7, fill = c("#1F78B4","#A6CEE3","#FB9A99"))+
      labs(y = "logCPM", x = "", title = "Expression" )+
      theme(legend.title=element_blank())+
      theme(plot.title = element_text(hjust = 0.5))+
      theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))

novel_expressed <- novel_concordant[novel_concordant$gene_id %in% mean_expression$id]
length(unique(novel_expressed$gene_id))

export(novel_expressed, file.path(data_dir, "03_novel_lncrnas/novel_expressed_external.gtf"))

```


#Stats 

```{r results='hide', message=FALSE, warning=FALSE}
# Palette for all plots below
palette <-brewer.pal(5,"Paired")
  # --------------------
## EXON COUNT
# --------------------

ec1 <- barplot_exon_count(novel_expressed, "Novel", palette[1])
ec2 <- barplot_exon_count(lncRNAs_ref, "lncRNAs - Reference Macaque", palette[2])
ec3 <- barplot_exon_count(lncRNAs_ref_human, "lncRNAs - Reference Human", palette[3])+labs(y = "Frequency")
ec4 <- barplot_exon_count(mrna_ref_human, "mRNAs - Reference Human", palette[4])
ec5 <- barplot_exon_count(mRNAs_ref, "mRNAs - Reference Macaque", palette[5])+labs(x = "Number of Exons") 
ggarrange( ec1,ec2,ec3,ec4,ec5,  ncol=1, nrow=5, top = "Transcripts Lengths") 
# Exon count generally lower in lncrnas than mrnas (Ok. same as sources)

# --------------------
## Transcript lengths: still weird, very long ones ...
# --------------------
tl1 <-transcript_length_density(novel_expressed," NOVEL",palette[1]) 
tl2 <-transcript_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
tl3 <-transcript_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
tl4 <-transcript_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
tl5 <-transcript_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Transcript length")
ggarrange( tl1,tl2,tl3,tl4,tl5,  ncol=1, nrow=5, top = "Transcripts Lengths",  legend="right")


# --------------------
## Transcript lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_transcript_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_transcript_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_transcript_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = c('Novel LncRNAs', 'Known lncRNAs: Macaque ', 'Known lncRNAs: Human', "Known mRNAs: Human", "Known mRNAs: Macaque")),  y = range )) +
        labs( x = "", y = "" , title = "Transcript length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,10000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1

# --------------------
## Exon lengths - Densities 
# --------------------
p1 <-exon_length_density(novel_expressed," NOVEL",palette[1]) 
p2 <-exon_length_density(lncRNAs_ref," REF MACAQUE",palette[2]) 
p3 <-exon_length_density(lncRNAs_ref_human,"REF HUMAN",palette[3]) + labs( y  = "Density")
p4 <-exon_length_density(mrna_ref_human," REF MACAQUE",palette[4]) 
p5 <-exon_length_density(mRNAs_ref,"REF HUMAN",palette[5]) + labs( x = " Exon length")
ggarrange( p1,p2,p3,p4,p5,  ncol=1, nrow=5, top = "Exon Lengths",  legend="right")

# --------------------
## Exon lengths - Boxplot
# --------------------
df <- data.frame()
df <- rbind(df,data.frame(calc_exon_length(novel_expressed, "Novel LncRNAs")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref, "Known lncRNAs: Macaque ")))
df <- rbind(df,data.frame(calc_exon_length(lncRNAs_ref_human, "Known lncRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mrna_ref_human, "Known mRNAs: Human")))
df <- rbind(df,data.frame(calc_exon_length(mRNAs_ref, "Known mRNAs: Macaque")))

p1 <- ggplot(df, aes(x = factor(type, level = c('Novel LncRNAs', 'Known lncRNAs: Macaque ', 'Known lncRNAs: Human', "Known mRNAs: Human", "Known mRNAs: Macaque")),  y = range )) +
        labs( x = "", y = "" , title = "Exon length")+
        theme(axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size=10))+
        geom_boxplot(outlier.shape=NA, fill = palette,color = "Darkgrey") + ylim(0,1000)+
        theme(legend.title=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1
```


# ------------------------------------
# Merge them together 
```{r results='hide', message=FALSE, warning=FALSE}

# Novel from ribodepletion compared to polyA 
import(file.path(data_dir, "03_novel_lncrnas/01_gffcompare/gff_compare_lncrnas"))
mask_u <- merged_annotated$class_code == "u";  mask_u[is.na(mask_u)] <- FALSE
table(merged_annotated[mask_u]$class_code)
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
novel_transcript_ids <- merged_annotated[mask_uix]$transcript_id
novel <- merged_annotated[merged_annotated$transcript_id %in% novel_transcript_ids]
print("Novel - gff compare")
print(length(unique(novel$transcript_id)))

## I could check some stats about the new ones.

# I just append one to the otherm

```















# -------------------------------------------------------------
# Known ones: Quick benchmark - should be prob put in the end
```{r results='hide', message=FALSE, warning=FALSE}
# Obtain novels from gff compare 
# remove genes, whose transcript are predicted as both lncrnas and mrnas
merged_annotated <- import(file.path(datadir, "02_RNA-Seq_all/05_lncrnaAnnotation_no_l_option/benchmark_known/05_lncrnaAnnotation_no_l_option/feelnc_gencode_linc/01_gffcompare/merged.annotated.gtf"))
mask_eq <- merged_annotated$class_code == "="  ;  mask_eq[is.na(mask_eq)] <- FALSE
# comparing the merged one with the reference. unkown, intergenic and intronic transcipts are kept.
known_identified <- (unique(merged_annotated[mask_eq]$transcript_id))
missed <- setdiff(unique(lncRNAs_ref$transcript_id), known_identified)

# -----------------
# PLOT 
# -----------------
df <- data.frame()
df <- rbind(df, data.frame(dataset = "Novel", number = length(unique(novel_expressed$transcript_id)), type = "Novel"))
df <- rbind(df, data.frame(dataset = "Correctly identified from rheMac10", number = length(known_identified), type = "Known correct"))
df <- rbind(df, data.frame(dataset = "Uncorrectly classified from rheMac10", number = length(missed), type = "Known missed"))


p1 <- ggplot(data=df, aes(x=type, y=number, fill=dataset)) +
  geom_bar(stat="identity", position=position_dodge(), width=0.85)+
  geom_text(aes(label=number),color="black",
            position = position_dodge(0.9), size=3.5)+
  scale_fill_brewer(palette="Paired")+
  labs(title=" LncRNAs prediction: novel lncRNAs vs known lncRNAs") +
  ylab("# Transcripts") +
  xlab("") +
  theme(legend.title=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+
        theme(plot.title = element_text(hjust = 0.5))
p1 

# calculate the proportion of the correctly idententified ones
prop_correctly = (length(correctly_identified)/(length(correctly_identified) + length(missed)))*100
print(paste0("Proportion correclty identified: ", prop_correctly))
```


 Save some files 
```{r results='hide', message=FALSE, warning=FALSE}
seqlevelsStyle(novel_expressed) <- "UCSC"
export(novel_expressed, file.path(datadir,"/02_RNA-Seq_all/07_liftover/novel_rhemac10_expressed_all.gtf"))
 
#novel_intronic <- novel_expressed[novel_expressed$type == "transcript" & novel_expressed$class_code == "i"]
novel_antisense<- novel_expressed[novel_expressed$type == "transcript" & novel_expressed$class_code == "x"]
novel_intergenic <- novel_expressed[novel_expressed$type == "transcript" & novel_expressed$class_code == "u"]


#export(novel_intronic, file.path(datadir,"/02_RNA-Seq_all/07_liftover/novel_rhemac10_expressed_intronic.gtf"))
export(novel_antisense, file.path(datadir,"/02_RNA-Seq_all/07_liftover/novel_rhemac10_expressed_antisense.gtf"))
export(novel_intergenic, file.path(datadir,"/02_RNA-Seq_all/07_liftover/novel_rhemac10_expressed_intergenic.gtf"))

?import
```


```{r}
# I have the gff with all the orthologs in human
seqlevelsStyle(human_annotation) <- "UCSC"
human_orthologs_all <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/02_RNA-Seq_all/07_liftover/human_novel_expressed.gtf")
#human_orthologs_intronic <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/02_RNA-Seq_all/07_liftover/human_novel_filtered_outliers_intronic.gtf")
human_orthologs_antisense <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/02_RNA-Seq_all/07_liftover/human_novel_expressed_antisense.gtf")
human_orthologs_intergenic <- import("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/02_RNA-Seq_all/07_liftover/human_novel_expressed_intergenic.gtf")

plot_biotype_orthologs <- function(human_orthologs, title){
    # Find overlaps and biotypes of orthologs 
    overlaps <- findOverlaps(human_orthologs, human_annotation)
    hits_in_human <- human_annotation[subjectHits(overlaps)]
    
    # Create DF with queryhit on the left and subject hit's biotype on the right
    df <- data.frame(matrix(ncol = 2, nrow = 0), stringsAsFactors = FALSE)
    colnames(df) <- c("i", "b")
    for(i in unique(queryHits(overlaps) )){
      overlaps_only <- overlaps[queryHits(overlaps) == i]
      biotypes <- unique(human_annotation[subjectHits(overlaps_only)]$gene_biotype)
      for (b in biotypes){df[nrow(df) + 1,] <-  c(i,b) }}
  
    # Check if assigned to multiple labels ( if so assign label more)
    double_found <- df %>%  dplyr::group_by(i) %>% dplyr::summarise(Unique_Elements =  dplyr::n_distinct(b)) %>%  dplyr::filter( Unique_Elements > 1)
    
    df[df$i %in% double_found$i, ]$b <- "more"
    
    o <- (df %>% distinct())$b
    o <- o[o == "protein_coding" | o == "lincRNA" | o == "antisense" | o == "more" ]
    p1 <- ggplot(data.frame(o), aes(x=o, fill = o)) + labs(title=title, x = "", y = "")+
      geom_bar()+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "darkgrey"))+theme(legend.title=element_blank())+ scale_fill_brewer(palette="Accent")
    p1
  return(p1)
}


# ---------------------------------------------------------
```


```{r}
plot_biotype_orthologs(human_orthologs_all, "Human orthologs all")
#plot_biotype_orthologs(human_orthologs_intronic, "human_orthologs_intronic")
plot_biotype_orthologs(human_orthologs_antisense, "human_orthologs_antisense")
plot_biotype_orthologs(human_orthologs_intergenic, "human_orthologs_intergenic")
```














