---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r setup, include=FALSE,results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(scater)
library(SingleCellExperiment)
library(purrr)
library(stringr)
library(mvoutlier)
library(dropbead)
library(scater)
library(stringr)
library(scran)
library(cowplot)
library(rtracklayer)
library(org.Mmu.eg.db)
library(graphics)
library(RCy3)

source("utils/sc_utils.R")
source("utils/de_utils.R")

datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_final_catalogue/rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))
result_path <- file.path("/home/luisas/Desktop/cluster/data/RObjects/results")
robjectsdir<- "/home/luisas/Desktop/cluster/data/RObjects/"
```

## Single Cell Analysis

------------------------
  Create Seurat Objects
------------------------

I import the data and merge them into one single Seurat object to proceed with the analysis.

Obtain 
```{r cars}
# Load the PBMC dataset
data_dir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_scRNA-Seq_inVivo_rhemac10/04_DigitalExpressionMatrix/"
files <- iterate_files(data_dir, ".dge.txt.gz")
files_umis <- files[!str_detect(files, "reads" )]

# Umis 
# all matrices files from single cell, still separated per lane
list <- map(files_umis[2:length(files_umis)], get_Seurat_object)
pbmc <- get_Seurat_object(files_umis[1])
list_names <- lapply(files_umis, function(x) str_replace_all(unlist(rev(unlist(str_split(x, pattern= "/"))[-1])[[1]]), "_", "-"))

length(list)
length(list_names)
unique(unlist(list_names))
# -------------
#   Merge
# -------------
# Merge all seurats object in one 

# Eemove objects where nocells were identified 
list_detected_only <- list[unlist(lapply(list, function(x) dim(x)[1] != 0 ))]
list_names <- list_names[unlist(lapply(list, function(x) dim(x)[1] != 0 ))]

pbmc.big <- merge(pbmc, y = list_detected_only, add.cell.ids = list_names, project = "PBMCbiginvivo")
# Check if all ids available
unique(sapply(X = strsplit(colnames(pbmc.big), split = "_"), FUN = "[", 1))
table(pbmc.big$orig.ident)
dim(pbmc.big)
saveRDS(pbmc.big, file.path(robjectsdir,"immune.combined_invivo_1.rds"))
```
##---------------------------------------------------------
##                           Cell QC 
##  Remove cells with too little or too high UMIS count 
##  Remove cell based on number of detected genes per cell
##----------------------------------------------------------
```{r pressure, results='hide', message=FALSE, warning=FALSE}
pbmc <- readRDS(file.path(robjectsdir,"invivo_immunecombined_cellqc.rds"))
pbmc
pbmc_sc <- as.SingleCellExperiment(pbmc.big)
pbmc_sc <- calculateQCMetrics(pbmc_sc)

# Library size
lib_size_hist <- plot_histogram( sce = pbmc_sc, qc_metric = "total_counts", title = "Library Size (total UMI)", log_scale = FALSE)

# Number of detected genes
n_detec_hist <- plot_histogram(sce = pbmc_sc,  qc_metric = "total_features_by_counts",title = "Number of detected genes", log_scale = FALSE)

lsizeplot <- lib_size_hist +
  geom_vline(xintercept = c(500, 10000), linetype = "dashed", color = "red") + xlim(0,2000)

ndectplot <- n_detec_hist +
  geom_vline(xintercept = c(600, 2000), linetype = "dashed", color = "red")

lsizeplot
ndectplot
```
# Filtering 

```{r}
pbmc <- pbmc_sc
keep_cells <-
  pbmc$total_counts > 1000 & pbmc$total_counts < 10000 &
  pbmc$total_features_by_counts > 600 & pbmc$total_features_by_counts < 2000 
table(keep_cells)
pbmc <- pbmc[, keep_cells]
saveRDS(pbmc,file.path(robjectsdir,"invivo_immunecombined_cellqc.rds"))
```


##----------------------------
##           Gene QC 
##  Exclude genes not showing up in a least 10 cells
##----------------------------  
Let us compute the number of cells that each gene has at least 1 UMI.
We see two peaks, the firt one of which corresponds to lowly expressed genes. As explained in [Luecken MD et al.](https://www.embopress.org/doi/pdf/10.15252/msb.20188746): "a guideline to setting this threshold is to use the minimum cell cluster size that is of interest and leaving some leeway for dropout effects". As we will not rely on clusters that have fewer than 15 cells, we will use it as a filter:

```{r}

# CLUSTER !!! 
# Do not delete this lines. The following steps were performed on the cluster for memory issues.
n_cells <- rowSums(as.matrix(counts(pbmc)) > 0)
#pbmc <- pbmc[n_cells > 10, ]
#saveRDS(pbmc, file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellandgeneqc.rds")
pbmc <- readRDS(file.path(robjectsdir,"invivo_immunecombined_cellqc_aftercellandgeneqc.rds"))

```

##-------------------------------------------------------------------------------
##    Doublet Detection: Scrublet 
##-------------------------------------------------------------------------------


```{r cars}
library(reshape2)
pbmc <- readRDS(file.path(robjectsdir,"invivo_immunecombined_cellqc_aftercellandgeneqc.rds"))
pbmc <- as.Seurat(pbmc)
# Extract count matrix
m <-as.matrix(GetAssayData(object = pbmc, slot = "counts"))
sparse.gbm <- Matrix(m , sparse = T )
writeMM(obj = sparse.gbm, file=file.path(robjectsdir,"sparse_invivo.mtx"))
```

# External Script Warning!! 
# --> Run Scrublet in Python

```{r cars}
cell_mask_scrublet <- read.table("/home/luisas/Desktop/cluster/data/RObjects/predicted_doublet_mask_invivo.txt")
doublet_scores_scrublet <- read.table("/home/luisas/Desktop/cluster/data/RObjects/predicted_doublet_scores_invivo.txt")

list <- as.list(as.character(cell_mask_scrublet$V1))
mask <- list == "True"
new <- pbmc[,!mask]

dim(new)
table(mask)
pbmc<- seurat_preprocess_standard(new)

# Now check which cells have less than X genes expressed --> cannot see anything special :/
cells <-  colnames(pbmc[,pbmc$total_features_by_counts > 1000])
table(pbmc$total_counts < 2000)
cells <- colnames(pbmc[,pbmc$total_counts > 5000])
plot_cells(pbmc, cells)

saveRDS(new, file.path(robjectsdir,"invivo_immunecombined_cellqc_aftercellandgeneqc_afterScrublet.rds"))
immune.combined <- readRDS(file.path(robjectsdir,"invivo_immunecombined_cellqc_aftercellandgeneqc_afterScrublet.rds"))
```

##-------------------------------------------------------------------------------
##                         Stimulated data analysis 
##-------------------------------------------------------------------------------



# ---- Integration
# Running integration on the cluster 

# -------------------------------------
#     After integration in cluster 
# -------------------------------------

```{r cars}
# Check the ones without integration for the moment 
#immune.combined <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet_after-integration.rds"))
# Import the newly computed cellswith 5k features and 
immune.combined <- readRDS(file.path(result_path, "invivo_immunecombined_cellqc_aftercellandgeneqc_afterScrublet.rds"))

# add some labeling 
immune.combined$orig.ident <- gsub("D_", "D-", gsub("-", "_",immune.combined$orig.ident))
immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "_")[[1]][1]))))
immune.combined$dpi <- immune.combined$cond <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][2])))
immune.combined$batch <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][3])))
immune.combined$analysis <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][4])))
immune.combined$sample <- paste0(immune.combined$individual," ", immune.combined$dpi)


# Identify and remove MDCKs 
immune.combined <- NormalizeData(immune.combined)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.5)

# Visualize clustering
DimPlot(immune.combined, reduction = "umap", group.by = "batch")
DimPlot(immune.combined, reduction = "umap", label = TRUE)
# Identify MCDK - remove
mcdk <- c("COL5A2", "EMP1")
plot_genes(immune.combined, mcdk, threshold = 2 , title = "MCDK-Cell markers expression", col = "#CA4866")
# Remove MCDK 
idents_to_keep <- setdiff(unique(Idents(immune.combined)), c(9,11))
immune.combined <- subset(immune.combined, idents = idents_to_keep)
DimPlot(immune.combined, reduction = "umap", group.by = "batch")
```



```{r cars}
#------------------
# Integration 
# -----------------
immune.combined.list <- SplitObject(immune.combined, split.by = "batch")

immune.combined.list <- lapply(X = immune.combined.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
immune.anchors <- FindIntegrationAnchors(object.list = immune.combined.list, dims = 1:20)
saveRDS(immune.anchors, file.path(robjectsdir, "immune_anchors_invivo.rds"))
immune.combined <- IntegrateData(anchorset = immune.anchors, dims = 1:20)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
DefaultAssay(immune.combined) <- "integrated"
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.5)
saveRDS(immune.combined, file.path(robjectsdir, "invivo_immunecombined_cellqc_aftercellandgeneqc_afterScrublet_integrated.rds"))

# ----------------------------
# Visualize the clusters
p1 <- DimPlot(immune.combined, reduction = "umap", group.by = "dpi", pt.size = 0.1, cols = brewer.pal(10, "Paired"))+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
p2 <- DimPlot(immune.combined, reduction = "umap", group.by = "individual")
p3 <- DimPlot(immune.combined, reduction = "umap", label = TRUE)
p4 <- DimPlot(immune.combined, reduction = "umap", group.by = "batch", pt.size = 0.2, cols = brewer.pal(3, "Paired"))
p5 <- DimPlot(immune.combined, reduction = "umap", group.by = "sample", pt.size = 0.00001)
p4 +theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

p1; p2; p3; p4; p5
```


```{r cars}
# ----------------------------
# FIND MARKERS  for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(immune.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
saveRDS(pbmc.markers, file.path(robjectsdir,"markers_invivo_normal.rds"))
pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)

b <- c("CD79B", "MS4A1", "CD19")
monocytes <- c("CD14", "LYZ", "CTSB", "PSAP")
t <- c("IL7R","CD3D", "CD28", "TRIB2", "SPOCK2")
nk <- c( "KLRB1")
dc <- c("IRF8")
neutrophil <- c("CD177", "SOD2")
platelet <- c("PPBP", "PF4")
mdk <- c("COL5A2", "EMP1")

marker.genes <- c(monocytes, nk, b,t)
plot_genes(immune.combined,monocytes,threshold= 2, title = "Monocytes marker expression", col = "purple")
plot_genes(immune.combined, nk, threshold = 2 , title = "NK markers expression", col = "purple")
plot_genes(immune.combined, b, threshold = 2 , title = "B markers expression", col = "purple")
plot_genes(immune.combined, t, threshold = 2 , title = "T-Cell markers expression", col = "purple")
plot_genes(immune.combined, neutrophil, threshold = 4 , title = "Neutrophil-Cell markers expression", col = "purple")


# No integration 
#immune.combined <- RenameIdents(immune.combined, `0` = "B", `1` = "T", `2` = "T", 
#    `3` = "T", `4` = "Myeloid", `5` = "B", `6` = "Myeloid", `7` = "Myeloid", `8` = "T", `9` = "MDCK", 
#    `10` = "Myeloid", `11` = "MDCK", `12` = "Myeloid", `13` = "Myeloid",`14` = "Myeloid", `15` = "Myeloid", #`16` = "T")
# Integrated w/mcdk
#immune.combined <- RenameIdents(immune.combined, `0` = "B", `1` = "T", `2` = "T", 
#    `3` = "T", `4` = "Myeloid", `5` = "Myeloid", `6` = "Myeloid", `7` = "MCDK", `8` = "Myeloid", `9` = "MDCK",
#    `10` = "Myeloid", `11` = "Myeloid", `12` = "Myeloid", `13` = "Myeloid")

immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "B", `2` = "T", 
   `3` = "T", `4` = "B", `5` = "Myeloid", `6` = "Myeloid", `7` = "Myeloid", `8` = "Myeloid", `9` = "Myeloid",
    `10` = "Myeloid", `11` = "Myeloid", `12` = "Myeloid", `13` = "Myeloid")
# Visualize Markers
DotPlot(immune.combined, features = rev(marker.genes), cols = c("blue", "red", "purple"), dot.scale = 8) + RotatedAxis()
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =brewer.pal(3, "Set2"), label.size = 6)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))


```
```{r cars}
# --------------------
#     EbolaGenes
#--------------------

ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unkown", sep ="-")]
ebola_genes
FeaturePlot(immune.combined, features  = ebola_genes)
plot_genes(immune.combined, ebola_genes,2, col = "darkred", "Ebola RNA expression")
```

# ------------------------------------
# Differential expression analysis
# ------------------------------------

# Group DPI to gather more 
```{r cars}
immune.combined$group <- immune.combined$dpi

immune.combined$group <- gsub("D-04" , "baseline",immune.combined$group)
immune.combined$group <- gsub("D000" , "baseline",immune.combined$group)

immune.combined$group <- gsub("D003" , "early",immune.combined$group)

immune.combined$group <- gsub("D004" , "middle",immune.combined$group)
immune.combined$group <- gsub("D005" , "middle",immune.combined$group)

immune.combined$group <- gsub("D006" , "late",immune.combined$group)
immune.combined$group <- gsub("D007" , "late",immune.combined$group)
immune.combined$group <- gsub("D008" , "late",immune.combined$group)

table(immune.combined$group)
DimPlot(immune.combined, group.by = "group", reduction = "umap")

saveRDS(immune.combined, file.path(robjectsdir, "immune.combined_invivo_aftercomplete_preprocessing.rds"))
```

# -----------------------------------
# Differential Expression Analysis 
# -----------------------------------

```{r cars}
source("utils/sc_utils.R")
source("utils/de_utils.R")

datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_final_catalogue/rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names.gtf"))
ebola_ref <- rtracklayer::import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

mrna_ref_id <- unique(ref$gene_name[ref$gene_biotype == "protein_coding"])
immune.combined <- readRDS( file.path(robjectsdir, "immune.combined_invivo_aftercomplete_preprocessing.rds"))
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(mrna_ref_id, "_" , "-")]

# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# if i want to add also mirnas
#lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype %in% c("lncRNA", "miRNA")])
lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype %in% c("lncRNA")])


# i remove the gene names so i can match 
annotation_with_no_genename <- lapply(lnc_ref_id, function(x) str_split(x, "_")[[1]][1])
annotated_lncrnas <- rownames(immune.combined)[ lapply(rownames(immune.combined), function(x) str_split(x, "-")[[1]][1]) %in% annotation_with_no_genename]
annotated_lncrna_gene_names_only <- lnc_ref_id[!grepl("_unkown",lnc_ref_id)] [lnc_ref_id[!grepl("_unkown",lnc_ref_id)] %in%  rownames(immune.combined)]
complete_set_annotated_lncrnas_in_object <- c(annotated_lncrnas, annotated_lncrna_gene_names_only)
length(unique(complete_set_annotated_lncrnas_in_object))

# Novel lncRNAs
ribodepl_genes <- rownames(immune.combined)[grepl("ribodepl*", rownames(immune.combined))]
polya_genes <- rownames(immune.combined)[grepl("polya*", rownames(immune.combined))]
novel_lncrnas <- c(ribodepl_genes, polya_genes)

# Complete Set 
all_lncrnas <- c(complete_set_annotated_lncrnas_in_object, novel_lncrnas)
all_lncrnas <- all_lncrnas[!grepl("ribodepl-MSTRG.72510-unkown",all_lncrnas)]
```


```{r cars}
get_mast_model  <- function(subset, relevel = "bystander", contarst = "conditioninfected"){
    # Convert to single cell experiment
    subset <- as.SingleCellExperiment(subset)
    # Extract ident column and use it as level
    cond<-factor(colData(subset)$ident)
    cond<-relevel(cond,relevel)
    colData(subset)$cond<-cond
    colData(subset)$wellKey<-colnames(subset)
    # add individual column in case is missing
    
    scam <- as(subset, 'SingleCellAssay')
    
    rowData(scam)$primerid <- rownames(rowData(scam))
   
    cdr2 <-colSums(assay(scam)>0)
     
    colData(scam)$cngeneson <- scale(cdr2)
    
    #We’ll fit a hurdle model, modeling the condition and (centered) ngeneson factor, thus adjusting for the cellular detection rate.
    #In order to have more interpretable coefficients, we’ll set the reference level of the factor to be the “unstimulated” cells.
    cond<-factor(colData(scam)$cond)
    cond<-relevel(cond,relevel)
    colData(scam)$condition<-cond
    
    zlmCond <- zlm(~condition + cngeneson+ freshfrozen, scam)
    # Summarize data
    print("here")
    summaryCond <- summary(zlmCond, doLRT=contarst) 
    summaryDt <- summaryCond$datatable
    fcHurdle <- merge(summaryDt[contrast==contarst & component=='H',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                          summaryDt[contrast==contarst & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
    fcHurdle[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
    return(fcHurdle)
}
```
# Set up differential expression analysis
```{r cars}
immune.combined$freshfrozen <- as.factor(ifelse(immune.combined$batch =="fresh", 0, 1))

myeloids <- subset(immune.combined, idents = "Myeloid"); Idents(myeloids) <- myeloids$group
#nk <- subset(immune.combined, idents = "NK"); Idents(nk) <- nk$group
t <- subset(immune.combined, idents = "T"); Idents(t) <- t$group
b <- subset(immune.combined, idents = "B"); Idents(b) <- b$group


mast_stateVSbaseline_perCelltype <- function(immune.combined , celltype, idents){
  subset <- subset(immune.combined, idents = celltype); Idents(subset) <- subset$group
  subset <- subset(subset, idents = idents)
  model_late_subset <- get_mast_model(subset, relevel = idents[1], contarst = paste0("condition", idents[2]))
  saveRDS(model_late_subset,file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_", idents[2],".rds")))
  
  return(model_late_subset)
}
immune.combined$individual

mast_myeloid_late <- mast_stateVSbaseline_perCelltype(immune.combined = immune.combined, "Myeloid", c("baseline", "late"))
mast_myeloid_middle <- mast_stateVSbaseline_perCelltype(immune.combined, "Myeloid", c("baseline", "middle"))
mast_myeloid_early <- mast_stateVSbaseline_perCelltype(immune.combined, "Myeloid", c("baseline", "early"))

mast_T_late <- mast_stateVSbaseline_perCelltype(immune.combined, "T", c("baseline", "late"))
mast_T_middle <- mast_stateVSbaseline_perCelltype(immune.combined, "T", c("baseline", "middle"))
mast_T_early <- mast_stateVSbaseline_perCelltype(immune.combined, "T", c("baseline", "early"))

mast_B_late <- mast_stateVSbaseline_perCelltype(immune.combined, "B", c("baseline", "late"))
mast_B_middle <- mast_stateVSbaseline_perCelltype(immune.combined, "B", c("baseline", "middle"))
mast_B_early <- mast_stateVSbaseline_perCelltype(immune.combined, "B", c("baseline", "early"))

```

```{r heatmap}
hm <- function(m,features, row_title, first  = FALSE, row_names = FALSE){
  # Only select lncrnas
  m <- m[rownames(m) %in% features,, drop = FALSE]
    rownames(m) <- gsub("-unkown","",rownames(m))


      column_ha <- HeatmapAnnotation(celltype = unlist(lapply( colnames(m), function(x) unlist(str_split(x, "_"))[1])), 
                        comparison = unlist(lapply( colnames(m), function(x) unlist(str_split(x, "_"))[2])),
                        show_annotation_name = FALSE
                        )
  row_ha <- rowAnnotation(genetype =substring(rownames(m), 1,3),
                        show_annotation_name = FALSE)
  if(!first){

    h1 <- Heatmap(m, show_row_dend = FALSE,
          show_column_dend = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          show_row_names = row_names, 
          row_title = row_title,
          row_title_rot = 0, 
          column_split = unlist(lapply( colnames(m), function(x) unlist(str_split(x, "_"))[1])), 
          show_heatmap_legend = FALSE
          )    
  }else{

    h1 <- Heatmap(m, show_row_dend = FALSE,
          show_column_dend = FALSE,
          cluster_columns = FALSE,
          show_column_names = FALSE,
          show_row_names = row_names, 
          row_title = row_title,
          row_title_rot = 0, 
          top_annotation = column_ha, 
          
          column_split = unlist(lapply( colnames(m), function(x) unlist(str_split(x, "_"))[1]))
          )
  }
  print("done")
  return(h1)
}
```

# ---------------------------------
#       LNCRNAS
# ---------------------------------

```{r heatmap}

set.seed(123)

# -------------------------------------------------------------------------
# Gather all information about changes of each gene in each comparison per celltype
# -------------------------------------------------------------------------

# ------- Myeloid
celltype <- "myeloid"
mast_myeloid_late$comparison <- paste0(celltype,"_late")
mast_myeloid_middle$comparison <- paste0(celltype,"_middle")
mast_myeloid_early$comparison <- paste0(celltype,"_early")
myeloids_de <- rbind(mast_myeloid_late, mast_myeloid_middle, mast_myeloid_early)

# ------- T
celltype <- "T"
mast_T_late$comparison <- paste0(celltype,"_late")
mast_T_middle$comparison <- paste0(celltype,"_middle")
mast_T_early$comparison <- paste0(celltype,"_early")
t_de <- rbind(mast_T_late, mast_T_middle, mast_T_early)


# ------- B
celltype <- "B"
mast_B_late$comparison <- paste0(celltype,"_late")
mast_B_middle$comparison <- paste0(celltype,"_middle")
mast_B_early$comparison <- paste0(celltype,"_early")
b_de <- rbind(mast_B_late, mast_B_middle, mast_B_early)

stats_complete <- rbind( myeloids_de, t_de, b_de)

# Convert into matrix
m <-xtabs(coef~primerid+comparison, stats_complete)
attr(m, "class") <- NULL
m <- m[,c("myeloid_early" ,"myeloid_middle","myeloid_late", "B_early" ,"B_middle","B_late", "T_early" ,"T_middle","T_late")]

# -------------------------------------------------------------------------
# Select genes
# -------------------------------------------------------------------------

FDRTHRESHOLD <- 0.1

get_de_names <- function(df_filtered, FDRTHRESHOLD = 0.1, FCTHRESHOLD = log(1), subset = all_lncrnas){
  # Only select lncrnas 
  df_filtered <- df_filtered[df_filtered$primerid %in% subset,]
  # Only select significant ones
  df_filtered <- df_filtered[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
  return(df_filtered$primerid)
}


# Select genes to plot
lnc_myeloid_late <- get_de_names(mast_myeloid_late, subset = all_lncrnas)
lnc_myeloid_middle <- get_de_names(mast_myeloid_middle, subset = all_lncrnas)
lnc_myeloid_early <- get_de_names(mast_myeloid_early, subset = all_lncrnas)

lnc_T_late <- get_de_names(mast_T_late, subset = all_lncrnas)
lnc_T_middle <- get_de_names(mast_T_middle, subset = all_lncrnas)
lnc_T_early <- get_de_names(mast_T_early, subset = all_lncrnas)

lnc_B_late <- get_de_names(mast_B_late, subset = all_lncrnas)
lnc_B_middle <- get_de_names(mast_B_middle, subset = all_lncrnas)
lnc_B_early <- get_de_names(mast_B_early, subset = all_lncrnas)

## mrnas
mrna_myeloid_late <- get_de_names(mast_myeloid_late, subset = annotated_mrnas)
mrna_myeloid_middle <- get_de_names(mast_myeloid_middle, subset = annotated_mrnas)
mrna_myeloid_early <- get_de_names(mast_myeloid_early, subset = annotated_mrnas)

mrna_T_late <- get_de_names(mast_T_late, subset = annotated_mrnas)
mrna_T_middle <- get_de_names(mast_T_middle, subset = annotated_mrnas)
mrna_T_early <- get_de_names(mast_T_early, subset = annotated_mrnas)

mrna_B_late <- get_de_names(mast_B_late, subset = annotated_mrnas)
mrna_B_middle <- get_de_names(mast_B_middle, subset = annotated_mrnas)
mrna_B_early <- get_de_names(mast_B_early, subset = annotated_mrnas)
# ----------------------------


de_lists_lnc <- list(lnc_myeloid_late, lnc_myeloid_middle, lnc_myeloid_early,lnc_T_late, lnc_T_middle, lnc_T_early, lnc_B_late, lnc_B_middle, lnc_B_early )
names(de_lists_lnc) <- c("lnc_myeloid_late", "lnc_myeloid_middle", "lnc_myeloid_early","lnc_T_late", "lnc_T_middle", "lnc_T_early", "lnc_B_late", "lnc_B_middle", "lnc_B_early")


de_lists_mrna <- list(mrna_myeloid_late, mrna_myeloid_middle, mrna_myeloid_early,mrna_T_late, mrna_T_middle, mrna_T_early, mrna_B_late, mrna_B_middle, mrna_B_early )
names(de_lists_mrna) <- c("mrna_myeloid_late", "mrna_myeloid_middle", "mrna_myeloid_early","mrna_T_late", "mrna_T_middle", "mrna_T_early", "mrna_B_late", "mrna_B_middle", "mrna_B_early")


# -------------------------------------------------------------------------
# Create Heatmap(s)
# -------------------------------------------------------------------------
de_lists_lnc <- de_lists_lnc[unlist(lapply(1:length(de_lists_lnc), function(index) length(de_lists_lnc[[index]]) != 0))]
h1_lnc<- hm(m= m,row_title = names(de_lists_lnc)[1], features  = de_lists_lnc[1][[1]], first = TRUE, row_names = TRUE)
hms <- lapply(2:length(de_lists_lnc), function(index) hm(m,row_title = names(de_lists_lnc)[index], features  = de_lists_lnc[index][[1]], row_names = TRUE))

length(hms)
heatmap_lnc <- h1_lnc %v% hms[[1]]%v%hms[[2]]%v%hms[[3]]%v%hms[[4]]%v%hms[[5]]
heatmap_lnc

# mrna ---------
de_lists_mrna <- de_lists_mrna[unlist(lapply(1:length(de_lists_mrna), function(index) length(de_lists_mrna[[index]]) != 0))]
h1 <- hm(m= m,row_title = names(de_lists_mrna)[1], features  = de_lists_mrna[1][[1]], first = TRUE, row_names = FALSE)
hms_mrna <- lapply(2:length(de_lists_mrna), function(index) hm(m,row_title = names(de_lists_mrna)[index], features  = de_lists_mrna[index][[1]], row_names = FALSE))

length(de_lists_mrna)

heatmap_mrna <- h1 %v% hms_mrna[[1]]%v%hms_mrna[[2]]%v%hms_mrna[[3]]%v%hms_mrna[[4]]%v%hms_mrna[[5]]%v%hms_mrna[[6]]%v%hms_mrna[[7]]%v%hms_mrna[[8]]
heatmap_mrna

do_go(unlist(de_lists_mrna))
```

```{r a}

# Only focus on Myeloid in lnc 

lnc_myeloid_late <- get_de_names(mast_myeloid_late, subset = all_lncrnas, FDRTHRESHOLD = 0.1, FCTHRESHOLD = log(1.2))
lnc_myeloid_middle <- get_de_names(mast_myeloid_middle, subset = all_lncrnas, FDRTHRESHOLD = 0.1, FCTHRESHOLD = log(1.2))
lnc_myeloid_early <- get_de_names(mast_myeloid_early, subset = all_lncrnas, FDRTHRESHOLD = 0.1, FCTHRESHOLD = log(1.2))

de_lists_lnc_my <- list(lnc_myeloid_late, lnc_myeloid_middle, lnc_myeloid_early )
names(de_lists_lnc_my) <- c("lnc_myeloid_late", "lnc_myeloid_middle", "lnc_myeloid_early")

de_lists_lnc_my <- de_lists_lnc_my[unlist(lapply(1:length(de_lists_lnc_my), function(index) length(de_lists_lnc_my[[index]]) != 0))]

# Heamap with all celltypes 
h1_lnc_my <- hm(m= m,row_title = names(de_lists_lnc_my)[1], features  = de_lists_lnc_my[1][[1]], first = TRUE, row_names = TRUE)
hms_lnc_my <- lapply(2:length(de_lists_lnc_my), function(index) hm(m,row_title = names(de_lists_lnc_my)[index], features  = de_lists_lnc_my[index][[1]], row_names = TRUE))

# Heamap with Myeloid only 
m_myeloid <- m[,c("myeloid_early", "myeloid_middle","myeloid_late")]
h1_lnc_my <- hm(m= m_myeloid,row_title = names(de_lists_lnc_my)[1], features  = de_lists_lnc_my[1][[1]], first = TRUE, row_names = TRUE)
hms_lnc_my <- lapply(2:length(de_lists_lnc_my), function(index) hm(m = m_myeloid,row_title = names(de_lists_lnc_my)[index], features  = de_lists_lnc_my[index][[1]], row_names = TRUE))

h1_lnc_my %v% hms_lnc_my[[1]]

# Visualize in expression heatmap 
# late only 
de_lnc_myelid_late_ordered <- de_lists_lnc_my$lnc_myeloid_late[order(m_myeloid[de_lists_lnc_my$lnc_myeloid_late,"myeloid_late" ])]


baseline_late <- subset(myeloids, idents = c("baseline", "late"))

DoHeatmap(myeloids, features = de_lnc_myelid_late_ordered, group.bar = TRUE)
DoHeatmap(baseline_late, features = de_lnc_myelid_late_ordered, group.bar = TRUE) 
#+     scale_fill_gradientn(colors = rev(RColorBrewer::brewer.pal(n = 5, name = "RdBu")))

DoHeatmap(myeloids, features = de_lists_lnc_my$lnc_myeloid_early, group.bar = TRUE)
```


# -------------------------------------------
# -------------------------------------------
# Analysis colocation in genomic regions

```{r colocation}
# 1) For each gene lncrna find neighbour protein coding genes. ( either overlapping or in close region) --> putative pairs
# 2) Filter: Check the pearson correlation of the identified pairs, reatain the ones w/ |pearson correlation coefficient| > 0.6
# ==> Only if something nice comes out 
# 3) Validation: Simulate random pairs 
# 4) Validation: Compare the correlation of the ones identified with the random ones. 
library(zeallot)

# 1. extract the count matrix
count_matrix <- log2(myeloids@assays$RNA@counts+.5)
count_matrix <- myeloids@assays$RNA@counts
rownames(count_matrix) <- gsub("-unkown","",rownames(count_matrix))
exp_matrix <-myeloids@assays$integrated@data

# 2. 
mrnas_ranges <- ref[ref$gene_name %in% mrna_ref_id  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unkown", "",all_lncrnas))]

# Add missing gene entries 
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(c,gr_list)

lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)



# --------------------------------------------------------------
#           Get 5 closest genes to each gene of interest
# --------------------------------------------------------------
# Find pairs under maximum overlap
get_n_closest_gr <- function(gr_gene, gr_hits =mrnas_ranges , n = 5, maxgap =1000000L ){
  p <- findOverlapPairs(gr_gene, get_gene_only(gr_hits), maxgap=maxgap, ignore.strand = TRUE)
  pairs_list <- zipup(p)
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  distances <- unlist(lapply(pairs_list, function(pair) GenomicRanges::distance(pair[1], pair[2], ignore.strand = TRUE) ))
  if(!is.null(distances)){
    n <- min(length(pairs_list), n)
    closest_5_pairs <-  pairs_list[order(distances)][1:n]
  }else{
    return(NULL)
  }
  return(closest_5_pairs)
}

#lnc_myeloid_late_annot<-lnc_myeloid_late[(unlist(lapply(lnc_myeloid_late, function(x) substr(x,1,3))) == "rib")]

lnc_myeloid_late_annot<-lnc_myeloid_late
# Only select the DE interestin
lnc_ranges_complete_de <- lnc_ranges_complete[lnc_ranges_complete$gene_id %in% gsub("-", "_",gsub("-unkown", "",lnc_myeloid_late_annot)), ]


ranges_5_cl_genes <- lapply(1:(length(lnc_ranges_complete_de)), function(index) unlist(get_n_closest_gr(lnc_ranges_complete_de[index], mrnas_ranges, n = 15, maxgap = 1000000L))[c(rep(FALSE,1), TRUE)]$gene_name)
ranges_5_cl_genes


mask_ensembl <- unlist(lapply(lnc_ranges_complete_de$gene_id, function(x) substr(x,1,3))) == "ENS"
table(mask_ensembl)
# GO En
do_go(unlist(ranges_5_cl_genes))
length(unique(ranges_5_cl_genes))
 

# ----------------------
# Find pairs allowing 10K gap on both sides
pairs <- findOverlapPairs(lnc_ranges_complete_de, get_gene_only(mrnas_ranges), maxgap=1000000L, ignore.strand = TRUE)
# Get a list of ranges objects in pairs
pairs_list <- zipup(pairs)
unlist(pairs_list)
do_go(unlist(pairs_list)[c(rep(FALSE,1), TRUE)]$gene_name)
do_go(unlist(pairs_list)[c(rep(FALSE,1), TRUE)]$gene_name, mmu = TRUE)



```


# -------------------------------------------
# -------------------------------------------
# Analysis cis-action/coregulation 

```{r cis-regulation}
# 1) For each gene lncrna find neighbour protein coding genes. ( either overlapping or in close region) --> putative pairs
# 2) Filter: Check the pearson correlation of the identified pairs, reatain the ones w/ |pearson correlation coefficient| > 0.6
# ==> Only if something nice comes out 
# 3) Validation: Simulate random pairs 
# 4) Validation: Compare the correlation of the ones identified with the random ones. 
library(zeallot)

# 1. extract the count matrix
count_matrix <- log2(myeloids@assays$RNA@counts+.5)
count_matrix <- myeloids@assays$RNA@counts
rownames(count_matrix) <- gsub("-unkown","",rownames(count_matrix))
exp_matrix <-myeloids@assays$integrated@data

# 2. 
mrnas_ranges <- ref[ref$gene_name %in% mrna_ref_id  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unkown", "",all_lncrnas))]

# Add missing gene entries 
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(c,gr_list)

lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)

# Only select the DE interestin

lnc_ranges_complete_de <- lnc_ranges_complete[lnc_ranges_complete$gene_id %in% gsub("-", "_",gsub("-unkown", "",de_lnc_myelid_late_ordered)), ]

# ----------------------
# Find pairs allowing 10K gap on both sides

pairs <- findOverlapPairs(lnc_ranges_complete_de, get_gene_only(mrnas_ranges), maxgap=10000L, ignore.strand = TRUE)
# Get a list of ranges objects in pairs
pairs_list <- zipup(pairs)
length(pairs_list)

# How many lncrnas in total vs how many are in pairs
length(unique(lnc_ranges$gene_id)); length(unique(mrnas_ranges$gene_id)); length(pairs)

# get the counts of the interesting genes
calc_correlation <- function(pair, count_matrix, type = "pearson"){
  gene1 <- str_replace_all(pair[1]$gene_id, "_" , "-")
  gene2 <- str_replace_all(pair[2]$gene_id, "_" , "-")
  print(gene1)
  print(gene2)
  c1 <- count_matrix[rownames(count_matrix)== gene1,]
  c2 <- count_matrix[rownames(count_matrix)== gene2,]
  print("----")
  # calculate the correlation coefficient 
  if(length(c1)> 0 && length(c2)>0){ 
    #pc <- bayes.cor.test(c1, c2) 
    pc <- cor.test(c1, c2, method = c(type)) 
  }else{ pc = cor.test(c(1:100),c(1:100), method = c(type))  }
  return(pc)
}
# Calc correlation values

correlations <- lapply(pairs_list, function(x) calc_correlation(x, count_matrix, type = "spearman"))
pvals <- (lapply(correlations, function(correlation) print(correlation$p.value)))
rhos <- as.vector(lapply(correlations, function(correlation) print(correlation$estimate)))

# Obtain significant correlations
significant_correlations <- correlations[pvals < 0.1 & rhos !=1]
significant_pairs <- pairs_list[pvals < 0.1 & rhos !=1]
significant_correlations

# Extract mrnas from significant correlations 
do_go <- function(list, keytype = "SYMBOL"){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =  org.Mmu.eg.db, keyType = keytype,ont = "BP",universe =unique(ref$gene_name))
  barplot(ego, showCategory=10)
  clusterProfiler::dotplot(ego)
}
correlated_mrnas <-unlist(significant_pairs)[c(rep(FALSE,1), TRUE)]$gene_id
do_go(correlated_mrnas, keytype = "ENSEMBL")

keytypes(org.Mmu.eg.db)
```