---
title: "DE_bulk"
author: "Luisa Santus"
date: "12/2/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
library(sva); library(edgeR)
library("genefilter")
library("gplots")
library(plyr)

# Import Utils 
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/","code/ebola/src/scripts/Analysis/utils")
source(file.path(scripts_dir,"de_utils.R")); source(file.path(scripts_dir,"de_utils.R"))

datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
novel_lnc <- import(file.path(datadir,"/02_RNA-Seq_rheMac10/07_liftover/novel_rhemac10_expressed_intergenic.gtf"))

dir_hisat_all =  file.path(datadir, "02_RNA-Seq_rheMac10/06_counts_all/")
htseq_files <- iterate_files(dir_hisat_all, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

dds <- create_dds(htseq_files,dir_hisat_all)

# Retain only blood and liver
dds <- dds[, dds$tissue == "Blood" | dds$tissue == "Liver"]
# remove bad sample
#dds <- dds[,!(dds$id == "RA0223")]
#dds <- dds[,!(dds$id == "RA1834")]
dds <- dds[, dds$condition == "Batch01"] 
dds <- dds[,(dds$dpo != "D03")]
dds <- dds[,(dds$dpo != "D04")]
# Create DGE
dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))

#all_lnc <- c(novel_lnc$gene_id,  lncRNAs_ref$gene_id) 
#dds_lnc <- dds[rownames(dds) %in% all_lnc,]

dge <- calcNormFactors(dge)
#assays(dds)$counts
assays(dds)$logCPM <- edgeR::cpm(dge, log = TRUE,prior.count = 0.5)
mask <- rowMeans(assays(dds)$logCPM) > 5
dds <- dds[mask, ]
dge <- dge[mask, ]







# Library sizes
# --------------------------------------------------------------------------------------
sorted_dge <- dge[,order(dge$samples$lib.size)]
colors <- brewer.pal(n = length(unique(sorted_dge$samples$tissue)), name = 'Paired')
col.cell <- colors[sorted_dge$samples$tissue]
barplot((sorted_dge$samples$lib.size)/1e+06, ylab = "Library sizes (Millions)", names.arg=(sorted_dge$samples$complete_id), las = 2, cex.names = 0.4, col = col.cell,legend = unique(sorted_dge$samples$tissue), args.legend = list(x = "topleft"))
# --------------------------------------------------------------------------------------------

# read metadata 
projdir = "/home/luisas/Desktop/cluster/data/"
metadata <- read.csv(file.path(projdir, "00_Metadata/batch_metadata.csv"), header = FALSE)
colnames(metadata) <- c("id", "gender", "weight", "birthday", "Type")

# Gender 
# Age
# Individual 


# Add metadata 
dds$infection <-ifelse(as.double(sub("D", "", dds$dpo))>0, "infected", "not-infected")
dds$gender <- metadata$gender[match(dds$id, metadata$id)]
dds$birthyear <- as.double(str_sub(metadata$birthday, -4))[match(dds$id,metadata$id)]
dds$weight <- as.double(metadata$weight)[match(dds$id,metadata$id)]


# Mixed model!!! sample id!!!
#mod <- model.matrix(~0+factor(dpo) + factor(tissue) + factor(age) + factor(gender), colData(dds))
mod <- model.matrix(~0 + factor(dpo)  , colData(dds))



v <- voom(dge, mod, plot=TRUE)
fit5 <- lmFit(v, mod)

FDRcutoff=0.01
fit5 <- eBayes(fit5)
res5 <- decideTests(fit5, p.value = FDRcutoff)
tt5 <- topTable(fit5, coef = 2, n = Inf)

DEgenes5 <- rownames(tt5)[tt5$adj.P.Val < FDRcutoff]
length(DEgenes5)

top300 <- head(tt5[order(tt5$adj.P.Val),], n=300)
```

```{r}
dds$dpo <- as.numeric(str_remove(dds$dpo , "D"))
colnames(dds) <- dds$dpo 
matrixDE <- assay(dds)[ DEgenes5, ][, order(as.integer(colnames(assay(dds)[ DEgenes5, ])))]
DEgenes5 <- top300$genes
matrixDE <- assay(dds)[ DEgenes5, ][, order(as.integer(colnames(assay(dds)[ DEgenes5, ])))]
print(matrixDE)
?heatmap.2
heatmap.2( matrixDE, scale="row", cexCol = 1.1,cexRow = 0.9,
     trace="none", dendrogram="row",Colv=colnames(matrixDE), 
     col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255))
```
..
```{r cars}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt5$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit5$t[, 2], df = fit5$df.prior + fit5$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```



First we need to get the counts into a readable format.
```{r echo=FALSE,  results='hide'}
directory_zyagen = file.path(datadir,"02_RNA-Seq_rheMac10/")
count_matrix<- iterate_files(directory_zyagen, 'gene_count_matrix.csv')

# Load gene(/transcript) count matrix and labels
countData <- as.matrix(read.csv(count_matrix, row.names="gene_id"))
colData <- get_colData(countData)
all(rownames(colData) %in% colnames(countData))
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)




dds <- dds[rownames(dds) %in% novel_lncrnas,]
dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))


dge <- calcNormFactors(dge)
assays(dds)$counts
assays(dds)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(dds)$logCPM) > 10
dds <- dds[mask, ]
dge <- dge[mask, ]


# Library sizes
sorted_dge <- dge[,order(dge$samples$lib.size)]
colors <- brewer.pal(n = length(levels(sorted_dge$samples$tissue)), name = 'Paired')
col.cell <- colors[sorted_dge$samples$tissue]
barplot((sorted_dge$samples$lib.size)/1e+06, ylab = "Library sizes (Millions)", names.arg=(sorted_dge$samples$complete_id), las = 2, cex.names = 0.4, col = col.cell,legend = unique(sorted_dge$samples$tissue), args.legend = list(x = "topleft"))
colData(dds)$dpo

mod <- model.matrix(~0+factor(dpo) + factor(tissue), colData(dds))
v <- voom(dge, mod, plot=TRUE)
fit5 <- lmFit(v, mod)
FDRcutoff=0.01
fit5 <- eBayes(fit5)
res5 <- decideTests(fit5, p.value = FDRcutoff)
tt5 <- topTable(fit5, coef = 2, n = Inf)
DEgenes5 <- rownames(tt5)[tt5$adj.P.Val < FDRcutoff]

dds$dpo <- as.numeric(str_remove(dds$dpo , "D"))
colnames(dds) <- dds$dpo 
```


## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
