---
title: "DE_bulk"
author: "Luisa Santus"
date: "12/2/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
library(sva); library(edgeR)
library("genefilter")
library("gplots")

dir_hisat_all =  file.path(datadir, "02_RNA-Seq_rheMac10/06_counts_all/")
htseq_files <- iterate_files(dir_hisat_all, "UMI.f3.q60.umi_dedup.HTseq.gene_counts.tab")

dds <- create_dds(htseq_files,dir_hisat_all)

sampleFiles <- htseq_files
directory <- 
  samples_info <- sapply(sampleFiles,extract_info_sample_from_filename)
  # Create ddsHTSeq Object 
  sampleTable <- data.frame(sampleName = sampleFiles,
                            fileName = sapply(sampleFiles,get_filename_and_dirs),
                            condition = samples_info[1,],
                            tissue = samples_info[2,], 
                            dpo = samples_info[3,], 
                            id = samples_info[4,],
                            complete_id = samples_info[5,]
  )
  dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                    directory = directory,
                                    design= ~ 1)


dds <- dds[rownames(dds) %in% lncRNAs_ref$gene_id,]
dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))


dge <- calcNormFactors(dge)
assays(dds)$counts
assays(dds)$logCPM <- edgeR::cpm(dge, log = TRUE,prior.count = 0.5)
mask <- rowMeans(assays(dds)$logCPM) > 10
dds <- dds[mask, ]
dge <- dge[mask, ]



# Library sizes
# --------------------------------------------------------------------------------------
sorted_dge <- dge[,order(dge$samples$lib.size)]
colors <- brewer.pal(n = length(unique(sorted_dge$samples$tissue)), name = 'Paired')
col.cell <- colors[sorted_dge$samples$tissue]
barplot((sorted_dge$samples$lib.size)/1e+06, ylab = "Library sizes (Millions)", names.arg=(sorted_dge$samples$complete_id), las = 2, cex.names = 0.4, col = col.cell,legend = unique(sorted_dge$samples$tissue), args.legend = list(x = "topleft"))
# --------------------------------------------------------------------------------------------

# read metadata 
projdir = "/home/luisas/Desktop/cluster/data/"
metadata <- read.csv(file.path(projdir, "00_Metadata/batch_metadata.csv"), header = FALSE)
metadata

# Gender 
# Age
# Individual 


# Mixed model!!! sample id!!!
mod <- model.matrix(~0+factor(dpo) + factor(tissue) + factor(age) + factor(gender), colData(dds))


v <- voom(dge, mod, plot=TRUE)
fit5 <- lmFit(v, mod)

FDRcutoff=0.01
fit5 <- eBayes(fit5)
res5 <- decideTests(fit5, p.value = FDRcutoff)
tt5 <- topTable(fit5, coef = 2, n = Inf)

DEgenes5 <- rownames(tt5)[tt5$adj.P.Val < FDRcutoff]
length(DEgenes5)
```

```{r}
dds$dpo <- as.numeric(str_remove(dds$dpo , "D"))
colnames(dds) <- dds$dpo 
matrixDE <- assay(dds)[ DEgenes5, ][, order(as.integer(colnames(assay(dds)[ DEgenes5, ])))]
print(matrixDE)
heatmap.2( matrixDE, scale="row", cexCol = 1.1,cexRow = 0.9,
     trace="none", dendrogram="row",Colv=colnames(matrixDE), 
     col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255))
```
..
```{r cars}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt5$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit5$t[, 2], df = fit5$df.prior + fit5$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```



First we need to get the counts into a readable format.
```{r echo=FALSE,  results='hide'}
directory_zyagen = file.path(datadir,"02_RNA-Seq_rheMac10/")
count_matrix<- iterate_files(directory_zyagen, 'gene_count_matrix.csv')

# Load gene(/transcript) count matrix and labels
countData <- as.matrix(read.csv(count_matrix, row.names="gene_id"))
colData <- get_colData(countData)
all(rownames(colData) %in% colnames(countData))
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~tissue)




dds <- dds[rownames(dds) %in% novel_lncrnas,]
dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))


dge <- calcNormFactors(dge)
assays(dds)$counts
assays(dds)$logCPM <- cpm(dge, log = TRUE)
mask <- rowMeans(assays(dds)$logCPM) > 10
dds <- dds[mask, ]
dge <- dge[mask, ]


# Library sizes
sorted_dge <- dge[,order(dge$samples$lib.size)]
colors <- brewer.pal(n = length(levels(sorted_dge$samples$tissue)), name = 'Paired')
col.cell <- colors[sorted_dge$samples$tissue]
barplot((sorted_dge$samples$lib.size)/1e+06, ylab = "Library sizes (Millions)", names.arg=(sorted_dge$samples$complete_id), las = 2, cex.names = 0.4, col = col.cell,legend = unique(sorted_dge$samples$tissue), args.legend = list(x = "topleft"))
colData(dds)$dpo

mod <- model.matrix(~0+factor(dpo) + factor(tissue), colData(dds))
v <- voom(dge, mod, plot=TRUE)
fit5 <- lmFit(v, mod)
FDRcutoff=0.01
fit5 <- eBayes(fit5)
res5 <- decideTests(fit5, p.value = FDRcutoff)
tt5 <- topTable(fit5, coef = 2, n = Inf)
DEgenes5 <- rownames(tt5)[tt5$adj.P.Val < FDRcutoff]

dds$dpo <- as.numeric(str_remove(dds$dpo , "D"))
colnames(dds) <- dds$dpo 
```


## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
