---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(ggthemes)
library(wesanderson)

source("../../utils/02_sc_utils.R")

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/")

# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready_neut.rds")
orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneid_ready.rds")

# -------------------------------------------
#    define threshold for the analysis
# -------------------------------------------

# Corrected q-value .05
FDRTHRESHOLD<- 0.05
# Fold-Change of 30%
FCTHRESHOLD <- log(1.23)
```

# --------------------------------------------------
# Identify infected Monocytes
# --------------------------------------------------

```{r readfilesDE}
# Define Ebola genes 
ebola_genes <- unique(ebola_ref$gene_id)

# Extract normalized expression values matrix
expression_matrix <- immune.combined@assays$RNA@counts
colnames(expression_matrix) <- Idents(immune.combined)


# 1. Per cell, check the % of ebola reads ( not normalized - just calculate the % of raw counts )
# 1a. Calculate number of ebola reads 
ebola_reads <- expression_matrix[gsub("-unknown","", rownames(expression_matrix)) %in% ebola_genes, ]
ebola_genome_reads <- colSums(ebola_reads)
# 1b. Calculate number of host reads
total_reads <- colSums(expression_matrix)
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- (ebola_genome_reads/total_reads)*100

# 1c. Store the % of ebola reads per cell in a dataframe
ebola_genome_percentage_df <- data.frame(percentage_viral_reads = ebola_genome_percentage,
                                         celltype = names(ebola_genome_percentage),
                                         cond =immune.combined$cond, 
                                         hour = immune.combined$dpi)


# 2. Define the threshold for the percentage of viral reads identified per cell for it to be called Infected 
unique(Idents(immune.combined))
# 2a. Per cell-type, calculate which is the threshold for which we get a maximum of 1% false positives
#threshold <- quantile(ebola_genome_percentage_df[!(ebola_genome_percentage_df$celltype %in% c("Monocyte")), ]$percentage_viral_reads, probs = c(0.99))
threshold <- max(unlist(lapply(setdiff(unique(ebola_genome_percentage_df$celltype), c("Monocyte")), function(x) quantile(ebola_genome_percentage_df[ebola_genome_percentage_df$celltype == x, ]$percentage_viral_reads, probs = c(0.99)))))

print(paste("Threshold", threshold))
#ebola_genome_percentage_df[ebola_genome_percentage_df$percentage_viral_reads > 0,]
```



```{r colocation}

# Separate the cells into infected and not infected based on the threshold
ebola_genome_percentage_df$classification <- ifelse(ebola_genome_percentage_df$percentage_viral_reads >= threshold, "Infected", "Not Infected")
immune.combined$infection <- ebola_genome_percentage_df$classification

table(immune.combined[,Idents(immune.combined) == "Monocyte" ]$infection)
table(immune.combined[,Idents(immune.combined) == "B" ]$infection)
table(immune.combined[,Idents(immune.combined) == "T" ]$infection)
table(immune.combined[,Idents(immune.combined) == "Neutrophil" ]$infection)

infected_cells <- rownames(ebola_genome_percentage_df[ebola_genome_percentage_df$classification == "Infected",])

# Plot infected cells 
DimPlot(immune.combined, label=F, cells.highlight= list(infected_cells), sizes.highlight = 0.2, cols.highlight = c("#DA0202"), cols= "grey", pt.size = 0.1)+ 
    labs(title = "Infected cells" )+ NoLegend()+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+
    theme(legend.position = "none", plot.title = element_text(size = 22), axis.title = element_text(size = 18))

# Or with labels (Same plot just with cell-type labels)
DimPlot(immune.combined, label=T, cells.highlight= list(infected_cells), sizes.highlight = 0.2, cols.highlight = c("#DA0202"), cols= "grey", pt.size = 0.1)+ 
    labs(title = "Infected cells" )+ NoLegend()+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+
    theme(legend.position = "none", plot.title = element_text(size = 22), axis.title = element_text(size = 18))
```


```{r colocation}
# Add information about how many cells are found per celltype and, per celltype, which is the proportion of cells infected
# How many cells are found per celltype

ebola_genome_percentage_df$hour <- as.character(ebola_genome_percentage_df$hour)
ebola_genome_percentage_df[as.character(ebola_genome_percentage_df$hour) %in% c("D007", "D008"),]$hour <- "D007&D008"

saveRDS(ebola_genome_percentage_df, "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds")

summary_celltype <- ebola_genome_percentage_df%>% dplyr::group_by(celltype) %>% tally()
summary_celltype <- ebola_genome_percentage_df%>% dplyr::group_by(celltype,hour) %>% tally()
summary_celltype$celltypedpi <-paste(summary_celltype$celltype, summary_celltype$hour, sep = "-")


infected_summary <- ebola_genome_percentage_df[as.character(ebola_genome_percentage_df$classification) == "Infected",]%>% dplyr::group_by(celltype,hour) %>% tally()
infected_summary$celltypedpi <- paste(infected_summary$celltype, infected_summary$hour, sep = "-")
infected_summary <- infected_summary[,c("celltypedpi", "n")]
names(infected_summary) <- c("celltypedpi", "infected")

percentage_cells_infected_per_celltype <- merge(infected_summary, summary_celltype, by = "celltypedpi")
percentage_cells_infected_per_celltype$perc_infected <- (percentage_cells_infected_per_celltype$infected / percentage_cells_infected_per_celltype$n)*100

# Visualize 
pal <- wes_palette("Zissou1", 5, type = "discrete")
ggplot(percentage_cells_infected_per_celltype, aes( x= celltype, y = perc_infected , fill = hour))+geom_bar(stat="identity",position = position_dodge(), alpha = 0.8)+theme_paper+xlab("")+ylab("% of cells infected")+theme(axis.text.y = element_text(size = 8))+theme(axis.text.x = element_text(angle=45, hjust=1), axis.title.y = element_text(size = 15), axis.text.y = element_text(size =15))+geom_hline(yintercept = 1.1, lty = 2)+scale_fill_manual(values=pal)

```

```{r colocation}

# Just visualize the threshold for calling a cell infected
ggplot(ebola_genome_percentage_df[ebola_genome_percentage_df$celltype == "Monocyte",], aes( x = percentage_viral_reads , col = classification, fill = classification))+geom_density( alpha = 0.6)+theme_paper+xlab("")+theme(axis.text.y = element_text(size = 8))+theme(axis.text.x = element_text(angle=45, hjust=1), axis.title.y = element_text(size = 15), axis.text.y = element_text(size =15))+scale_x_log10()+scale_fill_manual(values = c( "red", "grey"))+scale_color_manual(values = c( "red", "grey"))+xlab("Viral Load")+geom_vline(xintercept = threshold, lty =  2)


```



# -----------------------------
#  0. CALC CORRELATION ON CLUSTER
#-----------------------------

# ----------------------------------------------------------------
#        DE Analysis results: Bystander vs Infected  
# ----------------------------------------------------------------

```{r cars}
# General expression of gene
source("../../utils/04_utils_graph.R")
mono <-monocyte <-  subset(immune.combined, ident="Monocyte")

mast_mono <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE_InfectionStatus/Monocyte_MAST_infectedVSnotinfected_LIVE24_DPI58.rds")

de_mono <- mast_mono[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
de_mono <- de_mono[!(de_mono$primerid == "MSTRG.182795-unknown"),]
de_mono$direction <- ifelse(de_mono$coef < 0 , "down", "up")
de_mono$direction <- factor(de_mono$direction, c("down", "up"))
de_mono_lnc <- de_mono[de_mono$primerid %in% all_lncrnas, ]
de_mono_pc <- de_mono[de_mono$primerid %in% annotated_mrnas, ]

# Check wich lncRNAs are DE
print(de_mono_lnc)

# 1. Check enrichment for DE pc genes
# UP 
ego_pc_up <- clusterProfiler::enrichGO(gene = de_mono_pc[de_mono_pc$coef > 0 , ]$primerid,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
clusterProfiler::dotplot(ego_pc_up)
# DOWN 
ego_de_down <- clusterProfiler::enrichGO(gene = de_mono_pc[de_mono_pc$coef < 0 , ]$primerid,OrgDb =  org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP",universe =unique(gsub("-unknown", "", rownames(immune.combined))))
clusterProfiler::dotplot(ego_de_down)

de_stats_direction_plot <- de_mono %>% group_by(primerid) %>% dplyr::summarise(direction = toString((direction)))
ggplot(de_stats_direction_plot, aes(x=direction, fill = direction))+geom_bar(stat="count")+theme_paper+scale_fill_manual(values = c( "#C0E4C6", "#FF6666"))+ylab("Number of DE genes")+xlab("")+theme(legend.position = "")

# 2. Check wether we can also identify some that Broad people can, and compare directionality 
down_broad <- c("S100A9", "SELL", "MX1", "VCAN", "MS4A6A")
up_broad <- c("FTL", "C3", "CTSC")
de_mono[de_mono$primerid %in% down_broad,]
de_mono[de_mono$primerid %in% up_broad,]

# Explore DE lnc
string <- de_mono_lnc$primerid
orthologs[orthologs$lnc %in% unique(ref$transcript_id[grepl(string, ref$gene_id)]), ]

get_expression_summary_gene <- function(gene, mono = monocyte, ebola_genome_percentage_df_ = ebola_genome_percentage_df){
  expression_gene <- mono[rownames(mono)== gene, ]@assays$RNA@data
  df <- reshape2::melt(as.matrix(expression_gene))
  names(df) <- c("gene", "cell", "value")
  ebola_genome_percentage_df_$cell <- rownames(ebola_genome_percentage_df_)
  df <- merge(ebola_genome_percentage_df_, df, by="cell")
  df <- df[df$value !=0,]
  return(df)
}


boxplot_bystanders <- function(df){
  # Boxplot bystander
  df$class <- NA
  df[df$cond == "D000", ]$class <- "Baseline" 
  df[df$cond %in% c("D006", "D007", "D008") & df$classification == "Not Infected",]$class <- "Bystanders"
  df[df$cond %in% c("D006", "D007", "D008") & df$classification == "Infected",]$class <- "Infected"
  p <- ggplot(df[!is.na(df$class),], aes(x = class, y = value, fill = class))+geom_boxplot(notch = F, alpha=0.8, add = "jitter")+theme_paper+scale_fill_manual(values = wes_palette("Zissou1", 5, type = "discrete")[c(1,4,5)])+ggtitle(unique(df$gene))+geom_jitter(width=0.1)
  return(p)
}

# Visualize boxplot
gene <- de_mono_lnc$primerid
gene <- "MX1"
df <- get_expression_summary_gene(gene)
lapply(de_mono_lnc$primerid, function(gene) boxplot_bystanders(get_expression_summary_gene(gene)))

dp_data <- Dotplot_data(mono, features = de_mono_lnc$primerid, group.by = "group_inf", dot.scale =10,scale.by = "radius", cols = c( "lightblue", "darkred"))
ggplot(dp_data, aes(x  = avg.exp, y = pct.exp, col = id))+geom_point()+theme_paper+facet_grid(rows=vars(features.plot))
DotPlot(mono, features = c( de_mono_lnc$primerid), group.by = "group",scale.min = 0, scale.max = 100,  dot.scale =10,scale.by = "radius", cols = c( "lightblue", "darkred"))
```



# Visualize 
```{r cars}
FDRTHRESHOLD<- 0.05
RHOTHRESHOLD <- 0
correlations_infected_24 <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10//05_RObjects/06_correlation/03_viralload_infected_late/ViralLoad_spearman-correlations_standard.rds")
correlations_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval),]
sig_correlations_pc_infected_24 <- correlations_infected_24[!is.na(correlations_infected_24$pval) & correlations_infected_24$pval < FDRTHRESHOLD & abs(correlations_infected_24$rho) > RHOTHRESHOLD, ]

# Add Gene Type
sig_correlations_pc_infected_24$type <- NA
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% all_lncrnas ,]$type <- "lnc"
sig_correlations_pc_infected_24[sig_correlations_pc_infected_24$gene %in% annotated_mrnas ,]$type <- "pc"


table(sig_correlations_pc_infected_24$type)
correlations_infected_24[correlations_infected_24$gene %in% all_lncrnas ,]
as.data.frame(res)
```

# Which are the genes that present the strongest correlation ( protein coding - check if concordant with broad )
# Check which are the lnc which present the strongest correlation 



# PLot the expression of a gene in infected vs not infected cells

```{r colocation}
library(reshape2)

gene <- "NRAV"
gas5_genes <- unique(ref[!is.na(ref$transcript_id) & ref$transcript_id %in% unique(orthologs[orthologs$orthologGeneSymbol == "GAS5",]$lnc),]$gene_name)
gas5_gene <- rownames(subset(immune.combined, features = c(gas5_genes)))

gene <- gas5_gene
gene <- "MX1"

immune.combined.late <- immune.combined[,immune.combined$group == "late"]
infected_cells_labels <- rownames(ebola_genome_percentage_df[ebola_genome_percentage_df$classification == "Infected" & ebola_genome_percentage_df$celltype == "Monocyte",])
infected_cells_late <- subset(immune.combined.late, cells = infected_cells_labels)
infected_cells_labels <- colnames(infected_cells_late)

rownames(immune.combined)[grepl("GP", rownames(immune.combined))]

gene_expression <- subset(immune.combined.late, features = c(gene))@assays$RNA@data
colnames(gene_expression)<- ifelse(colnames(gene_expression) %in% infected_cells_labels, "infected", "other")
ge_df <- melt(as.matrix(gene_expression))
names(ge_df) <- c("gene", "status", "value")

pal <- wes_palette("Darjeeling1", 4, type = "discrete")
ggplot(ge_df, aes(x = status, y = value, col = status , fill = status))+geom_boxplot(alpha = 0.4)+geom_jitter()+theme_paper+scale_fill_manual(values = pal)+scale_color_manual(values = pal)+ggtitle(gene)


mono <- subset(immune.combined, idents = "Monocyte")
mono$group <- factor(mono$group, levels = c("baseline", "early", "middle", "late"))
DotPlot(mono, features = c(gene, neat, "SFPQ", "NP-unknown", string), group.by = "group", dot.scale =10,scale.by = "radius", cols = c( "lightblue", "darkred"))+coord_flip()+theme_paper+xlab("")+ylab("")+theme_bw()+ggtitle("Expression in Monocytes")

neat <- "ENSMMUG00000058532-unknown"
plot_seperate_features(mono, gene,ident = "group" )

plot_seperate_features


string <- "ENSMMUG00000058644"
orthologs[orthologs$lnc %in% unique(ref$transcript_id[grepl(string, ref$gene_id)]), ]

unique(orthologs$orthologGeneSymbol[grepl(string, orthologs$orthologGeneSymbol)])


ref[grepl("IFI6", ref$gene_name)]$gene_biotype
inspect_gene_expression(immune.combined, mono, gene)

expression_gene <- monocyte[rownames(monocyte)== gene, ]@assays$RNA@data
colnames(expression_gene) <- monocyte$group
df <- reshape2::melt(as.matrix(expression_gene))
names(df) <- c("gene", "group", "value")
df <- df[df$value !=0,]
# Visualize boxplot with z scores 
df$row <- rownames(df)
df$orderr <- order(df$value)

ggplot(df, aes(y = value, x = row, oder= orderr))+geom_point()



dp_data <- Dotplot_data(mono, features = c(gene, neat, "SFPQ", "NP-unknown"), group.by = "group", dot.scale =10,scale.by = "radius", cols = c( "lightblue", "darkred"))

ggplot(dp_data, aes(x  = avg.exp, y = pct.exp, col = features.plot))+geom_point()+theme_paper


```











