---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)
library(GENIE3)
library(igraph)
library(RCy3)
library(Rgraphviz)
library(reshape2)
library(org.Mmu.eg.db)
library(enrichplot)
library(networkD3)

source("../../utils/02_sc_utils.R")

# IN VIVO 
immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")

dim(immune.combined)

col_lnc = "#870052"
col_mrna = "#2e88bf"
de_all<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_stages.rds")
de_lnc <- de_all[de_all$type == "lnc", ]

# get DE lnc names


robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))


de_all$type <- ifelse(de_all$primerid %in% all_lncrnas, "lnc", "pc")
de_lnc_names <- gsub("-unknown", "",de_all[de_all$type == "lnc", ]$primerid)


ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))


orthologs <- readRDS("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs.rds")
get_ortholog <- function(gene_id){
  transcripts_of_gene <- ref[ref$gene_id == gene_id, ]$transcript_id
  return(orthologs[orthologs$lnc %in% transcripts_of_gene,])
}
get_gene_id <-function(transcript_id){
  return(unique(ref[!is.na(ref$transcript_id) & ref$transcript_id == transcript_id,]$gene_id))
}

#orthologs_gene_ids <- lapply(unique(orthologs$lnc), get_gene_id) 
#gene_ids <- apply(orthologs,1,function(x) get_gene_id(x[1]) )
#orthologs$gene_id <- gene_ids
#saveRDS(orthologs,"/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/05_orthologs/orthologs_geneids.rds" )


#distances <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/07_colocation/  ")
#distances$gene <-  gsub("-unknown", "",distances$gene)
#distances$second <-  gsub("-unknown", "",distances$second)


```

# Prepare the matrixes to calculate the networks on. One per celltype, only DE genes. 

```{r a}
myeloids <- immune.combined[,Idents(immune.combined)=="Myeloid"]
b <- immune.combined[,Idents(immune.combined)=="B"]
t <- immune.combined[,Idents(immune.combined)=="T"]

de_myeloids <- myeloids[rownames(myeloids) %in% de_all,]
de_b <- b[rownames(b) %in% de_all,]
de_t <- t[rownames(t) %in% de_all,]

exprMatr_myeloids_de <- de_myeloids@assays$integrated@data
exprMatr_b_de <- de_b@assays$integrated@data
exprMatr_t_de <- de_t@assays$integrated@data


universe_myeloids <- unique(rownames(myeloids))
universe_t <- unique(rownames(t))
universe_b <- unique(rownames(b))
```



# Prepare the matrixes with Ebola genes 

```{r a}
ebola_genes <- paste(ebola_ref$gene_id, "-unknown", sep = "")

de_myeloids_ebola <- myeloids[rownames(myeloids) %in% c(de_all,ebola_genes)]
de_b_ebola <- b[rownames(b) %in% c(de_all,ebola_genes)]
de_t_ebola <- t[rownames(t) %in% c(de_all,ebola_genes),]
```

# Save all the expression matrices

```{r a}
library(zeallot)
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))

de_all_genes<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_genes.rds")
length(unique(de_all_genes))

de_mrna_all <- de_all_genes[de_all_genes %in% annotated_mrnas]
length(unique(de_mrna_all))

de_lnc_all <- de_all_genes[de_all_genes %in% all_lncrnas]
length(unique(de_lnc_all))

mrnas_ranges <- ref[ref$gene_name %in% annotated_mrnas  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unknown", "",all_lncrnas))]
lnc_ranges_de <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unknown", "",de_lnc_all))]
mrnas_ranges_de<- ref[ref$gene_name %in% unique(unlist(de_mrna_all))  ]


# Add missing gene entries (as novel lncrnas only prsent transcript and exon entries)
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(base::c,gr_list)
lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)

# Only select the DE interestin
lnc_ranges_complete_de <- lnc_ranges_complete[lnc_ranges_complete$gene_id %in% gsub("-", "_",gsub("-unknown", "",unique(unlist(de_lnc_all)))), ]
length(unique(lnc_ranges_complete_de$gene_id))
```


# ---------------------- RUN ON CLUSTER GRNBOOST ---------------------------


# ---------------------------------------
#      Myeloids 
# ---------------------------------------


# 1) Visualize the hubs and do the enrichment 
# 2) Visualize sme specific examples 

```{r a}
library(reshape2)
# read in 
weightdf_myeloids <- read.table("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/output_myeloids.tsv")
weightMat_myeloids <- acast(weightdf_myeloids, V1~V2, value.var="V3")

# Clean names 
colnames(weightMat_myeloids) <- gsub(".unknown", "", colnames(weightMat_myeloids))
rownames(weightMat_myeloids) <- gsub(".unknown", "", rownames(weightMat_myeloids))

de_myeloids <- gsub("-unknown", "",de_lnc[de_lnc$celltype == "Myeloid",]$primerid)
```


# Only with lnc
```{r a}
set.seed(123)

# Get the graph and extract the clusters interesting for lncRNAS
prepare_hubs <- function(weightMat_myeloids, de_lnc_names = de_lnc_names){
  # Extract lncRNAS
  linkList <- getLinkList(weightMat_myeloids)
  max_n <- nrow(linkList)*0.5/100
  linkList <- getLinkList(weightMat_myeloids, reportMax = max_n)
  
  
  lnc_graphs <- linkList[as.character(linkList$regulatoryGene) %in% de_lnc_names |as.character(linkList$targetGene) %in% de_lnc_names, ]
  Gsi <- graph.data.frame(lnc_graphs,directed = F)
  
  hubs <- decompose.graph(Gsi, min.vertices = 5) 
  print(paste("#Hubs: ",length(hubs)))
  
  # Check how big they are
  print("Size of hubs: ")
  print(unlist(lapply(hubs, function(subgraph) length(V(subgraph)))))
  return(hubs)
}

# Visualize the hub and its connections

check_hub <- function(subgraph, universe_local, de_lnc_names = de_myeloids){
  # Visualize subgraph 
  V(subgraph)$color <- ifelse(names(as.list(V(subgraph))) %in% de_lnc_names, "red", "grey")
  V(subgraph)[startsWith(names(as.list(V(subgraph))), "MSTRG")]$color <- "purple"
  V(subgraph)$frame.color <- ifelse( names(as.list(V(subgraph))) %in% de_lnc_names & names(as.list(V(subgraph))) %in% orthologs$gene_id , "blue", "dark grey")

  plot(subgraph,  vertex.size = 8, vertex.frame.width =19 , vertex.label=NA, vertex.ltype=3)
  # Arrow size based on weigth
  #E(subgraph)$width <- E(subgraph)$weight/6
  #edge.start <- ends(subgraph, es=E(subgraph))[,1]
  #edge.col <- V(subgraph)$color[edge.start]
  #plot(subgraph, edge.color=edge.col, edge.curved=.1, vertex.label=NA, vertex.size=8)  

  #plot(subgraph, vertex.size=5, vertex.label.font=1, vertex.label.cex=.6)
  # Calculate enrichment
  ego <- clusterProfiler::enrichGO(gene = names(as.list(V(subgraph))),OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe_local)
  p0 <- (clusterProfiler::dotplot(ego))
  p1 <- emapplot(ego)
  p3 <- cnetplot(ego)
  netm <- as_adjacency_matrix(subgraph, attr="weight", sparse=F)
  palf <- colorRampPalette(c("light grey", "red")) 
  netm_filt <- netm[rownames(netm)[rownames(netm) %in% gsub("-unknown", "",de_lnc_names)], setdiff(rownames(netm), rownames(netm)[rownames(netm) %in% gsub("-unknown", "",de_lnc_names)]), drop = FALSE]
  if(nrow(netm_filt)>1){
      p2 <- heatmap.2(netm_filt, Rowv = TRUE, Colv = TRUE, col = palf(100), 
        scale="none", margins=c(10,10),trace='none',dendrogram='none',cexRow = 0.9)
  }else{
     p2 <- heatmap.2(rbind(netm_filt, netm_filt), Rowv = TRUE, Colv = TRUE, col = palf(100), 
        scale="none", margins=c(10,10),trace='none',dendrogram='none',cexRow = 0.9)
  }

  return(list(p0,p1,p2,p3))
}


plot_graph_heatmap <- function(subgraph){
  netm <- as_adjacency_matrix(subgraph, attr="weight", sparse=F)
  palf <- colorRampPalette(c("light grey", "red")) 
  netm_filt <- netm[rownames(netm)[rownames(netm) %in% gsub("-unknown", "",de_lnc_names)], setdiff(rownames(netm), rownames(netm)[rownames(netm) %in% gsub("-unknown", "",de_lnc_names)]), drop = FALSE]
  if(nrow(netm_filt)>1){
      p2 <- heatmap.2(netm_filt, Rowv = TRUE, Colv = TRUE, col = palf(100), 
        scale="none", margins=c(10,10),trace='none',dendrogram='none',cexRow = 0.9)
  }else{
     p2 <- heatmap.2(rbind(netm_filt, netm_filt), Rowv = TRUE, Colv = TRUE, col = palf(100), 
        scale="none", margins=c(10,10),trace='none',dendrogram='none',cexRow = 0.9)
  }
  return(p2)
}

```


# Find communities 

```{r a}
#universe_local <- universe_myeloids

linkList <- getLinkList(weightMat_myeloids)
max_n <- nrow(linkList)*0.5/100
linkList <- getLinkList(weightMat_myeloids, reportMax = max_n)
min(linkList$weight)

lnc_graphs <- linkList[as.character(linkList$regulatoryGene) %in% de_myeloids |as.character(linkList$targetGene) %in% de_myeloids, ]
Gsi <- graph.data.frame(lnc_graphs,directed = F)

# Cluster with louvain algorithm 
lou <- cluster_louvain(Gsi)
LO = layout_with_fr(Gsi)

# Only keep the ones haveing at least 5 nodes
Small = which(table(lou$membership) < 5)
Keep = V(Gsi)[!(lou$membership %in% Small)]

## Get subgraph & plot
gD2  = induced_subgraph(Gsi, Keep)
lou2 = cluster_louvain(gD2)
LO2 = LO[Keep,]

```

```{r a}

# Graph plotting Proprieties
V(gD2)$color <- ifelse(names(as.list(V(gD2))) %in% de_lnc_names, "red", "grey")
# highlight novel ones 
V(gD2)[startsWith(names(as.list(V(gD2))), "MSTRG")]$color <- "purple"
# Highlight orthologs
V(gD2)$frame.color <- ifelse( names(as.list(V(gD2))) %in% de_lnc_names & names(as.list(V(gD2))) %in% orthologs$gene_id , "blue", "#7A7A7A")

V(gD2)$community <- membership(lou2)

# Hihglight memberships to community: testing 
# V(gD2)$color <- ifelse(lou2$membership == 2, "black", "grey")

dev.new(width=7, height=4)
# ------------------------------------
#       Visualize the communities
# ------------------------------------
plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), edge.color= "#7A7A7A", vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"))

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)

saveRDS(gD2,"/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/gD2.rds" )
saveRDS(lou2,"/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/lou.rds" )


```

# Subselect in one community 

```{r a}

ortholog_genes <- unique(gsub("-unknown", "", ref[ref$transcript_id %in% unique(orthologs$lnc), ]$gene_id))


visualize_community_expression <- function(seuobj, gD2, n){
    sub <- V(gD2)[V(gD2)$community == n]
    genes <- names(V(gD2)[V(gD2)$community == n])
    print(n)
      calc_z <- function(expression, na.remove = F){
      mean <- rowMeans(expression, na.rm = na.remove); sd <- apply(expression, 1 , function(x) sd(x, na.rm = na.remove)); 
      z <- (expression-mean)/sd
      return(z)
    }
    
    get_z_scores <- function(seuratobject){
      # First computes z score per gene per cell and then calculate the mean of z score per celltype
      averaged_z_matrix <- sapply(unique(colnames(zmatrix)), function(x) rowMeans(zmatrix[,colnames(zmatrix) ==x]))
      return(averaged_z_matrix)
    }
    expression_genes <- seuobj[gsub("-unknown", "", rownames(seuobj)) %in% genes, ]@assays$RNA@data
    colnames(expression_genes) <- seuobj$group
   
    zmatrix <- as.matrix(calc_z(expression_genes, T))
    m<- sapply(unique(colnames(zmatrix)), function(x) rowMeans(zmatrix[,colnames(zmatrix) ==x]))
    rownames(m) <- gsub("-unknown", "", rownames(m))
    

    # Prepare dataframe for plotting
    df <- setNames(reshape2::melt(m), c('rows', 'vars', 'values'))
    names(df) <- c("gene", "group", "value")
    # Visualize boxplot with z scores 
    p1 <- ggplot(df, aes( x = group, y = value, col = group, fill = group) )+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+ylim(-1,1)+scale_fill_manual(values= c(brewer.pal(8, "Pastel2")[5:8]))+ggtitle(paste("Community: ",n)) 
    # Visualize heatmap
    column_ha <- HeatmapAnnotation(comparison = anno_text( unlist(lapply(colnames(m), function(x) toupper(substr(x,1,1)))), rot = 0, gp = gpar(fontsize = 25, fill =  brewer.pal(8, "Pastel2")[5:8]), location = 0.5, just = "center"))
   
  
  df_col<- data.frame(id = names(sub), col = sub$color, stringsAsFactors = F) 
  rownames(df_col) <- df_col$id
  row_ha_type <- rowAnnotation(genetype =df_col[rownames(m),]$col, show_legend = FALSE,
                               show_annotation_name = FALSE,
                               col = list(genetype = c("grey" = "grey", "red" = "red",
                                                       "purple" = "purple"
                               )))
  
    h1 <- Heatmap(m,  top_annotation = column_ha,right_annotation =row_ha_type,   column_order = c("baseline","early", "middle", "late"))+ggtitle(paste("Community: ",n)) 
    
    p1
    h1
    return(list(p1,h1))
}



lapply(unique(V(gD2)$community), function(n) visualize_community_expression(myeloids,gD2 , n))
```



```{r a}
# Pick one gene and visualize its expression across stages 
gene <- "MSTRG.146560"
inspect_gene_expression <- function(immune.combined, myeloids, gene){
  expression_gene <- myeloids[gsub("-unknown", "", rownames(myeloids))== gene, ]@assays$RNA@data
  colnames(expression_gene) <- myeloids$group
  df <- reshape2::melt(as.matrix(expression_gene))
  names(df) <- c("gene", "group", "value")
  df <- df[df$value !=0,]
  # Visualize boxplot with z scores 
  table(df$group)
  p1 <- ggplot(df, aes( x = group, y = value, col = group, fill = group) )+geom_violin(alpha = 0.7, color = "dark grey ")+theme_paper+scale_fill_manual(values= c(brewer.pal(8, "Pastel2")[5:9]))+ geom_boxplot(width=0.1, color = "dark grey ")
  
  
  p2 <- plot_seperate_features(immune.combined, gene  = c(paste(gene, "-unknown", sep = "")), ident = "group")
  return(list(p1, p2))
}
inspect_gene_expression(immune.combined, myeloids, gene)


zmatrix <- as.matrix(calc_z(expression_gene, T))
df <- reshape2::melt(as.matrix(zmatrix))
names(df) <- c("gene", "group", "value")
df <- df[df$value !=0,]
ggplot(df, aes( x = group, y = value, col = group, fill = group) )+geom_boxplot(alpha = 0.7, color = "dark grey ")+theme_paper+ geom_boxplot(width=0.1, color = "dark grey")

```


```{r a}
# ----------------------------------------
#     Separate hubs and communities 
# ----------------------------------------
hubs_myeloids <- prepare_hubs(weightMat_myeloids, de_myeloids)

# ---------------------------------------------
#       Check the enrichment per community 
# --------------------------------------------
do_go_community<- function(list, universe){
   ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
   p1 <- clusterProfiler::cnetplot(ego)
   p2 <- heatplot(ego, showCategory =5)
   return(list(p1,p2))
}

lapply(unique(lou2$membership)[1:2], function(n) do_go_community(lou2$names[lou2$membership == n],universe_myeloids))
```


```{r a}
# ----------------------
#     Community one 
# ----------------------

list <- lou2$names[lou2$membership == 1]
universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::heatplot(ego, showCategory = 5 )
clusterProfiler::dotplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)
plot_graph_heatmap(subgraph(gD2, V(gD2)[community==1]))
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == 1])[rownames(gD2[V(gD2)$community  == 1]) %in% de_lnc_names]
Reduce("rbind", lapply(lnc_comm, get_ortholog))

```

```{r a}
# ----------------------
#     Community two 
# ----------------------
n <- 2
list <- lou2$names[lou2$membership == n]
universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5 )
clusterProfiler::heatplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)
plot_graph_heatmap(subgraph(gD2, V(gD2)[community==n]))
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
Reduce("rbind", lapply(lnc_comm, get_ortholog))

```

```{r a}

# ----------------------
#     Community 3 
# ----------------------
n <- 3
list <- lou2$names[lou2$membership == n]
universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
barplot(ego, showCategory = 5)
clusterProfiler::heatplot(ego, showCategory = 5)


plot(subgraph(gD2, V(gD2)[community==n]))
gD2[V(gD2)$community  == 3,]
plot_graph_heatmap(subgraph(gD2, V(gD2)[community==n]))

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
```


```{r a}
# ----------------------
#     Community 4
# ----------------------

n <- 4
list <- lou2$names[lou2$membership == n]
universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))
gD2[V(gD2)$community  == 3,]

plot_graph_heatmap(subgraph(gD2, V(gD2)[community==n]))

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
```


```{r a}
# ----------------------
#     Community 5
# ----------------------

n <- 5
list <- lou2$names[lou2$membership == n]
universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))
clusterProfiler::cnetplot(ego, showCategory = 5 )
clusterProfiler::heatplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)
plot_graph_heatmap(subgraph(gD2, V(gD2)[community==n]))
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]

Reduce("rbind", lapply(lnc_comm, get_ortholog))

gene <- "CSF3R"
plot_seperate_features(immune.combined, gene  = gene, ident = "group", colors =c('#DCD9DE','navy'))

rownames(gD2[V(gD2)$community  == n])

lapply(rownames(gD2[V(gD2)$community  == n])[1:10], function(gene) plot_seperate_features(immune.combined, gene  = gene, ident = "group", colors =c('#DCD9DE','blue')))

plot_seperate_features(immune.combined, gene  = c(paste(gene, "-unknown", sep = "")), ident = "group")

```

```{r a}
# ----------------------
#     Community 6
# ----------------------

n <- 6
list <- lou2$names[lou2$membership == n]
universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]



plot(subgraph(gD2, V(gD2)[community==n]))
clusterProfiler::cnetplot(ego, showCategory = 5 )
clusterProfiler::heatplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)
plot_graph_heatmap(subgraph(gD2, V(gD2)[community==n]))
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
```


```{r a}
# ----------------------
#     Community 7
# ----------------------

n <- 7
list <- lou2$names[lou2$membership == n]
list <- list[!(list %in% de_lnc_names)]

universe <- universe_myeloids
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]




plot(subgraph(gD2, V(gD2)[community==n]))
clusterProfiler::cnetplot(ego, showCategory = 5 )
clusterProfiler::heatplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)
plot_graph_heatmap(subgraph(gD2, V(gD2)[community==n]))
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]


```


# ------------------------------------------------------------
#                   Distances 
# ------------------------------------------------------------

```{r a}
#  Find distance of a pair 
get_pair_distance <- function(one, two, distances){
  if(nrow(distances[distances$gene == one & distances$second == two, ])>0 ){
    return(data.frame(one, two, dist = distances[distances$gene == one & distances$second == two, ]$distances, bin = distances[distances$gene == one & distances$second == two, ]$bin))
  }else if(nrow(distances[distances$gene == two & distances$second == one, ])>0){
    return(data.frame(one, two, dist = distances[distances$gene == two & distances$second == one, ]$distances, bin = distances[distances$gene == two & distances$second == one, ]$bin))
  }else{
    return(data.frame(one, two, dist = "inf", bin = ">2Mb"))
  }
}


```




```{r a}
de_myeloids_all <- gsub("-unknown", "",de_all[de_all$celltype == "Monocyte",]$primerid)


lnc_ranges_myeloids_de <- lnc_ranges_complete_de[lnc_ranges_complete_de$gene_id %in% de_myeloids,]
mrna_ranges_myeloids_de <- mrnas_ranges_de[mrnas_ranges_de$gene_id %in% de_myeloids_all,]

# Get distances between DE genes 
distances <- do.call(rbind,lapply(1:(length(lnc_ranges_myeloids_de)), function(index) (get_n_closest_gr_name(lnc_ranges_myeloids_de[index], mrnas_ranges_de, maxgap = 100000000L))))
distances$gene <-  gsub("-unknown", "",distances$gene)
distances$second <-  gsub("-unknown", "",distances$second)
distances$bin <- (cut(distances$distances, breaks = c(-1,2000000, Inf), labels = c("<2Mb",">2Mb")))
# Test 
#get_pair_distance("CCL3", lnc_comm, distances)

pair_distance <- do.call(rbind,(apply(ends(gD2, es=E(gD2)), 1, function(x) get_pair_distance(x[1], x[2], distances))))
E(gD2)$bin_distance <- unlist(pair_distance$bin)

colors <- c("dark grey", "black")
width <- c(1,3)
col_int <- as.integer(cut(abs(E(gD2)$bin_distance-(nlevels(unlist(pair_distance$bin))+1)), breaks = 2))
bins <- pair_distance$bin
unique( pair_distance$bin)
```


# Plot with distances 
```{r a}

dev.new(width=7, height=4)

plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"), 
     edge.color=colors[col_int], edge.width = width[col_int])

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)

legend("topleft",
       legend=levels(pair_distance$bin),
       fill=rev(colors), border=NA)

```






# -------------------------------------------------------------------------
#         Visualize directionality of co-regulation 
# -------------------------------------------------------------------------







```{r a}
linkList <- getLinkList(weightMat_myeloids)
lnc_graphs <- linkList[as.character(linkList$regulatoryGene) %in% de_myeloids |as.character(linkList$targetGene) %in% de_myeloids, ]



# Get all pc connected with lnc 
# all_pc <- unique(unlist(list(lnc_graphs$regulatoryGene, lnc_graphs$targetGene)))
# ego <- clusterProfiler::enrichGO(gene = all_pc,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe_myeloids)
# (clusterProfiler::dotplot(ego))

# Get directionality 

#subset <- lnc_graphs[1:500,]

subset <- as.data.frame(ends(gD2, es=E(gD2)))
names(subset) <- c("regulatoryGene", "targetGene")
# Only for the ones in the graph 

de_all_myeloids <- de_all[de_all$celltype == "Myeloid",]

# IMPORTNT
subset$regulatoryGene <- gsub('.', "-", subset$regulatoryGene, fixed = T)
subset$targetGene <- gsub('.', "-", subset$targetGene, fixed = T)
de_all_myeloids$primerid <- gsub('.', "-", de_all_myeloids$primerid , fixed = T)
de_all$primerid <- gsub('.', "-", de_all$primerid , fixed = T)


# Obatin type of gene 
reg_type <- unlist(lapply(subset$regulatoryGene, function(gene) unique(de_all[gsub("-unknown", "", de_all$primerid)==gene,]$type)))

# Add regulatory direction 
directions <- lapply(subset$regulatoryGene, function(gene) unique(ifelse(de_all[gsub("-unknown", "", de_all$primerid)==gene,]$coef>0, "up", "down")))
directions_ready <- unlist(lapply(directions, function(x) ifelse(length(x)!=1, "uncertain", x)))
subset$regulatory_direction <- directions_ready

# Add target direction 
directions_target <- lapply(subset$targetGene, function(gene) unique(ifelse(de_all_myeloids[gsub("-unknown", "", de_all_myeloids$primerid)==gene,]$coef>0, "up", "down")))
directions_ready_t <- unlist(lapply(directions_target, function(x) ifelse(length(x)!=1, "uncertain", x)))
subset$target_direction <- directions_ready_t

# Add concordancy 
subset$concordance <- apply(subset,1,function(line) ifelse(line[3] == line[4], "concordant", "unconcordant"))
subset$concordance<- ifelse(subset$target_direction == "uncertain" | subset$regulatory_direction == "uncertain", "uncertain", subset$concordance )

# Plot concordancy 
subset$regulatory_type <- unlist(lapply(subset$regulatoryGene, function(gene) unique(de_all[gsub("-unknown", "",de_all$primerid) == gene, ]$type)))
subset$target_type <- unlist(lapply(subset$targetGene, function(gene) unique(de_all[gsub("-unknown", "",de_all$primerid) == gene, ]$type)))

# Remove lnc-lnc pairs 
subset <- subset[subset$target_type != subset$regulatory_type,]
# Remove pairs with uncertain 
subset <- subset[!(subset$regulatory_direction == "uncertain" | subset$target_direction == "uncertain"),]



subset$lnc_type  <- ifelse(subset$target_type == "lnc", subset$target_direction, subset$regulatory_direction)
subset$pc_type  <- ifelse(subset$target_type == "pc", subset$target_direction, subset$regulatory_direction)

```



```{r a}

prep_pairs <- function(subset){
  df_pairs <- data.frame(table(subset %>% dplyr::group_by(lnc_type, pc_type) %>% summarise(a = paste(`lnc_type`, `pc_type`))))
  names(df_pairs) <- c("pair", "value")
  df_pairs$source <- unlist(lapply(df_pairs$pair, function(x) str_split(x, " ")[[1]][1]))
  df_pairs$target <- unlist(lapply(df_pairs$pair, function(x) str_split(x, " ")[[1]][2]))
  df_pairs$target <- paste(df_pairs$target, "_pc")
  return(df_pairs)
}

plot_sankey <- function(df_pairs, group = c("type_a","type_a","type_b","type_b","type_b", "type_b", "type_b")){
  # A connection data frame is a list of flows with intensity for each flow
links <- data.frame(
  source=df_pairs$source, 
  target=df_pairs$target, 
  value=df_pairs$value
  )
 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes <- data.frame(
  name=c(as.character(links$source), 
  as.character(links$target)) %>% unique()
)
 
# With networkD3, connection must be provided using id, not using real name like in the links dataframe.. So we need to reformat it.
links$IDsource <- match(links$source, nodes$name)-1 
links$IDtarget <- match(links$target, nodes$name)-1
my_color <- 'd3.scaleOrdinal() .domain(["group_A", "group_B","group_C", "group_D"]) .range(["#7ABFF3", "#F04343" , "#7ABFF3", "#F04343"])'

p <- sankeyNetwork(Links = links, Nodes = nodes,
              Source = "IDsource", Target = "IDtarget",
              Value = "value", NodeID = "name", 
              sinksRight=FALSE, fontSize=18, colourScale = my_color)
p
return(p)
}

# Get concordance in graph 
get_concordance <- function(one, two, concordance){
  if(nrow(concordance[concordance$gene == one & concordance$second == two, ])>0 ){
    return(data.frame(one, two, dist = concordance[concordance$gene == one & concordance$second == two, ]$concordance))
  }else if(nrow(concordance[concordance$gene == two & concordance$second == one, ])>0){
    return(data.frame(one, two, dist = concordance[concordance$gene == two & concordance$second == one, ]$concordance))
  }
}



concordance <- subset
concordance$gene <- concordance$regulatoryGene
concordance$second <- concordance$targetGene



pair_concordance <- do.call(rbind,(apply(ends(gD2, es=E(gD2)), 1, function(x) get_concordance(x[1], x[2], concordance))))
E(gD2)$conc <- unlist(pair_concordance$dist)



colors <- c("#00CC00", "red")
width <- c(2,2)
col_int <- as.integer(cut(abs(E(gD2)$bin_distance-(nlevels(unlist(pair_concordance$dist))+1)), breaks = 2))
bins <- pair_concordance$dist
unique( pair_concordance$dist)



# Slankey plot 

df_pairs <- prep_pairs(subset)
plot_sankey(df_pairs)

ends(gD2, es=E(gD2))
ends(E(gD2)[1])

ggplot(subset, aes(x = concordance, fill = concordance))+geom_histogram(stat="count")+theme_paper+scale_fill_brewer(palette="Paired")

```


# Plot with distances 
```{r a}

dev.new(width=7, height=4)

plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"), 
     edge.color=colors[col_int], edge.width = width[col_int])

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)

legend("topleft",
       legend=levels(pair_concordance$dist),
       fill=rev(colors), border=NA)

```








# ---------------------------------------# ---------------------------------------# ---------------------------------------# -------------
# ---------------------------------------
# ---------------------------------------
#       Replicate for B and T cells   
# ---------------------------------------
# ---------------------------------------



# ---------------------------------------
#      B  
# ---------------------------------------





```{r a}

# read in 
weightdf_b <- read.table("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/output_b_de.tsv")
weightMat_b <- acast(weightdf_b, V1~V2, value.var="V3")

# Clean names 
colnames(weightMat_b) <- gsub(".unknown", "", colnames(weightMat_b))
rownames(weightMat_b) <- gsub(".unknown", "", rownames(weightMat_b))

de_b <- gsub("-unknown", "",de_lnc[de_lnc$celltype == "B",]$primerid)


################# Community detection 
linkList <- getLinkList(weightMat_b)
max_n <- nrow(linkList)*0.5/100
linkList <- getLinkList(weightMat_b, reportMax = max_n)

min(linkList$weight)

lnc_graphs <- linkList[as.character(linkList$regulatoryGene) %in% de_b |as.character(linkList$targetGene) %in% de_b, ]
Gsi <- graph.data.frame(lnc_graphs,directed = F)

# Cluster with louvain algorithm 
lou <- cluster_louvain(Gsi)
LO = layout_with_fr(Gsi)

# Only keep the ones haveing at least 5 nodes
Small = which(table(lou$membership) < 5)
Keep = V(Gsi)[!(lou$membership %in% Small)]

## Get subgraph & plot
gD2  = induced_subgraph(Gsi, Keep)
lou2 = cluster_louvain(gD2)
LO2 = LO[Keep,]

```




```{r a}

# Graph plotting Proprieties
V(gD2)$color <- ifelse(names(as.list(V(gD2))) %in% de_lnc_names, "red", "grey")
# highlight novel ones 
V(gD2)[startsWith(names(as.list(V(gD2))), "MSTRG")]$color <- "purple"
# Highlight orthologs
V(gD2)$frame.color <- ifelse( names(as.list(V(gD2))) %in% de_lnc_names & names(as.list(V(gD2))) %in% orthologs$gene_id , "blue", "#7A7A7A")

V(gD2)$community <- membership(lou2)

# Hihglight memberships to community: testing 
# V(gD2)$color <- ifelse(lou2$membership == 2, "black", "grey")

dev.new(width=7, height=4)
# ------------------------------------
#       Visualize the communities
# ------------------------------------
plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), edge.color= "#7A7A7A", vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"))

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)
```








```{r a}
de_b_all <- gsub("-unknown", "",de_all[de_all$celltype == "B",]$primerid)


lnc_ranges_b_de <- lnc_ranges_complete_de[lnc_ranges_complete_de$gene_id %in% de_b_all,]
mrna_ranges_b_de <- mrnas_ranges_de[mrnas_ranges_de$gene_id %in% de_b_all,]

# Get distances between DE genes 
distances <- do.call(rbind,lapply(1:(length(lnc_ranges_b_de)), function(index) (get_n_closest_gr_name(lnc_ranges_b_de[index], mrnas_ranges_de, maxgap = 100000000L))))
distances$gene <-  gsub("-unknown", "",distances$gene)
distances$second <-  gsub("-unknown", "",distances$second)
distances$bin <- (cut(distances$distances, breaks = c(-1,2000000, Inf), labels = c("<2Mb",">2Mb")))
# Test 
#get_pair_distance("CCL3", lnc_comm, distances)

pair_distance <- do.call(rbind,(apply(ends(gD2, es=E(gD2)), 1, function(x) get_pair_distance(x[1], x[2], distances))))
E(gD2)$bin_distance <- unlist(pair_distance$bin)

colors <- c("dark grey", "black")
width <- c(1,3)
col_int <- as.integer(cut(abs(E(gD2)$bin_distance-(nlevels(unlist(pair_distance$bin))+1)), breaks = 2))
bins <- pair_distance$bin
unique( pair_distance$bin)
```


# Plot with distances 
```{r a}

dev.new(width=7, height=4)

plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"), 
     edge.color=rev(colors)[col_int], edge.width = rev(width)[col_int])

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)

legend("topleft",
       legend=levels(pair_distance$bin),
       fill=(colors), border=NA)

```

```{r a}
# ----------------------------------------
#     Separate hubs and communities 
# ----------------------------------------
hubs_b<- prepare_hubs(weightMat_b, de_b)

check_hub(hubs_b[[1]], universe_b)
#check_hub(hubs_b[[2]], universe_b)
#check_hub(hubs_b[[3]], universe_b)

# ---------------------------------------------
#       Check the enrichment per community 
# --------------------------------------------
do_go_community<- function(list, universe){
   ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
   p1 <- clusterProfiler::cnetplot(ego)
   p2 <- heatplot(ego, showCategory =5)
   return(list(p1,p2))
}

lapply(unique(lou2$membership)[1:2], function(n) do_go_community(lou2$names[lou2$membership == n],universe_myeloids))



# ----------------------
#     Hub two 
# ----------------------
n <- 2
list <- lou2$names[lou2$membership == n]
universe <- universe_b
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5 )

plot_graph_heatmap(hubs_b[[2]])
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
get_ortholog(lnc_comm)
plot(subgraph(gD2, V(gD2)[community==n]))
# ----------------------
#     Community 3 
# ----------------------
n <- 3
list <- lou2$names[lou2$membership == n]
universe <- universe_b
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))
gD2[V(gD2)$community  == 3,]

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
get_ortholog(lnc_comm[1])
```





```{r a}



hubs_b <- prepare_hubs(weightMat_b, de_b)

check_hub(hubs_b[[1]], universe_b)
check_hub(hubs_b[[2]], universe_b)

# Plot second hub 
subgraph <- hubs_b[[2]]
V(subgraph)$color <- ifelse(names(as.list(V(subgraph))) %in% de_lnc_names, "red", "grey")
V(subgraph)[startsWith(names(as.list(V(subgraph))), "MSTRG")]$color <- "purple"
plot(subgraph, vertex.size=10,
vertex.label.color = "black", vertex.label.cex = 0.9, vertex.label.degree = -pi/2,
edge.arrow.size = 0.3, edge.arrow.width = 0.4, edge.color = "black")


plot_seperate_features(immune.combined, gene  = c("IL22"), ident = "group")
plot_seperate_features(immune.combined, gene  = c("MSTRG.31111-unknown"), ident = "group")
plot_seperate_features(immune.combined, gene  = c("GADD45B"), ident = "group")
#plot_seperate_features(immune.combined, gene  = c("ENSMMUG00000065174-unknown"), ident = "group")

# Investigate genes one by one
# ORTHOLOG 
gene_id <- "MSTRG.31111"
get_ortholog(gene_id)
# Find genomic location 
#ref[ref$gene_id == "MSTRG.31111",]

# Investigate genes one by one
# ORTHOLOG 
gene_id <- "ENSMMUG00000061913"
get_ortholog(gene_id)
# Find genomic location 
#ref[ref$gene_id == "MSTRG.31111",]
```






# ---------------------------------------
#      T 
# ---------------------------------------
```{r a}

# read in 
weightdf_t <- read.table("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/output_t_de.tsv")
weightMat_t <- acast(weightdf_t, V1~V2, value.var="V3")

# Clean names 
colnames(weightMat_t) <- gsub(".unknown", "", colnames(weightMat_t))
rownames(weightMat_t) <- gsub(".unknown", "", rownames(weightMat_t))

de_t <- gsub("-unknown", "",de_lnc[de_lnc$celltype == "T",]$primerid)


################# Community detection 
linkList <- getLinkList(weightMat_t)
max_n <- nrow(linkList)*0.5/100
linkList <- getLinkList(weightMat_t, reportMax = max_n)

min(linkList$weight)

lnc_graphs <- linkList[as.character(linkList$regulatoryGene) %in% de_t |as.character(linkList$targetGene) %in% de_t, ]
Gsi <- graph.data.frame(lnc_graphs,directed = F)

# Cluster with louvain algorithm 
lou <- cluster_louvain(Gsi)
LO = layout_with_fr(Gsi)

# Only keep the ones haveing at least 5 nodes
Small = which(table(lou$membership) < 5)
Keep = V(Gsi)[!(lou$membership %in% Small)]

## Get subgraph & plot
gD2  = induced_subgraph(Gsi, Keep)
lou2 = cluster_louvain(gD2)
LO2 = LO[Keep,]

```



```{r a}
de_t_all <- gsub("-unknown", "",de_all[de_all$celltype == "T",]$primerid)


lnc_ranges_t_de <- lnc_ranges_complete_de[lnc_ranges_complete_de$gene_id %in% de_t_all,]
mrna_ranges_t_de <- mrnas_ranges_de[mrnas_ranges_de$gene_id %in% de_t_all,]

# Get distances between DE genes 
distances <- do.call(rbind,lapply(1:(length(lnc_ranges_t_de)), function(index) (get_n_closest_gr_name(lnc_ranges_t_de[index], mrna_ranges_t_de, maxgap = 100000000L))))
distances$gene <-  gsub("-unknown", "",distances$gene)
distances$second <-  gsub("-unknown", "",distances$second)
distances$bin <- (cut(distances$distances, breaks = c(-1,2000000, Inf), labels = c("<2Mb",">2Mb")))
# Test 
#get_pair_distance("CCL3", lnc_comm, distances)

pair_distance <- do.call(rbind,(apply(ends(gD2, es=E(gD2)), 1, function(x) get_pair_distance(x[1], x[2], distances))))
E(gD2)$bin_distance <- unlist(pair_distance$bin)

colors <- c("dark grey", "black")
width <- c(1,3)
col_int <- as.integer(cut(abs(E(gD2)$bin_distance-(nlevels(unlist(pair_distance$bin))+1)), breaks = 2))
bins <- pair_distance$bin
unique( pair_distance$bin)
```


# Plot with distances 
```{r a}

# Graph plotting Proprieties
V(gD2)$color <- ifelse(names(as.list(V(gD2))) %in% de_lnc_names, "red", "grey")
# highlight novel ones 
V(gD2)[startsWith(names(as.list(V(gD2))), "MSTRG")]$color <- "purple"
# Highlight orthologs
V(gD2)$frame.color <- ifelse( names(as.list(V(gD2))) %in% de_lnc_names & names(as.list(V(gD2))) %in% orthologs$gene_id , "blue", "#7A7A7A")

V(gD2)$community <- membership(lou2)

# Hihglight memberships to community: testing 
# V(gD2)$color <- ifelse(lou2$membership == 2, "black", "grey")

dev.new(width=7, height=4)
# ------------------------------------
#       Visualize the communities
# ------------------------------------
plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), edge.color= "#7A7A7A", vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"))

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)

dev.new(width=7, height=4)

plot(lou2, gD2, vertex.label = NA, vertex.size=4, edge.arrow.size = .1, col=V(gD2)$color, layout=layout.fruchterman.reingold(gD2, niter=10000), vertex.frame.width = 8, mark.col=c(brewer.pal(11 , "Set3"), alpha.f = .3),  mark.border=brewer.pal(11 , "Set3"), 
     edge.color=(colors)[col_int], edge.width = (width)[col_int])

legend("topright",
       legend=unique(lou2$membership),
       fill=(brewer.pal(11 , "Set3")), border=NA)

legend("topleft",
       legend=levels(pair_distance$bin),
       fill=rev(colors), border=NA)

```




```{r a}

universe <- universe_t

hubs_t <- prepare_hubs(weightMat_t, de_t)

check_hub(hubs_t[[1]], universe_t, de_t)
check_hub(hubs_t[[2]], universe_t,de_t)

# ----------------------
#     Community one 
# ----------------------
n <- 1
list <- lou2$names[lou2$membership == n]
plot(subgraph(gD2, V(gD2)[community==n]))
plot_graph_heatmap(hubs_t[[1]])


# ----------------------
#     Community two 
# ----------------------
n <- 2
list <- lou2$names[lou2$membership == n]
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5 )
clusterProfiler::dotplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)
# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
get_ortholog(lnc_comm)

# ----------------------
#     Community 3 
# ----------------------
n <- 3
list <- lou2$names[lou2$membership == n]
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))
gD2[V(gD2)$community  == 3,]

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
get_ortholog(lnc_comm[1])


# ----------------------
#     Community 4
# ----------------------

n <- 4
list <- lou2$names[lou2$membership == n]

plot(subgraph(gD2, V(gD2)[community==n]))

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
get_ortholog(lnc_comm[2])




# ----------------------
#     Community 5
# ----------------------

n <- 5
list <- lou2$names[lou2$membership == n]
ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = "SYMBOL",ont = "BP",universe =universe)
clusterProfiler::cnetplot(ego, showCategory = 5)
barplot(ego, showCategory = 5)

plot(subgraph(gD2, V(gD2)[community==n]))
gD2[V(gD2)$community  == 3,]

# Extarct the lnc in the community 
lnc_comm <- rownames(gD2[V(gD2)$community  == n])[rownames(gD2[V(gD2)$community  == n]) %in% de_lnc_names]
ref[ref$gene_id == lnc_comm]
get_ortholog(lnc_comm[1])

```