---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

#-------- PREPARE EX VIVO OBJECT ---------------
# REMOVE DOUBLETS
# IDENTIFY CELL-TYPE MARKERS 
# PREPARE FINAL OBJECT for downstream analyses 

# Imports 
```{r a}
library(Seurat)
library(spatstat)
library(SingleCellExperiment)
library(stringr)
library(rtracklayer)
library(RColorBrewer)
library(wesanderson)
library(Matrix)

data_path <- "/home/mariasr/cluster/02.ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10_NOVELFILTERTEST"
project_path <- "/home/mariasr/cluster/02.ebola_sc/projects"
source(file.path(project_path, "code/ebola/analysis/utils/02_sc_utils.R"))
theme_sc <- theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

immune.combined <- readRDS(file.path(data_path,"05_RObjects/03_prep/immune.combined.rds"))

ebola_genes <- readRDS(file.path("/home/mariasr/cluster/02.ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds"))

# Load annotations
ref <- import(file.path("/home/mariasr/cluster/02.ebola_sc/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/mariasr/cluster/02.ebola_sc/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))


# Doublet detection results 
cell_mask_scrublet<- read.table(file.path(data_path,"/05_RObjects/02_DoubletDetection/immune.combined_scrublet_mask.txt"))

result_path<- file.path(data_path,"/05_RObjects/03_prep")
dir.create(file.path(result_path), showWarnings = FALSE)
print("Loaded")
```



# ---------------------------------------------------------
#    1. REMOVE DOUBLETS: Use Scrublet mask to remove doublets 
# ---------------------------------------------------------

```{r cars}
list <- as.list(as.character(cell_mask_scrublet$V1))
mask <- list == "True"
immune.combined <- immune.combined[,!mask]
```


# -------------------------------------
#    2. Metadata ordering 
# -------------------------------------

```{r cars}
#immune.combned <- readRDS(file.path(data_path,"/05_RObjects/03_prep/immune.combined.all_filt_doblets.rds"))

immune.combined$orig.ident <-  gsub("_moi1e_1", "", gsub("-", "_",immune.combined$orig.ident))
immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "_")[[1]][1]))))
immune.combined$individual_nhp <- ifelse(immune.combined$individual == "EV0003", "NHP1", "NHP2")

immune.combined$cond <- immune.combined$cond <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][2])))
immune.combined$dpi <- unlist(lapply(immune.combined$orig.ident, function(x) unlist(str_split(x, "_")[[1]][3])))
immune.combined$sample <- paste0(immune.combined$individual_nhp," ", immune.combined$dpi)
sum(table(immune.combined$dpi))

#saveRDS(immune.combined, file.path(data_path,"/05_RObjects/03_prep/immune.combined.all_filt_metadata.ordered.rds"))

```


# -------------------------------------------
#  3. Pre-processing and UMAP visualization       
# -------------------------------------------
```{r}

immune.combined <- NormalizeData(immune.combined)
immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.2)

DimPlot(immune.combined, reduction = "umap", label = TRUE)+theme_sc

```


# -------------------------------------
#    Cell-type assignation 
# -------------------------------------

```{r cars}
#pbmc.markers <- readRDS(file.path(data_path, "/05_RObjects/03_prep/markers_exvivo_normal_noHBB.rds"))
#saveRDS(pbmc.markers, file.path(result_path,"markers_exvivo_normal.rds"))
#test <- pbmc.markers %>% group_by(cluster)
#table(test[test$cluster %in% c(0,5,9,4),]$gene)
#m1<- FindMarkers(immune.combined,ident.1=2 )

#markers_2 <- m1
#markers_2 <- test[test$cluster == "2",]
#markers_2 %>% top_n(n = 10, wt = avg_logFC)

b <- c("CD79B", "MS4A1", "CD19", "IGHM")
CD8T <- c("CD3D", "GZMB", "GNLY")
CD4T <-  c("CD3D", "IL7R")
t <- c(CD4T, CD8T)
nk <- c("KLRB1", "GZMB","FCGR3")
mono <- c("LYZ", "PSAP", "CFD", "CD14") 
#mono <- c("IL1B") 
#dc <- c("IRF8", "GZMB")


marker.genes_red <- c(nk, t,b,mono)
#saveRDS(marker.genes_red, file.path(data_path,  "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10_NOVELFILTERTEST/05_RObjects/03_prep/marker.genes.rds"))

plot_genes(immune.combined,mono,threshold= 1, title = "Monocytes marker expression", col = "purple")
plot_genes(immune.combined, nk, threshold = 2 , title = "NK markers expression", col = "purple")
plot_genes(immune.combined, b, threshold = 1 , title = "B markers expression", col = "purple")
plot_genes(immune.combined, t, threshold = 2 , title = "T-Cell markers expression", col = "purple")

immune.combined <- RenameIdents(immune.combined, `0` = "B", `1` = "T",`2` = "NK", `3` = "B",  `4` = "B", `5` = "T", `6` = "T", `7` = "Monocyte", `8` = "Monocyte", `9` = "B", `10` = "Monocyte", `11` = "B")
#saveRDS(immune.combined, file.path(data_path,"05_RObjects/03_prep/03_immune.combined_ready.rds"))

```




```{r}
FeaturePlot(immune.combined, features="log10_total_counts")

immune.combined$log10_total_counts
```









# Plot by UMI 

```{r}
#immune.combined<- readRDS("/home/bscuser/Escritorio/cluster/02.Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10_NOVELFILTERTEST/05_RObjects/03_prep/03_immune.combined.normalized_20k.rds")
#ebola_genes <- readRDS(file.path("/Users/Maria/Desktop/cluster/02_Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds"))



# tSNE colored by high UMIs
immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$counts <- immune.combined$total_counts
immune.combined_df$umi <- "<10k"
immune.combined_df[immune.combined_df$counts > 10000,]$umi <- c("10-12k")
immune.combined_df[immune.combined_df$counts > 12000,]$umi <- c("12-15k")
#immune.combined_df[immune.combined_df$counts > 15000,]$umi <- c(">15k")
levels(immune.combined_df$umi) <- c("0", "10k-12k", "12-15k")

immune.combined$umi <- immune.combined_df$umi
colors <- c("#D3D3D3", "red", "blue", "green")

DimPlot(immune.combined, red = "umap", group.by = "umi", pt.size = 0.3, cols = colors, order = TRUE )+theme_sc

# tSNE colored by low UMIs
immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$counts <- immune.combined$total_counts
immune.combined_df$umi <- ">1700"     
immune.combined_df[immune.combined_df$counts < 1600 & immune.combined_df$counts > 1500,]$umi <- c("1500-1550")
immune.combined_df[immune.combined_df$counts < 1500 & immune.combined_df$counts > 1200,]$umi <- c("1200-1500")
immune.combined_df[immune.combined_df$counts < 1200 & immune.combined_df$counts > 1000,]$umi <- c("<1200")
levels(immune.combined_df$umi) <- c(">3000", "3000-2500", "2500-2000", "2000-1500")


immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$features <- immune.combined$total_features_by_counts
immune.combined_df$feat <- "1000-3500"     
immune.combined_df[immune.combined_df$features > 3500,]$feat <- c(">3500")
immune.combined_df[immune.combined_df$features < 900,]$feat <- c("<1000")
levels(immune.combined_df$feat)<- c("medium", "high", "low")


immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$counts <- immune.combined$total_counts
immune.combined_df$umi <- "1600-14000"     
immune.combined_df[immune.combined_df$counts < 1600,]$umi <- c("<1600")
immune.combined_df[immune.combined_df$counts > 13000 ,]$umi <- c(">13000")
levels(immune.combined_df$umi) <- c(">3000", "3000-2500", "2500-2000")

immune.combined$umi <- immune.combined_df$umi
levels(immune.combined$umi) <- c(">3000", "3000-2500", "2500-2000")




immune.combined$feat <- immune.combined_df$feat
colors <- c("#D3D3D3", "red")
colors <- c("red", "#D3D3D3", "green", "blue")

DimPlot(immune.combined, red = "umap", group.by = "umi", pt.size = 0.1, cols = colors, order = FALSE )+theme_sc



low_umi <- colnames(immune.combined[, immune.combined$total_counts < 1700])

DimPlot(object = immune.combined, cells.highlight = low_umi, pt.size = 0.002, cols.highlight = "red", cols = "gray", order = TRUE )+theme_sc

#immune.combined_df$viral_load <- ebola_genome_percentage$`(ebola_genome_reads/total_reads) * 100`
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
```


```{r}
# viral load and UMIs
expression_matrix <- immune.combined@assays$RNA@counts
ebola_reads <- expression_matrix[rownames(expression_matrix) %in% ebola_genes, ]
ebola_genome_reads <-colSums(ebola_reads)

total_reads <- colSums(expression_matrix)
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- as.data.frame((ebola_genome_reads/total_reads)*100)

immune.combined_df$viral_load <- ebola_genome_percentage$`(ebola_genome_reads/total_reads) * 100`

immune.combined.10k <- as.data.frame(colnames(immune.combined[,immune.combined$total_counts > 10000 & immune.combined$total_counts < 15000]))
colnames(immune.combined.10k) <- "id" 
immune.combined.10k$umis <- immune.combined[,immune.combined$total_counts > 10000 & immune.combined$total_counts < 15000]$total_counts
immune.combined.10k$viral_load <- immune.combined[,immune.combined$total_counts > 10000 & immune.combined$total_counts < 15000]$viral_load
immune.combined.10k$type <- ""
immune.combined.10k[immune.combined.10k$viral_load == 0,]$type <- "0"
immune.combined.10k[immune.combined.10k$viral_load > 0 & immune.combined.10k$viral_load < 1 ,]$type <- "0-1%"
immune.combined.10k[immune.combined.10k$viral_load > 1,]$type <- ">1%"
levels(immune.combined.10k$type) <- c("0", "0-1%", ">1%")

immune.combined.10k$class <- "10-15k UMIs"


ggplot(all, aes(x=type, fill=as.factor(class))) + geom_bar()+scale_y_continuous(breaks=seq(0, 700, by=100))+labs(fill="percentage of viral reads") 


immune.combined.15k <- as.data.frame(colnames(immune.combined[,immune.combined$total_counts > 15000 & immune.combined$total_counts < 20000]))
colnames(immune.combined.15k) <- "id"
immune.combined.15k$umis <- immune.combined[,immune.combined$total_counts > 15000 & immune.combined$total_counts < 20000]$total_counts
immune.combined.15k$viral_load <- immune.combined[,immune.combined$total_counts > 15000 & immune.combined$total_counts < 20000]$viral_load
immune.combined.15k$type <- ""
immune.combined.15k[immune.combined.15k$viral_load == 0,]$type <- "0"
immune.combined.15k[immune.combined.15k$viral_load > 0 & immune.combined.15k$viral_load < 1 ,]$type <- "0-1%"
immune.combined.15k[immune.combined.15k$viral_load > 1,]$type <- ">1%"
levels(immune.combined.15k$type) <- c("0", "0-1%", ">1%")
immune.combined.15k$class <- "15-20k UMIs"


immune.combined.20k <- as.data.frame(colnames(immune.combined[,immune.combined$total_counts > 20000]))
colnames(immune.combined.20k) <- "id"
immune.combined.20k$umis <- immune.combined[,immune.combined$total_counts > 20000]$total_counts
immune.combined.20k$viral_load <- immune.combined[,immune.combined$total_counts > 20000]$viral_load
immune.combined.20k$type <- ""
immune.combined.20k[immune.combined.20k$viral_load == 0,]$type <- "0"
immune.combined.20k[immune.combined.20k$viral_load > 0 & immune.combined.20k$viral_load < 1 ,]$type <- "0-1%"
immune.combined.20k[immune.combined.20k$viral_load > 1,]$type <- ">1%"
levels(immune.combined.10k$type) <- c("0", "0-1%", ">1%")
immune.combined.20k$class <- "> 20k UMIs"


all <- rbind(immune.combined.10k, immune.combined.15k, immune.combined.20k)
ggplot(all, aes(x=class, fill=as.factor(type))) + geom_bar()+labs(fill="percentage of viral reads") 

ggplot(immune.combined.10k[immune.combined.10k$viral_load > 0,], aes(x=viral_load)) + geom_histogram(fill="#2E86C1", binwidth = 0.1)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()


ggplot(immune.combined.15k[immune.combined.15k$viral_load > 0,], aes(x=viral_load)) + geom_histogram(fill="#2E86C1", binwidth = 0.1)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()


ggplot(immune.combined.20k[immune.combined.20k$viral_load > 0,], aes(x=viral_load)) + geom_histogram(fill="#2E86C1", binwidth = 0.1)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()
```


```{r}

#viral load distribution 
get_viral_load <- function(vl){
  print(1)
vl$per <- 0
  print(1)
vl[vl$viral_load > 1,]$per <- "1-10"
  print(1)
vl[vl$viral_load > 10,]$per <- "10-20"
  print(1)
vl[vl$viral_load > 20,]$per <- "20-30"
  print(1)
vl[vl$viral_load > 30,]$per <- "30-40"
  print(1)
vl[vl$viral_load > 40,]$per <- "40-50"
  print(1)
levels(vl$per) <- c("1-10", "10-20", "20-30", "30-40")
vl <- vl[vl$per != 0,]
return(vl)
}

viral_load.10k <- get_viral_load(immune.combined.10k)
viral_load.10k$type <- "10k-15k"
colnames(viral_load.10k)[1] <- "id"
  
viral_load.15k <- get_viral_load(immune.combined.15k)
viral_load.15k$type <- "15k-20k"
colnames(viral_load.15k)[1] <- "id"
viral_load.20k <- get_viral_load(immune.combined.20k)
viral_load.20k$type <- ">20k"
colnames(viral_load.20k)[1] <- "id"
all <- rbind(viral_load.10k, viral_load.15k, viral_load.20k)

ggplot(all, aes(x=type, fill=as.factor(per))) + geom_bar()+labs(fill="percentage of viral reads") 

table(all$type, all$per)




```


#scatterplot with bins 
```{r}
immune.combined_df <- immune.combined_df[immune.combined_df$viral_load > 0 & immune.combined_df$counts < 50000 & immune.combined_df$counts > 10000 ,]

ggplot(immune.combined_df, aes(x=counts,y=viral_load)) +
  geom_point(alpha = 0.4) +
  stat_summary_bin(fun.y='mean', bins=1000,
                   color='orange', size=2, geom='point')

ggplot(immune.combined_df, aes(x=counts,y=viral_load)) +
  geom_boxplot()

immune.combined_df %>%
  
  # Add a new column called 'bin': cut the initial 'carat' in bins
  mutate( bin=cut_width(counts, width=2000) ) %>%
  
  # plot
  ggplot( aes(x=bin, y=viral_load) ) +
    geom_boxplot(fill="#69b3a2") +
    theme_minimal() +
    xlab("UMIs")+ scale_x_discrete(labels=c("10k-14k", "14k-18k", "18k-22k", "22k-26k",
                              "26k-30k", "30k-34k", "34k-38k", "38k-42k", "42k-46k", "46k-50k"))



```
   
   
# number of fetaures vs number of counts 

```{r}
immune.combined = immune.combined[, immune.combined$total_counts < 15000]

features_counts = data.frame(counts=immune.combined$total_counts, features=immune.combined$total_features_by_counts)
ggplot(features_counts, aes(x=counts,y=features))+ geom_point(alpha = 0.4) 

```




















# Plot by UMI 

```{r}
#immune.combined<- readRDS("/home/bscuser/Escritorio/cluster/02.Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10_NOVELFILTERTEST/05_RObjects/03_prep/03_immune.combined.normalized_20k.rds")
#ebola_genes <- readRDS(file.path("/Users/Maria/Desktop/cluster/02_Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds"))



# tSNE colored by high UMIs
immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$counts <- immune.combined$total_counts
immune.combined_df$umi <- "<10k"
immune.combined_df[immune.combined_df$counts > 10000,]$umi <- c("10-12k")
immune.combined_df[immune.combined_df$counts > 12000,]$umi <- c("12-15k")
#immune.combined_df[immune.combined_df$counts > 15000,]$umi <- c(">15k")
levels(immune.combined_df$umi) <- c("0", "10k-12k", "12-15k")

immune.combined$umi <- immune.combined_df$umi
colors <- c("#D3D3D3", "red", "blue", "green")

DimPlot(immune.combined, red = "umap", group.by = "umi", pt.size = 0.3, cols = colors, order = TRUE )+theme_sc

# tSNE colored by low UMIs
immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$counts <- immune.combined$total_counts
immune.combined_df$umi <- ">1700"     
immune.combined_df[immune.combined_df$counts < 1550 & immune.combined_df$counts > 1500,]$umi <- c("1500-1550")
immune.combined_df[immune.combined_df$counts < 1500 & immune.combined_df$counts > 1200,]$umi <- c("1200-1500")
immune.combined_df[immune.combined_df$counts < 1200 & immune.combined_df$counts > 1000,]$umi <- c("<1200")
levels(immune.combined_df$umi) <- c(">3000", "3000-2500", "2500-2000", "2000-1500")


immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$features <- immune.combined$total_features_by_counts
immune.combined_df$feat <- "1000-3500"     
immune.combined_df[immune.combined_df$features > 3500,]$feat <- c(">3500")
immune.combined_df[immune.combined_df$features < 900,]$feat <- c("<1000")
levels(immune.combined_df$feat)<- c("medium", "high", "low")


immune.combined_df <- as.data.frame(colnames(immune.combined))
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
immune.combined_df$counts <- immune.combined$total_counts
immune.combined_df$umi <- "1600-14000"     
immune.combined_df[immune.combined_df$counts < 1600,]$umi <- c("<1600")
immune.combined_df[immune.combined_df$counts > 13000 ,]$umi <- c(">13000")
levels(immune.combined_df$umi) <- c(">3000", "3000-2500", "2500-2000")

immune.combined$umi <- immune.combined_df$umi
levels(immune.combined$umi) <- c(">3000", "3000-2500", "2500-2000")




immune.combined$feat <- immune.combined_df$feat
colors <- c("#D3D3D3", "red")
colors <- c("red", "#D3D3D3", "green", "blue")

DimPlot(immune.combined, red = "umap", group.by = "umi", pt.size = 0.1, cols = colors, order = FALSE )+theme_sc



low_umi <- colnames(immune.combined[, immune.combined$total_counts < 1600])

DimPlot(object = immune.combined, cells.highlight = low_umi, pt.size = 0.002, cols.highlight = "red", cols = "gray", order = TRUE )+theme_sc

#immune.combined_df$viral_load <- ebola_genome_percentage$`(ebola_genome_reads/total_reads) * 100`
#immune.combined_df <- immune.combined_df[immune.combined_df$counts > 10000,]
```


```{r}
# viral load and UMIs
expression_matrix <- immune.combined@assays$RNA@counts
ebola_reads <- expression_matrix[rownames(expression_matrix) %in% ebola_genes, ]
ebola_genome_reads <-colSums(ebola_reads)

total_reads <- colSums(expression_matrix)
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- as.data.frame((ebola_genome_reads/total_reads)*100)

immune.combined_df$viral_load <- ebola_genome_percentage$`(ebola_genome_reads/total_reads) * 100`

immune.combined.10k <- as.data.frame(colnames(immune.combined[,immune.combined$total_counts > 10000 & immune.combined$total_counts < 15000]))
colnames(immune.combined.10k) <- "id" 
immune.combined.10k$umis <- immune.combined[,immune.combined$total_counts > 10000 & immune.combined$total_counts < 15000]$total_counts
immune.combined.10k$viral_load <- immune.combined[,immune.combined$total_counts > 10000 & immune.combined$total_counts < 15000]$viral_load
immune.combined.10k$type <- ""
immune.combined.10k[immune.combined.10k$viral_load == 0,]$type <- "0"
immune.combined.10k[immune.combined.10k$viral_load > 0 & immune.combined.10k$viral_load < 1 ,]$type <- "0-1%"
immune.combined.10k[immune.combined.10k$viral_load > 1,]$type <- ">1%"
levels(immune.combined.10k$type) <- c("0", "0-1%", ">1%")

immune.combined.10k$class <- "10-15k UMIs"


ggplot(all, aes(x=type, fill=as.factor(class))) + geom_bar()+scale_y_continuous(breaks=seq(0, 700, by=100))+labs(fill="percentage of viral reads") 


immune.combined.15k <- as.data.frame(colnames(immune.combined[,immune.combined$total_counts > 15000 & immune.combined$total_counts < 20000]))
colnames(immune.combined.15k) <- "id"
immune.combined.15k$umis <- immune.combined[,immune.combined$total_counts > 15000 & immune.combined$total_counts < 20000]$total_counts
immune.combined.15k$viral_load <- immune.combined[,immune.combined$total_counts > 15000 & immune.combined$total_counts < 20000]$viral_load
immune.combined.15k$type <- ""
immune.combined.15k[immune.combined.15k$viral_load == 0,]$type <- "0"
immune.combined.15k[immune.combined.15k$viral_load > 0 & immune.combined.15k$viral_load < 1 ,]$type <- "0-1%"
immune.combined.15k[immune.combined.15k$viral_load > 1,]$type <- ">1%"
levels(immune.combined.15k$type) <- c("0", "0-1%", ">1%")
immune.combined.15k$class <- "15-20k UMIs"


immune.combined.20k <- as.data.frame(colnames(immune.combined[,immune.combined$total_counts > 20000]))
colnames(immune.combined.20k) <- "id"
immune.combined.20k$umis <- immune.combined[,immune.combined$total_counts > 20000]$total_counts
immune.combined.20k$viral_load <- immune.combined[,immune.combined$total_counts > 20000]$viral_load
immune.combined.20k$type <- ""
immune.combined.20k[immune.combined.20k$viral_load == 0,]$type <- "0"
immune.combined.20k[immune.combined.20k$viral_load > 0 & immune.combined.20k$viral_load < 1 ,]$type <- "0-1%"
immune.combined.20k[immune.combined.20k$viral_load > 1,]$type <- ">1%"
levels(immune.combined.10k$type) <- c("0", "0-1%", ">1%")
immune.combined.20k$class <- "> 20k UMIs"


all <- rbind(immune.combined.10k, immune.combined.15k, immune.combined.20k)
ggplot(all, aes(x=class, fill=as.factor(type))) + geom_bar()+labs(fill="percentage of viral reads") 

ggplot(immune.combined.10k[immune.combined.10k$viral_load > 0,], aes(x=viral_load)) + geom_histogram(fill="#2E86C1", binwidth = 0.1)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()


ggplot(immune.combined.15k[immune.combined.15k$viral_load > 0,], aes(x=viral_load)) + geom_histogram(fill="#2E86C1", binwidth = 0.1)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()


ggplot(immune.combined.20k[immune.combined.20k$viral_load > 0,], aes(x=viral_load)) + geom_histogram(fill="#2E86C1", binwidth = 0.1)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()
```


```{r}

#viral load distribution 
get_viral_load <- function(vl){
  print(1)
vl$per <- 0
  print(1)
vl[vl$viral_load > 1,]$per <- "1-10"
  print(1)
vl[vl$viral_load > 10,]$per <- "10-20"
  print(1)
vl[vl$viral_load > 20,]$per <- "20-30"
  print(1)
vl[vl$viral_load > 30,]$per <- "30-40"
  print(1)
vl[vl$viral_load > 40,]$per <- "40-50"
  print(1)
levels(vl$per) <- c("1-10", "10-20", "20-30", "30-40")
vl <- vl[vl$per != 0,]
return(vl)
}

viral_load.10k <- get_viral_load(immune.combined.10k)
viral_load.10k$type <- "10k-15k"
colnames(viral_load.10k)[1] <- "id"
  
viral_load.15k <- get_viral_load(immune.combined.15k)
viral_load.15k$type <- "15k-20k"
colnames(viral_load.15k)[1] <- "id"
viral_load.20k <- get_viral_load(immune.combined.20k)
viral_load.20k$type <- ">20k"
colnames(viral_load.20k)[1] <- "id"
all <- rbind(viral_load.10k, viral_load.15k, viral_load.20k)

ggplot(all, aes(x=type, fill=as.factor(per))) + geom_bar()+labs(fill="percentage of viral reads") 

table(all$type, all$per)




```


#scatterplot with bins 
```{r}
immune.combined_df <- immune.combined_df[immune.combined_df$viral_load > 0 & immune.combined_df$counts < 50000 & immune.combined_df$counts > 10000 ,]

ggplot(immune.combined_df, aes(x=counts,y=viral_load)) +
  geom_point(alpha = 0.4) +
  stat_summary_bin(fun.y='mean', bins=1000,
                   color='orange', size=2, geom='point')

ggplot(immune.combined_df, aes(x=counts,y=viral_load)) +
  geom_boxplot()

immune.combined_df %>%
  
  # Add a new column called 'bin': cut the initial 'carat' in bins
  mutate( bin=cut_width(counts, width=2000) ) %>%
  
  # plot
  ggplot( aes(x=bin, y=viral_load) ) +
    geom_boxplot(fill="#69b3a2") +
    theme_minimal() +
    xlab("UMIs")+ scale_x_discrete(labels=c("10k-14k", "14k-18k", "18k-22k", "22k-26k",
                              "26k-30k", "30k-34k", "34k-38k", "38k-42k", "42k-46k", "46k-50k"))



```
   
   
# number of fetaures vs number of counts 

```{r}
immune.combined = immune.combined[, immune.combined$total_counts < 15000]

features_counts = data.frame(counts=immune.combined$total_counts, features=immune.combined$total_features_by_counts)
ggplot(features_counts, aes(x=counts,y=features))+ geom_point(alpha = 0.4) 

```


# -------------------------------------
#    Visualize marker genes 
# -------------------------------------

```{r Find markers}
#immune.combined <- readRDS(file.path(data_path,"/05_RObjects/03_prep/03_immune.combined.ready.rds"))

# Visualize Markers
DotPlot(immune.combined, features = unique(marker.genes_red), cols = c("navy", "#FFD533", "grey"), dot.scale = 12) + RotatedAxis()+theme(axis.text = element_text(size = 20), axis.title = element_blank())+ theme(panel.background = element_rect(fill = "white", colour = "black"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))


pal_celltypes <- brewer.pal(4, "Set2")
pal_celltypes <-wes_palette("GrandBudapest1", 4)
pal_celltypes[3] <- "#AE4E4E"
pal <- pal_celltypes[c(2,1,4,3)]
pal[3] <- (brewer.pal(6,"Paired"))[2]
# Plot with correct
DimPlot(immune.combined, reduction = "umap", label = TRUE, cols =pal, label.size = 8)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))+theme(text = element_text(size=18))

```

```{r UMAPvisualization}

theme_sc <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "black"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Day post infection 
DimPlot(immune.combined, red = "umap", group.by = "dpi", pt.size = 0.1, cols = rev(brewer.pal(8, "Paired")))+theme_sc

# Individual 
DimPlot(immune.combined, reduction = "umap", group.by = "individual_nhp",cols = c(brewer.pal(10, "Paired")[c(1,2)]), pt.size = 0.1)+theme_sc

pal4 <- c(wes_palette("Zissou1", 5)[3],"#E55039",wes_palette("Zissou1", 5)[1] )
p4 <- DimPlot(immune.combined, reduction = "umap", group.by = "cond",cols =pal4, pt.size = 0.01)+theme_sc
p4

cc.genes



```



# -------------------------------------
#    Extract lncRNA and mRNA
# -------------------------------------


```{r cars}
# ------------------------------------
#     PC genes present in object 
# -------------------------------------
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% unique(ref$gene_name[ref$gene_biotype == "protein_coding"])]

# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# Annotated lncrnas id 
lnc_ref_id <- gsub("_","-",unique(ref[!is.na(ref$gene_biotype) & ref$gene_biotype == "lncRNA",]$gene_name))
unique(ref[ref$gene_name == lnc_ref_id[substr(lnc_ref_id,1,3) != "ENS"],]$gene_biotype)

# Annotated lncRNAs: remove the gene names so it can match 
annotated_lncrnas <- rownames(immune.combined)[rownames(immune.combined) %in% lnc_ref_id]
# Novel lncRNAs
novel_lncrnas <- rownames(immune.combined)[unlist(lapply(rownames(immune.combined), function(x) startsWith(x, "MSTRG")))]
# Complete Set 
all_lncrnas <- c(annotated_lncrnas, novel_lncrnas)

#------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated", "novel")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(annotated_lncrnas)), length(unique(novel_lncrnas))))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Dark2")
lnc_donut <- df.donut[df.donut$type == "lncRNAs",]
names(lnc_donut) <- c("-", "type", "count")
donut_plot(lnc_donut, "Greens")

# Save
robjectsdir<- file.path(data_path,"05_RObjects/05_stats")
dir.create(file.path(robjectsdir), showWarnings = TRUE)
saveRDS(all_lncrnas, file.path(robjectsdir, "all_lncrnas_4filt_1000.rds"))
saveRDS(annotated_mrnas, file.path(robjectsdir, "annotated_mrnas_4filt_1000.rds"))
```




############################################## later 

# -------------------------------------
# Preprocess and cluster identification
# -------------------------------------
```{r}
immune.combined <- readRDS(file.path(data_path, "/05_RObjects/03_prep/03_immune.combined.noebola.NOVELAPPROACH.rds"))
immune.combined <- NormalizeData(immune.combined)

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
immune.combined <- CellCycleScoring(immune.combined, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

pca <- DimPlot(immune.combined,
        reduction = "pca",
        group.by= "Phase")

#immune.combined <- ScaleData(immune.combined, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(immune.combined))
immune.combined <- ScaleData(immune.combined)


```



# -------------------------------------
#     Remove cycle-cell genes
# -------------------------------------



```{r}
immune.combined <- readRDS(file.path(data_path, "/05_RObjects/03_prep/03_immune.combined.normalized_noebola_NOVELAPPROACH.rds"))

immune.combined <- FindVariableFeatures(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)
immune.combined <- FindNeighbors(immune.combined, reduction = "pca",dims = 1:20 )
immune.combined <- FindClusters(immune.combined, resolution = 0.2)

#saveRDS(immune.combined, file.path(data_path,"05_RObjects/03_prep/03_immune.combined.normalized_noebola_NOVELAPPROACH.rds"))


DimPlot(immune.combined, reduction = "umap", label = TRUE)+theme_sc


```




# -------------------------------------
#     Add viral load
# -------------------------------------

```{r}
expression_matrix <- immune.combined@assays$RNA@counts

ebola_reads <- expression_matrix[rownames(expression_matrix) %in% ebola_genes, ]
ebola_genome_reads <-colSums(ebola_reads)

total_reads <- colSums2(as.matrix(expression_matrix))
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- (ebola_genome_reads/total_reads)*100

immune.combined$viral_load <- ebola_genome_percentage
```


# Remove ribosomal genes and HBB 

```{r}

RPS.genes <- grep(pattern = "^RPS", x = rownames(x = immune.combined), value = TRUE)
RPL.genes <- grep(pattern = "^RPL", x = rownames(x = immune.combined), value = TRUE)


percent.RPS <- Matrix::colSums(immune.combined@assays$RNA@counts[RPS.genes, ])/Matrix::colSums(immune.combined@assays$RNA@counts)
percent.RPL <- Matric::colSums(immune.combined@assays$RNA@counts[RPL.genes, ])/Matrix::colSums(immune.combined@assays$RNA@counts)
immune.combined <- AddMetaData(object = immune.combined, metadata = percent.RPS, col.name = "percent.RPS")
immune.combined <- AddMetaData(object = immune.combined, metadata = percent.RPL, col.name = "percent.RPL")





immune.combined <- immune.combined[!rownames(immune.combined) %in% RPS.genes,]
immune.combined <- immune.combined[!rownames(immune.combined) %in% RPL.genes,]



counts <- GetAssayData(immune.combined, assay = "RNA")
counts <- counts[-(which(rownames(counts) %in% c('HBB'))),]
immune.combined <- subset(immune.combined, features = rownames(counts))


saveRDS(immune.combined, file.path(data_path,"/05_RObjects/03_prep/03_immune.combined_metadata.ordered_noribo.rds"))

```










#significant genes vs number of cells

```{r}
df_counts <- readRDS(file.path(data_path, "/05_RObjects/06_correlation/03_viralload_infected_late/counts_df_monocytes.rds"))
#correlations_exvivo_noebola <- readRDS(file.path("/Users/Maria/Desktop/cluster/02_Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10_NOVELFILTERTEST/05_RObjects/06_correlation/03_viral_load_infected_late/mono_infected_late_ebolaremoved_correlation_significant_all.filt.rds"))

#genes_sig <- correlations_exvivo_noebola$gene
#immune.combined_corr <- immune.combined[rownames(immune.combined) %in% genes_sig,]
#expression_matrix <- as.data.frame(immune.combined_corr@assays$RNA@counts)
#counts <- apply(expression_matrix, 1, function(c)sum(c!=0))
#counts_df <- as.data.frame(counts)
#counts_df$rho <- correlations_exvivo_noebola$rho

names(df_counts)[1] <- "num_cells"
genes_h <-  c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
highlight_df <- df_counts[rownames(df_counts) %in% genes_h,]

lnc <- df_counts[rownames(df_counts) %in% sig_cor_lnc_exvivo_noebola$gene,]
ggplot(df_counts, aes(x=rho, y=num_cells)) + geom_point()+ geom_point(data=highlight_df,  color='red')
```


```{r}
cellsubset_F13A1 <- cellsubset["F13A1",]
F13A1 <- as.data.frame(cellsubset_F13A1$nCount_RNA)
F13A1$viral_load <- cellsubset_F13A1$viraload
F13A1$counts <- cellsubset_F13A1$nCount_RNA


# negative rho and low number of cells 
B2M <- readRDS(file.path(data_path,"/05_RObjects/06_correlation/03_viralload_infected_late/neg_corr.rds" ))
names(B2M)[3] <- "expression"
B2M <- B2M[B2M$viral_load > 0.09,]
ggplot(B2M, aes(x=expression, y=viral_load)) + geom_point() + ggtitle("B2M")


# high rho and low number of cells 
F13A1 <- readRDS(file.path(data_path,"/05_RObjects/06_correlation/03_viralload_infected_late/pos_corr.rds" ))
names(F13A1)[3] <- "expression"
F13A1 <- F13A1[F13A1$viral_load > 0.9,]
ggplot(F13A1, aes(x=expression, y=viral_load)) + geom_point() + ggtitle("F13A1")


#high rho and high number of cells 

FTH1  <- readRDS(file.path(data_path,"/05_RObjects/06_correlation/03_viral_load_infected_late/fth1.rds"))
names(FTH1)[3] <- "expression"
FTH1 <- FTH1[FTH1$viral_load > 0.9,]
ggplot(FTH1, aes(x=expression, y=viral_load)) + geom_point() + ggtitle("FTH1")


```






## Volcano plot (analyze ISG)
```{r}
EnhancedVolcano(res,
  lab = rownames(res),
  x = 'log2FoldChange',
  y = 'pvalue')


```









