---
title: "Broad test"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Seurat); library(ggplot2); library(readr);
library(Matrix);library(magrittr)
data_path <- "/home/bscuser/Escritorio/cluster/02.Ebola_sc/data"
cell_ids <- read_tsv(file = '/home/bscuser/Descargas/GSE158440_Kotliar2020_ExVivo_Seqwell_UMI_counts_barcodes.tsv.gz', col_names = FALSE)
#gene_ids <-read.table(file = '/home/bscuser/Descargas/GSE158440_Kotliar2020_ExVivo_Seqwell_UMI_counts_features.tsv.gz')
#expression_matrix <- readMM("/home/bscuser/Descargas/GSE158440_Kotliar2020_ExVivo_Seqwell_UMI_counts_expression.mtx.gz")
#expression_matrix <- t(expression_matrix)
#colnames(expression_matrix) <- cell_ids$X1
#rownames(expression_matrix) <- gene_ids$V1
#saveRDS(expression_matrix, "/home/bscuser/Escritorio/cluster/02.Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/expression_matrix.rds")


#Load broad data 
expression_matrix_broad <- readRDS(file.path(data_path,"/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/expression_matrix.rds"))
colnames(expression_matrix_broad) <- strsplit(as.character(colnames(expression_matrix_broad)), "_") %>% sapply(extract2, 2)
rownames(expression_matrix_broad) <- strsplit(as.character(rownames(expression_matrix_broad)), "__") %>% sapply(extract2, 2)

ebola_genes_broad <- c("EBOV-GP", "EBOV-L", "EBOV-NP","EBOV-VP24_EBOV-VP24", "EBOV-VP30","EBOV-VP35", "EBOV-VP40" )

# Per cell, check the % of ebola reads ( not normalized - just calculate the % of raw counts )
ebola_reads_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% ebola_genes_broad, ]
ebola_genome_reads_broad <-colSums(ebola_reads_broad)

total_reads_broad <- colSums(expression_matrix_broad)
ebola_gene_percentage_brad <- (ebola_reads_broad/total_reads_broad)*100
ebola_genome_percentage_broad<- (ebola_genome_reads_broad/total_reads_broad)*100

viral_reads_broad <-  data.frame(barcode = colnames(expression_matrix_broad), 
                                 percentage_viral_reads = ebola_genome_percentage_broad)


infected_broad <- viral_reads_broad[viral_reads_broad$percentage_viral_reads > 0, ]

```


# Check viral load our dataset 
```{r}
immune.combined <- readRDS(file.path(data_path, "/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")) 
immune.combined$barcode <- strsplit(as.character(colnames(immune.combined)), "_") %>% sapply(extract2, 2)
expression_matrix <- immune.combined@assays$RNA@counts
colnames(expression_matrix) <- strsplit(as.character(colnames(expression_matrix)), "_") %>% sapply(extract2, 2)

ebola_genes <- readRDS(file.path("/home/bscuser/Escritorio/cluster/02.Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds"))
ebola_reads <- expression_matrix[rownames(expression_matrix) %in% ebola_genes, ]
ebola_genome_reads <-colSums(ebola_reads)

total_reads <- colSums(expression_matrix)
ebola_gene_percentage <- (ebola_reads/total_reads)*100
ebola_genome_percentage<- (ebola_genome_reads/total_reads)*100

viral_reads <-  data.frame(barcode = colnames(expression_matrix), 
                        percentage_viral_reads = ebola_genome_percentage)

infected <- viral_reads[viral_reads$percentage_viral_reads > 0, ]
```



```{r}
#merge both datasets
viral_reads_broad_tomerge <- viral_reads_broad[viral_reads_broad$barcode %in% viral_reads$barcode,]
viral_reads_tomerge <- viral_reads[viral_reads$barcode %in% viral_reads_broad$barcode,]


merged <- merge(viral_reads_broad_tomerge, viral_reads_tomerge, by="barcode")
merged <- merged[merged$percentage_viral_reads.x > 0 & merged$percentage_viral_reads.y > 0,]
```


# Plot results 

```{r}

ggplot(infected_broad, aes(x=percentage_viral_reads)) + geom_histogram(fill="#2E86C1", binwidth = 0.03)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()
  #xlim(c(0,50))+ylim(c(0,15))

ggplot(infected, aes(x=percentage_viral_reads)) + geom_histogram(fill="#2E86C1", binwidth = 0.03)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()+ ggtitle("Our data")

ggplot(viral_reads_broad_tomerge, aes(x=percentage_viral_reads)) + geom_histogram(fill="#2E86C1", binwidth = 0.03)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()+ggtitle("Our data")

ggplot(viral_reads_tomerge, aes(x=percentage_viral_reads)) + geom_histogram(fill="#2E86C1", binwidth = 0.03)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()+ggtitle("Broad data")


```


# Viral correlation of TOP 5 genes 
```{r}
##### Our data 
immune.combined <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")
immune.combined$barcode <- strsplit(as.character(colnames(immune.combined)), "_") %>% sapply(extract2, 2)

#Select monocytes-late infection 
cellsubset <- subset(immune.combined, ident = "Monocyte")
# 2. Extract Infected cells 
cellsubset <- cellsubset[,cellsubset$infection == "Infected"]
# 3. Retain  late only 
cellsubset <- cellsubset[,cellsubset$cond == "live"]
cellsubset <- cellsubset[,cellsubset$dpi == "H024"]

barcode <- as.data.frame(cellsubset$barcode)
#cellsubset <- NormalizeData(cellsubset)


# remove ebola reads
ebola_genes <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds")

cellsubset <- cellsubset[setdiff(rownames(cellsubset), ebola_genes),]
#cellsubset <- NormalizeData(cellsubset)

list_genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")

calc_correlation_viralload <- function(cellsubset,viral_load,gene1){

  exp <- cellsubset[gene1,]@assays$RNA@data #c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
  mask <- as.vector(exp) > 0 
  #mask <- as.vector(exp)
  vir <- viral_load
  pc <- cor(as.vector(exp[mask]), as.vector(vir)[mask], method = c("spearman"))
  df <- data.frame(rho=pc)
  df$gene <- gene1
   return(df)
}

correlations <-lapply(list_genes, function(gene) calc_correlation_viralload(cellsubset, cellsubset$viral_load, gene))
correlations_df <- do.call(rbind, correlations)


#exp <- cellsubset["DYNLL1",]@assays$RNA@data  #c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
#mask <- as.vector(exp) > 0 
#vir <- cellsubset$viral_load
#cor(as.vector(exp[mask]), as.vector(vir)[mask], method = c("spearman"))


```




```{r}
##### Broad data 
#substact cells that pass the filter 
expression_matrix_broad <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/expression_matrix.rds")
colnames(expression_matrix_broad) <- strsplit(as.character(colnames(expression_matrix_broad)), "_") %>% sapply(extract2, 2)
rownames(expression_matrix_broad) <- strsplit(as.character(rownames(expression_matrix_broad)), "__") %>% sapply(extract2, 2)

#compute the viral load again 
ebola_reads_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% ebola_genes_broad, ]
ebola_genome_reads_broad <-colSums(ebola_reads_broad)

total_reads_broad <- colSums(expression_matrix_broad)
ebola_gene_percentage_brad <- (ebola_reads_broad/total_reads_broad)*100
ebola_genome_percentage_broad<- (ebola_genome_reads_broad/total_reads_broad)*100

viral_reads_broad <-  data.frame(barcode = colnames(expression_matrix_broad), 
                                 percentage_viral_reads = ebola_genome_percentage_broad)


expression_matrix_broad <- expression_matrix_broad[,colnames(expression_matrix_broad) %in% barcode$`cellsubset$barcode`]
expression_matrix_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% rownames(cellsubset),]

viral_reads_broad <- viral_reads_broad[barcode %in% expression_matrix_broad,]

#remove ebola reads 
expression_matrix_broad <- expression_matrix_broad[!rownames(expression_matrix_broad) %in% ebola_genes_broad,]
expression_matrix_broad <- NormalizeData(expression_matrix_broad)

exp <- expression_matrix_broad["DYNLL1",] #c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
mask <- as.vector(exp) > 0 
vir <- viral_reads_broad$percentage_viral_reads
cor(as.vector(exp[mask]), as.vector(vir)[mask], method = c("spearman"))


```


# Viral correlation of TOP 5 genes 
```{r}
##### Our data 
immune.combined <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/immune.combined.infectionstatus.rds")
immune.combined$barcode <- strsplit(as.character(colnames(immune.combined)), "_") %>% sapply(extract2, 2)

#Select monocytes-late infection 
cellsubset <- subset(immune.combined, ident = "Monocyte")
# 2. Extract Infected cells 
cellsubset <- cellsubset[,cellsubset$infection == "Infected"]
# 3. Retain  late only 
cellsubset <- cellsubset[,cellsubset$cond == "live"]
cellsubset <- cellsubset[,cellsubset$dpi == "H024"]

barcode <- as.data.frame(cellsubset$barcode)
#cellsubset <- NormalizeData(cellsubset)


# remove ebola reads
ebola_genes <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/05_stats/ebola_genes.rds")

cellsubset <- cellsubset[setdiff(rownames(cellsubset), ebola_genes),]
cellsubset <- NormalizeData(cellsubset)

list_genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")

calc_correlation_viralload <- function(cellsubset,viral_load,gene1){

  exp <- cellsubset[gene1,]@assays$RNA@data #c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
  mask <- rep(TRUE, length(gene1))
  vir <- log10(viral_load)
  pc <- cor(as.vector(exp[mask]), as.vector(vir)[mask], method = c("spearman"))
  df <- data.frame(rho=pc)
  df$gene <- gene1
   return(df)
}

correlations <-lapply(list_genes, function(gene) calc_correlation_viralload(cellsubset, cellsubset$viral_load, gene))
correlations_df <- do.call(rbind, correlations)

```




```{r}
##### Broad data subsetting as our dataset 
#substact cells that pass the filter 
expression_matrix_broad <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/expression_matrix.rds")
colnames(expression_matrix_broad) <- strsplit(as.character(colnames(expression_matrix_broad)), "_") %>% sapply(extract2, 2)
rownames(expression_matrix_broad) <- strsplit(as.character(rownames(expression_matrix_broad)), "__") %>% sapply(extract2, 2)
ebola_genes_broad <- c("EBOV-GP", "EBOV-L", "EBOV-NP","EBOV-VP24_EBOV-VP24", "EBOV-VP30","EBOV-VP35", "EBOV-VP40" )


#compute the viral load again 
ebola_reads_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% ebola_genes_broad, ]
ebola_genome_reads_broad <-colSums(ebola_reads_broad)

total_reads_broad <- colSums(expression_matrix_broad)
ebola_gene_percentage_brad <- (ebola_reads_broad/total_reads_broad)*100
ebola_genome_percentage_broad<- (ebola_genome_reads_broad/total_reads_broad)*100

viral_reads_broad <-  data.frame(barcode = colnames(expression_matrix_broad), 
                                 percentage_viral_reads = ebola_genome_percentage_broad)


expression_matrix_broad <- expression_matrix_broad[,colnames(expression_matrix_broad) %in% barcode$`cellsubset$barcode`]
expression_matrix_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% rownames(cellsubset),]

viral_reads_broad <- viral_reads_broad[viral_reads_broad$barcode %in% colnames(expression_matrix_broad),]

#remove ebola reads 
expression_matrix_broad <- expression_matrix_broad[!rownames(expression_matrix_broad) %in% ebola_genes_broad,]
expression_matrix_broad <- NormalizeData(expression_matrix_broad)

list_genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")

calc_correlation_viralload <- function(cellsubset,viral_load,gene1){

  exp <- cellsubset[gene1,] #c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
  mask <- as.vector(exp) > 0 
  vir <- log10(viral_load)
  pc <- cor(as.vector(exp[mask]), as.vector(vir)[mask], method = c("spearman"))
  df <- data.frame(rho=pc)
  df$gene <- gene1
   return(df)
}

correlations_broad <-lapply(list_genes, function(gene) calc_correlation_viralload(expression_matrix_broad, viral_reads_broad$percentage_viral_reads, gene))
correlations_df_broad <- do.call(rbind, correlations_broad)


```


```{r}
# Broad data selecting all cells 
expression_matrix_broad <- readRDS("/gpfs/projects/bsc83/Data/Ebola/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/expression_matrix.rds")
#expression_matrix_broad <- NormalizeData(expression_matrix_broad)
colnames(expression_matrix_broad) <- strsplit(as.character(colnames(expression_matrix_broad)), "_") %>% sapply(extract2, 2)
rownames(expression_matrix_broad) <- strsplit(as.character(rownames(expression_matrix_broad)), "__") %>% sapply(extract2, 2)
ebola_genes_broad <- c("EBOV-GP", "EBOV-L", "EBOV-NP","EBOV-VP24_EBOV-VP24", "EBOV-VP30","EBOV-VP35", "EBOV-VP40" )

ebola_reads_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% ebola_genes_broad, ]
ebola_genome_reads_broad <-colSums(ebola_reads_broad)


infected_cells <- expression_matrix_broad[,colSums(ebola_reads_broad) > 15]
expression_matrix_broad <- infected_cells
#expression_matrix_broad <- NormalizeData(expression_matrix_broad)


ebola_reads_broad <- expression_matrix_broad[rownames(expression_matrix_broad) %in% ebola_genes_broad, ]
ebola_genome_reads_broad <-colSums(ebola_reads_broad)
total_reads_broad <- colSums(expression_matrix_broad)
ebola_gene_percentage_broad <- (ebola_reads_broad/total_reads_broad)*100
ebola_genome_percentage_broad<- (ebola_genome_reads_broad/total_reads_broad)*100

viral_reads_broad <-  data.frame(barcode = colnames(expression_matrix_broad), 
                                 percentage_viral_reads = ebola_genome_percentage_broad)
#remove ebola reads 
expression_matrix_broad <- expression_matrix_broad[!rownames(expression_matrix_broad) %in% ebola_genes_broad,]
expression_matrix_broad <- NormalizeData(expression_matrix_broad)


list_genes <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")

calc_correlation_viralload <- function(cellsubset,viral_load,gene1){

  exp <- cellsubset[gene1,] #c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
  mask <- rep(TRUE, length(gene1))
  vir <- log10(viral_load)
  pc <- cor(as.vector(exp[mask]), as.vector(vir)[mask], method = c("spearman"))
  df <- data.frame(rho=pc)
  df$gene <- gene1
   return(df)
}

correlations_broad <-lapply(list_genes, function(gene) calc_correlation_viralload(expression_matrix_broad, viral_reads_broad$percentage_viral_reads, gene))
correlations_df_broad <- do.call(rbind, correlations_broad)


```


# load dataframe of correlation

```{r}

correlation_df <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/top5_correlation.rds"))
correlation_df_broad <- readRDS(file.path(data_path,"02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/top5_correlation_broad_infected.rds"))

```

```{r}

different_correlations <- readRDS("//Users/Maria/Desktop/cluster/02_Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/99_Test_BroadData/rho_df.rds")

different_correlations <- as.data.frame(different_correlations)
levels(different_correlations$gene) <- c("STAT1", "MX1", "MX2", "HSPA5", "DYNLL1")
different_correlations_pos <- different_correlations[different_correlations$gene == "DYNLL1" | different_correlations$gene == "HSPA5", ]
different_correlations_neg<- different_correlations[different_correlations$gene == "STAT1" | different_correlations$gene == "MX1" | different_correlations$gene == "MX2", ]


ggplot(different_correlations_pos, aes(x=thresh, y = rho, fill = gene, col=gene)) + geom_point()+theme_minimal()+geom_smooth(method = "lm")

ggplot(different_correlations_neg, aes(x=thresh, y = rho, fill = gene, col=gene)) + geom_point()+theme_minimal()+geom_smooth(method = "lm")

ggplot(different_correlations, aes(x=thresh, y = rho, fill = gene, col=gene)) + geom_point()+theme_minimal()+geom_smooth(method = "lm")


```






############################################ OLD ANALYSIS 

```{r}
# Gene counts in broad 
colnames(expression_matrix) <- strsplit(as.character(colnames(expression_matrix)), "_") %>% sapply(extract2, 2)
rownames(expression_matrix) <- strsplit(as.character(rownames(expression_matrix)), "__") %>% sapply(extract2, 2)
counts_broad <- as.data.frame(rowSums(expression_matrix))
counts_broad <- cbind(rownames(expression_matrix), counts_broad)


#gene counts in our data 
immune.combined$barcode <- strsplit(as.character(colnames(immune.combined)), "_") %>% sapply(extract2, 2)
barcode <- as.data.frame(immune.combined$barcode)
expression_matrix_data <- immune.combined.our@assays$RNA@counts
colnames(expression_matrix_data) <- strsplit(as.character(colnames(expression_matrix_data)), "_") %>% sapply(extract2, 2)
counts <- as.data.frame(rowSums(expression_matrix_data))
counts <- cbind(rownames(expression_matrix_data), counts)


counts_broad <- counts_broad[counts_broad$`colnames(expression_matrix)` %in% counts$`colnames(expression_matrix_data)`,]
names <- as.data.frame(counts_broad$`colnames(expression_matrix)`[counts$`colnames(expression_matrix_data),`])


counts <- counts[counts$`colnames(expression_matrix_data)` %in% counts_broad$`colnames(expression_matrix)`,]


rownames(expression_matrix_data) %in% ebola_genes

genes_excluded <- counts_broad[!counts_broad$`rownames(expression_matrix)` %in% counts$`rownames(expression_matrix_data)`,]


"EBOV-VP30" %in% rownames(expression_matrix_data)

```




#compare with our data 

```{r}
ebola_genome_percentage_df <- readRDS(file.path(data_path, "02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10/05_RObjects/03_prep/df_viralpercentage.rds"))

# extract the barcodes 
ebola_genome_percentage_df_broad$barcode <- strsplit(as.character(ebola_genome_percentage_df_broad$id), "_") %>% sapply(extract2, 2)
ebola_genome_percentage_df$barcode <- strsplit(as.character(rownames(ebola_genome_percentage_df)), "_") %>% sapply(extract2, 2)

#select the samples with > 40% of viral reads to check if they match 
higher_20 <- ebola_genome_percentage_df[ebola_genome_percentage_df$percentage_viral_reads > 20,]
length(higher_20$barcode) # 35 samples
higher20_broad <- ebola_genome_percentage_df_broad[ebola_genome_percentage_df_broad$percentage_viral_reads > 20,]
length(higher20_broad$barcode) # 166 samples 

higher_20$barcode %in% higher20_broad$barcode
h.20 <- merge (higher_20, higher20_broad, by = "barcode")
not_in_our <- higher20_broad[!higher20_broad$barcode %in% h.20,]


#all our dataset > 1 

all <- ebola_genome_percentage_df[as.numeric(ebola_genome_percentage_df$percentage_viral_reads) > 1, ]
all_broad <- ebola_genome_percentage_df_broad[ebola_genome_percentage_df_broad$percentage_viral_reads > 1,]

length(all$barcode) #689
all_merge <- merge(all, all_b, by = "barcode")

excluded <- all_b[!all_b$barcode %in% all_merge,]
excluded <- excluded[excluded$percentage_viral_reads > 5,]

ggplot(excluded, aes(x=percentage_viral_reads)) + geom_histogram(fill="#2E86C1", binwidth = 0.03)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()

```


# Comparison broad with the novel data without upper UMI filter 
```{r}
ebola_genome_percentage_df_new <- readRDS("/home/bscuser/Escritorio/cluster/02.Ebola_sc/data/02_scRNA-Seq_PBMCs/00_scRNA-Seq_exVivo_rhemac10_NOVELFILTERTEST/05_RObjects/03_prep/df_viralpercentage.rds")
# extract the barcodes 
ebola_genome_percentage_df_new$barcode <- strsplit(as.character(rownames(ebola_genome_percentage_df_new)), "_") %>% sapply(extract2, 2)

ggplot(ebola_genome_percentage_df, aes(x=percentage_viral_reads)) + geom_histogram(fill="#2E86C1", binwidth = 0.03)+ylab("# Cells")+xlab("Viral Load")+scale_x_log10()

#viral reads > 1 
all_new <- ebola_genome_percentage_df_new[as.numeric(ebola_genome_percentage_df_new$percentage_viral_reads) > 1, ]

length(all_new$barcode)

all_merge_new <- merge(all_new, all_b, by = "barcode")

length(all_merge$percentage_viral_reads.x > 30)
length(all_merge_new$percentage_viral_reads.x > 30) #1059
```



#check if the TOP 50 cells match 

```{r}
top10_old <- head(ebola_genome_percentage_df[order(ebola_genome_percentage_df$percentage_viral_reads, decreasing = TRUE),], 15)
top10_new <- head(ebola_genome_percentage_df_new[order(ebola_genome_percentage_df_new$percentage_viral_reads, decreasing = TRUE),], 31)
top10_broad <- head(ebola_genome_percentage_df_broad[order(ebola_genome_percentage_df_broad$percentage_viral_reads, decreasing = TRUE),], 50)

top10_new$type <- "new"
top10_broad$type <- "broad"
top10_old$type <- "original"

top10_new <- select(top10_new, percentage_viral_reads, barcode, type)
top10_broad <- select(top10_broad, percentage_viral_reads, barcode, type)
top10_old <- select(top10_old, percentage_viral_reads, barcode, type)
dataset <- rbind( top10_new,top10_broad, top10_old)

#dataset <-cbind(top10_broad, top10_new)

ggplot(
  dataset, 
  aes(
    y = as.numeric(percentage_viral_reads), 
    x = reorder(barcode, -percentage_viral_reads))) + geom_point() + 
facet_wrap(~type, ncol = 1)+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 



```


```{r}

excluded

get_viral_load <- function(vl){
vl$per <- 0
vl[vl$percentage_viral_reads > 1,]$per <- "1-10"
vl[vl$percentage_viral_read > 10,]$per <- "10-20"
vl[vl$percentage_viral_read > 20,]$per <- "20-30"
vl[vl$percentage_viral_read > 30,]$per <- "30-40"
vl[vl$percentage_viral_read > 40,]$per <- "40-50"
levels(vl$per) <- c("1-10", "10-20", "20-30", "30-40")
vl <- vl[vl$per != 0,]
return(vl)
}

viral_load.p <- get_viral_load(ebola_genome_percentage_df)
viral_load.p$type <- "used"
      

viral_load.discarted.p <- get_viral_load(excluded)
viral_load.discarted.p$type <- "discarted"
table(viral_load.discarted.p$per)

all <- rbind(viral_load.p, viral_load.discarted.p)


ggplot(all, aes(x=type, fill=as.factor(per))) + geom_bar()+scale_y_continuous(breaks=seq(0, 700, by=100))



table(all$type, all$per)





```

