---
title: "Untitled"
author: "Luisa Santus"
date: "3/23/2020"
output: html_document
---
# ------------------------------
# Mean Expression of LNCRNAs across celltypes and samples
# Each point represents the mean expression of one lncRNA across cells in one cell type
# Cells which show no expression of the lncRNAs taken into account are not considered for the mean #calculation
# -------
# -------
```{r examplecherrypicked}
get_expression_gene <- function(subset,ident = "", df,gene ,  threshold = 1 ){
  if(ident != ""){
    subset <- subset(subset, idents = ident)
  }
  subset <- subset(subset, features = c(gene))
  # Using normalized data 
  expression_matrix <- as.matrix(subset@assays$RNA@data)
  df_expression <- as.data.frame(t(expression_matrix))
  names(df_expression) <- c("Expression")

  # Save 
  dfo <- deparse(substitute(df))
  df_new <- data.frame(gene = gene, expression = df_expression$Expression, cell = rownames(df_expression))
  df_new$ident <- as.factor(ident)
  df_complete <- rbind(df, df_new)
  assign(dfo, df_complete, envir = .GlobalEnv);
}


# Plot separated by SAMPLE
df_expression_gene <- data.frame( gene = c(), expression = c(), ident = c(), cell = c())
identities <-as.vector(unique(Idents(immune.combined)))
invisible(lapply(identities, function(x) get_expression_gene(immune.combined, x, df_expression_gene, gene)))
ggplot(df_expression_gene, aes(x = ident, y = expression))+ geom_boxplot()

# Expression across cell types 
ggplot(df_expression_gene[df_expression_gene$expression > 0,], aes(x = ident, y = expression, fill = ident))+ geom_boxplot()+ theme_classic()+labs(title = "Expression of GENE across cells (cells with no expression filtered out)")



myeloids <- subset(immune.combined, ident = "Myeloid")
Idents(myeloids) <- myeloids$cond
#Idents(myeloids) <- myeloids$sample
myeloids_live_24 <- subset(myeloids, ident = "live")
#myeloids_live_24 <- subset(myeloids, ident = "live H024")
dim(myeloids_live_24)
# Identify which are the bystanders and which ones the infected ones 
subset_ebola <- subset(myeloids_live_24, features = ebola_genes)
# Retrieve unnnormalized expression data and separate the cells with at least one count od ebola
expression <-as.matrix(subset_ebola@assays$RNA@counts)
infected_cells <- colnames(expression)[colSums(expression) >0]
bystanders <- setdiff(colnames(expression), infected_cells)
length(infected_cells); length(bystanders)


ggplot(df_expression_gene[df_expression_gene$expression > 0,], aes(x = ident, y = expression, fill = status))+ geom_boxplot()+ theme_classic()+labs(title = "Expression of GENE across cells (cells with no expression filtered out)")
```

```{r expression }

df_expression_gene[df_expression_gene$cell %in% infected_cells,]$status <-  'infected'
df_expression_gene[df_expression_gene$cell %in% bystanders,]$status <-  'bystander'

myeloids <- subset(immune.combined, ident = "Myeloid")
Idents(myeloids) <- myeloids$cond
myeloids_live_24 <- subset(myeloids, ident = "live")
myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = infected_cells, value = 'infected')
myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = bystanders, value = 'bystander')
identities <- unique(Idents(myeloids_live_24))
df_lnc <- data.frame( gene_id = c(),meanexpr = c(), not_na_cells = c(), maxexpr = c(), medianexpr = c(), var = c(), var = c())
subset <- subset(myeloids_live_24, features = all_lncrnas)
invisible(lapply(identities, function(x) calc_mean_and_percentage_cell(subset, x, df_lnc)))

df_mrna <- data.frame( gene_id = c(),meanexpr = c(), not_na_cells = c(), maxexpr = c(), medianexpr = c(), var = c(), var = c())
subset <- subset(myeloids_live_24, features = annotated_mrnas)
dim(subset)
invisible(lapply(identities, function(x) calc_mean_and_percentage_cell(subset, x, df_mrna)))


df_lnc$type <- "lncRNA"
df_mrna$type   <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)
df_lnc_na <- nrow(df_lnc[is.na(df_lnc$meanexpr),])
df_mrna_na <- nrow(df_lnc[is.na(df_mrna$meanexpr),])
total_lnc <- length(unique(df_lnc$gene_id)) - df_lnc_na
total_mrna <- length(unique(df_mrna$gene_id)) - df_mrna_na

ggplot(df_complete, aes(x = type, y = maxexpr, fill = ident))+geom_boxplot()+theme_classic()+labs(title = paste0("MAX Expression values of mRNAs across cell type", "\n", "#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna), x = "")+theme(plot.title = element_text(hjust = 0.5))+scale_fill_brewer(palette = "Set3")

ggplot(df_complete, aes(x = type, y = medianexpr, fill = ident))+geom_boxplot()+theme_classic()+labs(title = paste0("MEDIAN Expression values of mRNAs across cell type", "\n", "#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna), x = "")+theme(plot.title = element_text(hjust = 0.5))+scale_fill_brewer(palette = "Set3")

ggplot(df_complete, aes(x = type, y = perc_cells_expressing, fill = ident))+geom_boxplot()+theme_classic()+labs(title = paste0("Percentage of cells expressing in Myeloids", "\n", "#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna), x = "")+theme(plot.title = element_text(hjust = 0.5))+scale_fill_brewer(palette = "Set3")


```
```{r percentage ebov reads}



ggplot(df_complete, aes(x = ident, y = meanexpr, fill = ident))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type", x = "")+ theme( axis.text.x=element_blank(),  axis.ticks.x=element_blank())
# mRNA
df <- data.frame( meanexpr = c(), not_na_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
Idents(subset) <- subset$ident
identities <-as.vector(unique(Idents(subset)))
invisible(lapply(identities, function(x) get_mean_expression_lnc_across_cells(subset, x, df, threshold)))
ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type", x = "")+ theme( axis.text.x=element_blank(),  axis.ticks.x=element_blank())


# Plot separated by Celltype
df_lnc <- data.frame( meanexpr = c(), not_na_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = all_lncrnas)
identities <-as.vector(unique(Idents(subset)))
invisible(lapply(identities, function(x) get_mean_expression_lnc_across_cells(subset, x, df_lnc, threshold)))
p <- ggplot(df_lnc, aes(x = ident, y = meanexpr, fill = ident))+geom_violin()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type", x = "") +geom_boxplot(width=0.1, fill = "white")

df_lnc$ident
library(ggpubr)
my_comparisons <- combn(as.character(unique(df_lnc$ident)), 2, simplify = FALSE)
p + stat_compare_means(comparisons = my_comparisons, label = "p.signif")
# mRNA
df_mrna <- data.frame( meanexpr = c(), not_na_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
identities <-as.vector(unique(Idents(subset)))
invisible(lapply(identities, function(x) get_mean_expression_lnc_across_cells(subset, x, df_mrna, threshold)))
ggplot(df_mrna, aes(x = ident, y = meanexpr, fill = ident))+ geom_violin()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type", x = "")+geom_boxplot(width=0.1, fill = "white") 

df_lnc$type <- "Lnc"
df_mrna$type <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)

# ---- Plot them together

ggplot(df_complete, aes(x = ident, y = meanexpr, fill = type))+geom_violin( trim = FALSE, position = position_dodge(0.9) )+geom_boxplot( width = 0.15, position = position_dodge(0.9))+theme_classic()+labs(title = "Mean Expression values of mRNAs across cell type", x = "")

ggplot(df_complete, aes(x = ident, y = meanexpr, fill = type))+geom_boxplot()+theme_classic()+labs(title = "Mean Expression values of mRNAs across cell type", x = "")


# Plot separated by SAMPLE
df <- data.frame( meanexpr = c(), not_na_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = all_lncrnas)
Idents(subset) <-paste(Idents(subset), subset$sample)
identities <-as.vector(unique(Idents(subset)))
invisible(lapply(identities, function(x) get_mean_expression_lnc_across_cells(subset, x, df, threshold)))
df$celltype <- unlist(lapply(df$ident, function(x) unlist(str_split(x, " "))[1]))
df$sample <- unlist(lapply(df$ident, function(x) paste(unlist(str_split(x, " "))[2:3], collapse = " ")))
ggplot(df, aes(x = celltype, y = meanexpr, fill = sample))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="Paired")+labs(title = "Mean Expression values of lncRNAs across cell type", x = "")
# mRNA
df <- data.frame( meanexpr = c(), not_na_cells = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
Idents(subset) <-paste(Idents(subset), subset$sample)
identities <-as.vector(unique(Idents(subset)))
invisible(lapply(identities, function(x) get_mean_expression_lnc_across_cells(subset, x, df, threshold)))
df$celltype <- unlist(lapply(df$ident, function(x) unlist(str_split(x, " "))[1]))
df$sample <- unlist(lapply(df$ident, function(x) paste(unlist(str_split(x, " "))[2:3], collapse = " ")))
ggplot(df, aes(x = celltype, y = meanexpr, fill = sample))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="Paired")+labs(title = "Mean Expression values of mRNAs across cell type", x = "")




# Outliers labeling
#------------------------
check_outlier <- function(v, coef=1.5){
  quantiles <- quantile(v,probs=c(0.05,0.95), na.rm = TRUE)
  IQR <- quantiles[2]-quantiles[1]
  res <- v < (quantiles[1]-coef*IQR)|v > (quantiles[2]+coef*IQR)
  return(res)
}
df$geneid <- rownames(df)
dat <- as.data.table(df)
dat[,outlier:=check_outlier(meanexpr), by =ident]
dat[,label:=ifelse(outlier,geneid,"")]
#------------------------



#+geom_text(aes(label=label),hjust=-0.3)
```

# Find markers that are lncrnas
```{r overallstats}
dim(immune.combined)
lnc_markers <- FindAllMarkers(immune.combined,features=annotated_lncrnas, logfc.threshold = 0 ,min.pct = 0)
top <- lnc_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
saveRDS(lnc_markers, "lncrmarkers.rds")

t <- top[top$cluster == "T",]$gene
t
nk <- top[top$cluster == "NK",]$gene
b <- top[top$cluster == "B",]$gene
marker.genes <- unique(c(t,nk,b))
DotPlot(immune.combined, features = rev(marker.genes), cols = c("blue", "red", "purple", "yellow"), dot.scale = 8, 
    split.by = "cond") + RotatedAxis()
marker.genes

nk<-c("ENSMMUG00000054210-unkown", "ENSMMUG00000063549-unkown","ENSMMUG00000060293-unkown")
FeaturePlot(immune.combined, features = nk)
plot_genes(immune.combined,nk,threshold= 3, title = "Monocytes marker expression", col = "purple")

pcHeatmap(nbt,pc.use = 1,use.full = TRUE,do.balanced = TRUE,remove.key = TRUE)

rownames(immune.combined@assays$RNA@scale.data)[grepl("ENSMMUG00000054210", rownames(immune.combined@assays$RNA@scale.data))]

a <- DoHeatmap(immune.combined,assay = "RNA",  slot = "data", features = top$gene[1:4], cells = 5000) + NoLegend()


immune.combined$celltype <- Idents(immune.combined)
plots <- VlnPlot(immune.combined, features = t[1:2], group.by = "cond",group.by = "celltype", 
    pt.size = 0, combine = FALSE)
plots
```




# Visualizing PCA 

```{r lncrnas}

df_lnc <- data.frame( meanexpr = c(), perc_cells_expressing = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = all_lncrnas)

dim(subset)
subset$individual <- unlist(lapply(subset$orig.ident , function(x) paste(unlist(str_split(x, "-")[[1]][1]))))
Idents(subset) <- paste(Idents(subset), subset$individual  ,subset$sample, sep = "_") 

identities <-as.vector(unique(Idents(subset)))
invisible(lapply(identities, function(x) get_mean_expression_lnc_across_cells(subset, x, df_lnc)))

df_lnc

df_lnc$celltype <- unlist(lapply(df_lnc$ident , function(x) paste(unlist(str_split(x, "_")[[1]][1]))))
df_lnc$individual <- unlist(lapply(df_lnc$ident , function(x) paste(unlist(str_split(x, "_")[[1]][2]))))
df_lnc$sample <- unlist(lapply(df_lnc$ident , function(x) paste(unlist(str_split(x, "_")[[1]][3]))))
df_lnc$sample <- factor(df_lnc$sample,levels = c("media H004", "media H024", "irrad H004", "irrad H024", "live H004", "live H024"))

df_lnc <- df_lnc[!is.na(df_lnc$meanexpr),]

ggplot(df_lnc, aes(x = sample, y = perc_cells_expressing, fill = individual))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="Paired")+labs(title = "Mean Expression values of ebola genes across cell type", x = "")


data_barplot <- data_summary(df_lnc, varname = "perc_cells_expressing", groupnames = c("sample", "individual", "celltype") )
ggplot(data_barplot, aes(x=celltype, y=perc_cells_expressing, fill=sample )) + 
   geom_bar(stat="identity", position=position_dodge())+ scale_fill_brewer(palette="Dark2") + theme_classic()+labs(title = "Percentage of cells expressing lncRNAs", y = "% cells")


# Only in Myeloid 
df_myeloid  <- df_lnc[df_lnc$celltype == "Myeloid",]
data_barplot <- data_summary(df_myeloid, varname = "perc_cells_expressing", groupnames = c("sample", "individual", "celltype") )
ggplot(data_barplot, aes(x=sample, y=perc_cells_expressing, fill=individual )) + 
   geom_bar(stat="identity", position=position_dodge())+ scale_fill_brewer(palette="Dark2") + theme_classic()+labs(title = "Percentage of Myeloid cells expressing lncRNAs", y = "% cells")
```

# --- DE analysis 

```{r DE }

# DE only in Infected Cells Live H024, Myeloid 
Idents(immune.combined)
myeloids <- subset(immune.combined, ident = "Myeloid")
Idents(myeloids) <- myeloids$sample
myeloids_live_24 <- subset(myeloids, ident = "live H024")
dim(myeloids_live_24)
# So we are left with 2,2k cells and 16k genes 

# Identify which are the bystanders and which ones the infected ones 
subset_ebola <- subset(myeloids_live_24, features = ebola_genes)
# Retrieve unnnormalized expression data and separate the cells with at least one count od ebola
expression <-as.matrix(subset_ebola@assays$RNA@counts)
infected_cells <- colnames(expression)[colSums(expression) >0]
bystanders <- setdiff(colnames(expression), infected_cells)
length(infected_cells); length(bystanders)

#Rename identities
myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = infected_cells, value = 'infected')
myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = bystanders, value = 'bystander')
unique(Idents(myeloids_live_24))

# DE analysis - ALL 
de_genes_all <- FindMarkers(myeloids_live_24, ident.1 ="infected", ident.2="bystander" )
de_genes_all <-de_genes_all[!grepl("ribodepl-MSTRG.72510", rownames(de_genes_all)),]
nrow(de_genes_all)
ggplot(de_genes_all, aes(x=avg_logFC, y=-log10(p_val_adj), label =  rownames(de_genes_all))) + geom_point() +
ggtitle("Volcano") + xlab("log2 FC") + ylab("-log10 adjusted p-value")+ theme_classic()+ labs(title = "Vulcano plot of all DE genes")+ geom_text(vjust = 0, nudge_y = 0.6, check_overlap = TRUE)

# DE analysis - mRNAs 
de_genes_mrna <- FindMarkers(myeloids_live_24, ident.1 ="infected", ident.2="bystander", features = annotated_mrnas )
nrow(de_genes_mrna)
ggplot(de_genes_mrna, aes(x=avg_logFC, y=-log10(p_val_adj), label =  rownames(de_genes_mrna))) + geom_point() +
ggtitle("Volcano") + xlab("log2 FC") + ylab("-log10 adjusted p-value")+ theme_classic()+ labs(title = "Vulcano plot of all DE genes")+ geom_text(vjust = 0, nudge_y = 0.6, check_overlap = TRUE)

de_genes_mrna <- de_genes_mrna[de_genes_mrna %in% keys(org.Mmu.eg.db, keytype="SYMBOL")]
ego <- clusterProfiler::enrichGO(gene = de_genes_mrna,OrgDb = org.Mmu.eg.db, keyType = 'SYMBOL')
keys(org.Mmu.eg.db,keytype = "SYMBOL")
barplot(ego, showCategory=10)
clusterProfiler::dotplot(ego)

do_go(de_genes_mrna)

# DE analysis - lncRNAs 
de_genes_lnc <- FindMarkers(myeloids_live_24, ident.1 ="infected", ident.2="bystander", features = all_lncrnas )
rownames(de_genes_lnc)
ggplot(de_genes_lnc, aes(x=avg_logFC, y=-log10(p_val_adj), label =  rownames(de_genes_lnc))) + geom_point() +
ggtitle("Volcano") + xlab("log2 FC") + ylab("-log10 adjusted p-value")+ theme_classic()+ labs(title = "Vulcano plot of all DE genes")+ geom_text(vjust = 0, nudge_y = 0.6, check_overlap = TRUE)

# Testing what they are 
lnc_ranges_complete[lnc_ranges_complete$gene_id == "ribodepl_MSTRG.72510-unkown",]
export(lnc_ranges_complete[grepl("polya_MSTRG.19809",lnc_ranges_complete$gene_id )], "/home/luisas/Desktop/testpolya.gtf")

# Check DE with MAST 
library("MAST")
library("scater")

myeloids_live_24$individual <- unlist(lapply(myeloids_live_24$orig.ident , function(x) paste(unlist(str_split(x, "-")[[1]][1]))))
myeloids_live_24$ident <- Idents(myeloids_live_24)

de_genes <- FindMarkers(myeloids_live_24, ident.1 ="infected", ident.2="bystander", test.use = "negbino", latent.vars = c("ident","individual"))



setdiff( rownames(de_genes_all),rownames(de_genes))
myeloids_live_24.sc <- as.SingleCellExperiment(myeloids_live_24)

cond<-factor(colData(myeloids_live_24.sc)$ident)
cond<-relevel(cond,"bystander")
cond<-relevel(cond,"Unstim")
colData(myeloids_live_24.sc)$cond<-cond
myeloids_live_24.sc$individual <- unlist(lapply(myeloids_live_24.sc$orig.ident , function(x) paste(unlist(str_split(x, "-")[[1]][1]))))



scaRaw <- FromMatrix(t(myeloids_live_24@assays$RNA@counts))
zlmCond <- zlm(~cond + individual, myeloids_live_24@assays$RNA@counts)


```



# Coexpression analysis
# Check coexpression with ebola genes 

```{r coexpression }

# extract the count matrix (not  normalized) 
count_matrix <- immune.combined@assays$RNA@counts
#count_matrix <- test@assays$RNA@counts


# Check the ranges 
mrnas_ranges <- ref[ref$gene_name %in% mrna_ref_id  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unkown", "",all_lncrnas))]

# The novel lncRNAs are only presenting exon entiries missing out the genes ones.
# Add missing gene entries 
# ------------------------------
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(c,gr_list)
lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)

# ----------------------
# Find pairs allowing 10K gap on both sides
mix <-cross2(ebola_genes[2],unique(lnc_ranges_complete$gene_id)[1:200])
length(mix)
correlations <- lapply(mix, function(x) calc_correlation_with_names(x[1],x[2],count_matrix))
correlations[is.na(correlations)] <- -1.2
correlations
threshold = 0
mix_filtered <- mix[correlations != threshold]
mix_filtered
# Visualize correlations in density plot
df <- as.data.frame(unlist(correlations)); names(df) <- c("corr")
ggplot(df, aes(x=corr)) + geom_density()

# Create nodes and edges 
source <- vector(); target <- vector(); all <- vector(); categories <- vector()
lapply(mix, function(x) create_nodes_and_edges(x, source, target,all, categories))
nodes <- data.frame(id=all, group=categories, stringsAsFactors=FALSE)
edges <- data.frame(source=source, target=target, stringsAsFactors=FALSE)
RCy3::createNetworkFromDataFrames(nodes,edges, title="my", collection="ex")
```
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r percentage ebov reads}

# --------------------------------------------
# Plot of expression across cells when pooled
# --------------------------------------------
df_filtered_lnc<- df_lnc[df_lnc$perc_cells_expressing > 0.01 , ]
df_filtered_mrna <- df_mrna[df_mrna$perc_cells_expressing > 0.01, ]

total_lnc <-  length(unique(df_filtered_lnc$gene_id))
total_mrna <-  length(unique(df_filtered_mrna$gene_id))

df_filtered_lnc$type <- "lncRNA"
df_filtered_mrna$type   <- "Protein Coding"
df_complete <- rbind(df_filtered_lnc, df_filtered_mrna)
ggplot(df_complete, aes(x = type, y = perc_cells_expressing, fill = type))+geom_boxplot()+theme_classic()+labs(title = paste0("Proportion of cells expressing genes", "\n", "#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna), x = "",  y = "Cells proportion")+theme(legend.title = element_blank())+theme(plot.title = element_text(hjust = 0.5))+scale_fill_brewer(palette = "Dark2")


total_lnc <-  length(unique(df_lnc$gene_id))
total_mrna <-  length(unique(df_mrna$gene_id))
df_lnc$type <- "lncRNA"
df_mrna$type   <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)

# Expression when pooled
ggplot(df_complete, aes(x = type, y = medianexpr, fill = type))+geom_boxplot()+theme_classic()+labs(title = paste0("Median expression", "\n", "#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna), x = "",  y = "median expression")+theme(legend.title = element_blank())+theme(plot.title = element_text(hjust = 0.5))+scale_fill_brewer(palette = "Dark2")
# Maxx
ggplot(df_complete, aes(x = type, y = maxexpr, fill = type))+geom_boxplot()+theme_classic()+labs(title = paste0("Max expression", "\n", "#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna), x = "",  y = "max expression ")+theme(legend.title = element_blank())+theme(plot.title = element_text(hjust = 0.5))+scale_fill_brewer(palette = "Dark2")


# Housekeeping genes
# list from https://www.genomics-online.com/resources/16/5049/housekeeping-genes/
hk_genes <- c("^ACTB$", "^GAPDH$", "^PGK1$", "^PPIA$", "^B2M$", "^YWHAZ$", "^SDHA$", "^TFRC$", "^GUSB$", "^HMBS$", "^HPRT1$", "^TBP$")
hk_genes <- c("^B2M$", "^BLM$","^PPIA$")

# df_mrna$tolabel <- df_mrna$gene_id
# df_mrna[which(df_mrna$perc_cells_expressing < 0.8 |  df_mrna$maxexpr < 5 ),]$tolabel <- ""
# df_mrna$tolabel[!is.na(df_mrna$tolabel)]
# df_mrna[which(!grepl(paste(hk_genes, collapse = "|"), df_mrna$gene_id)),]$tolabel <- ""

# max_mrna_scatter <- ggplot(df_mrna, aes(x = maxexpr , y = perc_cells_expressing))+geom_point(size = 1, col = col_mrna)+theme_classic()+labs(title = paste0("mRNAs. ",total_mrnas), x = " Normalized MAX Expression", y = "Cells proportion")+theme(plot.title = element_text(hjust = 0.5))+geom_point(data=df_mrna[df_mrna$tolabel != "", ], aes(x = maxexpr , y = perc_cells_expressing), colour="red", size=3)+geom_text(aes(label=df_mrna$tolabel),hjust=1, vjust=0.7)
# ggExtra::ggMarginal(max_mrna_scatter,type = 'boxplot',margins = 'both',size = 8,colour = "darkgrey",fill = col_mrna, alpha = 0.5)





```
## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

# Number of genes expressed per cell type ( All genes, lncRNAs, mRNAs )


```{r percentage ebov reads}
# Boxplot
df_filtered_lnc<- df_lnc[df_lnc$perc_cells_expressing > 0.01 , ]
df_filtered_mrna <- df_mrna[df_mrna$perc_cells_expressing > 0.01, ]

total_lnc <-  length(unique(df_filtered_lnc$gene_id))
total_mrna <-  length(unique(df_filtered_mrna$gene_id))

df_filtered_lnc$type <- "lncRNA"
df_filtered_mrna$type   <- "Protein Coding"
df_complete <- rbind(df_filtered_lnc, df_filtered_mrna)



df_lnc$type <- "lncRNA"
df_mrna$type   <- "Protein Coding"
df_complete <- rbind(df_lnc, df_mrna)
total_lnc <-  length(unique(df_lnc$gene_id))
total_mrna <-  length(unique(df_mrna$gene_id))

ggplot(df_complete, aes(x = type, y = var, fill = type))+geom_boxplot()+theme_classic()+theme(legend.position = "")+scale_fill_brewer(palette = "Dark2")+labs(y = "Variance", title = "Expression Variance ACROSS CELLS (Expression per gene) \n filtering only the ones showing expression in at least 1% of cells", x = "", subtitle = paste0("#lncRNAs: ", total_lnc,"\n", "#mRNAs: ", total_mrna))+theme(plot.title = element_text(hjust = 0.5))+theme(plot.subtitle= element_text(hjust = 0.5))


```

```{r overallstats}

# Plot cell proportion across cell types 
count_genes <- function(df_lnc, df_mrna, ident){
  x <- df_lnc[df_lnc$ident == ident,]
  count_lnc <- length(unique(x$gene_id))
  x <- df_mrna[df_mrna$ident == ident,]
  count_mrna <- length(unique(x$gene_id))
  return(paste0(ident, " lnc: ", count_lnc, " mrna: ", count_mrna))
}
unlist(lapply(unique(df_filtered_lnc$ident), function(ident) count_genes(df_filtered_lnc,df_filtered_mrna,ident) ))
# Select only one identity at the time - cell type

get_n_genes <- function(immune.combined,ident, df, threshold = 0 ){
  subset <- subset(immune.combined, idents = ident)
  n_genes <- as.vector((Matrix::colSums(subset@assays$RNA@scale.data>threshold)))
  dfo <- deparse(substitute(df))
  df_new <- data.frame(n_genes = n_genes)
  df_new$ident <- as.factor(ident)
  df_new$sample <- subset$sample
  df_complete <- rbind(df, df_new)
  assign(dfo, df_complete, envir = .GlobalEnv);
}


# Detect all different identities
# All
identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), n_genes = c())
# Collect number of genes expressed per cell type
lapply(identities, function(x) get_n_genes(immune.combined, x, df, threshold))
# Visualize - Violin plot
ggplot(df, aes(x=ident, y=n_genes, fill = ident)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") + geom_boxplot(width=0.1, fill="white")+ theme_classic()
ggplot(df, aes(x=ident, y=n_genes, fill = sample)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") +theme_classic()
data_barplot <- data_summary(df, varname = "n_genes", groupnames = c("sample", "ident") )
ggplot(data_barplot, aes(x=ident, y=n_genes, fill=sample)) + 
   geom_bar(stat="identity", position=position_dodge()) +
  geom_errorbar(aes(ymin=n_genes-sd, ymax=n_genes+sd), width=.2,
                 position=position_dodge(.9))+ scale_fill_brewer(palette="RdBu") + theme_minimal()



identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), n_genes = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = all_lncrnas)
lapply(identities, function(x) get_n_genes(subset, x, df, threshold))
# Visualize - Violin plot
ggplot(df, aes(x=ident, y=n_genes, fill = ident)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") + geom_boxplot(width=0.1, fill="white")+ theme_classic()
ggplot(df, aes(x=ident, y=n_genes, fill = sample)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") +theme_classic()
data_barplot <- data_summary(df, varname = "n_genes", groupnames = c("sample", "ident") )
ggplot(data_barplot, aes(x=ident, y=n_genes, fill=sample)) + 
   geom_bar(stat="identity", position=position_dodge()) +
  geom_errorbar(aes(ymin=n_genes-sd, ymax=n_genes+sd), width=.2,
                 position=position_dodge(.9))+ scale_fill_brewer(palette="RdBu") + theme_minimal()


identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), n_genes = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
lapply(identities, function(x) get_n_genes(subset, x, df, threshold))
# Visualize - Violin plot
ggplot(df, aes(x=ident, y=n_genes, fill = ident)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") + geom_boxplot(width=0.1, fill="white")+ theme_classic()
ggplot(df, aes(x=ident, y=n_genes, fill = sample)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") +theme_classic()
data_barplot <- data_summary(df, varname = "n_genes", groupnames = c("sample", "ident") )
ggplot(data_barplot, aes(x=ident, y=n_genes, fill=sample)) + 
   geom_bar(stat="identity", position=position_dodge()) +
  geom_errorbar(aes(ymin=n_genes-sd, ymax=n_genes+sd), width=.2,
                 position=position_dodge(.9))+ scale_fill_brewer(palette="RdBu") + theme_minimal()
```

```{r overallstats}

get_mean_expression <- function(immune.combined,ident, df, threshold = 0 ){
  subset <- subset(immune.combined, idents = ident)
  
  expr_gene_subset <- FetchData(subset, rownames(subset))
  meanexpr <- rowMeans(subset@assays$RNA@data)
  dfo <- deparse(substitute(df))
  df_new <- data.frame(meanexpr = meanexpr)
  df_new$ident <- as.factor(ident)
  df_new$sample <- subset$sample
  df_complete <- rbind(df, df_new)
  assign(dfo, df_complete, envir = .GlobalEnv);
}

# -------------------------
# Expression of lncrnas in each single cell
identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), meanexpr = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_lncrna)
lapply(identities, function(x) get_mean_expression(subset, x, df, threshold))

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_violin()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = sample))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type", x = "", y = "")


# -------------
# same for mrna 

# -------------------------
# Expression of lncrnas in each single cell
identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), meanexpr = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
lapply(identities, function(x) get_mean_expression(subset, x, df, threshold))

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_violin()+ theme_classic()+geom_boxplot()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = sample))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type", x = "", y = "")

```



```{r overallstats}

get_mean_expression <- function(immune.combined,ident, df, threshold = 0 ){
  subset <- subset(immune.combined, idents = ident)
  expr_gene_subset <- FetchData(subset, rownames(subset))
  meanexpr <- rowMeans(expr_gene_subset)
  dfo <- deparse(substitute(df))
  df_new <- data.frame(meanexpr = meanexpr)
  df_new$ident <- as.factor(ident)
  df_new$sample <- subset$sample
  df_complete <- rbind(df, df_new)
  assign(dfo, df_complete, envir = .GlobalEnv);
}

# -------------------------
# Expression of lncrnas in each single cell
identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), meanexpr = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_lncrna)
lapply(identities, function(x) get_mean_expression(subset, x, df, threshold))

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_violin()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = sample))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of lncRNAs across cell type", x = "", y = "")


# -------------
# same for mrna 

# -------------------------
# Expression of lncrnas in each single cell
identities <-as.vector(unique(Idents(immune.combined)))
#Initialze data frame for plotting
df <- data.frame(ident = c(), meanexpr = c())
# Collect number of genes expressed per cell type
subset <- subset(immune.combined, features = annotated_mrnas)
lapply(identities, function(x) get_mean_expression(subset, x, df, threshold))

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = ident))+ geom_violin()+ theme_classic()+geom_boxplot()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type")

ggplot(df, aes(x = ident, y = meanexpr, fill = sample))+ geom_boxplot()+ theme_classic()+ scale_fill_brewer(palette="RdBu")+labs(title = "Mean Expression values of mRNAs across cell type", x = "", y = "")

```
## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


# Percentage of Subset of Genes (e.g lncRNAs) expressed
# Calculate the number of genes expressed from subset per cell
# Calculate the total number of genes expressed in that cell
# calculate the percentage and plot 


```{r overallstats}

threshold<- 0
identities <-as.vector(unique(Idents(immune.combined)))

# -------------
#     EBOLA
# -------------
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unkown", sep ="-")]
genes_subset <- ebola_genes
#Initialze data frame for plotting
df <- data.frame(n_genes = c(), n_genes_subset = c(), perc = c())
# Collect number of genes expressed per cell type
lapply(identities, function(x) get_percentage_genes_expressed(immune.combined, x, genes_subset, threshold, df))
# Change data for barplot
data_barplot <- data_summary(df, varname = "perc", groupnames = c("sample", "celltype") )
data_barplot$perc <-  data_barplot$perc * 100

# Visualize - Violin plot
ggplot(df, aes(x=celltype, y=perc, fill = sample)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") + geom_boxplot(width=0.1, fill="white")+ theme_classic()
# Barplot 
ggplot(data_barplot, aes(x=celltype, y=perc, fill=sample)) + 
   geom_bar(stat="identity", position=position_dodge()) +
  labs(title = "Percentage of ebla genes expressed per cell", x = "")+
  geom_errorbar(aes(ymin=perc-sd, ymax=perc+sd), width=.2,
                 position=position_dodge(.9))+ scale_fill_brewer(palette="Spectral") + theme_classic()

# ----------------
#     LncRNAs
# ----------------
genes_subset <- all_lncrnas
#Initialze data frame for plotting
df <- data.frame(n_genes = c(), n_genes_subset = c(), perc = c())
# Collect number of genes expressed per cell type
lapply(identities, function(x) get_percentage_genes_expressed(immune.combined, x, genes_subset, threshold, df))
# Change data for barplot
data_barplot <- data_summary(df, varname = "perc", groupnames = c("sample", "celltype") )
data_barplot$perc <-  data_barplot$perc * 100

# Visualize - Violin plot
ggplot(df, aes(x=celltype, y=perc, fill = sample)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") + theme_classic()
# Barplot 
ggplot(data_barplot, aes(x=celltype, y=perc, fill=sample)) + 
   geom_bar(stat="identity", position=position_dodge()) +
  labs(title = "Percentage of lncRNAs genes expressed per cell", x = "", y = "")+
  geom_errorbar(aes(ymin=perc-sd, ymax=perc+sd), width=.2,
                 position=position_dodge(.9))+ scale_fill_brewer(palette="RdBu") + theme_classic()

# ----------------
#     LncRNAs
# ----------------
genes_subset <- annotated_mrnas
#Initialze data frame for plotting
df <- data.frame(n_genes = c(), n_genes_subset = c(), perc = c())
# Collect number of genes expressed per cell type
lapply(identities, function(x) get_percentage_genes_expressed(immune.combined, x, genes_subset, threshold, df))
# Change data for barplot
data_barplot <- data_summary(df, varname = "perc", groupnames = c("sample", "celltype") )
data_barplot$perc <-  data_barplot$perc * 100

# Visualize - Violin plot
ggplot(df, aes(x=celltype, y=perc, fill = sample)) + labs(title = "Number of genes expressed per CellType ")+geom_violin()+ scale_fill_brewer(palette="RdBu") + theme_classic()
# Barplot 
ggplot(data_barplot, aes(x=celltype, y=perc, fill=sample)) + 
   geom_bar(stat="identity", position=position_dodge()) +
  labs(title = "Percentage of mRNAs genes expressed per cell", x = "", y = "")+
  geom_errorbar(aes(ymin=perc-sd, ymax=perc+sd), width=.2,
                 position=position_dodge(.9))+ scale_fill_brewer(palette="RdBu") + theme_classic()
```
