---
title: "Untitled"
author: "Luisa Santus"
date: "2/28/2020"
output: html_document
---


# Supplementary figures

## Compare number of genes in macaques vs Human annotation 
```{r results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rtracklayer)
library(dplyr)
library(ggplot2)
library(plyr)
library(ggpubr)

# Import Utils
source(file.path("../utils/01_lncrna_annotation_utils.R"))

ref_macaque <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf"))
ref_human <- import(file.path("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf"))


ref_macaque_gene <- ref_macaque[ref_macaque$type =="gene",]

table(ref_macaque[ref_macaque$source == "ensembl",]$gene_biotype)
table(ref_macaque[ref_macaque$source == "RefSeq",]$gene_biotype)

table(ref_human$gene_biotype)
ref_macaque_gene <- ref_macaque[ref_macaque$type == "gene",]
ref_human_gene <- ref_human[ref_human$type == "gene",]

lnc_macaque <- sum(ref_macaque_gene$gene_biotype == "lncRNA")
pc_macaque <- sum(ref_macaque_gene$gene_biotype == "protein_coding")
lnc_human <- sum(ref_human_gene$gene_biotype == "lncRNA")
pc_human <- sum(ref_human_gene$gene_biotype == "protein_coding")

count <- c(lnc_macaque, pc_macaque, lnc_human, pc_human ) 
species <- c(rep("Macaque", 2), rep("Human",2))
gene_biotype <- c(rep(c("LncRNA", "Protein Coding"), 2 ) )
df_n_genes <- data.frame(count, species, gene_biotype)
```



```{r results='hide', message=FALSE, warning=FALSE}
ggplot(df_n_genes, aes(fill = gene_biotype, y = count, x = species))+
                geom_bar(stat="identity", position = position_dodge2())+
                theme_paper+scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)+
                labs(x = "", y = "")+
                scale_y_continuous( breaks = c(lnc_macaque, lnc_human))+
                geom_abline(slope=0, intercept=lnc_macaque,  col = "darkgrey",lty=2)+
                geom_abline(slope=0, intercept=lnc_human,  col = "darkgrey",lty=2)
```

```{r include=FALSE}
# Do only once: extract onluy lnc and pc from annotation file of macaque and human 
# Run gff compare 


# MACAQUE

#ref_ensembl <- import("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/rheMac10/Macaca_mulatta.Mmul_10.100.gtf") 

#ref_ensembl_lnc  <- ref_ensembl[ref_ensembl$gene_biotype == "lncRNA",]
#ref_ensembl_pc  <- ref_ensembl[ref_ensembl$gene_biotype == "protein_coding",]
#ref_ensembl_notlnc <- ref_ensembl[ref_ensembl$gene_biotype != "lncRNA",]

#export(ref_ensembl_lnc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/rheMac10/ensembl_100_lnc.gtf")
#export(ref_ensembl_pc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/rheMac10/ensembl_100_pc.gtf")
#export(ref_ensembl_notlnc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/rheMac10/ensembl_100_notlnc.gtf")

# Human 

#ref_ensembl <- import("/home/luisas/Desktop/cluster/gene_annotations/ensembl_release100/homo_sapiens/Homo_sapiens.GRCh38.100.gtf") 

#table(ref_ensembl$transcript_biotype)
#ref_ensembl_lnc  <- ref_ensembl[ref_ensembl$gene_biotype == "lncRNA",]
#ref_ensembl_pc  <- ref_ensembl[ref_ensembl$gene_biotype == "protein_coding",]
#ref_ensembl_notlnc <- ref_ensembl[ref_ensembl$gene_biotype != "lncRNA",]

#export(ref_ensembl_lnc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/Human//ensembl_100_lnc.gtf")
#export(ref_ensembl_pc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/Human//ensembl_100_pc.gtf")
#export(ref_ensembl_notlnc, "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/Human/ensembl_100_notlnc.gtf")
```

# Macaque 

```{r benchmark}
pcvslnc <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/rheMac10/01_gffcompare/pcvslnc.annotated.gtf")
notlncvslnc <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/rheMac10/01_gffcompare/notlncvslnc.annotated.gtf")

#plot_stats_benchmark(pcvslnc)
#plot_stats_benchmark(notlncvslnc)
```



# Human

```{r benchmark}
pcvslnc_human <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/Human/01_gffocompare/pcvslnc.annotated.gtf")
notlncvslnc_human <- import("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/01_PreliminaryFiles_rheMac10/benchmark/Human//01_gffocompare/notlncvslnc.annotated.gtf")

#plot_stats_benchmark(pcvslnc_human)
#plot_stats_benchmark(notlncvslnc_human)


prep_df_plot <- function(notlncvslnc,name){
  data <- data.frame(table(notlncvslnc[notlncvslnc$type == "transcript",]$class_code))
  colnames(data) <- c("group", "value")
  data <- data[data$group %in% c("u", "i", "x"), ]
  # Basic piechart
  # Compute the position of labels
  data <- data %>% 
      arrange(desc(group)) %>%
      mutate(prop = value / sum(data$value) *100) %>%
      mutate(ypos = cumsum(prop)- 0.5*prop )
  data$organism <- name
  return(data)
}

mac <- prep_df_plot(notlncvslnc, "Macaque")
human <- prep_df_plot(notlncvslnc_human, "Human")
data <- rbind(mac, human)


ggplot(data, aes(x = organism, y = value, fill = group))+geom_bar(stat="identity", position=position_dodge())+
      theme_paper+
      scale_fill_manual(values=c("orange","#994C00", "red"),labels = c("Antisense", "Intergenic", "Intronic"))+
      xlab("")+ylab("")+ggtitle("Annotatied LncRNAs (transcripts) stratified by type")
```



