---
title: "DE_bulk"
author: "Luisa Santus"
date: "12/2/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential expression Bulk 


```{r cars}
library(sva); library(edgeR)
library("genefilter")
library("gplots")
library(plyr)
library(geneplotter)
library(clusterProfiler)
library(org.Mmu.eg.db)
library(rtracklayer)
library(stringr)
library(edgeR)
library(zeallot)
library(DESeq2)
library(readr)
library(tximport)

# ----------------------------------
#           Import Utils 
# ----------------------------------
scripts_dir = file.path("/home/luisas/Desktop/cluster/proj/code/ebola/analysis/utils/")
source(file.path(scripts_dir,"03_de_utils.R"));

datadir <- "/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/"

ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))

dir_counts_ref = file.path(datadir, "04_quantification/")
htseq_files_ref <- iterate_files(dir_counts_ref, ".gene_counts.tab")
batch <- htseq_files_ref[str_detect(htseq_files_ref, "Batch01")]
zyagen <- htseq_files_ref[str_detect(htseq_files_ref, "Zyagen")]
htseq_files_ref <- c(batch,zyagen)
#htseq_files_ref <- batch


abundance_files <- iterate_files(dir_counts_ref, "t_data.ctab")
# Create Map for Transcript names to Gene names 
tmp <- read_tsv(abundance_files[1])
tx2gene <- tmp[,c("t_name", "gene_id")]
txi <- tximport(abundance_files, type = "stringtie", tx2gene = tx2gene)
colnames(txi$counts) <- abundance_files

dds <- create_dds(abundance_files,names_from_dir = TRUE,  tximport = TRUE, tx_obj = txi)
dds$tissue

samples_info <- sapply(sampleFiles,extract_info_sample_from_dir)

# Retain only blood and liver
dds <- dds[, dds$tissue == "Blood" | dds$tissue == "Liver"]
dds <- dds[!rownames(dds) %in% ebola_ref$gene_id,]

# remove one sample
#dds <- dds[,dds$complete_id != "Batch01_Blood_D07_RA0700"]
dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))
assays(dds)$logCPM <- edgeR::cpm(dge, log=TRUE, prior.count=0.5)
```

```{r cars}
# ----------------------------------------
#         Remove lowly expressed genes 
# ----------------------------------------
mask <- rowSums(edgeR::cpm(dge) > 0) > 0
dds.filt <- dds[mask, ]
dge.filt <- dge[mask, ]

# Plot 
multidensity(as.list(as.data.frame(assays(dds.filt)$logCPM)), xlab = "log2 CPM", legend = NULL,
             main = "After Filtering", cex.axis = 0.7,  las = 1)

multidensity(as.list(as.data.frame(assays(dds)$logCPM)), xlab = "log2 CPM", legend = NULL,
             main = "Before filtering", cex.axis = 0.7,  las = 1)

# How many lncRNAs I am retaining 

retained_lncRNAs_proportion <- length(rownames(dds.filt)[rownames(dds.filt) %in% lncRNAs_ref$gene_id])/length(unique(lncRNAs_ref$gene_id))*100
lost_lncRNAs_proportion <- 100-retained_lncRNAs_proportion
sprintf("Retained lncRNAs %s%s", retained_lncRNAs_proportion, "%")
sprintf("Lost lncRNAs %s%s", lost_lncRNAs_proportion, "%")

```

## Normalization

Since different samples may have different RNA compositions and this could be problematic for further analyses we need to take it into account and normalize the data.
We estimate a normalization factor for each library using the Trimmed Mean of M-Values and apply it to our data.

```{r}

dds <- dds.filt
dge <- dge.filt

boxplot(assays(dds)$logCPM , col = "pink", las=2, main="", names = dds$id)
title(main="Before Normalization ",ylab="Log-cpm")

dge <- calcNormFactors(dge) 
assays(dds)$logCPM <- edgeR::cpm(dge, log=TRUE, normalized.lib.sizes=TRUE, prior.count=0.25)



boxplot(assays(dds)$logCPM , col = "pink", las=2, main="" , names = dds$id)
title(main="After Normalization ",ylab="Log-cpm")

```

```{r cars}

# ---------------------------------------
#           Add Metadata
#-----------------------------------------
# Read metadata 
projdir = "/home/luisas/Desktop/cluster/data/"
metadata <- read.csv(file.path(projdir, "00_Metadata/batch_metadata.csv"), header = FALSE)
colnames(metadata) <- c("id", "gender", "weight", "birthday", "Type")
# Add metadata 
dds$infection <-ifelse(as.double(sub("D", "", dds$dpo))>0, "infected", "not-infected")
dds$gender <- metadata$gender[match(dds$id, metadata$id)]
dds$birthyear <- as.double(str_sub(metadata$birthday, -4))[match(dds$id,metadata$id)]
dds$weight <- as.double(metadata$weight)[match(dds$id,metadata$id)]

colnames(dds) <- dds$complete_id
# Order the levels
dds$infection <- factor(dds$infection, levels = c("infected","not-infected"), ordered = TRUE )
dds$gender <- factor(dds$gender, levels = c("Female","Male"), ordered = TRUE )
dds$birthyear <- factor(dds$birthyear, levels = c("2011","2012"), ordered = TRUE )

dge <- DGEList(counts = assays(dds)$counts, group = dds$dpo, genes = rownames(dds), samples = colData(dds))
dge <- calcNormFactors(dge) 
dge <- estimateCommonDisp(dge)
#mds <- plotMDS(cpm(dge, log = TRUE,prior.count = 0.01), main=" samples on genes CPM", top=50,  gene.selection="common")

```


```{r cars}
# Mixed model!!! sample id!!!
target <- colData(dds)
# I remove DPO because it makes the matrix not full rank!
design <- model.matrix(~factor(infection)+ factor(tissue)+ factor(birthyear)+ factor(gender), target) 
is.fullrank(design)
v <- voom(dge, design, plot=TRUE)
cor <- duplicateCorrelation(v, design, block = target$id)
v <- voom(dge, design, plot = TRUE, block = target$id, correlation = cor$consensus)
fit5 <- lmFit(v, design, block = target$id, correlation = cor$consensus)

FDRcutoff=0.05
fit5 <- eBayes(fit5)
res5 <- decideTests(fit5, p.value = FDRcutoff)
tt5 <- topTable(fit5, coef = 2, n = Inf)

DEgenes5 <- rownames(tt5)[tt5$adj.P.Val < FDRcutoff & !is.na(tt5$logFC)]
length(DEgenes5)

dds$dpo <- as.numeric(str_remove(dds$dpo , "D"))
colnames(dds) <- dds$dpo 

bulk_de_genes_all <- DEgenes5
saveRDS(bulk_de_genes_all, file.path(robjectsdir, "bulkDEgenes.rds"))
```


```{r}
# Plot the top 300 differentially expressed genes 

# Long non coding in reference macaque 
novel_and_ebola <-lncRNAs_ref[is.na(ref$gene_biotype)]
annotated_all <- ref[!is.na(ref$gene_biotype)]
annotated_lncrnas <- annotated_all[annotated_all$gene_biotype =="lncRNA"]

lnc_de_novel <-DEgenes5[DEgenes5 %in% novel_and_ebola$gene_id]
length(lnc_de_novel)

lnc_de_ref <-DEgenes5[DEgenes5 %in% annotated_lncrnas$gene_id]
length(lnc_de_ref)

lnc_de <- c(lnc_de_ref, lnc_de_novel)


matrixDE <- assays(dds)$logCPM[ lnc_de, ][, order(as.integer(colnames(assay(dds)[ lnc_de, ])))]

rowlabels <- replace(substring(rownames(matrixDE), 1,4),substring(rownames(matrixDE), 1,4) %in% substr(ebola_ref$gene_id, 1,4), NA)
row_ha_type <- rowAnnotation(genetype =rowlabels,
                        show_annotation_name = FALSE)
labels_viral <- ifelse(rownames(matrixDE) %in% ebola_ref$gene_id, "viralRNA", "na")
row_ha <- rowAnnotation(viral =labels_viral,
                        show_annotation_name = FALSE, col = list(viral = c("na" = "white", viralRNA = "darkred")))
Heatmap(calc_z(matrixDE), column_order = order(as.integer(colnames(matrixDE))),
        show_row_names = FALSE, right_annotation = row_ha,heatmap_legend_param = list(title = "Z-Score"))


#------ on protein coding
pc_de <- DEgenes5[DEgenes5 %in% annotated_all[annotated_all$gene_biotype == "protein_coding",]$gene_id]

matrixDE <- assays(dds)$logCPM[ pc_de, ][, order(as.integer(colnames(assay(dds)[ pc_de, ])))]
labels_viral <- ifelse(rownames(matrixDE) %in% ebola_ref$gene_id, "viralRNA", "na")
row_ha <- rowAnnotation(genetype =labels_viral,
                        show_annotation_name = FALSE, col = list(genetype = c("na" = "white", viralRNA = "orange")))
Heatmap(calc_z(matrixDE), column_order = order(as.integer(colnames(matrixDE))),
        show_row_names = FALSE, right_annotation = row_ha, heatmap_legend_param = list(title = "Z-Score") )


# General GO 
names_pc_de <- unique(annotated_all[annotated_all$gene_id %in% pc_de]$gene_name)
do_go(names_pc_de)

# GO in upregulated 
tt5_pc_de <- tt5[pc_de,]
pc_up <- tt5_pc_de[tt5_pc_de$logFC > 0, ]
pc_down <- tt5_pc_de[tt5_pc_de$logFC < 0, ]

names_up <- unique(annotated_all[annotated_all$gene_id %in% pc_up$genes]$gene_name)
do_go(names_up)

names_down <- unique(annotated_all[annotated_all$gene_id %in% pc_down$genes]$gene_name)
do_go(names_down)

```














