---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r setup, include=FALSE,results='hide', message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(scater)
library(SingleCellExperiment)
library(purrr)
library(stringr)
library(mvoutlier)
library(dropbead)
library(scater)
library(stringr)
library(scran)
library(cowplot)
library(rtracklayer)
library(org.Mmu.eg.db)
library(graphics)
library(RCy3)

source("utils/sc_utils.R")
source("utils/de_utils.R")

datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_final_catalogue/rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))
result_path <- file.path("/home/luisas/Desktop/cluster/data/RObjects/results")

```

## Single Cell Analysis

------------------------
  Create Seurat Objects
------------------------

I import the data and merge them into one single Seurat object to proceed with the analysis.

Obtain 
```{r cars}
# Load the PBMC dataset
data_dir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_scRNA-Seq_complete/04_DigitalExpressionMatrix/"
files <- iterate_files(data_dir, ".dge.txt.gz")
files_umis <- files[!str_detect(files, "reads" )]

# Umis 
# all matrices files from single cell, still separated per lane
list <- map(files_umis[2:length(files_umis)], get_Seurat_object)
pbmc <- get_Seurat_object(files_umis[1])
list_names <- lapply(files_umis, function(x) str_replace_all(unlist(rev(unlist(str_split(x, pattern= "/"))[-1])[[1]]), "_", "-"))

# -------------
#   Merge
# -------------
# Merge all seurats object in one 
pbmc.big <- merge(pbmc, y = list, add.cell.ids = list_names, project = "PBMCbig")
# Check if all ids available
unique(sapply(X = strsplit(colnames(pbmc.big), split = "_"), FUN = "[", 1))
table(pbmc.big$orig.ident)

saveRDS(pbmc.big, file.path(result_path,"seurat_pbmc_rhemac10_merged.rds"))
```
##---------------------------------------------------------
##                           Cell QC 
##  Remove cells with too little or too high UMIS count 
##  Remove cell based on number of detected genes per cell
##----------------------------------------------------------
```{r pressure, results='hide', message=FALSE, warning=FALSE}
pbmc.big <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged.rds"))

dim(pbmc.big)
pbmc_sc <- as.SingleCellExperiment(pbmc.big)
pbmc_sc <- calculateQCMetrics(pbmc_sc)

# Library size
lib_size_hist <- plot_histogram( sce = pbmc_sc, qc_metric = "total_counts", title = "Library Size (total UMI)", log_scale = FALSE)

# Number of detected genes
n_detec_hist <- plot_histogram(sce = pbmc_sc,  qc_metric = "total_features_by_counts",title = "Number of detected genes", log_scale = FALSE)

lsizeplot <- lib_size_hist +
  geom_vline(xintercept = c(1000, 10000), linetype = "dashed", color = "red")+ xlim(0,20000)


lsizeplot <- lib_size_hist +
  geom_vline(xintercept = c(1000, 10000), linetype = "dashed", color = "red") + xlim(0,2000)

ndectplot <- n_detec_hist +
  geom_vline(xintercept = c(600, 2000), linetype = "dashed", color = "red")

lsizeplot
ndectplot
```
# Filtering 

```{r}
pbmc <- pbmc_sc
keep_cells <-
  pbmc$total_counts > 1000 & pbmc$total_counts < 10000 &
  pbmc$total_features_by_counts > 600 & pbmc$total_features_by_counts < 2000 
table(keep_cells)
pbmc <- pbmc[, keep_cells]
saveRDS(pbmc,file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellqc.rds"))
```


##----------------------------
##           Gene QC 
##  Exclude genes not showing up in a least 10 cells
##----------------------------  
Let us compute the number of cells that each gene has at least 1 UMI.
We see two peaks, the firt one of which corresponds to lowly expressed genes. As explained in [Luecken MD et al.](https://www.embopress.org/doi/pdf/10.15252/msb.20188746): "a guideline to setting this threshold is to use the minimum cell cluster size that is of interest and leaving some leeway for dropout effects". As we will not rely on clusters that have fewer than 15 cells, we will use it as a filter:

```{r}
pbmc <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellqc.rds"))

# CLUSTER !!! 
# Do not delete this lines. The following steps were performed on the cluster for memory issues.
#n_cells <- rowSums(as.matrix(counts(pbmc)) > 0)
#pbmc <- pbmc[n_cells > 10, ]
#saveRDS(pbmc, file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellandgeneqc.rds")
pbmc <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellqc_aftercellandgeneqc.rds"))

```

##-------------------------------------------------------------------------------
##    Doublet Detection: Scrublet 
##-------------------------------------------------------------------------------


```{r cars}
library(reshape2)
pbmc <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellqc_aftercellandgeneqc.rds"))
pbmc <- as.Seurat(pbmc)
# Extract count matrix
m <-as.matrix(GetAssayData(object = pbmc, slot = "counts"))
sparse.gbm <- Matrix(m , sparse = T )
writeMM(obj = sparse.gbm, file=file.path(result_path,"sparse.mtx"))
```

# External Script Warning!! 
# --> Run Scrublet in Python

```{r cars}
cell_mask_scrublet <- read.table("/home/luisas/Desktop/cluster/proj/code/ebola/src/scripts/Analysis/results/predicted_doublet_mask_rhemac10.txt")
doublet_scores_scrublet <- read.table("/home/luisas/Desktop/cluster/proj/code/ebola/src/scripts/Analysis/results/predicted_doublet_scores_rhemac10.txt")

table(cell_mask_scrublet)
list <- as.list(as.character(cell_mask_scrublet$V1))
mask <- list == "True"
new <- pbmc[,!mask]
length(mask)
new <- as.Seurat(new)
dim(new)
table(mask)
pbmc<- seurat_preprocess_standard(new)

# Now check which cells have less than X genes expressed --> cannot see anything special :/
cells <-  colnames(pbmc[,pbmc$total_features_by_counts > 1000])
table(pbmc$total_counts < 2000)
cells <- colnames(pbmc[,pbmc$total_counts > 5000])
plot_cells(pbmc, cells)
DimPlot(pbmc, reduction = "umap", group.by = "sample")

saveRDS(new, file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet.rds"))
```

##-------------------------------------------------------------------------------
##                         Stimulated data analysis 
##-------------------------------------------------------------------------------
```{r cars}
# -----------------------------------
#         LncRNAs
# -----------------------------------
# Extract known lncrnas 

mrna_ref_id <- unique(ref$gene_name[ref$gene_biotype == "protein_coding"])
lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype == "lncRNA"])
ribodepl_genes <- rownames(immune.combined)[grepl("ribodepl*", rownames(immune.combined))]
polya_genes <- rownames(immune.combined)[grepl("polya*", rownames(immune.combined))]
novel_lncrnas <- c(ribodepl_genes, polya_genes)
annotated_lncrna <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(lnc_ref_id, "_" , "-")]
all_lncrnas <- c(annotated_lncrna, novel_lncrnas)




plot_genes_expression_nrgenes_thresholds(immune.combined , polya_genes, threshold_expression = 3, min_nr_genes = 1, "Cells expressing novel lncRNAs (from polyA data)", cond = "media")
length(unique(polya_genes))

plot_genes_expression_nrgenes_thresholds(immune.combined, ribodepl_genes, threshold_expression = 3, min_nr_genes = 1, "Cells expressing novel lncRNAs (from ribo-depleted data)")
length(unique(ribodepl_genes))

plot_genes_expression_nrgenes_thresholds(immune.combined, novel_lncrnas, threshold_expression = 3, min_nr_genes = 1,"Cells expressing all novel lncRNAs")
length(unique(novel_lncrnas))
plot_genes_expression_nrgenes_thresholds(immune.combined, annotated_lncrna,threshold_expression = 3, min_nr_genes = 1, "Cells expressing annotated lncRNAs")
length(unique(annotated_lncrna))
plot_genes_expression_nrgenes_thresholds(immune.combined, all_lncrnas, threshold_expression = 3, min_nr_genes = 1, "Cells expressing whole set of lncRNAs (annotated and novel)")
length(unique(all_lncrnas))
plot_genes_expression_nrgenes_thresholds(immune.combined, all_lncrnas, threshold_expression = 3, min_nr_genes = 1, "Cells expressing novel lncRNAs (from polyA data)", cond = "media")

plot_genes(immune.combined, novel_lncrnas,3, col = "darkred", "Ebola RNA expression")

# --------------------
#     EbolaGenes
#--------------------
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unkown", sep ="-")]
ebola_genes
plot_genes(immune.combined, ebola_genes,3, col = "darkred", "Ebola RNA expression")

plot_genes_expression_nrgenes_thresholds(immune.combined, ebola_genes, threshold_expression = 3, min_nr_genes = 1, "Cells expressing ebola genes")

```


# ---- Integration
# Running integration on the cluster 

# -------------------------------------
#     After integration in cluster 
# -------------------------------------

```{r cars}

# Normalize 
immune.combined <- NormalizeData(immune.combined)
immune.combined <- ScaleData(immune.combined)
immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)
immune.combined <- RunUMAP(immune.combined, reduction = "pca", dims = 1:20)

# Check the ones without integration for the moment 
#immune.combined <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet_after-integration.rds"))
# Import the newly computed cellswith 5k features and 
immune.combined <- readRDS(file.path(result_path, "seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet_dim_red.rds"))
DefaultAssay(immune.combined) <- "RNA"

immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "-")[[1]][1]))))
# ----------------------------
# FIND MARKERS  for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(immune.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
saveRDS(pbmc.markers, file.path(result_path,"markers_rhemac10_2k.rds"))
pbmc.markers <- readRDS(file.path(result_path,"markers_rhemac10_2k.rds"))
pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)

# ----------------------------
# Visualize the clusters
p4 <- DimPlot(immune.combined, reduction = "umap", group.by = "cond")
p5 <- DimPlot(immune.combined, reduction = "umap", group.by = "hour")
p6 <- DimPlot(immune.combined, reduction = "umap", group.by = "sample")
p8 <- DimPlot(immune.combined, reduction = "umap", group.by = "individual")
p7 <- DimPlot(immune.combined, reduction = "umap", label = TRUE)
p4
p5
p6
p7
p8

# Highlight marker genes
myeloid <- c("LYZ", "PSAP", "CCR2", "PPBP") 
monocytes <- c("CD14", "CD16")
b <- c( "CD79B", "MS4A1")
t <- c("CREM","IL7R","TRIB2")
t<-c("CD3","CD4", "CD8")
nk <- c("GNLY", "GZMB", "GZMK")
nk <- c("CD61", "CD16")
nk <- c("KLRB1")
neutrophil <- c("CD177", "SOD2")
monocytes <- c( "LYZ", "CTSB")
t <- c("CD3D", "IL7R")


immune.
DotPlot(immune.combined, features = rev(marker.genes), cols = c("blue", "red", "purple"), dot.scale = 8, 
    split.by = "cond") + RotatedAxis()

plot_genes(immune.combined,c("CD14", "LYZ", "CTSB"), threshold = 3 , title = "Myeloid markers expression", col = "purple")
plot_genes(immune.combined, nk, threshold = 3 , title = "NK markers expression", col = "purple")
plot_genes(immune.combined, b, threshold = 2 , title = "B markers expression", col = "purple")
plot_genes(immune.combined, t, threshold = 2 , title = "T-Cell markers expression", col = "purple")

rownames(immune.combined)
# Rename identities
dim(immune.combined)
Idents(immune.combined)
immune.combined <- RenameIdents(immune.combined, `0` = "Myeloid", `1` = "T", `2` = "T", 
    `3` = "NK", `4` = "B", `5` = "B", `6` = "T", `7` = "B", `8` = "NK", `9` = "Myeloid", 
    `10` = "B", `11` = "NK", `12` = "B", `13` = "NK",`14` = "NK", `15` = "NK", `16` = "B", `17` = "T" ,`18` = "Myeloid", `19` = "Myeloid",  `20` = "B")

# Sctransform 2k stim
immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "NK", `2` = "B", 
    `3` = "T", `4` = "B", `5` = "Myeloid", `6` = "B", `7` = "B", `8` = "Myeloid", `9` = "NK", 
    `10` = "T", `11` = "B", `12` = "T", `13` = "NK",`14` = "Myeloid", `15` = "T", `16` = "Myeloid", `17` = "Myeloid" ,`18` = "Myeloid")

# Sctranform 5k stim
immune.combined <- RenameIdents(immune.combined, `0` = "Myeloid", `1` = "T", `2` = "T", 
    `3` = "B", `4` = "NK", `5` = "B", `6` = "B", `7` = "T", `8` = "Myeloid", `9` = "NK", 
    `10` = "Myeloid", `11` = "T", `12` = "B", `13` = "NK",`14` = "NK", `15` = "Myeloid", `16` = "B", `17` = "T" ,`18` = "Myeloid", `19` = "Myeloid", `20` = "T")

# Sctransform 2k cond
immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "NK", `2` = "T", 
    `3` = "NK", `4` = "B", `5` = "Myeloid", `6` = "B", `7` = "B", `8` = "B", `9` = "Myeloid", 
    `10` = "Myeloid", `11` = "Myeloid", `12` = "B", `13` = "T",`14` = "Myeloid", `15` = "Myeloid")
DimPlot(immune.combined, label = TRUE)


markers.to.plot <- c(monocytes,b, nk,t)
DotPlot(immune.combined, features = rev(markers.to.plot), cols = c("blue", "red", "green"), dot.scale = 8, 
    split.by = "cond") + RotatedAxis()


VlnPlot(immune.combined, features = c("ribodepl-MSTRG.72510-unkown"), slot = "counts", log = TRUE)

ref$gene_id[grepl("ribodepl_MSTRG.72510",ref$gene_id)]
table(ref$gene_id == "ribodepl-MSTRG.72510_unkown")


FeaturePlot(immune.combined, features = "MX2")



```

# ------------------------------------
# Differential expression analysis
# ------------------------------------



# Try differential expression analysis on each celltype 

```{r cars}

# Add line to merge irrad and CTRL ( infected vs not infected )
immune.combined$stim  <- str_replace(str_replace(immune.combined$cond, "media", "irrad"), "irrad", "CTRL")
table(immune.combined$stim) 

# Rename identities with condition info
immune.combined$celltype.stim <- paste(Idents(immune.combined), immune.combined$cond, sep = "_")
immune.combined$celltype <- Idents(immune.combined)
Idents(immune.combined) <- "celltype.stim"
table(Idents(immune.combined))

# Select genes to test for DE 
ref_lnc <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(lnc_ref_id, "_" , "-")]
polya_genes <- rownames(immune.combined)[grepl("poly*", rownames(immune.combined))]
ribodepl_genes <- rownames(immune.combined)[grepl("ribodepl*", rownames(immune.combined))]
all_lnc <- c(ref_lnc, polya_genes, ribodepl_genes)
length(ref_lnc)
mrna_to_test <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(mrna_ref_id, "_" , "-")]

# Differential expresion analysis on a subset of genes 
cell_type <- "Myeloid"
live <- paste(cell_type,"live", sep ="_")
media <- paste(cell_type,"media", sep="_")
irrad <- paste(cell_type, "irrad", sep="_")
lnc_myeloid_live_irrad <- FindMarkers(immune.combined, ident.1 = live, ident.2 = irrad, verbose = FALSE, features =all_lnc)
lnc_myeloid_live_media <- FindMarkers(immune.combined, ident.1 = live, ident.2 = media, verbose = FALSE, features =all_lnc)
lnc_myeloid_irrad_media <- FindMarkers(immune.combined, ident.1 = irrad, ident.2 = media, verbose = FALSE, features =all_lnc)
nrow(lnc_myeloid_live_irrad)
nrow(lnc_myeloid_live_media)
nrow(lnc_myeloid_irrad_media)

rownames(lnc_myeloid_live_media)
rownames(lnc_myeloid_live_irrad)
rownames(lnc_myeloid_irrad_media)

de_lnc <- unique(c(rownames(lnc_myeloid_live_irrad),rownames(lnc_myeloid_live_media),rownames(lnc_myeloid_irrad_media)))

length(unique(de_lnc))
markers.to.plot <- de_lnc
p1 <- DotPlot(myeloids, assay ="integrated" , features = rev(markers.to.plot), cols = c("blue", "red", "green", "pink", "yellow", "black"), dot.scale = 4, 
    split.by = "sample") + RotatedAxis() +
  theme(axis.text.x=element_text(size=rel(0.5), angle=70), axis.text.y=element_text(size=rel(0.5)))
p1
myeloids <- subset(immune.combined, idents = "Myeloid")


```


```{r a}
library(VennDiagram)
grid.newpage()
vd_draw(list(rownames(lnc_myeloid_live_media),rownames(lnc_myeloid_irrad_media), rownames(lnc_myeloid_live_irrad)), c_names = c("lnc_myeloid_live_media", "lnc_myeloid_irrad_media", "lnc_myeloid_live_irrad"))
immune.combined@assays$
```
# -------------------------------------------
# -------------------------------------------
# Analysis cis-action/coregulation 

```{r cis-regulation}
# 1) For each gene lncrna find neighbour protein coding genes. ( either overlapping or in close region) --> putative pairs
# 2) Filter: Check the pearson correlation of the identified pairs, reatain the ones w/ |pearson correlation coefficient| > 0.6
# ==> Only if something nice comes out 
# 3) Validation: Simulate random pairs 
# 4) Validation: Compare the correlation of the ones identified with the random ones. 
library(zeallot)


test <-immune.combined[,grepl( "Myeloid",Idents(immune.combined))]
dim(immune.combined)
dim(test)

# extract the count matrix 
count_matrix <- log2(immune.combined@assays$RNA@counts+.5)
count_matrix <- test@assays$RNA@counts

mrnas_ranges <- ref[ref$gene_name %in% mrna_ref_id  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unkown", "",all_lncrnas))]
# chek that all ranges were correctly retrieved
ref$gene_id[grepl("polya*", ref$gene_id)]
length(unique(mrnas_ranges$gene_id))
length(unique(get_gene_only(lnc_ranges)$gene_id))
lnc_ranges$gene_id
# c(pairs, correlations) %<-% calc_correlation_lists(de_lncrnas,de_mrnas[1:100], count_matrix )
# mix <-cross2(ebola_genes,lnc_ref_id[1:100])
# correlations <- lapply(mix, function(x) calc_correlation_with_names(x[1],x[2],count_matrix))
# length(mix)
# length(correlations)
# mix[corre]

# Add missing gene entries 
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(c,gr_list)

lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)
length(unique(lnc_ranges$gene_id))
length(unique(de_lnc))

lnc_ranges_complete
# ----------------------
# Find pairs allowing 10K gap on both sides
pairs <- findOverlapPairs(lnc_ranges_complete, get_gene_only(mrnas_ranges), maxgap=5000)
# Get a list of ranges objects in pairs
pairs_list <- zipup(pairs)
length(pairs_list)

# How many lncrnas in total vs how many are in pairs
length(unique(lnc_ranges$gene_id)); length(unique(mrnas_ranges$gene_id)); length(pairs)

# Calc correlation values

correlations <- lapply(pairs_list, function(x) calc_correlation(x, count_matrix, type = "pearson"))
correlations[is.na(correlations)] <- -1.2
correlations
threshold = 0.1
filtered_pairs_list <- pairs_list[correlations > threshold]
length(filtered_pairs_list)

df <- as.data.frame(unlist(correlations)); names(df) <- c("corr")
ggplot(df, aes(x=corr)) + geom_density()


lnc_ranges[lnc_ranges$gene_id == "polya_MSTRG.30006"]

```

# --------------- Visualization ------------------- 

```{r network plot}
# ------------------------------------------
#   Cytoscape network visualization 
# -----------------------------------------
# Create nodes and edges 
source <- vector(); target <- vector(); all <- vector(); categories <- vector()

lapply(filtered_pairs_list, function(x) create_nodes_and_edges(x, source, target,all, categories))
nodes <- data.frame(id=all, group=categories, stringsAsFactors=FALSE)
edges <- data.frame(source=source, target=target, stringsAsFactors=FALSE)
RCy3::createNetworkFromDataFrames(nodes,edges, title="my", collection="ex")
```

```{r goEnrichment}
# -------------------------
#   GO enrichment of the cis affected mrnas
# -------------------------
mrna_cis <- unique(target)
mrna_cis <- mrna_cis[mrna_cis %in% keys(org.Mmu.eg.db, keytype="SYMBOL")]
do_go(mrna_cis)
```

# ---------------------- WCGNA test ------------------------

```{r wcgna}
source("/home/luisas/Desktop/RUN-WGCNA-master/WGCNA_functions.R")
library(reshape2)
library(ggplot2)
options(stringsAsFactors = FALSE);
enableWGCNAThreads()
dat <- read.table(file=unz("/home/luisas/Desktop/RUN-WGCNA-master/exprsExtimates.csv.zip",filename="exprsExtimates.csv"), header=TRUE, row.names=1, sep=",", )

immune.combined <- readRDS(file.path(result_path,"seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet.rds"))
count_matrix <- immune.combined@assays$RNA@counts
dim(count_matrix)
dat <- as.matrix(count_matrix[1:5000,])

determineSoftPowerWGCNA(data1=dat, outFile="powerPlots.png", propGenes=.1)
net <- runWGCNA(data1=dat, propGenes=.1, softPower=3, signedNetwork=TRUE)
plotModulesCut(referenceDataset=net, outFile="modClusterPlot.pdf")
e1 <- calculateModuleEigengenes(referenceDataset=net, split=2)

# Select the groups
groups <- gsub('-a.*dge.*$', '', x=rownames(e1))
groups <- t(data.frame(strsplit(groups, split="-", fixed=TRUE)))


groups <-groups[,c(2,5)]
groups[ ,2] <- gsub('H00', '', groups[ ,2])
rownames(groups) <- NULL
colnames(groups) <- c("Cond", "Hour")
e1 <- cbind(e1, groups)
# Melt
melted <- melt(data=e1, id.vars=c("Cond", "Hour"))
melted <- melted[order(as.numeric(melted$Hour)), ]
melted$variable <- gsub("ME", "", melted$variable)
# Plot
png(file="eigengenBoxplot.png")
q <- qplot(data=melted, y=value, x=Hour, facets=Cond~variable, 
           geom=c("boxplot", "point"), ylab="Eignegene expression",
           colour=Hour)
q
dev.off()

# Find Hub Genes
hubs <- moduleHubGenes(referenceDataset=net, MEs=calculateModuleEigengenes(referenceDataset=net, split=2), nGenes=10, split=2)
hubs[ ,colnames(hubs) == "blue"]


melted
```

```{r wcgna}


```

# -------------------------------------------
# -------------------------------------------


```{r cars}
library(knitr)
library(SummarizedExperiment)
library(edgeR)
library(geneplotter)
library(xtable)
library(org.Hs.eg.db)
library(GSVAdata)
library(GSVA)
data(c2BroadSets)
gsc <- GeneSetCollection(c2BroadSets)

length(gsc) # number of gene sets in the collection
head(names(gsc)) # some of the gene sets in the collection

mapIdentifiers(gsc, EntrezIdentifier())

Im <- incidence(gsc)
Im[1:2, 1:10]

Im <- Im[, colnames(Im) %in% rownames(all_DEgenes)]
dim(Im)
coadse.filt <- coadse.filt[colnames(Im), ]
dim(coadse.filt)
dge.filt <- dge.filt[colnames(Im),]
dim(dge.filt)

metadata(dds)$annotation
gsc <- mapIdentifiers(gsc, AnnoOrEntrezIdentifier(metadata(dds)$annotation))

```







### Identify highly expressed genes
As a validation, let us check whether the highest expressed genes are housekeeping genes like ACTB:

```{r}
plotHighestExprs(immune.combined)
```

# Normalization
To normalize for differences in library size and RNA composition, we will use the `scran` package:

```{r}
library(scran)
pbmc <- computeSumFactors(pbmc)
summary(sizeFactors(pbmc))
pbmc <- normalize(pbmc)
assays(pbmc)
logcounts(pbmc)[1:6, 1:6]
plot(sizeFactors(pbmc) ~ pbmc$total_counts)
```

## Save

```{r}
saveRDS(pbmc, file.path(result_path,"external_dataset_SCE.rds"))
```

# Seurat

```{r}

# Seurat pre-processing pipeline
pbmc_seu <- FindVariableFeatures(immune.combined)
pbmc_seu <- ScaleData(pbmc_seu)
pbmc_seu <- RunPCA(pbmc_seu)
ElbowPlot(pbmc_seu)
pbmc_seu <- FindNeighbors(pbmc_seu, reduction = "pca", dims = 1:4)
pbmc_seu <- FindClusters(pbmc_seu, resolution = 0.1)
pbmc_seu <- RunUMAP(pbmc_seu, reduction = "pca", dims = 1:4)
pbmc_seu <- RunTSNE(pbmc_seu, dims = 1:4)

DimPlot(pbmc_seu, reduction = "umap")

DefaultAssay(pbmc_seu) <- "RNA"
markers <- FindAllMarkers(
  pbmc_seu, 
  features = pbmc_seu@assays$RNA@var.features, 
  min.pct = 0.25
)

library(dplyr)
map(unique(markers$cluster), function(k) {
  markers %>% 
    dplyr::filter(cluster == k) %>% 
    head(30) %>% 
    select("cluster", "gene", "avg_logFC")
})
# Plot markers 


markers.to.plot <- c("ENSMMUG00000020171--GZMK", "ENSMMUG00000006222--ITM2C")
DotPlot(immune.combined, features = rev(markers.to.plot), cols = c("blue", "red", "green"), dot.scale = 8) + RotatedAxis()
# Save Seurat
saveRDS(pbmc_seu, file.path(result_path,"external_dataset_Seurat.rds"))
pbmc_seu <- readRDS(file.path(result_path,"external_dataset_Seurat.rds"))
```

Find markers:

```{r}
markers <- FindAllMarkers(
  pbmc_seu, 
  features = pbmc_seu@assays$RNA@var.features, 
  min.pct = 0.25
)

library(dplyr)
map(unique(markers$cluster), function(k) {
  markers %>% 
    dplyr::filter(cluster == k) %>% 
    head(30) %>% 
    select("cluster", "gene", "avg_logFC")
})
```

Judging by the previous markers, we can annotate the clusters as follows:

Cluster ID | Markers       | Cell Type
-----------|---------------|----------
0          | IL7R          | CD4 T cells
2          | GNLY, NKG7    | Natural Killer (NK)
3          | LYZ, S100A8   | CD14+ Monocytes
4          | MS4A1  | B cells

We can visualize these markers in UMAP space:

```{r}
clust_markers <- c("ENSMMUG00000001260--MS4A1", "ENSMMUG00000001676--IL7R", "ENSMMUG00000047204--GNLY", "ENSMMUG00000008987--LYZ")
FeaturePlot(pbmc_seu, features = clust_markers, reduction = "umap")
```

Finally, we can label clusters to its cell type:

```{r}
new_cluster_ids <- c("CD4 T", "NK", "CD14+ Mono",  "B")
names(new_cluster_ids) <- levels(pbmc_seu)
pbmc_seu <- RenameIdents(pbmc_seu, new_cluster_ids)
DimPlot(pbmc_seu, reduction = "umap")
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
