---
title: "Correaltion"
author: "Luisa Santus"
date: "3/31/2020"
output: html_document
---


```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr);library(Seurat); library(Matrix); library(scater); library(SingleCellExperiment); library(purrr); library(stringr)
library(mvoutlier); library(dropbead); library(scater); library(stringr); library(scran); library(cowplot); library(rtracklayer)
library(org.Mmu.eg.db); library(graphics); library(RCy3)
source("utils/sc_utils.R"); source("utils/de_utils.R")
datadir <- "/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq"
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_Ebola-RNASeq_all/03_novel_lncrnas/02_final_catalogue/rheMac10_EBOV-Kikivit_and_novelcatalogue_with_names.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))
result_path <- file.path("/home/luisas/Desktop/cluster/data/RObjects/results")
library(data.table)

immune.combined <- readRDS(file.path(result_path, "seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet_dim_red.rds"))
DefaultAssay(immune.combined) <- "RNA"
immune.combined$individual <- unlist(lapply(immune.combined$orig.ident, function(x) paste(unlist(str_split(x, "-")[[1]][1]))))

mrna_ref_id <- unique(ref$gene_name[ref$gene_biotype == "protein_coding"])
lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype == "lncRNA"])
ribodepl_genes <- rownames(immune.combined)[grepl("ribodepl*", rownames(immune.combined))]
polya_genes <- rownames(immune.combined)[grepl("polya*", rownames(immune.combined))]
novel_lncrnas <- c(ribodepl_genes, polya_genes)
annotated_lncrna <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(lnc_ref_id, "_" , "-")]
all_lncrnas <- c(annotated_lncrna, novel_lncrnas)
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% str_replace_all(mrna_ref_id, "_" , "-")]

# Change Identifiers 
immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "Myeloid", `2` = "NK", 
    `3` = "B", `4` = "T", `5` = "T", `6` = "B", `7` = "Myeloid", `8` = "B", `9` = "B", 
    `10` = "Myeloid", `11` = "B", `12` = "T", `13` = "Myeloid",`14` = "Myeloid", `15` = "T", `16` = "NK")
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unkown", sep ="-")]
```

```{r pressure, echo=FALSE}

infected <- get_infected_cells(immune.combined, ebola_genes)
infected
```

```{r correlation analysis}
# Calculate number of genes per cell
get_n_genes <- function(count_matrix, threshold = 0 ){
  n_genes <- as.vector((Matrix::colSums(count_matrix>threshold)))
  df_new <- data.frame(n_genes = n_genes)
  df_new$cell_id <- colnames(count_matrix)
  return(df_new)
}
# Per Gene Compute Linear Model and compute Adjusted P-Values
calc_linear_model <- function(gene1, count_matrix,individual,viral_load, n_genes ){
  gene1 <- str_replace_all(gene1, "_" , "-")
  exp <- count_matrix[rownames(count_matrix)== gene1,]
  model1 <- lm(exp ~ viral_load +n_genes+ individual)
  return(model1)
}

library(qvalue)
library(broom)

library(org.Hs.eg.db)
do_go <- function(list, mmu = FALSE){
  if(mmu){
      ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Mmu.eg.db, keyType = 'SYMBOL',ont = "BP")
      print("mu")
  }else{
    ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Hs.eg.db, keyType = 'SYMBOL',ont = "BP")
  }
  
  
  return(list(clusterProfiler::dotplot(ego),barplot(ego, showCategory=10) ))
}
```


# linear model 
```{r correlation analysis}








calc_models <- function(all_genes, infected, viral_load){
  subset <- subset(infected, features = all_genes)
  count_matrix <- subset@assays$RNA@data
  all_genes <- rownames(count_matrix)
  individual <- subset$individual
  number_genes_per_cell <- get_n_genes(count_matrix)
  models <- lapply(all_genes, function(gene1) calc_linear_model(gene1, count_matrix,individual,viral_load, number_genes_per_cell$n_genes))
  p_values <- as.vector(as.numeric(lapply(models, function(mod) glance(mod)$p.value)))
  adj.r.squared_values <- as.vector(as.numeric(lapply(models, function(mod) glance(mod)$adj.r.squared)))
  q_values <- qvalue(p_values)
  coefficients <- as.vector(unlist(lapply(models, function(mod1) as.numeric(as.data.frame(t(mod1$coefficients))$viral_load))))
  # Visualize, play around
  df <- data.frame(id = all_genes , log10qvalue =-log10(q_values$qvalues), q.val= q_values$qvalues, p.val = p_values , adj.r.squared = adj.r.squared_values, coef = coefficients )
  return(df)
}


# Calculate the viral load per cell 
ebola_genome <- subset(infected, features = ebola_genes)
count_matrix_ebola <- ebola_genome@assays$RNA@data
viral_load <- colSums(count_matrix_ebola)
# Calc models 
df <- calc_models(all_genes, infected, viral_load)
# Put Thresholds
df_not_na <- df[!is.na(df$adj.r.squared),]
df_radj <- df_not_na[df_not_na$adj.r.squared> 0.1,]
df_radj_pval <- df_radj[df_radj$p.val < 0.05,]
df_radj_pval_positive <- df_radj_pval[df_radj_pval$coef  > 0,]
df_radj_pval_negative <- df_radj_pval[df_radj_pval$coef  < 0,]
# Enrichment
do_go(df_radj_pval_negative$id)
do_go(df_radj_pval_positive$id)


# Calc models 
df_lnc <- calc_models(all_lncrnas, infected, viral_load)

df_not_na <- df_lnc[!is.na(df_lnc$adj.r.squared),]
df_radj <- df_not_na[df_not_na$adj.r.squared> 0.01,]
df_radj_pval <- df_radj[df_radj$q.val < 0.1,]
df_radj_pval_positive <- df_radj_pval[df_radj_pval$coef  > 0,]
df_radj_pval_negative <- df_radj_pval[df_radj_pval$coef  < 0,]

lowest_correlated_genes_to_lnc <- lapply(df_radj_pval_negative$id ,function(x)  get_correlated_genes(x, infected_immune.combined_no_ebola@assays$RNA@data, annotated_mrnas, "lowest"))

p1 <- do_go(unique(unlist(lowest_correlated_genes_to_lnc)))+labs(title = "Lowest", subtitle = paste0("#genes ",length(unique(unlist(lowest_correlated_genes_to_lnc))) ))
p1





```
# ----------------------------
# ----------------------------
#   Correlation analysis
# ----------------------------
# ----------------------------
```{r Helper Functions }
get_infected_cells <- function(immune.combined,ebola_genes, ident = "Myeloid", sample = "live H024", threshold = 2){
  # Identify Infected vs Bystanders
  myeloids <- subset(immune.combined, ident = ident)
  Idents(myeloids) <- myeloids$sample
  myeloids_live_24 <- subset(myeloids, ident = sample)
  
  subset_ebola <- subset(myeloids_live_24, features = ebola_genes)
  expression <-as.matrix(subset_ebola@assays$RNA@counts)
  
  viral_load <- calc_viral_load(ebola_genes, myeloids_live_24)
  infected_cells <- colnames(expression)[viral_load > threshold]
  bystanders <- setdiff(colnames(expression), infected_cells)
  length(infected_cells); length(bystanders)
  #Rename identities
  myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = infected_cells, value = 'infected')
  myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = bystanders, value = 'bystander')
  unique(Idents(myeloids_live_24))
  infected <- subset(myeloids_live_24, ident = "infected")
  return(infected)
}
calc_viral_load <- function(ebola_genes, cells){
  # Calculate the viral load per cell 
  ebola_genome <- subset(cells, features = ebola_genes)
  count_matrix_ebola <- ebola_genome@assays$RNA@counts
  count_matrix <- cells@assays$RNA@counts
  viral_counts_per_cell <- colSums(count_matrix_ebola)
  #viral_counts_per_cell <- count_matrix[rownames(count_matrix)== "NP-unkown",]
  total_counts_per_cell <- colSums(count_matrix)
  viral_load_percentage <- (viral_counts_per_cell/total_counts_per_cell)*100
  df <- data.frame(EBOV = viral_load_percentage)
  ggplot(df, aes(x = EBOV))+geom_histogram()
  return(viral_load_percentage)
}

# Calculate number of genes per cell
get_n_genes <- function(count_matrix, threshold = 0 ){
  n_genes <- as.vector((Matrix::colSums(count_matrix>threshold)))
  df_new <- data.frame(n_genes = n_genes)
  df_new$cell_id <- colnames(count_matrix)
  return(df_new)
}
calc_correlation_with_viralload <- function(gene_id, infected_immune.combined_no_ebola, viral_load_percentage){
  count_matrix <- infected_immune.combined_no_ebola@assays$RNA@data
  number_genes_per_cell <- get_n_genes(count_matrix)
  expression_gene1 <- count_matrix[rownames(count_matrix)== gene_id,]
  cor <- cor(as.numeric(expression_gene1), viral_load_percentage, method=c("spearman"))
  return(cor)
}
plot_correlation_coefficients <- function(all_genes, correlations){
  df <- data.frame(id = all_genes, cor = unlist(correlations))
  # Plot the correlation coefficients 
  lowest_df <- head(df[order(df$cor),], n =50 )
  lowest_df$type  <- "lowest"
  highest_df <- head(df[order(-df$cor),], n = 50)
  highest_df$type  <- "highest"
  df <- rbind(lowest_df, highest_df)
  p <- ggplot(df, aes(x = abs(cor), col = type)) + stat_ecdf(geom = "step")+ theme_classic()+labs(title = " |correlation coefficients| SPEARMAN", y = "") 
  return(p)
}

do_go <- function(list){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Hs.eg.db, keyType = 'SYMBOL',ont = "BP")
  clusterProfiler::dotplot(ego, , showCategory=9)
}

go_highest_lowest <- function(all_genes, correlations, subset = "highest"){
  df <- data.frame(id = all_genes, cor = unlist(correlations))
  if(subset == "highest"){
    set <- head(df[order(-df$cor),], n = 59)$id
  }else if(subset == "lowest"){
    set <- head(df[order(df$cor),], n = 50)$id
  }
  # Enrichment 
  p1 <- do_go(as.character(set))
  #+theme(axis.text.y=element_text(hjust=1, size = 5)
  return(p1)
}
```
# -----------------------------------------
#         Calculation Correlation 
# ------------------------------------------
```{r Function Inference LncRNAs}

# Get Infected Cells
infected <- get_infected_cells(immune.combined, ebola_genes, threshold = 1)
dim(infected)
# Get normalization values w/o ebola genes
infected_immune.combined_no_ebola <- subset(infected, features =setdiff(rownames(infected),ebola_genes))
#infected_immune.combined_no_ebola <- NormalizeData(infected_immune.combined_no_ebola)

# Calculate the viral load per cell 
viral_load_percentage <- calc_viral_load(ebola_genes, infected)

# Select cells in which gene can be detected 
all_genes <- rownames(infected_immune.combined_no_ebola)
correlations<- lapply(all_lncrnas, function(gene_id) calc_correlation_with_viralload(gene_id, infected_immune.combined_no_ebola, viral_load_percentage))
correlations_mrnas<- lapply(annotated_mrnas, function(gene_id) calc_correlation_with_viralload(gene_id, infected_immune.combined_no_ebola, viral_load_percentage))

plot_correlation_coefficients(all_lncrnas,correlations )

df<- data.frame(cor = unlist(correlations), type = "lncRNAs")
df2 <- data.frame(cor = unlist(correlations_mrnas), type = "mRNAs")
df <- rbind(df,df2 )
ggplot(df, aes(x = abs(cor), col = type, fill =  type)) + stat_ecdf(geom = "area", alpha = 0.5, size = 0.8)+ theme_classic()+labs( x= "|Correlation Coefficient|", title = " Correlation coefficients for viral load and normalized gene expression (lncRNAs) ", y = "")+xlim(0,0.3)
library(ggpubr)
ggboxplot(df[df$cor<0.59,], x = "type", y = "abs(cor)",
                color = "type", fill = "type", alpha = 0.5)+stat_compare_means(comparisons = list(c("lncRNAs", "mRNAs")))+labs(x = "", y = "|Correlation coefficient|")

# Get the most interesting from lncRNAs
df_corr <- data.frame( cor = unlist(correlations),all_lncrnas )
df_top <- df_corr[order(-(df_corr$cor)),][1:11,]
df_min <- df_corr[order((df_corr$cor)),][1:10,]
df_top$type <- "top"
df_min$type <- "min"
df <- rbind(df_top, df_min)
df$group <- as.factor(substr(df$all_lncrnas, 1 , 4))
df <- df[order(df$all_lncrnas),]
df$pos <- as.numeric(order(df$cor))
df[order(df$cor),]$pos <- 1:21
ggplot(df, aes(x = as.numeric(pos), y = cor, fill= group))+geom_bar(stat = "identity")+ylim(-0.3,0.3)+theme_classic()+scale_fill_brewer(palette= "Paired")+ theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+labs(title = "Top 10 max and min CC for lncRNAs")

# ----------------------
#    Top10
# ----------------------
gene_ids <- df_top$all_lncrnas

df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(cor(viral_load_percentage, as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data))), 2), nsmall = 2)))))

ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Inverse Correlation with Viral Load lncRNAs")+geom_line(stat = "summary_bin", binwidth = 10)+scale_color_brewer(palette = "Set3")+theme(legend.title = element_blank())+ylim(0,2)


# ----------------------
#    Negative
# ----------------------
gene_ids <- c("MX1", "MX2", "STAT1")

df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(cor(viral_load_percentage, as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data))), 2), nsmall = 2)))))

ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Inverse Correlation with Viral Load")+geom_line(stat = "summary_bin", binwidth = 10)+scale_color_brewer(palette = "Dark2")+theme(legend.title = element_blank())
# ----------------------
#    Positive
# ----------------------
gene_ids <- c( "DYNLL1", "HSPA5", "IARS")
df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(cor(viral_load_percentage, as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data))), 2), nsmall = 2)))))

ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Positive Correlation with Viral Load")+geom_line(stat = "summary_bin", binwidth = 10)+scale_color_brewer(palette = "Dark2")+theme(legend.title = element_blank())

```



```{r Function Inference LncRNAs}
# Go enrichment 
p1 <- go_highest_lowest(all_genes, correlations, subset = "highest")
p1 +theme(axis.text.y=element_text(hjust=1, size = 7))
p2 <- go_highest_lowest(all_genes, correlations, subset = "lowest")
p2 +theme(axis.text.y=element_text(hjust=1, size = 8))


# ----- LNC ONLY 
# Calculate the correlation only with lnc
subset_lnc <- subset(infected, features = all_lncrnas)
count_matrix <- subset_lnc@assays$RNA@data
all_genes <- rownames(count_matrix)
correlations_lnc <- lapply(all_genes, function(x) get_correlated_genes(x[1],viral_load_percentage,count_matrix))
plot_correlation_coefficients(all_genes,correlations_lnc )

#  transcriptional regulator ENSMMUG00000058644 0.1702529
subset_lnc_genes <- all_genes[!is.na(correlations_lnc) & correlations_lnc >0.1 ]

get_correlated_genes <- function(gene_id, count_matrix, genes, type, n_closest_cor = 20){
  expression_gene1 <- count_matrix[rownames(count_matrix)== gene_id,]
  correlations_with_genes <- unlist(lapply(genes, function(gene_id) as.numeric(cor(expression_gene1,  count_matrix[rownames(count_matrix)== gene_id,], method = c("spearman")))))
    # top 10 highest 
  if(type == "lowest"){
    lowest_genes <- genes[order(correlations_with_genes)][1:n_closest_cor]
    lowest_corr <- correlations_with_genes[order(correlations_with_genes)][1:n_closest_cor]
    print(type)
    print(lowest_corr)
    return(lowest_genes)
  }else if (type == "highest"){
    # top 10 lowest 
    print(type)
    highest_genes <- genes[order(-correlations_with_genes)][1:n_closest_cor]
    highest_corr <- correlations_with_genes[order(-correlations_with_genes)][1:n_closest_cor]
    return(highest_genes)
  }
}

# Lowest 
lowest_correlated_genes_to_lnc <- lapply(subset_lnc_genes ,function(x)  get_correlated_genes(x, infected_immune.combined_no_ebola@assays$RNA@data, annotated_mrnas, "lowest"))

p1 <- do_go(unique(unlist(lowest_correlated_genes_to_lnc)))+labs(title = "Lowest", subtitle = paste0("#genes ",length(unique(unlist(lowest_correlated_genes_to_lnc))) ))
p1
# Highest 
highest_correlated_genes_to_lnc <- lapply(subset_lnc_genes ,function(x)  get_correlated_genes(x, infected_immune.combined_no_ebola@assays$RNA@data, annotated_mrnas, "lowest"))

p2 <- do_go(unique(unlist(highest_correlated_genes_to_lnc)))+labs(title = "Highest", subtitle = paste0("Total number of genes used for enrichment",length(unique(unlist(highest_correlated_genes_to_lnc))) ))
p2



```


