---
title: "sc test analyses"
author: "Luisa Santus"
date: "3/16/2020"
output: html_document
---

# General Expression Patterns of lncRNAs Ex Vivo

```{r include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggpubr)
library(MatchIt)
robjectsdir<- "/home/luisas/Desktop/cluster/data/RObjects_old/"
source("../utils/02_sc_utils.R")

# Old version
ref<- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list_2/OLD_rheMac10_EBOV_and_novel_genenames.gtf"))

ebola_ref <- rtracklayer::import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/RObjects_old/results/seurat_pbmc_rhemac10_merged_aftercellandgeneqc_afterScrublet_dim_red.rds")

lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype %in% c("lncRNA")])

# Annotated lncRNAs: remove the gene names so it can match 
annotation_with_no_genename <- unlist(lapply(lnc_ref_id, function(x) str_split(x, "_")[[1]][1]))
annotated_lncrnas <- rownames(immune.combined)[ unlist(lapply(rownames(immune.combined), function(x) str_split(x, "-")[[1]][1])) %in% annotation_with_no_genename]
annotated_lncrna_gene_names_only <- lnc_ref_id[!grepl("_unkown",lnc_ref_id)] [lnc_ref_id[!grepl("_unkown",lnc_ref_id)] %in%  rownames(immune.combined)]
complete_set_annotated_lncrnas_in_object <- c(annotated_lncrnas, annotated_lncrna_gene_names_only)
print("done")
```

# Visualize how many lnc and mrnas are present in the object 

```{r check}
# ------------------------------------
#     mRNAs present in object 
# -------------------------------------
annotated_mrnas <- rownames(immune.combined)[rownames(immune.combined) %in% unique(ref$gene_name[ref$gene_biotype == "protein_coding"])]
length(unique(annotated_mrnas))
# ------------------------------------
#     LncRNAs present in object 
# -------------------------------------
# annotated lncrnas id 
#lnc_ref_id <- unique(ref$gene_name[ref$gene_biotype %in% c("lncRNA")])
# Annotated lncRNAs: remove the gene names so it can match 
# complete_set_annotated_lncrnas_in_object <- rownames(immune.combined)[rownames(immune.combined) %in% lnc_ref_id]

# Novel lncRNAs
ribodepl_genes <- rownames(immune.combined)[grepl("ribodepl*", rownames(immune.combined))]
polya_genes <- rownames(immune.combined)[grepl("polya*", rownames(immune.combined))]
novel_lncrnas <- c(ribodepl_genes, polya_genes)


# Complete Set 
all_lncrnas <- c(complete_set_annotated_lncrnas_in_object, novel_lncrnas)
# Remove gene overlapping EBOV genome
all_lncrnas <- all_lncrnas[!grepl("ribodepl-MSTRG.72510-unkown",all_lncrnas)]

# Check how big are the individual sets of genes
length(unique(complete_set_annotated_lncrnas_in_object)); length(unique(novel_lncrnas))
length(unique(all_lncrnas))

# ------------------------------------------------------------
# Visualize number of lncRNAs with donut plot
# ------------------------------------------------------------
type <- as.character(c("mRNAs", "lncRNAs", "lncRNAs"))
subtype <- c("annotated", "annotated", "novel")
count <- as.numeric(c(length(unique(annotated_mrnas)), length(unique(complete_set_annotated_lncrnas_in_object)), length(unique(novel_lncrnas))))

df.donut <-data.frame(type = type, subtype = subtype, count= count)
aggregated_df.donut <- aggregate(df.donut$count, by=list(type=df.donut$type), FUN=sum)
names(aggregated_df.donut) <- c("type", "count")
donut_plot(aggregated_df.donut, "Dark2")
lnc_donut <- df.donut[df.donut$type == "lncRNAs",]
names(lnc_donut) <- c("-", "type", "count")
donut_plot(lnc_donut, "Greens")
```

```{r visualize}
# ----------------------------
#           Visualize
# ----------------------------
p1 <- DimPlot(immune.combined, reduction = "umap", group.by = "cond")+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
p2 <- DimPlot(immune.combined, reduction = "umap", group.by = "hour")+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
p3 <- DimPlot(immune.combined, reduction = "umap", group.by = "sample", pt.size  = 0.01, cols = brewer.pal(8, "Set1"))+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
#p4 <- DimPlot(immune.combined, reduction = "umap", group.by = "individual")+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))
p5 <- DimPlot(immune.combined, reduction = "umap", label = TRUE)+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))

p1;p2; p3; p5
```

# Batch effect investigation 

```{r batchinvestigation}
plot_genes(immune.combined,c("MX1", "MX2"),threshold= 2, title = "Monocytes marker expression", col = "purple")
FeaturePlot(immune.combined, features = c("MX1"))
```

# Markers!
```{r cars}
b <- c("CD79B", "MS4A1")
monocytes <- c("CD14", "LYZ", "CTSB", "PSAP", "SOD2")
t <- c("IL7R","CD3D")
nk <- c("GZMB", "GZMK", "KLRB1")
dc <- c("IRF8")

#pbmc.markers <- readRDS(file.path(result_path,"EXVIVO_markers_rhemac10_2k.rds"))
#pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)


plot_genes(immune.combined,monocytes,threshold= 2, title = "Monocytes marker expression", col = "purple")
plot_genes(immune.combined, nk, threshold = 3 , title = "NK markers expression", col = "purple")
plot_genes(immune.combined, b, threshold = 2 , title = "B markers expression", col = "purple")
plot_genes(immune.combined, t, threshold = 3 , title = "T-Cell markers expression", col = "purple")

# Change Identifiers 
immune.combined <- RenameIdents(immune.combined, `0` = "T", `1` = "Myeloid", `2` = "NK", 
    `3` = "B", `4` = "T", `5` = "T", `6` = "B", `7` = "Myeloid", `8` = "B", `9` = "B", 
    `10` = "Myeloid", `11` = "B", `12` = "T", `13` = "Myeloid",`14` = "Myeloid", `15` = "T", `16` = "NK")

#saveRDS(immune.combined, paste0(robjectsdir,"EXVIVO_immune.combined_idents.rds" ))


# Visualize identities 
DimPlot(immune.combined, label = TRUE, pt.size = 0.1, cols=  brewer.pal(4, "Set2")[c(1,3,4,2)])+theme_minimal()+ theme(panel.background = element_rect(fill = "white", colour = "grey50"), panel.grid.major = element_blank(), panel.grid.minor = element_blank())+theme(legend.text = element_text(size=18))


# DotPlot of the marker genes
marker.genes <- c(b, nk, monocytes, t)
DotPlot(immune.combined, features = rev(marker.genes), cols = c("blue", "red", "purple"), dot.scale = 8, split.by = "cond") + RotatedAxis()+theme(axis.text = element_text(size = 15), axis.title = element_blank())
ebola_genes <- rownames(immune.combined)[rownames(immune.combined) %in% paste(ebola_ref$gene_id, "unkown", sep ="-")]

```

# Calculate viral load 

```{r Helper Functions }
get_infected_cells <- function(immune.combined,ebola_genes, ident = "Myeloid", sample = "live H024", threshold = 2){
  # Identify Infected vs Bystanders
  myeloids <- subset(immune.combined, ident = ident)
  Idents(myeloids) <- myeloids$sample
  myeloids_live_24 <- subset(myeloids, ident = sample)
  
  subset_ebola <- subset(myeloids_live_24, features = ebola_genes)
  expression <-as.matrix(subset_ebola@assays$RNA@counts)
  
  viral_load <- calc_viral_load(ebola_genes, myeloids_live_24)
  infected_cells <- colnames(expression)[viral_load > threshold]
  bystanders <- setdiff(colnames(expression), infected_cells)
  length(infected_cells); length(bystanders)
  #Rename identities
  myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = infected_cells, value = 'infected')
  myeloids_live_24 <- SetIdent(object = myeloids_live_24, cells = bystanders, value = 'bystander')
  unique(Idents(myeloids_live_24))
  infected <- subset(myeloids_live_24, ident = "infected")
  return(infected)
}
cells <- immune.combined
calc_viral_load <- function(ebola_genes, cells){
  # Calculate the viral load per cell 
  ebola_genome <- subset(cells, features = ebola_genes)
  count_matrix_ebola <- ebola_genome@assays$RNA@counts
  count_matrix <- cells@assays$RNA@counts
  viral_counts_per_cell <- colSums(count_matrix_ebola)
  
  #viral_counts_per_cell <- count_matrix[rownames(count_matrix)== "NP-unkown",]
  total_counts_per_cell <- colSums(count_matrix)
  viral_load_percentage <- (viral_counts_per_cell/total_counts_per_cell)*100
  df <- data.frame(EBOV = viral_load_percentage, nebov = viral_counts_per_cell, ntot = total_counts_per_cell)
  df <- data.frame(EBOV = viral_load_percentage)
  ggplot(df[df$EBOV>0.9,], aes(x = EBOV))+geom_histogram()
  ggplot(df, aes(x = total_counts_per_cell))+geom_histogram()
  df[df$EBOV > 20,, drop = FALSE]
  return(viral_load_percentage)
}

# Calculate number of genes per cell
get_n_genes <- function(count_matrix, threshold = 0 ){
  n_genes <- as.vector((Matrix::colSums(count_matrix>threshold)))
  df_new <- data.frame(n_genes = n_genes)
  df_new$cell_id <- colnames(count_matrix)
  return(df_new)
}
calc_correlation_with_viralload <- function(gene_id, infected_immune.combined_no_ebola, viral_load_percentage){
  count_matrix <- infected_immune.combined_no_ebola@assays$RNA@data
  number_genes_per_cell <- get_n_genes(count_matrix)
  expression_gene1 <- count_matrix[rownames(count_matrix)== gene_id,]
  cor <- cor(as.numeric(expression_gene1), viral_load_percentage, method=c("spearman"))
  return(cor)
}

gene_id <- all_lncrnas[2]
calc_correlation_with_viralload <- function(gene_id, infected_immune.combined_no_ebola, viral_load_percentage, onlyexpressed = FALSE){
  count_matrix <- infected_immune.combined_no_ebola@assays$RNA@data
  number_genes_per_cell <- get_n_genes(count_matrix)
  expression_gene1 <- count_matrix[rownames(count_matrix)== gene_id,]
  print(gene_id)
  # Only calc correlation when 
  if(onlyexpressed == TRUE){
    # Only calculate it with the cells in which it expresses 
    mask <- expression_gene1>0
    table(mask)
    expression_gene1  <- expression_gene1[mask]
    viral_load_percentage <- viral_load_percentage[mask]
  }
  if(length(expression_gene1)>10){
  #cor <- cor(as.numeric(expression_gene1), viral_load_percentage, method=c("spearman"))
  rhos <- cor.test(as.numeric(expression_gene1),viral_load_percentage , method = "spearman")$estimate
  pvals <- cor.test(as.numeric(expression_gene1),viral_load_percentage , method = "spearman")$p.value
  df <- data.frame( genename = gene_id, rho = rhos , pvals = pvals)
  }else{
    df <- data.frame()
  }
  return(df)
}


plot_correlation_coefficients <- function(all_genes, correlations){
  df <- data.frame(id = all_genes, cor = unlist(correlations))
  # Plot the correlation coefficients 
  lowest_df <- head(df[order(df$cor),], n =50 )
  lowest_df$type  <- "lowest"
  highest_df <- head(df[order(-df$cor),], n = 50)
  highest_df$type  <- "highest"
  df <- rbind(lowest_df, highest_df)
  p <- ggplot(df, aes(x = abs(cor), col = type)) + stat_ecdf(geom = "step")+ theme_classic()+labs(title = " |correlation coefficients| SPEARMAN", y = "") 
  return(p)
}


do_go <- function(list){
  ego <- clusterProfiler::enrichGO(gene = list,OrgDb =org.Hs.eg.db, keyType = 'SYMBOL',ont = "BP")
  clusterProfiler::dotplot(ego, , showCategory=9)
}

go_highest_lowest <- function(all_genes, correlations, subset = "highest"){
  df <- data.frame(id = all_genes, cor = unlist(correlations))
  if(subset == "highest"){
    set <- head(df[order(-df$cor),], n = 59)$id
  }else if(subset == "lowest"){
    set <- head(df[order(df$cor),], n = 50)$id
  }
  # Enrichment 
  p1 <- do_go(as.character(set))
  #+theme(axis.text.y=element_text(hjust=1, size = 5)
  return(p1)
}
```



# -----------------------------------------
#         Calculation Correlation 
# ------------------------------------------
```{r Function Inference LncRNAs}
# Get Infected Cells
infected <- get_infected_cells(immune.combined, ebola_genes, threshold = 1)
dim(infected)
# Get normalization values w/o ebola genes
infected_immune.combined_no_ebola <- subset(infected, features =setdiff(rownames(infected),ebola_genes))
#infected_immune.combined_no_ebola <- NormalizeData(infected_immune.combined_no_ebola)
# Calculate the viral load per cell 
viral_load_percentage <- calc_viral_load(ebola_genes, infected)

# Select cells in which gene can be detected 
all_genes <- rownames(infected_immune.combined_no_ebola)
correlations<-  rbindlist(lapply(all_lncrnas, function(gene_id) calc_correlation_with_viralload(gene_id, infected_immune.combined_no_ebola, viral_load_percentage, onlyexpressed = FALSE)))
correlations_mrnas<- rbindlist(lapply(annotated_mrnas, function(gene_id) calc_correlation_with_viralload(gene_id, infected_immune.combined_no_ebola, viral_load_percentage)))
print("done")


saveRDS(correlations, file.path(robjectsdir, "EXVIVO_lnc_cor.rds"))
saveRDS(correlations_mrnas, file.path(robjectsdir, "EXVIVO_mrna_cor.rds"))

correlations <- readRDS(file.path(robjectsdir, "EXVIVO_lnc_cor.rds"))
correlations_mrnas <- readRDS(file.path(robjectsdir, "EXVIVO_mrna_cor.rds"))
correlations$type <- "lncRNA"
correlations_mrnas$type <- "mRNA"
correlations_complete <- rbind(correlations, correlations_mrnas)

# Calculate q Value 
library(qvalue)
correlations_complete$qval <- qvalue(correlations_complete$pvals)$qvalues


# Check which are the significant ones
significant_correlations <- correlations_complete[!is.na(correlations_complete$qval) & correlations_complete$qval < 0.1,]

# how many are down and how many are up 
significant_up <- significant_correlations[significant_correlations$rho>0,]
table(significant_up$type)
significant_up_lnc <- significant_up[significant_up$type == "lncRNA",]
significant_up_mrna <- significant_up[significant_up$type == "mRNA",]
do_go(unique(significant_up_mrna$genename))

significant_down <- significant_correlations[significant_correlations$rho<0,]
table(significant_down$type)
significant_down_lnc <- significant_down[significant_down$type == "lncRNA",]
significant_down_mrna <- significant_down[significant_down$type == "mRNA",]
do_go(unique(significant_down_mrna$genename))


significant_mrna <- rbind(significant_down_mrna, significant_up_mrna)
do_go(unique(significant_mrna$genename))

table(significant_correlations_$type)
p1 <- go_highest_lowest(significant_correlations_mrna$genename, unlist(significant_correlations_mrna$rho), subset = "highest")

p2 <- go_highest_lowest(significant_correlations_mrna$genename, unlist(significant_correlations_mrna$rho), subset = "lowest")

p2 <- go_highest_lowest(significant_correlations_mrna$genename, unlist(significant_correlations_mrna$qval), subset = "lowest")


p2

```

# Visualize protein coding vs lncrnas correlation coefficients 
```{r Function Inference LncRNAs}
library(ggpubr)
col_lnc = "#870052"
col_mrna = "#2e88bf"
palette_plot_percentage <- c(col_lnc, col_mrna)


# --------------------------------------------------------------------------------------
# Compare correlation coefficients of lncrnas and protein coding genes 
# --------------------------------------------------------------------------------------

ggboxplot(correlations_complete, x = "type", y = "abs(rho)",
                color = "type", fill = "type", alpha = 0.5)+stat_compare_means(comparisons = list(c("lncRNA", "mRNA")))+labs(x = "", y = "|Correlation coefficient|")+scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)

ggboxplot(correlations_complete, x = "type", y = "qval",
                color = "type", fill = "type", alpha = 0.5)+stat_compare_means(comparisons = list(c("lncRNA", "mRNA")))+labs(x = "", y = "Q value")+scale_fill_manual(values = palette_plot_percentage)+
                scale_color_manual(values = palette_plot_percentage)

```

# Visualize top 10 

```{r Function Inference LncRNAs}
# Get the most interesting from lncRNAs
df_corr <- significant_correlations_lnc
df_top <- df_corr[order(-(df_corr$rho)),][1:11,]
df_min <- df_corr[order((df_corr$rho)),][1:10,]
df_top$type <- "top"
df_min$type <- "min"
df <- rbind(df_top, df_min)
df$group <- as.factor(substr(df$genename, 1 , 4))
df <- df[order(df$genename),]
df$pos <- as.numeric(order(df$rho))
df[order(df$rho),]$pos <- 1:21
ggplot(df, aes(x = as.numeric(pos), y = rho, fill= group))+geom_bar(stat = "identity")+ylim(-0.3,0.3)+theme_classic()+scale_fill_brewer(palette= "Paired")+ theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+labs(title = "Top 10 max and min CC for lncRNAs")

# ----------------------
#    Top10
# ----------------------
gene_ids <- df_top$genename

df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(cor(viral_load_percentage, as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data))), 2), nsmall = 2)))))

ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Inverse Correlation with Viral Load lncRNAs")+geom_line(stat = "summary_bin", binwidth = 10)+scale_color_brewer(palette = "Set3")+theme(legend.title = element_blank())+ylim(0,2)
```

# Check example with broad 

```{r Function Inference LncRNAs}

# ----------------------
#    Negative
# ----------------------
gene_ids <- c("MX1", "MX2", "STAT1")
correlations_mrnas[correlations_mrnas$genename == "MX1", ]

df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(correlations_mrnas[correlations_mrnas$genename == gene_id, ]$rho, 2), nsmall = 2), "   pval: ",correlations_mrnas[correlations_mrnas$genename == gene_id, ]$pvals))))

ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Inverse Correlation with Viral Load")+geom_line(stat = "summary_bin", binwidth = 10)+scale_color_brewer(palette = "Dark2")+theme(legend.title = element_blank())
# ----------------------
#    Positive
# ----------------------
gene_ids <- c( "DYNLL1", "HSPA5", "IARS")
df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(correlations_mrnas[correlations_mrnas$genename == gene_id, ]$rho, 2), nsmall = 2), "   pval: ",correlations_mrnas[correlations_mrnas$genename == gene_id, ]$pvals))))

ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Inverse Correlation with Viral Load")+geom_line(stat = "summary_bin", binwidth = 5)+scale_color_brewer(palette = "Dark2")+theme(legend.title = element_blank())

```



# check overlap with in vivo 

```{r invivo_overlap}
celltype <- "Myeloid"
mast_myeloid_late <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","late",".rds")))
mast_myeloid_middle <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","middle",".rds")))
mast_myeloid_early <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","early",".rds")))

celltype <- "B"
mast_B_late <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","late",".rds")))
mast_B_middle <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","middle",".rds")))
mast_B_early <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","early",".rds")))

celltype <- "T"
mast_T_late <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","late",".rds")))
mast_T_middle <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","middle",".rds")))
mast_T_early <- readRDS(file.path(robjectsdir, paste0("MAST_invivo_model_", celltype,"_","early",".rds")))


# ------- Myeloid
celltype <- "myeloid"
mast_myeloid_late$comparison <- paste0(celltype,"_late")
mast_myeloid_middle$comparison <- paste0(celltype,"_middle")
mast_myeloid_early$comparison <- paste0(celltype,"_early")
myeloids_de <- rbind(mast_myeloid_late, mast_myeloid_middle, mast_myeloid_early)

# ------- T
celltype <- "T"
mast_T_late$comparison <- paste0(celltype,"_late")
mast_T_middle$comparison <- paste0(celltype,"_middle")
mast_T_early$comparison <- paste0(celltype,"_early")
t_de <- rbind(mast_T_late, mast_T_middle, mast_T_early)


# ------- B
celltype <- "B"
mast_B_late$comparison <- paste0(celltype,"_late")
mast_B_middle$comparison <- paste0(celltype,"_middle")
mast_B_early$comparison <- paste0(celltype,"_early")
b_de <- rbind(mast_B_late, mast_B_middle, mast_B_early)

stats_complete <- rbind( myeloids_de, t_de, b_de)

# Convert into matrix
m <-xtabs(coef~primerid+comparison, stats_complete)
attr(m, "class") <- NULL
colnames(m)
m <- m[,c("myeloid_early" ,"myeloid_middle","myeloid_late", "B_early" ,"B_middle","B_late", "T_early" ,"T_middle","T_late")]

m_significance<-xtabs(fdr~primerid+comparison, stats_complete)
attr(m_significance, "class") <- NULL
m_significance <- m_significance[,c("myeloid_early" ,"myeloid_middle","myeloid_late", "B_early" ,"B_middle","B_late", "T_early" ,"T_middle","T_late")]

# -------------------------------------------------------------------------
# Select genes
# -------------------------------------------------------------------------

FDRTHRESHOLD <- 0.1

get_de_names <- function(df_filtered, FDRTHRESHOLD = 0.1, FCTHRESHOLD = log(1.2), subset = all_lncrnas){
  # Only select lncrnas 
  df_filtered <- df_filtered[df_filtered$primerid %in% subset,]
  # Only select significant ones
  df_filtered <- df_filtered[fdr<FDRTHRESHOLD &abs(coef) > FCTHRESHOLD,]
  return(df_filtered$primerid)
}


# Select genes to plot
lnc_myeloid_late <- get_de_names(mast_myeloid_late, subset = all_lncrnas)
#intersect(bulk_de_genes_all, gsub("-", "_",gsub("-unkown", "", lnc_myeloid_late)))

lnc_myeloid_middle <- get_de_names(mast_myeloid_middle, subset = all_lncrnas)
lnc_myeloid_early <- get_de_names(mast_myeloid_early, subset = all_lncrnas)

lnc_T_late <- get_de_names(mast_T_late, subset = all_lncrnas)
lnc_T_middle <- get_de_names(mast_T_middle, subset = all_lncrnas)
lnc_T_early <- get_de_names(mast_T_early, subset = all_lncrnas)

lnc_B_late <- get_de_names(mast_B_late, subset = all_lncrnas)
lnc_B_middle <- get_de_names(mast_B_middle, subset = all_lncrnas)
lnc_B_early <- get_de_names(mast_B_early, subset = all_lncrnas)

## mrnas
mrna_myeloid_late <- get_de_names(mast_myeloid_late, subset = annotated_mrnas)
mrna_myeloid_middle <- get_de_names(mast_myeloid_middle, subset = annotated_mrnas)
mrna_myeloid_early <- get_de_names(mast_myeloid_early, subset = annotated_mrnas)

mrna_T_late <- get_de_names(mast_T_late, subset = annotated_mrnas)
mrna_T_middle <- get_de_names(mast_T_middle, subset = annotated_mrnas)
mrna_T_early <- get_de_names(mast_T_early, subset = annotated_mrnas)

mrna_B_late <- get_de_names(mast_B_late, subset = annotated_mrnas)
mrna_B_middle <- get_de_names(mast_B_middle, subset = annotated_mrnas)
mrna_B_early <- get_de_names(mast_B_early, subset = annotated_mrnas)
# ----------------------------


de_lists_lnc <- list(lnc_myeloid_late, lnc_myeloid_middle, lnc_myeloid_early,lnc_T_late, lnc_T_middle, lnc_T_early, lnc_B_late, lnc_B_middle, lnc_B_early )
names(de_lists_lnc) <- c("lnc_myeloid_late", "lnc_myeloid_middle", "lnc_myeloid_early","lnc_T_late", "lnc_T_middle", "lnc_T_early", "lnc_B_late", "lnc_B_middle", "lnc_B_early")


de_lists_mrna <- list(mrna_myeloid_late, mrna_myeloid_middle, mrna_myeloid_early,mrna_T_late, mrna_T_middle, mrna_T_early, mrna_B_late, mrna_B_middle, mrna_B_early )
names(de_lists_mrna) <- c("mrna_myeloid_late", "mrna_myeloid_middle", "mrna_myeloid_early","mrna_T_late", "mrna_T_middle", "mrna_T_early", "mrna_B_late", "mrna_B_middle", "mrna_B_early")

all_de_mrna_myeloid <- unique(c(mrna_myeloid_early, mrna_myeloid_late, mrna_myeloid_middle))

# Overlap DE and significantly correlated mrna with viral load

intersect <- intersect(significant_correlations_mrna$genename, all_de_mrna_myeloid)
length(unique(intersect))
length(unique(significant_correlations_mrna$genename))
length(all_de_mrna_myeloid)

intersect <- intersect(significant_correlations$genename, all_de_lnc_myeloid)
length(intersect)
table(significant_correlations$type)
length(all_de_lnc_myeloid)
```


```{r invivo_overlap}

# Lnc Myeloid Late

visualize_cor <- function(gene_list, correlations = correlations, threshold = 0.1){
  myeloid_late_cor <- rbindlist(lapply(gene_list ,function(gene_id) correlations[correlations$genename == gene_id, ]))
  myeloid_late_cor$sig <- ifelse(myeloid_late_cor$qval<threshold, TRUE, FALSE)
  
  # Plot only the significant ones 
  gene_ids <- myeloid_late_cor[myeloid_late_cor$qval<threshold, ]$genename
  df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = paste0(gene_id, "  ", format(round(correlations[correlations$genename == gene_id, ]$rho, 2), nsmall = 2), "   qval: ",correlations[correlations$genename == gene_id, ]$qval))))
  
  p <- ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = "Ex Vivo:  Inverse Correlation with Viral Load")+geom_line(stat = "summary_bin", binwidth = 5)+scale_color_brewer(palette = "Dark2")+theme(legend.title = element_blank())
  return(p)
}


gene_list <- all_de_lnc_myeloid
all_de_lnc_myeloid <- c(lnc_myeloid_late, lnc_myeloid_middle,lnc_myeloid_early )
visualize_cor(unique(all_de_lnc_myeloid), correlations_complete)
visualize_cor(lnc_myeloid_late, correlations_complete)
visualize_cor(lnc_myeloid_middle, correlations_complete)
visualize_cor(lnc_myeloid_early, correlations_complete)


# Overlap 
intersect(significant_correlations_lnc$genename , all_de_lnc_myeloid)

visualize_cor <- function(gene_list,title = "Ex vivo",  correlations = correlations, threshold = 0.1){
  myeloid_late_cor <- rbindlist(lapply(gene_list ,function(gene_id) correlations[correlations$genename == gene_id, ]))
  myeloid_late_cor$sig <- ifelse(myeloid_late_cor$qval<threshold, TRUE, FALSE)
  
  # Plot only the significant ones 
  gene_ids <- myeloid_late_cor[myeloid_late_cor$qval<threshold, ]$genename
  df_complete <- rbindlist(lapply(gene_ids, function(gene_id) data.frame( viral_load_percentage = viral_load_percentage, expression_gene = as.vector(as.matrix(subset(infected_immune.combined_no_ebola, features = gene_id)@assays$RNA@data)), gene_id = gene_id)))
  
  df_complete <- df_complete[df_complete$expression_gene > 0, ]
  
  p <- ggplot(df_complete, aes(x = viral_load_percentage, y = expression_gene, col =  gene_id))+theme_classic()+labs(y = "Normalized Gene Expression", x = "%Viral Load", title = title)+theme(legend.title = element_blank())+
    geom_point()+geom_smooth(se = FALSE)
  p
  return(p)
}

correlations <- correlations_complete
gene_ids <- unique(all_de_lnc_myeloid)
a <- visualize_cor(unique(all_de_lnc_myeloid), title = "EX vivo ", correlations_complete, threshold = 1)
a + xlim(0,20)
a <- visualize_cor(lnc_myeloid_late, title = "late",correlations =  correlations_complete, threshold  = 1)
a

visualize_cor(lnc_myeloid_middle, correlations_complete, threshold = 1, title = "middle")
visualize_cor(lnc_myeloid_early, correlations_complete, threshold = 1, title = "early")
dim(immune.combined)


pair <- c("THY", "ENSMMUG00000061948-unkown")
annotated_mrnas[annotated_mrnas]
all_lncrnas[grepl("ENSMMUG00000061948-unkown", all_lncrnas)]
rbind(correlations_mrnas[correlations_mrnas$genename == "THY",], )

visualize_cor(pair, correlations = )




```









