---
title: "Sc analysis"
author: "Luisa Santus"
date: "1/22/2020"
output: html_document
---

```{r a}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(Matrix)
library(SingleCellExperiment)
library(stringr)
library(scales)
library(rtracklayer)
library(ComplexHeatmap)
library(RColorBrewer)
library(scales)

source("../../utils/02_sc_utils.R")

theme_paper <- theme(legend.title = element_blank())+theme(panel.background = element_rect(fill = "white", colour = "white"))+theme(panel.background = element_rect(fill = "white", colour = "grey50"))+theme(axis.text = element_text(size = 18), axis.title = element_text(size = 20), legend.text = element_text(size = 18))

# Gene annotation 
ref <- import(file.path("/home/luisas/Desktop/cluster/data/01_bulk_RNA-Seq_lncRNAs_annotation/03_novel_lncRNAs_list/rheMac10_EBOV_and_novel_genenames.gtf"))
ebola_ref <- import(file.path("/home/luisas/Desktop/cluster/data/00_RawData/pardis_shared_data/sabeti-txnomics/shared-resources/HISAT2/EBOV-Kikwit/KU182905.1.gtf"))

# Output paths
robjectsdir <- file.path("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/")

# LncRNAs
robjectsdir_stats <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/05_stats/"
all_lncrnas <- readRDS(file.path(robjectsdir_stats, "all_lncrnas.rds"))
annotated_mrnas <- readRDS(file.path(robjectsdir_stats,"annotated_mrnas.rds"))
print("--")

immune.combined <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/03_prep/03_immune.combined.ready.rds")

de_all_genes<- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/de_all_genes.rds")
length(unique(de_all_genes))

de_mrna_all <- de_all_genes[de_all_genes %in% annotated_mrnas]
length(unique(de_mrna_all))

de_lnc_all <- de_all_genes[de_all_genes %in% all_lncrnas]
length(unique(de_lnc_all))

stats_complete <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/04_DE/stats_complete.rds")


```


# -----------------------------------
#          Compute distances 
# ------------------------------------

```{r colocation}
library(zeallot)

mrnas_ranges <- ref[ref$gene_name %in% annotated_mrnas  ]
lnc_ranges <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unknown", "",all_lncrnas))]
lnc_ranges_de <- ref[ref$gene_id %in% gsub("-", "_",gsub("-unknown", "",de_lnc_all))]
mrnas_ranges_de<- ref[ref$gene_name %in% unique(unlist(de_mrna_all))  ]


# Add missing gene entries (as novel lncrnas only prsent transcript and exon entries)
genes_lnc_ranges <- get_gene_only(lnc_ranges)$gene_id
missing_gene_lines <- unique(lnc_ranges$gene_id)[!(unique(lnc_ranges$gene_id) %in% genes_lnc_ranges)]
gr_list <- lapply(missing_gene_lines, function(x) get_gene_range(x, lnc_ranges))
missing_gene_ranges <- do.call(base::c,gr_list)
lnc_ranges_complete <-c(get_gene_only(lnc_ranges), missing_gene_ranges)

# Only select the DE interestin
lnc_ranges_complete_de <- lnc_ranges_complete[lnc_ranges_complete$gene_id %in% gsub("-", "_",gsub("-unknown", "",unique(unlist(de_lnc_all)))), ]
length(unique(lnc_ranges_complete_de$gene_id))

# --------------------------------------------------------------
#           Get 5 closest genes to each gene of interest
# --------------------------------------------------------------
# Find pairs under maximum overlap
get_all_colocations <- function(gr_gene, gr_hits =mrnas_ranges){
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  distances <- unlist(lapply(1:length(mrnas_ranges), function(index) GenomicRanges::distance(gr_gene, mrnas_ranges[index], ignore.strand = TRUE) ))
  df <- data.frame(gene=gr_gene$gene_id, distances=distances, second = p@second$gene_id, stringsAsFactors = F, chr = seqnames(gr_gene) )
  return(df)
}



get_n_closest_gr <- function(gr_gene, gr_hits =mrnas_ranges, maxgap =2000000L ){
  p <- findOverlapPairs(gr_gene, get_gene_only(gr_hits), maxgap=maxgap, ignore.strand = TRUE)
  
  print(gr_gene$gene_name)
  # Compute thedistances and select the first 5
  if(length(p@second) == 0 ){
    df <- data.frame(gene=gr_gene$gene_id, distances=NA, second = NA, chr = seqnames(gr_gene), stringsAsFactors = F )
  }else{
    distances <- unlist(lapply(1:length(p@second), function(index) GenomicRanges::distance(gr_gene, p@second[index], ignore.strand = TRUE) ))
    df <- data.frame(gene=gr_gene$gene_id, distances=distances, second = p@second$gene_id, stringsAsFactors = F, chr = seqnames(gr_gene) )
  }
  return(df)
}



get_nearest <- function(gr_gene, gr_hits =mrnas_ranges){
  p <- nearest(gr_gene, get_gene_only(gr_hits),ignore.strand = TRUE)
  if(is.na(p)){
    return(NA)
  }
  df <- data.frame(gene = gr_gene$gene_id, second =get_gene_only(gr_hits)[p]$gene_id, distance = GenomicRanges::distance(gr_gene, get_gene_only(gr_hits)[p], ignore.strand = TRUE) )
  return(df)
}
```
# Check where lncRNAs are located - compared to each other 

```{r colocation}
df_distances_de <- do.call(rbind, lapply(1:(length(lnc_ranges_complete_de)), function(index) (get_nearest(lnc_ranges_complete_de[index],gr_hits=lnc_ranges_complete_de[setdiff(1:length(lnc_ranges_complete_de), index)]))))
df_distances_de$type <- ifelse(substr(df_distances_de$gene,1,4)=="MSTR", "Novel", "Annotated")
df_distances_de <- df_distances_de[!is.na(df_distances_de$distance),]
options(scipen=999)

df_distances_de$bin <- (cut(df_distances_de$distance, breaks = c(-1,20000, 100000,500000,1000000, 2000000,10000000, Inf), labels = c("20kb", "100kb","500kb","1Mb","2Mb", "10Mb",">10Mb" )))
df_lnc_distance_count <- df_distances_de %>% group_by(bin) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100

ggplot(df_lnc_distance_count , aes(bin, y = perc))+geom_bar(stat = "identity", position = position_dodge(), col = "red", fill = "darkred")+theme_paper+ggtitle("Closest DE lncRNA")

ggplot(df_distances_de, aes(y = distance, x = type,col = type, fill = type))+geom_boxplot( alpha = 0.7)+theme_paper+scale_fill_brewer(palette = "Paired")+scale_color_brewer(palette = "Paired")

```

```{r colocation}
# Compared to DE protein coding genes 

df_distances_de <- do.call(rbind, lapply(1:(length(lnc_ranges_complete_de)), function(index) (get_nearest(lnc_ranges_complete_de[index],gr_hits=mrnas_ranges_de))))
df_distances_de$type <- ifelse(substr(df_distances_de$gene,1,4)=="MSTR", "Novel", "Annotated")
df_distances_de <- df_distances_de[!is.na(df_distances_de$distance),]

# Create Bins 
options(scipen=999)
ggplot(df_distances_de, aes(x = distance, fill = type))+geom_histogram(alpha=0.6)+theme_paper+ggtitle("Distance to closest DE mRNA")+scale_fill_brewer(palette = "Paired")
df_distances_de$bin <- (cut(df_distances_de$distance, breaks = c(-1,20000, 100000,500000,1000000, 2000000,10000000, Inf), labels = c("20kb", "100kb","500kb","1Mb","2Mb", "10Mb",">10Mb" )))

df_lnc_distance_count <- df_distances_de %>% group_by(bin) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100

ggplot(df_lnc_distance_count , aes(bin, y = perc))+geom_bar(stat = "identity", position = position_dodge(),fill = "Orange")+theme_paper+xlab("Distance to closest DE mRNA") + ylab("% DE lncRNAs")


df_distances_de_novel <- df_distances_de[df_distances_de$type == "Novel",]
df_lnc_distance_count <- df_distances_de_novel %>% group_by(bin, type) %>% tally()
df_lnc_distance_count$perc <- (df_lnc_distance_count$n/sum(df_lnc_distance_count$n))*100

df_distances_de_annot <- df_distances_de[df_distances_de$type == "Annotated",]
df_lnc_distance_count_annot <- df_distances_de_annot %>% group_by(bin, type) %>% tally()
df_lnc_distance_count_annot$perc <- (df_lnc_distance_count_annot$n/sum(df_lnc_distance_count_annot$n))*100


ggplot(rbind(df_lnc_distance_count,df_lnc_distance_count_annot ) , aes(bin, y = perc, fill = type))+geom_bar(stat = "identity", position = position_dodge())+theme_paper+xlab("Distance to closest DE mRNA") + ylab("% DE lncRNAs")+scale_fill_brewer(palette = "Paired")

```





```{r colocation}
library(org.Hs.eg.db)
library(clusterProfiler) 

# maximum 1Mb
all_lnc_2mb <- do.call(rbind,lapply(1:(length(lnc_ranges_complete_de)), function(index) (get_n_closest_gr(lnc_ranges_complete_de[index], mrnas_ranges_de, maxgap = 10000000L))))
all_lnc_2mb$type <- ifelse(substr(all_lnc_2mb$gene,1,4)=="MSTR", "Novel", "Annotated")

# How many DE pc genes they have in 2Mbp range nearby 
ggplot(all_lnc_2mb %>% group_by(gene, type) %>% tally(), aes(x=n, fill = type ))+geom_histogram()+theme_paper+scale_fill_brewer(palette = "Paired")+ggtitle("How many DE mRNAs found in 1Mb")

# ------------------------------------------------------------
# Visualize enrichment of PC genes found in those regions
# -----------------------------------------------------------
all_lnc_2mb_names <- ref[ref$gene_id %in% all_lnc_2mb$second, ]$gene_name
all_lnc_2mb_stats <- stats_complete[stats_complete$primerid %in% unlist(unique(all_lnc_2mb_names)),]
logfc <- all_lnc_2mb_stats$coef

all_lnc_2mb_names_up <- unique(all_lnc_2mb_stats[logfc>0, ]$primerid)
all_lnc_2mb_names_down <- unique(all_lnc_2mb_stats[logfc<0, ]$primerid)
ego <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))
ego_up <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names_up)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))
ego_down <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names_down)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))
#cnetplot(ego, foldChange = 2^logfc, colorEdge = TRUE, node_label = "gene")
dotplot(ego)+theme(axis.text.y = element_text(size = 10))
dotplot(ego_up)+theme(axis.text.y = element_text(size = 10))
dotplot(ego_down)+theme(axis.text.y = element_text(size = 10))



# Visualize enrichment of PC genes found in those regions close to Novel 
all_lnc_2mb_novel <- all_lnc_2mb[all_lnc_2mb$type == "Novel",]
all_lnc_2mb_names <- ref[ref$gene_id %in% all_lnc_2mb_novel$second, ]$gene_name
ego <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))
#cnetplot(ego, foldChange = 2^logfc, colorEdge = TRUE, node_label = "gene")
dotplot(ego)+theme(axis.text.y = element_text(size = 10))


# Visualize enrichment of PC genes found in those regions close to Annotated 
all_lnc_2mb_novel <- all_lnc_2mb[all_lnc_2mb$type == "Annotated",]
all_lnc_2mb_names <- ref[ref$gene_id %in% all_lnc_2mb_novel$second, ]$gene_name
logfc <- stats_complete[stats_complete$primerid %in% unlist(unique(all_lnc_2mb_names)),]$coef
ego <- clusterProfiler::enrichGO(gene = unlist(unique(all_lnc_2mb_names)),OrgDb =org.Hs.eg.db, keyType = "SYMBOL",ont = "BP",universe =unique(rownames(immune.combined)))
#cnetplot(ego, foldChange = 2^logfc, colorEdge = TRUE, node_label = "gene")
dotplot(ego)+theme(axis.text.y = element_text(size = 10))

```

```{r colocation}

all_lnc_2mb <- all_lnc_2mb[!is.na(all_lnc_2mb$distances),]
all_lnc_2mb$bin <- (cut(all_lnc_2mb$distances, breaks = c(-1,20000, 100000,500000,1000000, 2000000,10000000, Inf), labels = c("20kb", "100kb","500kb","1Mb","2Mb", "10Mb",">10Mb" )))
all_lnc_2mb <- all_lnc_2mb[!is.na(all_lnc_2mb$bin), ]
ggplot(all_lnc_2mb %>% group_by(gene, type, bin) %>% tally(), aes(y= n, col = type, fill = type, x = bin))+geom_boxplot(alpha = 0.5)+theme_paper+ scale_fill_brewer(palette = "Paired")+scale_color_brewer(palette = "Paired")+xlab("Distance from DE lnc allowed")+ylab("#DE mRNAs found")

```

# -------------------------------
#  Pearson Correlations
# -------------------------------

```{r correlation}

cor_dir <- "/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/06_correlation/01_pearson/"
files <- iterate_files(cor_dir, "pearson_*")
get_celltype <- function(x){gsub(" ", "", strsplit(strsplit(basename(x), "_")[[1]][3], '.', fixed = T ))}
get_celltype <- function(x){strsplit(gsub(" ", "",strsplit(basename(x), "_")[[1]][3]), '.', fixed = T )[[1]][1]}


# Read Rds files resulting from DE Analysis and assign to correct variable name 
correlations_celltype <- unlist(lapply(files, function(name) { 
                                varname=paste("cor", get_celltype(name),  sep="_") ;
                                assign(varname, readRDS(name), envir = .GlobalEnv);
                                print(varname);
                                
                                # Assign right comparison depending on variable name
                                eval(parse(text=paste0(varname, "<- do.call(rbind,", varname,"[!is.na(",varname,")])")), envir = .GlobalEnv);
                                eval(parse(text=paste0(varname,"$celltype <- \"", get_celltype(name), "\"")), envir = .GlobalEnv);
                                
                                return(varname)
                              }))


cor_complete <- rbind(cor_B, cor_T, cor_Monocyte, cor_Neutrophil, cor_all) 
```

# Check the directionality of co-regulation 
```{r correlation}
# For how many lncrnas do i find a correlation 
length(unique(cor_complete[cor_complete$g1 %in% all_lncrnas, ]$g1))

cor_significant <- cor_complete[cor_complete$pval < 0.05 & !is.na(cor_complete$rho),]


ggplot(cor_significant, aes( x=rho, col = celltype))+geom_density(alpha = 0.2, size= 0.8)+theme_paper+xlab("Correlation coefficient")+scale_color_brewer(palette="Dark2")+ggtitle("Significant correlation coefficients")

cor_significant$direction <- ifelse(cor_significant$rho > 0 , "Positive", "Negative")

ggplot(cor_significant, aes( x=direction, col = celltype, fill = celltype))+geom_bar(alpha = 0.9, size= 0.8, position = position_dodge())+theme_paper+scale_color_brewer(palette="Dark2")+scale_fill_brewer(palette="Dark2")+ggtitle("Directionality of coregulation with DE Protein Coding Genes")+xlab("")

```


# Pair with the highest correlation coefficient 
```{r correlation}

df_sig <- cor_significant[rev(order(cor_significant$rho)),]

index <- 3
plot_pair <- function(index){
  pos_example_gene <- df_sig[index,]$g1
  pos_example_gene_2 <- df_sig[index,]$g2
  cor_complete[cor_complete$g1==pos_example_gene & cor_complete$g2==pos_example_gene_2 ,]
  
  plot_seperate_features(immune.combined, gene  = c(pos_example_gene), ident = "group")
  plot_seperate_features(immune.combined, gene  = c(pos_example_gene_2), ident = "group", colors =c('#DCD9DE','navy'))
}

lapply(1:10, function(x) plot_pair(x))


cor_significant[(order(cor_significant$rho)),]

pos_example_gene <- "ENSMMUG00000056793-unknown"
pos_example_gene_2 <- "GLUL"
cor_complete[cor_complete$g1==pos_example_gene & cor_complete$g2==pos_example_gene_2 ,]

plot_seperate_features(immune.combined, gene  = c(pos_example_gene), ident = "group")
plot_seperate_features(immune.combined, gene  = c(pos_example_gene_2), ident = "group", colors =c('#DCD9DE','navy'))

```

# Relationship correlation - colocation 

```{r correlation}
get_cor <- function(cor_complete, gene, second){
  a <- cor_complete[cor_complete$g1 == gene & cor_complete$g2 == second, ]$rho
  b <- cor_complete[cor_complete$g2 == gene & cor_complete$g1 == second, ]$rho
  print(paste(a,"--", b))
}

distances <- data.frame(g1=all_lnc_2mb$gene, g2=all_lnc_2mb$second, distance = all_lnc_2mb$distances, bin = all_lnc_2mb$bin, stringsAsFactors = F)

get_col_cor_df <- function(distances, cor_all){
  cor_all_prep <- cor_all
  cor_all_prep$g1 <-  unlist(lapply(cor_all_prep$g1, function(x) gsub("-unknown", "", x)))
  cor_all_prep$g2 <-  unlist(lapply(cor_all_prep$g2, function(x) gsub("-unknown", "", x)))
  correlation_colocation_all <- merge(cor_all_prep,distances, by=c("g1","g2"))
  return(correlation_colocation_all)
}

correlation_colocation_all <- get_col_cor_df(distances, cor_all)
correlation_colocation_Monocyte <- get_col_cor_df(distances, cor_Monocyte)
correlation_colocation_B <- get_col_cor_df(distances, cor_B)
correlation_colocation_T <- get_col_cor_df(distances, cor_T)
correlation_colocation_neutrophil <- get_col_cor_df(distances, cor_Neutrophil)
df <- rbind(correlation_colocation_all, correlation_colocation_T, correlation_colocation_B, correlation_colocation_Monocyte, correlation_colocation_neutrophil)


length(unique(correlation_colocation_all$g1))

ggplot(df[df$pval < 0.05, ], aes(x = bin, y =abs(rho), col = celltype, fill = celltype))+geom_boxplot(alpha = 0.5)+theme_paper+scale_fill_brewer(palette="Dark2")+scale_color_brewer(palette="Dark2")


ggplot(df[df$pval < 0.05 & df$celltype == "all", ], aes(x = bin, y =abs(rho), col = celltype, fill = celltype))+geom_boxplot(alpha = 0.5)+theme_paper+scale_fill_brewer(palette="Dark2")+scale_color_brewer(palette="Dark2")


ggplot(df[df$pval < 0.05, ], aes(x = bin, y =pval, col = celltype, fill = celltype))+geom_boxplot(alpha = 0.5)+theme_paper+scale_fill_brewer(palette="Dark2")+scale_color_brewer(palette="Dark2")
```

# Prepare data for cytoscape mapping gene type 
```{r correlation}
df <-data.frame(name = all_lncrnas)
df$type <- "lnc"

df_2 <- data.frame(name = setdiff(annotated_mrnas, all_lncrnas))
df_2$type <- "pc"

write.table(rbind(df, df_2), "/home/luisas/Desktop/types.csv", row.names = FALSE, col.names = FALSE, quote = F, sep=",")
df
rownames(weightMat)
```



```{r correlation}
library(GENIE3)
library(igraph)
library(RCy3)
library(Rgraphviz)

weightMat <- readRDS("/home/luisas/Desktop/cluster/data/02_scRNA-Seq_PBMCs/01_scRNA-Seq_inVivo_rhemac10/05_RObjects/09_graph/genie3_res.rds")
linkList <- getLinkList(weightMat, reportMax = 200)

edge_listsi <- linkList
Gsi <- graph.data.frame(edge_listsi,directed = F)
Asi <- get.adjacency(Gsi,sparse = F,attr = "weight",type = "both")

g_arasi <- graph.adjacency(Asi,mode = "undirected",weighted = T)
g.cyto <- igraph.to.graphNEL(g_arasi)

cw = createNetworkFromGraph("net", graph=g.cyto)


write.csv(t(exprMatr), "/home/luisas/Desktop/cluster/data/matrix.csv", row.names = FALSE)

t_names <- colnames(t(exprMatr))[1:10]
write.table((t_names), "/home/luisas/Desktop/cluster/data/tnames.csv", row.names = FALSE, col.names = FALSE, quote = F)

as.matrix(weightMat)


```



